<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />



<meta name="date" content="2024-11-08" />

<title>Bioconductor annotation packages</title>

<script src="site_libs/header-attrs-2.29/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.13.2/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-6.4.2/css/all.min.css" rel="stylesheet" />
<link href="site_libs/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet" />

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/main/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">My blog</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/davetang/muse">
    <span class="fa fa-github"></span>
     
    Source code
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Bioconductor annotation packages</h1>
<h4 class="date">2024-11-08</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span>
workflowr <span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2024-11-08
</p>
<p>
<strong>Checks:</strong> <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 7
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong> <code>muse/</code> <span
class="glyphicon glyphicon-question-sign" aria-hidden="true"
title="This is the local directory in which the code in this file was executed.">
</span>
</p>
<p>
This reproducible <a href="https://rmarkdown.rstudio.com">R Markdown</a>
analysis was created with <a
  href="https://github.com/workflowr/workflowr">workflowr</a> (version
1.7.1). The <em>Checks</em> tab describes the reproducibility checks
that were applied when the results were created. The <em>Past
versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date
</a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git
repository, you know the exact version of the code that produced these
results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the
global environment can affect the analysis in your R Markdown file in
unknown ways. For reproduciblity it’s best to always run the code in an
empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20200712code">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Seed:</strong>
<code>set.seed(20200712)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20200712code"
class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20200712)</code> was run prior to running
the code in the R Markdown file. Setting a seed ensures that any results
that rely on randomness, e.g. subsampling or permutations, are
reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Session information:</strong>
recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package
versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be
confident that you successfully produced the results during this
run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr
project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomdavetangmusetree2993db0bd2266a77096c4bd4b83e2a31d1560c87targetblank2993db0a">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Repository version:</strong>
<a href="https://github.com/davetang/muse/tree/2993db0bd2266a77096c4bd4b83e2a31d1560c87" target="_blank">2993db0</a>
</a>
</p>
</div>
<div
id="strongRepositoryversionstrongahrefhttpsgithubcomdavetangmusetree2993db0bd2266a77096c4bd4b83e2a31d1560c87targetblank2993db0a"
class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development
and connecting the code version to the results is critical for
reproducibility.
</p>
<p>
The results in this page were generated with repository version
<a href="https://github.com/davetang/muse/tree/2993db0bd2266a77096c4bd4b83e2a31d1560c87" target="_blank">2993db0</a>.
See the <em>Past versions</em> tab to see a history of the changes made
to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for
the analysis have been committed to Git prior to generating the results
(you can use <code>wflow_publish</code> or
<code>wflow_git_commit</code>). workflowr only checks the R Markdown
file, but you know if there are other scripts or data files that it
depends on. Below is the status of the Git repository when the results
were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .Rhistory
    Ignored:    .Rproj.user/
    Ignored:    data/pbmc3k.csv
    Ignored:    data/pbmc3k.csv.gz
    Ignored:    data/pbmc3k/
    Ignored:    r_packages_4.4.0/
    Ignored:    r_packages_4.4.1/

Untracked files:
    Untracked:  Homo_sapiens.GRCh38.113.gtf.gz

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not
included in this status report because it is ok for generated content to
have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">
<p>
These are the previous versions of the repository in which changes were
made to the R Markdown (<code>analysis/annotation_dbi.Rmd</code>) and
HTML (<code>docs/annotation_dbi.html</code>) files. If you’ve configured
a remote Git repository (see <code>?wflow_git_remote</code>), click on
the hyperlinks in the table below to view the files as they were in that
past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/davetang/muse/blob/2993db0bd2266a77096c4bd4b83e2a31d1560c87/analysis/annotation_dbi.Rmd" target="_blank">2993db0</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-11-08
</td>
<td>
Mouse reference
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/davetang/muse/c8bf5517aaeacc4678a673e8b989d6745ebe52d3/docs/annotation_dbi.html" target="_blank">c8bf551</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-11-08
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/davetang/muse/blob/f5eef9941496db6e7f3d028ae979d75d5222307b/analysis/annotation_dbi.Rmd" target="_blank">f5eef99</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-11-08
</td>
<td>
Output info from org.Hs.eg.db
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/davetang/muse/def5ab0e21099a06f653428581319601be1d65c5/docs/annotation_dbi.html" target="_blank">def5ab0</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-10-23
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/davetang/muse/blob/58216aecf17a4080d377fa5748fb837eb700530a/analysis/annotation_dbi.Rmd" target="_blank">58216ae</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-10-23
</td>
<td>
Entrez Gene IDs are species specific
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/davetang/muse/4ee2a0f17fa24ab59da9d467d85bbfd5157b5cb0/docs/annotation_dbi.html" target="_blank">4ee2a0f</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-10-23
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/davetang/muse/blob/b3485f31832823d7392fdbe2d26dfa5825cc084d/analysis/annotation_dbi.Rmd" target="_blank">b3485f3</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-10-23
</td>
<td>
reactome.db
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/davetang/muse/de5bf16b6fce3a65ad1497282f4df0d9bfb80bdd/docs/annotation_dbi.html" target="_blank">de5bf16</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-10-23
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/davetang/muse/blob/250626520e1f74381d11320378fd180318793492/analysis/annotation_dbi.Rmd" target="_blank">2506265</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-10-23
</td>
<td>
Using AnnotationDbi
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<p>The <a
href="https://www.bioconductor.org/packages/release/data/annotation/">Bioconductor
annotation packages</a> are an extensive collection of annotations. For
this post I simply illustrate the basics of probing these annotation
packages. For the first example I will use the <a
href="https://bioconductor.org/packages/release/data/annotation/html/org.Hs.eg.db.html">org.Hs.eg.db</a>
package, which provides genome wide annotations for the human
genome.</p>
<div id="installation" class="section level2">
<h2>Installation</h2>
<p>To begin, install the {org.Hs.eg.db} package.</p>
<pre class="r"><code>if (!require(&quot;BiocManager&quot;, quietly = TRUE))
    install.packages(&quot;BiocManager&quot;)

BiocManager::install(&quot;org.Hs.eg.db&quot;)</code></pre>
</div>
<div id="package" class="section level2">
<h2>Package</h2>
<p>Load package.</p>
<pre class="r"><code>packageVersion(&quot;org.Hs.eg.db&quot;)</code></pre>
<pre><code>[1] &#39;3.20.0&#39;</code></pre>
<pre class="r"><code>suppressPackageStartupMessages(library(org.Hs.eg.db))
class(org.Hs.eg.db)</code></pre>
<pre><code>[1] &quot;OrgDb&quot;
attr(,&quot;package&quot;)
[1] &quot;AnnotationDbi&quot;</code></pre>
<p>{org.Hs.eg.db} info.</p>
<pre class="r"><code>org.Hs.eg.db</code></pre>
<pre><code>OrgDb object:
| DBSCHEMAVERSION: 2.1
| Db type: OrgDb
| Supporting package: AnnotationDbi
| DBSCHEMA: HUMAN_DB
| ORGANISM: Homo sapiens
| SPECIES: Human
| EGSOURCEDATE: 2024-Sep20
| EGSOURCENAME: Entrez Gene
| EGSOURCEURL: ftp://ftp.ncbi.nlm.nih.gov/gene/DATA
| CENTRALID: EG
| TAXID: 9606
| GOSOURCENAME: 
| GOSOURCEURL: 
| GOSOURCEDATE: 
| GOEGSOURCEDATE: 2024-Sep20
| GOEGSOURCENAME: Entrez Gene
| GOEGSOURCEURL: ftp://ftp.ncbi.nlm.nih.gov/gene/DATA
| KEGGSOURCENAME: KEGG GENOME
| KEGGSOURCEURL: ftp://ftp.genome.jp/pub/kegg/genomes
| KEGGSOURCEDATE: 2011-Mar15
| GPSOURCENAME: UCSC Genome Bioinformatics (Homo sapiens)
| GPSOURCEURL: ftp://hgdownload.cse.ucsc.edu/goldenPath/hg38/database
| GPSOURCEDATE: 2024-Sep22
| ENSOURCEDATE: 2024-May14
| ENSOURCENAME: Ensembl
| ENSOURCEURL: ftp://ftp.ensembl.org/pub/current_fasta
| UPSOURCENAME: Uniprot
| UPSOURCEURL: http://www.UniProt.org/
| UPSOURCEDATE: Mon Sep 23 15:46:45 2024</code></pre>
<pre><code>
Please see: help(&#39;select&#39;) for usage information</code></pre>
</div>
<div id="getting-started" class="section level2">
<h2>Getting started</h2>
<p>We can query the package by using the
<code>AnnotationDbi::select()</code> function; to find out what we can
select and return we can use the <code>keys()</code>,
<code>columns()</code>, and <code>keytypes()</code> functions.</p>
<pre class="r"><code>head(keys(org.Hs.eg.db))</code></pre>
<pre><code>[1] &quot;1&quot;  &quot;2&quot;  &quot;3&quot;  &quot;9&quot;  &quot;10&quot; &quot;11&quot;</code></pre>
<pre class="r"><code>#store the first six keys
my_keys &lt;- head(keys(org.Hs.eg.db))
 
keytypes(org.Hs.eg.db)</code></pre>
<pre><code> [1] &quot;ACCNUM&quot;       &quot;ALIAS&quot;        &quot;ENSEMBL&quot;      &quot;ENSEMBLPROT&quot;  &quot;ENSEMBLTRANS&quot;
 [6] &quot;ENTREZID&quot;     &quot;ENZYME&quot;       &quot;EVIDENCE&quot;     &quot;EVIDENCEALL&quot;  &quot;GENENAME&quot;    
[11] &quot;GENETYPE&quot;     &quot;GO&quot;           &quot;GOALL&quot;        &quot;IPI&quot;          &quot;MAP&quot;         
[16] &quot;OMIM&quot;         &quot;ONTOLOGY&quot;     &quot;ONTOLOGYALL&quot;  &quot;PATH&quot;         &quot;PFAM&quot;        
[21] &quot;PMID&quot;         &quot;PROSITE&quot;      &quot;REFSEQ&quot;       &quot;SYMBOL&quot;       &quot;UCSCKG&quot;      
[26] &quot;UNIPROT&quot;     </code></pre>
<pre class="r"><code>columns(org.Hs.eg.db)</code></pre>
<pre><code> [1] &quot;ACCNUM&quot;       &quot;ALIAS&quot;        &quot;ENSEMBL&quot;      &quot;ENSEMBLPROT&quot;  &quot;ENSEMBLTRANS&quot;
 [6] &quot;ENTREZID&quot;     &quot;ENZYME&quot;       &quot;EVIDENCE&quot;     &quot;EVIDENCEALL&quot;  &quot;GENENAME&quot;    
[11] &quot;GENETYPE&quot;     &quot;GO&quot;           &quot;GOALL&quot;        &quot;IPI&quot;          &quot;MAP&quot;         
[16] &quot;OMIM&quot;         &quot;ONTOLOGY&quot;     &quot;ONTOLOGYALL&quot;  &quot;PATH&quot;         &quot;PFAM&quot;        
[21] &quot;PMID&quot;         &quot;PROSITE&quot;      &quot;REFSEQ&quot;       &quot;SYMBOL&quot;       &quot;UCSCKG&quot;      
[26] &quot;UNIPROT&quot;     </code></pre>
<pre class="r"><code>AnnotationDbi::select(
  org.Hs.eg.db,
  keys = my_keys,
  columns=c(&quot;ENTREZID&quot;,&quot;SYMBOL&quot;,&quot;GENENAME&quot;),
  keytype=&quot;ENTREZID&quot;
)</code></pre>
<pre><code>&#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<pre><code>  ENTREZID SYMBOL                           GENENAME
1        1   A1BG             alpha-1-B glycoprotein
2        2    A2M              alpha-2-macroglobulin
3        3  A2MP1 alpha-2-macroglobulin pseudogene 1
4        9   NAT1              N-acetyltransferase 1
5       10   NAT2              N-acetyltransferase 2
6       11   NATP     N-acetyltransferase pseudogene</code></pre>
<p>Within the annotation package are many objects, which we can also
probe.</p>
<pre class="r"><code>ls(&quot;package:org.Hs.eg.db&quot;)</code></pre>
<pre><code> [1] &quot;org.Hs.eg&quot;                &quot;org.Hs.eg_dbconn&quot;        
 [3] &quot;org.Hs.eg_dbfile&quot;         &quot;org.Hs.eg_dbInfo&quot;        
 [5] &quot;org.Hs.eg_dbschema&quot;       &quot;org.Hs.eg.db&quot;            
 [7] &quot;org.Hs.egACCNUM&quot;          &quot;org.Hs.egACCNUM2EG&quot;      
 [9] &quot;org.Hs.egALIAS2EG&quot;        &quot;org.Hs.egCHR&quot;            
[11] &quot;org.Hs.egCHRLENGTHS&quot;      &quot;org.Hs.egCHRLOC&quot;         
[13] &quot;org.Hs.egCHRLOCEND&quot;       &quot;org.Hs.egENSEMBL&quot;        
[15] &quot;org.Hs.egENSEMBL2EG&quot;      &quot;org.Hs.egENSEMBLPROT&quot;    
[17] &quot;org.Hs.egENSEMBLPROT2EG&quot;  &quot;org.Hs.egENSEMBLTRANS&quot;   
[19] &quot;org.Hs.egENSEMBLTRANS2EG&quot; &quot;org.Hs.egENZYME&quot;         
[21] &quot;org.Hs.egENZYME2EG&quot;       &quot;org.Hs.egGENENAME&quot;       
[23] &quot;org.Hs.egGENETYPE&quot;        &quot;org.Hs.egGO&quot;             
[25] &quot;org.Hs.egGO2ALLEGS&quot;       &quot;org.Hs.egGO2EG&quot;          
[27] &quot;org.Hs.egMAP&quot;             &quot;org.Hs.egMAP2EG&quot;         
[29] &quot;org.Hs.egMAPCOUNTS&quot;       &quot;org.Hs.egOMIM&quot;           
[31] &quot;org.Hs.egOMIM2EG&quot;         &quot;org.Hs.egORGANISM&quot;       
[33] &quot;org.Hs.egPATH&quot;            &quot;org.Hs.egPATH2EG&quot;        
[35] &quot;org.Hs.egPFAM&quot;            &quot;org.Hs.egPMID&quot;           
[37] &quot;org.Hs.egPMID2EG&quot;         &quot;org.Hs.egPROSITE&quot;        
[39] &quot;org.Hs.egREFSEQ&quot;          &quot;org.Hs.egREFSEQ2EG&quot;      
[41] &quot;org.Hs.egSYMBOL&quot;          &quot;org.Hs.egSYMBOL2EG&quot;      
[43] &quot;org.Hs.egUCSCKG&quot;          &quot;org.Hs.egUNIPROT&quot;        </code></pre>
<p>Ensembl Gene IDs to Entrez Gene IDs.</p>
<pre class="r"><code>class(org.Hs.egENSEMBL2EG)</code></pre>
<pre><code>[1] &quot;AnnDbBimap&quot;
attr(,&quot;package&quot;)
[1] &quot;AnnotationDbi&quot;</code></pre>
<pre class="r"><code>head(keys(org.Hs.egENSEMBL2EG))</code></pre>
<pre><code>[1] &quot;ENSG00000121410&quot; &quot;ENSG00000175899&quot; &quot;ENSG00000291190&quot; &quot;ENSG00000171428&quot;
[5] &quot;ENSG00000156006&quot; &quot;ENSG00000196136&quot;</code></pre>
<p>Create a lookup table.</p>
<pre class="r"><code>my_ensg_keys &lt;- head(keys(org.Hs.egENSEMBL2EG))
 
AnnotationDbi::select(
  org.Hs.eg.db,
  keys = my_ensg_keys,
  columns=c(&quot;ENSEMBL&quot;,&quot;ENTREZID&quot;,&quot;SYMBOL&quot;,&quot;GENENAME&quot;),
  keytype=&quot;ENSEMBL&quot;
)</code></pre>
<pre><code>&#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<pre><code>          ENSEMBL ENTREZID   SYMBOL                           GENENAME
1 ENSG00000121410        1     A1BG             alpha-1-B glycoprotein
2 ENSG00000175899        2      A2M              alpha-2-macroglobulin
3 ENSG00000291190        3    A2MP1 alpha-2-macroglobulin pseudogene 1
4 ENSG00000171428        9     NAT1              N-acetyltransferase 1
5 ENSG00000156006       10     NAT2              N-acetyltransferase 2
6 ENSG00000196136       12 SERPINA3           serpin family A member 3</code></pre>
<p>Using the annotation package {org.Hs.eg.db} we can easily convert
different gene identifiers, obtain their gene symbols, and descriptions
(as well as all other keytypes).</p>
</div>
<div id="mouse" class="section level2">
<h2>Mouse</h2>
<p>Install {org.Mm.eg.db}.</p>
<pre class="r"><code>BiocManager::install(&quot;org.Mm.eg.db&quot;)</code></pre>
<p>Package info.</p>
<pre class="r"><code>packageVersion(&quot;org.Mm.eg.db&quot;)</code></pre>
<pre><code>[1] &#39;3.20.0&#39;</code></pre>
<pre class="r"><code>suppressPackageStartupMessages(library(org.Mm.eg.db))
class(org.Mm.eg.db)</code></pre>
<pre><code>[1] &quot;OrgDb&quot;
attr(,&quot;package&quot;)
[1] &quot;AnnotationDbi&quot;</code></pre>
<p>{org.Mm.eg.db} info.</p>
<pre class="r"><code>org.Mm.eg.db</code></pre>
<pre><code>OrgDb object:
| DBSCHEMAVERSION: 2.1
| Db type: OrgDb
| Supporting package: AnnotationDbi
| DBSCHEMA: MOUSE_DB
| ORGANISM: Mus musculus
| SPECIES: Mouse
| EGSOURCEDATE: 2024-Sep20
| EGSOURCENAME: Entrez Gene
| EGSOURCEURL: ftp://ftp.ncbi.nlm.nih.gov/gene/DATA
| CENTRALID: EG
| TAXID: 10090
| GOSOURCENAME: 
| GOSOURCEURL: 
| GOSOURCEDATE: 
| GOEGSOURCEDATE: 2024-Sep20
| GOEGSOURCENAME: Entrez Gene
| GOEGSOURCEURL: ftp://ftp.ncbi.nlm.nih.gov/gene/DATA
| KEGGSOURCENAME: KEGG GENOME
| KEGGSOURCEURL: ftp://ftp.genome.jp/pub/kegg/genomes
| KEGGSOURCEDATE: 2011-Mar15
| GPSOURCENAME: UCSC Genome Bioinformatics (Mus musculus)
| GPSOURCEURL: ftp://hgdownload.cse.ucsc.edu/goldenPath/mm39/database
| GPSOURCEDATE: 2024-Sep22
| ENSOURCEDATE: 2024-May14
| ENSOURCENAME: Ensembl
| ENSOURCEURL: ftp://ftp.ensembl.org/pub/current_fasta
| UPSOURCENAME: Uniprot
| UPSOURCEURL: http://www.UniProt.org/
| UPSOURCEDATE: Mon Sep 23 16:04:35 2024</code></pre>
<pre><code>
Please see: help(&#39;select&#39;) for usage information</code></pre>
<p>Mouse query.</p>
<pre class="r"><code>AnnotationDbi::select(
  org.Mm.eg.db,
  keys = c(&#39;ENSMUSG00000121796&#39;, &#39;ENSMUSG00000121918&#39;, &#39;ENSMUSG00000121583&#39;, &#39;ENSMUSG00000121584&#39;),
  columns=c(&quot;ENSEMBL&quot;,&quot;ENTREZID&quot;,&quot;SYMBOL&quot;,&quot;GENENAME&quot;),
  keytype=&quot;ENSEMBL&quot;
)</code></pre>
<pre><code>&#39;select()&#39; returned 1:many mapping between keys and columns</code></pre>
<pre><code>             ENSEMBL  ENTREZID  SYMBOL                           GENENAME
1 ENSMUSG00000121796      &lt;NA&gt;    &lt;NA&gt;                               &lt;NA&gt;
2 ENSMUSG00000121918      &lt;NA&gt;    &lt;NA&gt;                               &lt;NA&gt;
3 ENSMUSG00000121583     54151 Zftraf1 zinc finger TRAF type containing 1
4 ENSMUSG00000121583 125488414 Gm57852              predicted gene, 57852
5 ENSMUSG00000121584 125480935 Tmem276          transmembrane protein 276</code></pre>
</div>
<div id="reactome.db" class="section level2">
<h2>reactome.db</h2>
<p><a
href="https://bioconductor.org/packages/release/data/annotation/html/reactome.db.html">reactome.db</a>
is a set of annotation maps for Reactome, assembled using data from
Reactome. This package has been created by a third-party developer, and
is not affiliated with the Reactome team.</p>
<p>To begin, install the {reactome.db} package.</p>
<pre class="r"><code>BiocManager::install(&quot;reactome.db&quot;)</code></pre>
<p>Load the package.</p>
<pre class="r"><code>packageVersion(&#39;reactome.db&#39;)</code></pre>
<pre><code>[1] &#39;1.89.0&#39;</code></pre>
<pre class="r"><code>suppressPackageStartupMessages(library(reactome.db))
class(reactome.db)</code></pre>
<pre><code>[1] &quot;ReactomeDb&quot;
attr(,&quot;package&quot;)
[1] &quot;AnnotationDbi&quot;</code></pre>
<p>Demo.</p>
<pre class="r"><code>keytypes(reactome.db)</code></pre>
<pre><code>[1] &quot;ENTREZID&quot;   &quot;GO&quot;         &quot;PATHID&quot;     &quot;PATHNAME&quot;   &quot;REACTOMEID&quot;</code></pre>
<pre class="r"><code>columns(reactome.db)</code></pre>
<pre><code>[1] &quot;ENTREZID&quot;   &quot;GO&quot;         &quot;PATHID&quot;     &quot;PATHNAME&quot;   &quot;REACTOMEID&quot;</code></pre>
<pre class="r"><code>head(keys(reactome.db))</code></pre>
<pre><code>[1] &quot;1&quot;         &quot;10&quot;        &quot;100&quot;       &quot;1000&quot;      &quot;10000&quot;     &quot;100000160&quot;</code></pre>
<pre class="r"><code>my_keys &lt;- head(keys(reactome.db))

AnnotationDbi::select(
  reactome.db,
  keys = my_keys,
  columns=&quot;REACTOMEID&quot;,
  keytype=&quot;ENTREZID&quot;
) |&gt;
  head()</code></pre>
<pre><code>&#39;select()&#39; returned 1:many mapping between keys and columns</code></pre>
<pre><code>  ENTREZID    REACTOMEID
1        1  R-HSA-109582
2        1  R-HSA-114608
3        1  R-HSA-168249
4        1  R-HSA-168256
5        1 R-HSA-6798695
6        1   R-HSA-76002</code></pre>
<p>Note that Entrez Gene IDs are species-specific; TP53 in humans has
Entrez Gene ID 7157 and TP53 in mice has Entrez Gene ID 22059.</p>
<pre class="r"><code>AnnotationDbi::select(
  reactome.db,
  keys = &#39;7157&#39;,
  columns=c(&quot;REACTOMEID&quot;, &quot;PATHNAME&quot;),
  keytype=&quot;ENTREZID&quot;
) |&gt;
  head()</code></pre>
<pre><code>&#39;select()&#39; returned 1:many mapping between keys and columns</code></pre>
<pre><code>  ENTREZID    REACTOMEID
1     7157  R-HSA-109581
2     7157  R-HSA-109582
3     7157  R-HSA-109606
4     7157  R-HSA-111448
5     7157  R-HSA-114452
6     7157 R-HSA-1169410
                                                            PATHNAME
1                                            Homo sapiens: Apoptosis
2                                           Homo sapiens: Hemostasis
3                      Homo sapiens: Intrinsic Pathway for Apoptosis
4 Homo sapiens: Activation of NOXA and translocation to mitochondria
5                      Homo sapiens: Activation of BH3-only proteins
6          Homo sapiens: Antiviral mechanism by IFN-stimulated genes</code></pre>
<pre class="r"><code>AnnotationDbi::select(
  org.Hs.eg.db,
  keys = &#39;7157&#39;,
  columns=c(&quot;ENTREZID&quot;,&quot;SYMBOL&quot;,&quot;GENENAME&quot;),
  keytype=&quot;ENTREZID&quot;
)</code></pre>
<pre><code>&#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<pre><code>  ENTREZID SYMBOL          GENENAME
1     7157   TP53 tumor protein p53</code></pre>
<pre class="r"><code>AnnotationDbi::select(
  reactome.db,
  keys = &#39;22059&#39;,
  columns=c(&quot;REACTOMEID&quot;, &quot;PATHNAME&quot;),
  keytype=&quot;ENTREZID&quot;
) |&gt;
  head()</code></pre>
<pre><code>&#39;select()&#39; returned 1:many mapping between keys and columns</code></pre>
<pre><code>  ENTREZID    REACTOMEID
1    22059 R-MMU-1169410
2    22059 R-MMU-1280215
3    22059 R-MMU-1640170
4    22059  R-MMU-168256
5    22059  R-MMU-212436
6    22059 R-MMU-2262752
                                                   PATHNAME
1 Mus musculus: Antiviral mechanism by IFN-stimulated genes
2         Mus musculus: Cytokine Signaling in Immune system
3                                  Mus musculus: Cell Cycle
4                               Mus musculus: Immune System
5               Mus musculus: Generic Transcription Pathway
6                Mus musculus: Cellular responses to stress</code></pre>
</div>
<div id="go.db" class="section level2">
<h2>GO.db</h2>
<p>There are also other annotation packages, such as <a
href="https://bioconductor.org/packages/release/data/annotation/html/GO.db.html">GO.db</a>,
which contains a set of annotation maps describing the entire Gene
Ontology that we can probe in the same manner.</p>
<p>To begin, install the {GO.db} package.</p>
<pre class="r"><code>BiocManager::install(&quot;GO.db&quot;)</code></pre>
<p>Load the package.</p>
<pre class="r"><code>packageVersion(&#39;GO.db&#39;)</code></pre>
<pre><code>[1] &#39;3.20.0&#39;</code></pre>
<pre class="r"><code>suppressPackageStartupMessages(library(GO.db))
class(GO.db)</code></pre>
<pre><code>[1] &quot;GODb&quot;
attr(,&quot;package&quot;)
[1] &quot;AnnotationDbi&quot;</code></pre>
<p>Use the same functions as before on the <code>AnnotationDbi</code>
object.</p>
<pre class="r"><code>keytypes(GO.db)</code></pre>
<pre><code>[1] &quot;DEFINITION&quot; &quot;GOID&quot;       &quot;ONTOLOGY&quot;   &quot;TERM&quot;      </code></pre>
<pre class="r"><code>my_go_keys &lt;- head(keys(GO.db))
my_go_keys</code></pre>
<pre><code>[1] &quot;GO:0000001&quot; &quot;GO:0000002&quot; &quot;GO:0000006&quot; &quot;GO:0000007&quot; &quot;GO:0000009&quot;
[6] &quot;GO:0000010&quot;</code></pre>
<pre class="r"><code>AnnotationDbi::select(
  GO.db,
  keys = my_go_keys,
  columns=c(&quot;GOID&quot;, &quot;TERM&quot;, &quot;ONTOLOGY&quot;),
  keytype=&quot;GOID&quot;
)</code></pre>
<pre><code>&#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<pre><code>        GOID                                                     TERM ONTOLOGY
1 GO:0000001                                mitochondrion inheritance       BP
2 GO:0000002                         mitochondrial genome maintenance       BP
3 GO:0000006    high-affinity zinc transmembrane transporter activity       MF
4 GO:0000007 low-affinity zinc ion transmembrane transporter activity       MF
5 GO:0000009                   alpha-1,6-mannosyltransferase activity       MF
6 GO:0000010                heptaprenyl diphosphate synthase activity       MF</code></pre>
</div>
<div id="genomicfeatures" class="section level2">
<h2>GenomicFeatures</h2>
<p>For genome annotation purposes, one may be interested in the gene
models, which are provided in the <a
href="https://www.bioconductor.org/packages/release/data/annotation/html/TxDb.Hsapiens.UCSC.hg38.knownGene.html">TxDb.Hsapiens.UCSC.hg38.knownGene</a>
package.</p>
<pre class="r"><code>BiocManager::install(&quot;TxDb.Hsapiens.UCSC.hg38.knownGene&quot;)</code></pre>
<p>Load the package.</p>
<pre class="r"><code>packageVersion(&#39;TxDb.Hsapiens.UCSC.hg38.knownGene&#39;)</code></pre>
<pre><code>[1] &#39;3.20.0&#39;</code></pre>
<pre class="r"><code>suppressPackageStartupMessages(library(TxDb.Hsapiens.UCSC.hg38.knownGene))
class(TxDb.Hsapiens.UCSC.hg38.knownGene)</code></pre>
<pre><code>[1] &quot;TxDb&quot;
attr(,&quot;package&quot;)
[1] &quot;GenomicFeatures&quot;</code></pre>
<p>To see what columns can be returned use the <code>columns()</code>
function</p>
<pre class="r"><code>columns(TxDb.Hsapiens.UCSC.hg38.knownGene)</code></pre>
<pre><code> [1] &quot;CDSCHROM&quot;   &quot;CDSEND&quot;     &quot;CDSID&quot;      &quot;CDSNAME&quot;    &quot;CDSPHASE&quot;  
 [6] &quot;CDSSTART&quot;   &quot;CDSSTRAND&quot;  &quot;EXONCHROM&quot;  &quot;EXONEND&quot;    &quot;EXONID&quot;    
[11] &quot;EXONNAME&quot;   &quot;EXONRANK&quot;   &quot;EXONSTART&quot;  &quot;EXONSTRAND&quot; &quot;GENEID&quot;    
[16] &quot;TXCHROM&quot;    &quot;TXEND&quot;      &quot;TXID&quot;       &quot;TXNAME&quot;     &quot;TXSTART&quot;   
[21] &quot;TXSTRAND&quot;   &quot;TXTYPE&quot;    </code></pre>
<p>Selecting.</p>
<pre class="r"><code>AnnotationDbi::select(
  TxDb.Hsapiens.UCSC.hg38.knownGene,
  keys = head(keys(TxDb.Hsapiens.UCSC.hg38.knownGene), 2),
  columns=c(&#39;GENEID&#39;, &#39;TXCHROM&#39;, &#39;TXSTART&#39;, &#39;TXEND&#39;, &#39;TXID&#39;),
  keytype=&quot;GENEID&quot;
)</code></pre>
<pre><code>&#39;select()&#39; returned 1:many mapping between keys and columns</code></pre>
<pre><code>   GENEID   TXID TXCHROM  TXSTART    TXEND
1       1 230133   chr19 58345178 58347634
2       1 230134   chr19 58345183 58353492
3       1 230135   chr19 58346854 58356225
4       1 230136   chr19 58346858 58353491
5       1 230137   chr19 58346860 58347657
6       1 230138   chr19 58348466 58362751
7       1 230139   chr19 58350594 58353129
8       1 230140   chr19 58353021 58356083
9      10 104998    chr8 18386311 18388323
10     10 105000    chr8 18391282 18401218
11     10 105001    chr8 18391287 18400993</code></pre>
</div>
<div id="combining-databases" class="section level2">
<h2>Combining databases</h2>
<p>We can combine different annotation packages too; say I have two
genes LRRK2 and PINK1, and I want their genomic locations as well as
OMIM ids.</p>
<pre class="r"><code>my_genes &lt;- c(&quot;LRRK2&quot;,&quot;PINK1&quot;)
 
AnnotationDbi::select(
  org.Hs.eg.db,
  keys = my_genes,
  columns=c(&quot;ENTREZID&quot;, &quot;SYMBOL&quot;,&quot;OMIM&quot;),
  keytype=&quot;SYMBOL&quot;
) -&gt; a</code></pre>
<pre><code>&#39;select()&#39; returned 1:many mapping between keys and columns</code></pre>
<pre class="r"><code>a </code></pre>
<pre><code>  SYMBOL ENTREZID   OMIM
1  LRRK2   120892 607060
2  LRRK2   120892 609007
3  PINK1    65018 605909
4  PINK1    65018 608309</code></pre>
<pre class="r"><code># symbol and transcript location
AnnotationDbi::select(
  TxDb.Hsapiens.UCSC.hg38.knownGene,
  keys = a$ENTREZID,
  columns=c(&#39;GENEID&#39;, &#39;TXCHROM&#39;, &#39;TXSTART&#39;, &#39;TXEND&#39;, &#39;TXID&#39;),
  keytype=&quot;GENEID&quot;
) -&gt; b</code></pre>
<pre><code>&#39;select()&#39; returned many:many mapping between keys and columns</code></pre>
<pre class="r"><code>head(b)</code></pre>
<pre><code>  GENEID   TXID TXCHROM  TXSTART    TXEND
1 120892 149454   chr12 40196744 40283952
2 120892 149456   chr12 40224977 40242782
3 120892 149457   chr12 40224997 40367077
4 120892 149458   chr12 40224997 40369285
5 120892 149459   chr12 40225011 40304976
6 120892 149460   chr12 40225080 40369259</code></pre>
<pre class="r"><code># change GENEID column to ENTREZID for consistency and for merging
names(b) &lt;- c(&#39;ENTREZID&#39;, &#39;TXID&#39;, &#39;TXCHROM&#39;, &#39;TXSTART&#39;, &#39;TXEND&#39;)
c &lt;- merge(a, b, &#39;ENTREZID&#39;)
head(c)</code></pre>
<pre><code>  ENTREZID SYMBOL   OMIM   TXID TXCHROM  TXSTART    TXEND
1   120892  LRRK2 607060 149454   chr12 40196744 40283952
2   120892  LRRK2 607060 149456   chr12 40224977 40242782
3   120892  LRRK2 607060 149457   chr12 40224997 40367077
4   120892  LRRK2 607060 149458   chr12 40224997 40369285
5   120892  LRRK2 607060 149459   chr12 40225011 40304976
6   120892  LRRK2 607060 149460   chr12 40225080 40369259</code></pre>
</div>
<div id="conclusions" class="section level2">
<h2>Conclusions</h2>
<p>One important point to be aware of when using the Bioconductor
annotation packages is to appreciate the lag between updates to the
packages. From the <a
href="https://master.bioconductor.org/help/faq/#annotation-faq">FAQ</a>
on the Bioconductor website:</p>
<blockquote>
<p>Different sources take different approaches to managing annotations.
The annotation packages in Bioconductor are based on downloads obtained
shortly before each Bioconductor release, and so can lag by six months
(at the end of the release cycle) compared to on-line resources. The
advantage of this approach is that the annotations do not change
unexpectedly during development of an analysis, while the disadvantage
is that the resource is not quite up-to-date with current
understanding.</p>
</blockquote>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span>
Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 4.4.1 (2024-06-14)
Platform: x86_64-pc-linux-gnu
Running under: Ubuntu 22.04.5 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 
LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

time zone: Etc/UTC
tzcode source: system (glibc)

attached base packages:
[1] stats4    stats     graphics  grDevices utils     datasets  methods  
[8] base     

other attached packages:
 [1] TxDb.Hsapiens.UCSC.hg38.knownGene_3.20.0
 [2] GenomicFeatures_1.58.0                  
 [3] GenomicRanges_1.58.0                    
 [4] GenomeInfoDb_1.42.0                     
 [5] GO.db_3.20.0                            
 [6] reactome.db_1.89.0                      
 [7] org.Mm.eg.db_3.20.0                     
 [8] org.Hs.eg.db_3.20.0                     
 [9] AnnotationDbi_1.68.0                    
[10] IRanges_2.40.0                          
[11] S4Vectors_0.44.0                        
[12] Biobase_2.66.0                          
[13] BiocGenerics_0.52.0                     
[14] workflowr_1.7.1                         

loaded via a namespace (and not attached):
 [1] blob_1.2.4                  Biostrings_2.74.0          
 [3] bitops_1.0-9                fastmap_1.2.0              
 [5] RCurl_1.98-1.16             GenomicAlignments_1.42.0   
 [7] promises_1.3.0              XML_3.99-0.17              
 [9] digest_0.6.37               lifecycle_1.0.4            
[11] processx_3.8.4              KEGGREST_1.46.0            
[13] RSQLite_2.3.7               magrittr_2.0.3             
[15] compiler_4.4.1              rlang_1.1.4                
[17] sass_0.4.9                  tools_4.4.1                
[19] utf8_1.2.4                  yaml_2.3.10                
[21] rtracklayer_1.66.0          knitr_1.48                 
[23] S4Arrays_1.6.0              bit_4.5.0                  
[25] curl_6.0.0                  DelayedArray_0.32.0        
[27] abind_1.4-8                 BiocParallel_1.40.0        
[29] grid_4.4.1                  fansi_1.0.6                
[31] git2r_0.35.0                SummarizedExperiment_1.36.0
[33] cli_3.6.3                   rmarkdown_2.29             
[35] crayon_1.5.3                rstudioapi_0.17.1          
[37] httr_1.4.7                  rjson_0.2.23               
[39] DBI_1.2.3                   cachem_1.1.0               
[41] stringr_1.5.1               zlibbioc_1.52.0            
[43] parallel_4.4.1              XVector_0.46.0             
[45] restfulr_0.0.15             matrixStats_1.4.1          
[47] vctrs_0.6.5                 Matrix_1.7-0               
[49] jsonlite_1.8.9              callr_3.7.6                
[51] bit64_4.5.2                 jquerylib_0.1.4            
[53] glue_1.8.0                  codetools_0.2-20           
[55] ps_1.8.1                    stringi_1.8.4              
[57] later_1.3.2                 BiocIO_1.16.0              
[59] UCSC.utils_1.2.0            tibble_3.2.1               
[61] pillar_1.9.0                htmltools_0.5.8.1          
[63] GenomeInfoDbData_1.2.13     R6_2.5.1                   
[65] rprojroot_2.0.4             lattice_0.22-6             
[67] evaluate_1.0.1              png_0.1-8                  
[69] Rsamtools_2.22.0            memoise_2.0.1              
[71] httpuv_1.6.15               bslib_0.8.0                
[73] Rcpp_1.0.13-1               SparseArray_1.6.0          
[75] whisker_0.4.1               xfun_0.49                  
[77] fs_1.6.5                    MatrixGenerics_1.18.0      
[79] getPass_0.2-4               pkgconfig_2.0.3            </code></pre>
</div>
</div>
</div>
</div>
</div>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
https://docs.mathjax.org/en/latest/web/configuration.html. This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>




</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
