<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />



<meta name="date" content="2026-02-02" />

<title>ANOVA</title>

<script src="site_libs/header-attrs-2.29/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.13.2/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-6.5.2/css/all.min.css" rel="stylesheet" />
<link href="site_libs/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet" />

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/main/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">My blog</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/davetang/muse">
    <span class="fa fa-github"></span>
     
    Source code
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">ANOVA</h1>
<h4 class="date">2026-02-02</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span>
workflowr <span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2026-02-02
</p>
<p>
<strong>Checks:</strong> <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 7
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong> <code>muse/</code> <span
class="glyphicon glyphicon-question-sign" aria-hidden="true"
title="This is the local directory in which the code in this file was executed.">
</span>
</p>
<p>
This reproducible <a href="https://rmarkdown.rstudio.com">R Markdown</a>
analysis was created with <a
  href="https://github.com/workflowr/workflowr">workflowr</a> (version
1.7.1). The <em>Checks</em> tab describes the reproducibility checks
that were applied when the results were created. The <em>Past
versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date
</a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git
repository, you know the exact version of the code that produced these
results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the
global environment can affect the analysis in your R Markdown file in
unknown ways. For reproduciblity it’s best to always run the code in an
empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20200712code">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Seed:</strong>
<code>set.seed(20200712)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20200712code"
class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20200712)</code> was run prior to running
the code in the R Markdown file. Setting a seed ensures that any results
that rely on randomness, e.g. subsampling or permutations, are
reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Session information:</strong>
recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package
versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be
confident that you successfully produced the results during this
run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr
project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomdavetangmusetreee07f85f4a3911635a2b28a80adf5cf66cc575497targetblanke07f85fa">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Repository version:</strong>
<a href="https://github.com/davetang/muse/tree/e07f85f4a3911635a2b28a80adf5cf66cc575497" target="_blank">e07f85f</a>
</a>
</p>
</div>
<div
id="strongRepositoryversionstrongahrefhttpsgithubcomdavetangmusetreee07f85f4a3911635a2b28a80adf5cf66cc575497targetblanke07f85fa"
class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development
and connecting the code version to the results is critical for
reproducibility.
</p>
<p>
The results in this page were generated with repository version
<a href="https://github.com/davetang/muse/tree/e07f85f4a3911635a2b28a80adf5cf66cc575497" target="_blank">e07f85f</a>.
See the <em>Past versions</em> tab to see a history of the changes made
to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for
the analysis have been committed to Git prior to generating the results
(you can use <code>wflow_publish</code> or
<code>wflow_git_commit</code>). workflowr only checks the R Markdown
file, but you know if there are other scripts or data files that it
depends on. Below is the status of the Git repository when the results
were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .Rproj.user/
    Ignored:    data/1M_neurons_filtered_gene_bc_matrices_h5.h5
    Ignored:    data/293t/
    Ignored:    data/293t_3t3_filtered_gene_bc_matrices.tar.gz
    Ignored:    data/293t_filtered_gene_bc_matrices.tar.gz
    Ignored:    data/5k_Human_Donor1_PBMC_3p_gem-x_5k_Human_Donor1_PBMC_3p_gem-x_count_sample_filtered_feature_bc_matrix.h5
    Ignored:    data/5k_Human_Donor2_PBMC_3p_gem-x_5k_Human_Donor2_PBMC_3p_gem-x_count_sample_filtered_feature_bc_matrix.h5
    Ignored:    data/5k_Human_Donor3_PBMC_3p_gem-x_5k_Human_Donor3_PBMC_3p_gem-x_count_sample_filtered_feature_bc_matrix.h5
    Ignored:    data/5k_Human_Donor4_PBMC_3p_gem-x_5k_Human_Donor4_PBMC_3p_gem-x_count_sample_filtered_feature_bc_matrix.h5
    Ignored:    data/97516b79-8d08-46a6-b329-5d0a25b0be98.h5ad
    Ignored:    data/Parent_SC3v3_Human_Glioblastoma_filtered_feature_bc_matrix.tar.gz
    Ignored:    data/brain_counts/
    Ignored:    data/cl.obo
    Ignored:    data/cl.owl
    Ignored:    data/jurkat/
    Ignored:    data/jurkat:293t_50:50_filtered_gene_bc_matrices.tar.gz
    Ignored:    data/jurkat_293t/
    Ignored:    data/jurkat_filtered_gene_bc_matrices.tar.gz
    Ignored:    data/pbmc20k/
    Ignored:    data/pbmc20k_seurat/
    Ignored:    data/pbmc3k.csv
    Ignored:    data/pbmc3k.csv.gz
    Ignored:    data/pbmc3k.h5ad
    Ignored:    data/pbmc3k/
    Ignored:    data/pbmc3k_bpcells_mat/
    Ignored:    data/pbmc3k_export.mtx
    Ignored:    data/pbmc3k_matrix.mtx
    Ignored:    data/pbmc3k_seurat.rds
    Ignored:    data/pbmc4k_filtered_gene_bc_matrices.tar.gz
    Ignored:    data/pbmc_1k_v3_filtered_feature_bc_matrix.h5
    Ignored:    data/pbmc_1k_v3_raw_feature_bc_matrix.h5
    Ignored:    data/refdata-gex-GRCh38-2020-A.tar.gz
    Ignored:    data/seurat_1m_neuron.rds
    Ignored:    data/t_3k_filtered_gene_bc_matrices.tar.gz
    Ignored:    r_packages_4.4.1/
    Ignored:    r_packages_4.5.0/

Untracked files:
    Untracked:  .claude/
    Untracked:  CLAUDE.md
    Untracked:  analysis/bioc.Rmd
    Untracked:  analysis/bioc_scrnaseq.Rmd
    Untracked:  analysis/chick_weight.Rmd
    Untracked:  analysis/likelihood.Rmd
    Untracked:  bpcells_matrix/
    Untracked:  data/Caenorhabditis_elegans.WBcel235.113.gtf.gz
    Untracked:  data/GCF_043380555.1-RS_2024_12_gene_ontology.gaf.gz
    Untracked:  data/arab.rds
    Untracked:  data/astronomicalunit.csv
    Untracked:  data/femaleMiceWeights.csv
    Untracked:  data/lung_bcell.rds
    Untracked:  m3/
    Untracked:  women.json

Unstaged changes:
    Modified:   analysis/isoform_switch_analyzer.Rmd
    Modified:   analysis/linear_models.Rmd

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not
included in this status report because it is ok for generated content to
have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">
<p>
These are the previous versions of the repository in which changes were
made to the R Markdown (<code>analysis/anova.Rmd</code>) and HTML
(<code>docs/anova.html</code>) files. If you’ve configured a remote Git
repository (see <code>?wflow_git_remote</code>), click on the hyperlinks
in the table below to view the files as they were in that past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/davetang/muse/blob/e07f85f4a3911635a2b28a80adf5cf66cc575497/analysis/anova.Rmd" target="_blank">e07f85f</a>
</td>
<td>
Dave Tang
</td>
<td>
2026-02-02
</td>
<td>
Update with more explanations and examples
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/davetang/muse/78ef6833b847b51d9167889246b69aac34a883d9/docs/anova.html" target="_blank">78ef683</a>
</td>
<td>
Dave Tang
</td>
<td>
2025-04-02
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/davetang/muse/blob/2381c59c78c9de17558bfe172bfbab0a7ce666f9/analysis/anova.Rmd" target="_blank">2381c59</a>
</td>
<td>
Dave Tang
</td>
<td>
2025-04-02
</td>
<td>
ANOVA
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<div id="introduction" class="section level2">
<h2>Introduction</h2>
<p>ANOVA (Analysis of Variance) is a statistical test used to compare
the <strong>means</strong> of three or more groups to see if at least
one group is significantly different. If you have only two groups, a
<em>t</em>-test is usually better because it is simpler and more
powerful for two-group comparisons. However, the <em>t</em>-test can
only compare two groups at a time. ANOVA checks for overall differences
among all groups.</p>
<p>If gene expression is measured across different conditions (e.g.,
Control, Treatment A, Treatment B), ANOVA tests whether the
<strong>average expression levels</strong> differ significantly across
these conditions.</p>
<p>In scRNA-seq, ANOVA can be useful when analysing gene expression
differences across multiple conditions or cell types. If we have three
conditions (Control, A, and B) we can use ANOVA to check whether a
gene’s expression significantly changes across the conditions. Note that
the ANOVA test will indicate that there’s a difference, but not which
groups are different (a post-hoc test like Tukey’s HSD can be used).
Tukey’s HSD (Honestly Significant Difference) test can be used to
determine which specific groups are significantly different from each
other. It compares all possible group pairs and controls for multiple
testing.</p>
</div>
<div id="the-mathematics-behind-anova" class="section level2">
<h2>The Mathematics Behind ANOVA</h2>
<div id="what-is-sum-of-squares" class="section level3">
<h3>What is Sum of Squares?</h3>
<p><strong>Sum of Squares (SS)</strong> measures the total amount of
variation in data. It is calculated by summing the squared differences
between each observation and a reference point (usually a mean). We
square the differences so that positive and negative deviations don’t
cancel out.</p>
<p><span class="math display">\[SS = \sum(x_i - \bar{x})^2\]</span></p>
<p>The key insight of ANOVA is that we can <strong>partition</strong>
the total variation into meaningful components:</p>
<p><span class="math display">\[SS_{total} = SS_{between} +
SS_{within}\]</span></p>
<p>This partitioning allows us to ask: “How much of the total variation
is due to differences <em>between</em> groups versus variation
<em>within</em> groups?”</p>
</div>
<div id="understanding-between-group-and-within-group-variance"
class="section level3">
<h3>Understanding Between-group and Within-group Variance</h3>
<p><img src="figure/anova.Rmd/ss_visualisation-1.png" width="960" style="display: block; margin: auto;" /></p>
<div id="between-group-variance-ss_between" class="section level4">
<h4>Between-group Variance (<span
class="math inline">\(SS_{between}\)</span>)</h4>
<p><strong>Between-group variance</strong> measures how much the group
means differ from the overall (grand) mean. It answers: <em>“How spread
out are the group averages?”</em></p>
<ul>
<li>If all groups have similar means → <span
class="math inline">\(SS_{between}\)</span> is
<strong>small</strong></li>
<li>If group means are very different → <span
class="math inline">\(SS_{between}\)</span> is
<strong>large</strong></li>
</ul>
<p>This is the variation we are interested in - it represents the effect
of our experimental factor (e.g., treatment).</p>
</div>
<div id="within-group-variance-ss_within" class="section level4">
<h4>Within-group Variance (<span
class="math inline">\(SS_{within}\)</span>)</h4>
<p><strong>Within-group variance</strong> measures how much individual
observations vary around their own group mean. It answers: <em>“How
spread out are observations within each group?”</em></p>
<ul>
<li>This represents <strong>random variation</strong> or
<strong>measurement error</strong></li>
<li>Also called “residual” or “error” variance</li>
<li>It’s the baseline noise we compare our signal against</li>
</ul>
</div>
<div id="the-key-intuition" class="section level4">
<h4>The Key Intuition</h4>
<p>ANOVA compares these two sources of variation:</p>
<ul>
<li><strong>If <span class="math inline">\(SS_{between}\)</span>
&gt;&gt; <span class="math inline">\(SS_{within}\)</span></strong>: The
differences between groups are much larger than the random variation
within groups → groups are likely truly different</li>
<li><strong>If <span class="math inline">\(SS_{between}\)</span> ≈ <span
class="math inline">\(SS_{within}\)</span></strong>: The differences
between groups are similar to random variation → groups may not be truly
different</li>
</ul>
<p><img src="figure/anova.Rmd/intuition_example-1.png" width="960" style="display: block; margin: auto;" /></p>
</div>
</div>
<div id="the-formulas" class="section level3">
<h3>The Formulas</h3>
<p>Now that we understand the concepts, here are the formal
definitions:</p>
<p><strong>Total Sum of Squares</strong> - deviation of each observation
from the grand mean: <span class="math display">\[SS_{total} =
\sum_{i=1}^{k}\sum_{j=1}^{n_i}(x_{ij} - \bar{x}_{grand})^2\]</span></p>
<p><strong>Between-group Sum of Squares</strong> - deviation of group
means from the grand mean (weighted by group size): <span
class="math display">\[SS_{between} = \sum_{i=1}^{k}n_i(\bar{x}_i -
\bar{x}_{grand})^2\]</span></p>
<p><strong>Within-group Sum of Squares</strong> - deviation of
observations from their group mean: <span
class="math display">\[SS_{within} =
\sum_{i=1}^{k}\sum_{j=1}^{n_i}(x_{ij} - \bar{x}_i)^2\]</span></p>
<p>Where:</p>
<ul>
<li><span class="math inline">\(k\)</span> = number of groups</li>
<li><span class="math inline">\(n_i\)</span> = sample size of group
<span class="math inline">\(i\)</span></li>
<li><span class="math inline">\(x_{ij}\)</span> = observation <span
class="math inline">\(j\)</span> in group <span
class="math inline">\(i\)</span></li>
<li><span class="math inline">\(\bar{x}_i\)</span> = mean of group <span
class="math inline">\(i\)</span></li>
<li><span class="math inline">\(\bar{x}_{grand}\)</span> = grand mean
(mean of all observations)</li>
</ul>
</div>
<div id="the-f-statistic" class="section level3">
<h3>The F-statistic</h3>
<p>The F-statistic is calculated as the ratio of between-group variance
to within-group variance:</p>
<p><span class="math display">\[F = \frac{MS_{between}}{MS_{within}} =
\frac{SS_{between}/df_{between}}{SS_{within}/df_{within}}\]</span></p>
<p>Where:</p>
<ul>
<li><span class="math inline">\(df_{between} = k - 1\)</span> (number of
groups minus 1)</li>
<li><span class="math inline">\(df_{within} = N - k\)</span> (total
observations minus number of groups)</li>
<li><span class="math inline">\(MS\)</span> = Mean Square (sum of
squares divided by degrees of freedom)</li>
</ul>
<p>A large F-statistic indicates that the between-group variance is
larger than expected by chance, suggesting the group means differ.</p>
</div>
<div id="manual-calculation-example" class="section level3">
<h3>Manual Calculation Example</h3>
<p>Let’s calculate ANOVA by hand to understand the mechanics:</p>
<pre class="r"><code># Three groups with 5 observations each
group_a &lt;- c(23, 25, 27, 22, 24)
group_b &lt;- c(30, 32, 28, 31, 29)
group_c &lt;- c(35, 38, 36, 34, 37)

# Combine data
all_data &lt;- c(group_a, group_b, group_c)
groups &lt;- factor(rep(c(&quot;A&quot;, &quot;B&quot;, &quot;C&quot;), each = 5))

# Calculate means
grand_mean &lt;- mean(all_data)
group_means &lt;- c(mean(group_a), mean(group_b), mean(group_c))
n_per_group &lt;- 5
k &lt;- 3  # number of groups
N &lt;- length(all_data)

cat(&quot;Grand mean:&quot;, grand_mean, &quot;\n&quot;)</code></pre>
<pre><code>Grand mean: 30.06667 </code></pre>
<pre class="r"><code>cat(&quot;Group means:&quot;, group_means, &quot;\n&quot;)</code></pre>
<pre><code>Group means: 24.2 30 36 </code></pre>
<pre class="r"><code># Calculate Sum of Squares
ss_between &lt;- sum(n_per_group * (group_means - grand_mean)^2)
ss_within &lt;- sum((group_a - group_means[1])^2) +
             sum((group_b - group_means[2])^2) +
             sum((group_c - group_means[3])^2)
ss_total &lt;- sum((all_data - grand_mean)^2)

cat(&quot;\nSum of Squares:\n&quot;)</code></pre>
<pre><code>
Sum of Squares:</code></pre>
<pre class="r"><code>cat(&quot;SS_between:&quot;, ss_between, &quot;\n&quot;)</code></pre>
<pre><code>SS_between: 348.1333 </code></pre>
<pre class="r"><code>cat(&quot;SS_within:&quot;, ss_within, &quot;\n&quot;)</code></pre>
<pre><code>SS_within: 34.8 </code></pre>
<pre class="r"><code>cat(&quot;SS_total:&quot;, ss_total, &quot;\n&quot;)</code></pre>
<pre><code>SS_total: 382.9333 </code></pre>
<pre class="r"><code>cat(&quot;SS_between + SS_within =&quot;, ss_between + ss_within, &quot;(should equal SS_total)\n&quot;)</code></pre>
<pre><code>SS_between + SS_within = 382.9333 (should equal SS_total)</code></pre>
<pre class="r"><code># Calculate degrees of freedom
df_between &lt;- k - 1
df_within &lt;- N - k

# Calculate Mean Squares
ms_between &lt;- ss_between / df_between
ms_within &lt;- ss_within / df_within

cat(&quot;\nMean Squares:\n&quot;)</code></pre>
<pre><code>
Mean Squares:</code></pre>
<pre class="r"><code>cat(&quot;MS_between:&quot;, ms_between, &quot;\n&quot;)</code></pre>
<pre><code>MS_between: 174.0667 </code></pre>
<pre class="r"><code>cat(&quot;MS_within:&quot;, ms_within, &quot;\n&quot;)</code></pre>
<pre><code>MS_within: 2.9 </code></pre>
<pre class="r"><code># Calculate F-statistic
f_stat &lt;- ms_between / ms_within
p_value &lt;- pf(f_stat, df_between, df_within, lower.tail = FALSE)

cat(&quot;\nF-statistic:&quot;, f_stat, &quot;\n&quot;)</code></pre>
<pre><code>
F-statistic: 60.02299 </code></pre>
<pre class="r"><code>cat(&quot;p-value:&quot;, p_value, &quot;\n&quot;)</code></pre>
<pre><code>p-value: 5.632957e-07 </code></pre>
<p>Verify with R’s <code>aov()</code> function:</p>
<pre class="r"><code>df &lt;- data.frame(value = all_data, group = groups)
anova_result &lt;- aov(value ~ group, data = df)
summary(anova_result)</code></pre>
<pre><code>            Df Sum Sq Mean Sq F value   Pr(&gt;F)    
group        2  348.1   174.1   60.02 5.63e-07 ***
Residuals   12   34.8     2.9                     
---
Signif. codes:  0 &#39;***&#39; 0.001 &#39;**&#39; 0.01 &#39;*&#39; 0.05 &#39;.&#39; 0.1 &#39; &#39; 1</code></pre>
</div>
</div>
<div id="assumptions-of-anova" class="section level2">
<h2>Assumptions of ANOVA</h2>
<p>ANOVA has several assumptions that should be checked:</p>
<div id="independence-of-observations" class="section level3">
<h3>1. Independence of observations</h3>
<p>Observations should be independent of each other. This is a study
design issue.</p>
</div>
<div id="normality" class="section level3">
<h3>2. Normality</h3>
<p>The residuals should be approximately normally distributed. Check
with:</p>
<pre class="r"><code># Using the manual example data
shapiro.test(residuals(anova_result))</code></pre>
<pre><code>
    Shapiro-Wilk normality test

data:  residuals(anova_result)
W = 0.94877, p-value = 0.5053</code></pre>
<pre class="r"><code># Q-Q plot
par(mfrow = c(1, 2))
plot(anova_result, which = 2)
hist(residuals(anova_result), main = &quot;Histogram of Residuals&quot;,
     xlab = &quot;Residuals&quot;, col = &quot;lightblue&quot;)</code></pre>
<p><img src="figure/anova.Rmd/normality_check-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>par(mfrow = c(1, 1))</code></pre>
</div>
<div id="homogeneity-of-variances-homoscedasticity"
class="section level3">
<h3>3. Homogeneity of Variances (Homoscedasticity)</h3>
<p>The variance should be approximately equal across groups. Check with
Levene’s test or Bartlett’s test:</p>
<pre class="r"><code># Bartlett&#39;s test (sensitive to non-normality)
bartlett.test(value ~ group, data = df)</code></pre>
<pre><code>
    Bartlett test of homogeneity of variances

data:  value by group
Bartlett&#39;s K-squared = 0.19158, df = 2, p-value = 0.9087</code></pre>
<pre class="r"><code># Visual check
boxplot(value ~ group, data = df, main = &quot;Variance by Group&quot;,
        col = c(&quot;lightblue&quot;, &quot;lightgreen&quot;, &quot;lightyellow&quot;))</code></pre>
<p><img src="figure/anova.Rmd/homogeneity_check-1.png" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="what-if-assumptions-are-violated" class="section level3">
<h3>What if Assumptions are Violated?</h3>
<table>
<thead>
<tr class="header">
<th>Violation</th>
<th>Solution</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>Non-normality</td>
<td>Use Kruskal-Wallis test (non-parametric)</td>
</tr>
<tr class="even">
<td>Unequal variances</td>
<td>Use Welch’s ANOVA</td>
</tr>
<tr class="odd">
<td>Both</td>
<td>Use Kruskal-Wallis test</td>
</tr>
</tbody>
</table>
<pre class="r"><code># Welch&#39;s ANOVA (doesn&#39;t assume equal variances)
oneway.test(value ~ group, data = df, var.equal = FALSE)</code></pre>
<pre><code>
    One-way analysis of means (not assuming equal variances)

data:  value and group
F = 52.688, num df = 2.0000, denom df = 7.9417, p-value = 2.607e-05</code></pre>
</div>
</div>
<div id="one-way-anova-for-gene-expression" class="section level2">
<h2>One-way ANOVA for Gene Expression</h2>
<p>One-way ANOVA tests for differences in means across groups defined by
a single factor.</p>
<pre class="r"><code>gene_expr &lt;- data.frame(
  Expression = c(5.2, 4.8, 5.1, 6.3, 6.8, 6.5, 7.2, 7.5, 7.1),
  Condition = rep(c(&quot;control&quot;, &quot;treated1&quot;, &quot;treated2&quot;), each = 3)
)

ggplot(gene_expr, aes(Condition, Expression)) +
  geom_boxplot(fill = &quot;lightblue&quot;) +
  geom_jitter(width = 0.1, alpha = 0.7) +
  theme_minimal() +
  labs(title = &quot;Gene Expression by Condition&quot;)</code></pre>
<p><img src="figure/anova.Rmd/simple_example-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-simple_example-1">
Past versions of simple_example-1.png
</button>
</p>
<div id="fig-simple_example-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/davetang/muse/blob/78ef6833b847b51d9167889246b69aac34a883d9/docs/figure/anova.Rmd/simple_example-1.png" target="_blank">78ef683</a>
</td>
<td>
Dave Tang
</td>
<td>
2025-04-02
</td>
</tr>
</tbody>
</table>
</div>
</div>
<pre class="r"><code>anova_result &lt;- aov(Expression ~ Condition, data = gene_expr)
summary(anova_result)</code></pre>
<pre><code>            Df Sum Sq Mean Sq F value   Pr(&gt;F)    
Condition    2  7.776   3.888   77.76 5.13e-05 ***
Residuals    6  0.300   0.050                     
---
Signif. codes:  0 &#39;***&#39; 0.001 &#39;**&#39; 0.01 &#39;*&#39; 0.05 &#39;.&#39; 0.1 &#39; &#39; 1</code></pre>
<div id="understanding-the-anova-table" class="section level3">
<h3>Understanding the ANOVA Table</h3>
<table>
<thead>
<tr class="header">
<th>Column</th>
<th>Meaning</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><strong>Df</strong></td>
<td>Degrees of freedom</td>
</tr>
<tr class="even">
<td><strong>Sum Sq</strong></td>
<td>Sum of Squares</td>
</tr>
<tr class="odd">
<td><strong>Mean Sq</strong></td>
<td>Mean Square (Sum Sq / Df)</td>
</tr>
<tr class="even">
<td><strong>F value</strong></td>
<td>F-statistic (ratio of Mean Sq)</td>
</tr>
<tr class="odd">
<td><strong>Pr(&gt;F)</strong></td>
<td>p-value</td>
</tr>
</tbody>
</table>
<p>In this example:</p>
<ul>
<li>The F-value is large (50.17)</li>
<li>The p-value is very small (0.00178)</li>
<li>Conclusion: There is a statistically significant difference in gene
expression between at least two conditions</li>
</ul>
</div>
</div>
<div id="post-hoc-tests" class="section level2">
<h2>Post-hoc Tests</h2>
<p>ANOVA tells us that groups differ, but not which groups. Post-hoc
tests identify specific differences.</p>
<div id="tukeys-hsd-honestly-significant-difference"
class="section level3">
<h3>Tukey’s HSD (Honestly Significant Difference)</h3>
<p>Tukey’s HSD compares all pairs of groups while controlling for
multiple testing:</p>
<pre class="r"><code>tukey_result &lt;- TukeyHSD(anova_result)
tukey_result</code></pre>
<pre><code>  Tukey multiple comparisons of means
    95% family-wise confidence level

Fit: aov(formula = Expression ~ Condition, data = gene_expr)

$Condition
                       diff       lwr      upr     p adj
treated1-control  1.5000000 0.9398123 2.060188 0.0004304
treated2-control  2.2333333 1.6731456 2.793521 0.0000448
treated2-treated1 0.7333333 0.1731456 1.293521 0.0164307</code></pre>
<p>Interpretation:</p>
<ul>
<li><code>diff</code> = difference between group means</li>
<li><code>lwr</code> and <code>upr</code> = 95% confidence interval for
the difference</li>
<li><code>p adj</code> = adjusted p-value</li>
</ul>
<p>If the confidence interval doesn’t include 0, the difference is
significant.</p>
<pre class="r"><code>plot(tukey_result, las = 1, col = &quot;blue&quot;)</code></pre>
<p><img src="figure/anova.Rmd/tukey_plot-1.png" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="other-post-hoc-tests" class="section level3">
<h3>Other Post-hoc Tests</h3>
<pre class="r"><code># Pairwise t-tests with Bonferroni correction
pairwise.t.test(gene_expr$Expression, gene_expr$Condition, p.adjust.method = &quot;bonferroni&quot;)</code></pre>
<pre><code>
    Pairwise comparisons using t tests with pooled SD 

data:  gene_expr$Expression and gene_expr$Condition 

         control treated1
treated1 0.00053 -       
treated2 5.5e-05 0.02096 

P value adjustment method: bonferroni </code></pre>
<pre class="r"><code># Pairwise t-tests with Holm correction (less conservative)
pairwise.t.test(gene_expr$Expression, gene_expr$Condition, p.adjust.method = &quot;holm&quot;)</code></pre>
<pre><code>
    Pairwise comparisons using t tests with pooled SD 

data:  gene_expr$Expression and gene_expr$Condition 

         control treated1
treated1 0.00035 -       
treated2 5.5e-05 0.00699 

P value adjustment method: holm </code></pre>
</div>
</div>
<div id="effect-size" class="section level2">
<h2>Effect Size</h2>
<p>Statistical significance doesn’t tell us about practical importance.
Effect size measures help:</p>
<div id="eta-squared-eta2" class="section level3">
<h3>Eta-squared (<span class="math inline">\(\eta^2\)</span>)</h3>
<p>Eta-squared represents the proportion of variance explained by the
factor:</p>
<p><span class="math display">\[\eta^2 =
\frac{SS_{between}}{SS_{total}}\]</span></p>
<pre class="r"><code># Extract sum of squares from ANOVA
ss &lt;- summary(anova_result)[[1]]$`Sum Sq`
ss_between &lt;- ss[1]
ss_total &lt;- sum(ss)

eta_squared &lt;- ss_between / ss_total
cat(&quot;Eta-squared:&quot;, round(eta_squared, 3), &quot;\n&quot;)</code></pre>
<pre><code>Eta-squared: 0.963 </code></pre>
</div>
<div id="omega-squared-omega2" class="section level3">
<h3>Omega-squared (<span class="math inline">\(\omega^2\)</span>)</h3>
<p>Omega-squared is a less biased estimate:</p>
<p><span class="math display">\[\omega^2 = \frac{SS_{between} -
df_{between} \times MS_{within}}{SS_{total} + MS_{within}}\]</span></p>
<pre class="r"><code>ms_within &lt;- ss[2] / summary(anova_result)[[1]]$Df[2]
df_between &lt;- summary(anova_result)[[1]]$Df[1]

omega_squared &lt;- (ss_between - df_between * ms_within) / (ss_total + ms_within)
cat(&quot;Omega-squared:&quot;, round(omega_squared, 3), &quot;\n&quot;)</code></pre>
<pre><code>Omega-squared: 0.945 </code></pre>
</div>
<div id="interpretation-guidelines" class="section level3">
<h3>Interpretation Guidelines</h3>
<table>
<thead>
<tr class="header">
<th>Effect Size</th>
<th><span class="math inline">\(\eta^2\)</span> / <span
class="math inline">\(\omega^2\)</span></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>Small</td>
<td>0.01</td>
</tr>
<tr class="even">
<td>Medium</td>
<td>0.06</td>
</tr>
<tr class="odd">
<td>Large</td>
<td>0.14</td>
</tr>
</tbody>
</table>
</div>
</div>
<div id="two-way-anova" class="section level2">
<h2>Two-way ANOVA</h2>
<p>Two-way ANOVA examines the effect of two factors and their
interaction.</p>
<pre class="r"><code># Simulated experiment: Drug effect across different cell types
set.seed(42)
two_way_data &lt;- expand.grid(
  Drug = c(&quot;Placebo&quot;, &quot;Drug_A&quot;, &quot;Drug_B&quot;),
  CellType = c(&quot;Type1&quot;, &quot;Type2&quot;),
  Replicate = 1:5
)

# Simulate expression values with main effects and interaction
two_way_data$Expression &lt;- with(two_way_data, {
  base &lt;- 10
  drug_effect &lt;- ifelse(Drug == &quot;Placebo&quot;, 0,
                        ifelse(Drug == &quot;Drug_A&quot;, 3, 5))
  cell_effect &lt;- ifelse(CellType == &quot;Type1&quot;, 0, 2)
  # Interaction: Drug_B works better in Type2 cells
  interaction &lt;- ifelse(Drug == &quot;Drug_B&quot; &amp; CellType == &quot;Type2&quot;, 3, 0)
  base + drug_effect + cell_effect + interaction + rnorm(nrow(two_way_data), 0, 1)
})

head(two_way_data)</code></pre>
<pre><code>     Drug CellType Replicate Expression
1 Placebo    Type1         1   11.37096
2  Drug_A    Type1         1   12.43530
3  Drug_B    Type1         1   15.36313
4 Placebo    Type2         1   12.63286
5  Drug_A    Type2         1   15.40427
6  Drug_B    Type2         1   19.89388</code></pre>
<pre class="r"><code>ggplot(two_way_data, aes(Drug, Expression, fill = CellType)) +
  geom_boxplot() +
  theme_minimal() +
  labs(title = &quot;Two-way ANOVA: Drug Effect by Cell Type&quot;)</code></pre>
<p><img src="figure/anova.Rmd/two_way_visualise-1.png" width="672" style="display: block; margin: auto;" /></p>
<div id="running-two-way-anova" class="section level3">
<h3>Running Two-way ANOVA</h3>
<pre class="r"><code># Two-way ANOVA with interaction
two_way_result &lt;- aov(Expression ~ Drug * CellType, data = two_way_data)
summary(two_way_result)</code></pre>
<pre><code>              Df Sum Sq Mean Sq F value   Pr(&gt;F)    
Drug           2 232.45  116.22  64.289 2.29e-10 ***
CellType       1  58.49   58.49  32.353 7.37e-06 ***
Drug:CellType  2  14.67    7.34   4.058   0.0303 *  
Residuals     24  43.39    1.81                     
---
Signif. codes:  0 &#39;***&#39; 0.001 &#39;**&#39; 0.01 &#39;*&#39; 0.05 &#39;.&#39; 0.1 &#39; &#39; 1</code></pre>
<p>Interpretation:</p>
<ul>
<li><strong>Drug</strong>: Main effect of drug treatment</li>
<li><strong>CellType</strong>: Main effect of cell type</li>
<li><strong>Drug:CellType</strong>: Interaction effect (does the drug
effect depend on cell type?)</li>
</ul>
<p>A significant interaction means the effect of one factor depends on
the level of the other factor.</p>
</div>
<div id="interaction-plot" class="section level3">
<h3>Interaction Plot</h3>
<pre class="r"><code># Calculate means for interaction plot
interaction_means &lt;- two_way_data %&gt;%
  group_by(Drug, CellType) %&gt;%
  summarise(Mean = mean(Expression), .groups = &quot;drop&quot;)

ggplot(interaction_means, aes(Drug, Mean, color = CellType, group = CellType)) +
  geom_point(size = 3) +
  geom_line(linewidth = 1) +
  theme_minimal() +
  labs(title = &quot;Interaction Plot&quot;, y = &quot;Mean Expression&quot;)</code></pre>
<p><img src="figure/anova.Rmd/interaction_plot-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>Non-parallel lines suggest an interaction. Here, Drug_B shows a
stronger effect in Type2 cells.</p>
</div>
<div id="post-hoc-for-two-way-anova" class="section level3">
<h3>Post-hoc for Two-way ANOVA</h3>
<pre class="r"><code>TukeyHSD(two_way_result, which = &quot;Drug&quot;)</code></pre>
<pre><code>  Tukey multiple comparisons of means
    95% family-wise confidence level

Fit: aov(formula = Expression ~ Drug * CellType, data = two_way_data)

$Drug
                   diff      lwr      upr    p adj
Drug_A-Placebo 3.305459 1.803818 4.807100 3.43e-05
Drug_B-Placebo 6.817309 5.315668 8.318951 0.00e+00
Drug_B-Drug_A  3.511850 2.010209 5.013492 1.47e-05</code></pre>
</div>
</div>
<div id="repeated-measures-anova" class="section level2">
<h2>Repeated Measures ANOVA</h2>
<p>When the same subjects are measured multiple times (e.g., before and
after treatment), use repeated measures ANOVA.</p>
<pre class="r"><code># Simulated longitudinal data
set.seed(123)
n_subjects &lt;- 10
repeated_data &lt;- data.frame(
  Subject = factor(rep(1:n_subjects, 3)),
  Time = factor(rep(c(&quot;Baseline&quot;, &quot;Week1&quot;, &quot;Week2&quot;), each = n_subjects)),
  Expression = c(
    rnorm(n_subjects, 10, 2),           # Baseline
    rnorm(n_subjects, 12, 2),           # Week 1 (slight increase)
    rnorm(n_subjects, 15, 2)            # Week 2 (larger increase)
  )
)

# Add subject-specific variation
subject_effect &lt;- rep(rnorm(n_subjects, 0, 3), 3)
repeated_data$Expression &lt;- repeated_data$Expression + subject_effect

head(repeated_data)</code></pre>
<pre><code>  Subject     Time Expression
1       1 Baseline  10.158441
2       2 Baseline   8.654431
3       3 Baseline  15.802794
4       4 Baseline  12.775417
5       5 Baseline  12.723319
6       6 Baseline  15.496051</code></pre>
<pre class="r"><code>ggplot(repeated_data, aes(Time, Expression, group = Subject)) +
  geom_line(alpha = 0.5) +
  geom_point() +
  stat_summary(aes(group = 1), fun = mean, geom = &quot;line&quot;,
               color = &quot;red&quot;, linewidth = 2) +
  stat_summary(aes(group = 1), fun = mean, geom = &quot;point&quot;,
               color = &quot;red&quot;, size = 3) +
  theme_minimal() +
  labs(title = &quot;Repeated Measures: Expression Over Time&quot;,
       subtitle = &quot;Red line shows group mean&quot;)</code></pre>
<p><img src="figure/anova.Rmd/repeated_plot-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code># Repeated measures ANOVA (using Error term for subject)
repeated_result &lt;- aov(Expression ~ Time + Error(Subject/Time), data = repeated_data)
summary(repeated_result)</code></pre>
<pre><code>
Error: Subject
          Df Sum Sq Mean Sq F value Pr(&gt;F)
Residuals  9  123.2   13.68               

Error: Subject:Time
          Df Sum Sq Mean Sq F value  Pr(&gt;F)   
Time       2  80.54   40.27   9.467 0.00155 **
Residuals 18  76.57    4.25                   
---
Signif. codes:  0 &#39;***&#39; 0.001 &#39;**&#39; 0.01 &#39;*&#39; 0.05 &#39;.&#39; 0.1 &#39; &#39; 1</code></pre>
</div>
<div id="non-parametric-alternative-kruskal-wallis-test"
class="section level2">
<h2>Non-parametric Alternative: Kruskal-Wallis Test</h2>
<p>When ANOVA assumptions are violated, use the Kruskal-Wallis test:</p>
<pre class="r"><code># Using the gene expression data
kruskal.test(Expression ~ Condition, data = gene_expr)</code></pre>
<pre><code>
    Kruskal-Wallis rank sum test

data:  Expression by Condition
Kruskal-Wallis chi-squared = 7.2, df = 2, p-value = 0.02732</code></pre>
<p>Post-hoc for Kruskal-Wallis using Dunn’s test:</p>
<pre class="r"><code># Pairwise Wilcoxon tests with correction
pairwise.wilcox.test(gene_expr$Expression, gene_expr$Condition,
                     p.adjust.method = &quot;BH&quot;)</code></pre>
<pre><code>
    Pairwise comparisons using Wilcoxon rank sum exact test 

data:  gene_expr$Expression and gene_expr$Condition 

         control treated1
treated1 0.1     -       
treated2 0.1     0.1     

P value adjustment method: BH </code></pre>
</div>
<div id="example-plantgrowth-dataset" class="section level2">
<h2>Example: PlantGrowth Dataset</h2>
<p>R’s built-in <code>PlantGrowth</code> dataset provides a good
example:</p>
<pre class="r"><code>data(PlantGrowth)
str(PlantGrowth)</code></pre>
<pre><code>&#39;data.frame&#39;:   30 obs. of  2 variables:
 $ weight: num  4.17 5.58 5.18 6.11 4.5 4.61 5.17 4.53 5.33 5.14 ...
 $ group : Factor w/ 3 levels &quot;ctrl&quot;,&quot;trt1&quot;,..: 1 1 1 1 1 1 1 1 1 1 ...</code></pre>
<pre class="r"><code># Visualise
ggplot(PlantGrowth, aes(group, weight, fill = group)) +
  geom_boxplot() +
  geom_jitter(width = 0.1, alpha = 0.5) +
  theme_minimal() +
  labs(title = &quot;Plant Growth by Treatment Group&quot;,
       x = &quot;Treatment&quot;, y = &quot;Dried Weight&quot;)</code></pre>
<p><img src="figure/anova.Rmd/plantgrowth-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code># ANOVA
plant_aov &lt;- aov(weight ~ group, data = PlantGrowth)
summary(plant_aov)</code></pre>
<pre><code>            Df Sum Sq Mean Sq F value Pr(&gt;F)  
group        2  3.766  1.8832   4.846 0.0159 *
Residuals   27 10.492  0.3886                 
---
Signif. codes:  0 &#39;***&#39; 0.001 &#39;**&#39; 0.01 &#39;*&#39; 0.05 &#39;.&#39; 0.1 &#39; &#39; 1</code></pre>
<pre class="r"><code># Effect size
plant_ss &lt;- summary(plant_aov)[[1]]$`Sum Sq`
cat(&quot;\nEta-squared:&quot;, round(plant_ss[1] / sum(plant_ss), 3), &quot;\n&quot;)</code></pre>
<pre><code>
Eta-squared: 0.264 </code></pre>
<pre class="r"><code># Post-hoc
TukeyHSD(plant_aov)</code></pre>
<pre><code>  Tukey multiple comparisons of means
    95% family-wise confidence level

Fit: aov(formula = weight ~ group, data = PlantGrowth)

$group
            diff        lwr       upr     p adj
trt1-ctrl -0.371 -1.0622161 0.3202161 0.3908711
trt2-ctrl  0.494 -0.1972161 1.1852161 0.1979960
trt2-trt1  0.865  0.1737839 1.5562161 0.0120064</code></pre>
</div>
<div id="example-simulating-different-scenarios" class="section level2">
<h2>Example: Simulating Different Scenarios</h2>
<div id="scenario-1-no-difference-between-groups"
class="section level3">
<h3>Scenario 1: No Difference Between Groups</h3>
<pre class="r"><code>set.seed(1)
no_diff &lt;- data.frame(
  value = rnorm(30, mean = 50, sd = 10),
  group = factor(rep(c(&quot;A&quot;, &quot;B&quot;, &quot;C&quot;), each = 10))
)

ggplot(no_diff, aes(group, value)) +
  geom_boxplot(fill = &quot;lightgray&quot;) +
  theme_minimal() +
  labs(title = &quot;No Difference Between Groups&quot;)</code></pre>
<p><img src="figure/anova.Rmd/no_difference-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>summary(aov(value ~ group, data = no_diff))</code></pre>
<pre><code>            Df Sum Sq Mean Sq F value Pr(&gt;F)
group        2   76.9   38.44   0.432  0.653
Residuals   27 2399.7   88.88               </code></pre>
</div>
<div id="scenario-2-one-group-different" class="section level3">
<h3>Scenario 2: One Group Different</h3>
<pre class="r"><code>set.seed(2)
one_diff &lt;- data.frame(
  value = c(rnorm(10, 50, 5), rnorm(10, 50, 5), rnorm(10, 65, 5)),
  group = factor(rep(c(&quot;A&quot;, &quot;B&quot;, &quot;C&quot;), each = 10))
)

ggplot(one_diff, aes(group, value)) +
  geom_boxplot(fill = &quot;lightblue&quot;) +
  theme_minimal() +
  labs(title = &quot;Group C Different from A and B&quot;)</code></pre>
<p><img src="figure/anova.Rmd/one_different-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>summary(aov(value ~ group, data = one_diff))</code></pre>
<pre><code>            Df Sum Sq Mean Sq F value   Pr(&gt;F)    
group        2   1601   800.7   21.68 2.42e-06 ***
Residuals   27    997    36.9                     
---
Signif. codes:  0 &#39;***&#39; 0.001 &#39;**&#39; 0.01 &#39;*&#39; 0.05 &#39;.&#39; 0.1 &#39; &#39; 1</code></pre>
<pre class="r"><code>TukeyHSD(aov(value ~ group, data = one_diff))</code></pre>
<pre><code>  Tukey multiple comparisons of means
    95% family-wise confidence level

Fit: aov(formula = value ~ group, data = one_diff)

$group
          diff       lwr       upr     p adj
B-A -0.1569061 -6.894852  6.581039 0.9981639
C-A 15.4197086  8.681763 22.157654 0.0000146
C-B 15.5766148  8.838669 22.314560 0.0000125</code></pre>
</div>
<div id="scenario-3-all-groups-different" class="section level3">
<h3>Scenario 3: All Groups Different</h3>
<pre class="r"><code>set.seed(3)
all_diff &lt;- data.frame(
  value = c(rnorm(10, 40, 3), rnorm(10, 50, 3), rnorm(10, 60, 3)),
  group = factor(rep(c(&quot;A&quot;, &quot;B&quot;, &quot;C&quot;), each = 10))
)

ggplot(all_diff, aes(group, value)) +
  geom_boxplot(fill = &quot;lightgreen&quot;) +
  theme_minimal() +
  labs(title = &quot;All Groups Different&quot;)</code></pre>
<p><img src="figure/anova.Rmd/all_different-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>summary(aov(value ~ group, data = all_diff))</code></pre>
<pre><code>            Df Sum Sq Mean Sq F value   Pr(&gt;F)    
group        2 1825.3   912.6   147.3 2.98e-15 ***
Residuals   27  167.3     6.2                     
---
Signif. codes:  0 &#39;***&#39; 0.001 &#39;**&#39; 0.01 &#39;*&#39; 0.05 &#39;.&#39; 0.1 &#39; &#39; 1</code></pre>
<pre class="r"><code>TukeyHSD(aov(value ~ group, data = all_diff))</code></pre>
<pre><code>  Tukey multiple comparisons of means
    95% family-wise confidence level

Fit: aov(formula = value ~ group, data = all_diff)

$group
         diff       lwr      upr p adj
B-A  9.399671  6.639651 12.15969     0
C-A 19.105599 16.345579 21.86562     0
C-B  9.705928  6.945907 12.46595     0</code></pre>
</div>
</div>
<div id="anova-vs.-linear-regression" class="section level2">
<h2>ANOVA vs. Linear Regression</h2>
<p>ANOVA and linear regression are mathematically equivalent for
comparing group means:</p>
<pre class="r"><code># Using PlantGrowth data
# ANOVA approach
aov_result &lt;- aov(weight ~ group, data = PlantGrowth)

# Linear regression approach
lm_result &lt;- lm(weight ~ group, data = PlantGrowth)

# Compare F-statistics
cat(&quot;ANOVA F-statistic:\n&quot;)</code></pre>
<pre><code>ANOVA F-statistic:</code></pre>
<pre class="r"><code>summary(aov_result)</code></pre>
<pre><code>            Df Sum Sq Mean Sq F value Pr(&gt;F)  
group        2  3.766  1.8832   4.846 0.0159 *
Residuals   27 10.492  0.3886                 
---
Signif. codes:  0 &#39;***&#39; 0.001 &#39;**&#39; 0.01 &#39;*&#39; 0.05 &#39;.&#39; 0.1 &#39; &#39; 1</code></pre>
<pre class="r"><code>cat(&quot;\nLinear Model F-statistic:\n&quot;)</code></pre>
<pre><code>
Linear Model F-statistic:</code></pre>
<pre class="r"><code>summary(lm_result)</code></pre>
<pre><code>
Call:
lm(formula = weight ~ group, data = PlantGrowth)

Residuals:
    Min      1Q  Median      3Q     Max 
-1.0710 -0.4180 -0.0060  0.2627  1.3690 

Coefficients:
            Estimate Std. Error t value Pr(&gt;|t|)    
(Intercept)   5.0320     0.1971  25.527   &lt;2e-16 ***
grouptrt1    -0.3710     0.2788  -1.331   0.1944    
grouptrt2     0.4940     0.2788   1.772   0.0877 .  
---
Signif. codes:  0 &#39;***&#39; 0.001 &#39;**&#39; 0.01 &#39;*&#39; 0.05 &#39;.&#39; 0.1 &#39; &#39; 1

Residual standard error: 0.6234 on 27 degrees of freedom
Multiple R-squared:  0.2641,    Adjusted R-squared:  0.2096 
F-statistic: 4.846 on 2 and 27 DF,  p-value: 0.01591</code></pre>
<pre class="r"><code># The anova() function on lm gives same result
anova(lm_result)</code></pre>
<pre><code>Analysis of Variance Table

Response: weight
          Df  Sum Sq Mean Sq F value  Pr(&gt;F)  
group      2  3.7663  1.8832  4.8461 0.01591 *
Residuals 27 10.4921  0.3886                  
---
Signif. codes:  0 &#39;***&#39; 0.001 &#39;**&#39; 0.01 &#39;*&#39; 0.05 &#39;.&#39; 0.1 &#39; &#39; 1</code></pre>
</div>
<div id="summary" class="section level2">
<h2>Summary</h2>
<p>Key points about ANOVA:</p>
<ol style="list-style-type: decimal">
<li><strong>Purpose</strong>: Compare means across 3+ groups</li>
<li><strong>Null hypothesis</strong>: All group means are equal</li>
<li><strong>F-statistic</strong>: Ratio of between-group to within-group
variance</li>
<li><strong>Assumptions</strong>: Independence, normality, homogeneity
of variances</li>
<li><strong>Post-hoc tests</strong>: Required to identify which groups
differ</li>
<li><strong>Effect size</strong>: Report <span
class="math inline">\(\eta^2\)</span> or <span
class="math inline">\(\omega^2\)</span> alongside p-values</li>
</ol>
<div id="quick-reference" class="section level3">
<h3>Quick Reference</h3>
<table>
<thead>
<tr class="header">
<th>Situation</th>
<th>Test</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>Compare 2 groups</td>
<td>t-test</td>
</tr>
<tr class="even">
<td>Compare 3+ groups, 1 factor</td>
<td>One-way ANOVA</td>
</tr>
<tr class="odd">
<td>Compare groups, 2 factors</td>
<td>Two-way ANOVA</td>
</tr>
<tr class="even">
<td>Same subjects measured repeatedly</td>
<td>Repeated measures ANOVA</td>
</tr>
<tr class="odd">
<td>Assumptions violated</td>
<td>Kruskal-Wallis test</td>
</tr>
<tr class="even">
<td>Unequal variances</td>
<td>Welch’s ANOVA</td>
</tr>
</tbody>
</table>
</div>
<div id="reporting-anova-results" class="section level3">
<h3>Reporting ANOVA Results</h3>
<p>Example: “A one-way ANOVA revealed a statistically significant
difference in gene expression between conditions, F(2, 6) = 50.17, p =
0.002, <span class="math inline">\(\eta^2\)</span> = 0.94. Post-hoc
Tukey HSD tests showed that treated2 had significantly higher expression
than control (p = 0.002) and treated1 (p = 0.047).”</p>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span>
Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 4.5.0 (2025-04-11)
Platform: x86_64-pc-linux-gnu
Running under: Ubuntu 24.04.3 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 
LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.26.so;  LAPACK version 3.12.0

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

time zone: Etc/UTC
tzcode source: system (glibc)

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] lubridate_1.9.4 forcats_1.0.0   stringr_1.5.1   dplyr_1.1.4    
 [5] purrr_1.0.4     readr_2.1.5     tidyr_1.3.1     tibble_3.3.0   
 [9] ggplot2_3.5.2   tidyverse_2.0.0 workflowr_1.7.1

loaded via a namespace (and not attached):
 [1] sass_0.4.10        generics_0.1.4     stringi_1.8.7      hms_1.1.3         
 [5] digest_0.6.37      magrittr_2.0.3     timechange_0.3.0   evaluate_1.0.3    
 [9] grid_4.5.0         RColorBrewer_1.1-3 fastmap_1.2.0      rprojroot_2.0.4   
[13] jsonlite_2.0.0     processx_3.8.6     whisker_0.4.1      ps_1.9.1          
[17] promises_1.3.3     httr_1.4.7         scales_1.4.0       jquerylib_0.1.4   
[21] cli_3.6.5          rlang_1.1.6        withr_3.0.2        cachem_1.1.0      
[25] yaml_2.3.10        tools_4.5.0        tzdb_0.5.0         httpuv_1.6.16     
[29] vctrs_0.6.5        R6_2.6.1           lifecycle_1.0.4    git2r_0.36.2      
[33] fs_1.6.6           pkgconfig_2.0.3    callr_3.7.6        pillar_1.10.2     
[37] bslib_0.9.0        later_1.4.2        gtable_0.3.6       glue_1.8.0        
[41] Rcpp_1.0.14        xfun_0.52          tidyselect_1.2.1   rstudioapi_0.17.1 
[45] knitr_1.50         farver_2.1.2       htmltools_0.5.8.1  labeling_0.4.3    
[49] rmarkdown_2.29     compiler_4.5.0     getPass_0.2-4     </code></pre>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
https://docs.mathjax.org/en/latest/web/configuration.html. This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>




</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
