<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />



<meta name="date" content="2025-02-23" />

<title>Pokedex for Cell Types</title>

<script src="site_libs/header-attrs-2.28/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.13.2/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-6.4.2/css/all.min.css" rel="stylesheet" />
<link href="site_libs/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet" />

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/main/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">My blog</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/davetang/muse">
    <span class="fa fa-github"></span>
     
    Source code
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Pokedex for Cell Types</h1>
<h4 class="date">2025-02-23</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span>
workflowr <span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2025-02-23
</p>
<p>
<strong>Checks:</strong> <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 7
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong> <code>muse/</code> <span
class="glyphicon glyphicon-question-sign" aria-hidden="true"
title="This is the local directory in which the code in this file was executed.">
</span>
</p>
<p>
This reproducible <a href="https://rmarkdown.rstudio.com">R Markdown</a>
analysis was created with <a
  href="https://github.com/workflowr/workflowr">workflowr</a> (version
1.7.1). The <em>Checks</em> tab describes the reproducibility checks
that were applied when the results were created. The <em>Past
versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date
</a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git
repository, you know the exact version of the code that produced these
results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the
global environment can affect the analysis in your R Markdown file in
unknown ways. For reproduciblity it’s best to always run the code in an
empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20200712code">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Seed:</strong>
<code>set.seed(20200712)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20200712code"
class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20200712)</code> was run prior to running
the code in the R Markdown file. Setting a seed ensures that any results
that rely on randomness, e.g. subsampling or permutations, are
reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Session information:</strong>
recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package
versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be
confident that you successfully produced the results during this
run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr
project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomdavetangmusetree3b7d52a9ead799f9529b0084954e725c1905bc6ftargetblank3b7d52aa">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Repository version:</strong>
<a href="https://github.com/davetang/muse/tree/3b7d52a9ead799f9529b0084954e725c1905bc6f" target="_blank">3b7d52a</a>
</a>
</p>
</div>
<div
id="strongRepositoryversionstrongahrefhttpsgithubcomdavetangmusetree3b7d52a9ead799f9529b0084954e725c1905bc6ftargetblank3b7d52aa"
class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development
and connecting the code version to the results is critical for
reproducibility.
</p>
<p>
The results in this page were generated with repository version
<a href="https://github.com/davetang/muse/tree/3b7d52a9ead799f9529b0084954e725c1905bc6f" target="_blank">3b7d52a</a>.
See the <em>Past versions</em> tab to see a history of the changes made
to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for
the analysis have been committed to Git prior to generating the results
(you can use <code>wflow_publish</code> or
<code>wflow_git_commit</code>). workflowr only checks the R Markdown
file, but you know if there are other scripts or data files that it
depends on. Below is the status of the Git repository when the results
were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .Rproj.user/
    Ignored:    data/1M_neurons_filtered_gene_bc_matrices_h5.h5
    Ignored:    data/293t/
    Ignored:    data/293t_3t3_filtered_gene_bc_matrices.tar.gz
    Ignored:    data/293t_filtered_gene_bc_matrices.tar.gz
    Ignored:    data/5k_Human_Donor1_PBMC_3p_gem-x_5k_Human_Donor1_PBMC_3p_gem-x_count_sample_filtered_feature_bc_matrix.h5
    Ignored:    data/5k_Human_Donor2_PBMC_3p_gem-x_5k_Human_Donor2_PBMC_3p_gem-x_count_sample_filtered_feature_bc_matrix.h5
    Ignored:    data/5k_Human_Donor3_PBMC_3p_gem-x_5k_Human_Donor3_PBMC_3p_gem-x_count_sample_filtered_feature_bc_matrix.h5
    Ignored:    data/5k_Human_Donor4_PBMC_3p_gem-x_5k_Human_Donor4_PBMC_3p_gem-x_count_sample_filtered_feature_bc_matrix.h5
    Ignored:    data/Parent_SC3v3_Human_Glioblastoma_filtered_feature_bc_matrix.tar.gz
    Ignored:    data/brain_counts/
    Ignored:    data/cl.obo
    Ignored:    data/cl.owl
    Ignored:    data/jurkat/
    Ignored:    data/jurkat:293t_50:50_filtered_gene_bc_matrices.tar.gz
    Ignored:    data/jurkat_293t/
    Ignored:    data/jurkat_filtered_gene_bc_matrices.tar.gz
    Ignored:    data/pbmc20k/
    Ignored:    data/pbmc3k/
    Ignored:    data/pbmc4k_filtered_gene_bc_matrices.tar.gz
    Ignored:    data/refdata-gex-GRCh38-2020-A.tar.gz
    Ignored:    data/seurat_1m_neuron.rds
    Ignored:    data/t_3k_filtered_gene_bc_matrices.tar.gz
    Ignored:    r_packages_4.4.1/

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not
included in this status report because it is ok for generated content to
have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">
<p>
These are the previous versions of the repository in which changes were
made to the R Markdown (<code>analysis/celldex.Rmd</code>) and HTML
(<code>docs/celldex.html</code>) files. If you’ve configured a remote
Git repository (see <code>?wflow_git_remote</code>), click on the
hyperlinks in the table below to view the files as they were in that
past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/davetang/muse/blob/3b7d52a9ead799f9529b0084954e725c1905bc6f/analysis/celldex.Rmd" target="_blank">3b7d52a</a>
</td>
<td>
Dave Tang
</td>
<td>
2025-02-23
</td>
<td>
Human primary cell atlas
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/davetang/muse/1c8436b386472477d6bab7d669603a6894933608/docs/celldex.html" target="_blank">1c8436b</a>
</td>
<td>
Dave Tang
</td>
<td>
2025-02-22
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/davetang/muse/blob/5b60bcc40a66867f511050f70a3c71621244c6c4/analysis/celldex.Rmd" target="_blank">5b60bcc</a>
</td>
<td>
Dave Tang
</td>
<td>
2025-02-22
</td>
<td>
Monaco reference
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/davetang/muse/5917ecee01ca00e447c772a0ec65d11443918d21/docs/celldex.html" target="_blank">5917ece</a>
</td>
<td>
Dave Tang
</td>
<td>
2025-02-13
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/davetang/muse/blob/329c88173996a40b6f78701295e9c5879a9e9fc1/analysis/celldex.Rmd" target="_blank">329c881</a>
</td>
<td>
Dave Tang
</td>
<td>
2025-02-13
</td>
<td>
Pokedex for cell types
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<blockquote>
<p>The celldex package provides convenient access to several cell type
reference datasets. Most of these references are derived from bulk
RNA-seq or microarray data of cell populations that (hopefully) consist
of a pure cell type after sorting and/or culturing. The aim is to
provide a common resource for further analysis like cell type annotation
of single cell expression data or deconvolution of cell type proportions
in bulk expression datasets.</p>
</blockquote>
<div id="installation" class="section level2">
<h2>Installation</h2>
<p>Install <a
href="https://bioconductor.org/packages/3.20/data/experiment/html/celldex.html">celldex</a>.</p>
<pre class="r"><code>if (!require(&quot;BiocManager&quot;, quietly = TRUE))
    install.packages(&quot;BiocManager&quot;)

BiocManager::install(&quot;celldex&quot;)</code></pre>
</div>
<div id="references" class="section level2">
<h2>References</h2>
<p>We can examine the available references using the
<code>surveyReferences()</code> function. This returns a DataFrame of
the reference’s name and version, along with additional information like
the title, description, species, number of samples, available labels,
and so on.</p>
<pre class="r"><code>suppressPackageStartupMessages(library(celldex))
surveyReferences()</code></pre>
<pre><code>DataFrame with 7 rows and 10 columns
                    name     version        path                  title
             &lt;character&gt; &lt;character&gt; &lt;character&gt;            &lt;character&gt;
1                   dice  2024-02-26          NA Human bulk RNA-seq d..
2       blueprint_encode  2024-02-26          NA Human bulk RNA-seq d..
3                 immgen  2024-02-26          NA Mouse microarray exp..
4           mouse_rnaseq  2024-02-26          NA Bulk RNA-seq data of..
5                   hpca  2024-02-26          NA Microarray data from..
6 novershtern_hematopo..  2024-02-26          NA Bulk microarray expr..
7          monaco_immune  2024-02-26          NA Human bulk RNA-seq d..
             description taxonomy_id  genome   samples
             &lt;character&gt;      &lt;List&gt;  &lt;List&gt; &lt;integer&gt;
1 Human bulk RNA-seq d..        9606              1561
2 Human bulk RNA-seq d..        9606               259
3 Mouse microarray exp..       10090               830
4 Bulk RNA-seq data of..       10090 MGSCv37       358
5 Microarray data from..        9606               713
6 Bulk microarray expr..        9606               211
7 Human bulk RNA-seq d..        9606  GRCh38       114
                           labels
                           &lt;List&gt;
1 label.main,label.fine,label.ont
2 label.main,label.fine,label.ont
3 label.main,label.fine,label.ont
4 label.main,label.fine,label.ont
5 label.main,label.fine,label.ont
6 label.main,label.fine,label.ont
7 label.main,label.fine,label.ont
                                                                               sources
                                                                  &lt;SplitDataFrameList&gt;
1                   PubMed:30449622:NA,ExperimentHub:EH3488:NA,ExperimentHub:EH3489:NA
2                             PubMed:22955616:NA,PubMed:24091925:NA,PubMed:30643263:NA
3                                   PubMed:18800157:NA,GEO:GSE15907:NA,GEO:GSE37448:NA
4                  PubMed:30858345:NA,URL:https://github.com/B..:NA,PubMed:30643263:NA
5 PubMed:24053356:NA,PubMed:30643263:NA,GitHub:dviraran/SingleR:adc4a0e4d5cfa79db18f..
6                           PubMed:21241896:NA,GEO:GSE24759:NA,ExperimentHub:EH3490:NA
7                          PubMed:30726743:NA,GEO:GSE107011:NA,ExperimentHub:EH3496:NA</code></pre>
<p>Alternatively, users can search the text of each reference’s metadata
to identify relevant datasets. This may require some experimentation as
it depends on the level of detail in the metadata supplied by the
uploader.</p>
<pre class="r"><code>searchReferences(
    defineTextQuery(&quot;immun%&quot;, partial=TRUE) &amp; 
    defineTextQuery(&quot;9606&quot;, field=&quot;taxonomy_id&quot;)
)</code></pre>
<pre><code>DataFrame with 2 rows and 10 columns
              name     version        path                  title
       &lt;character&gt; &lt;character&gt; &lt;character&gt;            &lt;character&gt;
1             dice  2024-02-26          NA Human bulk RNA-seq d..
2 blueprint_encode  2024-02-26          NA Human bulk RNA-seq d..
             description taxonomy_id genome   samples
             &lt;character&gt;      &lt;List&gt; &lt;List&gt; &lt;integer&gt;
1 Human bulk RNA-seq d..        9606             1561
2 Human bulk RNA-seq d..        9606              259
                           labels
                           &lt;List&gt;
1 label.main,label.fine,label.ont
2 label.main,label.fine,label.ont
                                                             sources
                                                &lt;SplitDataFrameList&gt;
1 PubMed:30449622:NA,ExperimentHub:EH3488:NA,ExperimentHub:EH3489:NA
2           PubMed:22955616:NA,PubMed:24091925:NA,PubMed:30643263:NA</code></pre>
<div id="human-primary-cell-atlas" class="section level3">
<h3>Human primary cell atlas</h3>
<p>The HPCA reference consists of publicly available microarray datasets
derived from human primary cells (Mabbott et al. 2013). Most of the
labels refer to blood subpopulations but cell types from other tissues
are also available.</p>
<pre class="r"><code>hpca &lt;- fetchReference(&quot;hpca&quot;, &quot;2024-02-26&quot;)
colData(hpca) |&gt;
  as.data.frame() |&gt;
  dplyr::pull(&quot;label.main&quot;) |&gt;
  table() |&gt;
  as.data.frame() |&gt;
  dplyr::arrange(-Freq)</code></pre>
<pre><code>                   Var1 Freq
1            Macrophage   90
2                    DC   88
3               T_cells   68
4     Endothelial_cells   64
5              Monocyte   60
6     Tissue_stem_cells   55
7             iPS_cells   42
8                B_cell   26
9         Keratinocytes   25
10          Neutrophils   21
11 Embryonic_stem_cells   17
12     Epithelial_cells   16
13              Neurons   16
14  Smooth_muscle_cells   16
15          Osteoblasts   15
16          Fibroblasts   10
17           HSC_-G-CSF   10
18                  MSC    9
19         Chondrocytes    8
20         Erythroblast    8
21                   BM    7
22            HSC_CD34+    6
23          Gametocytes    5
24              NK_cell    5
25            Platelets    5
26          Hepatocytes    3
27            Astrocyte    2
28                  CMP    2
29                  GMP    2
30                  MEP    2
31            Myelocyte    2
32     Pre-B_cell_CD34-    2
33     Pro-B_cell_CD34+    2
34        Pro-Myelocyte    2
35           BM &amp; Prog.    1
36 Neuroepithelial_cell    1</code></pre>
</div>
<div id="monaco-reference" class="section level3">
<h3>Monaco reference</h3>
<p>The <a
href="https://bioconductor.org/packages/3.20/data/experiment/vignettes/celldex/inst/doc/userguide.html#44_Monaco_immune_data">Monaco
reference</a> consists of bulk RNA-seq samples of sorted immune cell
populations from GSE107011 (Monaco et al. 2019).</p>
<p>This is the human immune reference that best covers all of the bases
for a typical PBMC sample. It provides expansive B and T cell subsets,
differentiates between classical and non-classical monocytes, includes
basic dendritic cell subsets, and also includes neutrophil and basophil
samples to help identify small contaminating populations that may have
slipped into a PBMC preparation.</p>
<pre class="r"><code>monaco_immune &lt;- fetchReference(&quot;monaco_immune&quot;, &quot;2024-02-26&quot;)
colData(monaco_immune) |&gt;
  as.data.frame()</code></pre>
<pre><code>                       label.main                    label.fine  label.ont
DZQV_CD8_naive       CD8+ T cells             Naive CD8 T cells CL:0000900
DZQV_CD8_CM          CD8+ T cells    Central memory CD8 T cells CL:0000907
DZQV_CD8_EM          CD8+ T cells   Effector memory CD8 T cells CL:0000913
DZQV_CD8_TE          CD8+ T cells Terminal effector CD8 T cells CL:0001062
DZQV_MAIT                 T cells                    MAIT cells CL:0000940
DZQV_VD2+                 T cells                Vd2 gd T cells CL:0000798
DZQV_VD2-                 T cells            Non-Vd2 gd T cells CL:0000798
DZQV_TFH             CD4+ T cells     Follicular helper T cells CL:0002038
DZQV_Treg            CD4+ T cells            T regulatory cells CL:0000815
DZQV_Th1             CD4+ T cells                     Th1 cells CL:0000545
DZQV_Th1/Th17        CD4+ T cells                Th1/Th17 cells CL:0000912
DZQV_Th17            CD4+ T cells                    Th17 cells CL:0000899
DZQV_Th2             CD4+ T cells                     Th2 cells CL:0000546
DZQV_CD4_naive       CD4+ T cells             Naive CD4 T cells CL:0000895
DZQV_Progenitor       Progenitors              Progenitor cells CL:0002043
DZQV_B_naive              B cells                 Naive B cells CL:0000788
DZQV_B_NSM                B cells   Non-switched memory B cells CL:0000970
DZQV_B_Ex                 B cells             Exhausted B cells CL:0000236
DZQV_B_SM                 B cells       Switched memory B cells CL:0000972
DZQV_Plasmablasts         B cells                  Plasmablasts CL:0000980
DZQV_C_mono             Monocytes           Classical monocytes CL:0000860
DZQV_I_mono             Monocytes        Intermediate monocytes CL:0002393
DZQV_NC_mono            Monocytes       Non classical monocytes CL:0000875
DZQV_NK                  NK cells          Natural killer cells CL:0000623
DZQV_pDC          Dendritic cells  Plasmacytoid dendritic cells CL:0000784
DZQV_mDC          Dendritic cells       Myeloid dendritic cells CL:0000782
DZQV_Neutrophils      Neutrophils       Low-density neutrophils CL:0000096
DZQV_Basophils          Basophils         Low-density basophils CL:0000043
925L_CD8_naive       CD8+ T cells             Naive CD8 T cells CL:0000900
925L_CD8_CM          CD8+ T cells    Central memory CD8 T cells CL:0000907
925L_CD8_EM          CD8+ T cells   Effector memory CD8 T cells CL:0000913
925L_CD8_TE          CD8+ T cells Terminal effector CD8 T cells CL:0001062
925L_MAIT                 T cells                    MAIT cells CL:0000940
925L_VD2+                 T cells                Vd2 gd T cells CL:0000798
925L_VD2-                 T cells            Non-Vd2 gd T cells CL:0000798
925L_TFH             CD4+ T cells     Follicular helper T cells CL:0002038
925L_Treg            CD4+ T cells            T regulatory cells CL:0000815
925L_Th1             CD4+ T cells                     Th1 cells CL:0000545
925L_Th1/Th17        CD4+ T cells                Th1/Th17 cells CL:0000912
925L_Th17            CD4+ T cells                    Th17 cells CL:0000899
925L_Th2             CD4+ T cells                     Th2 cells CL:0000546
925L_CD4_naive       CD4+ T cells             Naive CD4 T cells CL:0000895
925L_CD4_TE          CD4+ T cells Terminal effector CD4 T cells CL:0001044
925L_Progenitor       Progenitors              Progenitor cells CL:0002043
925L_B_naive              B cells                 Naive B cells CL:0000788
925L_B_NSM                B cells   Non-switched memory B cells CL:0000970
925L_B_Ex                 B cells             Exhausted B cells CL:0000236
925L_B_SM                 B cells       Switched memory B cells CL:0000972
925L_Plasmablasts         B cells                  Plasmablasts CL:0000980
925L_C_mono             Monocytes           Classical monocytes CL:0000860
925L_I_mono             Monocytes        Intermediate monocytes CL:0002393
925L_NC_mono            Monocytes       Non classical monocytes CL:0000875
925L_NK                  NK cells          Natural killer cells CL:0000623
925L_pDC          Dendritic cells  Plasmacytoid dendritic cells CL:0000784
925L_mDC          Dendritic cells       Myeloid dendritic cells CL:0000782
925L_Neutrophils      Neutrophils       Low-density neutrophils CL:0000096
925L_Basophils          Basophils         Low-density basophils CL:0000043
9JD4_CD8_naive       CD8+ T cells             Naive CD8 T cells CL:0000900
9JD4_CD8_CM          CD8+ T cells    Central memory CD8 T cells CL:0000907
9JD4_CD8_EM          CD8+ T cells   Effector memory CD8 T cells CL:0000913
9JD4_CD8_TE          CD8+ T cells Terminal effector CD8 T cells CL:0001062
9JD4_MAIT                 T cells                    MAIT cells CL:0000940
9JD4_VD2+                 T cells                Vd2 gd T cells CL:0000798
9JD4_VD2-                 T cells            Non-Vd2 gd T cells CL:0000798
9JD4_TFH             CD4+ T cells     Follicular helper T cells CL:0002038
9JD4_Treg            CD4+ T cells            T regulatory cells CL:0000815
9JD4_Th1             CD4+ T cells                     Th1 cells CL:0000545
9JD4_Th1/Th17        CD4+ T cells                Th1/Th17 cells CL:0000912
9JD4_Th17            CD4+ T cells                    Th17 cells CL:0000899
9JD4_Th2             CD4+ T cells                     Th2 cells CL:0000546
9JD4_CD4_naive       CD4+ T cells             Naive CD4 T cells CL:0000895
9JD4_CD4_TE          CD4+ T cells Terminal effector CD4 T cells CL:0001044
9JD4_Progenitor       Progenitors              Progenitor cells CL:0002043
9JD4_B_naive              B cells                 Naive B cells CL:0000788
9JD4_B_NSM                B cells   Non-switched memory B cells CL:0000970
9JD4_B_Ex                 B cells             Exhausted B cells CL:0000236
9JD4_B_SM                 B cells       Switched memory B cells CL:0000972
9JD4_Plasmablasts         B cells                  Plasmablasts CL:0000980
9JD4_C_mono             Monocytes           Classical monocytes CL:0000860
9JD4_I_mono             Monocytes        Intermediate monocytes CL:0002393
9JD4_NC_mono            Monocytes       Non classical monocytes CL:0000875
9JD4_NK                  NK cells          Natural killer cells CL:0000623
9JD4_pDC          Dendritic cells  Plasmacytoid dendritic cells CL:0000784
9JD4_mDC          Dendritic cells       Myeloid dendritic cells CL:0000782
9JD4_Neutrophils      Neutrophils       Low-density neutrophils CL:0000096
9JD4_Basophils          Basophils         Low-density basophils CL:0000043
G4YW_CD8_naive       CD8+ T cells             Naive CD8 T cells CL:0000900
G4YW_CD8_CM          CD8+ T cells    Central memory CD8 T cells CL:0000907
G4YW_CD8_EM          CD8+ T cells   Effector memory CD8 T cells CL:0000913
G4YW_CD8_TE          CD8+ T cells Terminal effector CD8 T cells CL:0001062
G4YW_MAIT                 T cells                    MAIT cells CL:0000940
G4YW_VD2+                 T cells                Vd2 gd T cells CL:0000798
G4YW_VD2-                 T cells            Non-Vd2 gd T cells CL:0000798
G4YW_TFH             CD4+ T cells     Follicular helper T cells CL:0002038
G4YW_Treg            CD4+ T cells            T regulatory cells CL:0000815
G4YW_Th1             CD4+ T cells                     Th1 cells CL:0000545
G4YW_Th1/Th17        CD4+ T cells                Th1/Th17 cells CL:0000912
G4YW_Th17            CD4+ T cells                    Th17 cells CL:0000899
G4YW_Th2             CD4+ T cells                     Th2 cells CL:0000546
G4YW_CD4_naive       CD4+ T cells             Naive CD4 T cells CL:0000895
G4YW_Progenitor       Progenitors              Progenitor cells CL:0002043
G4YW_B_naive              B cells                 Naive B cells CL:0000788
G4YW_B_NSM                B cells   Non-switched memory B cells CL:0000970
G4YW_B_Ex                 B cells             Exhausted B cells CL:0000236
G4YW_B_SM                 B cells       Switched memory B cells CL:0000972
G4YW_Plasmablasts         B cells                  Plasmablasts CL:0000980
G4YW_C_mono             Monocytes           Classical monocytes CL:0000860
G4YW_I_mono             Monocytes        Intermediate monocytes CL:0002393
G4YW_NC_mono            Monocytes       Non classical monocytes CL:0000875
G4YW_NK                  NK cells          Natural killer cells CL:0000623
G4YW_pDC          Dendritic cells  Plasmacytoid dendritic cells CL:0000784
G4YW_mDC          Dendritic cells       Myeloid dendritic cells CL:0000782
G4YW_Neutrophils      Neutrophils       Low-density neutrophils CL:0000096
G4YW_Basophils          Basophils         Low-density basophils CL:0000043</code></pre>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span>
Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 4.4.1 (2024-06-14)
Platform: x86_64-pc-linux-gnu
Running under: Ubuntu 22.04.5 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 
LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

time zone: Etc/UTC
tzcode source: system (glibc)

attached base packages:
[1] stats4    stats     graphics  grDevices utils     datasets  methods  
[8] base     

other attached packages:
 [1] celldex_1.16.0              SummarizedExperiment_1.36.0
 [3] Biobase_2.66.0              GenomicRanges_1.58.0       
 [5] GenomeInfoDb_1.42.3         IRanges_2.40.1             
 [7] S4Vectors_0.44.0            BiocGenerics_0.52.0        
 [9] MatrixGenerics_1.18.1       matrixStats_1.4.1          
[11] workflowr_1.7.1            

loaded via a namespace (and not attached):
 [1] tidyselect_1.2.1          alabaster.se_1.6.0       
 [3] dplyr_1.1.4               blob_1.2.4               
 [5] filelock_1.0.3            Biostrings_2.74.1        
 [7] fastmap_1.2.0             BiocFileCache_2.14.0     
 [9] promises_1.3.0            digest_0.6.37            
[11] lifecycle_1.0.4           alabaster.matrix_1.6.1   
[13] processx_3.8.4            KEGGREST_1.46.0          
[15] alabaster.base_1.6.1      RSQLite_2.3.7            
[17] magrittr_2.0.3            compiler_4.4.1           
[19] rlang_1.1.4               sass_0.4.9               
[21] tools_4.4.1               utf8_1.2.4               
[23] yaml_2.3.10               knitr_1.48               
[25] S4Arrays_1.6.0            bit_4.5.0                
[27] curl_5.2.3                DelayedArray_0.32.0      
[29] abind_1.4-8               HDF5Array_1.34.0         
[31] gypsum_1.2.0              withr_3.0.2              
[33] grid_4.4.1                fansi_1.0.6              
[35] ExperimentHub_2.14.0      git2r_0.35.0             
[37] Rhdf5lib_1.28.0           cli_3.6.3                
[39] rmarkdown_2.28            crayon_1.5.3             
[41] generics_0.1.3            rstudioapi_0.17.1        
[43] httr_1.4.7                DelayedMatrixStats_1.28.1
[45] rhdf5_2.50.2              DBI_1.2.3                
[47] cachem_1.1.0              stringr_1.5.1            
[49] zlibbioc_1.52.0           parallel_4.4.1           
[51] AnnotationDbi_1.68.0      BiocManager_1.30.25      
[53] XVector_0.46.0            alabaster.schemas_1.6.0  
[55] vctrs_0.6.5               Matrix_1.7-0             
[57] jsonlite_1.8.9            callr_3.7.6              
[59] bit64_4.5.2               alabaster.ranges_1.6.0   
[61] jquerylib_0.1.4           glue_1.8.0               
[63] ps_1.8.1                  stringi_1.8.4            
[65] BiocVersion_3.20.0        later_1.3.2              
[67] UCSC.utils_1.2.0          tibble_3.2.1             
[69] pillar_1.9.0              rhdf5filters_1.18.0      
[71] rappdirs_0.3.3            htmltools_0.5.8.1        
[73] GenomeInfoDbData_1.2.13   httr2_1.0.5              
[75] R6_2.5.1                  dbplyr_2.5.0             
[77] sparseMatrixStats_1.18.0  rprojroot_2.0.4          
[79] evaluate_1.0.1            lattice_0.22-6           
[81] AnnotationHub_3.14.0      png_0.1-8                
[83] memoise_2.0.1             httpuv_1.6.15            
[85] bslib_0.8.0               Rcpp_1.0.13              
[87] SparseArray_1.6.1         whisker_0.4.1            
[89] xfun_0.48                 fs_1.6.4                 
[91] getPass_0.2-4             pkgconfig_2.0.3          </code></pre>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
https://docs.mathjax.org/en/latest/web/configuration.html. This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>




</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
