<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />



<meta name="date" content="2025-11-12" />

<title>Getting started with CELLxGENE Discover Census</title>

<script src="site_libs/header-attrs-2.29/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.13.2/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-6.5.2/css/all.min.css" rel="stylesheet" />
<link href="site_libs/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet" />

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/main/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">My blog</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/davetang/muse">
    <span class="fa fa-github"></span>
     
    Source code
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Getting started with CELLxGENE Discover
Census</h1>
<h4 class="date">2025-11-12</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span>
workflowr <span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2025-11-12
</p>
<p>
<strong>Checks:</strong> <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 7
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong> <code>muse/</code> <span
class="glyphicon glyphicon-question-sign" aria-hidden="true"
title="This is the local directory in which the code in this file was executed.">
</span>
</p>
<p>
This reproducible <a href="https://rmarkdown.rstudio.com">R Markdown</a>
analysis was created with <a
  href="https://github.com/workflowr/workflowr">workflowr</a> (version
1.7.1). The <em>Checks</em> tab describes the reproducibility checks
that were applied when the results were created. The <em>Past
versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date
</a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git
repository, you know the exact version of the code that produced these
results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the
global environment can affect the analysis in your R Markdown file in
unknown ways. For reproduciblity it’s best to always run the code in an
empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20200712code">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Seed:</strong>
<code>set.seed(20200712)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20200712code"
class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20200712)</code> was run prior to running
the code in the R Markdown file. Setting a seed ensures that any results
that rely on randomness, e.g. subsampling or permutations, are
reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Session information:</strong>
recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package
versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be
confident that you successfully produced the results during this
run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr
project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomdavetangmusetree98e75c742386f51b666ba17e7aac43fc31f1e09atargetblank98e75c7a">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Repository version:</strong>
<a href="https://github.com/davetang/muse/tree/98e75c742386f51b666ba17e7aac43fc31f1e09a" target="_blank">98e75c7</a>
</a>
</p>
</div>
<div
id="strongRepositoryversionstrongahrefhttpsgithubcomdavetangmusetree98e75c742386f51b666ba17e7aac43fc31f1e09atargetblank98e75c7a"
class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development
and connecting the code version to the results is critical for
reproducibility.
</p>
<p>
The results in this page were generated with repository version
<a href="https://github.com/davetang/muse/tree/98e75c742386f51b666ba17e7aac43fc31f1e09a" target="_blank">98e75c7</a>.
See the <em>Past versions</em> tab to see a history of the changes made
to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for
the analysis have been committed to Git prior to generating the results
(you can use <code>wflow_publish</code> or
<code>wflow_git_commit</code>). workflowr only checks the R Markdown
file, but you know if there are other scripts or data files that it
depends on. Below is the status of the Git repository when the results
were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .Rproj.user/
    Ignored:    data/1M_neurons_filtered_gene_bc_matrices_h5.h5
    Ignored:    data/293t/
    Ignored:    data/293t_3t3_filtered_gene_bc_matrices.tar.gz
    Ignored:    data/293t_filtered_gene_bc_matrices.tar.gz
    Ignored:    data/5k_Human_Donor1_PBMC_3p_gem-x_5k_Human_Donor1_PBMC_3p_gem-x_count_sample_filtered_feature_bc_matrix.h5
    Ignored:    data/5k_Human_Donor2_PBMC_3p_gem-x_5k_Human_Donor2_PBMC_3p_gem-x_count_sample_filtered_feature_bc_matrix.h5
    Ignored:    data/5k_Human_Donor3_PBMC_3p_gem-x_5k_Human_Donor3_PBMC_3p_gem-x_count_sample_filtered_feature_bc_matrix.h5
    Ignored:    data/5k_Human_Donor4_PBMC_3p_gem-x_5k_Human_Donor4_PBMC_3p_gem-x_count_sample_filtered_feature_bc_matrix.h5
    Ignored:    data/97516b79-8d08-46a6-b329-5d0a25b0be98.h5ad
    Ignored:    data/Parent_SC3v3_Human_Glioblastoma_filtered_feature_bc_matrix.tar.gz
    Ignored:    data/brain_counts/
    Ignored:    data/cl.obo
    Ignored:    data/cl.owl
    Ignored:    data/jurkat/
    Ignored:    data/jurkat:293t_50:50_filtered_gene_bc_matrices.tar.gz
    Ignored:    data/jurkat_293t/
    Ignored:    data/jurkat_filtered_gene_bc_matrices.tar.gz
    Ignored:    data/pbmc20k/
    Ignored:    data/pbmc20k_seurat/
    Ignored:    data/pbmc3k.h5ad
    Ignored:    data/pbmc3k/
    Ignored:    data/pbmc3k_bpcells_mat/
    Ignored:    data/pbmc3k_export.mtx
    Ignored:    data/pbmc3k_matrix.mtx
    Ignored:    data/pbmc3k_seurat.rds
    Ignored:    data/pbmc4k_filtered_gene_bc_matrices.tar.gz
    Ignored:    data/pbmc_1k_v3_filtered_feature_bc_matrix.h5
    Ignored:    data/pbmc_1k_v3_raw_feature_bc_matrix.h5
    Ignored:    data/refdata-gex-GRCh38-2020-A.tar.gz
    Ignored:    data/seurat_1m_neuron.rds
    Ignored:    data/t_3k_filtered_gene_bc_matrices.tar.gz
    Ignored:    r_packages_4.4.1/
    Ignored:    r_packages_4.5.0/

Untracked files:
    Untracked:  analysis/bioc_scrnaseq.Rmd
    Untracked:  bpcells_matrix/
    Untracked:  data/Caenorhabditis_elegans.WBcel235.113.gtf.gz
    Untracked:  data/GCF_043380555.1-RS_2024_12_gene_ontology.gaf.gz
    Untracked:  data/arab.rds
    Untracked:  data/astronomicalunit.csv
    Untracked:  data/femaleMiceWeights.csv
    Untracked:  data/lung_bcell.rds
    Untracked:  m3/
    Untracked:  women.json

Unstaged changes:
    Modified:   analysis/isoform_switch_analyzer.Rmd

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not
included in this status report because it is ok for generated content to
have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">
<p>
These are the previous versions of the repository in which changes were
made to the R Markdown (<code>analysis/cellxgene_census.Rmd</code>) and
HTML (<code>docs/cellxgene_census.html</code>) files. If you’ve
configured a remote Git repository (see <code>?wflow_git_remote</code>),
click on the hyperlinks in the table below to view the files as they
were in that past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/davetang/muse/blob/98e75c742386f51b666ba17e7aac43fc31f1e09a/analysis/cellxgene_census.Rmd" target="_blank">98e75c7</a>
</td>
<td>
Dave Tang
</td>
<td>
2025-11-12
</td>
<td>
Available assays
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/davetang/muse/164872b9d24e1210cba49d5253385df566469950/docs/cellxgene_census.html" target="_blank">164872b</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-04-08
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/davetang/muse/blob/b3568c0ffa80ed11b0dd2cd8564fb893f99aba38/analysis/cellxgene_census.Rmd" target="_blank">b3568c0</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-04-08
</td>
<td>
CELLxGENE Census
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<div id="preparation" class="section level2">
<h2>Preparation</h2>
<p>Following the <a
href="https://chanzuckerberg.github.io/cellxgene-census/r/index.html">documentation</a>
for the R package {cellxgene.census} which is part of CZ CELLxGENE
Discover Census.</p>
<p>{cellxgene.census} provides an API to efficiently access the
cloud-hosted Census single-cell data from R. In just a few seconds users
can access any slice of Census data using cell or gene filters across
hundreds of single-cell datasets.</p>
<p>Census data can be fetched in an iterative fashion for
bigger-than-memory slices of data, or quickly exported to basic R
structures, as well as {Seurat} or {SingleCellExperiment} objects for
downstream analysis.</p>
<p>Install from the <a
href="https://ropensci.org/r-universe/">R-universe platform</a>. If
installing using Ubuntu/Debian, you may need to install the following
libraries via APT:</p>
<ul>
<li>libxml2-dev</li>
<li>libssl-dev</li>
<li>libcurl4-openssl-dev</li>
</ul>
<p>In addition you must have at least <code>cmake</code> v3.21. Install
the {tiledbsoma} dependency first, since this takes some time to compile
and install, it’s better to fail first.</p>
<pre class="r"><code>install.packages(
  &#39;tiledbsoma&#39;,
  repos = c(&#39;https://tiledb-inc.r-universe.dev&#39;, &#39;https://cloud.r-project.org&#39;)
)</code></pre>
<p>Now install {cellxgene.census}, which should be a breeze now.</p>
<pre class="r"><code>install.packages(
  &quot;cellxgene.census&quot;,
  repos=c(&#39;https://chanzuckerberg.r-universe.dev&#39;, &#39;https://cloud.r-project.org&#39;)
)</code></pre>
<p>Now install {Seurat}.</p>
<pre class="r"><code>install.packages(&quot;Seurat&quot;)</code></pre>
</div>
<div id="querying-the-metadata" class="section level2">
<h2>Querying the metadata</h2>
<p><a
href="https://chanzuckerberg.github.io/cellxgene-census/r/articles/census_query_extract.html">Querying
and fetching the single-cell data and cell/gene metadata</a></p>
<pre class="r"><code>library(&quot;cellxgene.census&quot;)
packageVersion(&quot;cellxgene.census&quot;)</code></pre>
<pre><code>[1] &#39;1.16.1&#39;</code></pre>
<p>The human gene metadata of the Census, for RNA assays, is located at
<code>census$get("census_data")$get("homo_sapiens")$obs</code>. The
mouse cell metadata is at
<code>census$get("census_data")$get("mus_musculus").obs</code>.</p>
<p>To learn what metadata columns are available for fetching and
filtering we can directly look at the keys of the cell metadata.</p>
<pre class="r"><code>census &lt;- open_soma()</code></pre>
<pre><code>The stable Census release is currently 2025-01-30. Specify census_version = &quot;2025-01-30&quot; in future calls to open_soma() to ensure data consistency.</code></pre>
<pre class="r"><code>my_keys &lt;- census$get(&quot;census_data&quot;)$get(&quot;homo_sapiens&quot;)$obs$colnames()
my_keys</code></pre>
<pre><code> [1] &quot;soma_joinid&quot;                             
 [2] &quot;dataset_id&quot;                              
 [3] &quot;assay&quot;                                   
 [4] &quot;assay_ontology_term_id&quot;                  
 [5] &quot;cell_type&quot;                               
 [6] &quot;cell_type_ontology_term_id&quot;              
 [7] &quot;development_stage&quot;                       
 [8] &quot;development_stage_ontology_term_id&quot;      
 [9] &quot;disease&quot;                                 
[10] &quot;disease_ontology_term_id&quot;                
[11] &quot;donor_id&quot;                                
[12] &quot;is_primary_data&quot;                         
[13] &quot;observation_joinid&quot;                      
[14] &quot;self_reported_ethnicity&quot;                 
[15] &quot;self_reported_ethnicity_ontology_term_id&quot;
[16] &quot;sex&quot;                                     
[17] &quot;sex_ontology_term_id&quot;                    
[18] &quot;suspension_type&quot;                         
[19] &quot;tissue&quot;                                  
[20] &quot;tissue_ontology_term_id&quot;                 
[21] &quot;tissue_type&quot;                             
[22] &quot;tissue_general&quot;                          
[23] &quot;tissue_general_ontology_term_id&quot;         
[24] &quot;raw_sum&quot;                                 
[25] &quot;nnz&quot;                                     
[26] &quot;raw_mean_nnz&quot;                            
[27] &quot;raw_variance_nnz&quot;                        
[28] &quot;n_measured_vars&quot;                         </code></pre>
<p>Assays.</p>
<pre class="r"><code>suppressPackageStartupMessages({
  library(dplyr)
  library(ggplot2)
})
census$get(&quot;census_data&quot;)$get(&quot;homo_sapiens&quot;)$obs$read(column_names = c(&#39;dataset_id&#39;, &quot;assay&quot;))$concat() |&gt;
  as.data.frame() -&gt; avail_assays

avail_assays |&gt;
  distinct(dataset_id, assay) |&gt;
  dplyr::summarise(n = n(), .by = assay) |&gt;
  slice_max(order_by = n, n = 10) |&gt;
  arrange(n) |&gt;
  dplyr::mutate(assay = factor(assay, levels = assay)) |&gt;
  ggplot(aes(assay, n)) +
  geom_col() +
  theme_minimal() +
  coord_flip() +
  labs(x = &#39;Assay type&#39;, y = &#39;Datasets&#39;, title = &#39;CELLxGENE human datasets&#39;)</code></pre>
<p><img src="figure/cellxgene_census.Rmd/assays-1.png" width="672" style="display: block; margin: auto;" /></p>
<p><code>soma_joinid</code> is a special <code>SOMADataFrame</code>
column that is used for join operations. All of the keys can be used to
fetch specific columns or specific rows matching a condition. For the
latter we need to know the values we are looking for <em>a
priori</em>.</p>
<p>For example let’s see what are the possible values available for
<code>sex</code>. To this we can load all cell metadata but fetching
only for the column sex; <code>column_names</code> are character vector
indicating what metadata columns to fetch.</p>
<pre class="r"><code>census$get(&quot;census_data&quot;)$get(&quot;homo_sapiens&quot;)$obs$read(column_names = &quot;sex&quot;)$concat() |&gt;
  as.data.frame() |&gt;
  unique()</code></pre>
<pre><code>         sex
1     female
651     male
1034 unknown</code></pre>
<p>With this information we can fetch all cell metadata for a specific
sex value, for example “unknown”; the <code>value_filter</code> is an R
expression with selection conditions to fetch rows.</p>
<pre class="r"><code>census$get(&quot;census_data&quot;)$get(&quot;homo_sapiens&quot;)$obs$read(value_filter = &quot;sex == &#39;unknown&#39;&quot;)$concat() |&gt;
  as.data.frame() -&gt; sex_unknown

head(sex_unknown)</code></pre>
<pre><code>  soma_joinid                           dataset_id     assay
1        1033 fbd69faa-b0c5-45ba-89c9-da938a7f5a14 10x 3&#39; v2
2        1034 fbd69faa-b0c5-45ba-89c9-da938a7f5a14 10x 3&#39; v2
3        1035 fbd69faa-b0c5-45ba-89c9-da938a7f5a14 10x 3&#39; v2
4        1036 fbd69faa-b0c5-45ba-89c9-da938a7f5a14 10x 3&#39; v2
5        1037 fbd69faa-b0c5-45ba-89c9-da938a7f5a14 10x 3&#39; v2
6        1038 fbd69faa-b0c5-45ba-89c9-da938a7f5a14 10x 3&#39; v2
  assay_ontology_term_id   cell_type cell_type_ontology_term_id
1            EFO:0009899 erythrocyte                 CL:0000232
2            EFO:0009899 erythrocyte                 CL:0000232
3            EFO:0009899 erythrocyte                 CL:0000232
4            EFO:0009899 erythrocyte                 CL:0000232
5            EFO:0009899 erythrocyte                 CL:0000232
6            EFO:0009899 erythrocyte                 CL:0000232
                   development_stage development_stage_ontology_term_id disease
1 20th week post-fertilization stage                     HsapDv:0000057  normal
2 20th week post-fertilization stage                     HsapDv:0000057  normal
3 20th week post-fertilization stage                     HsapDv:0000057  normal
4 20th week post-fertilization stage                     HsapDv:0000057  normal
5 20th week post-fertilization stage                     HsapDv:0000057  normal
6 20th week post-fertilization stage                     HsapDv:0000057  normal
  disease_ontology_term_id        donor_id is_primary_data observation_joinid
1             PATO:0000461 fetal_19w_donor           FALSE         h_r5~dgk?K
2             PATO:0000461 fetal_19w_donor           FALSE         w*rWbY-D$r
3             PATO:0000461 fetal_19w_donor           FALSE         jYiZT4_2bP
4             PATO:0000461 fetal_19w_donor           FALSE         (5msji51MZ
5             PATO:0000461 fetal_19w_donor           FALSE         MW+Yf)(ITf
6             PATO:0000461 fetal_19w_donor           FALSE         uY82G&gt;CnDL
  self_reported_ethnicity self_reported_ethnicity_ontology_term_id     sex
1                 unknown                                  unknown unknown
2                 unknown                                  unknown unknown
3                 unknown                                  unknown unknown
4                 unknown                                  unknown unknown
5                 unknown                                  unknown unknown
6                 unknown                                  unknown unknown
  sex_ontology_term_id suspension_type tissue tissue_ontology_term_id
1              unknown            cell thymus          UBERON:0002370
2              unknown            cell thymus          UBERON:0002370
3              unknown            cell thymus          UBERON:0002370
4              unknown            cell thymus          UBERON:0002370
5              unknown            cell thymus          UBERON:0002370
6              unknown            cell thymus          UBERON:0002370
  tissue_type  tissue_general tissue_general_ontology_term_id raw_sum  nnz
1      tissue endocrine gland                  UBERON:0002368    6444 1125
2      tissue endocrine gland                  UBERON:0002368    6330 1180
3      tissue endocrine gland                  UBERON:0002368    6314 1419
4      tissue endocrine gland                  UBERON:0002368    6270 1225
5      tissue endocrine gland                  UBERON:0002368    6115 1074
6      tissue endocrine gland                  UBERON:0002368    5735 1125
  raw_mean_nnz raw_variance_nnz n_measured_vars
1     5.728000         3991.949           36406
2     5.364407         2470.166           36406
3     4.449612         1411.458           36406
4     5.118367         1791.147           36406
5     5.693669         3619.081           36406
6     5.097778         1956.028           36406</code></pre>
<p>We can use both <code>column_names</code> and
<code>value_filter</code> to perform specific queries. For example
fetching the disease column for the cell_type “B cell” in the
<code>tissue_general</code> “lung”.</p>
<pre class="r"><code>cell_metadata_b_cell &lt;- census$get(&quot;census_data&quot;)$get(&quot;homo_sapiens&quot;)$obs$read(
  value_filter = &quot;cell_type == &#39;B cell&#39; &amp; tissue_general == &#39;lung&#39;&quot;,
  column_names = &quot;disease&quot;
)

cell_metadata_b_cell &lt;- as.data.frame(cell_metadata_b_cell$concat())

table(cell_metadata_b_cell)</code></pre>
<pre><code>disease
                                                       Alzheimer disease 
                                                                       0 
                                     B-cell acute lymphoblastic leukemia 
                                                                       0 
                                             B-cell non-Hodgkin lymphoma 
                                                                       0 
                                                       Barrett esophagus 
                                                                       0 
                                                                COVID-19 
                                                                    7421 
                                                           Crohn disease 
                                                                       0 
                                                           Crohn ileitis 
                                                                       0 
                                                           Down syndrome 
                                                                       0 
                                          HER2 positive breast carcinoma 
                                                                       0 
                                                  HIV infectious disease 
                                                                       0 
                                                      Lewy body dementia 
                                                                       0 
                                                       Parkinson disease 
                                                                       0 
                                                            Pick disease 
                                                                       0 
                                             Plasmodium malariae malaria 
                                                                       0 
                                                        Sjogren syndrome 
                                                                       0 
                                                             Wilms tumor 
                                                                       0 
                                                    acute kidney failure 
                                                                       0 
                                                  acute myeloid leukemia 
                                                                       0 
                                             acute myocardial infarction 
                                                                       0 
                                            acute promyelocytic leukemia 
                                                                       0 
                                                          adenocarcinoma 
                                                                       0 
                                      age related macular degeneration 7 
                                                                       0 
                                           amyotrophic lateral sclerosis 
                                                                       0 
amyotrophic lateral sclerosis 26 with or without frontotemporal dementia 
                                                                       0 
                                                             anencephaly 
                                                                      79 
                         arrhythmogenic right ventricular cardiomyopathy 
                                                                       0 
                                                    aspiration pneumonia 
                                                                       0 
                                                         atherosclerosis 
                                                                       0 
                                                     atrial fibrillation 
                                                                       0 
                                                    basal cell carcinoma 
                                                                       0 
                                                    basal laminar drusen 
                                                                       0 
                                            benign prostatic hyperplasia 
                                                                       0 
                                                                blastoma 
                                                                       0 
                                                           breast cancer 
                                                                     157 
                                                        breast carcinoma 
                                                                       0 
                                                          cardiomyopathy 
                                                                       0 
                                                                cataract 
                                                                       0 
                                        chromophobe renal cell carcinoma 
                                                                       0 
                                                  chronic kidney disease 
                                                                       0 
                                   chronic obstructive pulmonary disease 
                                                                    6714 
                                                        chronic rhinitis 
                                                                       0 
                                              clear cell renal carcinoma 
                                                                       0 
                                                    clonal hematopoiesis 
                                                                       0 
                                    colon sessile serrated adenoma/polyp 
                                                                       0 
                                                       colorectal cancer 
                                                                       0 
                                                     colorectal neoplasm 
                                                                       0 
                                        common variable immunodeficiency 
                                                                       0 
                                                congenital heart disease 
                                                                       0 
                                                         cystic fibrosis 
                                                                       0 
                                                                dementia 
                                                                       0 
                                           diffuse large B-cell lymphoma 
                                                                       0 
                                               digestive system disorder 
                                                                       0 
                                                  dilated cardiomyopathy 
                                                                       0 
                                                           enamel caries 
                                                                       0 
                                                                epilepsy 
                                                                       0 
                                estrogen-receptor positive breast cancer 
                                                                       0 
                                                     follicular lymphoma 
                                                                       0 
                                                 frontotemporal dementia 
                                                                       0 
                                                          gastric cancer 
                                                                       0 
                                           gastric intestinal metaplasia 
                                                                       0 
                                                               gastritis 
                                                                       0 
                                                              gingivitis 
                                                                       0 
                                                            glioblastoma 
                                                                       0 
                                                          heart disorder 
                                                                       0 
                                                           heart failure 
                                                                       0 
                                                    hematologic disorder 
                                                                       0 
                                                hepatocellular carcinoma 
                                                                       0 
                                                            hydrosalpinx 
                                                                       0 
                                                      hyperplastic polyp 
                                                                       0 
                                            hypersensitivity pneumonitis 
                                                                      52 
                                             hypertrophic cardiomyopathy 
                                                                       0 
                                                               influenza 
                                                                       0 
                                                                  injury 
                                                                       0 
                                               interstitial lung disease 
                                                                     376 
                                         intrahepatic cholangiocarcinoma 
                                                                       0 
                                        invasive ductal breast carcinoma 
                                                                       0 
                                       invasive lobular breast carcinoma 
                                                                       0 
                                                juvenile dermatomyositis 
                                                                       0 
                                                                  keloid 
                                                                       0 
                                                  kidney benign neoplasm 
                                                                       0 
                                                       kidney oncocytoma 
                                                                       0 
               leukoencephalopathy, diffuse hereditary, with spheroids 1 
                                                                       0 
                                                             listeriosis 
                                                                       0 
                                                   localized scleroderma 
                                                                       0 
                                                           long COVID-19 
                                                                       0 
                                              luminal A breast carcinoma 
                                                                       0 
                                              luminal B breast carcinoma 
                                                                       0 
                                                     lung adenocarcinoma 
                                                                   62351 
                                               lung large cell carcinoma 
                                                                    1534 
                                                lymphangioleiomyomatosis 
                                                                     133 
                                                    macular degeneration 
                                                                       0 
                                          malignant ovarian serous tumor 
                                                                       0 
                                           malignant pancreatic neoplasm 
                                                                       0 
                                                                melanoma 
                                                                       0 
                                                     metastatic melanoma 
                                                                       0 
                                                      multiple sclerosis 
                                                                       0 
                                                myelodysplastic syndrome 
                                                                       0 
                              myelodysplastic/myeloproliferative disease 
                                                                       0 
                                                   myocardial infarction 
                                                                       0 
                                                neuroendocrine carcinoma 
                                                                       0 
                                           non-compaction cardiomyopathy 
                                                                       0 
                                           non-small cell lung carcinoma 
                                                                   17484 
                                     non-specific interstitial pneumonia 
                                                                     231 
                                       nonpapillary renal cell carcinoma 
                                                                       0 
                                                                  normal 
                                                                   32365 
                                                       opiate dependence 
                                                                       0 
                                     oral cavity squamous cell carcinoma 
                                                                       0 
                                      oropharynx squamous cell carcinoma 
                                                                       0 
                                                           periodontitis 
                                                                       0 
                                                   pilocytic astrocytoma 
                                                                       0 
                                                     plasma cell myeloma 
                                                                       0 
                                                   pleomorphic carcinoma 
                                                                    1210 
                                                               pneumonia 
                                                                      50 
                                                  post-COVID-19 disorder 
                                                                       0 
                                             primary biliary cholangitis 
                                                                       0 
               primary cutaneous diffuse large B-cell lymphoma, Leg type 
                                                                       0 
                                          primary sclerosing cholangitis 
                                                                       0 
                                          progressive supranuclear palsy 
                                                                       0 
                                                     pulmonary emphysema 
                                                                    1512 
                                                      pulmonary fibrosis 
                                                                    6798 
                                                   pulmonary sarcoidosis 
                                                                       6 
                                                                pulpitis 
                                                                       0 
                                                     respiratory failure 
                                                                       0 
                                             respiratory system disorder 
                                                                       0 
                                                    rheumatoid arthritis 
                                                                       0 
                                               small cell lung carcinoma 
                                                                     583 
                                                 squamous cell carcinoma 
                                                                       0 
                                            squamous cell lung carcinoma 
                                                                   11920 
                                            systemic lupus erythematosus 
                                                                       0 
                                                  temporal lobe epilepsy 
                                                                       0 
                                                           tongue cancer 
                                                                       0 
                                                           toxoplasmosis 
                                                                       0 
                                        triple-negative breast carcinoma 
                                                                       0 
                                                              trisomy 18 
                                                                       0 
                                                         tubular adenoma 
                                                                       0 
                                                   tubulovillous adenoma 
                                                                       0 
                                                type 1 diabetes mellitus 
                                                                       0 
                                                type 2 diabetes mellitus 
                                                                       0 </code></pre>
<div id="querying-expression-data-as-seurat" class="section level3">
<h3>Querying expression data as Seurat</h3>
<p>Use <code>get_seurat()</code> to perform the same type of filtering
but returning a Seurat object.</p>
<ul>
<li><code>obs_column_names</code> — character vector indicating the
columns to select for cell metadata.</li>
<li><code>obs_value_filter</code> — expression with selection conditions
to fetch cells meeting a criteria.</li>
<li><code>var_column_names</code> — character vector indicating the
columns to select for gene metadata.</li>
<li><code>var_value_filter</code> — expression with selection conditions
to fetch genes meeting a criteria.</li>
</ul>
<pre class="r"><code>suppressPackageStartupMessages(library(&quot;Seurat&quot;))

seurat_obj &lt;- get_seurat(
  census, &quot;Homo sapiens&quot;,
  obs_column_names = c(&quot;cell_type&quot;, &quot;tissue_general&quot;, &quot;disease&quot;, &quot;sex&quot;),
  var_value_filter = &quot;feature_id %in% c(&#39;ENSG00000161798&#39;, &#39;ENSG00000188229&#39;)&quot;,
  obs_value_filter = &quot;cell_type == &#39;B cell&#39; &amp; tissue_general == &#39;lung&#39; &amp; disease == &#39;COVID-19&#39;&quot;
)

saveRDS(object = seurat_obj, file = &quot;data/lung_bcell.rds&quot;)</code></pre>
<p>Seurat object.</p>
<pre class="r"><code>seurat_obj &lt;- readRDS(file = &quot;data/lung_bcell.rds&quot;)
seurat_obj</code></pre>
<pre><code>Loading required namespace: SeuratObject</code></pre>
<pre><code>An object of class Seurat 
2 features across 7421 samples within 1 assay 
Active assay: RNA (2 features, 0 variable features)
 2 layers present: counts, data</code></pre>
</div>
<div id="close-the-census" class="section level3">
<h3>Close the census</h3>
<p>After use, the census object should be closed to release memory and
other resources. This also closes all SOMA objects accessed via the
top-level census. Closing can be automated using
<code>on.exit(census$close(), add = TRUE)</code> immediately after
<code>census &lt;- open_soma()</code>.</p>
<pre class="r"><code>census$close()</code></pre>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span>
Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 4.5.0 (2025-04-11)
Platform: x86_64-pc-linux-gnu
Running under: Ubuntu 24.04.3 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 
LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.26.so;  LAPACK version 3.12.0

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

time zone: Etc/UTC
tzcode source: system (glibc)

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] ggplot2_3.5.2           dplyr_1.1.4             RcppSpdlog_0.0.23      
[4] cellxgene.census_1.16.1 workflowr_1.7.1        

loaded via a namespace (and not attached):
 [1] dotCall64_1.2       spam_2.11-1         gtable_0.3.6       
 [4] xfun_0.52           bslib_0.9.0         processx_3.8.6     
 [7] lattice_0.22-6      callr_3.7.6         vctrs_0.6.5        
[10] tools_4.5.0         ps_1.9.1            generics_0.1.4     
[13] parallel_4.5.0      curl_6.4.0          tibble_3.3.0       
[16] pkgconfig_2.0.3     Matrix_1.7-3        RColorBrewer_1.1-3 
[19] assertthat_0.2.1    lifecycle_1.0.4     compiler_4.5.0     
[22] farver_2.1.2        stringr_1.5.1       git2r_0.36.2       
[25] codetools_0.2-20    getPass_0.2-4       SeuratObject_5.1.0 
[28] tiledbsoma_2.1.0    httpuv_1.6.16       htmltools_0.5.8.1  
[31] nanoarrow_0.7.0-1   sass_0.4.10         yaml_2.3.10        
[34] later_1.4.2         pillar_1.10.2       jquerylib_0.1.4    
[37] whisker_0.4.1       aws.s3_0.3.22       cachem_1.1.0       
[40] RcppCCTZ_0.2.13     tiledb_0.33.1.1     parallelly_1.45.0  
[43] tidyselect_1.2.1    digest_0.6.37       future_1.58.0      
[46] stringi_1.8.7       listenv_0.9.1       purrr_1.0.4        
[49] labeling_0.4.3      arrow_22.0.0        rprojroot_2.0.4    
[52] fastmap_1.2.0       grid_4.5.0          cli_3.6.5          
[55] magrittr_2.0.3      base64enc_0.1-3     spdl_0.0.5         
[58] future.apply_1.20.0 aws.signature_0.6.0 withr_3.0.2        
[61] scales_1.4.0        promises_1.3.3      sp_2.2-0           
[64] bit64_4.6.0-1       nanotime_0.3.12     rmarkdown_2.29     
[67] httr_1.4.7          globals_0.18.0      bit_4.6.0          
[70] progressr_0.15.1    zoo_1.8-14          evaluate_1.0.3     
[73] knitr_1.50          rlang_1.1.6         Rcpp_1.0.14        
[76] glue_1.8.0          xml2_1.3.8          rstudioapi_0.17.1  
[79] jsonlite_2.0.0      R6_2.6.1            fs_1.6.6           </code></pre>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
https://docs.mathjax.org/en/latest/web/configuration.html. This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>




</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
