<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />



<meta name="date" content="2025-03-06" />

<title>Getting started with clusterProfiler</title>

<script src="site_libs/header-attrs-2.29/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.13.2/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-6.5.2/css/all.min.css" rel="stylesheet" />
<link href="site_libs/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet" />

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/main/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">My blog</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/davetang/muse">
    <span class="fa fa-github"></span>
     
    Source code
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Getting started with clusterProfiler</h1>
<h4 class="date">2025-03-06</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span>
workflowr <span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2025-03-06
</p>
<p>
<strong>Checks:</strong> <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 7
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong> <code>muse/</code> <span
class="glyphicon glyphicon-question-sign" aria-hidden="true"
title="This is the local directory in which the code in this file was executed.">
</span>
</p>
<p>
This reproducible <a href="https://rmarkdown.rstudio.com">R Markdown</a>
analysis was created with <a
  href="https://github.com/workflowr/workflowr">workflowr</a> (version
1.7.1). The <em>Checks</em> tab describes the reproducibility checks
that were applied when the results were created. The <em>Past
versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date
</a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git
repository, you know the exact version of the code that produced these
results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the
global environment can affect the analysis in your R Markdown file in
unknown ways. For reproduciblity it’s best to always run the code in an
empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20200712code">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Seed:</strong>
<code>set.seed(20200712)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20200712code"
class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20200712)</code> was run prior to running
the code in the R Markdown file. Setting a seed ensures that any results
that rely on randomness, e.g. subsampling or permutations, are
reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Session information:</strong>
recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package
versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be
confident that you successfully produced the results during this
run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr
project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomdavetangmusetree1fa9528fb1a905b623649fff33f384aa8d0c7586targetblank1fa9528a">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Repository version:</strong>
<a href="https://github.com/davetang/muse/tree/1fa9528fb1a905b623649fff33f384aa8d0c7586" target="_blank">1fa9528</a>
</a>
</p>
</div>
<div
id="strongRepositoryversionstrongahrefhttpsgithubcomdavetangmusetree1fa9528fb1a905b623649fff33f384aa8d0c7586targetblank1fa9528a"
class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development
and connecting the code version to the results is critical for
reproducibility.
</p>
<p>
The results in this page were generated with repository version
<a href="https://github.com/davetang/muse/tree/1fa9528fb1a905b623649fff33f384aa8d0c7586" target="_blank">1fa9528</a>.
See the <em>Past versions</em> tab to see a history of the changes made
to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for
the analysis have been committed to Git prior to generating the results
(you can use <code>wflow_publish</code> or
<code>wflow_git_commit</code>). workflowr only checks the R Markdown
file, but you know if there are other scripts or data files that it
depends on. Below is the status of the Git repository when the results
were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .Rproj.user/
    Ignored:    data/1M_neurons_filtered_gene_bc_matrices_h5.h5
    Ignored:    data/293t/
    Ignored:    data/293t_3t3_filtered_gene_bc_matrices.tar.gz
    Ignored:    data/293t_filtered_gene_bc_matrices.tar.gz
    Ignored:    data/5k_Human_Donor1_PBMC_3p_gem-x_5k_Human_Donor1_PBMC_3p_gem-x_count_sample_filtered_feature_bc_matrix.h5
    Ignored:    data/5k_Human_Donor2_PBMC_3p_gem-x_5k_Human_Donor2_PBMC_3p_gem-x_count_sample_filtered_feature_bc_matrix.h5
    Ignored:    data/5k_Human_Donor3_PBMC_3p_gem-x_5k_Human_Donor3_PBMC_3p_gem-x_count_sample_filtered_feature_bc_matrix.h5
    Ignored:    data/5k_Human_Donor4_PBMC_3p_gem-x_5k_Human_Donor4_PBMC_3p_gem-x_count_sample_filtered_feature_bc_matrix.h5
    Ignored:    data/Parent_SC3v3_Human_Glioblastoma_filtered_feature_bc_matrix.tar.gz
    Ignored:    data/brain_counts/
    Ignored:    data/cl.obo
    Ignored:    data/cl.owl
    Ignored:    data/jurkat/
    Ignored:    data/jurkat:293t_50:50_filtered_gene_bc_matrices.tar.gz
    Ignored:    data/jurkat_293t/
    Ignored:    data/jurkat_filtered_gene_bc_matrices.tar.gz
    Ignored:    data/pbmc20k/
    Ignored:    data/pbmc20k_seurat/
    Ignored:    data/pbmc3k/
    Ignored:    data/pbmc4k_filtered_gene_bc_matrices.tar.gz
    Ignored:    data/pbmc_1k_v3_raw_feature_bc_matrix.h5
    Ignored:    data/refdata-gex-GRCh38-2020-A.tar.gz
    Ignored:    data/seurat_1m_neuron.rds
    Ignored:    data/t_3k_filtered_gene_bc_matrices.tar.gz
    Ignored:    r_packages_4.4.1/

Untracked files:
    Untracked:  analysis/bioc_scrnaseq.Rmd
    Untracked:  data/97516b79-8d08-46a6-b329-5d0a25b0be98.h5ad
    Untracked:  data/pbmc_1k_v3_filtered_feature_bc_matrix.h5

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not
included in this status report because it is ok for generated content to
have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">
<p>
These are the previous versions of the repository in which changes were
made to the R Markdown (<code>analysis/cluster_profiler.Rmd</code>) and
HTML (<code>docs/cluster_profiler.html</code>) files. If you’ve
configured a remote Git repository (see <code>?wflow_git_remote</code>),
click on the hyperlinks in the table below to view the files as they
were in that past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/davetang/muse/blob/1fa9528fb1a905b623649fff33f384aa8d0c7586/analysis/cluster_profiler.Rmd" target="_blank">1fa9528</a>
</td>
<td>
Dave Tang
</td>
<td>
2025-03-06
</td>
<td>
Universal enrichment analyser
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/davetang/muse/2c080dce2c4547d90fe19cf649b4683a4aaa5142/docs/cluster_profiler.html" target="_blank">2c080dc</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-11-01
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/davetang/muse/blob/124a8d99ab0e6bfd8b2d0cedc7eeeeea6d822f3e/analysis/cluster_profiler.Rmd" target="_blank">124a8d9</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-11-01
</td>
<td>
Using clusterProfiler
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<div id="getting-started" class="section level2">
<h2>Getting started</h2>
<p>Install.</p>
<pre class="r"><code>if (!require(&quot;BiocManager&quot;, quietly = TRUE))
  install.packages(&quot;BiocManager&quot;)

install.packages(&#39;ggarchery&#39;)
install.packages(&#39;ggtangle&#39;)
BiocManager::install(&quot;clusterProfiler&quot;)
BiocManager::install(&quot;org.Hs.eg.db&quot;)
BiocManager::install(&quot;GO.db&quot;)</code></pre>
<p>Load libraries.</p>
<pre class="r"><code>suppressPackageStartupMessages(library(tidyverse))
suppressPackageStartupMessages(library(ggarchery))
suppressPackageStartupMessages(library(clusterProfiler))
suppressPackageStartupMessages(library(org.Hs.eg.db))
suppressPackageStartupMessages(library(GO.db))</code></pre>
</div>
<div id="example-data" class="section level2">
<h2>Example data</h2>
<p>Use <a
href="https://davetang.org/muse/2024/10/31/an-example-differential-gene-expression-results-table/">An
example differential gene expression results table</a>.</p>
<pre class="r"><code>de_res &lt;- read_csv(&quot;https://raw.githubusercontent.com/davetang/muse/refs/heads/main/data/13970886_edger_res.csv&quot;, show_col_types = FALSE)
head(de_res)</code></pre>
<pre><code># A tibble: 6 × 6
  ensembl_gene_id  logFC logCPM      F  PValue adjusted_pvalue
  &lt;chr&gt;            &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt;   &lt;dbl&gt;           &lt;dbl&gt;
1 ENSG00000000003  2.73   4.83   4.28  0.0684           0.109 
2 ENSG00000000005 -7.00   0.541 17.6   0.00216          0.0138
3 ENSG00000000419  0.120  5.34   0.114 0.743            0.776 
4 ENSG00000000457 -0.708  5.31   3.35  0.0993           0.145 
5 ENSG00000000460 -0.897  3.95   2.66  0.136            0.186 
6 ENSG00000000938  1.54   5.60   1.86  0.205            0.258 </code></pre>
</div>
<div id="gene-ids" class="section level2">
<h2>Gene IDs</h2>
<p>The {clusterProfiler} package uses the <code>enrichGO()</code>
function for performing a <a
href="https://yulab-smu.top/biomedical-knowledge-mining-book/clusterprofiler-go.html#clusterprofiler-go-ora">Gene
Ontology over-representation test</a>. The input for <code>gene</code>
is a vector of Entrez Gene IDs.</p>
<p>Our example data uses Ensembl gene IDs, so we need to convert these
into Entrez Gene IDs. We can use the {org.Hs.eg.db} package provided by
<a
href="https://bioconductor.org/packages/release/data/annotation/html/org.Hs.eg.db.html">Bioconductor</a>,
which provides genome wide annotation for human, primarily based on
mapping using Entrez Gene identifiers.</p>
<pre class="r"><code>ensembl_to_entrez &lt;- AnnotationDbi::select(
  org.Hs.eg.db,
  keys = de_res$ensembl_gene_id,
  columns = c(&quot;ENSEMBL&quot;, &quot;ENTREZID&quot;), 
  keytype = &quot;ENSEMBL&quot;
)</code></pre>
<pre><code>&#39;select()&#39; returned 1:many mapping between keys and columns</code></pre>
<pre class="r"><code>de_res |&gt;
  dplyr::inner_join(ensembl_to_entrez, by = dplyr::join_by(ensembl_gene_id == ENSEMBL)) |&gt;
  dplyr::select(ensembl_gene_id, ENTREZID, dplyr::everything()) -&gt; de_res

head(de_res)</code></pre>
<pre><code># A tibble: 6 × 7
  ensembl_gene_id ENTREZID  logFC logCPM      F  PValue adjusted_pvalue
  &lt;chr&gt;           &lt;chr&gt;     &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt;   &lt;dbl&gt;           &lt;dbl&gt;
1 ENSG00000000003 7105      2.73   4.83   4.28  0.0684           0.109 
2 ENSG00000000005 64102    -7.00   0.541 17.6   0.00216          0.0138
3 ENSG00000000419 8813      0.120  5.34   0.114 0.743            0.776 
4 ENSG00000000457 57147    -0.708  5.31   3.35  0.0993           0.145 
5 ENSG00000000460 55732    -0.897  3.95   2.66  0.136            0.186 
6 ENSG00000000938 2268      1.54   5.60   1.86  0.205            0.258 </code></pre>
</div>
<div id="clusterprofiler" class="section level2">
<h2>clusterProfiler</h2>
<p>The example data contains results of a comparison between cancer
samples and normal samples; positive fold change indicates that genes
were expressed higher in cancer. There are a lot of genes up-regulated
in cancer.</p>
<pre class="r"><code>de_res |&gt;
  dplyr::filter(logFC &gt; 0, adjusted_pvalue &lt; 0.05) |&gt;
  nrow()</code></pre>
<pre><code>[1] 1440</code></pre>
<p>Instead of examining the list of up-regulated genes individually,
which would take a long time, we can see whether these genes have common
Gene Ontology terms more often than expected.</p>
<p>First we’ll get the top 500 most significantly up-regulated
genes.</p>
<pre class="r"><code>de_res |&gt;
  dplyr::filter(logFC &gt; 0, adjusted_pvalue &lt; 0.05) |&gt;
  dplyr::slice_min(order_by = adjusted_pvalue, n = 500) |&gt;
  dplyr::filter(!is.na(ENTREZID)) |&gt;
  dplyr::pull(ENTREZID) -&gt; sig_up_genes</code></pre>
<p>Create the background, i.e., the universe, so we can test for
over-representation.</p>
<pre class="r"><code>de_res |&gt;
  dplyr::filter(!is.na(ENTREZID)) |&gt;
  dplyr::pull(ENTREZID) -&gt; the_universe</code></pre>
<p>Gene Ontology terms are grouped into three ontologies:</p>
<ol style="list-style-type: decimal">
<li>Molecular Function (MF)</li>
<li>Biological Process (BP), and</li>
<li>Cellular Component (CC)</li>
</ol>
<p>Here, we’ll perform an over-representation analysis using biological
processes.</p>
<pre class="r"><code>ego_bp &lt;- enrichGO(
  gene          = sig_up_genes,
  universe      = the_universe,
  OrgDb         = org.Hs.eg.db,
  ont           = &quot;BP&quot;,
  pAdjustMethod = &quot;BH&quot;,
  pvalueCutoff  = 0.01,
  qvalueCutoff  = 0.05,
  minGSSize     = 10,
  maxGSSize     = 500,
  readable      = TRUE
)

head(ego_bp)</code></pre>
<pre><code>                   ID
GO:0016064 GO:0016064
GO:0019724 GO:0019724
GO:0002460 GO:0002460
GO:0002449 GO:0002449
GO:0002443 GO:0002443
GO:0006959 GO:0006959
                                                                                                                         Description
GO:0016064                                                                                   immunoglobulin mediated immune response
GO:0019724                                                                                                  B cell mediated immunity
GO:0002460 adaptive immune response based on somatic recombination of immune receptors built from immunoglobulin superfamily domains
GO:0002449                                                                                              lymphocyte mediated immunity
GO:0002443                                                                                               leukocyte mediated immunity
GO:0006959                                                                                                   humoral immune response
           GeneRatio   BgRatio RichFactor FoldEnrichment    zScore       pvalue
GO:0016064    30/385 196/17518 0.15306122       6.964484 12.587673 4.569415e-17
GO:0019724    30/385 199/17518 0.15075377       6.859492 12.461452 7.046005e-17
GO:0002460    37/385 382/17518 0.09685864       4.407194 10.092931 3.590714e-14
GO:0002449    35/385 368/17518 0.09510870       4.327569  9.670801 3.134308e-13
GO:0002443    38/385 466/17518 0.08154506       3.710406  8.889621 3.440135e-12
GO:0006959    30/385 299/17518 0.10033445       4.565348  9.321312 4.278292e-12
               p.adjust       qvalue
GO:0016064 1.319717e-13 1.180762e-13
GO:0019724 1.319717e-13 1.180762e-13
GO:0002460 4.483604e-11 4.011520e-11
GO:0002449 2.935280e-10 2.626220e-10
GO:0002443 2.577349e-09 2.305977e-09
GO:0006959 2.671080e-09 2.389839e-09
                                                                                                                                                                                                                                                                     geneID
GO:0016064                                               IGLC1/IGHG4/IGHG2/IGHG3/IGHV5-51/IGLL5/IGHV4-39/IGLC3/IGHV3-23/IGHV2-5/CD27/IGHV3-30/IGHV3-48/BATF/IGHM/IGHG1/IGLC2/IGHV4-61/IGHA2/IGHV3-7/IGHV3-21/IGHV3-33/IGHV4-59/C4A/IGHV1-69D/FOXP3/IGKC/IGHV3-15/TREM2/CD28
GO:0019724                                               IGLC1/IGHG4/IGHG2/IGHG3/IGHV5-51/IGLL5/IGHV4-39/IGLC3/IGHV3-23/IGHV2-5/CD27/IGHV3-30/IGHV3-48/BATF/IGHM/IGHG1/IGLC2/IGHV4-61/IGHA2/IGHV3-7/IGHV3-21/IGHV3-33/IGHV4-59/C4A/IGHV1-69D/FOXP3/IGKC/IGHV3-15/TREM2/CD28
GO:0002460   IGLC1/IGHG4/IGHG2/IGHG3/IGHV5-51/IGLL5/IGHV4-39/IGLC3/IL4I1/IRF4/JAK3/IGHV3-23/IGHV2-5/CD27/IGHV3-30/CCL19/IGHV3-48/CXCL13/LILRB4/BATF/IGHM/IGHG1/IGLC2/IGHV4-61/IGHA2/IGHV3-7/IGHV3-21/IGHV3-33/IGHV4-59/C4A/IGHV1-69D/FOXP3/IGKC/IGHV3-15/PLA2G4A/TREM2/CD28
GO:0002449                IGLC1/IGHG4/IGHG2/IGHG3/IGHV5-51/IGLL5/IGHV4-39/IGLC3/IL4I1/IGHV3-23/IGHV2-5/CD27/IGHV3-30/SLAMF7/IGHV3-48/LILRB4/BATF/IGHM/IGHG1/IGLC2/IGHV4-61/IGHA2/IGHV3-7/IGHV3-21/IGHV3-33/IGHV4-59/C4A/IGHV1-69D/CD2/FOXP3/LGALS9/IGKC/IGHV3-15/TREM2/CD28
GO:0002443 IGLC1/IGHG4/IGHG2/IGHG3/IGHV5-51/IGLL5/IGHV4-39/IGLC3/IL4I1/JAK3/IGHV3-23/IGHV2-5/CD27/IGHV3-30/SLAMF7/IGHV3-48/LILRB4/BATF/IGHM/IGHG1/IGLC2/IGHV4-61/IGHA2/IGHV3-7/IGHV3-21/IGHV3-33/IGHV4-59/C4A/CCL3/IGHV1-69D/CD84/CD2/FOXP3/LGALS9/IGKC/IGHV3-15/TREM2/CD28
GO:0006959                                                                             WFDC2/IGHG4/IGHG2/IGHG3/IGKV3-20/CXCL14/POU2AF1/MMP7/CCL22/BPIFA1/CXCL9/CCL19/CXCL13/CXCL1/LTF/CXCL10/MS4A1/IGHM/IGHG1/POU2F2/CCL7/TNFRSF21/ADM/IGHA2/C4A/CCL3/CXCL8/CCL8/TREM2/CD28
           Count
GO:0016064    30
GO:0019724    30
GO:0002460    37
GO:0002449    35
GO:0002443    38
GO:0006959    30</code></pre>
<p>Bar plot showing each enriched GO term coloured by the adjusted
p-value.</p>
<pre class="r"><code>barplot(ego_bp, showCategory=10)</code></pre>
<p><img src="figure/cluster_profiler.Rmd/barplot-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-barplot-1">
Past versions of barplot-1.png
</button>
</p>
<div id="fig-barplot-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/davetang/muse/blob/2c080dce2c4547d90fe19cf649b4683a4aaa5142/docs/figure/cluster_profiler.Rmd/barplot-1.png" target="_blank">2c080dc</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-11-01
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>Dot plot showing each enriched GO term with associated
statistics.</p>
<pre class="r"><code>dotplot(ego_bp, showCategory=10)</code></pre>
<p><img src="figure/cluster_profiler.Rmd/dotplot-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-dotplot-1">
Past versions of dotplot-1.png
</button>
</p>
<div id="fig-dotplot-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/davetang/muse/blob/2c080dce2c4547d90fe19cf649b4683a4aaa5142/docs/figure/cluster_profiler.Rmd/dotplot-1.png" target="_blank">2c080dc</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-11-01
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>Heat plot showing the enriched GO terms on the y-axis and the genes
on the x-axis. Genes with the associated GO term are highlighted.</p>
<pre class="r"><code>heatplot(ego_bp, showCategory=10)</code></pre>
<p><img src="figure/cluster_profiler.Rmd/heatplot-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-heatplot-1">
Past versions of heatplot-1.png
</button>
</p>
<div id="fig-heatplot-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/davetang/muse/blob/2c080dce2c4547d90fe19cf649b4683a4aaa5142/docs/figure/cluster_profiler.Rmd/heatplot-1.png" target="_blank">2c080dc</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-11-01
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p><code>goplot</code> shows the gene ontology graph with the enriched
GO terms highlighted.</p>
<pre class="r"><code>goplot(ego_bp)</code></pre>
<pre><code>Warning: ggrepel: 1 unlabeled data points (too many overlaps). Consider
increasing max.overlaps</code></pre>
<p><img src="figure/cluster_profiler.Rmd/goplot-1.png" width="1152" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-goplot-1">
Past versions of goplot-1.png
</button>
</p>
<div id="fig-goplot-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/davetang/muse/blob/2c080dce2c4547d90fe19cf649b4683a4aaa5142/docs/figure/cluster_profiler.Rmd/goplot-1.png" target="_blank">2c080dc</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-11-01
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>Another nice feature of {clusterProfiler} is that you can plot
multiple gene lists together. We can create a list of down-regulated
genes.</p>
<pre class="r"><code>de_res |&gt;
  dplyr::filter(logFC &lt; 0, adjusted_pvalue &lt; 0.05) |&gt;
  dplyr::slice_min(order_by = adjusted_pvalue, n = 500) |&gt;
  dplyr::filter(!is.na(ENTREZID)) |&gt;
  dplyr::pull(ENTREZID) -&gt; sig_down_genes</code></pre>
<p>Perform GO enrichment on two gene lists.</p>
<pre class="r"><code>my_gene_list &lt;- list(
  up_gene = sig_up_genes,
  down_gene = sig_down_genes
)

ego_bp_both &lt;- compareCluster(
  geneCluster = my_gene_list,
  fun = &quot;enrichGO&quot;,
  universe = the_universe,
  OrgDb = org.Hs.eg.db,
  keyType = &quot;ENTREZID&quot;,
  ont = &quot;BP&quot;,
  pvalueCutoff = 0.01,
  pAdjustMethod = &quot;BH&quot;,
  qvalueCutoff = 0.05,
  minGSSize = 10,
  maxGSSize = 500,
  readable = TRUE
)

head(as.data.frame(ego_bp_both))</code></pre>
<pre><code>  Cluster         ID
1 up_gene GO:0016064
2 up_gene GO:0019724
3 up_gene GO:0002460
4 up_gene GO:0002449
5 up_gene GO:0002443
6 up_gene GO:0006959
                                                                                                                Description
1                                                                                   immunoglobulin mediated immune response
2                                                                                                  B cell mediated immunity
3 adaptive immune response based on somatic recombination of immune receptors built from immunoglobulin superfamily domains
4                                                                                              lymphocyte mediated immunity
5                                                                                               leukocyte mediated immunity
6                                                                                                   humoral immune response
  GeneRatio   BgRatio RichFactor FoldEnrichment    zScore       pvalue
1    30/385 196/17518 0.15306122       6.964484 12.587673 4.569415e-17
2    30/385 199/17518 0.15075377       6.859492 12.461452 7.046005e-17
3    37/385 382/17518 0.09685864       4.407194 10.092931 3.590714e-14
4    35/385 368/17518 0.09510870       4.327569  9.670801 3.134308e-13
5    38/385 466/17518 0.08154506       3.710406  8.889621 3.440135e-12
6    30/385 299/17518 0.10033445       4.565348  9.321312 4.278292e-12
      p.adjust       qvalue
1 1.319717e-13 1.180762e-13
2 1.319717e-13 1.180762e-13
3 4.483604e-11 4.011520e-11
4 2.935280e-10 2.626220e-10
5 2.577349e-09 2.305977e-09
6 2.671080e-09 2.389839e-09
                                                                                                                                                                                                                                                            geneID
1                                               IGLC1/IGHG4/IGHG2/IGHG3/IGHV5-51/IGLL5/IGHV4-39/IGLC3/IGHV3-23/IGHV2-5/CD27/IGHV3-30/IGHV3-48/BATF/IGHM/IGHG1/IGLC2/IGHV4-61/IGHA2/IGHV3-7/IGHV3-21/IGHV3-33/IGHV4-59/C4A/IGHV1-69D/FOXP3/IGKC/IGHV3-15/TREM2/CD28
2                                               IGLC1/IGHG4/IGHG2/IGHG3/IGHV5-51/IGLL5/IGHV4-39/IGLC3/IGHV3-23/IGHV2-5/CD27/IGHV3-30/IGHV3-48/BATF/IGHM/IGHG1/IGLC2/IGHV4-61/IGHA2/IGHV3-7/IGHV3-21/IGHV3-33/IGHV4-59/C4A/IGHV1-69D/FOXP3/IGKC/IGHV3-15/TREM2/CD28
3   IGLC1/IGHG4/IGHG2/IGHG3/IGHV5-51/IGLL5/IGHV4-39/IGLC3/IL4I1/IRF4/JAK3/IGHV3-23/IGHV2-5/CD27/IGHV3-30/CCL19/IGHV3-48/CXCL13/LILRB4/BATF/IGHM/IGHG1/IGLC2/IGHV4-61/IGHA2/IGHV3-7/IGHV3-21/IGHV3-33/IGHV4-59/C4A/IGHV1-69D/FOXP3/IGKC/IGHV3-15/PLA2G4A/TREM2/CD28
4                IGLC1/IGHG4/IGHG2/IGHG3/IGHV5-51/IGLL5/IGHV4-39/IGLC3/IL4I1/IGHV3-23/IGHV2-5/CD27/IGHV3-30/SLAMF7/IGHV3-48/LILRB4/BATF/IGHM/IGHG1/IGLC2/IGHV4-61/IGHA2/IGHV3-7/IGHV3-21/IGHV3-33/IGHV4-59/C4A/IGHV1-69D/CD2/FOXP3/LGALS9/IGKC/IGHV3-15/TREM2/CD28
5 IGLC1/IGHG4/IGHG2/IGHG3/IGHV5-51/IGLL5/IGHV4-39/IGLC3/IL4I1/JAK3/IGHV3-23/IGHV2-5/CD27/IGHV3-30/SLAMF7/IGHV3-48/LILRB4/BATF/IGHM/IGHG1/IGLC2/IGHV4-61/IGHA2/IGHV3-7/IGHV3-21/IGHV3-33/IGHV4-59/C4A/CCL3/IGHV1-69D/CD84/CD2/FOXP3/LGALS9/IGKC/IGHV3-15/TREM2/CD28
6                                                                             WFDC2/IGHG4/IGHG2/IGHG3/IGKV3-20/CXCL14/POU2AF1/MMP7/CCL22/BPIFA1/CXCL9/CCL19/CXCL13/CXCL1/LTF/CXCL10/MS4A1/IGHM/IGHG1/POU2F2/CCL7/TNFRSF21/ADM/IGHA2/C4A/CCL3/CXCL8/CCL8/TREM2/CD28
  Count
1    30
2    30
3    37
4    35
5    38
6    30</code></pre>
<p>Dot plot with enriched GO terms by gene list.</p>
<pre class="r"><code>dotplot(ego_bp_both, showCategory = 15)</code></pre>
<p><img src="figure/cluster_profiler.Rmd/dotplot_gene_list-1.png" width="1152" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-dotplot_gene_list-1">
Past versions of dotplot_gene_list-1.png
</button>
</p>
<div id="fig-dotplot_gene_list-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/davetang/muse/blob/2c080dce2c4547d90fe19cf649b4683a4aaa5142/docs/figure/cluster_profiler.Rmd/dotplot_gene_list-1.png" target="_blank">2c080dc</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-11-01
</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<div id="enricher" class="section level2">
<h2>enricher</h2>
<p><code>enricher()</code> is a universal enrichment analyzer.</p>
<ul>
<li>gene - a vector of gene id</li>
<li>universe - background genes. If missing, the all genes listed in the
database (eg TERM2GENE table) will be used as background.</li>
<li>minGSSize - minimal size of genes annotated for testing</li>
<li>maxGSSize - maximal size of genes annotated for testing</li>
<li>TERM2GENE - user input annotation of TERM TO GENE mapping, a
data.frame of 2 column with term and gene. Only used when gson is
NULL.</li>
<li>TERM2NAME - user input of TERM TO NAME mapping, a data.frame of 2
column with term and name. Only used when gson is NULL.</li>
</ul>
<p><a href="https://support.bioconductor.org/p/9144621/">TERM2NAME</a>
is needed to map the GOID to its term (description).</p>
<p>An easy way of obtaining this mapping is by extracting the
information from the Bioconductor GO annotation database package
({GO.db}).</p>
<pre class="r"><code># extract a named vector of all terms
goterms &lt;- Term(GOTERM)

#convert into a data frame
term2name &lt;- data.frame(
  &quot;term&quot;=names(goterms),
  &quot;name&quot;=goterms
)

dim(term2name)</code></pre>
<pre><code>[1] 40940     2</code></pre>
<pre class="r"><code>head(term2name)</code></pre>
<pre><code>                 term                                                     name
GO:0000001 GO:0000001                                mitochondrion inheritance
GO:0000002 GO:0000002                         mitochondrial genome maintenance
GO:0000006 GO:0000006    high-affinity zinc transmembrane transporter activity
GO:0000007 GO:0000007 low-affinity zinc ion transmembrane transporter activity
GO:0000009 GO:0000009                   alpha-1,6-mannosyltransferase activity
GO:0000010 GO:0000010                heptaprenyl diphosphate synthase activity</code></pre>
<p><code>term2name</code> can then be used when calling
<code>enrichr()</code>, by specifying <code>TERM2NAME=term2name</code>.
Be sure, though, to check that <code>term2name</code> contains all GOIDs
present in your TERM2GENE mapping.</p>
<p>Demo input.</p>
<pre class="r"><code>demo &lt;- readr::read_csv(&quot;data/nfurzeri_gene_id_to_go_id.csv.gz&quot;, show_col_types = FALSE)
head(demo)</code></pre>
<pre><code># A tibble: 6 × 2
  ensembl_gene_id    go_id     
  &lt;chr&gt;              &lt;chr&gt;     
1 ENSNFUG00015000040 &lt;NA&gt;      
2 ENSNFUG00015000041 GO:0007156
3 ENSNFUG00015000041 GO:0005886
4 ENSNFUG00015000041 GO:0005737
5 ENSNFUG00015000041 GO:0050808
6 ENSNFUG00015000041 GO:0007411</code></pre>
<p>Prepare data frame of 2 column with term and gene.</p>
<pre class="r"><code>demo |&gt;
  dplyr::filter(!is.na(go_id)) |&gt;
  dplyr::rename(term = go_id, gene = ensembl_gene_id) |&gt;
  dplyr::select(term, gene) |&gt;
  dplyr::arrange(term) -&gt; term2gene

head(term2gene)</code></pre>
<pre><code># A tibble: 6 × 2
  term       gene              
  &lt;chr&gt;      &lt;chr&gt;             
1 GO:0000002 ENSNFUG00015006992
2 GO:0000002 ENSNFUG00015020070
3 GO:0000002 ENSNFUG00015013825
4 GO:0000002 ENSNFUG00015019642
5 GO:0000009 ENSNFUG00015008531
6 GO:0000012 ENSNFUG00015012664</code></pre>
<p>Test with random genes.</p>
<pre class="r"><code>set.seed(1984)
my_genes &lt;- sample(x = unique(term2gene$gene), size = 100, replace = FALSE)

res &lt;- enricher(
  gene = my_genes,
  TERM2GENE = term2gene,
  TERM2NAME = term2name
)

res</code></pre>
<pre><code>#
# over-representation test
#
#...@organism    UNKNOWN 
#...@ontology    UNKNOWN 
#...@gene    chr [1:100] &quot;ENSNFUG00015003803&quot; &quot;ENSNFUG00015000174&quot; &quot;ENSNFUG00015015379&quot; ...
#...pvalues adjusted by &#39;BH&#39; with cutoff &lt;0.05 
#...0 enriched terms found
#...Citation
S Xu, E Hu, Y Cai, Z Xie, X Luo, L Zhan, W Tang, Q Wang, B Liu, R Wang, W Xie, T Wu, L Xie, G Yu. Using clusterProfiler to characterize multiomics data. Nature Protocols. 2024, 19(11):3292-3320 </code></pre>
<p>Test with genes associated with same term.</p>
<pre class="r"><code>term2gene |&gt;
  dplyr::group_by(term) |&gt;
  dplyr::summarise(n = n()) |&gt;
  dplyr::arrange(-n) -&gt; dev_null

my_term &lt;- &quot;GO:0006397&quot;
dplyr::filter(term2name, term == my_term)</code></pre>
<pre><code>                 term            name
GO:0006397 GO:0006397 mRNA processing</code></pre>
<pre class="r"><code>term2gene |&gt;
  dplyr::filter(term == my_term) |&gt;
  dplyr::pull(gene) |&gt;
  head(75) -&gt; enriched_genes

set.seed(1984)
my_genes &lt;- sample(x = unique(term2gene$gene), size = 25, replace = FALSE)
my_genes &lt;- union(enriched_genes, my_genes)

res &lt;- enricher(
  gene = my_genes,
  TERM2GENE = term2gene,
  TERM2NAME = term2name
)

res |&gt;
  as.data.frame()</code></pre>
<pre><code>                   ID                                              Description
GO:0006397 GO:0006397                                          mRNA processing
GO:0008380 GO:0008380                                             RNA splicing
GO:0005681 GO:0005681                                     spliceosomal complex
GO:0000398 GO:0000398                           mRNA splicing, via spliceosome
GO:0071011 GO:0071011                                 precatalytic spliceosome
GO:0046540 GO:0046540                             U4/U6 x U5 tri-snRNP complex
GO:0000381 GO:0000381 regulation of alternative mRNA splicing, via spliceosome
GO:0071013 GO:0071013                             catalytic step 2 spliceosome
GO:0005682 GO:0005682                                                 U5 snRNP
GO:1990904 GO:1990904                                ribonucleoprotein complex
GO:0016607 GO:0016607                                            nuclear speck
GO:0005688 GO:0005688                                                 U6 snRNP
GO:0003729 GO:0003729                                             mRNA binding
GO:0005685 GO:0005685                                                 U1 snRNP
GO:0000387 GO:0000387                              spliceosomal snRNP assembly
GO:0043484 GO:0043484                               regulation of RNA splicing
GO:0004527 GO:0004527                                     exonuclease activity
GO:0005686 GO:0005686                                                 U2 snRNP
GO:0003730 GO:0003730                                      mRNA 3&#39;-UTR binding
GO:0005654 GO:0005654                                              nucleoplasm
GO:0000932 GO:0000932                                                   P-body
GO:0080090 GO:0080090                  regulation of primary metabolic process
GO:0030968 GO:0030968          endoplasmic reticulum unfolded protein response
GO:0005687 GO:0005687                                                 U4 snRNP
GO:0004521 GO:0004521                                RNA endonuclease activity
           GeneRatio   BgRatio RichFactor FoldEnrichment    zScore
GO:0006397    75/100 125/18585 0.60000000     111.510000 91.180214
GO:0008380    43/100  89/18585 0.48314607      89.792697 61.757998
GO:0005681    22/100  54/18585 0.40740741      75.716667 40.441272
GO:0000398    21/100  71/18585 0.29577465      54.969718 33.511130
GO:0071011     7/100  16/18585 0.43750000      81.309375 23.636973
GO:0046540     9/100  43/18585 0.20930233      38.898837 18.299604
GO:0000381     7/100  20/18585 0.35000000      65.047500 21.078008
GO:0071013     7/100  20/18585 0.35000000      65.047500 21.078008
GO:0005682     5/100  19/18585 0.26315789      48.907895 15.366841
GO:1990904    10/100 228/18585 0.04385965       8.151316  7.991207
GO:0016607     5/100  32/18585 0.15625000      29.039062 11.675924
GO:0005688     4/100  15/18585 0.26666667      49.560000 13.838156
GO:0003729     7/100 104/18585 0.06730769      12.509135  8.656782
GO:0005685     4/100  26/18585 0.15384615      28.592308 10.355171
GO:0000387     3/100  12/18585 0.25000000      46.462500 11.586780
GO:0043484     3/100  14/18585 0.21428571      39.825000 10.688523
GO:0004527     3/100  19/18585 0.15789474      29.344737  9.091801
GO:0005686     3/100  30/18585 0.10000000      18.585000  7.089774
GO:0003730     3/100  31/18585 0.09677419      17.985484  6.961453
GO:0005654     5/100 129/18585 0.03875969       7.203488  5.200222
GO:0000932     3/100  35/18585 0.08571429      15.930000  6.502521
GO:0080090     2/100  12/18585 0.16666667      30.975000  7.639565
GO:0030968     2/100  14/18585 0.14285714      26.550000  7.033916
GO:0005687     2/100  15/18585 0.13333333      24.780000  6.776593
GO:0004521     2/100  17/18585 0.11764706      21.864706  6.330150
                  pvalue      p.adjust        qvalue
GO:0006397 1.059713e-152 1.653153e-150 1.372050e-150
GO:0008380  2.787322e-77  2.174111e-75  1.804424e-75
GO:0005681  6.857863e-37  3.566089e-35  2.959709e-35
GO:0000398  1.048636e-31  4.089682e-30  3.394270e-30
GO:0071011  1.159786e-12  3.298732e-11  2.737814e-11
GO:0046540  1.268743e-12  3.298732e-11  2.737814e-11
GO:0000381  7.722209e-12  1.505831e-10  1.249779e-10
GO:0071013  7.722209e-12  1.505831e-10  1.249779e-10
GO:0005682  4.465615e-08  7.740399e-07  6.424217e-07
GO:1990904  4.185349e-07  6.529145e-06  5.418926e-06
GO:0016607  7.316677e-07  1.037638e-05  8.611974e-06
GO:0005688  1.029158e-06  1.337905e-05  1.110407e-05
GO:0003729  1.464263e-06  1.757115e-05  1.458335e-05
GO:0005685  1.077073e-05  1.200167e-04  9.960900e-05
GO:0000387  3.210311e-05  3.338724e-04  2.771006e-04
GO:0043484  5.270183e-05  5.138429e-04  4.264688e-04
GO:0004527  1.375797e-04  1.262496e-03  1.047821e-03
GO:0005686  5.522099e-04  4.785820e-03  3.972036e-03
GO:0003730  6.089961e-04  5.000179e-03  4.149946e-03
GO:0005654  6.625769e-04  5.168100e-03  4.289314e-03
GO:0000932  8.730213e-04  6.485301e-03  5.382538e-03
GO:0080090  1.826457e-03  1.295124e-02  1.074900e-02
GO:0030968  2.500695e-03  1.696123e-02  1.407714e-02
GO:0005687  2.875322e-03  1.868959e-02  1.551160e-02
GO:0004521  3.698227e-03  2.307694e-02  1.915293e-02
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     geneID
GO:0006397 ENSNFUG00015000047/ENSNFUG00015000176/ENSNFUG00015000433/ENSNFUG00015000717/ENSNFUG00015000995/ENSNFUG00015002878/ENSNFUG00015003278/ENSNFUG00015003412/ENSNFUG00015003645/ENSNFUG00015006374/ENSNFUG00015008466/ENSNFUG00015010592/ENSNFUG00015011476/ENSNFUG00015011629/ENSNFUG00015012572/ENSNFUG00015015893/ENSNFUG00015018395/ENSNFUG00015021027/ENSNFUG00015023111/ENSNFUG00015024532/ENSNFUG00015000276/ENSNFUG00015000347/ENSNFUG00015000470/ENSNFUG00015000633/ENSNFUG00015000679/ENSNFUG00015001197/ENSNFUG00015001330/ENSNFUG00015002687/ENSNFUG00015003553/ENSNFUG00015003613/ENSNFUG00015004037/ENSNFUG00015004391/ENSNFUG00015008722/ENSNFUG00015009124/ENSNFUG00015010820/ENSNFUG00015011580/ENSNFUG00015013273/ENSNFUG00015014123/ENSNFUG00015014399/ENSNFUG00015014448/ENSNFUG00015019426/ENSNFUG00015020527/ENSNFUG00015020545/ENSNFUG00015021436/ENSNFUG00015004299/ENSNFUG00015006219/ENSNFUG00015006538/ENSNFUG00015007623/ENSNFUG00015008859/ENSNFUG00015008871/ENSNFUG00015008954/ENSNFUG00015013147/ENSNFUG00015014506/ENSNFUG00015014956/ENSNFUG00015015816/ENSNFUG00015017783/ENSNFUG00015019674/ENSNFUG00015019903/ENSNFUG00015020693/ENSNFUG00015021220/ENSNFUG00015022748/ENSNFUG00015023101/ENSNFUG00015023143/ENSNFUG00015004762/ENSNFUG00015005084/ENSNFUG00015005095/ENSNFUG00015006742/ENSNFUG00015008749/ENSNFUG00015009426/ENSNFUG00015013451/ENSNFUG00015014267/ENSNFUG00015014760/ENSNFUG00015014763/ENSNFUG00015018080/ENSNFUG00015018929
GO:0008380                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 ENSNFUG00015000047/ENSNFUG00015000433/ENSNFUG00015002878/ENSNFUG00015003412/ENSNFUG00015008466/ENSNFUG00015010592/ENSNFUG00015011629/ENSNFUG00015012572/ENSNFUG00015015893/ENSNFUG00015018395/ENSNFUG00015021027/ENSNFUG00015023111/ENSNFUG00015000276/ENSNFUG00015000347/ENSNFUG00015000470/ENSNFUG00015001197/ENSNFUG00015002687/ENSNFUG00015004391/ENSNFUG00015009124/ENSNFUG00015010820/ENSNFUG00015011580/ENSNFUG00015013273/ENSNFUG00015014123/ENSNFUG00015014448/ENSNFUG00015019426/ENSNFUG00015020527/ENSNFUG00015008859/ENSNFUG00015008871/ENSNFUG00015013147/ENSNFUG00015014506/ENSNFUG00015019674/ENSNFUG00015019903/ENSNFUG00015020693/ENSNFUG00015021220/ENSNFUG00015022748/ENSNFUG00015023101/ENSNFUG00015005084/ENSNFUG00015008749/ENSNFUG00015014267/ENSNFUG00015014760/ENSNFUG00015014763/ENSNFUG00015018080/ENSNFUG00015018929
GO:0005681                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                ENSNFUG00015000433/ENSNFUG00015002878/ENSNFUG00015010592/ENSNFUG00015011629/ENSNFUG00015012572/ENSNFUG00015015893/ENSNFUG00015018395/ENSNFUG00015021027/ENSNFUG00015023111/ENSNFUG00015001197/ENSNFUG00015002687/ENSNFUG00015014123/ENSNFUG00015019426/ENSNFUG00015014506/ENSNFUG00015019903/ENSNFUG00015020693/ENSNFUG00015022748/ENSNFUG00015023101/ENSNFUG00015005084/ENSNFUG00015014760/ENSNFUG00015014763/ENSNFUG00015018929
GO:0000398                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   ENSNFUG00015000433/ENSNFUG00015002878/ENSNFUG00015003412/ENSNFUG00015010592/ENSNFUG00015011629/ENSNFUG00015012572/ENSNFUG00015015893/ENSNFUG00015023111/ENSNFUG00015001197/ENSNFUG00015002687/ENSNFUG00015014123/ENSNFUG00015008871/ENSNFUG00015014506/ENSNFUG00015019903/ENSNFUG00015020693/ENSNFUG00015022748/ENSNFUG00015005084/ENSNFUG00015008749/ENSNFUG00015014267/ENSNFUG00015014760/ENSNFUG00015014763
GO:0071011                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                             ENSNFUG00015000433/ENSNFUG00015011629/ENSNFUG00015018395/ENSNFUG00015014123/ENSNFUG00015019674/ENSNFUG00015005084/ENSNFUG00015014763
GO:0046540                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                       ENSNFUG00015002878/ENSNFUG00015011629/ENSNFUG00015014123/ENSNFUG00015008871/ENSNFUG00015019674/ENSNFUG00015019903/ENSNFUG00015020693/ENSNFUG00015005084/ENSNFUG00015014763
GO:0000381                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                             ENSNFUG00015024532/ENSNFUG00015000470/ENSNFUG00015011580/ENSNFUG00015014123/ENSNFUG00015014448/ENSNFUG00015008859/ENSNFUG00015013147
GO:0071013                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                             ENSNFUG00015010592/ENSNFUG00015018395/ENSNFUG00015001197/ENSNFUG00015004391/ENSNFUG00015019674/ENSNFUG00015005084/ENSNFUG00015014267
GO:0005682                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   ENSNFUG00015018395/ENSNFUG00015001197/ENSNFUG00015008871/ENSNFUG00015019674/ENSNFUG00015020693
GO:1990904                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    ENSNFUG00015000047/ENSNFUG00015018395/ENSNFUG00015024532/ENSNFUG00015008722/ENSNFUG00015010820/ENSNFUG00015014956/ENSNFUG00015019674/ENSNFUG00015021220/ENSNFUG00015005084/ENSNFUG00015014763
GO:0016607                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   ENSNFUG00015003412/ENSNFUG00015006374/ENSNFUG00015000276/ENSNFUG00015009124/ENSNFUG00015019426
GO:0005688                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      ENSNFUG00015002878/ENSNFUG00015011629/ENSNFUG00015019903/ENSNFUG00015005084
GO:0003729                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                             ENSNFUG00015000047/ENSNFUG00015000995/ENSNFUG00015008466/ENSNFUG00015000347/ENSNFUG00015000470/ENSNFUG00015014448/ENSNFUG00015008954
GO:0005685                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      ENSNFUG00015008466/ENSNFUG00015010592/ENSNFUG00015018395/ENSNFUG00015019674
GO:0000387                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         ENSNFUG00015010592/ENSNFUG00015018395/ENSNFUG00015019674
GO:0043484                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         ENSNFUG00015000995/ENSNFUG00015014448/ENSNFUG00015018080
GO:0004527                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         ENSNFUG00015014399/ENSNFUG00015005095/ENSNFUG00015000334
GO:0005686                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         ENSNFUG00015000433/ENSNFUG00015018395/ENSNFUG00015019674
GO:0003730                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         ENSNFUG00015024532/ENSNFUG00015008722/ENSNFUG00015004762
GO:0005654                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   ENSNFUG00015021027/ENSNFUG00015023111/ENSNFUG00015010820/ENSNFUG00015019426/ENSNFUG00015018836
GO:0000932                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         ENSNFUG00015002878/ENSNFUG00015006219/ENSNFUG00015005084
GO:0080090                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            ENSNFUG00015011476/ENSNFUG00015021436
GO:0030968                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            ENSNFUG00015011476/ENSNFUG00015021436
GO:0005687                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            ENSNFUG00015018395/ENSNFUG00015014123
GO:0004521                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            ENSNFUG00015011476/ENSNFUG00015021436
           Count
GO:0006397    75
GO:0008380    43
GO:0005681    22
GO:0000398    21
GO:0071011     7
GO:0046540     9
GO:0000381     7
GO:0071013     7
GO:0005682     5
GO:1990904    10
GO:0016607     5
GO:0005688     4
GO:0003729     7
GO:0005685     4
GO:0000387     3
GO:0043484     3
GO:0004527     3
GO:0005686     3
GO:0003730     3
GO:0005654     5
GO:0000932     3
GO:0080090     2
GO:0030968     2
GO:0005687     2
GO:0004521     2</code></pre>
<p>Check some of the results.</p>
<pre class="r"><code>go_term &lt;- &quot;GO:0007156&quot;

dplyr::filter(term2name, term == go_term)</code></pre>
<pre><code>                 term
GO:0007156 GO:0007156
                                                                      name
GO:0007156 homophilic cell adhesion via plasma membrane adhesion molecules</code></pre>
<pre class="r"><code>dplyr::filter(term2gene, gene %in% my_genes, term == go_term) |&gt;
  nrow()</code></pre>
<pre><code>[1] 1</code></pre>
<pre class="r"><code>dplyr::filter(term2gene, term == go_term) |&gt;
  nrow()</code></pre>
<pre><code>[1] 130</code></pre>
<pre class="r"><code>dotplot(res, showCategory=10)</code></pre>
<p><img src="figure/cluster_profiler.Rmd/dotplot_res-1.png" width="768" style="display: block; margin: auto;" /></p>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span>
Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 4.4.1 (2024-06-14)
Platform: x86_64-pc-linux-gnu
Running under: Ubuntu 22.04.5 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 
LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

time zone: Etc/UTC
tzcode source: system (glibc)

attached base packages:
[1] stats4    stats     graphics  grDevices utils     datasets  methods  
[8] base     

other attached packages:
 [1] GO.db_3.20.0           org.Hs.eg.db_3.20.0    AnnotationDbi_1.68.0  
 [4] IRanges_2.40.1         S4Vectors_0.44.0       Biobase_2.66.0        
 [7] BiocGenerics_0.52.0    clusterProfiler_4.14.6 ggarchery_0.4.3       
[10] lubridate_1.9.4        forcats_1.0.0          stringr_1.5.1         
[13] dplyr_1.1.4            purrr_1.0.4            readr_2.1.5           
[16] tidyr_1.3.1            tibble_3.2.1           ggplot2_3.5.1         
[19] tidyverse_2.0.0        workflowr_1.7.1       

loaded via a namespace (and not attached):
  [1] DBI_1.2.3               gson_0.1.0              rlang_1.1.5            
  [4] magrittr_2.0.3          DOSE_4.0.0              git2r_0.35.0           
  [7] compiler_4.4.1          RSQLite_2.3.9           getPass_0.2-4          
 [10] png_0.1-8               callr_3.7.6             vctrs_0.6.5            
 [13] reshape2_1.4.4          pkgconfig_2.0.3         crayon_1.5.3           
 [16] fastmap_1.2.0           XVector_0.46.0          labeling_0.4.3         
 [19] utf8_1.2.4              promises_1.3.2          rmarkdown_2.29         
 [22] tzdb_0.4.0              enrichplot_1.26.6       UCSC.utils_1.2.0       
 [25] ps_1.9.0                bit_4.5.0.1             xfun_0.51              
 [28] zlibbioc_1.52.0         cachem_1.1.0            aplot_0.2.5            
 [31] GenomeInfoDb_1.42.3     jsonlite_1.9.1          blob_1.2.4             
 [34] later_1.4.1             BiocParallel_1.40.0     parallel_4.4.1         
 [37] R6_2.6.1                bslib_0.9.0             stringi_1.8.4          
 [40] RColorBrewer_1.1-3      jquerylib_0.1.4         GOSemSim_2.32.0        
 [43] Rcpp_1.0.14             knitr_1.49              ggtangle_0.0.6         
 [46] R.utils_2.13.0          igraph_2.1.4            httpuv_1.6.15          
 [49] Matrix_1.7-0            splines_4.4.1           timechange_0.3.0       
 [52] tidyselect_1.2.1        qvalue_2.38.0           rstudioapi_0.17.1      
 [55] yaml_2.3.10             codetools_0.2-20        curl_6.2.1             
 [58] processx_3.8.6          lattice_0.22-6          plyr_1.8.9             
 [61] treeio_1.30.0           withr_3.0.2             KEGGREST_1.46.0        
 [64] evaluate_1.0.3          gridGraphics_0.5-1      Biostrings_2.74.1      
 [67] ggtree_3.14.0           pillar_1.10.1           whisker_0.4.1          
 [70] ggfun_0.1.8             generics_0.1.3          vroom_1.6.5            
 [73] rprojroot_2.0.4         hms_1.1.3               tidytree_0.4.6         
 [76] munsell_0.5.1           scales_1.3.0            glue_1.8.0             
 [79] lazyeval_0.2.2          tools_4.4.1             data.table_1.17.0      
 [82] fgsea_1.32.2            fs_1.6.5                fastmatch_1.1-6        
 [85] cowplot_1.1.3           grid_4.4.1              ape_5.8-1              
 [88] colorspace_2.1-1        nlme_3.1-164            patchwork_1.3.0        
 [91] GenomeInfoDbData_1.2.13 cli_3.6.4               gtable_0.3.6           
 [94] R.methodsS3_1.8.2       yulab.utils_0.2.0       sass_0.4.9             
 [97] digest_0.6.37           ggrepel_0.9.6           ggplotify_0.1.2        
[100] farver_2.1.2            memoise_2.0.1           htmltools_0.5.8.1      
[103] R.oo_1.27.0             lifecycle_1.0.4         httr_1.4.7             
[106] bit64_4.6.0-1          </code></pre>
</div>
</div>
</div>
</div>
</div>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
https://docs.mathjax.org/en/latest/web/configuration.html. This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>




</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
