<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />



<meta name="date" content="2025-12-01" />

<title>Getting started with clusterProfiler</title>

<script src="site_libs/header-attrs-2.29/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.13.2/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-6.5.2/css/all.min.css" rel="stylesheet" />
<link href="site_libs/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet" />

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/main/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">My blog</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/davetang/muse">
    <span class="fa fa-github"></span>
     
    Source code
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Getting started with clusterProfiler</h1>
<h4 class="date">2025-12-01</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span>
workflowr <span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2025-12-01
</p>
<p>
<strong>Checks:</strong> <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 7
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong> <code>muse/</code> <span
class="glyphicon glyphicon-question-sign" aria-hidden="true"
title="This is the local directory in which the code in this file was executed.">
</span>
</p>
<p>
This reproducible <a href="https://rmarkdown.rstudio.com">R Markdown</a>
analysis was created with <a
  href="https://github.com/workflowr/workflowr">workflowr</a> (version
1.7.1). The <em>Checks</em> tab describes the reproducibility checks
that were applied when the results were created. The <em>Past
versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date
</a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git
repository, you know the exact version of the code that produced these
results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the
global environment can affect the analysis in your R Markdown file in
unknown ways. For reproduciblity it’s best to always run the code in an
empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20200712code">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Seed:</strong>
<code>set.seed(20200712)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20200712code"
class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20200712)</code> was run prior to running
the code in the R Markdown file. Setting a seed ensures that any results
that rely on randomness, e.g. subsampling or permutations, are
reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Session information:</strong>
recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package
versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be
confident that you successfully produced the results during this
run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr
project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomdavetangmusetree3bc33aabcb24474b28e2c81844b6bc7733f7798etargetblank3bc33aaa">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Repository version:</strong>
<a href="https://github.com/davetang/muse/tree/3bc33aabcb24474b28e2c81844b6bc7733f7798e" target="_blank">3bc33aa</a>
</a>
</p>
</div>
<div
id="strongRepositoryversionstrongahrefhttpsgithubcomdavetangmusetree3bc33aabcb24474b28e2c81844b6bc7733f7798etargetblank3bc33aaa"
class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development
and connecting the code version to the results is critical for
reproducibility.
</p>
<p>
The results in this page were generated with repository version
<a href="https://github.com/davetang/muse/tree/3bc33aabcb24474b28e2c81844b6bc7733f7798e" target="_blank">3bc33aa</a>.
See the <em>Past versions</em> tab to see a history of the changes made
to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for
the analysis have been committed to Git prior to generating the results
(you can use <code>wflow_publish</code> or
<code>wflow_git_commit</code>). workflowr only checks the R Markdown
file, but you know if there are other scripts or data files that it
depends on. Below is the status of the Git repository when the results
were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .Rproj.user/
    Ignored:    data/1M_neurons_filtered_gene_bc_matrices_h5.h5
    Ignored:    data/293t/
    Ignored:    data/293t_3t3_filtered_gene_bc_matrices.tar.gz
    Ignored:    data/293t_filtered_gene_bc_matrices.tar.gz
    Ignored:    data/5k_Human_Donor1_PBMC_3p_gem-x_5k_Human_Donor1_PBMC_3p_gem-x_count_sample_filtered_feature_bc_matrix.h5
    Ignored:    data/5k_Human_Donor2_PBMC_3p_gem-x_5k_Human_Donor2_PBMC_3p_gem-x_count_sample_filtered_feature_bc_matrix.h5
    Ignored:    data/5k_Human_Donor3_PBMC_3p_gem-x_5k_Human_Donor3_PBMC_3p_gem-x_count_sample_filtered_feature_bc_matrix.h5
    Ignored:    data/5k_Human_Donor4_PBMC_3p_gem-x_5k_Human_Donor4_PBMC_3p_gem-x_count_sample_filtered_feature_bc_matrix.h5
    Ignored:    data/97516b79-8d08-46a6-b329-5d0a25b0be98.h5ad
    Ignored:    data/Parent_SC3v3_Human_Glioblastoma_filtered_feature_bc_matrix.tar.gz
    Ignored:    data/brain_counts/
    Ignored:    data/cl.obo
    Ignored:    data/cl.owl
    Ignored:    data/jurkat/
    Ignored:    data/jurkat:293t_50:50_filtered_gene_bc_matrices.tar.gz
    Ignored:    data/jurkat_293t/
    Ignored:    data/jurkat_filtered_gene_bc_matrices.tar.gz
    Ignored:    data/pbmc20k/
    Ignored:    data/pbmc20k_seurat/
    Ignored:    data/pbmc3k.h5ad
    Ignored:    data/pbmc3k/
    Ignored:    data/pbmc3k_bpcells_mat/
    Ignored:    data/pbmc3k_export.mtx
    Ignored:    data/pbmc3k_matrix.mtx
    Ignored:    data/pbmc3k_seurat.rds
    Ignored:    data/pbmc4k_filtered_gene_bc_matrices.tar.gz
    Ignored:    data/pbmc_1k_v3_filtered_feature_bc_matrix.h5
    Ignored:    data/pbmc_1k_v3_raw_feature_bc_matrix.h5
    Ignored:    data/refdata-gex-GRCh38-2020-A.tar.gz
    Ignored:    data/seurat_1m_neuron.rds
    Ignored:    data/t_3k_filtered_gene_bc_matrices.tar.gz
    Ignored:    r_packages_4.4.1/
    Ignored:    r_packages_4.5.0/

Untracked files:
    Untracked:  analysis/bioc.Rmd
    Untracked:  analysis/bioc_scrnaseq.Rmd
    Untracked:  bpcells_matrix/
    Untracked:  data/Caenorhabditis_elegans.WBcel235.113.gtf.gz
    Untracked:  data/GCF_043380555.1-RS_2024_12_gene_ontology.gaf.gz
    Untracked:  data/arab.rds
    Untracked:  data/astronomicalunit.csv
    Untracked:  data/femaleMiceWeights.csv
    Untracked:  data/lung_bcell.rds
    Untracked:  m3/
    Untracked:  women.json

Unstaged changes:
    Modified:   analysis/isoform_switch_analyzer.Rmd

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not
included in this status report because it is ok for generated content to
have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">
<p>
These are the previous versions of the repository in which changes were
made to the R Markdown (<code>analysis/cluster_profiler.Rmd</code>) and
HTML (<code>docs/cluster_profiler.html</code>) files. If you’ve
configured a remote Git repository (see <code>?wflow_git_remote</code>),
click on the hyperlinks in the table below to view the files as they
were in that past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/davetang/muse/blob/3bc33aabcb24474b28e2c81844b6bc7733f7798e/analysis/cluster_profiler.Rmd" target="_blank">3bc33aa</a>
</td>
<td>
Dave Tang
</td>
<td>
2025-12-01
</td>
<td>
Manually calculate RichFactor and FoldEnrichment
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/davetang/muse/18e8e9b7e30957ba8e89e2e3844d045616cb3a86/docs/cluster_profiler.html" target="_blank">18e8e9b</a>
</td>
<td>
Dave Tang
</td>
<td>
2025-05-26
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/davetang/muse/blob/ec4e2515caca03020a6be3dc7914be6231d90208/analysis/cluster_profiler.Rmd" target="_blank">ec4e251</a>
</td>
<td>
Dave Tang
</td>
<td>
2025-05-26
</td>
<td>
Trying to reproduce the numerator
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/davetang/muse/94680bd73fa5fd3b9ff81bae1f6dda032ef2f4d2/docs/cluster_profiler.html" target="_blank">94680bd</a>
</td>
<td>
Dave Tang
</td>
<td>
2025-05-24
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/davetang/muse/blob/bc9f220452162c36b6fde9b7d9d6bb709fe803f6/analysis/cluster_profiler.Rmd" target="_blank">bc9f220</a>
</td>
<td>
Dave Tang
</td>
<td>
2025-05-24
</td>
<td>
Trying to recreate the ratios
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/davetang/muse/0ac82102a0064cf6f70ad9fdec39e3272f0aefe0/docs/cluster_profiler.html" target="_blank">0ac8210</a>
</td>
<td>
Dave Tang
</td>
<td>
2025-05-24
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/davetang/muse/blob/845e8649fe20396618d985ec6f972fcd317b332a/analysis/cluster_profiler.Rmd" target="_blank">845e864</a>
</td>
<td>
Dave Tang
</td>
<td>
2025-05-24
</td>
<td>
Better explanations
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/davetang/muse/c8e80cd43b29fa99ef8a009be39c1b496db2de7f/docs/cluster_profiler.html" target="_blank">c8e80cd</a>
</td>
<td>
Dave Tang
</td>
<td>
2025-05-23
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/davetang/muse/blob/8f0da0060e58f212dd8866784eebf82d06589cce/analysis/cluster_profiler.Rmd" target="_blank">8f0da00</a>
</td>
<td>
Dave Tang
</td>
<td>
2025-05-23
</td>
<td>
Check background length
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/davetang/muse/006e824308df24e7f43c3d664e30d96617250941/docs/cluster_profiler.html" target="_blank">006e824</a>
</td>
<td>
Dave Tang
</td>
<td>
2025-03-09
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/davetang/muse/blob/fc1274d1c95357d2b0b1dbffe8584e8f00296111/analysis/cluster_profiler.Rmd" target="_blank">fc1274d</a>
</td>
<td>
Dave Tang
</td>
<td>
2025-03-09
</td>
<td>
Subset GO terms to only biological processes
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/davetang/muse/f0166b0791270b7f3b51540dc73a64d37947e451/docs/cluster_profiler.html" target="_blank">f0166b0</a>
</td>
<td>
Dave Tang
</td>
<td>
2025-03-09
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/davetang/muse/blob/6ad265795436739eeafc0d41461cfcb4bea982c2/analysis/cluster_profiler.Rmd" target="_blank">6ad2657</a>
</td>
<td>
Dave Tang
</td>
<td>
2025-03-09
</td>
<td>
Checking missing GO terms
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/davetang/muse/e815348c6fc2adfec65202ff3718880a944b088c/docs/cluster_profiler.html" target="_blank">e815348</a>
</td>
<td>
Dave Tang
</td>
<td>
2025-03-06
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/davetang/muse/blob/1fa9528fb1a905b623649fff33f384aa8d0c7586/analysis/cluster_profiler.Rmd" target="_blank">1fa9528</a>
</td>
<td>
Dave Tang
</td>
<td>
2025-03-06
</td>
<td>
Universal enrichment analyser
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/davetang/muse/2c080dce2c4547d90fe19cf649b4683a4aaa5142/docs/cluster_profiler.html" target="_blank">2c080dc</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-11-01
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/davetang/muse/blob/124a8d99ab0e6bfd8b2d0cedc7eeeeea6d822f3e/analysis/cluster_profiler.Rmd" target="_blank">124a8d9</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-11-01
</td>
<td>
Using clusterProfiler
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<p><a
href="https://bioconductor.org/packages/release/bioc/html/clusterProfiler.html">clusterProfiler</a>:</p>
<blockquote>
<p>This package supports functional characteristics of both coding and
non-coding genomics data for thousands of species with up-to-date gene
annotation. It provides a univeral interface for gene functional
annotation from a variety of sources and thus can be applied in diverse
scenarios. It provides a tidy interface to access, manipulate, and
visualize enrichment results to help users achieve efficient data
interpretation. Datasets obtained from multiple treatments and time
points can be analyzed and compared in a single run, easily revealing
functional consensus and differences among distinct conditions.</p>
</blockquote>
<div id="getting-started" class="section level2">
<h2>Getting started</h2>
<p>Install {clusterProfiler} and some dependencies.</p>
<pre class="r"><code>if (!require(&quot;BiocManager&quot;, quietly = TRUE))
  install.packages(&quot;BiocManager&quot;)

install.packages(&#39;ggarchery&#39;)
install.packages(&#39;ggtangle&#39;)
BiocManager::install(&quot;clusterProfiler&quot;)
BiocManager::install(&quot;org.Hs.eg.db&quot;)
BiocManager::install(&quot;GO.db&quot;)</code></pre>
<p>Load required libraries.</p>
<pre class="r"><code>suppressPackageStartupMessages(library(tidyverse))
suppressPackageStartupMessages(library(ggarchery))
suppressPackageStartupMessages(library(clusterProfiler))
suppressPackageStartupMessages(library(org.Hs.eg.db))
suppressPackageStartupMessages(library(GO.db))</code></pre>
</div>
<div id="example-data" class="section level2">
<h2>Example data</h2>
<p>We will use <a
href="https://davetang.org/muse/2024/10/31/an-example-differential-gene-expression-results-table/">An
example differential gene expression results table</a> that contains
{edgeR} results comparing normal versus cancer samples.</p>
<pre class="r"><code>de_res &lt;- read_csv(&quot;https://raw.githubusercontent.com/davetang/muse/refs/heads/main/data/13970886_edger_res.csv&quot;, show_col_types = FALSE)
head(de_res)</code></pre>
<pre><code># A tibble: 6 × 6
  ensembl_gene_id  logFC logCPM      F  PValue adjusted_pvalue
  &lt;chr&gt;            &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt;   &lt;dbl&gt;           &lt;dbl&gt;
1 ENSG00000000003  2.73   4.83   4.28  0.0684           0.109 
2 ENSG00000000005 -7.00   0.541 17.6   0.00216          0.0138
3 ENSG00000000419  0.120  5.34   0.114 0.743            0.776 
4 ENSG00000000457 -0.708  5.31   3.35  0.0993           0.145 
5 ENSG00000000460 -0.897  3.95   2.66  0.136            0.186 
6 ENSG00000000938  1.54   5.60   1.86  0.205            0.258 </code></pre>
</div>
<div id="gene-ids" class="section level2">
<h2>Gene IDs</h2>
<p>The {clusterProfiler} package uses the <code>enrichGO()</code>
function for performing a <a
href="https://yulab-smu.top/biomedical-knowledge-mining-book/clusterprofiler-go.html#clusterprofiler-go-ora">Gene
Ontology over-representation test</a>. The arguments are:</p>
<ul>
<li><code>gene</code> - a vector of entrez gene id.</li>
<li><code>OrgDb</code> - OrgDb database</li>
<li><code>keyType</code> - keytype of input gene</li>
<li><code>ont</code> - One of “BP”, “MF”, and “CC” subontologies, or
“ALL” for all three.</li>
<li><code>pvalueCutoff</code> - adjusted pvalue cutoff on enrichment
tests to report</li>
<li><code>pAdjustMethod</code> - one of “holm”, “hochberg”, “hommel”,
“bonferroni”, “BH”, “BY”, “fdr”, “none”</li>
<li><code>universe</code> - background genes. If missing, the all genes
listed in the database (e.g. TERM2GENE table) will be used as
background.</li>
<li><code>qvalueCutoff</code> - qvalue cutoff on enrichment tests to
report as significant. Tests must pass i) pvalueCutoff on unadjusted
pvalues, ii) pvalueCutoff on adjusted pvalues and iii) qvalueCutoff on
qvalues to be reported.</li>
<li><code>minGSSize</code> - minimal size of genes annotated by Ontology
term for testing.</li>
<li><code>maxGSSize</code> - maximal size of genes annotated for
testing</li>
<li><code>readable</code> - whether mapping gene ID to gene Name</li>
<li><code>pool</code> - If ont=‘ALL’, whether pool 3 GO
sub-ontologies</li>
</ul>
<p>Our example data uses Ensembl gene IDs, so we need to convert these
into Entrez Gene IDs. We can use the {org.Hs.eg.db} package provided by
<a
href="https://bioconductor.org/packages/release/data/annotation/html/org.Hs.eg.db.html">Bioconductor</a>,
which provides genome wide annotation for human, primarily based on
mapping using Entrez Gene identifiers.</p>
<pre class="r"><code>ensembl_to_entrez &lt;- AnnotationDbi::select(
  org.Hs.eg.db,
  keys = de_res$ensembl_gene_id,
  columns = c(&quot;ENSEMBL&quot;, &quot;ENTREZID&quot;), 
  keytype = &quot;ENSEMBL&quot;
)</code></pre>
<pre><code>&#39;select()&#39; returned 1:many mapping between keys and columns</code></pre>
<pre class="r"><code>de_res |&gt;
  dplyr::inner_join(ensembl_to_entrez, by = dplyr::join_by(ensembl_gene_id == ENSEMBL)) |&gt;
  dplyr::select(ensembl_gene_id, ENTREZID, dplyr::everything()) -&gt; de_res

head(de_res)</code></pre>
<pre><code># A tibble: 6 × 7
  ensembl_gene_id ENTREZID  logFC logCPM      F  PValue adjusted_pvalue
  &lt;chr&gt;           &lt;chr&gt;     &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt;   &lt;dbl&gt;           &lt;dbl&gt;
1 ENSG00000000003 7105      2.73   4.83   4.28  0.0684           0.109 
2 ENSG00000000005 64102    -7.00   0.541 17.6   0.00216          0.0138
3 ENSG00000000419 8813      0.120  5.34   0.114 0.743            0.776 
4 ENSG00000000457 57147    -0.708  5.31   3.35  0.0993           0.145 
5 ENSG00000000460 55732    -0.897  3.95   2.66  0.136            0.186 
6 ENSG00000000938 2268      1.54   5.60   1.86  0.205            0.258 </code></pre>
</div>
<div id="clusterprofiler" class="section level2">
<h2>clusterProfiler</h2>
<p>The example data contains results of a comparison between cancer
samples and normal samples; positive fold change indicates that genes
were expressed higher in cancer. There are a lot of genes up-regulated
in cancer.</p>
<pre class="r"><code>de_res |&gt;
  dplyr::filter(logFC &gt; 0, adjusted_pvalue &lt; 0.05) |&gt;
  nrow()</code></pre>
<pre><code>[1] 1442</code></pre>
<p>Instead of examining the list of up-regulated genes individually,
which would take a long time, we can see whether these genes are
associated with Gene Ontology terms more often than expected. Our
expectation will be based on the background.</p>
<p>First we’ll get the top 500 (an arbitrary number) most significantly
up-regulated genes.</p>
<pre class="r"><code>de_res |&gt;
  dplyr::filter(logFC &gt; 0, adjusted_pvalue &lt; 0.05) |&gt;
  dplyr::slice_min(order_by = adjusted_pvalue, n = 500) |&gt;
  dplyr::filter(!is.na(ENTREZID)) |&gt;
  dplyr::pull(ENTREZID) -&gt; sig_up_genes</code></pre>
<p>Create the background, i.e., the universe, using all genes tested for
differential expression so we can test for over-representation.</p>
<pre class="r"><code>de_res |&gt;
  dplyr::filter(!is.na(ENTREZID)) |&gt;
  dplyr::pull(ENTREZID) -&gt; the_universe</code></pre>
<p>Gene Ontology terms are grouped into three ontologies:</p>
<ol style="list-style-type: decimal">
<li>Molecular Function (MF)</li>
<li>Biological Process (BP), and</li>
<li>Cellular Component (CC)</li>
</ol>
<p>Here, we’ll perform an over-representation analysis using biological
processes.</p>
<pre class="r"><code>ego_bp &lt;- enrichGO(
  gene          = sig_up_genes,
  universe      = the_universe,
  OrgDb         = org.Hs.eg.db,
  ont           = &quot;BP&quot;,
  pAdjustMethod = &quot;BH&quot;,
  pvalueCutoff  = 0.01,
  qvalueCutoff  = 0.05,
  minGSSize     = 10,
  maxGSSize     = 500,
  readable      = TRUE
)

head(ego_bp)</code></pre>
<pre><code>                   ID
GO:0016064 GO:0016064
GO:0019724 GO:0019724
GO:0002460 GO:0002460
GO:0002449 GO:0002449
GO:0002443 GO:0002443
GO:0006959 GO:0006959
                                                                                                                         Description
GO:0016064                                                                                   immunoglobulin mediated immune response
GO:0019724                                                                                                  B cell mediated immunity
GO:0002460 adaptive immune response based on somatic recombination of immune receptors built from immunoglobulin superfamily domains
GO:0002449                                                                                              lymphocyte mediated immunity
GO:0002443                                                                                               leukocyte mediated immunity
GO:0006959                                                                                                   humoral immune response
           GeneRatio   BgRatio RichFactor FoldEnrichment    zScore       pvalue
GO:0016064    29/386 192/17354 0.15104167       6.790614 12.169056 3.046858e-16
GO:0019724    29/386 195/17354 0.14871795       6.686143 12.043556 4.660827e-16
GO:0002460    36/386 379/17354 0.09498681       4.270469  9.709355 2.085276e-13
GO:0002449    35/386 367/17354 0.09536785       4.287600  9.601062 4.074276e-13
GO:0002443    38/386 465/17354 0.08172043       3.674032  8.815698 4.599706e-12
GO:0006959    30/386 298/17354 0.10067114       4.526028  9.260213 5.283568e-12
               p.adjust       qvalue
GO:0016064 8.722738e-13 7.827737e-13
GO:0019724 8.722738e-13 7.827737e-13
GO:0002460 2.601729e-10 2.334777e-10
GO:0002449 3.812504e-10 3.421319e-10
GO:0002443 3.296066e-09 2.957871e-09
GO:0006959 3.296066e-09 2.957871e-09
                                                                                                                                                                                                                                                                        geneID
GO:0016064                                                       IGLC1/IGHG4/IGHG2/IGHG3/IGHV5-51/IGLL5/IGHV4-39/IGLC3/IGHV3-23/IGHV2-5/CD27/IGHV3-30/IGHV3-48/BATF/IGHG1/IGLC2/IGHV4-61/IGHA2/IGHV3-7/IGHV3-21/IGHV3-33/IGHV4-59/C4A/IGHV1-69D/FOXP3/IGKC/IGHV3-15/TREM2/CD28
GO:0019724                                                       IGLC1/IGHG4/IGHG2/IGHG3/IGHV5-51/IGLL5/IGHV4-39/IGLC3/IGHV3-23/IGHV2-5/CD27/IGHV3-30/IGHV3-48/BATF/IGHG1/IGLC2/IGHV4-61/IGHA2/IGHV3-7/IGHV3-21/IGHV3-33/IGHV4-59/C4A/IGHV1-69D/FOXP3/IGKC/IGHV3-15/TREM2/CD28
GO:0002460           IGLC1/IGHG4/IGHG2/IGHG3/IGHV5-51/IGLL5/IGHV4-39/IGLC3/IL4I1/IRF4/JAK3/IGHV3-23/IGHV2-5/CD27/IGHV3-30/CCL19/IGHV3-48/CXCL13/LILRB4/BATF/IGHG1/IGLC2/IGHV4-61/IGHA2/IGHV3-7/IGHV3-21/IGHV3-33/IGHV4-59/C4A/IGHV1-69D/FOXP3/IGKC/IGHV3-15/PLA2G4A/TREM2/CD28
GO:0002449                IGLC1/IGHG4/IGHG2/IGHG3/IGHV5-51/IGLL5/IGHV4-39/IGLC3/IL4I1/IGHV3-23/IGHV2-5/CD27/IGHV3-30/SLAMF7/IGHV3-48/NECTIN4/LILRB4/BATF/IGHG1/IGLC2/IGHV4-61/IGHA2/IGHV3-7/IGHV3-21/IGHV3-33/IGHV4-59/C4A/IGHV1-69D/CD2/FOXP3/LGALS9/IGKC/IGHV3-15/TREM2/CD28
GO:0002443 IGLC1/IGHG4/IGHG2/IGHG3/IGHV5-51/IGLL5/IGHV4-39/IGLC3/IL4I1/JAK3/IGHV3-23/IGHV2-5/CD27/IGHV3-30/SLAMF7/IGHV3-48/NECTIN4/LILRB4/BATF/IGHG1/IGLC2/IGHV4-61/IGHA2/IGHV3-7/IGHV3-21/IGHV3-33/IGHV4-59/C4A/CCL3/IGHV1-69D/CD84/CD2/FOXP3/LGALS9/IGKC/IGHV3-15/TREM2/CD28
GO:0006959                                                                                WFDC2/IGHG4/IGHG2/IGHG3/IGKV3-20/CXCL14/POU2AF1/MMP7/CCL22/BPIFA1/CXCL9/CCL19/CXCL13/CXCL1/LTF/CXCL10/MS4A1/IGHM/IGHG1/POU2F2/CCL7/TNFRSF21/ADM/IGHA2/C4A/CCL3/CXCL8/CCL8/TREM2/CD28
           Count
GO:0016064    29
GO:0019724    29
GO:0002460    36
GO:0002449    35
GO:0002443    38
GO:0006959    30</code></pre>
<p>Here, we’ll perform an over-representation analysis using molecular
functions.</p>
<pre class="r"><code>ego_mf &lt;- enrichGO(
  gene          = sig_up_genes,
  universe      = the_universe,
  OrgDb         = org.Hs.eg.db,
  ont           = &quot;MF&quot;,
  pAdjustMethod = &quot;BH&quot;,
  pvalueCutoff  = 0.01,
  qvalueCutoff  = 0.05,
  minGSSize     = 10,
  maxGSSize     = 500,
  readable      = TRUE
)

head(ego_mf)</code></pre>
<pre><code>                   ID                     Description GeneRatio   BgRatio
GO:0003823 GO:0003823                 antigen binding    50/387 158/17900
GO:0008009 GO:0008009              chemokine activity    10/387  44/17900
GO:0034987 GO:0034987 immunoglobulin receptor binding     7/387  18/17900
GO:0042379 GO:0042379      chemokine receptor binding    11/387  68/17900
GO:1901681 GO:1901681         sulfur compound binding    19/387 273/17900
GO:0005539 GO:0005539       glycosaminoglycan binding    17/387 238/17900
           RichFactor FoldEnrichment    zScore       pvalue     p.adjust
GO:0003823 0.31645570      14.637098 25.594038 1.316692e-44 9.348515e-42
GO:0008009 0.22727273      10.512098  9.390747 2.560692e-08 9.090455e-06
GO:0034987 0.38888889      17.987367 10.718743 5.420961e-08 1.282961e-05
GO:0042379 0.16176471       7.482140  7.960905 2.126713e-07 3.774916e-05
GO:1901681 0.06959707       3.219089  5.492328 8.292546e-06 1.177542e-03
GO:0005539 0.07142857       3.303802  5.318673 1.769284e-05 2.057883e-03
                 qvalue
GO:0003823 8.787188e-42
GO:0008009 8.544623e-06
GO:0034987 1.205926e-05
GO:0042379 3.548253e-05
GO:1901681 1.106837e-03
GO:0005539 1.934319e-03
                                                                                                                                                                                                                                                                                                                                                                                                                               geneID
GO:0003823 IGKV4-1/IGLV1-51/IGLV1-44/IGLV1-40/IGLV3-21/IGLC1/IGHG4/IGHG2/IGHG3/IGHV5-51/IGKV3-20/IGKV1-5/IGKV3-15/IGLL5/IGHV4-39/IGLV1-47/IGLV2-23/IGLV2-14/IGLC3/IGKV1D-39/IGHV3-23/IGHV2-5/HLA-DQB2/IGLV3-19/IGHV3-30/IGHV3-48/IGKV2D-28/IGLV6-57/IGLV3-25/IGHM/IGHG1/IGLC2/SLC7A5/IGHV4-61/IGLV2-8/IGLV7-43/IGLV3-27/IGHA2/IGHV3-7/IGHV3-21/IGHV3-33/IGHV4-59/HLA-DQA1/HLA-DQA2/IGKV1-39/IGKV2-30/IGHV1-69D/IGKC/IGHV3-15/IGKV1-17
GO:0008009                                                                                                                                                                                                                                                                                                                                                                CXCL14/CCL22/CXCL9/CCL19/CXCL13/CXCL10/CCL7/CCL3/CXCL8/CCL8
GO:0034987                                                                                                                                                                                                                                                                                                                                                                                  IGHG4/IGHG2/IGHG3/IGHM/IGHG1/ADAM28/IGHA2
GO:0042379                                                                                                                                                                                                                                                                                                                                                          CXCL14/CCL22/CXCL9/CCL19/CXCL13/CXCL10/CCL7/CCL3/CXCL8/STAT1/CCL8
GO:1901681                                                                                                                                                                                                                                                                                                            PLA2G2D/CXCL13/LTF/CXCL10/TMEM184A/MDK/PTGES/CCN4/COMP/CCL7/NRP2/RCC1/CHST15/ADGRG1/THBS2/CXCL8/PTGES2/GSS/CCL8
GO:0005539                                                                                                                                                                                                                                                                                                                        PLA2G2D/CXCL13/LTF/CXCL10/IGHM/TMEM184A/MDK/CEMIP/CCN4/COMP/CCL7/NRP2/ADGRG1/THBS2/CXCL8/CCL8/TREM2
           Count
GO:0003823    50
GO:0008009    10
GO:0034987     7
GO:0042379    11
GO:1901681    19
GO:0005539    17</code></pre>
<div id="examining-the-results" class="section level3">
<h3>Examining the results</h3>
<p>Results are stored in an <code>enrichResult</code> class.</p>
<pre class="r"><code>class(ego_mf)</code></pre>
<pre><code>[1] &quot;enrichResult&quot;
attr(,&quot;package&quot;)
[1] &quot;DOSE&quot;</code></pre>
<p>Slot names of the class.</p>
<pre class="r"><code>slotNames(ego_mf)</code></pre>
<pre><code> [1] &quot;result&quot;        &quot;pvalueCutoff&quot;  &quot;pAdjustMethod&quot; &quot;qvalueCutoff&quot; 
 [5] &quot;organism&quot;      &quot;ontology&quot;      &quot;gene&quot;          &quot;keytype&quot;      
 [9] &quot;universe&quot;      &quot;gene2Symbol&quot;   &quot;geneSets&quot;      &quot;readable&quot;     
[13] &quot;termsim&quot;       &quot;method&quot;        &quot;dr&quot;           </code></pre>
<p>One question that has been asked (and not conclusively answered) on
the interweb is why do the universe lengths differ when the same list of
background genes is provided?</p>
<pre class="r"><code>length(the_universe)</code></pre>
<pre><code>[1] 28981</code></pre>
<pre class="r"><code>length(ego_mf@universe)</code></pre>
<pre><code>[1] 17900</code></pre>
<pre class="r"><code>length(ego_bp@universe)</code></pre>
<pre><code>[1] 17354</code></pre>
<p>In order to address this, we will first create a Entrez Gene ID to GO
term lookup table.</p>
<pre class="r"><code>entrez_to_go &lt;- select(
  org.Hs.eg.db,
  keys = keys(org.Hs.eg.db, keytype = &quot;ENTREZID&quot;),
  columns = c(&quot;GO&quot;, &quot;ONTOLOGY&quot;),
  keytype = &quot;ENTREZID&quot;
)</code></pre>
<pre><code>&#39;select()&#39; returned 1:many mapping between keys and columns</code></pre>
<pre class="r"><code>head(entrez_to_go)</code></pre>
<pre><code>  ENTREZID         GO EVIDENCE ONTOLOGY
1        1 GO:0002764      IBA       BP
2        1 GO:0005576      HDA       CC
3        1 GO:0005576      IDA       CC
4        1 GO:0005576      TAS       CC
5        1 GO:0005615      HDA       CC
6        1 GO:0005886      IBA       CC</code></pre>
<p>Intuitively (to me), the only reason why the universe lengths are
different per ontology group is because there is some sort of filtering
performed per ontology group. One gene is associated to various GO terms
and if you group them into the three respective groups, their numbers
might be different. Here’s an example.</p>
<pre class="r"><code>entrez_to_go |&gt;
  dplyr::filter(ENTREZID == &quot;1&quot;) |&gt;
  dplyr::summarise(n = n(), .by = ONTOLOGY)</code></pre>
<pre><code>  ONTOLOGY  n
1       BP  1
2       CC 11</code></pre>
<p>Let’s reproduce the universe list for molecular functions. The
pipeline below is:</p>
<ul>
<li>Filter the Entrez Gene ID to GO term lookup table to only contain GO
terms in the MF group</li>
<li>Output only the <code>ENTREZID</code> and <code>GO</code>
columns</li>
<li>Keep only the unique rows</li>
<li>Keep only the Entrez Gene IDs in our background</li>
<li>Output all the Entrez Gene IDs</li>
<li>Keep only unique IDs</li>
<li>Calculate the length</li>
</ul>
<pre class="r"><code>length(ego_mf@universe)</code></pre>
<pre><code>[1] 17900</code></pre>
<pre class="r"><code>dplyr::filter(entrez_to_go, ONTOLOGY == &quot;MF&quot;) |&gt;
  dplyr::select(ENTREZID, GO) |&gt;
  dplyr::distinct() |&gt;
  dplyr::filter(ENTREZID %in% the_universe) |&gt;
  dplyr::pull(ENTREZID) |&gt;
  unique() |&gt;
  length()</code></pre>
<pre><code>[1] 17900</code></pre>
<p>Repeat the same workflow on biological processes.</p>
<pre class="r"><code>length(ego_bp@universe)</code></pre>
<pre><code>[1] 17354</code></pre>
<pre class="r"><code>dplyr::filter(entrez_to_go, ONTOLOGY == &quot;BP&quot;) |&gt;
  dplyr::select(ENTREZID, GO) |&gt;
  dplyr::distinct() |&gt;
  dplyr::filter(ENTREZID %in% the_universe) |&gt;
  dplyr::pull(ENTREZID) |&gt;
  unique() |&gt;
  length()</code></pre>
<pre><code>[1] 17354</code></pre>
<p>Getting the same lengths confirms our intuition that gene lists are
created by only keeping genes that have at least one GO term in the Gene
Ontology group of interest.</p>
</div>
<div id="ratios" class="section level3">
<h3>Ratios</h3>
<p>Another question asked on the interweb and not conclusively answered
is how are <code>GeneRatio</code> and <code>BgRatio</code> calculated.
Since we now know that genes are filtered per ontology group we need to
take this into account. Let’s reproduce the first result for biological
processes.</p>
<pre class="r"><code>ego_bp |&gt;
  as.data.frame() |&gt;
  dplyr::slice_head(n = 1)</code></pre>
<pre><code>                   ID                             Description GeneRatio
GO:0016064 GO:0016064 immunoglobulin mediated immune response    29/386
             BgRatio RichFactor FoldEnrichment   zScore       pvalue
GO:0016064 192/17354  0.1510417       6.790614 12.16906 3.046858e-16
               p.adjust       qvalue
GO:0016064 8.722738e-13 7.827737e-13
                                                                                                                                                                                                                  geneID
GO:0016064 IGLC1/IGHG4/IGHG2/IGHG3/IGHV5-51/IGLL5/IGHV4-39/IGLC3/IGHV3-23/IGHV2-5/CD27/IGHV3-30/IGHV3-48/BATF/IGHG1/IGLC2/IGHV4-61/IGHA2/IGHV3-7/IGHV3-21/IGHV3-33/IGHV4-59/C4A/IGHV1-69D/FOXP3/IGKC/IGHV3-15/TREM2/CD28
           Count
GO:0016064    29</code></pre>
<p>Re-create denominators of <code>GeneRatio</code> and
<code>BgRatio</code>.</p>
<pre class="r"><code>gene_list_to_go &lt;- function(gene_list, ont = &quot;BP&quot;){
  dplyr::filter(entrez_to_go, ONTOLOGY == ont) |&gt;
    dplyr::select(ENTREZID, GO) |&gt;
    dplyr::filter(ENTREZID %in% gene_list)
}

gene_list_to_go(sig_up_genes) |&gt;
  pull(ENTREZID) |&gt;
  unique() -&gt; sig_up_genes_entrez
length(sig_up_genes_entrez)</code></pre>
<pre><code>[1] 386</code></pre>
<pre class="r"><code>gene_list_to_go(the_universe) |&gt;
  pull(ENTREZID) |&gt;
  unique() -&gt; the_universe_entrez
length(the_universe_entrez)</code></pre>
<pre><code>[1] 17354</code></pre>
<p><a href="GO:0016064" class="uri">GO:0016064</a> is associated to 19
Entrez Gene IDs but the results above are 29.</p>
<pre class="r"><code>entrez_to_go |&gt;
  dplyr::filter(ENTREZID %in% sig_up_genes_entrez) |&gt;
  dplyr::summarise(n = n(), .by = &quot;GO&quot;) |&gt;
  dplyr::filter(GO == &quot;GO:0016064&quot;)</code></pre>
<pre><code>          GO  n
1 GO:0016064 19</code></pre>
<p>The numerator can also be calculated by tallying the
<code>geneID</code> column.</p>
<pre class="r"><code>ego_bp |&gt;
  as.data.frame() |&gt;
  dplyr::slice_head(n = 1) |&gt;
  dplyr::pull(geneID) |&gt;
  stringr::str_split(&quot;/&quot;) |&gt;
  unlist() |&gt;
  unique() -&gt; ego_bp_first_gs

ego_bp_first_gs</code></pre>
<pre><code> [1] &quot;IGLC1&quot;     &quot;IGHG4&quot;     &quot;IGHG2&quot;     &quot;IGHG3&quot;     &quot;IGHV5-51&quot;  &quot;IGLL5&quot;    
 [7] &quot;IGHV4-39&quot;  &quot;IGLC3&quot;     &quot;IGHV3-23&quot;  &quot;IGHV2-5&quot;   &quot;CD27&quot;      &quot;IGHV3-30&quot; 
[13] &quot;IGHV3-48&quot;  &quot;BATF&quot;      &quot;IGHG1&quot;     &quot;IGLC2&quot;     &quot;IGHV4-61&quot;  &quot;IGHA2&quot;    
[19] &quot;IGHV3-7&quot;   &quot;IGHV3-21&quot;  &quot;IGHV3-33&quot;  &quot;IGHV4-59&quot;  &quot;C4A&quot;       &quot;IGHV1-69D&quot;
[25] &quot;FOXP3&quot;     &quot;IGKC&quot;      &quot;IGHV3-15&quot;  &quot;TREM2&quot;     &quot;CD28&quot;     </code></pre>
<p>A clue to reproducing the numerator is that gene symbols are used, so
there must be some Entrez Gene ID to gene symbol conversion (potentially
contributing to a one-to-many lookup). We will create a Entrez Gene ID
to gene symbol lookup table.</p>
<pre class="r"><code># columns(org.Hs.eg.db)
entrez_to_gene_symbol &lt;- select(
  org.Hs.eg.db,
  keys = keys(org.Hs.eg.db, keytype = &quot;ENTREZID&quot;),
  columns = c(&quot;SYMBOL&quot;, &quot;ALIAS&quot;),
  keytype = &quot;ENTREZID&quot;
)</code></pre>
<pre><code>&#39;select()&#39; returned 1:many mapping between keys and columns</code></pre>
<pre class="r"><code>head(entrez_to_gene_symbol)</code></pre>
<pre><code>  ENTREZID SYMBOL    ALIAS
1        1   A1BG      A1B
2        1   A1BG      ABG
3        1   A1BG      GAB
4        1   A1BG HYST2477
5        1   A1BG     A1BG
6        2    A2M     A2MD</code></pre>
<p>Get the list of Entrez Gene IDs for up-regulated genes for the first
result of the biological processes enrichment analysis.</p>
<p>Note that this length does not match the numerator of
<code>GeneRatio</code>.</p>
<pre class="r"><code>gene_list_to_go(sig_up_genes) |&gt;
  dplyr::filter(GO == &quot;GO:0016064&quot;) |&gt;
  dplyr::pull(ENTREZID) |&gt;
  unique() -&gt; eg1

eg1</code></pre>
<pre><code> [1] &quot;939&quot;       &quot;3514&quot;      &quot;3537&quot;      &quot;3538&quot;      &quot;3539&quot;      &quot;28388&quot;    
 [7] &quot;28391&quot;     &quot;28392&quot;     &quot;28394&quot;     &quot;28424&quot;     &quot;28434&quot;     &quot;28439&quot;    
[13] &quot;28442&quot;     &quot;28444&quot;     &quot;28448&quot;     &quot;28452&quot;     &quot;28457&quot;     &quot;100423062&quot;
[19] &quot;102723169&quot;</code></pre>
<p>Convert to gene symbols.</p>
<pre class="r"><code>entrez_to_gene_symbol |&gt;
  dplyr::filter(ENTREZID %in% eg1) -&gt; eg1_gene_symbol

eg1_gene_symbol</code></pre>
<pre><code>    ENTREZID    SYMBOL       ALIAS
1        939      CD27        S152
2        939      CD27 S152. LPFS2
3        939      CD27         T14
4        939      CD27     TNFRSF7
5        939      CD27        Tp55
6        939      CD27        CD27
7       3514      IGKC       HCAK1
8       3514      IGKC       IGKCD
9       3514      IGKC          Km
10      3514      IGKC        IGKC
11      3537     IGLC1        IGLC
12      3537     IGLC1       IGLC1
13      3538     IGLC2        IGLC
14      3538     IGLC2       IGLC2
15      3539     IGLC3        IGLC
16      3539     IGLC3       IGLC3
17     28388  IGHV5-51     IGHV551
18     28388  IGHV5-51          VH
19     28388  IGHV5-51    IGHV5-51
20     28391  IGHV4-61     IGHV461
21     28391  IGHV4-61          VH
22     28391  IGHV4-61    IGHV4-61
23     28392  IGHV4-59     IGHV459
24     28392  IGHV4-59          VH
25     28392  IGHV4-59    IGHV4-59
26     28394  IGHV4-39     IGHV439
27     28394  IGHV4-39          VH
28     28394  IGHV4-39    IGHV4-39
29     28424  IGHV3-48     IGHV348
30     28424  IGHV3-48          VH
31     28424  IGHV3-48    IGHV3-48
32     28434  IGHV3-33     IGHV333
33     28434  IGHV3-33          VH
34     28434  IGHV3-33    IGHV3-33
35     28439  IGHV3-30     IGHV330
36     28439  IGHV3-30          VH
37     28439  IGHV3-30    IGHV3-30
38     28442  IGHV3-23        DP47
39     28442  IGHV3-23     IGHV323
40     28442  IGHV3-23       V3-23
41     28442  IGHV3-23        VH26
42     28442  IGHV3-23    IGHV3-23
43     28444  IGHV3-21     IGHV321
44     28444  IGHV3-21          VH
45     28444  IGHV3-21    IGHV3-21
46     28448  IGHV3-15     IGHV315
47     28448  IGHV3-15          VH
48     28448  IGHV3-15    IGHV3-15
49     28452   IGHV3-7      IGHV37
50     28452   IGHV3-7          VH
51     28452   IGHV3-7     IGHV3-7
52     28457   IGHV2-5      IGHV25
53     28457   IGHV2-5          VH
54     28457   IGHV2-5     IGHV2-5
55 100423062     IGLL5         IGL
56 100423062     IGLL5        IGLV
57 100423062     IGLL5      VL-MAR
58 100423062     IGLL5       IGLL5
59 102723169 IGHV1-69D   IGHV1-69D</code></pre>
<p>Are the gene symbols in the enriched result in our table of gene
symbol and aliases?</p>
<pre class="r"><code>ego_bp_first_gs %in% c(unique(eg1_gene_symbol$SYMBOL), unique(eg1_gene_symbol$ALIAS))</code></pre>
<pre><code> [1]  TRUE FALSE FALSE FALSE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE
[13]  TRUE FALSE FALSE  TRUE  TRUE FALSE  TRUE  TRUE  TRUE  TRUE FALSE  TRUE
[25] FALSE  TRUE  TRUE FALSE FALSE</code></pre>
<p>Are all the gene symbols in the enriched results in our Entrez Gene
ID to gene symbol lookup table? (Yes.)</p>
<pre class="r"><code>entrez_to_gene_symbol |&gt;
  dplyr::filter(SYMBOL %in% ego_bp_first_gs) |&gt;
  dplyr::pull(SYMBOL) |&gt;
  unique() |&gt;
  length()</code></pre>
<pre><code>[1] 29</code></pre>
<p>Get the Entrez Gene IDs from the gene symbols of the results.</p>
<pre class="r"><code>entrez_to_gene_symbol |&gt;
  dplyr::filter(SYMBOL %in% ego_bp_first_gs) |&gt;
  dplyr::pull(ENTREZID) |&gt;
  unique() -&gt; ego_bp_first_gs_entrez

ego_bp_first_gs_entrez</code></pre>
<pre><code> [1] &quot;720&quot;       &quot;939&quot;       &quot;940&quot;       &quot;3494&quot;      &quot;3500&quot;      &quot;3501&quot;     
 [7] &quot;3502&quot;      &quot;3503&quot;      &quot;3514&quot;      &quot;3537&quot;      &quot;3538&quot;      &quot;3539&quot;     
[13] &quot;10538&quot;     &quot;28388&quot;     &quot;28391&quot;     &quot;28392&quot;     &quot;28394&quot;     &quot;28424&quot;    
[19] &quot;28434&quot;     &quot;28439&quot;     &quot;28442&quot;     &quot;28444&quot;     &quot;28448&quot;     &quot;28452&quot;    
[25] &quot;28457&quot;     &quot;50943&quot;     &quot;54209&quot;     &quot;100423062&quot; &quot;102723169&quot;</code></pre>
<p>Why are these Entrez Gene IDs included?</p>
<pre class="r"><code>additional_ids &lt;- setdiff(ego_bp_first_gs_entrez, eg1)
additional_ids</code></pre>
<pre><code> [1] &quot;720&quot;   &quot;940&quot;   &quot;3494&quot;  &quot;3500&quot;  &quot;3501&quot;  &quot;3502&quot;  &quot;3503&quot;  &quot;10538&quot; &quot;50943&quot;
[10] &quot;54209&quot;</code></pre>
<p>Are these additional IDs in our results?</p>
<pre class="r"><code>additional_ids %in% sig_up_genes</code></pre>
<pre><code> [1] TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE</code></pre>
<p>Build GO to Entrez Gene ID lookup.</p>
<pre class="r"><code>go_to_entrez &lt;- select(
  org.Hs.eg.db,
  keys = keys(GO.db),
  columns = &quot;ENTREZID&quot;,
  keytype = &quot;GO&quot;
)</code></pre>
<pre><code>&#39;select()&#39; returned 1:many mapping between keys and columns</code></pre>
<pre class="r"><code>head(go_to_entrez)</code></pre>
<pre><code>          GO EVIDENCE ONTOLOGY ENTREZID
1 GO:0000001     &lt;NA&gt;     &lt;NA&gt;     &lt;NA&gt;
2 GO:0000002      TAS       BP      291
3 GO:0000002      IMP       BP     1890
4 GO:0000002      ISS       BP     4205
5 GO:0000002      IMP       BP     4358
6 GO:0000002      IMP       BP     4976</code></pre>
<p>I still get 19.</p>
<pre class="r"><code>go_to_entrez |&gt;
  dplyr::filter(GO == &quot;GO:0016064&quot;, ENTREZID %in% sig_up_genes) |&gt;
  dplyr::pull(ENTREZID) |&gt;
  unique() |&gt;
  length()</code></pre>
<pre><code>[1] 19</code></pre>
<p>Include GO.</p>
<pre class="r"><code>entrez_to_symbol_and_go &lt;- select(
  org.Hs.eg.db,
  keys = keys(org.Hs.eg.db, keytype = &quot;ENTREZID&quot;),
  columns = c(&quot;SYMBOL&quot;, &quot;GO&quot;),
  keytype = &quot;ENTREZID&quot;
)</code></pre>
<pre><code>&#39;select()&#39; returned 1:many mapping between keys and columns</code></pre>
<pre class="r"><code>entrez_to_symbol_and_go |&gt;
  dplyr::filter(GO == &quot;GO:0016064&quot;, ENTREZID %in% sig_up_genes)</code></pre>
<pre><code>    ENTREZID    SYMBOL         GO EVIDENCE ONTOLOGY
1        939      CD27 GO:0016064      NAS       BP
2       3514      IGKC GO:0016064      IBA       BP
3       3537     IGLC1 GO:0016064      IBA       BP
4       3538     IGLC2 GO:0016064      IBA       BP
5       3539     IGLC3 GO:0016064      IBA       BP
6      28388  IGHV5-51 GO:0016064      IBA       BP
7      28391  IGHV4-61 GO:0016064      IBA       BP
8      28392  IGHV4-59 GO:0016064      IBA       BP
9      28394  IGHV4-39 GO:0016064      IBA       BP
10     28424  IGHV3-48 GO:0016064      IBA       BP
11     28434  IGHV3-33 GO:0016064      IBA       BP
12     28439  IGHV3-30 GO:0016064      IBA       BP
13     28442  IGHV3-23 GO:0016064      IBA       BP
14     28444  IGHV3-21 GO:0016064      IBA       BP
15     28448  IGHV3-15 GO:0016064      IBA       BP
16     28452   IGHV3-7 GO:0016064      IBA       BP
17     28457   IGHV2-5 GO:0016064      IBA       BP
18 100423062     IGLL5 GO:0016064      IBA       BP
19 102723169 IGHV1-69D GO:0016064      IBA       BP</code></pre>
<p>Still a mystery on the number 29.</p>
</div>
<div id="richfactor" class="section level3">
<h3>RichFactor</h3>
<p>The <code>RichFactor</code> is calculated by <code>Count</code> /
<code>M</code>, where <code>Count</code> is the number of genes from
your input list annotated to the term, and <code>M</code> is the total
number of genes in the background that are annotated to that term (the
numerator of <code>BgRatio</code>).</p>
<pre class="r"><code>ego_bp |&gt;
  as.data.frame() |&gt;
  dplyr::slice_head(n = 1) -&gt; my_df

my_df$Count / as.numeric(unlist(stringr::str_split(string = my_df$BgRatio, pattern = &#39;/&#39;))[1]) == my_df$RichFactor</code></pre>
<pre><code>[1] TRUE</code></pre>
</div>
<div id="foldenrichment" class="section level3">
<h3>FoldEnrichment</h3>
<p><code>FoldEnrichment</code> is simply <code>GeneRatio</code> divided
by <code>BgRatio</code>.</p>
<pre class="r"><code>ego_bp |&gt;
  as.data.frame() |&gt;
  dplyr::slice_head(n = 1) -&gt; my_df

all.equal(my_df$FoldEnrichment, eval(parse(text = my_df$GeneRatio)) / eval(parse(text = my_df$BgRatio)))</code></pre>
<pre><code>[1] TRUE</code></pre>
</div>
<div id="visualisations" class="section level3">
<h3>Visualisations</h3>
<p>Bar plot showing each enriched GO term coloured by the adjusted
p-value.</p>
<pre class="r"><code>barplot(ego_bp, showCategory=10)</code></pre>
<p><img src="figure/cluster_profiler.Rmd/barplot-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-barplot-1">
Past versions of barplot-1.png
</button>
</p>
<div id="fig-barplot-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/davetang/muse/blob/18e8e9b7e30957ba8e89e2e3844d045616cb3a86/docs/figure/cluster_profiler.Rmd/barplot-1.png" target="_blank">18e8e9b</a>
</td>
<td>
Dave Tang
</td>
<td>
2025-05-26
</td>
</tr>
<tr>
<td>
<a href="https://github.com/davetang/muse/blob/e815348c6fc2adfec65202ff3718880a944b088c/docs/figure/cluster_profiler.Rmd/barplot-1.png" target="_blank">e815348</a>
</td>
<td>
Dave Tang
</td>
<td>
2025-03-06
</td>
</tr>
<tr>
<td>
<a href="https://github.com/davetang/muse/blob/2c080dce2c4547d90fe19cf649b4683a4aaa5142/docs/figure/cluster_profiler.Rmd/barplot-1.png" target="_blank">2c080dc</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-11-01
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>Dot plot showing each enriched GO term with associated
statistics.</p>
<pre class="r"><code>dotplot(ego_bp, showCategory=10)</code></pre>
<p><img src="figure/cluster_profiler.Rmd/dotplot-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-dotplot-1">
Past versions of dotplot-1.png
</button>
</p>
<div id="fig-dotplot-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/davetang/muse/blob/18e8e9b7e30957ba8e89e2e3844d045616cb3a86/docs/figure/cluster_profiler.Rmd/dotplot-1.png" target="_blank">18e8e9b</a>
</td>
<td>
Dave Tang
</td>
<td>
2025-05-26
</td>
</tr>
<tr>
<td>
<a href="https://github.com/davetang/muse/blob/e815348c6fc2adfec65202ff3718880a944b088c/docs/figure/cluster_profiler.Rmd/dotplot-1.png" target="_blank">e815348</a>
</td>
<td>
Dave Tang
</td>
<td>
2025-03-06
</td>
</tr>
<tr>
<td>
<a href="https://github.com/davetang/muse/blob/2c080dce2c4547d90fe19cf649b4683a4aaa5142/docs/figure/cluster_profiler.Rmd/dotplot-1.png" target="_blank">2c080dc</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-11-01
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>Heat plot showing the enriched GO terms on the y-axis and the genes
on the x-axis. Genes with the associated GO term are highlighted.</p>
<pre class="r"><code>heatplot(ego_bp, showCategory=10)</code></pre>
<p><img src="figure/cluster_profiler.Rmd/heatplot-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-heatplot-1">
Past versions of heatplot-1.png
</button>
</p>
<div id="fig-heatplot-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/davetang/muse/blob/18e8e9b7e30957ba8e89e2e3844d045616cb3a86/docs/figure/cluster_profiler.Rmd/heatplot-1.png" target="_blank">18e8e9b</a>
</td>
<td>
Dave Tang
</td>
<td>
2025-05-26
</td>
</tr>
<tr>
<td>
<a href="https://github.com/davetang/muse/blob/e815348c6fc2adfec65202ff3718880a944b088c/docs/figure/cluster_profiler.Rmd/heatplot-1.png" target="_blank">e815348</a>
</td>
<td>
Dave Tang
</td>
<td>
2025-03-06
</td>
</tr>
<tr>
<td>
<a href="https://github.com/davetang/muse/blob/2c080dce2c4547d90fe19cf649b4683a4aaa5142/docs/figure/cluster_profiler.Rmd/heatplot-1.png" target="_blank">2c080dc</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-11-01
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p><code>goplot</code> shows the gene ontology graph with the enriched
GO terms highlighted.</p>
<pre class="r"><code>goplot(ego_bp)</code></pre>
<pre><code>Warning: ggrepel: 1 unlabeled data points (too many overlaps). Consider
increasing max.overlaps</code></pre>
<p><img src="figure/cluster_profiler.Rmd/goplot-1.png" width="1152" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-goplot-1">
Past versions of goplot-1.png
</button>
</p>
<div id="fig-goplot-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/davetang/muse/blob/18e8e9b7e30957ba8e89e2e3844d045616cb3a86/docs/figure/cluster_profiler.Rmd/goplot-1.png" target="_blank">18e8e9b</a>
</td>
<td>
Dave Tang
</td>
<td>
2025-05-26
</td>
</tr>
<tr>
<td>
<a href="https://github.com/davetang/muse/blob/0ac82102a0064cf6f70ad9fdec39e3272f0aefe0/docs/figure/cluster_profiler.Rmd/goplot-1.png" target="_blank">0ac8210</a>
</td>
<td>
Dave Tang
</td>
<td>
2025-05-24
</td>
</tr>
<tr>
<td>
<a href="https://github.com/davetang/muse/blob/c8e80cd43b29fa99ef8a009be39c1b496db2de7f/docs/figure/cluster_profiler.Rmd/goplot-1.png" target="_blank">c8e80cd</a>
</td>
<td>
Dave Tang
</td>
<td>
2025-05-23
</td>
</tr>
<tr>
<td>
<a href="https://github.com/davetang/muse/blob/006e824308df24e7f43c3d664e30d96617250941/docs/figure/cluster_profiler.Rmd/goplot-1.png" target="_blank">006e824</a>
</td>
<td>
Dave Tang
</td>
<td>
2025-03-09
</td>
</tr>
<tr>
<td>
<a href="https://github.com/davetang/muse/blob/f0166b0791270b7f3b51540dc73a64d37947e451/docs/figure/cluster_profiler.Rmd/goplot-1.png" target="_blank">f0166b0</a>
</td>
<td>
Dave Tang
</td>
<td>
2025-03-09
</td>
</tr>
<tr>
<td>
<a href="https://github.com/davetang/muse/blob/e815348c6fc2adfec65202ff3718880a944b088c/docs/figure/cluster_profiler.Rmd/goplot-1.png" target="_blank">e815348</a>
</td>
<td>
Dave Tang
</td>
<td>
2025-03-06
</td>
</tr>
<tr>
<td>
<a href="https://github.com/davetang/muse/blob/2c080dce2c4547d90fe19cf649b4683a4aaa5142/docs/figure/cluster_profiler.Rmd/goplot-1.png" target="_blank">2c080dc</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-11-01
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>Another nice feature of {clusterProfiler} is that you can plot
multiple gene lists together. We can create a list of down-regulated
genes.</p>
<pre class="r"><code>de_res |&gt;
  dplyr::filter(logFC &lt; 0, adjusted_pvalue &lt; 0.05) |&gt;
  dplyr::slice_min(order_by = adjusted_pvalue, n = 500) |&gt;
  dplyr::filter(!is.na(ENTREZID)) |&gt;
  dplyr::pull(ENTREZID) -&gt; sig_down_genes</code></pre>
<p>Perform GO enrichment on two gene lists.</p>
<pre class="r"><code>my_gene_list &lt;- list(
  up_gene = sig_up_genes,
  down_gene = sig_down_genes
)

ego_bp_both &lt;- compareCluster(
  geneCluster = my_gene_list,
  fun = &quot;enrichGO&quot;,
  universe = the_universe,
  OrgDb = org.Hs.eg.db,
  keyType = &quot;ENTREZID&quot;,
  ont = &quot;BP&quot;,
  pvalueCutoff = 0.01,
  pAdjustMethod = &quot;BH&quot;,
  qvalueCutoff = 0.05,
  minGSSize = 10,
  maxGSSize = 500,
  readable = TRUE
)

head(as.data.frame(ego_bp_both))</code></pre>
<pre><code>  Cluster         ID
1 up_gene GO:0016064
2 up_gene GO:0019724
3 up_gene GO:0002460
4 up_gene GO:0002449
5 up_gene GO:0002443
6 up_gene GO:0006959
                                                                                                                Description
1                                                                                   immunoglobulin mediated immune response
2                                                                                                  B cell mediated immunity
3 adaptive immune response based on somatic recombination of immune receptors built from immunoglobulin superfamily domains
4                                                                                              lymphocyte mediated immunity
5                                                                                               leukocyte mediated immunity
6                                                                                                   humoral immune response
  GeneRatio   BgRatio RichFactor FoldEnrichment    zScore       pvalue
1    29/386 192/17354 0.15104167       6.790614 12.169056 3.046858e-16
2    29/386 195/17354 0.14871795       6.686143 12.043556 4.660827e-16
3    36/386 379/17354 0.09498681       4.270469  9.709355 2.085276e-13
4    35/386 367/17354 0.09536785       4.287600  9.601062 4.074276e-13
5    38/386 465/17354 0.08172043       3.674032  8.815698 4.599706e-12
6    30/386 298/17354 0.10067114       4.526028  9.260213 5.283568e-12
      p.adjust       qvalue
1 8.722738e-13 7.827737e-13
2 8.722738e-13 7.827737e-13
3 2.601729e-10 2.334777e-10
4 3.812504e-10 3.421319e-10
5 3.296066e-09 2.957871e-09
6 3.296066e-09 2.957871e-09
                                                                                                                                                                                                                                                               geneID
1                                                       IGLC1/IGHG4/IGHG2/IGHG3/IGHV5-51/IGLL5/IGHV4-39/IGLC3/IGHV3-23/IGHV2-5/CD27/IGHV3-30/IGHV3-48/BATF/IGHG1/IGLC2/IGHV4-61/IGHA2/IGHV3-7/IGHV3-21/IGHV3-33/IGHV4-59/C4A/IGHV1-69D/FOXP3/IGKC/IGHV3-15/TREM2/CD28
2                                                       IGLC1/IGHG4/IGHG2/IGHG3/IGHV5-51/IGLL5/IGHV4-39/IGLC3/IGHV3-23/IGHV2-5/CD27/IGHV3-30/IGHV3-48/BATF/IGHG1/IGLC2/IGHV4-61/IGHA2/IGHV3-7/IGHV3-21/IGHV3-33/IGHV4-59/C4A/IGHV1-69D/FOXP3/IGKC/IGHV3-15/TREM2/CD28
3           IGLC1/IGHG4/IGHG2/IGHG3/IGHV5-51/IGLL5/IGHV4-39/IGLC3/IL4I1/IRF4/JAK3/IGHV3-23/IGHV2-5/CD27/IGHV3-30/CCL19/IGHV3-48/CXCL13/LILRB4/BATF/IGHG1/IGLC2/IGHV4-61/IGHA2/IGHV3-7/IGHV3-21/IGHV3-33/IGHV4-59/C4A/IGHV1-69D/FOXP3/IGKC/IGHV3-15/PLA2G4A/TREM2/CD28
4                IGLC1/IGHG4/IGHG2/IGHG3/IGHV5-51/IGLL5/IGHV4-39/IGLC3/IL4I1/IGHV3-23/IGHV2-5/CD27/IGHV3-30/SLAMF7/IGHV3-48/NECTIN4/LILRB4/BATF/IGHG1/IGLC2/IGHV4-61/IGHA2/IGHV3-7/IGHV3-21/IGHV3-33/IGHV4-59/C4A/IGHV1-69D/CD2/FOXP3/LGALS9/IGKC/IGHV3-15/TREM2/CD28
5 IGLC1/IGHG4/IGHG2/IGHG3/IGHV5-51/IGLL5/IGHV4-39/IGLC3/IL4I1/JAK3/IGHV3-23/IGHV2-5/CD27/IGHV3-30/SLAMF7/IGHV3-48/NECTIN4/LILRB4/BATF/IGHG1/IGLC2/IGHV4-61/IGHA2/IGHV3-7/IGHV3-21/IGHV3-33/IGHV4-59/C4A/CCL3/IGHV1-69D/CD84/CD2/FOXP3/LGALS9/IGKC/IGHV3-15/TREM2/CD28
6                                                                                WFDC2/IGHG4/IGHG2/IGHG3/IGKV3-20/CXCL14/POU2AF1/MMP7/CCL22/BPIFA1/CXCL9/CCL19/CXCL13/CXCL1/LTF/CXCL10/MS4A1/IGHM/IGHG1/POU2F2/CCL7/TNFRSF21/ADM/IGHA2/C4A/CCL3/CXCL8/CCL8/TREM2/CD28
  Count
1    29
2    29
3    36
4    35
5    38
6    30</code></pre>
<p>Dot plot with enriched GO terms by gene list.</p>
<pre class="r"><code>dotplot(ego_bp_both, showCategory = 15)</code></pre>
<p><img src="figure/cluster_profiler.Rmd/dotplot_gene_list-1.png" width="1152" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-dotplot_gene_list-1">
Past versions of dotplot_gene_list-1.png
</button>
</p>
<div id="fig-dotplot_gene_list-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/davetang/muse/blob/18e8e9b7e30957ba8e89e2e3844d045616cb3a86/docs/figure/cluster_profiler.Rmd/dotplot_gene_list-1.png" target="_blank">18e8e9b</a>
</td>
<td>
Dave Tang
</td>
<td>
2025-05-26
</td>
</tr>
<tr>
<td>
<a href="https://github.com/davetang/muse/blob/e815348c6fc2adfec65202ff3718880a944b088c/docs/figure/cluster_profiler.Rmd/dotplot_gene_list-1.png" target="_blank">e815348</a>
</td>
<td>
Dave Tang
</td>
<td>
2025-03-06
</td>
</tr>
<tr>
<td>
<a href="https://github.com/davetang/muse/blob/2c080dce2c4547d90fe19cf649b4683a4aaa5142/docs/figure/cluster_profiler.Rmd/dotplot_gene_list-1.png" target="_blank">2c080dc</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-11-01
</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
</div>
<div id="enricher" class="section level2">
<h2>enricher</h2>
<p><code>enricher()</code> is a universal enrichment analyzer.</p>
<ul>
<li>gene - a vector of gene id</li>
<li>universe - background genes. If missing, the all genes listed in the
database (eg TERM2GENE table) will be used as background.</li>
<li>minGSSize - minimal size of genes annotated for testing</li>
<li>maxGSSize - maximal size of genes annotated for testing</li>
<li>TERM2GENE - user input annotation of TERM TO GENE mapping, a
data.frame of 2 column with term and gene. Only used when gson is
NULL.</li>
<li>TERM2NAME - user input of TERM TO NAME mapping, a data.frame of 2
column with term and name. Only used when gson is NULL.</li>
</ul>
<p><a href="https://support.bioconductor.org/p/9144621/">TERM2NAME</a>
is needed to map the GOID to its term (description).</p>
<p>An easy way of obtaining this mapping is by extracting the
information from the Bioconductor GO annotation database package
({GO.db}).</p>
<pre class="r"><code># extract a named vector of all terms
goterms &lt;- AnnotationDbi::Term(GOTERM)

#convert into a data frame
term2name &lt;- data.frame(
  &quot;term&quot;=names(goterms),
  &quot;name&quot;=goterms
)

dim(term2name)</code></pre>
<pre><code>[1] 40268     2</code></pre>
<pre class="r"><code>head(term2name)</code></pre>
<pre><code>                 term                                                     name
GO:0000001 GO:0000001                                mitochondrion inheritance
GO:0000002 GO:0000002                         mitochondrial genome maintenance
GO:0000006 GO:0000006    high-affinity zinc transmembrane transporter activity
GO:0000007 GO:0000007 low-affinity zinc ion transmembrane transporter activity
GO:0000009 GO:0000009                   alpha-1,6-mannosyltransferase activity
GO:0000010 GO:0000010                heptaprenyl diphosphate synthase activity</code></pre>
<p>Get the ontologies and store in <code>term2name</code>.</p>
<pre class="r"><code>ontologies &lt;- AnnotationDbi::select(x = GO.db, keys = names(goterms), columns = c(&quot;GOID&quot;, &quot;ONTOLOGY&quot;))</code></pre>
<pre><code>&#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<pre class="r"><code>length(unique(term2name$term))</code></pre>
<pre><code>[1] 40268</code></pre>
<pre class="r"><code>length(unique(ontologies$GOID))</code></pre>
<pre><code>[1] 40268</code></pre>
<pre class="r"><code>stopifnot(all(term2name$term == ontologies$GOID))

term2name$ontology &lt;- ontologies$ONTOLOGY
head(term2name)</code></pre>
<pre><code>                 term                                                     name
GO:0000001 GO:0000001                                mitochondrion inheritance
GO:0000002 GO:0000002                         mitochondrial genome maintenance
GO:0000006 GO:0000006    high-affinity zinc transmembrane transporter activity
GO:0000007 GO:0000007 low-affinity zinc ion transmembrane transporter activity
GO:0000009 GO:0000009                   alpha-1,6-mannosyltransferase activity
GO:0000010 GO:0000010                heptaprenyl diphosphate synthase activity
           ontology
GO:0000001       BP
GO:0000002       BP
GO:0000006       MF
GO:0000007       MF
GO:0000009       MF
GO:0000010       MF</code></pre>
<pre class="r"><code>any(is.na(term2name$term))</code></pre>
<pre><code>[1] FALSE</code></pre>
<p><code>term2name</code> can then be used when calling
<code>enrichr()</code>, by specifying <code>TERM2NAME=term2name</code>.
Be sure, though, to check that <code>term2name</code> contains all GOIDs
present in your TERM2GENE mapping.</p>
<p>Demo input.</p>
<pre class="r"><code>demo &lt;- readr::read_csv(&quot;data/nfurzeri_gene_id_to_go_id.csv.gz&quot;, show_col_types = FALSE)
head(demo)</code></pre>
<pre><code># A tibble: 6 × 2
  ensembl_gene_id    go_id     
  &lt;chr&gt;              &lt;chr&gt;     
1 ENSNFUG00015000040 &lt;NA&gt;      
2 ENSNFUG00015000041 GO:0007156
3 ENSNFUG00015000041 GO:0005886
4 ENSNFUG00015000041 GO:0005737
5 ENSNFUG00015000041 GO:0050808
6 ENSNFUG00015000041 GO:0007411</code></pre>
<p>Check if we have all the GO IDs.</p>
<pre class="r"><code>goids &lt;- demo$go_id
goids &lt;- goids[!is.na(goids)]

table(unique(goids) %in% term2name$term)</code></pre>
<pre><code>
FALSE  TRUE 
   48  6499 </code></pre>
<p>Which GO IDs are missing?</p>
<pre class="r"><code>missing_goids &lt;- setdiff(unique(goids), term2name$term)
missing_goids</code></pre>
<pre><code> [1] &quot;GO:0001054&quot; &quot;GO:0008272&quot; &quot;GO:0102769&quot; &quot;GO:0032508&quot; &quot;GO:0006211&quot;
 [6] &quot;GO:0006880&quot; &quot;GO:0014070&quot; &quot;GO:0097428&quot; &quot;GO:0043402&quot; &quot;GO:0045263&quot;
[11] &quot;GO:0045261&quot; &quot;GO:0030374&quot; &quot;GO:0004024&quot; &quot;GO:0044255&quot; &quot;GO:0106035&quot;
[16] &quot;GO:0003896&quot; &quot;GO:0097153&quot; &quot;GO:0001056&quot; &quot;GO:0001055&quot; &quot;GO:0046538&quot;
[21] &quot;GO:0061780&quot; &quot;GO:0090179&quot; &quot;GO:0097159&quot; &quot;GO:0051103&quot; &quot;GO:0004310&quot;
[26] &quot;GO:0006268&quot; &quot;GO:0035308&quot; &quot;GO:0042989&quot; &quot;GO:0042543&quot; &quot;GO:0090309&quot;
[31] &quot;GO:0052794&quot; &quot;GO:0052795&quot; &quot;GO:0052796&quot; &quot;GO:0071407&quot; &quot;GO:0034998&quot;
[36] &quot;GO:1901615&quot; &quot;GO:0022417&quot; &quot;GO:0102148&quot; &quot;GO:0097199&quot; &quot;GO:1901006&quot;
[41] &quot;GO:0005355&quot; &quot;GO:0046177&quot; &quot;GO:0046950&quot; &quot;GO:0002094&quot; &quot;GO:0003867&quot;
[46] &quot;GO:0045226&quot; &quot;GO:0060775&quot; &quot;GO:1901617&quot;</code></pre>
<p>They are missing because they have become obsolete.</p>
<pre class="r"><code>missing_goids %in% keys(GOOBSOLETE)</code></pre>
<pre><code> [1] TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE
[16] TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE
[31] TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE
[46] TRUE TRUE TRUE</code></pre>
<p>Prepare data frame of 2 column with term and gene.</p>
<pre class="r"><code>demo |&gt;
  dplyr::filter(!is.na(go_id)) |&gt;
  dplyr::rename(term = go_id, gene = ensembl_gene_id) |&gt;
  dplyr::select(term, gene) |&gt;
  dplyr::arrange(term) -&gt; term2gene

lookup &lt;- AnnotationDbi::select(x = GO.db, keys = term2gene$term, columns = c(&quot;GOID&quot;, &quot;ONTOLOGY&quot;))</code></pre>
<pre><code>&#39;select()&#39; returned many:1 mapping between keys and columns</code></pre>
<pre class="r"><code>stopifnot(all(lookup$GOID == term2gene$term))

term2gene$ontology &lt;- lookup$ONTOLOGY

head(term2gene)</code></pre>
<pre><code># A tibble: 6 × 3
  term       gene               ontology
  &lt;chr&gt;      &lt;chr&gt;              &lt;chr&gt;   
1 GO:0000002 ENSNFUG00015006992 BP      
2 GO:0000002 ENSNFUG00015020070 BP      
3 GO:0000002 ENSNFUG00015013825 BP      
4 GO:0000002 ENSNFUG00015019642 BP      
5 GO:0000009 ENSNFUG00015008531 MF      
6 GO:0000012 ENSNFUG00015012664 BP      </code></pre>
<p>Test with random genes.</p>
<pre class="r"><code>set.seed(1984)
my_genes &lt;- sample(x = unique(term2gene$gene), size = 100, replace = FALSE)

res &lt;- enricher(
  gene = my_genes,
  TERM2GENE = dplyr::filter(term2gene, ontology == &quot;BP&quot;),
  TERM2NAME = dplyr::filter(term2name, ontology == &quot;BP&quot;)
)

res</code></pre>
<pre><code>#
# over-representation test
#
#...@organism    UNKNOWN 
#...@ontology    UNKNOWN 
#...@gene    chr [1:100] &quot;ENSNFUG00015003803&quot; &quot;ENSNFUG00015000174&quot; &quot;ENSNFUG00015015379&quot; ...
#...pvalues adjusted by &#39;BH&#39; with cutoff &lt;0.05 
#...0 enriched terms found
#...Citation
S Xu, E Hu, Y Cai, Z Xie, X Luo, L Zhan, W Tang, Q Wang, B Liu, R Wang, W Xie, T Wu, L Xie, G Yu. Using clusterProfiler to characterize multiomics data. Nature Protocols. 2024, 19(11):3292-3320 </code></pre>
<p>Test with genes associated with same term.</p>
<pre class="r"><code>term2gene |&gt;
  dplyr::group_by(term) |&gt;
  dplyr::summarise(n = n()) |&gt;
  dplyr::arrange(-n) -&gt; dev_null

my_term &lt;- &quot;GO:0006397&quot;
dplyr::filter(term2name, term == my_term)</code></pre>
<pre><code>                 term            name ontology
GO:0006397 GO:0006397 mRNA processing       BP</code></pre>
<pre class="r"><code>term2gene |&gt;
  dplyr::filter(term == my_term) |&gt;
  dplyr::pull(gene) |&gt;
  head(75) -&gt; enriched_genes

set.seed(1984)
my_genes &lt;- sample(x = unique(term2gene$gene), size = 25, replace = FALSE)
my_genes &lt;- union(enriched_genes, my_genes)

res &lt;- enricher(
  gene = my_genes,
  TERM2GENE = dplyr::filter(term2gene, ontology == &quot;BP&quot;),
  TERM2NAME = dplyr::filter(term2name, ontology == &quot;BP&quot;)
)

res |&gt;
  as.data.frame()</code></pre>
<pre><code>                   ID                                              Description
GO:0006397 GO:0006397                                          mRNA processing
GO:0008380 GO:0008380                                             RNA splicing
GO:0000398 GO:0000398                           mRNA splicing, via spliceosome
GO:0000381 GO:0000381 regulation of alternative mRNA splicing, via spliceosome
GO:0000387 GO:0000387                              spliceosomal snRNP assembly
GO:0043484 GO:0043484                               regulation of RNA splicing
GO:0080090 GO:0080090                  regulation of primary metabolic process
GO:0030968 GO:0030968          endoplasmic reticulum unfolded protein response
           GeneRatio   BgRatio RichFactor FoldEnrichment    zScore
GO:0006397     75/92 125/12879  0.6000000       83.99348 79.087811
GO:0008380     43/92  89/12879  0.4831461       67.63520 53.505351
GO:0000398     21/92  71/12879  0.2957746       41.40524 28.957426
GO:0000381      7/92  20/12879  0.3500000       48.99620 18.220152
GO:0000387      3/92  12/12879  0.2500000       34.99728  9.993784
GO:0043484      3/92  14/12879  0.2142857       29.99767  9.207803
GO:0080090      2/92  12/12879  0.1666667       23.33152  6.564537
GO:0030968      2/92  14/12879  0.1428571       19.99845  6.032690
                  pvalue      p.adjust        qvalue
GO:0006397 5.819132e-146 4.306157e-144 3.736495e-144
GO:0008380  1.727409e-72  6.391414e-71  5.545893e-71
GO:0000398  3.075882e-29  7.587174e-28  6.583466e-28
GO:0000381  5.410001e-11  1.000850e-09  8.684475e-10
GO:0000387  7.407252e-05  1.096273e-03  9.512471e-04
GO:0043484  1.212924e-04  1.495940e-03  1.298042e-03
GO:0080090  3.179859e-03  3.361565e-02  2.916863e-02
GO:0030968  4.343815e-03  4.018029e-02  3.486483e-02
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     geneID
GO:0006397 ENSNFUG00015000047/ENSNFUG00015000176/ENSNFUG00015000433/ENSNFUG00015000717/ENSNFUG00015000995/ENSNFUG00015002878/ENSNFUG00015003278/ENSNFUG00015003412/ENSNFUG00015003645/ENSNFUG00015006374/ENSNFUG00015008466/ENSNFUG00015010592/ENSNFUG00015011476/ENSNFUG00015011629/ENSNFUG00015012572/ENSNFUG00015015893/ENSNFUG00015018395/ENSNFUG00015021027/ENSNFUG00015023111/ENSNFUG00015024532/ENSNFUG00015000276/ENSNFUG00015000347/ENSNFUG00015000470/ENSNFUG00015000633/ENSNFUG00015000679/ENSNFUG00015001197/ENSNFUG00015001330/ENSNFUG00015002687/ENSNFUG00015003553/ENSNFUG00015003613/ENSNFUG00015004037/ENSNFUG00015004391/ENSNFUG00015008722/ENSNFUG00015009124/ENSNFUG00015010820/ENSNFUG00015011580/ENSNFUG00015013273/ENSNFUG00015014123/ENSNFUG00015014399/ENSNFUG00015014448/ENSNFUG00015019426/ENSNFUG00015020527/ENSNFUG00015020545/ENSNFUG00015021436/ENSNFUG00015004299/ENSNFUG00015006219/ENSNFUG00015006538/ENSNFUG00015007623/ENSNFUG00015008859/ENSNFUG00015008871/ENSNFUG00015008954/ENSNFUG00015013147/ENSNFUG00015014506/ENSNFUG00015014956/ENSNFUG00015015816/ENSNFUG00015017783/ENSNFUG00015019674/ENSNFUG00015019903/ENSNFUG00015020693/ENSNFUG00015021220/ENSNFUG00015022748/ENSNFUG00015023101/ENSNFUG00015023143/ENSNFUG00015004762/ENSNFUG00015005084/ENSNFUG00015005095/ENSNFUG00015006742/ENSNFUG00015008749/ENSNFUG00015009426/ENSNFUG00015013451/ENSNFUG00015014267/ENSNFUG00015014760/ENSNFUG00015014763/ENSNFUG00015018080/ENSNFUG00015018929
GO:0008380                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 ENSNFUG00015000047/ENSNFUG00015000433/ENSNFUG00015002878/ENSNFUG00015003412/ENSNFUG00015008466/ENSNFUG00015010592/ENSNFUG00015011629/ENSNFUG00015012572/ENSNFUG00015015893/ENSNFUG00015018395/ENSNFUG00015021027/ENSNFUG00015023111/ENSNFUG00015000276/ENSNFUG00015000347/ENSNFUG00015000470/ENSNFUG00015001197/ENSNFUG00015002687/ENSNFUG00015004391/ENSNFUG00015009124/ENSNFUG00015010820/ENSNFUG00015011580/ENSNFUG00015013273/ENSNFUG00015014123/ENSNFUG00015014448/ENSNFUG00015019426/ENSNFUG00015020527/ENSNFUG00015008859/ENSNFUG00015008871/ENSNFUG00015013147/ENSNFUG00015014506/ENSNFUG00015019674/ENSNFUG00015019903/ENSNFUG00015020693/ENSNFUG00015021220/ENSNFUG00015022748/ENSNFUG00015023101/ENSNFUG00015005084/ENSNFUG00015008749/ENSNFUG00015014267/ENSNFUG00015014760/ENSNFUG00015014763/ENSNFUG00015018080/ENSNFUG00015018929
GO:0000398                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   ENSNFUG00015000433/ENSNFUG00015002878/ENSNFUG00015003412/ENSNFUG00015010592/ENSNFUG00015011629/ENSNFUG00015012572/ENSNFUG00015015893/ENSNFUG00015023111/ENSNFUG00015001197/ENSNFUG00015002687/ENSNFUG00015014123/ENSNFUG00015008871/ENSNFUG00015014506/ENSNFUG00015019903/ENSNFUG00015020693/ENSNFUG00015022748/ENSNFUG00015005084/ENSNFUG00015008749/ENSNFUG00015014267/ENSNFUG00015014760/ENSNFUG00015014763
GO:0000381                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                             ENSNFUG00015024532/ENSNFUG00015000470/ENSNFUG00015011580/ENSNFUG00015014123/ENSNFUG00015014448/ENSNFUG00015008859/ENSNFUG00015013147
GO:0000387                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         ENSNFUG00015010592/ENSNFUG00015018395/ENSNFUG00015019674
GO:0043484                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         ENSNFUG00015000995/ENSNFUG00015014448/ENSNFUG00015018080
GO:0080090                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            ENSNFUG00015011476/ENSNFUG00015021436
GO:0030968                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            ENSNFUG00015011476/ENSNFUG00015021436
           Count
GO:0006397    75
GO:0008380    43
GO:0000398    21
GO:0000381     7
GO:0000387     3
GO:0043484     3
GO:0080090     2
GO:0030968     2</code></pre>
<p>Check some of the results.</p>
<pre class="r"><code>go_term &lt;- &quot;GO:0007156&quot;

dplyr::filter(term2name, term == go_term)</code></pre>
<pre><code>                 term
GO:0007156 GO:0007156
                                                                      name
GO:0007156 homophilic cell adhesion via plasma membrane adhesion molecules
           ontology
GO:0007156       BP</code></pre>
<pre class="r"><code>dplyr::filter(term2gene, gene %in% my_genes, term == go_term) |&gt;
  nrow()</code></pre>
<pre><code>[1] 1</code></pre>
<pre class="r"><code>dplyr::filter(term2gene, term == go_term) |&gt;
  nrow()</code></pre>
<pre><code>[1] 130</code></pre>
<p>Dot plot.</p>
<pre class="r"><code>dotplot(res, showCategory=10) +
  ggtitle(&quot;Biological Processes&quot;)</code></pre>
<p><img src="figure/cluster_profiler.Rmd/dotplot_res-1.png" width="768" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-dotplot_res-1">
Past versions of dotplot_res-1.png
</button>
</p>
<div id="fig-dotplot_res-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/davetang/muse/blob/18e8e9b7e30957ba8e89e2e3844d045616cb3a86/docs/figure/cluster_profiler.Rmd/dotplot_res-1.png" target="_blank">18e8e9b</a>
</td>
<td>
Dave Tang
</td>
<td>
2025-05-26
</td>
</tr>
<tr>
<td>
<a href="https://github.com/davetang/muse/blob/006e824308df24e7f43c3d664e30d96617250941/docs/figure/cluster_profiler.Rmd/dotplot_res-1.png" target="_blank">006e824</a>
</td>
<td>
Dave Tang
</td>
<td>
2025-03-09
</td>
</tr>
<tr>
<td>
<a href="https://github.com/davetang/muse/blob/e815348c6fc2adfec65202ff3718880a944b088c/docs/figure/cluster_profiler.Rmd/dotplot_res-1.png" target="_blank">e815348</a>
</td>
<td>
Dave Tang
</td>
<td>
2025-03-06
</td>
</tr>
</tbody>
</table>
</div>
</div>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span>
Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 4.5.0 (2025-04-11)
Platform: x86_64-pc-linux-gnu
Running under: Ubuntu 24.04.3 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 
LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.26.so;  LAPACK version 3.12.0

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

time zone: Etc/UTC
tzcode source: system (glibc)

attached base packages:
[1] stats4    stats     graphics  grDevices utils     datasets  methods  
[8] base     

other attached packages:
 [1] GO.db_3.21.0           org.Hs.eg.db_3.21.0    AnnotationDbi_1.70.0  
 [4] IRanges_2.42.0         S4Vectors_0.46.0       Biobase_2.68.0        
 [7] BiocGenerics_0.54.0    generics_0.1.4         clusterProfiler_4.16.0
[10] ggarchery_0.4.4        lubridate_1.9.4        forcats_1.0.0         
[13] stringr_1.5.1          dplyr_1.1.4            purrr_1.0.4           
[16] readr_2.1.5            tidyr_1.3.1            tibble_3.3.0          
[19] ggplot2_3.5.2          tidyverse_2.0.0        workflowr_1.7.1       

loaded via a namespace (and not attached):
  [1] DBI_1.2.3               gson_0.1.0              rlang_1.1.6            
  [4] magrittr_2.0.3          DOSE_4.2.0              git2r_0.36.2           
  [7] compiler_4.5.0          RSQLite_2.4.2           getPass_0.2-4          
 [10] png_0.1-8               callr_3.7.6             vctrs_0.6.5            
 [13] reshape2_1.4.4          pkgconfig_2.0.3         crayon_1.5.3           
 [16] fastmap_1.2.0           XVector_0.48.0          labeling_0.4.3         
 [19] utf8_1.2.6              promises_1.3.3          rmarkdown_2.29         
 [22] tzdb_0.5.0              enrichplot_1.28.4       UCSC.utils_1.4.0       
 [25] ps_1.9.1                bit_4.6.0               xfun_0.52              
 [28] cachem_1.1.0            aplot_0.2.8             GenomeInfoDb_1.44.1    
 [31] jsonlite_2.0.0          blob_1.2.4              later_1.4.2            
 [34] BiocParallel_1.42.1     parallel_4.5.0          R6_2.6.1               
 [37] bslib_0.9.0             stringi_1.8.7           RColorBrewer_1.1-3     
 [40] jquerylib_0.1.4         GOSemSim_2.34.0         Rcpp_1.0.14            
 [43] knitr_1.50              ggtangle_0.0.7          R.utils_2.13.0         
 [46] igraph_2.1.4            httpuv_1.6.16           Matrix_1.7-3           
 [49] splines_4.5.0           timechange_0.3.0        tidyselect_1.2.1       
 [52] qvalue_2.40.0           rstudioapi_0.17.1       yaml_2.3.10            
 [55] codetools_0.2-20        curl_6.4.0              processx_3.8.6         
 [58] lattice_0.22-6          plyr_1.8.9              treeio_1.32.0          
 [61] withr_3.0.2             KEGGREST_1.48.1         evaluate_1.0.3         
 [64] gridGraphics_0.5-1      Biostrings_2.76.0       ggtree_3.16.3          
 [67] pillar_1.10.2           whisker_0.4.1           ggfun_0.2.0            
 [70] vroom_1.6.5             rprojroot_2.0.4         hms_1.1.3              
 [73] tidytree_0.4.6          scales_1.4.0            glue_1.8.0             
 [76] lazyeval_0.2.2          tools_4.5.0             data.table_1.17.4      
 [79] fgsea_1.34.2            fs_1.6.6                fastmatch_1.1-6        
 [82] cowplot_1.2.0           grid_4.5.0              ape_5.8-1              
 [85] nlme_3.1-168            patchwork_1.3.0         GenomeInfoDbData_1.2.14
 [88] cli_3.6.5               gtable_0.3.6            R.methodsS3_1.8.2      
 [91] yulab.utils_0.2.0       sass_0.4.10             digest_0.6.37          
 [94] ggrepel_0.9.6           ggplotify_0.1.2         farver_2.1.2           
 [97] memoise_2.0.1           htmltools_0.5.8.1       R.oo_1.27.1            
[100] lifecycle_1.0.4         httr_1.4.7              bit64_4.6.0-1          </code></pre>
</div>
</div>
</div>
</div>
</div>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
https://docs.mathjax.org/en/latest/web/configuration.html. This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>




</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
