<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />



<meta name="date" content="2024-08-06" />

<title>Differential gene expression analysis with DESeq2</title>

<script src="site_libs/header-attrs-2.27/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.13.2/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-6.4.2/css/all.min.css" rel="stylesheet" />
<link href="site_libs/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet" />

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/main/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">My blog</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/davetang/muse">
    <span class="fa fa-github"></span>
     
    Source code
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Differential gene expression analysis with
DESeq2</h1>
<h4 class="date">2024-08-06</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span>
workflowr <span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2024-08-06
</p>
<p>
<strong>Checks:</strong> <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 7
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong> <code>muse/</code> <span
class="glyphicon glyphicon-question-sign" aria-hidden="true"
title="This is the local directory in which the code in this file was executed.">
</span>
</p>
<p>
This reproducible <a href="https://rmarkdown.rstudio.com">R Markdown</a>
analysis was created with <a
  href="https://github.com/workflowr/workflowr">workflowr</a> (version
1.7.1). The <em>Checks</em> tab describes the reproducibility checks
that were applied when the results were created. The <em>Past
versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date
</a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git
repository, you know the exact version of the code that produced these
results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the
global environment can affect the analysis in your R Markdown file in
unknown ways. For reproduciblity it’s best to always run the code in an
empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20200712code">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Seed:</strong>
<code>set.seed(20200712)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20200712code"
class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20200712)</code> was run prior to running
the code in the R Markdown file. Setting a seed ensures that any results
that rely on randomness, e.g. subsampling or permutations, are
reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Session information:</strong>
recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package
versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be
confident that you successfully produced the results during this
run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr
project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomdavetangmusetree665782bb8465a8773d6924849ed75eae1772f1b5targetblank665782ba">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Repository version:</strong>
<a href="https://github.com/davetang/muse/tree/665782bb8465a8773d6924849ed75eae1772f1b5" target="_blank">665782b</a>
</a>
</p>
</div>
<div
id="strongRepositoryversionstrongahrefhttpsgithubcomdavetangmusetree665782bb8465a8773d6924849ed75eae1772f1b5targetblank665782ba"
class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development
and connecting the code version to the results is critical for
reproducibility.
</p>
<p>
The results in this page were generated with repository version
<a href="https://github.com/davetang/muse/tree/665782bb8465a8773d6924849ed75eae1772f1b5" target="_blank">665782b</a>.
See the <em>Past versions</em> tab to see a history of the changes made
to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for
the analysis have been committed to Git prior to generating the results
(you can use <code>wflow_publish</code> or
<code>wflow_git_commit</code>). workflowr only checks the R Markdown
file, but you know if there are other scripts or data files that it
depends on. Below is the status of the Git repository when the results
were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .Rhistory
    Ignored:    .Rproj.user/
    Ignored:    r_packages_4.3.3/
    Ignored:    r_packages_4.4.0/

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not
included in this status report because it is ok for generated content to
have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">
<p>
These are the previous versions of the repository in which changes were
made to the R Markdown (<code>analysis/deseq2.Rmd</code>) and HTML
(<code>docs/deseq2.html</code>) files. If you’ve configured a remote Git
repository (see <code>?wflow_git_remote</code>), click on the hyperlinks
in the table below to view the files as they were in that past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/davetang/muse/blob/665782bb8465a8773d6924849ed75eae1772f1b5/analysis/deseq2.Rmd" target="_blank">665782b</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-08-06
</td>
<td>
Noise and replicates
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/davetang/muse/3a1d43a87502d6cefd79f879373e86fd32928482/docs/deseq2.html" target="_blank">3a1d43a</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-08-06
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/davetang/muse/blob/3738c6f4da763bc34b6e6aff1bdd004b1fca040d/analysis/deseq2.Rmd" target="_blank">3738c6f</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-08-06
</td>
<td>
Multifactorial design
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/davetang/muse/f4da2ce124efa9c5771f1c7ea1779f7073531f6a/docs/deseq2.html" target="_blank">f4da2ce</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-08-05
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/davetang/muse/blob/9293da5953964d919bbacd4c01f53c4674f6e8e7/analysis/deseq2.Rmd" target="_blank">9293da5</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-08-05
</td>
<td>
Visualisations after a DGE analysis
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/davetang/muse/365890ed39ee3f688897840df5742a07f4cfca27/docs/deseq2.html" target="_blank">365890e</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-08-05
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/davetang/muse/blob/3780bfab5474f613432aac97371b9b7e50b22f4d/analysis/deseq2.Rmd" target="_blank">3780bfa</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-08-05
</td>
<td>
A basic DESeq2 analysis
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/davetang/muse/1383bf704e651ba4ae4e663988680c68099b3262/docs/deseq2.html" target="_blank">1383bf7</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-08-05
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/davetang/muse/blob/d5e67bcb4bfb2c2c64e7f45cd6d750551f5ba11d/analysis/deseq2.Rmd" target="_blank">d5e67bc</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-08-05
</td>
<td>
Getting started with DESeq2
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<p><a
href="https://bioconductor.org/packages/release/bioc/html/DESeq2.html">DESeq2</a>
is used to:</p>
<blockquote>
<p>Estimate variance-mean dependence in count data from high-throughput
sequencing assays and test for differential expression based on a model
using the negative binomial distribution.</p>
</blockquote>
<div id="installation" class="section level2">
<h2>Installation</h2>
<p>Install using <code>BiocManager::install()</code>.</p>
<pre class="r"><code>if (!require(&quot;BiocManager&quot;, quietly = TRUE))
    install.packages(&quot;BiocManager&quot;)

BiocManager::install(&quot;DESeq2&quot;)</code></pre>
<p>We will use data from the <a
href="https://www.bioconductor.org/packages/release/data/experiment/html/pasilla.html">pasilla
package</a> so install it too.</p>
<pre class="r"><code>BiocManager::install(&quot;pasilla&quot;)</code></pre>
</div>
<div id="example-data" class="section level2">
<h2>Example data</h2>
<p>Example dataset in the experiment data package {pasilla}.</p>
<pre class="r"><code>fn &lt;- system.file(&quot;extdata&quot;, &quot;pasilla_gene_counts.tsv&quot;, package = &quot;pasilla&quot;, mustWork = TRUE)
counts &lt;- as.matrix(read.csv(fn, sep=&quot;\t&quot;, row.names = &quot;gene_id&quot;))
dim(counts)</code></pre>
<pre><code>[1] 14599     7</code></pre>
<p>The matrix tallies the number of reads assigned for each gene in each
sample.</p>
<pre class="r"><code>tail(counts)</code></pre>
<pre><code>            untreated1 untreated2 untreated3 untreated4 treated1 treated2
FBgn0261570       3296       4910       2156       2060     5077     3069
FBgn0261571          0          0          0          0        1        0
FBgn0261572          4         13          4         11        7        3
FBgn0261573       2651       3653       1571       1612     3334     1848
FBgn0261574       6385       9318       3110       2819    10455     3508
FBgn0261575          6         53          1          3       42        3
            treated3
FBgn0261570     3022
FBgn0261571        0
FBgn0261572        3
FBgn0261573     1908
FBgn0261574     3047
FBgn0261575        4</code></pre>
</div>
<div id="size-factors" class="section level2">
<h2>Size factors</h2>
<p>Estimate size factors.</p>
<pre class="r"><code>DESeq2::estimateSizeFactorsForMatrix(counts)</code></pre>
<pre><code>untreated1 untreated2 untreated3 untreated4   treated1   treated2   treated3 
 1.1382630  1.7930004  0.6495470  0.7516892  1.6355751  0.7612698  0.8326526 </code></pre>
</div>
<div id="mean-variance-relationship" class="section level2">
<h2>Mean-variance relationship</h2>
<p>Variance versus mean for the (size factor adjusted)
<code>counts</code> data. The axes are logarithmic. Also shown are lines
through the origin with slopes 1 (green) and 2 (red).</p>
<pre class="r"><code>sf &lt;- DESeq2::estimateSizeFactorsForMatrix(counts)

ncounts &lt;- t(t(counts) / sf)

# untreated samples
uncounts &lt;- ncounts[, grep(&quot;^untreated&quot;, colnames(ncounts)), drop = FALSE]

ggplot(
  tibble(
    mean = rowMeans(uncounts),
    var  = rowVars(uncounts)
  ),
  aes(x = log1p(mean), y = log1p(var))
) +
  geom_hex() +
  coord_fixed() +
  theme_minimal() +
  theme(legend.position = &quot;none&quot;) +
  geom_abline(slope = 1:2, color = c(&quot;forestgreen&quot;, &quot;red&quot;))</code></pre>
<p><img src="figure/deseq2.Rmd/mean_variance-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-mean_variance-1">
Past versions of mean_variance-1.png
</button>
</p>
<div id="fig-mean_variance-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/davetang/muse/blob/1383bf704e651ba4ae4e663988680c68099b3262/docs/figure/deseq2.Rmd/mean_variance-1.png" target="_blank">1383bf7</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-08-05
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>The green line is what we expect if the variance equals the mean, as
is the case for a Poisson-distributed random variable. This
approximately fits the data in the lower range. The red line corresponds
to the quadratic mean-variance relationship <span
class="math inline">\(v = m^2\)</span>. We can see that in the upper
range of the data, the quadratic relationship approximately fits the
data.</p>
</div>
<div id="a-basic-analysis" class="section level2">
<h2>A basic analysis</h2>
<p>The {pasilla} data set is from an experiment on <em>Drosophila
melanogaster</em> cell cultures that tested the effect of RNAi knockdown
of the splicing factor <em>pasilla</em> on the cells’ transcriptome.
There were two experimental conditions, termed <em>untreated</em> and
<em>treated</em> in the column headers and they correspond to negative
control and siRNA against <em>pasilla</em>. The experimental metadata of
the seven samples in this dataset are loaded below.</p>
<pre class="r"><code>annotationFile &lt;- system.file(&quot;extdata&quot;, &quot;pasilla_sample_annotation.csv&quot;, package = &quot;pasilla&quot;, mustWork = TRUE)
pasillaSampleAnno &lt;- readr::read_csv(annotationFile)</code></pre>
<pre><code>Rows: 7 Columns: 6
── Column specification ────────────────────────────────────────────────────────
Delimiter: &quot;,&quot;
chr (4): file, condition, type, total number of reads
dbl (2): number of lanes, exon counts

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.</code></pre>
<pre class="r"><code>pasillaSampleAnno</code></pre>
<pre><code># A tibble: 7 × 6
  file    condition type  `number of lanes` total number of read…¹ `exon counts`
  &lt;chr&gt;   &lt;chr&gt;     &lt;chr&gt;             &lt;dbl&gt; &lt;chr&gt;                          &lt;dbl&gt;
1 treate… treated   sing…                 5 35158667                    15679615
2 treate… treated   pair…                 2 12242535 (x2)               15620018
3 treate… treated   pair…                 2 12443664 (x2)               12733865
4 untrea… untreated sing…                 2 17812866                    14924838
5 untrea… untreated sing…                 6 34284521                    20764558
6 untrea… untreated pair…                 2 10542625 (x2)               10283129
7 untrea… untreated pair…                 2 12214974 (x2)               11653031
# ℹ abbreviated name: ¹​`total number of reads`</code></pre>
<p>The dataset was produced in two batches, the first consisting of
three sequencing libraries using single read sequencing and the second
using using paired-end sequencing.</p>
<p>Create factors.</p>
<pre class="r"><code>mutate(
  pasillaSampleAnno,
  condition = factor(condition, levels = c(&quot;untreated&quot;, &quot;treated&quot;)),
  type = factor(sub(&quot;-.*&quot;, &quot;&quot;, type), levels = c(&quot;single&quot;, &quot;paired&quot;))
) -&gt; pasillaSampleAnno</code></pre>
<p>Note that the design is approximately balanced between the factor of
interest, <code>condition</code>, and the “nuisance factor”,
<code>type</code>.</p>
<pre class="r"><code>with(pasillaSampleAnno, table(condition, type))</code></pre>
<pre><code>           type
condition   single paired
  untreated      2      2
  treated        1      2</code></pre>
<p>{DESeq2} uses <code>DESeqDataSet</code> objects and is an extension
of the <code>SummarizedExperiment</code> class in Bioconductor. We can
create a <code>DESeqDataSet</code> object using the constructor function
<code>DESeqDataSetFromMatrix</code>.</p>
<pre class="r"><code>mt &lt;- match(colnames(counts), sub(&quot;fb$&quot;, &quot;&quot;, pasillaSampleAnno$file))
stopifnot(!any(is.na(mt)))

pasilla &lt;- DESeqDataSetFromMatrix(
  countData = counts,
  colData   = pasillaSampleAnno[mt, ],
  design    = ~ condition
)

class(pasilla)</code></pre>
<pre><code>[1] &quot;DESeqDataSet&quot;
attr(,&quot;package&quot;)
[1] &quot;DESeq2&quot;</code></pre>
<pre class="r"><code>is(pasilla, &quot;SummarizedExperiment&quot;)</code></pre>
<pre><code>[1] TRUE</code></pre>
<p>After creating a <code>DESeqDataSet</code> we are ready to carry out
a differential expression analysis. The aim is to identify genes that
are differentially abundant between the treated and untreated cells. A
test that is conceptually similar to the <em>t</em>-test is used. A
choice of standard analysis steps are wrapped into a single function,
<code>DESeq()</code>.</p>
<pre class="r"><code>pasilla &lt;- DESeq(pasilla)</code></pre>
<pre><code>estimating size factors</code></pre>
<pre><code>estimating dispersions</code></pre>
<pre><code>gene-wise dispersion estimates</code></pre>
<pre><code>mean-dispersion relationship</code></pre>
<pre><code>final dispersion estimates</code></pre>
<pre><code>fitting model and testing</code></pre>
<p>The <code>DESeq()</code> function is a wrapper that calls:</p>
<ul>
<li><code>estimateSizeFactors</code> for normalisation</li>
<li><code>estimateDispersions</code> for dispersion estimation</li>
<li><code>nbinomWaldTest</code> for hypothesis tests for differential
abundance</li>
</ul>
<p>The test is between the two levels <code>untreated</code> and
<code>treated</code> of the factor <code>condition</code>, since this is
what was specified in the design: <code>design = ~ condition</code>.</p>
<pre class="r"><code>res &lt;- results(pasilla)
res[order(res$padj), ] |&gt; head()</code></pre>
<pre><code>log2 fold change (MLE): condition treated vs untreated 
Wald test p-value: condition treated vs untreated 
DataFrame with 6 rows and 6 columns
             baseMean log2FoldChange     lfcSE      stat       pvalue
            &lt;numeric&gt;      &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;    &lt;numeric&gt;
FBgn0039155   730.596       -4.61901 0.1687068  -27.3789 4.88599e-165
FBgn0025111  1501.411        2.89986 0.1269205   22.8479 1.53430e-115
FBgn0029167  3706.117       -2.19700 0.0969888  -22.6521 1.33042e-113
FBgn0003360  4343.035       -3.17967 0.1435264  -22.1539 9.56283e-109
FBgn0035085   638.233       -2.56041 0.1372952  -18.6490  1.28772e-77
FBgn0039827   261.916       -4.16252 0.2325888  -17.8965  1.25663e-71
                    padj
               &lt;numeric&gt;
FBgn0039155 4.06661e-161
FBgn0025111 6.38497e-112
FBgn0029167 3.69104e-110
FBgn0003360 1.98979e-105
FBgn0035085  2.14354e-74
FBgn0039827  1.74316e-68</code></pre>
<div id="exploring-the-results" class="section level3">
<h3>Exploring the results</h3>
<p>The first step after a differential expression analysis is to
visualise the following plots:</p>
<ul>
<li>the histogram of p-values</li>
<li>the MA plot</li>
<li>an ordination plot</li>
<li>a heatmap</li>
</ul>
<p>Histogram of p-values of a differential expression analysis.”</p>
<pre class="r"><code>ggplot(as(res, &quot;data.frame&quot;), aes(x = pvalue)) +
  geom_histogram(binwidth = 0.01, fill = &quot;Royalblue&quot;, boundary = 0) +
  geom_hline(yintercept = 100, lty = 3) +
  theme_minimal()</code></pre>
<pre><code>Warning: Removed 2241 rows containing non-finite outside the scale range
(`stat_bin()`).</code></pre>
<p><img src="figure/deseq2.Rmd/pvalue_hist-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-pvalue_hist-1">
Past versions of pvalue_hist-1.png
</button>
</p>
<div id="fig-pvalue_hist-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/davetang/muse/blob/f4da2ce124efa9c5771f1c7ea1779f7073531f6a/docs/figure/deseq2.Rmd/pvalue_hist-1.png" target="_blank">f4da2ce</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-08-05
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>The p-value distribution shows a uniform background (non-DE genes)
with values between 0 and 1 and a peak of small p-values at the left (DE
genes). The ratio of the level of the background to the height of the
peak provides us with a rough indication of the false discovery rate
(FDR) that would be associated with calling the genes in the leftmost
bin differentially expressed.</p>
<p>In this case, the leftmost bin contains 993 genes and the background
level (dotted line) is around 100, so the FDR associated with calling
all genes in the leftmost bin would be around 10%.</p>
<pre class="r"><code>table(res$pvalue &lt; 0.01)</code></pre>
<pre><code>
FALSE  TRUE 
11365   993 </code></pre>
<p>A background distribution that is not uniform, but shows a tilted
shape with an increase towards the right, tends to indicate batch
effects, that is, underlying systematic variation that makes replicates
look more different than expected.</p>
<p>The {DESeq2} package can produce the MA plot.</p>
<pre class="r"><code>plotMA(pasilla, ylim = c(-2, 2))</code></pre>
<p><img src="figure/deseq2.Rmd/plot_ma-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-plot_ma-1">
Past versions of plot_ma-1.png
</button>
</p>
<div id="fig-plot_ma-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/davetang/muse/blob/f4da2ce124efa9c5771f1c7ea1779f7073531f6a/docs/figure/deseq2.Rmd/plot_ma-1.png" target="_blank">f4da2ce</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-08-05
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>{DESeq2} also has a function <code>plotPCA()</code> that can produce
PCA plots. PC1 is mostly aligned with the experimental covariate of
interest (<code>untreated</code>/<code>treated</code>), while PC2 is
roughly aligned with the sequencing protocol
(<code>single</code>/<code>paired</code>).</p>
<pre class="r"><code># transform using regularised logarithm
pas_rlog &lt;- rlogTransformation(pasilla)
plotPCA(pas_rlog, intgroup=c(&quot;condition&quot;, &quot;type&quot;)) + coord_fixed()</code></pre>
<pre><code>using ntop=500 top features by variance</code></pre>
<pre><code>Coordinate system already present. Adding new coordinate system, which will
replace the existing one.</code></pre>
<p><img src="figure/deseq2.Rmd/deseq2_pca-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-deseq2_pca-1">
Past versions of deseq2_pca-1.png
</button>
</p>
<div id="fig-deseq2_pca-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/davetang/muse/blob/f4da2ce124efa9c5771f1c7ea1779f7073531f6a/docs/figure/deseq2.Rmd/deseq2_pca-1.png" target="_blank">f4da2ce</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-08-05
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>Heatmap using the regularised logarithm (rlog) transformed data of
the 30 most variable genes.</p>
<pre class="r"><code>library(&quot;pheatmap&quot;)
wanted &lt;- order(rowMeans(assay(pas_rlog)), decreasing = TRUE)[1:30]

pheatmap(
  assay(pas_rlog)[wanted, ],
  scale = &quot;row&quot;,
  annotation_col = as.data.frame(
    colData(pas_rlog)[, c(&quot;condition&quot;, &quot;type&quot;)]
  )
)</code></pre>
<p><img src="figure/deseq2.Rmd/pheatmap-1.png" width="768" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-pheatmap-1">
Past versions of pheatmap-1.png
</button>
</p>
<div id="fig-pheatmap-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/davetang/muse/blob/f4da2ce124efa9c5771f1c7ea1779f7073531f6a/docs/figure/deseq2.Rmd/pheatmap-1.png" target="_blank">f4da2ce</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-08-05
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>Note the clustering of the sequencing protocol and this highlights
that our differential expression analysis was probably too naive and
that we should adjust for the “nuisance” factor.</p>
</div>
</div>
<div id="multifactorial-designs-and-linear-models"
class="section level2">
<h2>Multifactorial designs and linear models</h2>
<p>Let’s assume that, in addition to the siRNA knockdown of the pasilla
gene, we also want to test the effect of a certain drug. We would then
envisage an experiment in which the experimenter treats the cells either
with negative control, with the siRNA against pasilla, with the drug, or
with both. To analyse this experiment, we can introduce the
notation:</p>
<p><span class="math display">\[\begin{equation}
   y = \beta_0 + x_1 \beta_1 + x_2 \beta_2 + x_1 x_2 \beta_{12}
\end{equation}\]</span></p>
<p>The left-hand size, <span class="math inline">\(y\)</span>, is the
experimental measurement of interest, which in our case is the suitably
transformed expression level of a gene. Since in an RNA-seq experiment
there are lots of genes, we’ll have as many copies of the above
equation, one for each.</p>
<p>The coefficient <span class="math inline">\(\beta_0\)</span> is the
base level of the measurement in the negative control; often it is
called the <strong>intercept</strong>.</p>
<p>The design factors <span class="math inline">\(x_1\)</span> and <span
class="math inline">\(x_2\)</span> are binary indicator variables: <span
class="math inline">\(x_1\)</span> takes the value 1 if the siRNA was
transfected and 0 if not, and similarly, <span
class="math inline">\(x_2\)</span> indicates whether the drug was
administered.</p>
<p>In the experiment where only where only the siRNA is used, <span
class="math inline">\(x_1\)</span> = 1 and <span
class="math inline">\(x_2\)</span> = 0, and the third and fourth terms
of the equation vanish. The equation then simplifies to <span
class="math inline">\(y = \beta_0 + \beta_1\)</span>. This means that
<span class="math inline">\(\beta_1\)</span> represents the difference
between treatment and control. If our measurements are on a logarithmic
scale, then:</p>
<p><span class="math display">\[
y = \beta_0 + \beta_1 \\
\beta_1 = y - \beta_0 = log_2(expression_{treated}) -
log2(expression_{untreated}) \\
= log_2\frac{expression_{treated}}{expression_{untreated}}
\]</span></p>
<p>is the logarithmic fold change due to treatment with the siRNA. In
exactly the same way, <span class="math inline">\(\beta_2\)</span> is
the logarithmic fold change due to treatment with the drug.</p>
<p>If we treat the cells with both siRNA and drug, <span
class="math inline">\(x_1\)</span> = <span
class="math inline">\(x_2\)</span> = 1:</p>
<p><span class="math display">\[
y = \beta_0 + x_1 \beta_1 + x_2 \beta_2 + x_1 x_2 \beta_{12} \\
y = \beta_0 + \beta_1 + \beta_2 + \beta_{12} \\
\beta_{12} = y - (\beta_0 + \beta_1 + \beta_2)
\]</span></p>
<p>This means that <span class="math inline">\(\beta_{12}\)</span> is
the difference between the observed outcome, <span
class="math inline">\(y\)</span>, and the outcome expected from the
individual treatments, obtained by adding to the baseline the effect of
siRNA alone, <span class="math inline">\(\beta_1\)</span>, and of drug
alone, <span class="math inline">\(\beta_2\)</span>.</p>
<p>We call <span class="math inline">\(\beta_{12}\)</span> the
<em>interaction</em> effect of siRNA and drug. It has nothing to do with
a physical interaction: the terminology indicates that the effects of
these two different experimental factors do not simply add up, but
combine in a more complicated fashion. For instance, if the target of
the drug and of the siRNA were equivalent, leading to the same effect on
the cells, then we would biologically expect that <span
class="math inline">\(\beta_1\)</span> = <span
class="math inline">\(\beta_2\)</span>. We would also expect that their
combination has no further effect, so that <span
class="math inline">\(\beta_{12}\)</span> = <span
class="math inline">\(-\beta_1\)</span>.</p>
<p>If, on the other hand, the targets of the drug and of the siRNA are
in parallel pathways that can buffer each other, we would expect that
<span class="math inline">\(\beta_1\)</span> and <span
class="math inline">\(\beta_2\)</span> are both relatively small, but
the combined effect is synergistic, and <span
class="math inline">\(\beta_{12}\)</span> is large.</p>
<p>We don’t always care about interactions. Many experiments are
designed with multiple factors where we care most about each of their
individual effects. In that case, the combinatorial treatment might not
be present in the experimental design, and the model to use for the
analysis is a version of the equation with the rightmost term
removed.</p>
<p>We can succinctly encode the design of the experiment in the
<strong>design matrix</strong>. The combinatorial experiment would
be:</p>
<table>
<thead>
<tr class="header">
<th align="center"><span class="math inline">\(x_0\)</span></th>
<th align="center"><span class="math inline">\(x_1\)</span></th>
<th align="center"><span class="math inline">\(x_2\)</span></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="center">1</td>
<td align="center">0</td>
<td align="center">0</td>
</tr>
<tr class="even">
<td align="center">1</td>
<td align="center">1</td>
<td align="center">0</td>
</tr>
<tr class="odd">
<td align="center">1</td>
<td align="center">0</td>
<td align="center">1</td>
</tr>
<tr class="even">
<td align="center">1</td>
<td align="center">1</td>
<td align="center">1</td>
</tr>
</tbody>
</table>
<p>The columns of the design matrix correspond to the experimental
factors, and its rows represent the different experimental conditions,
four in this case. If the combinatorial treatment is not performed, then
the design matrix is reduced to only the first three rows.</p>
</div>
<div id="noise-and-replicates" class="section level2">
<h2>Noise and replicates</h2>
<p>The equation:</p>
<p><span class="math display">\[\begin{equation}
   y = \beta_0 + x_1 \beta_1 + x_2 \beta_2 + x_1 x_2 \beta_{12}
\end{equation}\]</span></p>
<p>provides a conceptual decomposition of the observed data into the
effects caused by the different experimental variables. If our data (the
<span class="math inline">\(y\)</span>s) were absolutely precise, we
could set up a linear system of equations, one equation for each of the
four possible experimental conditions represented by the <span
class="math inline">\(x\)</span>s, and solve for the <span
class="math inline">\(\beta\)</span>s.</p>
<p>However real data is usually affected by noise and we need replicates
to estimate the levels of noise and assess the uncertainty of our
estimated <span class="math inline">\(\beta\)</span>s. Only then can we
empirically assess whether any of the observed changes between
conditions are significantly larger than those occurring due just to
experimental or natural variation. We extend the equation to:</p>
<p><span class="math display">\[
y_j = x_{j0} \beta_0 + x_{j1} \beta_1 + x_{j2} \beta_2 + x_{j1} x_{j2}
\beta_{12} + \epsilon_j
\]</span></p>
<p>We have included the index <span class="math inline">\(j\)</span> and
a new term <span class="math inline">\(\epsilon_j\)</span>. The index
<span class="math inline">\(j\)</span> now explicitly counts over our
individual replicate experiments; for example, if for each of the four
conditions we perform three replicates, then <span
class="math inline">\(j\)</span> counts from 1 to 12. The design matrix
now has 12 rows, and <span class="math inline">\(x_{jk}\)</span> is the
value of the matrix in its <span class="math inline">\(j\)</span>th row
and <span class="math inline">\(k\)</span>th column.</p>
<p>The additional terms <span class="math inline">\(\epsilon_j\)</span>,
which we call the <strong>residuals</strong>, are there to absorb
differences between replicates. However, one additional modeling
component is needed: the system of 12 equations would be underdetermined
without further information, since it now has more variables (12 <span
class="math inline">\(\epsilon\)</span>s and four <span
class="math inline">\(\beta\)</span>s) than it has equations (12, one
for each <span class="math inline">\(j\)</span>). To address this, we
require that the <span class="math inline">\(\epsilon_j\)</span> be
small. One popular way to overcome this is to minimise the sum of
squared residuals,</p>
<p><span class="math display">\[
\sum_j \epsilon^2_j \rightarrow min.
\]</span></p>
<p>It turns out that with this requirement satisfied, the <span
class="math inline">\(\beta\)</span>s represent the <em>average</em>
effects of each of the experimental factors, while the residuals <span
class="math inline">\(\epsilon_j\)</span> reflect the experimental
fluntuations around the mean between the replicates. This approach,
which is called <strong>least sum of squares fitting</strong>, is
mathematically convenient, since it can be achieved by straight-forward
matrix algebra and it is what the <code>lm()</code> function does.</p>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span>
Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 4.4.0 (2024-04-24)
Platform: x86_64-pc-linux-gnu
Running under: Ubuntu 22.04.4 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 
LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

time zone: Etc/UTC
tzcode source: system (glibc)

attached base packages:
[1] stats4    stats     graphics  grDevices utils     datasets  methods  
[8] base     

other attached packages:
 [1] pheatmap_1.0.12             hexbin_1.28.3              
 [3] pasilla_1.32.0              DEXSeq_1.50.0              
 [5] RColorBrewer_1.1-3          AnnotationDbi_1.66.0       
 [7] BiocParallel_1.38.0         DESeq2_1.44.0              
 [9] SummarizedExperiment_1.34.0 Biobase_2.64.0             
[11] MatrixGenerics_1.16.0       matrixStats_1.3.0          
[13] GenomicRanges_1.56.1        GenomeInfoDb_1.40.1        
[15] IRanges_2.38.1              S4Vectors_0.42.1           
[17] BiocGenerics_0.50.0         lubridate_1.9.3            
[19] forcats_1.0.0               stringr_1.5.1              
[21] dplyr_1.1.4                 purrr_1.0.2                
[23] readr_2.1.5                 tidyr_1.3.1                
[25] tibble_3.2.1                ggplot2_3.5.1              
[27] tidyverse_2.0.0             workflowr_1.7.1            

loaded via a namespace (and not attached):
  [1] rstudioapi_0.16.0       jsonlite_1.8.8          magrittr_2.0.3         
  [4] farver_2.1.2            rmarkdown_2.27          fs_1.6.4               
  [7] zlibbioc_1.50.0         vctrs_0.6.5             memoise_2.0.1          
 [10] Rsamtools_2.20.0        htmltools_0.5.8.1       S4Arrays_1.4.1         
 [13] progress_1.2.3          curl_5.2.1              SparseArray_1.4.8      
 [16] sass_0.4.9              bslib_0.7.0             httr2_1.0.1            
 [19] cachem_1.1.0            whisker_0.4.1           lifecycle_1.0.4        
 [22] pkgconfig_2.0.3         Matrix_1.7-0            R6_2.5.1               
 [25] fastmap_1.2.0           GenomeInfoDbData_1.2.12 digest_0.6.35          
 [28] colorspace_2.1-0        ps_1.7.6                rprojroot_2.0.4        
 [31] geneplotter_1.82.0      RSQLite_2.3.7           hwriter_1.3.2.1        
 [34] filelock_1.0.3          labeling_0.4.3          fansi_1.0.6            
 [37] timechange_0.3.0        httr_1.4.7              abind_1.4-5            
 [40] compiler_4.4.0          bit64_4.0.5             withr_3.0.0            
 [43] DBI_1.2.3               highr_0.11              biomaRt_2.60.1         
 [46] rappdirs_0.3.3          DelayedArray_0.30.1     tools_4.4.0            
 [49] httpuv_1.6.15           glue_1.7.0              callr_3.7.6            
 [52] promises_1.3.0          grid_4.4.0              getPass_0.2-4          
 [55] generics_0.1.3          gtable_0.3.5            tzdb_0.4.0             
 [58] hms_1.1.3               xml2_1.3.6              utf8_1.2.4             
 [61] XVector_0.44.0          pillar_1.9.0            vroom_1.6.5            
 [64] genefilter_1.86.0       later_1.3.2             splines_4.4.0          
 [67] BiocFileCache_2.12.0    lattice_0.22-6          survival_3.5-8         
 [70] bit_4.0.5               annotate_1.82.0         tidyselect_1.2.1       
 [73] locfit_1.5-9.9          Biostrings_2.72.1       knitr_1.47             
 [76] git2r_0.33.0            xfun_0.44               statmod_1.5.0          
 [79] stringi_1.8.4           UCSC.utils_1.0.0        yaml_2.3.8             
 [82] evaluate_0.24.0         codetools_0.2-20        cli_3.6.2              
 [85] xtable_1.8-4            munsell_0.5.1           processx_3.8.4         
 [88] jquerylib_0.1.4         Rcpp_1.0.12             dbplyr_2.5.0           
 [91] png_0.1-8               XML_3.99-0.16.1         parallel_4.4.0         
 [94] blob_1.2.4              prettyunits_1.2.0       bitops_1.0-7           
 [97] scales_1.3.0            crayon_1.5.2            rlang_1.1.4            
[100] KEGGREST_1.44.1        </code></pre>
</div>
</div>
</div>
</div>
</div>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
https://docs.mathjax.org/en/latest/web/configuration.html. This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>




</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
