<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />



<meta name="date" content="2026-01-16" />

<title>Diversity metrics</title>

<script src="site_libs/header-attrs-2.29/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.13.2/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-6.5.2/css/all.min.css" rel="stylesheet" />
<link href="site_libs/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet" />

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/main/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">My blog</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/davetang/muse">
    <span class="fa fa-github"></span>
     
    Source code
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Diversity metrics</h1>
<h4 class="date">2026-01-16</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span>
workflowr <span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2026-01-16
</p>
<p>
<strong>Checks:</strong> <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 7
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong> <code>muse/</code> <span
class="glyphicon glyphicon-question-sign" aria-hidden="true"
title="This is the local directory in which the code in this file was executed.">
</span>
</p>
<p>
This reproducible <a href="https://rmarkdown.rstudio.com">R Markdown</a>
analysis was created with <a
  href="https://github.com/workflowr/workflowr">workflowr</a> (version
1.7.1). The <em>Checks</em> tab describes the reproducibility checks
that were applied when the results were created. The <em>Past
versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date
</a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git
repository, you know the exact version of the code that produced these
results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the
global environment can affect the analysis in your R Markdown file in
unknown ways. For reproduciblity it’s best to always run the code in an
empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20200712code">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Seed:</strong>
<code>set.seed(20200712)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20200712code"
class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20200712)</code> was run prior to running
the code in the R Markdown file. Setting a seed ensures that any results
that rely on randomness, e.g. subsampling or permutations, are
reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Session information:</strong>
recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package
versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be
confident that you successfully produced the results during this
run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr
project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomdavetangmusetree898022415254d426af18eb166d69bce00c77c763targetblank8980224a">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Repository version:</strong>
<a href="https://github.com/davetang/muse/tree/898022415254d426af18eb166d69bce00c77c763" target="_blank">8980224</a>
</a>
</p>
</div>
<div
id="strongRepositoryversionstrongahrefhttpsgithubcomdavetangmusetree898022415254d426af18eb166d69bce00c77c763targetblank8980224a"
class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development
and connecting the code version to the results is critical for
reproducibility.
</p>
<p>
The results in this page were generated with repository version
<a href="https://github.com/davetang/muse/tree/898022415254d426af18eb166d69bce00c77c763" target="_blank">8980224</a>.
See the <em>Past versions</em> tab to see a history of the changes made
to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for
the analysis have been committed to Git prior to generating the results
(you can use <code>wflow_publish</code> or
<code>wflow_git_commit</code>). workflowr only checks the R Markdown
file, but you know if there are other scripts or data files that it
depends on. Below is the status of the Git repository when the results
were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .Rproj.user/
    Ignored:    data/1M_neurons_filtered_gene_bc_matrices_h5.h5
    Ignored:    data/293t/
    Ignored:    data/293t_3t3_filtered_gene_bc_matrices.tar.gz
    Ignored:    data/293t_filtered_gene_bc_matrices.tar.gz
    Ignored:    data/5k_Human_Donor1_PBMC_3p_gem-x_5k_Human_Donor1_PBMC_3p_gem-x_count_sample_filtered_feature_bc_matrix.h5
    Ignored:    data/5k_Human_Donor2_PBMC_3p_gem-x_5k_Human_Donor2_PBMC_3p_gem-x_count_sample_filtered_feature_bc_matrix.h5
    Ignored:    data/5k_Human_Donor3_PBMC_3p_gem-x_5k_Human_Donor3_PBMC_3p_gem-x_count_sample_filtered_feature_bc_matrix.h5
    Ignored:    data/5k_Human_Donor4_PBMC_3p_gem-x_5k_Human_Donor4_PBMC_3p_gem-x_count_sample_filtered_feature_bc_matrix.h5
    Ignored:    data/97516b79-8d08-46a6-b329-5d0a25b0be98.h5ad
    Ignored:    data/Parent_SC3v3_Human_Glioblastoma_filtered_feature_bc_matrix.tar.gz
    Ignored:    data/brain_counts/
    Ignored:    data/cl.obo
    Ignored:    data/cl.owl
    Ignored:    data/jurkat/
    Ignored:    data/jurkat:293t_50:50_filtered_gene_bc_matrices.tar.gz
    Ignored:    data/jurkat_293t/
    Ignored:    data/jurkat_filtered_gene_bc_matrices.tar.gz
    Ignored:    data/pbmc20k/
    Ignored:    data/pbmc20k_seurat/
    Ignored:    data/pbmc3k.csv
    Ignored:    data/pbmc3k.csv.gz
    Ignored:    data/pbmc3k.h5ad
    Ignored:    data/pbmc3k/
    Ignored:    data/pbmc3k_bpcells_mat/
    Ignored:    data/pbmc3k_export.mtx
    Ignored:    data/pbmc3k_matrix.mtx
    Ignored:    data/pbmc3k_seurat.rds
    Ignored:    data/pbmc4k_filtered_gene_bc_matrices.tar.gz
    Ignored:    data/pbmc_1k_v3_filtered_feature_bc_matrix.h5
    Ignored:    data/pbmc_1k_v3_raw_feature_bc_matrix.h5
    Ignored:    data/refdata-gex-GRCh38-2020-A.tar.gz
    Ignored:    data/seurat_1m_neuron.rds
    Ignored:    data/t_3k_filtered_gene_bc_matrices.tar.gz
    Ignored:    r_packages_4.4.1/
    Ignored:    r_packages_4.5.0/

Untracked files:
    Untracked:  analysis/bioc.Rmd
    Untracked:  analysis/bioc_scrnaseq.Rmd
    Untracked:  analysis/likelihood.Rmd
    Untracked:  bpcells_matrix/
    Untracked:  data/Caenorhabditis_elegans.WBcel235.113.gtf.gz
    Untracked:  data/GCF_043380555.1-RS_2024_12_gene_ontology.gaf.gz
    Untracked:  data/arab.rds
    Untracked:  data/astronomicalunit.csv
    Untracked:  data/femaleMiceWeights.csv
    Untracked:  data/lung_bcell.rds
    Untracked:  m3/
    Untracked:  women.json

Unstaged changes:
    Modified:   analysis/isoform_switch_analyzer.Rmd

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not
included in this status report because it is ok for generated content to
have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">
<p>
These are the previous versions of the repository in which changes were
made to the R Markdown (<code>analysis/diversity.Rmd</code>) and HTML
(<code>docs/diversity.html</code>) files. If you’ve configured a remote
Git repository (see <code>?wflow_git_remote</code>), click on the
hyperlinks in the table below to view the files as they were in that
past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/davetang/muse/blob/898022415254d426af18eb166d69bce00c77c763/analysis/diversity.Rmd" target="_blank">8980224</a>
</td>
<td>
Dave Tang
</td>
<td>
2026-01-16
</td>
<td>
Diversity metrics
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<div id="introduction" class="section level1">
<h1>Introduction</h1>
<p>This notebook demonstrates various diversity metrics. Install
packages, if necessary.</p>
<pre class="r"><code>install.packages(&quot;vegan&quot;)</code></pre>
</div>
<div id="data" class="section level1">
<h1>Data</h1>
<p>Creating a simulated cell type by organoid count matrix representing
scRNA-seq data from organoids generated using 5 different methods.</p>
<pre class="r"><code>set.seed(1984)

cell_types &lt;- c(
  &quot;Neural Progenitors&quot;,
  &quot;Neurons&quot;,
  &quot;Astrocytes&quot;, 
  &quot;Oligodendrocytes&quot;,
  &quot;Microglia&quot;,
  &quot;Endothelial&quot;,
  &quot;Radial Glia&quot;,
  &quot;Intermediate Progenitors&quot;
)

count_matrix &lt;- base::matrix(c(
  450, 480, 420, 380, 350, 340, 400, 380,
  200, 800, 150, 120, 80, 100, 250, 200,
  100, 1200, 50, 30, 20, 40, 80, 60,
  500, 300, 250, 200, 150, 100, 80, 50,
  300, 500, 200, 180, 150, 120, 200, 150
), nrow = 8, ncol = 5, byrow = FALSE)
count_matrix</code></pre>
<pre><code>     [,1] [,2] [,3] [,4] [,5]
[1,]  450  200  100  500  300
[2,]  480  800 1200  300  500
[3,]  420  150   50  250  200
[4,]  380  120   30  200  180
[5,]  350   80   20  150  150
[6,]  340  100   40  100  120
[7,]  400  250   80   80  200
[8,]  380  200   60   50  150</code></pre>
<pre class="r"><code>base::rownames(count_matrix) &lt;- cell_types
base::colnames(count_matrix) &lt;- letters[1:ncol(count_matrix)]

count_matrix</code></pre>
<pre><code>                           a   b    c   d   e
Neural Progenitors       450 200  100 500 300
Neurons                  480 800 1200 300 500
Astrocytes               420 150   50 250 200
Oligodendrocytes         380 120   30 200 180
Microglia                350  80   20 150 150
Endothelial              340 100   40 100 120
Radial Glia              400 250   80  80 200
Intermediate Progenitors 380 200   60  50 150</code></pre>
</div>
<div id="composition" class="section level1">
<h1>Composition</h1>
<p>Visualise the composition.</p>
<pre class="r"><code>prop_matrix &lt;- apply(count_matrix, 2, \(x) x/sum(x))

prop_df &lt;- base::as.data.frame(prop_matrix) |&gt;
  tibble::rownames_to_column(&quot;cell_type&quot;) |&gt;
  tidyr::pivot_longer(-cell_type, names_to = &quot;method&quot;, values_to = &quot;proportion&quot;)

ggplot(prop_df, aes(x = method, y = proportion, fill = cell_type)) +
  geom_bar(stat = &quot;identity&quot;) +
  theme_bw() +
  theme(axis.text.x = ggplot2::element_text(angle = 45, hjust = 1)) +
  labs(
    title = &quot;Cell type composition by method&quot;,
    y = &quot;Proportion&quot;,
    x = &quot;Method&quot;
  ) +
  scale_fill_brewer(palette = &quot;Set3&quot;)</code></pre>
<p><img src="figure/diversity.Rmd/composition-1.png" width="960" style="display: block; margin: auto;" /></p>
</div>
<div id="diversity-metrics" class="section level1">
<h1>Diversity Metrics</h1>
<div id="species-richness-s" class="section level2">
<h2>Species Richness (S)</h2>
<p><strong>What it measures:</strong> Simply the number of different
cell types present (with &gt;0 cells).</p>
<p><strong>Interpretation:</strong> Higher values indicate more cell
types are present.</p>
<p><strong>Key characteristic:</strong> Does NOT account for abundance -
treats rare and common types equally.</p>
<pre class="r"><code>richness &lt;- vegan::specnumber(base::t(count_matrix))

richness_df &lt;- base::data.frame(
  Method = base::names(richness),
  Richness = richness
)

richness_df</code></pre>
<pre><code>  Method Richness
a      a        8
b      b        8
c      c        8
d      d        8
e      e        8</code></pre>
</div>
<div id="shannon-diversity-index" class="section level2">
<h2>Shannon Diversity Index</h2>
<p><strong>What it measures:</strong> Both richness and evenness
combined. Quantifies uncertainty in predicting the cell type of a
randomly selected cell.</p>
<p><strong>Interpretation:</strong> - Higher values indicate greater
diversity - Ranges from 0 (one cell type only) to ln(S) (all types
equally abundant) - Values typically between 1.5-3.5 for biological
communities</p>
<p><strong>Key characteristic:</strong> More sensitive to rare types
than Simpson’s index.</p>
<pre class="r"><code>shannon &lt;- vegan::diversity(base::t(count_matrix), index = &quot;shannon&quot;)

shannon_df &lt;- base::data.frame(
  Method = base::names(shannon),
  Shannon = base::round(shannon, 3)
)

shannon_df</code></pre>
<pre><code>  Method Shannon
a      a   2.073
b      b   1.768
c      c   0.992
d      d   1.865
e      e   1.968</code></pre>
</div>
<div id="simpson-diversity-index" class="section level2">
<h2>Simpson Diversity Index</h2>
<p><strong>What it measures:</strong> The probability that two randomly
selected cells are of different types.</p>
<p><strong>Interpretation:</strong> - Ranges from 0 to 1 - Higher values
indicate greater diversity - Less sensitive to rare species than
Shannon</p>
<p><strong>Key characteristic:</strong> Emphasizes dominant cell types
more than Shannon.</p>
<pre class="r"><code>simpson &lt;- vegan::diversity(base::t(count_matrix), index = &quot;simpson&quot;)

simpson_df &lt;- base::data.frame(
  Method = base::names(simpson),
  Simpson = base::round(simpson, 3)
)

simpson_df</code></pre>
<pre><code>  Method Simpson
a      a   0.873
b      b   0.768
c      c   0.413
d      d   0.818
e      e   0.842</code></pre>
</div>
<div id="inverse-simpson" class="section level2">
<h2>Inverse Simpson</h2>
<p><strong>What it measures:</strong> The effective number of cell
types. Represents the number of equally abundant types needed to produce
the observed diversity.</p>
<p><strong>Interpretation:</strong> - Minimum value of 1 (only one type)
- Maximum value equals richness (S) when all types equally abundant -
More intuitive than Simpson’s D</p>
<p><strong>Key characteristic:</strong> Can be directly compared to
species richness.</p>
<pre class="r"><code>inv_simpson &lt;- vegan::diversity(base::t(count_matrix), index = &quot;invsimpson&quot;)

inv_simpson_df &lt;- base::data.frame(
  Method = base::names(inv_simpson),
  InvSimpson = base::round(inv_simpson, 2)
)

inv_simpson_df</code></pre>
<pre><code>  Method InvSimpson
a      a       7.90
b      b       4.32
c      c       1.70
d      d       5.49
e      e       6.33</code></pre>
</div>
<div id="pielous-evenness-index" class="section level2">
<h2>Pielou’s Evenness Index</h2>
<p><strong>What it measures:</strong> How evenly cells are distributed
among different types, independent of richness.</p>
<p><strong>Interpretation:</strong> - Ranges from 0 to 1 - 1 indicates
perfect evenness (all types equally abundant) - 0 indicates complete
unevenness (one type dominates)</p>
<p><strong>Key characteristic:</strong> Separates evenness from
richness, useful for comparing samples with different numbers of cell
types.</p>
<pre class="r"><code>pielou &lt;- shannon / base::log(richness)

pielou_df &lt;- base::data.frame(
  Method = base::names(pielou),
  Pielou = base::round(pielou, 3)
)

pielou_df</code></pre>
<pre><code>  Method Pielou
a      a  0.997
b      b  0.850
c      c  0.477
d      d  0.897
e      e  0.946</code></pre>
</div>
<div id="berger-parker-dominance-index" class="section level2">
<h2>Berger-Parker Dominance Index</h2>
<p><strong>What it measures:</strong> The proportion of the most
abundant cell type.</p>
<p><strong>Interpretation:</strong> - Ranges from 0 to 1 - Higher values
indicate one type strongly dominates - Inverse of diversity (high
dominance = low diversity)</p>
<p><strong>Key characteristic:</strong> Simple, intuitive measure
focusing only on the most abundant type.</p>
<pre class="r"><code>berger_parker &lt;- base::apply(count_matrix, 2, \(x) base::max(x) / base::sum(x))

bp_df &lt;- base::data.frame(
  Method = base::names(berger_parker),
  BergerParker = base::round(berger_parker, 3)
)

bp_df</code></pre>
<pre><code>  Method BergerParker
a      a        0.150
b      b        0.421
c      c        0.759
d      d        0.307
e      e        0.278</code></pre>
</div>
</div>
<div id="comparison" class="section level1">
<h1>Comparison</h1>
<p>Summary.</p>
<pre class="r"><code>summary_df &lt;- base::data.frame(
  Richness = richness,
  Shannon = base::round(shannon, 3),
  Simpson = base::round(simpson, 3),
  InvSimpson = base::round(inv_simpson, 2),
  Pielou = base::round(pielou, 3),
  BergerParker = base::round(berger_parker, 3)
)

summary_df</code></pre>
<pre><code>  Richness Shannon Simpson InvSimpson Pielou BergerParker
a        8   2.073   0.873       7.90  0.997        0.150
b        8   1.768   0.768       4.32  0.850        0.421
c        8   0.992   0.413       1.70  0.477        0.759
d        8   1.865   0.818       5.49  0.897        0.307
e        8   1.968   0.842       6.33  0.946        0.278</code></pre>
</div>
<div id="key-differences-between-metrics" class="section level1">
<h1>Key Differences Between Metrics</h1>
<ul>
<li>Richness vs. Diversity Indices
<ul>
<li><strong>Richness</strong> only counts types (presence/absence)</li>
<li><strong>Shannon, Simpson, Inverse Simpson</strong> combine richness
and evenness</li>
</ul></li>
<li>Shannon vs. Simpson
<ul>
<li><strong>Shannon</strong> is more sensitive to rare cell types</li>
<li><strong>Simpson</strong> emphasizes dominant cell types</li>
<li><strong>Shannon</strong> ranges 0 to ln(S); <strong>Simpson</strong>
ranges 0 to 1</li>
</ul></li>
<li>Simpson vs. Inverse Simpson
<ul>
<li>Same information, different scale</li>
<li><strong>Inverse Simpson</strong> is more intuitive (effective number
of types)</li>
</ul></li>
<li>Evenness vs. Diversity
<ul>
<li><strong>Pielou’s J’</strong> isolates evenness from richness</li>
<li>Allows comparing organoids with different numbers of cell types</li>
</ul></li>
<li>Dominance vs. Diversity
<ul>
<li><strong>Berger-Parker</strong> focuses only on the most abundant
type</li>
<li>Opposite of diversity (high dominance = low diversity)</li>
</ul></li>
</ul>
</div>
<div id="conclusion" class="section level1">
<h1>Conclusion</h1>
<ol style="list-style-type: decimal">
<li><strong>For overall diversity:</strong> Use <strong>Shannon</strong>
or <strong>Inverse Simpson</strong></li>
<li><strong>For evenness specifically:</strong> Use <strong>Pielou’s
J’</strong></li>
<li><strong>To identify dominant types:</strong> Use
<strong>Berger-Parker</strong></li>
<li><strong>For simple comparison:</strong> Use
<strong>Richness</strong> and <strong>Inverse Simpson</strong>
together</li>
</ol>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span>
Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 4.5.0 (2025-04-11)
Platform: x86_64-pc-linux-gnu
Running under: Ubuntu 24.04.3 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 
LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.26.so;  LAPACK version 3.12.0

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

time zone: Etc/UTC
tzcode source: system (glibc)

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] vegan_2.7-1     permute_0.9-8   lubridate_1.9.4 forcats_1.0.0  
 [5] stringr_1.5.1   dplyr_1.1.4     purrr_1.0.4     readr_2.1.5    
 [9] tidyr_1.3.1     tibble_3.3.0    ggplot2_3.5.2   tidyverse_2.0.0
[13] workflowr_1.7.1

loaded via a namespace (and not attached):
 [1] sass_0.4.10        generics_0.1.4     lattice_0.22-6     stringi_1.8.7     
 [5] hms_1.1.3          digest_0.6.37      magrittr_2.0.3     timechange_0.3.0  
 [9] evaluate_1.0.3     grid_4.5.0         RColorBrewer_1.1-3 fastmap_1.2.0     
[13] Matrix_1.7-3       rprojroot_2.0.4    jsonlite_2.0.0     processx_3.8.6    
[17] whisker_0.4.1      ps_1.9.1           promises_1.3.3     mgcv_1.9-1        
[21] httr_1.4.7         scales_1.4.0       jquerylib_0.1.4    cli_3.6.5         
[25] rlang_1.1.6        splines_4.5.0      withr_3.0.2        cachem_1.1.0      
[29] yaml_2.3.10        parallel_4.5.0     tools_4.5.0        tzdb_0.5.0        
[33] httpuv_1.6.16      vctrs_0.6.5        R6_2.6.1           lifecycle_1.0.4   
[37] git2r_0.36.2       fs_1.6.6           MASS_7.3-65        cluster_2.1.8.1   
[41] pkgconfig_2.0.3    callr_3.7.6        pillar_1.10.2      bslib_0.9.0       
[45] later_1.4.2        gtable_0.3.6       glue_1.8.0         Rcpp_1.0.14       
[49] xfun_0.52          tidyselect_1.2.1   rstudioapi_0.17.1  knitr_1.50        
[53] farver_2.1.2       nlme_3.1-168       htmltools_0.5.8.1  labeling_0.4.3    
[57] rmarkdown_2.29     compiler_4.5.0     getPass_0.2-4     </code></pre>
</div>
</div>
</div>
</div>
</div>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
https://docs.mathjax.org/en/latest/web/configuration.html. This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>




</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
