<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />



<meta name="date" content="2024-07-31" />

<title>Normalisation methods implemented in edgeR</title>

<script src="site_libs/header-attrs-2.27/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.13.2/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-6.4.2/css/all.min.css" rel="stylesheet" />
<link href="site_libs/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet" />

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/main/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">My blog</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/davetang/muse">
    <span class="fa fa-github"></span>
     
    Source code
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Normalisation methods implemented in
edgeR</h1>
<h4 class="date">2024-07-31</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span>
workflowr <span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2024-07-31
</p>
<p>
<strong>Checks:</strong> <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 7
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong> <code>muse/</code> <span
class="glyphicon glyphicon-question-sign" aria-hidden="true"
title="This is the local directory in which the code in this file was executed.">
</span>
</p>
<p>
This reproducible <a href="https://rmarkdown.rstudio.com">R Markdown</a>
analysis was created with <a
  href="https://github.com/workflowr/workflowr">workflowr</a> (version
1.7.1). The <em>Checks</em> tab describes the reproducibility checks
that were applied when the results were created. The <em>Past
versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date
</a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git
repository, you know the exact version of the code that produced these
results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the
global environment can affect the analysis in your R Markdown file in
unknown ways. For reproduciblity it’s best to always run the code in an
empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20200712code">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Seed:</strong>
<code>set.seed(20200712)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20200712code"
class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20200712)</code> was run prior to running
the code in the R Markdown file. Setting a seed ensures that any results
that rely on randomness, e.g. subsampling or permutations, are
reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Session information:</strong>
recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package
versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be
confident that you successfully produced the results during this
run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr
project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomdavetangmusetree566ad8cd53be8041cecdee7688cd154dca82df69targetblank566ad8ca">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Repository version:</strong>
<a href="https://github.com/davetang/muse/tree/566ad8cd53be8041cecdee7688cd154dca82df69" target="_blank">566ad8c</a>
</a>
</p>
</div>
<div
id="strongRepositoryversionstrongahrefhttpsgithubcomdavetangmusetree566ad8cd53be8041cecdee7688cd154dca82df69targetblank566ad8ca"
class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development
and connecting the code version to the results is critical for
reproducibility.
</p>
<p>
The results in this page were generated with repository version
<a href="https://github.com/davetang/muse/tree/566ad8cd53be8041cecdee7688cd154dca82df69" target="_blank">566ad8c</a>.
See the <em>Past versions</em> tab to see a history of the changes made
to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for
the analysis have been committed to Git prior to generating the results
(you can use <code>wflow_publish</code> or
<code>wflow_git_commit</code>). workflowr only checks the R Markdown
file, but you know if there are other scripts or data files that it
depends on. Below is the status of the Git repository when the results
were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .Rhistory
    Ignored:    .Rproj.user/
    Ignored:    r_packages_4.3.3/
    Ignored:    r_packages_4.4.0/

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not
included in this status report because it is ok for generated content to
have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">
<p>
These are the previous versions of the repository in which changes were
made to the R Markdown (<code>analysis/edger.Rmd</code>) and HTML
(<code>docs/edger.html</code>) files. If you’ve configured a remote Git
repository (see <code>?wflow_git_remote</code>), click on the hyperlinks
in the table below to view the files as they were in that past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/davetang/muse/blob/566ad8cd53be8041cecdee7688cd154dca82df69/analysis/edger.Rmd" target="_blank">566ad8c</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-07-31
</td>
<td>
Include analysis
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/davetang/muse/c34f664aceea989b58fa9937da8b5a07ae4f7673/docs/edger.html" target="_blank">c34f664</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-07-29
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/davetang/muse/blob/0c101211fd757b37b6c3b758beb1a55a98519cf5/analysis/edger.Rmd" target="_blank">0c10121</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-07-29
</td>
<td>
Normalisation methods using edgeR
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/davetang/muse/67bf9ca3cd1817402b882925d69927057915442f/docs/edger.html" target="_blank">67bf9ca</a>
</td>
<td>
Dave Tang
</td>
<td>
2023-10-13
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/davetang/muse/blob/4bc5f6a79b1ce56dc34cd125f914593fb25af065/analysis/edger.Rmd" target="_blank">4bc5f6a</a>
</td>
<td>
Dave Tang
</td>
<td>
2023-10-13
</td>
<td>
edgeR normalisation
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<blockquote>
<p>Differential expression analysis of RNA-seq expression profiles with
biological replication. Implements a range of statistical methodology
based on the negative binomial distributions, including empirical Bayes
estimation, exact tests, generalized linear models and quasi-likelihood
tests. As well as RNA-seq, it be applied to differential signal analysis
of other types of genomic data that produce read counts, including
ChIP-seq, ATAC-seq, Bisulfite-seq, SAGE and CAGE.</p>
</blockquote>
<div id="installation" class="section level2">
<h2>Installation</h2>
<p>Install using <code>BiocManager::install()</code>.</p>
<pre class="r"><code>if (!require(&quot;BiocManager&quot;, quietly = TRUE))
    install.packages(&quot;BiocManager&quot;)

BiocManager::install(&quot;edgeR&quot;)</code></pre>
</div>
<div id="documentation" class="section level2">
<h2>Documentation</h2>
<p>From <code>?edgeR::normLibSizes</code>:</p>
<blockquote>
<p>Calculate scaling factors to convert the raw library sizes for a set
of sequenced samples into normalized effective library sizes.</p>
<p>This function computes scaling factors to convert observed library
sizes into normalized library sizes, also called “effective library
sizes”. The effective library sizes for use in downstream analysis are
lib.size * norm.factors where lib.size contains the original library
sizes and norm.factors is the vector of scaling factors computed by this
function.</p>
<p>The TMM method implements the trimmed mean of M-values method
proposed by Robinson and Oshlack (2010). By default, the M-values are
weighted according to inverse variances, as computed by the delta method
for logarithms of binomial random variables. If refColumn is
unspecified, then the column whose count-per-million upper quartile is
closest to the mean upper quartile is set as the reference library.</p>
<p>The TMMwsp method stands for “TMM with singleton pairing”. This is a
variant of TMM that is intended to have more stable performance when the
counts have a high proportion of zeros. In the TMM method, genes that
have zero count in either library are ignored when comparing pairs of
libraries. In the TMMwsp method, the positive counts from such genes are
reused to increase the number of features by which the libraries are
compared. The singleton positive counts are paired up between the
libraries in decreasing order of size and then a slightly modified TMM
method is applied to the re-ordered libraries. If refColumn is
unspecified, then the column with largest sum of square-root counts is
used as the reference library.</p>
<p>RLE is the scaling factor method proposed by Anders and Huber (2010).
We call it “relative log expression”, as median library is calculated
from the geometric mean of all columns and the median ratio of each
sample to the median library is taken as the scale factor.</p>
<p>The upperquartile method is the upper-quartile normalization method
of Bullard et al (2010), in which the scale factors are calculated from
the 75% quantile of the counts for each library, after removing genes
that are zero in all libraries. The idea is generalized here to allow
normalization by any quantile of the count distributions.</p>
<p>If method=“none”, then the normalization factors are set to 1.</p>
<p>For symmetry, normalization factors are adjusted to multiply to 1.
Rows of object that have zero counts for all columns are removed before
normalization factors are computed. The number of such rows does not
affect the estimated normalization factors.</p>
</blockquote>
</div>
<div id="a-test-dataset" class="section level2">
<h2>A test dataset</h2>
<p>I created a dataset to test the different normalisation methods.
There are four samples; column one and two are the controls and column
three and four are the patients. 25 transcripts are in all four samples
in equal amount. Another 25 transcripts are <strong>only</strong>
present in the controls and their expression is the same as the first 25
transcripts in the controls. The four samples have exactly the same
depth (500 counts). However, since the patient samples have half the
number of transcripts than the controls (25 versus 50), they are
sequenced at twice the depth. This hypothetical situation was described
in <a
href="https://genomebiology.biomedcentral.com/articles/10.1186/gb-2010-11-3-r25">Robinson
and Oshlack Genome Biology 2010</a>.</p>
<pre class="r"><code>control_1 &lt;- rep(10, 50)
control_2 &lt;- rep(10, 50)
patient_1 &lt;- c(rep(20, 25),rep(0,25))
patient_2 &lt;- c(rep(20, 25),rep(0,25))

df &lt;- data.frame(c1=control_1,
                 c2=control_2,
                 p1=patient_1,
                 p2=patient_2)

head(df)</code></pre>
<pre><code>  c1 c2 p1 p2
1 10 10 20 20
2 10 10 20 20
3 10 10 20 20
4 10 10 20 20
5 10 10 20 20
6 10 10 20 20</code></pre>
<p>Tail of the dataset.</p>
<pre class="r"><code>tail(df)</code></pre>
<pre><code>   c1 c2 p1 p2
45 10 10  0  0
46 10 10  0  0
47 10 10  0  0
48 10 10  0  0
49 10 10  0  0
50 10 10  0  0</code></pre>
<p>Equal depth.</p>
<pre class="r"><code>colSums(df)</code></pre>
<pre><code> c1  c2  p1  p2 
500 500 500 500 </code></pre>
</div>
<div id="no-normalisation" class="section level2">
<h2>No normalisation</h2>
<p>Let’s run the differential expression analysis without any
normalisation step:</p>
<pre class="r"><code>library(edgeR)</code></pre>
<pre><code>Loading required package: limma</code></pre>
<pre class="r"><code># create group vector
group &lt;- c(&#39;control&#39;,&#39;control&#39;,&#39;patient&#39;,&#39;patient&#39;)

# create DGEList object
d &lt;- DGEList(counts=df, group=group)

# check out the DGEList object
d</code></pre>
<pre><code>An object of class &quot;DGEList&quot;
$counts
  c1 c2 p1 p2
1 10 10 20 20
2 10 10 20 20
3 10 10 20 20
4 10 10 20 20
5 10 10 20 20
45 more rows ...

$samples
     group lib.size norm.factors
c1 control      500            1
c2 control      500            1
p1 patient      500            1
p2 patient      500            1</code></pre>
<p>Perform a differential gene expression analysis.</p>
<pre class="r"><code>d &lt;- DGEList(counts=df, group=group)
d &lt;- estimateCommonDisp(d)

# perform the DE test
de &lt;- exactTest(d)

# how many differentially expressed transcripts?
table(p.adjust(de$table$PValue, method=&quot;BH&quot;)&lt;0.05)</code></pre>
<pre><code>
TRUE 
  50 </code></pre>
<p>Without normalisation, every transcript is differentially
expressed.</p>
</div>
<div id="tmm-normalisation" class="section level2">
<h2>TMM normalisation</h2>
<p>From the <a
href="https://bioconductor.org/packages/release/bioc/vignettes/edgeR/inst/doc/edgeRUsersGuide.pdf">edgeR
manual</a>:</p>
<p>{edgeR} is concerned with differential expression analysis rather
than with the quantification of expression levels. It is concerned with
relative changes in expression levels between conditions, but not
directly with estimating absolute expression levels.</p>
<p>The <code>normLibSizes()</code> function normalises the library sizes
in such a way to minimise the log-fold changes between the samples for
most genes. The default method for computing these scale factors uses a
trimmed mean of M-values (TMM) between each pair of samples. We call the
product of the original library size and the scaling factor the
effective library size, i.e., the normalised library size. The effective
library size replaces the original library size in all downstream
analyses</p>
<p>Let’s test the weighted trimmed mean of M-values method:</p>
<pre class="r"><code>d_tmm &lt;- normLibSizes(d, method=&quot;TMM&quot;)
d_tmm</code></pre>
<pre><code>An object of class &quot;DGEList&quot;
$counts
  c1 c2 p1 p2
1 10 10 20 20
2 10 10 20 20
3 10 10 20 20
4 10 10 20 20
5 10 10 20 20
45 more rows ...

$samples
     group lib.size norm.factors
c1 control      500    0.7071068
c2 control      500    0.7071068
p1 patient      500    1.4142136
p2 patient      500    1.4142136

$common.dispersion
[1] 0.0001005378

$pseudo.counts
  c1 c2 p1 p2
1 10 10 20 20
2 10 10 20 20
3 10 10 20 20
4 10 10 20 20
5 10 10 20 20
45 more rows ...

$pseudo.lib.size
[1] 500

$AveLogCPM
[1] 15.04175 15.04175 15.04175 15.04175 15.04175
45 more elements ...</code></pre>
<p>If we use the scaling, we will observe that the differences are now
less pronounced.</p>
<pre class="r"><code>head(d$counts)</code></pre>
<pre><code>  c1 c2 p1 p2
1 10 10 20 20
2 10 10 20 20
3 10 10 20 20
4 10 10 20 20
5 10 10 20 20
6 10 10 20 20</code></pre>
<pre class="r"><code>head(d_tmm$counts * d_tmm$samples$norm.factors)</code></pre>
<pre><code>         c1        c2       p1       p2
1  7.071068 14.142136 14.14214 28.28427
2  7.071068 14.142136 14.14214 28.28427
3 14.142136  7.071068 28.28427 14.14214
4 14.142136  7.071068 28.28427 14.14214
5  7.071068 14.142136 14.14214 28.28427
6  7.071068 14.142136 14.14214 28.28427</code></pre>
<p>Perform the differential expression test.</p>
<pre class="r"><code>d_tmm &lt;- estimateCommonDisp(d_tmm)
d_tmm &lt;- exactTest(d_tmm)
table(p.adjust(d_tmm$table$PValue, method=&quot;BH&quot;)&lt;0.05)</code></pre>
<pre><code>
FALSE  TRUE 
   25    25 </code></pre>
<p>Only half of the transcripts are differentially expressed (the last
25 transcripts).</p>
</div>
<div id="rle-normalisation" class="section level2">
<h2>RLE normalisation</h2>
<p>Let’s test the relative log expression normalisation method:</p>
<pre class="r"><code>d_rle &lt;- normLibSizes(d, method=&quot;RLE&quot;)
d_rle</code></pre>
<pre><code>An object of class &quot;DGEList&quot;
$counts
  c1 c2 p1 p2
1 10 10 20 20
2 10 10 20 20
3 10 10 20 20
4 10 10 20 20
5 10 10 20 20
45 more rows ...

$samples
     group lib.size norm.factors
c1 control      500    0.7071068
c2 control      500    0.7071068
p1 patient      500    1.4142136
p2 patient      500    1.4142136

$common.dispersion
[1] 0.0001005378

$pseudo.counts
  c1 c2 p1 p2
1 10 10 20 20
2 10 10 20 20
3 10 10 20 20
4 10 10 20 20
5 10 10 20 20
45 more rows ...

$pseudo.lib.size
[1] 500

$AveLogCPM
[1] 15.04175 15.04175 15.04175 15.04175 15.04175
45 more elements ...</code></pre>
<p>Perform the differential gene expression analysis.</p>
<pre class="r"><code>d_rle &lt;- estimateCommonDisp(d_rle)
d_rle &lt;- exactTest(d_rle)
table(p.adjust(d_rle$table$PValue, method=&quot;BH&quot;)&lt;0.05)</code></pre>
<pre><code>
FALSE  TRUE 
   25    25 </code></pre>
</div>
<div id="upper-quartile-normalisation" class="section level2">
<h2>Upper-quartile normalisation</h2>
<p>Lastly let’s try the upper-quartile normalisation method:</p>
<pre class="r"><code>d_uq &lt;- normLibSizes(d, method=&quot;upperquartile&quot;)
d_uq</code></pre>
<pre><code>An object of class &quot;DGEList&quot;
$counts
  c1 c2 p1 p2
1 10 10 20 20
2 10 10 20 20
3 10 10 20 20
4 10 10 20 20
5 10 10 20 20
45 more rows ...

$samples
     group lib.size norm.factors
c1 control      500    0.7071068
c2 control      500    0.7071068
p1 patient      500    1.4142136
p2 patient      500    1.4142136

$common.dispersion
[1] 0.0001005378

$pseudo.counts
  c1 c2 p1 p2
1 10 10 20 20
2 10 10 20 20
3 10 10 20 20
4 10 10 20 20
5 10 10 20 20
45 more rows ...

$pseudo.lib.size
[1] 500

$AveLogCPM
[1] 15.04175 15.04175 15.04175 15.04175 15.04175
45 more elements ...</code></pre>
<p>DE test.</p>
<pre class="r"><code>d_uq &lt;- estimateCommonDisp(d_uq)
d_uq &lt;- exactTest(d_uq)
table(p.adjust(d_uq$table$PValue, method=&quot;BH&quot;)&lt;0.05)</code></pre>
<pre><code>
FALSE  TRUE 
   25    25 </code></pre>
</div>
<div id="testing-a-real-dataset" class="section level2">
<h2>Testing a real dataset</h2>
<p>Perform differential gene expression analysis using various
normalisation methods on the <a
href="https://pubmed.ncbi.nlm.nih.gov/19088194/">pnas_expression.txt</a>
dataset.</p>
<pre class="r"><code>my_url &lt;- &quot;https://davetang.org/file/pnas_expression.txt&quot;
data &lt;- read.table(my_url, header=TRUE, sep=&quot;\t&quot;)

dim(data)</code></pre>
<pre><code>[1] 37435     9</code></pre>
<p>Prepare a DGEList object.</p>
<pre class="r"><code>d &lt;- data[,2:8]
rownames(d) &lt;- data[,1]
group &lt;- c(rep(&quot;Control&quot;,4),rep(&quot;DHT&quot;,3))
d &lt;- DGEList(counts = d, group=group)</code></pre>
<p>Filter out lowly expressed genes.</p>
<pre class="r"><code>keep &lt;- rowSums(cpm(d) &gt; 0.5) &gt;= 2
d &lt;- d[keep, , keep.lib.sizes=FALSE]</code></pre>
<p>Carry out differential gene expression analysis with no
normalisation.</p>
<pre class="r"><code>no_norm &lt;- estimateCommonDisp(d)
no_norm &lt;- exactTest(no_norm)
table(p.adjust(no_norm$table$PValue, method=&quot;BH&quot;)&lt;0.05)</code></pre>
<pre><code>
FALSE  TRUE 
14683  4782 </code></pre>
<p>With TMM normalisation.</p>
<pre class="r"><code>TMM &lt;- normLibSizes(d, method=&quot;TMM&quot;)
TMM</code></pre>
<pre><code>An object of class &quot;DGEList&quot;
$counts
                lane1 lane2 lane3 lane4 lane5 lane6 lane8
ENSG00000124208   478   619   628   744   483   716   240
ENSG00000182463    27    20    27    26    48    55    24
ENSG00000124201   180   218   293   275   373   301    88
ENSG00000124205     0     0     5     5     0     0     0
ENSG00000124207    76    80    85    97    80    81    37
19460 more rows ...

$samples
        group lib.size norm.factors
lane1 Control   978233    1.0327868
lane2 Control  1156431    1.0371519
lane3 Control  1441598    1.0278995
lane4 Control  1484958    1.0294240
lane5     DHT  1822854    0.9366119
lane6     DHT  1833704    0.9340634
lane8     DHT   681509    1.0084769</code></pre>
<pre class="r"><code>TMM &lt;- estimateCommonDisp(TMM)
TMM &lt;- exactTest(TMM)
table(p.adjust(TMM$table$PValue, method=&quot;BH&quot;)&lt;0.05)</code></pre>
<pre><code>
FALSE  TRUE 
14913  4552 </code></pre>
<p>With RLE.</p>
<pre class="r"><code>RLE &lt;- normLibSizes(d, method=&quot;RLE&quot;)
RLE</code></pre>
<pre><code>An object of class &quot;DGEList&quot;
$counts
                lane1 lane2 lane3 lane4 lane5 lane6 lane8
ENSG00000124208   478   619   628   744   483   716   240
ENSG00000182463    27    20    27    26    48    55    24
ENSG00000124201   180   218   293   275   373   301    88
ENSG00000124205     0     0     5     5     0     0     0
ENSG00000124207    76    80    85    97    80    81    37
19460 more rows ...

$samples
        group lib.size norm.factors
lane1 Control   978233    1.0149859
lane2 Control  1156431    1.0236589
lane3 Control  1441598    1.0345742
lane4 Control  1484958    1.0400447
lane5     DHT  1822854    0.9706371
lane6     DHT  1833704    0.9734747
lane8     DHT   681509    0.9466503</code></pre>
<pre class="r"><code>RLE &lt;- estimateCommonDisp(RLE)
RLE &lt;- exactTest(RLE)
table(p.adjust(RLE$table$PValue, method=&quot;BH&quot;)&lt;0.05)</code></pre>
<pre><code>
FALSE  TRUE 
14801  4664 </code></pre>
<p>With the upper quartile method.</p>
<pre class="r"><code>uq &lt;- normLibSizes(d, method=&quot;upperquartile&quot;)
uq</code></pre>
<pre><code>An object of class &quot;DGEList&quot;
$counts
                lane1 lane2 lane3 lane4 lane5 lane6 lane8
ENSG00000124208   478   619   628   744   483   716   240
ENSG00000182463    27    20    27    26    48    55    24
ENSG00000124201   180   218   293   275   373   301    88
ENSG00000124205     0     0     5     5     0     0     0
ENSG00000124207    76    80    85    97    80    81    37
19460 more rows ...

$samples
        group lib.size norm.factors
lane1 Control   978233    1.0110884
lane2 Control  1156431    1.0263441
lane3 Control  1441598    1.0291497
lane4 Control  1484958    1.0324023
lane5     DHT  1822854    0.9766795
lane6     DHT  1833704    0.9843852
lane8     DHT   681509    0.9433508</code></pre>
<pre class="r"><code>uq &lt;- estimateCommonDisp(uq)
uq &lt;- exactTest(uq)
table(p.adjust(uq$table$PValue, method=&quot;BH&quot;)&lt;0.05)</code></pre>
<pre><code>
FALSE  TRUE 
14796  4669 </code></pre>
<p>Finding the overlaps between the differential gene expression
analyses.</p>
<pre class="r"><code>library(gplots)</code></pre>
<pre><code>
Attaching package: &#39;gplots&#39;</code></pre>
<pre><code>The following object is masked from &#39;package:stats&#39;:

    lowess</code></pre>
<pre class="r"><code>get_de &lt;- function(x, pvalue){
  my_i &lt;- p.adjust(x$PValue, method=&quot;BH&quot;) &lt; pvalue
  row.names(x)[my_i]
}

my_de_no_norm &lt;- get_de(no_norm$table, 0.05)
my_de_tmm &lt;- get_de(TMM$table, 0.05)
my_de_rle &lt;- get_de(RLE$table, 0.05)
my_de_uq &lt;- get_de(uq$table, 0.05)

gplots::venn(list(no_norm = my_de_no_norm, TMM = my_de_tmm, RLE = my_de_rle, UQ = my_de_uq))</code></pre>
<p><img src="figure/edger.Rmd/unnamed-chunk-8-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>There is a large overlap of differentially expressed genes in the
different normalisation approaches.</p>
<pre class="r"><code>gplots::venn(list(TMM = my_de_tmm, RLE = my_de_rle, UQ = my_de_uq))</code></pre>
<p><img src="figure/edger.Rmd/unnamed-chunk-9-1.png" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="summary" class="section level2">
<h2>Summary</h2>
<p>The normalisation factors were quite similar between all
normalisation methods, which is why the results of the differential
expression were quite concordant. Most methods down sized the DHT
samples with a normalisation factor of less than one to account for the
larger library sizes of these samples.</p>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span>
Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 4.4.0 (2024-04-24)
Platform: x86_64-pc-linux-gnu
Running under: Ubuntu 22.04.4 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 
LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

time zone: Etc/UTC
tzcode source: system (glibc)

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] gplots_3.1.3.1  edgeR_4.2.1     limma_3.60.4    lubridate_1.9.3
 [5] forcats_1.0.0   stringr_1.5.1   dplyr_1.1.4     purrr_1.0.2    
 [9] readr_2.1.5     tidyr_1.3.1     tibble_3.2.1    ggplot2_3.5.1  
[13] tidyverse_2.0.0 workflowr_1.7.1

loaded via a namespace (and not attached):
 [1] gtable_0.3.5       xfun_0.44          bslib_0.7.0        caTools_1.18.2    
 [5] processx_3.8.4     lattice_0.22-6     callr_3.7.6        tzdb_0.4.0        
 [9] vctrs_0.6.5        tools_4.4.0        ps_1.7.6           bitops_1.0-7      
[13] generics_0.1.3     fansi_1.0.6        highr_0.11         pkgconfig_2.0.3   
[17] KernSmooth_2.23-22 lifecycle_1.0.4    compiler_4.4.0     git2r_0.33.0      
[21] statmod_1.5.0      munsell_0.5.1      getPass_0.2-4      httpuv_1.6.15     
[25] htmltools_0.5.8.1  sass_0.4.9         yaml_2.3.8         later_1.3.2       
[29] pillar_1.9.0       jquerylib_0.1.4    whisker_0.4.1      cachem_1.1.0      
[33] gtools_3.9.5       tidyselect_1.2.1   locfit_1.5-9.9     digest_0.6.35     
[37] stringi_1.8.4      rprojroot_2.0.4    fastmap_1.2.0      grid_4.4.0        
[41] colorspace_2.1-0   cli_3.6.2          magrittr_2.0.3     utf8_1.2.4        
[45] withr_3.0.0        scales_1.3.0       promises_1.3.0     timechange_0.3.0  
[49] rmarkdown_2.27     httr_1.4.7         hms_1.1.3          evaluate_0.24.0   
[53] knitr_1.47         rlang_1.1.4        Rcpp_1.0.12        glue_1.7.0        
[57] rstudioapi_0.16.0  jsonlite_1.8.8     R6_2.5.1           fs_1.6.4          </code></pre>
</div>
</div>
</div>
</div>
</div>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
https://docs.mathjax.org/en/latest/web/configuration.html. This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>




</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
