<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />



<meta name="date" content="2024-10-31" />

<title>Differential gene expression analysis using edgeR</title>

<script src="site_libs/header-attrs-2.27/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.13.2/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-6.4.2/css/all.min.css" rel="stylesheet" />
<link href="site_libs/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet" />

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/main/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">My blog</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/davetang/muse">
    <span class="fa fa-github"></span>
     
    Source code
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Differential gene expression analysis using
edgeR</h1>
<h4 class="date">2024-10-31</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span>
workflowr <span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2024-10-31
</p>
<p>
<strong>Checks:</strong> <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 7
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong> <code>muse/</code> <span
class="glyphicon glyphicon-question-sign" aria-hidden="true"
title="This is the local directory in which the code in this file was executed.">
</span>
</p>
<p>
This reproducible <a href="https://rmarkdown.rstudio.com">R Markdown</a>
analysis was created with <a
  href="https://github.com/workflowr/workflowr">workflowr</a> (version
1.7.1). The <em>Checks</em> tab describes the reproducibility checks
that were applied when the results were created. The <em>Past
versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date
</a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git
repository, you know the exact version of the code that produced these
results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the
global environment can affect the analysis in your R Markdown file in
unknown ways. For reproduciblity it’s best to always run the code in an
empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20200712code">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Seed:</strong>
<code>set.seed(20200712)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20200712code"
class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20200712)</code> was run prior to running
the code in the R Markdown file. Setting a seed ensures that any results
that rely on randomness, e.g. subsampling or permutations, are
reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Session information:</strong>
recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package
versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be
confident that you successfully produced the results during this
run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr
project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomdavetangmusetree04948c54e3d2024b33286d089cac99d0cfbb1173targetblank04948c5a">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Repository version:</strong>
<a href="https://github.com/davetang/muse/tree/04948c54e3d2024b33286d089cac99d0cfbb1173" target="_blank">04948c5</a>
</a>
</p>
</div>
<div
id="strongRepositoryversionstrongahrefhttpsgithubcomdavetangmusetree04948c54e3d2024b33286d089cac99d0cfbb1173targetblank04948c5a"
class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development
and connecting the code version to the results is critical for
reproducibility.
</p>
<p>
The results in this page were generated with repository version
<a href="https://github.com/davetang/muse/tree/04948c54e3d2024b33286d089cac99d0cfbb1173" target="_blank">04948c5</a>.
See the <em>Past versions</em> tab to see a history of the changes made
to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for
the analysis have been committed to Git prior to generating the results
(you can use <code>wflow_publish</code> or
<code>wflow_git_commit</code>). workflowr only checks the R Markdown
file, but you know if there are other scripts or data files that it
depends on. Below is the status of the Git repository when the results
were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .Rhistory
    Ignored:    .Rproj.user/
    Ignored:    data/pbmc3k.csv
    Ignored:    data/pbmc3k.csv.gz
    Ignored:    data/pbmc3k/
    Ignored:    r_packages_4.4.0/

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not
included in this status report because it is ok for generated content to
have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">
<p>
These are the previous versions of the repository in which changes were
made to the R Markdown (<code>analysis/edger_de.Rmd</code>) and HTML
(<code>docs/edger_de.html</code>) files. If you’ve configured a remote
Git repository (see <code>?wflow_git_remote</code>), click on the
hyperlinks in the table below to view the files as they were in that
past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/davetang/muse/blob/04948c54e3d2024b33286d089cac99d0cfbb1173/analysis/edger_de.Rmd" target="_blank">04948c5</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-10-31
</td>
<td>
Export edgeR results
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/davetang/muse/cfbb929582e7c0386338bce4e96900f406e7a447/docs/edger_de.html" target="_blank">cfbb929</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-10-31
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/davetang/muse/blob/92b951ca79310340c50fbcf9d3d8a59c7278a96b/analysis/edger_de.Rmd" target="_blank">92b951c</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-10-31
</td>
<td>
Export edgeR results
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/davetang/muse/6aa4e2d485d9f744f1e0119c2d3a7aac934a1eee/docs/edger_de.html" target="_blank">6aa4e2d</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-10-22
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/davetang/muse/blob/dda9d85bb9e9c9117d4f65bbdae02328ae957c9b/analysis/edger_de.Rmd" target="_blank">dda9d85</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-10-22
</td>
<td>
Visualise DEGs
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/davetang/muse/6625ffc621574b3db7638f082be40db4d548d0c2/docs/edger_de.html" target="_blank">6625ffc</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-10-22
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/davetang/muse/blob/3661a2e11475b2b827b70a86800c1e939966510d/analysis/edger_de.Rmd" target="_blank">3661a2e</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-10-22
</td>
<td>
Additional sample
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/davetang/muse/c9037d1312e24959a4f379125e2f1732175787f0/docs/edger_de.html" target="_blank">c9037d1</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-10-22
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/davetang/muse/blob/f7369ced3bf60f895b1ed68e76165c2eb43a5c7e/analysis/edger_de.Rmd" target="_blank">f7369ce</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-10-22
</td>
<td>
Differential gene expression analysis using edgeR
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<p><a
href="https://bioconductor.org/packages/release/bioc/html/edgeR.html">edgeR</a>
carries out:</p>
<blockquote>
<p>Differential expression analysis of RNA-seq expression profiles with
biological replication. Implements a range of statistical methodology
based on the negative binomial distributions, including empirical Bayes
estimation, exact tests, generalized linear models and quasi-likelihood
tests. As well as RNA-seq, it be applied to differential signal analysis
of other types of genomic data that produce read counts, including
ChIP-seq, ATAC-seq, Bisulfite-seq, SAGE and CAGE.</p>
</blockquote>
<p><a
href="https://www.bioconductor.org/packages/release/bioc/vignettes/edgeR/inst/doc/edgeRUsersGuide.pdf">edgeR
Manual</a>.</p>
<div id="installation" class="section level2">
<h2>Installation</h2>
<p>Install using <code>BiocManager::install()</code>.</p>
<pre class="r"><code>if (!require(&quot;BiocManager&quot;, quietly = TRUE))
    install.packages(&quot;BiocManager&quot;)

BiocManager::install(&quot;edgeR&quot;)</code></pre>
</div>
<div id="data" class="section level2">
<h2>Data</h2>
<p>Use data from <a href="https://zenodo.org/records/13970886">An
example RNA-seq count table using nf-core/rnaseq</a>.</p>
<pre class="r"><code>my_url &lt;- &#39;https://zenodo.org/records/13970886/files/rsem.merged.gene_counts.tsv?download=1&#39;
my_file &lt;- &#39;rsem.merged.gene_counts.tsv&#39;

if(file.exists(my_file) == FALSE){
  download.file(url = my_url, destfile = my_file)
}

gene_counts &lt;- read_tsv(&quot;rsem.merged.gene_counts.tsv&quot;, show_col_types = FALSE)
head(gene_counts)</code></pre>
<pre><code># A tibble: 6 × 10
  gene_id   `transcript_id(s)` ERR160122 ERR160123 ERR160124 ERR164473 ERR164550
  &lt;chr&gt;     &lt;chr&gt;                  &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;
1 ENSG0000… ENST00000373020,E…        2         6         5      374       1637 
2 ENSG0000… ENST00000373031,E…       19        40        28        0          1 
3 ENSG0000… ENST00000371582,E…      268.      274.      429.     489        637 
4 ENSG0000… ENST00000367770,E…      360.      449.      566.     363.       606.
5 ENSG0000… ENST00000286031,E…      156.      185.      265.      85.4      312.
6 ENSG0000… ENST00000374003,E…       24        23        40     1181        423 
# ℹ 3 more variables: ERR164551 &lt;dbl&gt;, ERR164552 &lt;dbl&gt;, ERR164554 &lt;dbl&gt;</code></pre>
<p>Metadata.</p>
<pre class="r"><code>tibble::tribble(
  ~sample, ~run_id, ~group,
  &quot;C2_norm&quot;, &quot;ERR160122&quot;, &quot;normal&quot;,
  &quot;C3_norm&quot;, &quot;ERR160123&quot;, &quot;normal&quot;,
  &quot;C5_norm&quot;, &quot;ERR160124&quot;, &quot;normal&quot;,
  &quot;C1_norm&quot;, &quot;ERR164473&quot;, &quot;normal&quot;,
  &quot;C1_cancer&quot;, &quot;ERR164550&quot;, &quot;cancer&quot;,
  &quot;C2_cancer&quot;, &quot;ERR164551&quot;, &quot;cancer&quot;,
  &quot;C3_cancer&quot;, &quot;ERR164552&quot;, &quot;cancer&quot;,
  &quot;C5_cancer&quot;, &quot;ERR164554&quot;, &quot;cancer&quot;
) -&gt; my_metadata

my_metadata$group &lt;- factor(my_metadata$group, levels = c(&#39;normal&#39;, &#39;cancer&#39;))</code></pre>
</div>
<div id="dgelist" class="section level2">
<h2>DGEList</h2>
<p>The input to <code>edgeR</code> is the <code>DGEList</code> object.
The required inputs for creating a <code>DGEList</code> object is the
count table and a grouping factor.</p>
<pre class="r"><code>gene_counts |&gt;
  dplyr::select(starts_with(&quot;ERR&quot;)) |&gt;
  mutate(across(everything(), as.integer)) |&gt;
  as.matrix() -&gt; gene_counts_mat

row.names(gene_counts_mat) &lt;- gene_counts$gene_id

idx &lt;- match(colnames(gene_counts_mat), my_metadata$run_id)
colnames(gene_counts_mat) &lt;- my_metadata$sample[idx]

y &lt;- DGEList(
  counts = gene_counts_mat,
  group = my_metadata$group[idx]
)

y</code></pre>
<pre><code>An object of class &quot;DGEList&quot;
$counts
                C2_norm C3_norm C5_norm C1_norm C1_cancer C2_cancer C3_cancer
ENSG00000000003       2       6       5     374      1637       650      1015
ENSG00000000005      19      40      28       0         1         0         0
ENSG00000000419     268     273     428     489       637       879      1157
ENSG00000000457     360     449     566     362       605       708       632
ENSG00000000460     155     184     264      85       312       239       147
                C5_cancer
ENSG00000000003       562
ENSG00000000005         0
ENSG00000000419       729
ENSG00000000457       478
ENSG00000000460       156
63135 more rows ...

$samples
           group lib.size norm.factors
C2_norm   normal  4441588            1
C3_norm   normal  5349291            1
C5_norm   normal  7613422            1
C1_norm   normal 15972904            1
C1_cancer cancer 22329493            1
C2_cancer cancer 29928921            1
C3_cancer cancer 24891206            1
C5_cancer cancer 23703606            1</code></pre>
</div>
<div id="filtering-to-remove-low-counts" class="section level2">
<h2>Filtering to remove low counts</h2>
<p>Remove genes that are lowly expressed.</p>
<pre class="r"><code>keep &lt;- rowSums(cpm(y) &gt; 0.5) &gt;= 2
y &lt;- y[keep, , keep.lib.sizes=FALSE]
y</code></pre>
<pre><code>An object of class &quot;DGEList&quot;
$counts
                C2_norm C3_norm C5_norm C1_norm C1_cancer C2_cancer C3_cancer
ENSG00000000003       2       6       5     374      1637       650      1015
ENSG00000000005      19      40      28       0         1         0         0
ENSG00000000419     268     273     428     489       637       879      1157
ENSG00000000457     360     449     566     362       605       708       632
ENSG00000000460     155     184     264      85       312       239       147
                C5_cancer
ENSG00000000003       562
ENSG00000000005         0
ENSG00000000419       729
ENSG00000000457       478
ENSG00000000460       156
37419 more rows ...

$samples
           group lib.size norm.factors
C2_norm   normal  4431282            1
C3_norm   normal  5337400            1
C5_norm   normal  7594512            1
C1_norm   normal 15964680            1
C1_cancer cancer 22317658            1
C2_cancer cancer 29912740            1
C3_cancer cancer 24876336            1
C5_cancer cancer 23693355            1</code></pre>
</div>
<div id="normalisation-for-composition-bias" class="section level2">
<h2>Normalisation for composition bias</h2>
<blockquote>
<p>The <code>normLibSizes()</code> function normalizes the library sizes
in such a way to minimize the log-fold changes between the samples for
most genes. The default method for computing these scale factors uses a
trimmed mean of M-values (TMM) between each pair of samples. We call the
product of the original library size and the scaling factor the
effective library size, i.e., the normalized library size. The effective
library size replaces the original library size in all downstream
analyses</p>
</blockquote>
<pre class="r"><code>y &lt;- normLibSizes(y)
y</code></pre>
<pre><code>An object of class &quot;DGEList&quot;
$counts
                C2_norm C3_norm C5_norm C1_norm C1_cancer C2_cancer C3_cancer
ENSG00000000003       2       6       5     374      1637       650      1015
ENSG00000000005      19      40      28       0         1         0         0
ENSG00000000419     268     273     428     489       637       879      1157
ENSG00000000457     360     449     566     362       605       708       632
ENSG00000000460     155     184     264      85       312       239       147
                C5_cancer
ENSG00000000003       562
ENSG00000000005         0
ENSG00000000419       729
ENSG00000000457       478
ENSG00000000460       156
37419 more rows ...

$samples
           group lib.size norm.factors
C2_norm   normal  4431282    1.4420477
C3_norm   normal  5337400    1.4377942
C5_norm   normal  7594512    1.3836402
C1_norm   normal 15964680    0.8389637
C1_cancer cancer 22317658    0.8199077
C2_cancer cancer 29912740    0.7972139
C3_cancer cancer 24876336    0.8151023
C5_cancer cancer 23693355    0.7798397</code></pre>
</div>
<div id="mds" class="section level2">
<h2>MDS</h2>
<pre class="r"><code>plotMDS(y, plot = FALSE)$eigen.vectors[, 1:2] |&gt;
  as.data.frame() |&gt;
  cbind(my_metadata) |&gt;
  dplyr::rename(`Eigenvector 1` = V1, `Eigenvector 2` = V2) |&gt;
  ggplot(aes(`Eigenvector 1`, `Eigenvector 2`, colour = group, label = sample)) +
  geom_point(size = 2) +
  geom_text_repel(show.legend = FALSE) +
  theme_minimal() +
  ggtitle(&quot;MDS plot&quot;)</code></pre>
<p><img src="figure/edger_de.Rmd/mds-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-mds-1">
Past versions of mds-1.png
</button>
</p>
<div id="fig-mds-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/davetang/muse/blob/6625ffc621574b3db7638f082be40db4d548d0c2/docs/figure/edger_de.Rmd/mds-1.png" target="_blank">6625ffc</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-10-22
</td>
</tr>
<tr>
<td>
<a href="https://github.com/davetang/muse/blob/c9037d1312e24959a4f379125e2f1732175787f0/docs/figure/edger_de.Rmd/mds-1.png" target="_blank">c9037d1</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-10-22
</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<div id="differential-expression" class="section level2">
<h2>Differential expression</h2>
<pre class="r"><code>design &lt;- model.matrix(~y$samples$group)
y &lt;- estimateDisp(y, design, robust=TRUE)
fit &lt;- glmQLFit(y, design, robust=TRUE)
res &lt;- glmQLFTest(fit)
topTags(res)</code></pre>
<pre><code>Coefficient:  y$samples$groupcancer 
                    logFC    logCPM         F       PValue          FDR
ENSG00000289381 -7.411609 2.3413727 156.14405 3.485881e-09 0.0001304556
ENSG00000100985  5.744629 5.7690594 110.89086 1.418804e-06 0.0058527061
ENSG00000275332 -7.648126 2.5721238  59.27705 1.721074e-06 0.0058527061
ENSG00000220695 -9.132437 4.0374574  58.73122 1.982088e-06 0.0058527061
ENSG00000256422 -5.529837 0.5541825  54.99786 2.395813e-06 0.0058527061
ENSG00000151834 -8.035376 3.6441016  61.94292 2.733010e-06 0.0058527061
ENSG00000167910 -8.018047 3.4530411  55.86149 3.928929e-06 0.0058527061
ENSG00000226443 -5.742734 0.9948963  51.37638 4.209874e-06 0.0058527061
ENSG00000250696 -8.580216 3.8092969  52.88815 4.376155e-06 0.0058527061
ENSG00000196778 -8.809149 4.2377739  54.66541 4.439027e-06 0.0058527061</code></pre>
<p>CPM.</p>
<pre class="r"><code>my_cpms &lt;- cpm(y, prior.count=2, log=TRUE)
head(my_cpms)</code></pre>
<pre><code>                  C2_norm    C3_norm    C5_norm   C1_norm C1_cancer C2_cancer
ENSG00000000003 -1.159752 -0.1258585 -0.7121002  4.810340  6.485360  4.775676
ENSG00000000005  1.635961  2.4187173  1.4850268 -2.893182 -2.401590 -2.893182
ENSG00000000419  5.394869  5.1582110  5.3528004  5.195507  5.127070  5.209251
ENSG00000000457  5.819453  5.8738922  5.7548326  4.763521  5.053005  4.898404
ENSG00000000460  4.608265  4.5916439  4.6586766  2.696179  4.103099  3.344387
ENSG00000000938  1.959920  1.6469490  1.9786359  6.464501  4.539243  7.133878
                C3_cancer C5_cancer
ENSG00000000003  5.649383  4.933136
ENSG00000000005 -2.893182 -2.893182
ENSG00000000419  5.837817  5.307028
ENSG00000000457  4.968242  4.700693
ENSG00000000460  2.884458  3.100561
ENSG00000000938  5.947951  5.961622</code></pre>
<p>Down-regulated genes with respect to the cancer group.</p>
<pre class="r"><code>topTags(res, n = Inf) |&gt;
  as.data.frame() |&gt;
  dplyr::filter(FDR &lt; 0.01) |&gt;
  dplyr::filter(logFC &lt; 0) |&gt;
  dplyr::slice_min(order_by = FDR, n = 20, with_ties = FALSE) |&gt;
  row.names() -&gt; down_genes

my_cpms[down_genes, ] |&gt;
  pheatmap()</code></pre>
<p><img src="figure/edger_de.Rmd/down_regulated_genes-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-down_regulated_genes-1">
Past versions of down_regulated_genes-1.png
</button>
</p>
<div id="fig-down_regulated_genes-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/davetang/muse/blob/cfbb929582e7c0386338bce4e96900f406e7a447/docs/figure/edger_de.Rmd/down_regulated_genes-1.png" target="_blank">cfbb929</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-10-31
</td>
</tr>
<tr>
<td>
<a href="https://github.com/davetang/muse/blob/6aa4e2d485d9f744f1e0119c2d3a7aac934a1eee/docs/figure/edger_de.Rmd/down_regulated_genes-1.png" target="_blank">6aa4e2d</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-10-22
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>Up-regulated genes with respect to the cancer group.</p>
<pre class="r"><code>topTags(res, n = Inf) |&gt;
  as.data.frame() |&gt;
  dplyr::filter(FDR &lt; 0.01) |&gt;
  dplyr::filter(logFC &gt; 0) |&gt;
  dplyr::slice_min(order_by = FDR, n = 20, with_ties = FALSE) |&gt;
  row.names() -&gt; up_genes

my_cpms[up_genes, ] |&gt;
  pheatmap()</code></pre>
<p><img src="figure/edger_de.Rmd/up_regulated_genes-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-up_regulated_genes-1">
Past versions of up_regulated_genes-1.png
</button>
</p>
<div id="fig-up_regulated_genes-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/davetang/muse/blob/cfbb929582e7c0386338bce4e96900f406e7a447/docs/figure/edger_de.Rmd/up_regulated_genes-1.png" target="_blank">cfbb929</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-10-31
</td>
</tr>
<tr>
<td>
<a href="https://github.com/davetang/muse/blob/6aa4e2d485d9f744f1e0119c2d3a7aac934a1eee/docs/figure/edger_de.Rmd/up_regulated_genes-1.png" target="_blank">6aa4e2d</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-10-22
</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<div id="export-results" class="section level2">
<h2>Export results</h2>
<p>Results are stored in <code>res</code>.</p>
<pre class="r"><code>class(res)</code></pre>
<pre><code>[1] &quot;DGELRT&quot;
attr(,&quot;package&quot;)
[1] &quot;edgeR&quot;</code></pre>
<p>Specifically, in <code>res$table</code>.</p>
<pre class="r"><code>head(res$table)</code></pre>
<pre><code>                     logFC   logCPM          F      PValue
ENSG00000000003  2.7275199 4.832838  4.2763383 0.068387013
ENSG00000000005 -7.0023947 0.540980 17.6315738 0.002159975
ENSG00000000419  0.1201226 5.339964  0.1140066 0.743147316
ENSG00000000457 -0.7079087 5.306090  3.3541835 0.099292833
ENSG00000000460 -0.8967285 3.947381  2.6625206 0.136187708
ENSG00000000938  1.5368400 5.597107  1.8649388 0.205160993</code></pre>
<p>Add adjusted p-values and export.</p>
<pre class="r"><code>res$table |&gt;
  dplyr::mutate(adjusted_pvalue = p.adjust(PValue, method = &quot;BH&quot;)) |&gt;
  tibble::rownames_to_column(&quot;ensembl_gene_id&quot;) |&gt;
  readr::write_csv(file = &quot;data/13970886_edger_res.csv&quot;)</code></pre>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span>
Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 4.4.0 (2024-04-24)
Platform: x86_64-pc-linux-gnu
Running under: Ubuntu 22.04.4 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 
LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

time zone: Etc/UTC
tzcode source: system (glibc)

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] pheatmap_1.0.12 ggrepel_0.9.5   edgeR_4.2.2     limma_3.60.6   
 [5] lubridate_1.9.3 forcats_1.0.0   stringr_1.5.1   dplyr_1.1.4    
 [9] purrr_1.0.2     readr_2.1.5     tidyr_1.3.1     tibble_3.2.1   
[13] ggplot2_3.5.1   tidyverse_2.0.0 workflowr_1.7.1

loaded via a namespace (and not attached):
 [1] gtable_0.3.5       xfun_0.44          bslib_0.7.0        processx_3.8.4    
 [5] lattice_0.22-6     callr_3.7.6        tzdb_0.4.0         vctrs_0.6.5       
 [9] tools_4.4.0        ps_1.7.6           generics_0.1.3     parallel_4.4.0    
[13] fansi_1.0.6        highr_0.11         pkgconfig_2.0.3    RColorBrewer_1.1-3
[17] lifecycle_1.0.4    farver_2.1.2       compiler_4.4.0     git2r_0.33.0      
[21] statmod_1.5.0      munsell_0.5.1      getPass_0.2-4      httpuv_1.6.15     
[25] htmltools_0.5.8.1  sass_0.4.9         yaml_2.3.8         crayon_1.5.2      
[29] later_1.3.2        pillar_1.9.0       jquerylib_0.1.4    whisker_0.4.1     
[33] cachem_1.1.0       tidyselect_1.2.1   locfit_1.5-9.9     digest_0.6.37     
[37] stringi_1.8.4      splines_4.4.0      labeling_0.4.3     rprojroot_2.0.4   
[41] fastmap_1.2.0      grid_4.4.0         colorspace_2.1-0   cli_3.6.3         
[45] magrittr_2.0.3     utf8_1.2.4         withr_3.0.1        scales_1.3.0      
[49] promises_1.3.0     bit64_4.0.5        timechange_0.3.0   rmarkdown_2.27    
[53] httr_1.4.7         bit_4.0.5          hms_1.1.3          evaluate_0.24.0   
[57] knitr_1.47         rlang_1.1.4        Rcpp_1.0.12        glue_1.7.0        
[61] rstudioapi_0.16.0  vroom_1.6.5        jsonlite_1.8.8     R6_2.5.1          
[65] fs_1.6.4          </code></pre>
</div>
</div>
</div>
</div>
</div>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
https://docs.mathjax.org/en/latest/web/configuration.html. This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>




</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
