<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />



<meta name="date" content="2025-05-26" />

<title>Using the Fast preranked Gene Set Enrichment Analysis (fgsea) package</title>

<script src="site_libs/header-attrs-2.29/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.13.2/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-6.5.2/css/all.min.css" rel="stylesheet" />
<link href="site_libs/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet" />

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/main/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">My blog</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/davetang/muse">
    <span class="fa fa-github"></span>
     
    Source code
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Using the Fast preranked Gene Set
Enrichment Analysis (fgsea) package</h1>
<h4 class="date">2025-05-26</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span>
workflowr <span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2025-05-26
</p>
<p>
<strong>Checks:</strong> <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 7
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong> <code>muse/</code> <span
class="glyphicon glyphicon-question-sign" aria-hidden="true"
title="This is the local directory in which the code in this file was executed.">
</span>
</p>
<p>
This reproducible <a href="https://rmarkdown.rstudio.com">R Markdown</a>
analysis was created with <a
  href="https://github.com/workflowr/workflowr">workflowr</a> (version
1.7.1). The <em>Checks</em> tab describes the reproducibility checks
that were applied when the results were created. The <em>Past
versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date
</a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git
repository, you know the exact version of the code that produced these
results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the
global environment can affect the analysis in your R Markdown file in
unknown ways. For reproduciblity it’s best to always run the code in an
empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20200712code">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Seed:</strong>
<code>set.seed(20200712)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20200712code"
class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20200712)</code> was run prior to running
the code in the R Markdown file. Setting a seed ensures that any results
that rely on randomness, e.g. subsampling or permutations, are
reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Session information:</strong>
recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package
versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be
confident that you successfully produced the results during this
run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr
project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomdavetangmusetree1d87e1e248cb4d1908f7d244cbe19d2b93295574targetblank1d87e1ea">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Repository version:</strong>
<a href="https://github.com/davetang/muse/tree/1d87e1e248cb4d1908f7d244cbe19d2b93295574" target="_blank">1d87e1e</a>
</a>
</p>
</div>
<div
id="strongRepositoryversionstrongahrefhttpsgithubcomdavetangmusetree1d87e1e248cb4d1908f7d244cbe19d2b93295574targetblank1d87e1ea"
class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development
and connecting the code version to the results is critical for
reproducibility.
</p>
<p>
The results in this page were generated with repository version
<a href="https://github.com/davetang/muse/tree/1d87e1e248cb4d1908f7d244cbe19d2b93295574" target="_blank">1d87e1e</a>.
See the <em>Past versions</em> tab to see a history of the changes made
to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for
the analysis have been committed to Git prior to generating the results
(you can use <code>wflow_publish</code> or
<code>wflow_git_commit</code>). workflowr only checks the R Markdown
file, but you know if there are other scripts or data files that it
depends on. Below is the status of the Git repository when the results
were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .Rproj.user/
    Ignored:    data/1M_neurons_filtered_gene_bc_matrices_h5.h5
    Ignored:    data/293t/
    Ignored:    data/293t_3t3_filtered_gene_bc_matrices.tar.gz
    Ignored:    data/293t_filtered_gene_bc_matrices.tar.gz
    Ignored:    data/5k_Human_Donor1_PBMC_3p_gem-x_5k_Human_Donor1_PBMC_3p_gem-x_count_sample_filtered_feature_bc_matrix.h5
    Ignored:    data/5k_Human_Donor2_PBMC_3p_gem-x_5k_Human_Donor2_PBMC_3p_gem-x_count_sample_filtered_feature_bc_matrix.h5
    Ignored:    data/5k_Human_Donor3_PBMC_3p_gem-x_5k_Human_Donor3_PBMC_3p_gem-x_count_sample_filtered_feature_bc_matrix.h5
    Ignored:    data/5k_Human_Donor4_PBMC_3p_gem-x_5k_Human_Donor4_PBMC_3p_gem-x_count_sample_filtered_feature_bc_matrix.h5
    Ignored:    data/97516b79-8d08-46a6-b329-5d0a25b0be98.h5ad
    Ignored:    data/Parent_SC3v3_Human_Glioblastoma_filtered_feature_bc_matrix.tar.gz
    Ignored:    data/brain_counts/
    Ignored:    data/cl.obo
    Ignored:    data/cl.owl
    Ignored:    data/jurkat/
    Ignored:    data/jurkat:293t_50:50_filtered_gene_bc_matrices.tar.gz
    Ignored:    data/jurkat_293t/
    Ignored:    data/jurkat_filtered_gene_bc_matrices.tar.gz
    Ignored:    data/pbmc20k/
    Ignored:    data/pbmc20k_seurat/
    Ignored:    data/pbmc3k/
    Ignored:    data/pbmc3k_bpcells_mat/
    Ignored:    data/pbmc3k_seurat.rds
    Ignored:    data/pbmc4k_filtered_gene_bc_matrices.tar.gz
    Ignored:    data/pbmc_1k_v3_filtered_feature_bc_matrix.h5
    Ignored:    data/pbmc_1k_v3_raw_feature_bc_matrix.h5
    Ignored:    data/refdata-gex-GRCh38-2020-A.tar.gz
    Ignored:    data/seurat_1m_neuron.rds
    Ignored:    data/t_3k_filtered_gene_bc_matrices.tar.gz
    Ignored:    r_packages_4.4.1/
    Ignored:    r_packages_4.5.0/

Untracked files:
    Untracked:  Nothobranchius_furzeri.Nfu_20140520.113.gtf.gz
    Untracked:  analysis/bioc_scrnaseq.Rmd
    Untracked:  bpcells_matrix/
    Untracked:  data/GCF_043380555.1-RS_2024_12_gene_ontology.gaf.gz
    Untracked:  m3/
    Untracked:  pbmc3k_before_filtering.rds
    Untracked:  pbmc3k_save_rds.rds
    Untracked:  rsem.merged.gene_counts.tsv

Unstaged changes:
    Modified:   analysis/cluster_profiler.Rmd

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not
included in this status report because it is ok for generated content to
have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">
<p>
These are the previous versions of the repository in which changes were
made to the R Markdown (<code>analysis/fgsea.Rmd</code>) and HTML
(<code>docs/fgsea.html</code>) files. If you’ve configured a remote Git
repository (see <code>?wflow_git_remote</code>), click on the hyperlinks
in the table below to view the files as they were in that past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/davetang/muse/blob/1d87e1e248cb4d1908f7d244cbe19d2b93295574/analysis/fgsea.Rmd" target="_blank">1d87e1e</a>
</td>
<td>
Dave Tang
</td>
<td>
2025-05-26
</td>
<td>
Modify plotEnrichment plot
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/davetang/muse/6f2ce9bade8cedffe20e0ef397ea6e46efdc9583/docs/fgsea.html" target="_blank">6f2ce9b</a>
</td>
<td>
Dave Tang
</td>
<td>
2025-01-24
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/davetang/muse/a24419ba7e6d4b09e1345f9a62f696dc138b7b29/docs/fgsea.html" target="_blank">a24419b</a>
</td>
<td>
Dave Tang
</td>
<td>
2023-02-13
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/davetang/muse/blob/380c988691bee78a82f4c4a98a08b027c56d3fe0/analysis/fgsea.Rmd" target="_blank">380c988</a>
</td>
<td>
Dave Tang
</td>
<td>
2023-02-13
</td>
<td>
Fix leading edge gene count
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/davetang/muse/f48870229f7197cbea7eca54661b1a1d6a26167f/docs/fgsea.html" target="_blank">f488702</a>
</td>
<td>
Dave Tang
</td>
<td>
2023-02-13
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/davetang/muse/blob/f90ae592bd3971c4a362ab00880294996f019c8f/analysis/fgsea.Rmd" target="_blank">f90ae59</a>
</td>
<td>
Dave Tang
</td>
<td>
2023-02-13
</td>
<td>
fgsea analysis and summary
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/davetang/muse/b82b4fccdedc76ac273db0a488b52fbe64437164/docs/fgsea.html" target="_blank">b82b4fc</a>
</td>
<td>
Dave Tang
</td>
<td>
2023-02-01
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/davetang/muse/blob/fb915380459c4cba93cbbaa75828886048e1935a/analysis/fgsea.Rmd" target="_blank">fb91538</a>
</td>
<td>
Dave Tang
</td>
<td>
2023-02-01
</td>
<td>
fgsea example data
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<p>From the <a href="https://pubmed.ncbi.nlm.nih.gov/16199517/">original
paper</a> describing the Gene Set Enrichment Analysis (GSEA):</p>
<blockquote>
<p>The goal of GSEA is to determine whether members of a gene set
<strong>S</strong> tend to occur toward the top (or bottom) of the list
<strong>L</strong>, in which case the gene set is correlated with the
phenotypic class distinction.</p>
</blockquote>
<p>To provide an example, consider carrying out RNA-seq experiments with
two conditions, where one set of experiments are the control and another
set are the treatment. A differential gene expression analysis is
carried out and each gene is <strong>ranked</strong> by the fold-change
between the control and treatment; genes that are up-regulated
(fold-change &gt; 1) are ranked at the top and genes that are
down-regulated (fold-change &lt; 1) are ranked at the bottom.</p>
<p>Besides inspecting the individual genes that are up- or
down-regulated, you wish to find out whether a pathway of genes is up-
or down-regulated. Therefore you carry out GSEA using a set of gene
pathways (<strong>S</strong>) and check whether genes belonging to the
same pathway tend to occur at the start or the end of the list of ranked
genes (<strong>L</strong>).</p>
<div id="installation" class="section level2">
<h2>Installation</h2>
<p>First install <a
href="https://bioconductor.org/packages/release/bioc/html/fgsea.html">fgsea</a>.</p>
<pre class="r"><code>if (!require(&quot;BiocManager&quot;, quietly = TRUE))
  install.packages(&quot;BiocManager&quot;)

if (!require(&quot;fgsea&quot;, quietly = TRUE))
  BiocManager::install(&quot;fgsea&quot;)

library(fgsea)</code></pre>
</div>
<div id="example-data" class="section level2">
<h2>Example data</h2>
<p>The example pathways are packaged with <code>fgsea</code> and can be
loaded with <code>data()</code>. The example pathways are stored in a
list.</p>
<pre class="r"><code>data(&quot;examplePathways&quot;, package = &quot;fgsea&quot;)
class(examplePathways)</code></pre>
<pre><code>[1] &quot;list&quot;</code></pre>
<p>There are 1457 example pathways.</p>
<pre class="r"><code>length(examplePathways)</code></pre>
<pre><code>[1] 1457</code></pre>
<p>The first pathway 1221633_Meiotic_Synapsis contains Entrez Gene IDs
that belong to this gene set.</p>
<pre class="r"><code>examplePathways[1]</code></pre>
<pre><code>$`1221633_Meiotic_Synapsis`
 [1] &quot;12189&quot;     &quot;13006&quot;     &quot;15077&quot;     &quot;15078&quot;     &quot;15270&quot;     &quot;15512&quot;    
 [7] &quot;16905&quot;     &quot;16906&quot;     &quot;19357&quot;     &quot;20842&quot;     &quot;20843&quot;     &quot;20957&quot;    
[13] &quot;20962&quot;     &quot;21749&quot;     &quot;21750&quot;     &quot;22196&quot;     &quot;23856&quot;     &quot;24061&quot;    
[19] &quot;28113&quot;     &quot;50878&quot;     &quot;56739&quot;     &quot;57321&quot;     &quot;64009&quot;     &quot;66654&quot;    
[25] &quot;69386&quot;     &quot;71846&quot;     &quot;74075&quot;     &quot;77053&quot;     &quot;94244&quot;     &quot;97114&quot;    
[31] &quot;97122&quot;     &quot;97908&quot;     &quot;101185&quot;    &quot;140557&quot;    &quot;223697&quot;    &quot;260423&quot;   
[37] &quot;319148&quot;    &quot;319149&quot;    &quot;319150&quot;    &quot;319151&quot;    &quot;319152&quot;    &quot;319153&quot;   
[43] &quot;319154&quot;    &quot;319155&quot;    &quot;319156&quot;    &quot;319157&quot;    &quot;319158&quot;    &quot;319159&quot;   
[49] &quot;319160&quot;    &quot;319161&quot;    &quot;319565&quot;    &quot;320332&quot;    &quot;320558&quot;    &quot;326619&quot;   
[55] &quot;326620&quot;    &quot;360198&quot;    &quot;497652&quot;    &quot;544973&quot;    &quot;625328&quot;    &quot;667250&quot;   
[61] &quot;100041230&quot; &quot;102641229&quot; &quot;102641751&quot; &quot;102642045&quot;</code></pre>
<p>The gene ranks are also packaged with <code>fgsea</code> but we will
re-generate the ranks based on the <a
href="https://github.com/ctlab/fgsea/blob/master/inst/gen_gene_ranks.R">author’s
code</a> to see how the ranks were created. Several other Bioconductor
packages are required to generate the ranks. The Reactome database is
also installed here for later use.</p>
<pre class="r"><code>bioc_pac &lt;- c(&quot;GEOquery&quot;, &quot;limma&quot;, &quot;org.Mm.eg.db&quot;, &quot;reactome.db&quot;)
cran_pac &lt;- c(&quot;data.table&quot;, &quot;pheatmap&quot;)

install_pac &lt;- function(x, repo){
  if (!require(x, quietly = TRUE, character.only = TRUE)){
    if(repo == &quot;bioc&quot;){
      BiocManager::install(x, character.only = TRUE)
    } else if (repo == &quot;cran&quot;){
      install.packages(x, character.only = TRUE)
    } else {
      stop(&quot;Unknown repo&quot;)
    }
  }
}

sapply(bioc_pac, install_pac, repo = &quot;bioc&quot;)</code></pre>
<pre><code>
Attaching package: &#39;generics&#39;</code></pre>
<pre><code>The following objects are masked from &#39;package:base&#39;:

    as.difftime, as.factor, as.ordered, intersect, is.element, setdiff,
    setequal, union</code></pre>
<pre><code>
Attaching package: &#39;BiocGenerics&#39;</code></pre>
<pre><code>The following objects are masked from &#39;package:stats&#39;:

    IQR, mad, sd, var, xtabs</code></pre>
<pre><code>The following objects are masked from &#39;package:base&#39;:

    anyDuplicated, aperm, append, as.data.frame, basename, cbind,
    colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,
    get, grep, grepl, is.unsorted, lapply, Map, mapply, match, mget,
    order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,
    rbind, Reduce, rownames, sapply, saveRDS, table, tapply, unique,
    unsplit, which.max, which.min</code></pre>
<pre><code>Welcome to Bioconductor

    Vignettes contain introductory material; view with
    &#39;browseVignettes()&#39;. To cite Bioconductor, see
    &#39;citation(&quot;Biobase&quot;)&#39;, and for packages &#39;citation(&quot;pkgname&quot;)&#39;.</code></pre>
<pre><code>Setting options(&#39;download.file.method.GEOquery&#39;=&#39;auto&#39;)</code></pre>
<pre><code>Setting options(&#39;GEOquery.inmemory.gpl&#39;=FALSE)</code></pre>
<pre><code>
Attaching package: &#39;limma&#39;</code></pre>
<pre><code>The following object is masked from &#39;package:BiocGenerics&#39;:

    plotMA</code></pre>
<pre><code>
Attaching package: &#39;S4Vectors&#39;</code></pre>
<pre><code>The following object is masked from &#39;package:utils&#39;:

    findMatches</code></pre>
<pre><code>The following objects are masked from &#39;package:base&#39;:

    expand.grid, I, unname</code></pre>
<pre><code></code></pre>
<pre><code>$GEOquery
NULL

$limma
NULL

$org.Mm.eg.db
NULL

$reactome.db
NULL</code></pre>
<pre class="r"><code>sapply(cran_pac, install_pac, repo = &quot;cran&quot;)</code></pre>
<pre><code>
Attaching package: &#39;data.table&#39;</code></pre>
<pre><code>The following object is masked from &#39;package:IRanges&#39;:

    shift</code></pre>
<pre><code>The following objects are masked from &#39;package:S4Vectors&#39;:

    first, second</code></pre>
<pre><code>$data.table
NULL

$pheatmap
NULL</code></pre>
<p>Load packages.</p>
<pre class="r"><code>library(GEOquery)
library(limma)
library(org.Mm.eg.db)
library(data.table)
# for collapseBy
source(&quot;https://raw.githubusercontent.com/assaron/r-utils/master/R/exprs.R&quot;)</code></pre>
<p>The example gene ranks were created using <a
href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE14308">mouse
microarray data</a>, which we can download using <code>getGEO</code>
from the <code>GEOquery</code> package.</p>
<pre class="r"><code>gse14308 &lt;- getGEO(&quot;GSE14308&quot;)[[1]]</code></pre>
<pre><code>Found 1 file(s)</code></pre>
<pre><code>GSE14308_series_matrix.txt.gz</code></pre>
<p>Conditions (are extracted from the title and) are added to the
phenotypic data.</p>
<pre class="r"><code>pData(gse14308)$condition &lt;- sub(&quot;-.*$&quot;, &quot;&quot;, gse14308$title)
pData(gse14308)[, c(&#39;platform_id&#39;, &#39;type&#39;, &#39;condition&#39;)]</code></pre>
<pre><code>          platform_id type condition
GSM357839     GPL1261  RNA       Th2
GSM357841     GPL1261  RNA       Th2
GSM357842     GPL1261  RNA       Th1
GSM357843     GPL1261  RNA      Th17
GSM357844     GPL1261  RNA       Th1
GSM357845     GPL1261  RNA      Th17
GSM357847     GPL1261  RNA     Naive
GSM357848     GPL1261  RNA     Naive
GSM357849     GPL1261  RNA     iTreg
GSM357850     GPL1261  RNA     iTreg
GSM357852     GPL1261  RNA     nTreg
GSM357853     GPL1261  RNA     nTreg</code></pre>
<p><code>fData</code> retrieves information on features, which are
microarray probe sets.</p>
<pre class="r"><code>feature_dat &lt;- fData(gse14308)
colnames(feature_dat)</code></pre>
<pre><code> [1] &quot;ID&quot;                               &quot;GB_ACC&quot;                          
 [3] &quot;SPOT_ID&quot;                          &quot;Species Scientific Name&quot;         
 [5] &quot;Annotation Date&quot;                  &quot;Sequence Type&quot;                   
 [7] &quot;Sequence Source&quot;                  &quot;Target Description&quot;              
 [9] &quot;Representative Public ID&quot;         &quot;Gene Title&quot;                      
[11] &quot;Gene Symbol&quot;                      &quot;ENTREZ_GENE_ID&quot;                  
[13] &quot;RefSeq Transcript ID&quot;             &quot;Gene Ontology Biological Process&quot;
[15] &quot;Gene Ontology Cellular Component&quot; &quot;Gene Ontology Molecular Function&quot;</code></pre>
<p>The <code>collapseBy</code> function is sourced from <a
href="https://raw.githubusercontent.com/assaron/r-utils/master/R/exprs.R">exprs.R</a>
and the source is shown below but named as <code>collapseBy_</code> so
we do not overwrite the sourced function. We will go through each line
of code to try to understand what the function is doing.</p>
<pre class="r"><code>collapseBy_ &lt;- function(es, factor, FUN=median) {
  ranks &lt;- apply(exprs(es), 1, FUN)
  t &lt;- data.frame(f=factor, i=seq_along(ranks), r=ranks)
  t &lt;- t[order(t$r, decreasing=T), ]
  keep &lt;- t[!duplicated(t$f) &amp; !is.na(t$f),]$i
  res &lt;- es[keep, ]
  fData(res)$origin &lt;- rownames(res)
  rownames(res) &lt;- factor[keep]
  res
}</code></pre>
<p><code>ranks</code> contains the median probe intensity across all
samples.</p>
<pre class="r"><code>ranks &lt;- apply(exprs(gse14308), 1, median)
head(ranks)</code></pre>
<pre><code>  1415670_at   1415671_at   1415672_at   1415673_at 1415674_a_at   1415675_at 
    3324.740     4933.035    11202.750     3239.865     4007.050     1830.655 </code></pre>
<p>The <code>ENTREZ_GENE_ID</code>, index, and median get saved into a
data frame. (I’ve used <code>my_df</code> here because <code>t</code> is
the name of the transpose function.) Next, the data frame is ordered
from highest intensity to lowest.</p>
<pre class="r"><code>my_df &lt;- data.frame(
  f=fData(gse14308)$ENTREZ_GENE_ID,
  i=seq_along(ranks),
  r=ranks
)
my_df &lt;- my_df[order(my_df$r, decreasing=TRUE), ]

head(my_df)</code></pre>
<pre><code>                                             f     i        r
1438859_x_at                             20090 23165 99737.85
1454859_a_at 65019 /// 100044627 /// 100862455 39154 99078.75
1416404_s_at 20055 /// 100039355 /// 100862433   735 97947.30
1455485_x_at                             22121 39780 97510.75
1422475_a_at                             20091  6781 97177.05
1435873_a_at               22121 /// 100504632 20179 96558.85</code></pre>
<p>A vector called <code>keep</code> is created to keep only rows with
non-duplicated and non-missing Entrez Gene IDs.</p>
<pre class="r"><code>keep &lt;- my_df[!duplicated(my_df$f) &amp; !is.na(my_df$f),]$i
length(keep)</code></pre>
<pre><code>[1] 21603</code></pre>
<p>Finally, <code>keep</code> is used to subset the data;
<code>origin</code> is created to store the original probe IDs before
replacing the row names with Entrez Gene IDs.</p>
<pre class="r"><code>res &lt;- gse14308[keep, ]
fData(res)$origin &lt;- rownames(res)
rownames(res) &lt;- fData(gse14308)$ENTREZ_GENE_ID[keep]
res</code></pre>
<pre><code>ExpressionSet (storageMode: lockedEnvironment)
assayData: 21603 features, 12 samples 
  element names: exprs 
protocolData: none
phenoData
  sampleNames: GSM357839 GSM357841 ... GSM357853 (12 total)
  varLabels: title geo_accession ... condition (34 total)
  varMetadata: labelDescription
featureData
  featureNames: 20090 65019 /// 100044627 /// 100862455 ... 194227
    (21603 total)
  fvarLabels: ID GB_ACC ... origin (17 total)
  fvarMetadata: Column Description labelDescription
experimentData: use &#39;experimentData(object)&#39;
  pubMedIds: 19144320 
Annotation: GPL1261 </code></pre>
<p>Therefore, the <code>collapseBy</code> function is used to calculate
the median intensity across samples/experiments, which is used to rank
features, and to remove features that are duplicated or have no Entrez
Gene ID.</p>
<p>Now that we know what <code>collapseBy</code> does, we can use
it.</p>
<pre class="r"><code>es &lt;- collapseBy(gse14308, fData(gse14308)$ENTREZ_GENE_ID, FUN=median)
es</code></pre>
<pre><code>ExpressionSet (storageMode: lockedEnvironment)
assayData: 21603 features, 12 samples 
  element names: exprs 
protocolData: none
phenoData
  sampleNames: GSM357839 GSM357841 ... GSM357853 (12 total)
  varLabels: title geo_accession ... condition (34 total)
  varMetadata: labelDescription
featureData
  featureNames: 20090 65019 /// 100044627 /// 100862455 ... 194227
    (21603 total)
  fvarLabels: ID GB_ACC ... origin (17 total)
  fvarMetadata: Column Description labelDescription
experimentData: use &#39;experimentData(object)&#39;
  pubMedIds: 19144320 
Annotation: GPL1261 </code></pre>
<p>Probe IDs that mapped to several Entrez Gene IDs and empty entries
are also removed.</p>
<pre class="r"><code>es &lt;- es[!grepl(&quot;///&quot;, rownames(es)), ]
es &lt;- es[rownames(es) != &quot;&quot;, ]
dim(exprs(es))</code></pre>
<pre><code>[1] 20770    12</code></pre>
<p>Quantile normalisation is carried out.</p>
<pre class="r"><code>exprs(es) &lt;- normalizeBetweenArrays(log2(exprs(es)+1), method=&quot;quantile&quot;)</code></pre>
<p>A <a href="https://f1000research.com/articles/9-1444">design
matrix</a> is defined.</p>
<pre class="r"><code>es.design &lt;- model.matrix(~0+condition, data=pData(es))
es.design</code></pre>
<pre><code>          conditioniTreg conditionNaive conditionnTreg conditionTh1
GSM357839              0              0              0            0
GSM357841              0              0              0            0
GSM357842              0              0              0            1
GSM357843              0              0              0            0
GSM357844              0              0              0            1
GSM357845              0              0              0            0
GSM357847              0              1              0            0
GSM357848              0              1              0            0
GSM357849              1              0              0            0
GSM357850              1              0              0            0
GSM357852              0              0              1            0
GSM357853              0              0              1            0
          conditionTh17 conditionTh2
GSM357839             0            1
GSM357841             0            1
GSM357842             0            0
GSM357843             1            0
GSM357844             0            0
GSM357845             1            0
GSM357847             0            0
GSM357848             0            0
GSM357849             0            0
GSM357850             0            0
GSM357852             0            0
GSM357853             0            0
attr(,&quot;assign&quot;)
[1] 1 1 1 1 1 1
attr(,&quot;contrasts&quot;)
attr(,&quot;contrasts&quot;)$condition
[1] &quot;contr.treatment&quot;</code></pre>
<p>A linear model is fit given the design.</p>
<pre class="r"><code>fit &lt;- lmFit(es, es.design)</code></pre>
<p>A contrasts matrix is used to compute contrasts using our fitted
linear model. Here we’re contrasting naive T-cells to T-helper 1
cells.</p>
<pre class="r"><code>makeContrasts(
  conditionTh1-conditionNaive,
  levels=es.design
)</code></pre>
<pre><code>                Contrasts
Levels           conditionTh1 - conditionNaive
  conditioniTreg                             0
  conditionNaive                            -1
  conditionnTreg                             0
  conditionTh1                               1
  conditionTh17                              0
  conditionTh2                               0</code></pre>
<pre class="r"><code>fit2 &lt;- contrasts.fit(
  fit,
  makeContrasts(
    conditionTh1-conditionNaive,
    levels=es.design
  )
)

fit2</code></pre>
<pre><code>An object of class &quot;MArrayLM&quot;
$coefficients
       Contrasts
        conditionTh1 - conditionNaive
  20090                   0.011485524
  22121                  -0.003602942
  20091                   0.035573263
  67671                   0.092303053
  19241                   0.013452656
20765 more rows ...

$rank
[1] 6

$assign
[1] 1 1 1 1 1 1

$qr
$qr
          conditioniTreg conditionNaive conditionnTreg conditionTh1
GSM357839      -1.414214       0.000000       0.000000    0.0000000
GSM357841       0.000000      -1.414214       0.000000    0.0000000
GSM357842       0.000000       0.000000      -1.414214    0.0000000
GSM357843       0.000000       0.000000       0.000000   -1.4142136
GSM357844       0.000000       0.000000       0.000000    0.7071068
          conditionTh17 conditionTh2
GSM357839      0.000000            0
GSM357841      0.000000            0
GSM357842      0.000000            0
GSM357843      0.000000            0
GSM357844      1.414214            0
7 more rows ...

$qraux
[1] 1.0 1.0 1.0 1.0 1.5 1.0

$pivot
[1] 1 2 3 4 5 6

$tol
[1] 1e-07

$rank
[1] 6


$df.residual
[1] 6 6 6 6 6
20765 more elements ...

$sigma
     20090      22121      20091      67671      19241 
0.03820456 0.03984248 0.05377004 0.03935491 0.03240994 
20765 more elements ...

$cov.coefficients
                               Contrasts
Contrasts                       conditionTh1 - conditionNaive
  conditionTh1 - conditionNaive                             1

$stdev.unscaled
       Contrasts
        conditionTh1 - conditionNaive
  20090                             1
  22121                             1
  20091                             1
  67671                             1
  19241                             1
20765 more rows ...

$pivot
[1] 1 2 3 4 5 6

$genes
                ID    GB_ACC SPOT_ID Species Scientific Name Annotation Date
20090 1438859_x_at  AV037157                    Mus musculus     Oct 6, 2014
22121 1455485_x_at  AI324936                    Mus musculus     Oct 6, 2014
20091 1422475_a_at NM_016959                    Mus musculus     Oct 6, 2014
67671 1433472_x_at  AA050777                    Mus musculus     Oct 6, 2014
19241   1415906_at NM_021278                    Mus musculus     Oct 6, 2014
           Sequence Type Sequence Source
20090 Consensus sequence         GenBank
22121 Consensus sequence         GenBank
20091 Consensus sequence         GenBank
67671 Consensus sequence         GenBank
19241 Consensus sequence         GenBank
                                                                                                                                                                                                                                                                Target Description
20090                                                                              gb:AV037157 /DB_XREF=gi:4856822 /DB_XREF=AV037157 /CLONE=1600022O04 /FEA=EST /CNT=10 /TID=Mm.154915.5 /TIER=Stack /STK=9 /UG=Mm.154915 /LL=20090 /UG_GENE=Rps29 /UG_TITLE=ribosomal protein S29
22121                                                                         gb:AI324936 /DB_XREF=gi:4059365 /DB_XREF=mb49d08.x1 /CLONE=IMAGE:332751 /FEA=EST /CNT=24 /TID=Mm.13020.6 /TIER=Stack /STK=12 /UG=Mm.13020 /LL=22121 /UG_GENE=Rpl13a /UG_TITLE=ribosomal protein L13a
20091                                         gb:NM_016959.1 /DB_XREF=gi:8394217 /GEN=Rps3a /FEA=FLmRNA /CNT=122 /TID=Mm.6957.1 /TIER=FL+Stack /STK=78 /UG=Mm.6957 /LL=20091 /DEF=Mus musculus ribosomal protein S3a (Rps3a), mRNA. /PROD=ribosomal protein S3a /FL=gb:NM_016959.1
67671                                                            gb:AA050777 /DB_XREF=gi:1530594 /DB_XREF=mg72a01.r1 /CLONE=IMAGE:438504 /FEA=EST /CNT=215 /TID=Mm.43330.3 /TIER=Stack /STK=151 /UG=Mm.43330 /LL=67671 /UG_GENE=0610025G13Rik /UG_TITLE=RIKEN cDNA 0610025G13 gene
19241 gb:NM_021278.1 /DB_XREF=gi:10946577 /GEN=Tmsb4x /FEA=FLmRNA /CNT=360 /TID=Mm.142729.1 /TIER=FL+Stack /STK=172 /UG=Mm.142729 /LL=19241 /DEF=Mus musculus thymosin, beta 4, X chromosome (Tmsb4x), mRNA. /PROD=thymosin, beta 4, X chromosome /FL=gb:NM_021278.1 gb:BC018286.1
      Representative Public ID                     Gene Title Gene Symbol
20090                 AV037157          ribosomal protein S29       Rps29
22121                 AI324936         ribosomal protein L13A      Rpl13a
20091                NM_016959         ribosomal protein S3A1      Rps3a1
67671                 AA050777          ribosomal protein L38       Rpl38
19241                NM_021278 thymosin, beta 4, X chromosome      Tmsb4x
      ENTREZ_GENE_ID
20090          20090
22121          22121
20091          20091
67671          67671
19241          19241
                                                                                RefSeq Transcript ID
20090                                                                                      NM_009093
22121                                                                                      NM_009438
20091                                                                                      NM_016959
67671 NM_001048057 /// NM_001048058 /// NM_023372 /// XM_006534007 /// XM_006534008 /// XM_006534009
19241                                                                     NM_021278 /// XM_006528759
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    Gene Ontology Biological Process
20090                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         0006412 // translation // not recorded
22121 0006351 // transcription, DNA-templated // inferred from electronic annotation /// 0006355 // regulation of transcription, DNA-templated // inferred from electronic annotation /// 0006412 // translation // inferred from electronic annotation /// 0006417 // regulation of translation // inferred from electronic annotation /// 0017148 // negative regulation of translation // inferred from direct assay /// 0017148 // negative regulation of translation // not recorded /// 0032496 // response to lipopolysaccharide // inferred from mutant phenotype /// 0032844 // regulation of homeostatic process // inferred from mutant phenotype /// 0048246 // macrophage chemotaxis // inferred from mutant phenotype /// 0060425 // lung morphogenesis // inferred from mutant phenotype /// 0071346 // cellular response to interferon-gamma // inferred from direct assay /// 0071346 // cellular response to interferon-gamma // not recorded /// 1901194 // negative regulation of formation of translation preinitiation complex // not recorded
20091                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      0002181 // cytoplasmic translation // not recorded /// 0006412 // translation // inferred from electronic annotation /// 0030154 // cell differentiation // inferred from electronic annotation /// 0043066 // negative regulation of apoptotic process // inferred from sequence or structural similarity /// 0097194 // execution phase of apoptosis // inferred from sequence or structural similarity
67671                                                                                                                                                                                                                                                                                                                                                                                          0001501 // skeletal system development // inferred from mutant phenotype /// 0001501 // skeletal system development // traceable author statement /// 0001503 // ossification // inferred from mutant phenotype /// 0006412 // translation // inferred from electronic annotation /// 0006417 // regulation of translation // inferred from mutant phenotype /// 0007605 // sensory perception of sound // inferred from mutant phenotype /// 0034463 // 90S preribosome assembly // inferred from direct assay /// 0042474 // middle ear morphogenesis // inferred from mutant phenotype /// 0048318 // axial mesoderm development // inferred from mutant phenotype
19241                                                                                                                                                                                                                                                                                                                                                                                                                      0007010 // cytoskeleton organization // inferred from electronic annotation /// 0014911 // positive regulation of smooth muscle cell migration // traceable author statement /// 0030036 // actin cytoskeleton organization // inferred from electronic annotation /// 0030334 // regulation of cell migration // inferred from direct assay /// 0042989 // sequestering of actin monomers // inferred from electronic annotation /// 0045893 // positive regulation of transcription, DNA-templated // traceable author statement /// 0051152 // positive regulation of smooth muscle cell differentiation // traceable author statement
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      Gene Ontology Cellular Component
20090                                                                                                                                                                                   0005622 // intracellular // inferred from electronic annotation /// 0005737 // cytoplasm // not recorded /// 0005840 // ribosome // inferred from electronic annotation /// 0015935 // small ribosomal subunit // not recorded /// 0022627 // cytosolic small ribosomal subunit // not recorded /// 0030529 // ribonucleoprotein complex // inferred from electronic annotation /// 0070062 // extracellular vesicular exosome // not recorded
22121 0005576 // extracellular region // inferred from electronic annotation /// 0005634 // nucleus // not recorded /// 0005840 // ribosome // inferred from electronic annotation /// 0015934 // large ribosomal subunit // inferred from electronic annotation /// 0016020 // membrane // inferred from electronic annotation /// 0022625 // cytosolic large ribosomal subunit // not recorded /// 0030529 // ribonucleoprotein complex // inferred from direct assay /// 0030529 // ribonucleoprotein complex // not recorded /// 0097452 // GAIT complex // inferred from direct assay /// 0097452 // GAIT complex // not recorded
20091                                                                                                                                                                                               0005622 // intracellular // inferred from electronic annotation /// 0005634 // nucleus // inferred from direct assay /// 0005737 // cytoplasm // inferred from electronic annotation /// 0005829 // cytosol // inferred from direct assay /// 0005840 // ribosome // inferred from electronic annotation /// 0022627 // cytosolic small ribosomal subunit // not recorded /// 0030529 // ribonucleoprotein complex // not recorded
67671                                                                                                                                                                                                                                                                   0005622 // intracellular // inferred from electronic annotation /// 0005840 // ribosome // inferred from electronic annotation /// 0022625 // cytosolic large ribosomal subunit // not recorded /// 0030529 // ribonucleoprotein complex // inferred from electronic annotation /// 0033291 // eukaryotic 80S initiation complex // inferred from direct assay
19241                                                                                                                                                                                                                                                                                                                                                                                         0005634 // nucleus // inferred from direct assay /// 0005737 // cytoplasm // inferred from electronic annotation /// 0005829 // cytosol // inferred from direct assay /// 0005856 // cytoskeleton // inferred from electronic annotation
                                                                                                                                                                                                                                                                                                                                                                                                                                                                    Gene Ontology Molecular Function
20090                                                                                                                                                                                                                                                                                                          0003735 // structural constituent of ribosome // not recorded /// 0008270 // zinc ion binding // not recorded /// 0046872 // metal ion binding // inferred from electronic annotation
22121 0003676 // nucleic acid binding // inferred from electronic annotation /// 0003677 // DNA binding // inferred from electronic annotation /// 0003729 // mRNA binding // inferred from direct assay /// 0003735 // structural constituent of ribosome // inferred from electronic annotation /// 0005125 // cytokine activity // inferred from electronic annotation /// 0044822 // poly(A) RNA binding // not recorded /// 0046872 // metal ion binding // inferred from electronic annotation
20091                                                                                                                                                                                                                                                                                        0003735 // structural constituent of ribosome // not recorded /// 0005515 // protein binding // inferred from physical interaction /// 0031369 // translation initiation factor binding // not recorded
67671                                                                                                                                                                                                                                                                                                                                                                                                           0003735 // structural constituent of ribosome // inferred from electronic annotation
19241                                                                                                                                                                                                                                                                                                                                                           0003779 // actin binding // inferred from electronic annotation /// 0005515 // protein binding // inferred from physical interaction
            origin
20090 1438859_x_at
22121 1455485_x_at
20091 1422475_a_at
67671 1433472_x_at
19241   1415906_at
20765 more rows ...

$Amean
   20090    22121    20091    67671    19241 
16.28272 16.27241 16.25759 16.23408 16.23061 
20765 more elements ...

$method
[1] &quot;ls&quot;

$design
          conditioniTreg conditionNaive conditionnTreg conditionTh1
GSM357839              0              0              0            0
GSM357841              0              0              0            0
GSM357842              0              0              0            1
GSM357843              0              0              0            0
GSM357844              0              0              0            1
          conditionTh17 conditionTh2
GSM357839             0            1
GSM357841             0            1
GSM357842             0            0
GSM357843             1            0
GSM357844             0            0
7 more rows ...

$contrasts
                Contrasts
Levels           conditionTh1 - conditionNaive
  conditioniTreg                             0
  conditionNaive                            -1
  conditionnTreg                             0
  conditionTh1                               1
  conditionTh17                              0
  conditionTh2                               0</code></pre>
<p>Finally, the differential expression analysis is carried out and the
results saved. The results are ranked by <a
href="https://support.bioconductor.org/p/6124/">limma’s moderated
t-statistic</a> and this creates the ranked list of genes.</p>
<pre class="r"><code>fit2 &lt;- eBayes(fit2)
names(topTable(fit2, adjust.method=&quot;BH&quot;, number=12000, sort.by = &quot;B&quot;))</code></pre>
<pre><code> [1] &quot;ID&quot;                               &quot;GB_ACC&quot;                          
 [3] &quot;SPOT_ID&quot;                          &quot;Species.Scientific.Name&quot;         
 [5] &quot;Annotation.Date&quot;                  &quot;Sequence.Type&quot;                   
 [7] &quot;Sequence.Source&quot;                  &quot;Target.Description&quot;              
 [9] &quot;Representative.Public.ID&quot;         &quot;Gene.Title&quot;                      
[11] &quot;Gene.Symbol&quot;                      &quot;ENTREZ_GENE_ID&quot;                  
[13] &quot;RefSeq.Transcript.ID&quot;             &quot;Gene.Ontology.Biological.Process&quot;
[15] &quot;Gene.Ontology.Cellular.Component&quot; &quot;Gene.Ontology.Molecular.Function&quot;
[17] &quot;origin&quot;                           &quot;logFC&quot;                           
[19] &quot;AveExpr&quot;                          &quot;t&quot;                               
[21] &quot;P.Value&quot;                          &quot;adj.P.Val&quot;                       
[23] &quot;B&quot;                               </code></pre>
<pre class="r"><code>de &lt;- data.table(topTable(fit2, adjust.method=&quot;BH&quot;, number=12000, sort.by = &quot;B&quot;), keep.rownames = TRUE)
ranks &lt;- de[order(t), list(rn, t)]
ranks</code></pre>
<pre><code>           rn         t
       &lt;char&gt;     &lt;num&gt;
    1: 170942 -62.22877
    2: 109711 -49.47829
    3:  18124 -43.40540
    4:  12775 -41.16952
    5:  72148 -33.23463
   ---                 
11996:  58801  49.10222
11997:  13730  50.02863
11998:  15937  50.29120
11999:  12772  50.52651
12000:  80876  52.59930</code></pre>
<p>Load <code>exampleRanks</code>.</p>
<pre class="r"><code>data(&quot;exampleRanks&quot;, package = &quot;fgsea&quot;)
head(exampleRanks)</code></pre>
<pre><code>   170942    109711     18124     12775     72148     16010 
-63.33703 -49.74779 -43.63878 -41.51889 -33.26039 -32.77626 </code></pre>
<p>The names of the vector are the Entrez Gene IDs and the values are
the rank metric.</p>
<p>Compare with our results.</p>
<pre class="r"><code>wanted &lt;- head(names(exampleRanks))

ranks[rn %in% wanted]</code></pre>
<pre><code>       rn         t
   &lt;char&gt;     &lt;num&gt;
1: 170942 -62.22877
2: 109711 -49.47829
3:  18124 -43.40540
4:  12775 -41.16952
5:  72148 -33.23463
6:  16010 -32.38200</code></pre>
<p>Visualise the six most significantly down- and up-regulated
genes.</p>
<pre class="r"><code>library(pheatmap)

my_sample &lt;- pData(es)$condition == &quot;Th1&quot; | pData(es)$condition == &quot;Naive&quot; 
my_group &lt;- data.frame(group = pData(es)$condition[my_sample])
row.names(my_group) &lt;- colnames(exprs(es))[my_sample]
 
pheatmap(
  mat = es[c(head(de[order(t), 1])$rn, tail(de[order(t), 1])$rn), my_sample],
  annotation_col = my_group,
  cluster_rows = FALSE,
  cellwidth=25,
  cellheight=15,
  scale = &quot;row&quot;
)</code></pre>
<p><img src="figure/fgsea.Rmd/pheatmap-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-pheatmap-1">
Past versions of pheatmap-1.png
</button>
</p>
<div id="fig-pheatmap-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/davetang/muse/blob/6f2ce9bade8cedffe20e0ef397ea6e46efdc9583/docs/figure/fgsea.Rmd/pheatmap-1.png" target="_blank">6f2ce9b</a>
</td>
<td>
Dave Tang
</td>
<td>
2025-01-24
</td>
</tr>
<tr>
<td>
<a href="https://github.com/davetang/muse/blob/b82b4fccdedc76ac273db0a488b52fbe64437164/docs/figure/fgsea.Rmd/pheatmap-1.png" target="_blank">b82b4fc</a>
</td>
<td>
Dave Tang
</td>
<td>
2023-02-01
</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<div id="vignette-analysis" class="section level2">
<h2>Vignette analysis</h2>
<p>The following section is based on the <a
href="https://bioconductor.org/packages/release/bioc/vignettes/fgsea/inst/doc/fgsea-tutorial.html">fgsea
tutorial</a> but with my elaborations. The pathways are stored in
<code>examplePathways</code> and the ranked gene list in
<code>exampleRanks</code>.</p>
<p>The <code>fgsea()</code> function runs the pre-ranked gene set
enrichment analysis.</p>
<pre class="r"><code>data(examplePathways)
data(exampleRanks)
set.seed(42)

system.time(
  fgseaRes &lt;- fgsea(
    pathways = examplePathways, 
    stats = exampleRanks,
    minSize=15,
    maxSize=500
  )
) </code></pre>
<pre><code>   user  system elapsed 
  4.324   0.137   3.312 </code></pre>
<pre class="r"><code>class(fgseaRes)</code></pre>
<pre><code>[1] &quot;data.table&quot; &quot;data.frame&quot;</code></pre>
<p>Top 6 enriched pathways; see the <a
href="https://github.com/Rdatatable/data.table/wiki">data.table Wiki</a>
for more information on the <code>data.table</code> package.</p>
<pre class="r"><code>head(fgseaRes[order(pval), ])</code></pre>
<pre><code>                                           pathway         pval         padj
                                            &lt;char&gt;        &lt;num&gt;        &lt;num&gt;
1:                     5990979_Cell_Cycle,_Mitotic 6.690481e-27 3.920622e-24
2:                              5990980_Cell_Cycle 3.312565e-26 9.705816e-24
3:                    5991851_Mitotic_Prometaphase 8.470173e-19 1.654507e-16
4: 5992217_Resolution_of_Sister_Chromatid_Cohesion 2.176649e-18 3.188791e-16
5:                                 5991454_M_Phase 1.873997e-14 2.196325e-12
6:         5991599_Separation_of_Sister_Chromatids 8.733223e-14 8.529448e-12
     log2err        ES      NES  size  leadingEdge
       &lt;num&gt;     &lt;num&gt;    &lt;num&gt; &lt;int&gt;       &lt;list&gt;
1: 1.3422338 0.5594755 2.769070   317 66336, 6....
2: 1.3267161 0.5388497 2.705894   369 66336, 6....
3: 1.1239150 0.7253270 2.972690    82 66336, 6....
4: 1.1053366 0.7347987 2.957518    74 66336, 6....
5: 0.9759947 0.5576247 2.554076   173 66336, 6....
6: 0.9545416 0.6164600 2.670030   116 66336, 6....</code></pre>
<p>Number of significant pathways at padj &lt; 0.01.</p>
<pre class="r"><code>sum(fgseaRes[, padj &lt; 0.01])</code></pre>
<pre><code>[1] 76</code></pre>
<p>Plot the most significantly enriched pathway.</p>
<pre class="r"><code>library(ggplot2)
plotEnrichment(
  examplePathways[[head(fgseaRes[order(pval), ], 1)$pathway]],
  exampleRanks,
  ticksSize = 0.2 # default
) +
  labs(title=head(fgseaRes[order(pval), ], 1)$pathway) -&gt; p

p</code></pre>
<p><img src="figure/fgsea.Rmd/plot_sign_pathway-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-plot_sign_pathway-1">
Past versions of plot_sign_pathway-1.png
</button>
</p>
<div id="fig-plot_sign_pathway-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/davetang/muse/blob/6f2ce9bade8cedffe20e0ef397ea6e46efdc9583/docs/figure/fgsea.Rmd/plot_sign_pathway-1.png" target="_blank">6f2ce9b</a>
</td>
<td>
Dave Tang
</td>
<td>
2025-01-24
</td>
</tr>
<tr>
<td>
<a href="https://github.com/davetang/muse/blob/f48870229f7197cbea7eca54661b1a1d6a26167f/docs/figure/fgsea.Rmd/plot_sign_pathway-1.png" target="_blank">f488702</a>
</td>
<td>
Dave Tang
</td>
<td>
2023-02-13
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>Check out the object returned by <code>plotEnrichment()</code>.</p>
<pre class="r"><code>class(p)</code></pre>
<pre><code>[1] &quot;gg&quot;     &quot;ggplot&quot;</code></pre>
<pre class="r"><code>names(p)</code></pre>
<pre><code> [1] &quot;data&quot;        &quot;layers&quot;      &quot;scales&quot;      &quot;guides&quot;      &quot;mapping&quot;    
 [6] &quot;theme&quot;       &quot;coordinates&quot; &quot;facet&quot;       &quot;plot_env&quot;    &quot;layout&quot;     
[11] &quot;labels&quot;     </code></pre>
<p>Look inside layers.</p>
<pre class="r"><code>p$layers</code></pre>
<pre><code>[[1]]
mapping: x = ~rank, y = ~ES 
geom_line: na.rm = FALSE, orientation = NA
stat_identity: na.rm = FALSE
position_identity 

[[2]]
mapping: x = ~rank, y = ~-spreadES/16, xend = ~rank, yend = ~spreadES/16 
geom_segment: arrow = NULL, arrow.fill = NULL, lineend = butt, linejoin = round, na.rm = FALSE
stat_identity: na.rm = FALSE
position_identity 

[[3]]
mapping: yintercept = ~yintercept 
geom_hline: na.rm = FALSE
stat_identity: na.rm = FALSE
position_identity 

[[4]]
mapping: yintercept = ~yintercept 
geom_hline: na.rm = FALSE
stat_identity: na.rm = FALSE
position_identity 

[[5]]
mapping: yintercept = ~yintercept 
geom_hline: na.rm = FALSE
stat_identity: na.rm = FALSE
position_identity </code></pre>
<p>Change line width of <code>geom_line</code>.</p>
<pre class="r"><code>p$layers[[1]]$aes_params$size &lt;- 2
p</code></pre>
<p><img src="figure/fgsea.Rmd/change_line_width-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>Plot the top 10 pathways enriched at the top and bottom of the ranked
list, respectively.</p>
<pre class="r"><code>topPathwaysUp &lt;- fgseaRes[ES &gt; 0][head(order(pval), n=10), pathway]
topPathwaysDown &lt;- fgseaRes[ES &lt; 0][head(order(pval), n=10), pathway]
topPathways &lt;- c(topPathwaysUp, rev(topPathwaysDown))
plotGseaTable(
  examplePathways[topPathways],
  exampleRanks,
  fgseaRes, 
  gseaParam = 0.5
)</code></pre>
<p><img src="figure/fgsea.Rmd/plot_gsea_table-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-plot_gsea_table-1">
Past versions of plot_gsea_table-1.png
</button>
</p>
<div id="fig-plot_gsea_table-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/davetang/muse/blob/6f2ce9bade8cedffe20e0ef397ea6e46efdc9583/docs/figure/fgsea.Rmd/plot_gsea_table-1.png" target="_blank">6f2ce9b</a>
</td>
<td>
Dave Tang
</td>
<td>
2025-01-24
</td>
</tr>
<tr>
<td>
<a href="https://github.com/davetang/muse/blob/f48870229f7197cbea7eca54661b1a1d6a26167f/docs/figure/fgsea.Rmd/plot_gsea_table-1.png" target="_blank">f488702</a>
</td>
<td>
Dave Tang
</td>
<td>
2023-02-13
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>Since the genes were ranked according to their differential
expression significance and fold change, with the most significantly
down-regulated genes at the top and up-regulated genes at the bottom of
the list, the enriched gene sets provides us with some idea of the
function of these genes. For example, <code>exampleRanks</code> was
generated by testing naive T cells against T helper 1 cells, so the
enriched pathways may suggest how naive cells are differentiated into T
helper 1 cells.</p>
</div>
<div id="reactome" class="section level2">
<h2>Reactome</h2>
<p><a href="https://reactome.org/">Reactome pathways</a> can also be
used with fgsea and this database is available via Bioconductor (and was
installed earlier in this post).</p>
<p>The <code>reactomePathways()</code> function returns a list of
Reactome pathways given a list of Entrez Gene IDs. In the example data,
the IDs are stored as the names in the <code>exampleRanks</code>
vector.</p>
<pre class="r"><code>my_pathways &lt;- reactomePathways(names(exampleRanks))</code></pre>
<pre><code>&#39;select()&#39; returned 1:many mapping between keys and columns</code></pre>
<pre><code>&#39;select()&#39; returned 1:1 mapping between keys and columns</code></pre>
<p>Out of interest, let’s check the median number of genes in the
pathways.</p>
<pre class="r"><code>summary(sapply(my_pathways, length))</code></pre>
<pre><code>   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
   1.00    4.00   11.00   30.07   29.00 1250.00 </code></pre>
<p>The results are very similar to the example pathways, where the
pathways are related to the cell cycle and mitosis.</p>
<pre class="r"><code>set.seed(42)

system.time(
  fgsea_reactome &lt;- fgsea(
    pathways = my_pathways, 
    stats = exampleRanks,
    minSize=15,
    maxSize=500
  )
)</code></pre>
<pre><code>   user  system elapsed 
  3.813   0.212   2.228 </code></pre>
<pre class="r"><code>head(fgsea_reactome[order(pval), ])</code></pre>
<pre><code>                                   pathway         pval         padj   log2err
                                    &lt;char&gt;        &lt;num&gt;        &lt;num&gt;     &lt;num&gt;
1:                              Cell Cycle 8.200375e-28 5.986274e-25 1.3651796
2:                     Cell Cycle, Mitotic 6.852361e-27 2.501112e-24 1.3422338
3:                  Cell Cycle Checkpoints 1.080789e-20 2.629921e-18 1.1778933
4: Resolution of Sister Chromatid Cohesion 3.987687e-16 7.277529e-14 1.0376962
5:                    Mitotic Prometaphase 1.121259e-15 1.637039e-13 1.0175448
6:                                 S Phase 2.413133e-14 2.935978e-12 0.9759947
          ES      NES  size  leadingEdge
       &lt;num&gt;    &lt;num&gt; &lt;int&gt;       &lt;list&gt;
1: 0.5250345 2.661002   422 66336, 6....
2: 0.5440528 2.724387   357 66336, 6....
3: 0.6068232 2.830899   190 66336, 6....
4: 0.6806110 2.815193    91 66336, 6....
5: 0.5948613 2.697816   149 66336, 6....
6: 0.6295907 2.706269   107 12428, 5....</code></pre>
</div>
<div id="leading-edge" class="section level2">
<h2>Leading edge</h2>
<p>Again quote the <a
href="https://pubmed.ncbi.nlm.nih.gov/16199517/">original paper</a>:</p>
<blockquote>
<p>Often it is useful to extract the core members of high scoring gene
sets that contribute to the enrichment score. We define the leading-edge
subset to be those genes in the gene set <strong>S</strong> that appear
in the ranked list <strong>L</strong> at, or before, the point where the
running sum reaches its maximum deviation from zero.</p>
</blockquote>
<p>The <code>fgsea</code> results includes the leading edge, which
contains the genes that contributed to the enrichment score. To
illustrate, let’s take the leader edge genes from the most significant
Reactome pathway.</p>
<pre class="r"><code>fgsea_reactome[order(pval),][1,]</code></pre>
<pre><code>      pathway         pval         padj log2err        ES      NES  size
       &lt;char&gt;        &lt;num&gt;        &lt;num&gt;   &lt;num&gt;     &lt;num&gt;    &lt;num&gt; &lt;int&gt;
1: Cell Cycle 8.200375e-28 5.986274e-25 1.36518 0.5250345 2.661002   422
    leadingEdge
         &lt;list&gt;
1: 66336, 6....</code></pre>
<p>The set of leading edge genes is made up of 197 genes.</p>
<pre class="r"><code>fgsea_reactome[order(pval),][1,]$leadingEdge[[1]]</code></pre>
<pre><code>  [1] &quot;66336&quot;  &quot;66977&quot;  &quot;15366&quot;  &quot;12442&quot;  &quot;107995&quot; &quot;66442&quot;  &quot;19361&quot;  &quot;12571&quot; 
  [9] &quot;12428&quot;  &quot;52276&quot;  &quot;54392&quot;  &quot;66311&quot;  &quot;215387&quot; &quot;67629&quot;  &quot;54124&quot;  &quot;12649&quot; 
 [17] &quot;72415&quot;  &quot;56150&quot;  &quot;57441&quot;  &quot;20877&quot;  &quot;67121&quot;  &quot;12615&quot;  &quot;66468&quot;  &quot;67849&quot; 
 [25] &quot;19053&quot;  &quot;73804&quot;  &quot;76044&quot;  &quot;20878&quot;  &quot;15270&quot;  &quot;13555&quot;  &quot;60411&quot;  &quot;60530&quot; 
 [33] &quot;17219&quot;  &quot;69270&quot;  &quot;12575&quot;  &quot;69263&quot;  &quot;12448&quot;  &quot;14211&quot;  &quot;20873&quot;  &quot;18005&quot; 
 [41] &quot;72119&quot;  &quot;71988&quot;  &quot;12189&quot;  &quot;72107&quot;  &quot;17215&quot;  &quot;12534&quot;  &quot;66156&quot;  &quot;208628&quot;
 [49] &quot;237911&quot; &quot;22390&quot;  &quot;68240&quot;  &quot;228421&quot; &quot;68014&quot;  &quot;269582&quot; &quot;19348&quot;  &quot;12236&quot; 
 [57] &quot;72151&quot;  &quot;18817&quot;  &quot;21781&quot;  &quot;18968&quot;  &quot;217653&quot; &quot;66934&quot;  &quot;272551&quot; &quot;227613&quot;
 [65] &quot;67141&quot;  &quot;67951&quot;  &quot;68612&quot;  &quot;68298&quot;  &quot;108000&quot; &quot;23834&quot;  &quot;106344&quot; &quot;242705&quot;
 [73] &quot;18141&quot;  &quot;223921&quot; &quot;26886&quot;  &quot;13557&quot;  &quot;26909&quot;  &quot;72787&quot;  &quot;268697&quot; &quot;72155&quot; 
 [81] &quot;56371&quot;  &quot;17535&quot;  &quot;107823&quot; &quot;12531&quot;  &quot;327762&quot; &quot;12567&quot;  &quot;229841&quot; &quot;67052&quot; 
 [89] &quot;16319&quot;  &quot;66634&quot;  &quot;171567&quot; &quot;26931&quot;  &quot;67203&quot;  &quot;381306&quot; &quot;12235&quot;  &quot;19891&quot; 
 [97] &quot;74470&quot;  &quot;72083&quot;  &quot;66570&quot;  &quot;17216&quot;  &quot;76308&quot;  &quot;19687&quot;  &quot;17218&quot;  &quot;102920&quot;
[105] &quot;29870&quot;  &quot;18973&quot;  &quot;16881&quot;  &quot;17463&quot;  &quot;68964&quot;  &quot;75786&quot;  &quot;19645&quot;  &quot;19075&quot; 
[113] &quot;26417&quot;  &quot;69736&quot;  &quot;19357&quot;  &quot;76816&quot;  &quot;70385&quot;  &quot;70645&quot;  &quot;22628&quot;  &quot;225182&quot;
[121] &quot;22627&quot;  &quot;52683&quot;  &quot;19076&quot;  &quot;18972&quot;  &quot;26932&quot;  &quot;12544&quot;  &quot;17997&quot;  &quot;26440&quot; 
[129] &quot;68549&quot;  &quot;12445&quot;  &quot;269113&quot; &quot;26444&quot;  &quot;19324&quot;  &quot;93765&quot;  &quot;103733&quot; &quot;59001&quot; 
[137] &quot;107976&quot; &quot;19179&quot;  &quot;12579&quot;  &quot;232987&quot; &quot;17420&quot;  &quot;219072&quot; &quot;26445&quot;  &quot;105988&quot;
[145] &quot;69745&quot;  &quot;18538&quot;  &quot;69928&quot;  &quot;11651&quot;  &quot;235559&quot; &quot;68097&quot;  &quot;57296&quot;  &quot;63955&quot; 
[153] &quot;19170&quot;  &quot;17246&quot;  &quot;17220&quot;  &quot;12144&quot;  &quot;208084&quot; &quot;50793&quot;  &quot;77605&quot;  &quot;18392&quot; 
[161] &quot;236930&quot; &quot;70024&quot;  &quot;66296&quot;  &quot;16906&quot;  &quot;109145&quot; &quot;71819&quot;  &quot;67733&quot;  &quot;50883&quot; 
[169] &quot;12447&quot;  &quot;75430&quot;  &quot;12532&quot;  &quot;68942&quot;  &quot;26442&quot;  &quot;19177&quot;  &quot;230376&quot; &quot;245688&quot;
[177] &quot;229776&quot; &quot;214901&quot; &quot;16563&quot;  &quot;16328&quot;  &quot;18971&quot;  &quot;77011&quot;  &quot;225659&quot; &quot;68953&quot; 
[185] &quot;100088&quot; &quot;66671&quot;  &quot;19718&quot;  &quot;78177&quot;  &quot;66578&quot;  &quot;59046&quot;  &quot;22194&quot;  &quot;52530&quot; 
[193] &quot;23996&quot;  &quot;101565&quot; &quot;16475&quot;  &quot;381280&quot; &quot;235661&quot;</code></pre>
<p>The Cell Cycle contains 422 genes, which was indicated in the
<code>fgsea</code> result as size (but it’s always good to
double-check).</p>
<pre class="r"><code>length(my_pathways[[fgsea_reactome[order(pval),][1,]$pathway]])</code></pre>
<pre><code>[1] 422</code></pre>
</div>
<div id="summary" class="section level2">
<h2>Summary</h2>
<p>The Gene Set Enrichment Analysis (GSEA) has been around since 2005
and has become a routine step when analysing gene expression data. It
differs from Gene Ontology enrichment analysis in that it considers all
genes in contrast to taking only significantly differentially expressed
genes.</p>
<p>The <code>fgsea</code> package allows one to conduct a pre-ranked
GSEA in R, which is one approach in a GSEA. A p-value is estimated by
permuting the genes in a gene set, which leads to randomly assigned gene
sets of the same size. Note that “This approach is not strictly accurate
because it ignores gene-gene correlations and will overestimate the
significance levels and may lead to false positives.” However, it is
still useful in getting an idea of the potential roles of the genes that
are up- and down-regulated. If your pre-ranked GSEA returns no
significant gene sets, you may still get an idea of what roles the up-
and down-regulated genes may be involved in by examining the leading
edge set. This set indicates the genes that contributed to the
enrichment score.</p>
<p>The example ranks in the <code>fgsea</code> package were ranked based
on the moderated t-statistic. <a
href="https://genomespot.blogspot.com/2014/09/data-analysis-step-8-pathway-analysis.html">Another
metric</a> that you may consider is to take the signed fold change and
multiply it by the <span
class="math inline">\(-log_{10}\)</span>p-value. For example, if you
performed your differential expression analysis with <code>edgeR</code>,
you can simply multiply the signed fold change column to the <span
class="math inline">\(-log_{10}\)</span>p-value column. This metric is
based on the paper: <a
href="https://pubmed.ncbi.nlm.nih.gov/20660011/">Rank-rank
hypergeometric overlap: identification of statistically significant
overlap between gene-expression signatures</a>.</p>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span>
Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 4.5.0 (2025-04-11)
Platform: x86_64-pc-linux-gnu
Running under: Ubuntu 24.04.2 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 
LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.26.so;  LAPACK version 3.12.0

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

time zone: Etc/UTC
tzcode source: system (glibc)

attached base packages:
[1] stats4    stats     graphics  grDevices utils     datasets  methods  
[8] base     

other attached packages:
 [1] ggplot2_3.5.2        pheatmap_1.0.12      data.table_1.17.2   
 [4] reactome.db_1.92.0   org.Mm.eg.db_3.21.0  AnnotationDbi_1.70.0
 [7] IRanges_2.42.0       S4Vectors_0.46.0     limma_3.64.0        
[10] GEOquery_2.76.0      Biobase_2.68.0       BiocGenerics_0.54.0 
[13] generics_0.1.4       fgsea_1.34.0         BiocManager_1.30.25 
[16] workflowr_1.7.1     

loaded via a namespace (and not attached):
 [1] DBI_1.2.3                   httr2_1.1.2                
 [3] rlang_1.1.6                 magrittr_2.0.3             
 [5] git2r_0.36.2                matrixStats_1.5.0          
 [7] compiler_4.5.0              RSQLite_2.3.11             
 [9] getPass_0.2-4               png_0.1-8                  
[11] callr_3.7.6                 vctrs_0.6.5                
[13] stringr_1.5.1               pkgconfig_2.0.3            
[15] crayon_1.5.3                fastmap_1.2.0              
[17] XVector_0.48.0              labeling_0.4.3             
[19] promises_1.3.2              rmarkdown_2.29             
[21] tzdb_0.5.0                  UCSC.utils_1.4.0           
[23] ps_1.9.1                    purrr_1.0.4                
[25] bit_4.6.0                   xfun_0.52                  
[27] cachem_1.1.0                GenomeInfoDb_1.44.0        
[29] jsonlite_2.0.0              blob_1.2.4                 
[31] later_1.4.2                 DelayedArray_0.34.1        
[33] BiocParallel_1.42.0         parallel_4.5.0             
[35] R6_2.6.1                    bslib_0.9.0                
[37] stringi_1.8.7               RColorBrewer_1.1-3         
[39] GenomicRanges_1.60.0        jquerylib_0.1.4            
[41] Rcpp_1.0.14                 SummarizedExperiment_1.38.1
[43] knitr_1.50                  R.utils_2.13.0             
[45] readr_2.1.5                 httpuv_1.6.16              
[47] rentrez_1.2.3               Matrix_1.7-3               
[49] tidyselect_1.2.1            rstudioapi_0.17.1          
[51] abind_1.4-8                 yaml_2.3.10                
[53] codetools_0.2-20            curl_6.2.3                 
[55] processx_3.8.6              lattice_0.22-6             
[57] tibble_3.2.1                withr_3.0.2                
[59] KEGGREST_1.48.0             evaluate_1.0.3             
[61] xml2_1.3.8                  Biostrings_2.76.0          
[63] pillar_1.10.2               MatrixGenerics_1.20.0      
[65] whisker_0.4.1               rprojroot_2.0.4            
[67] hms_1.1.3                   scales_1.4.0               
[69] glue_1.8.0                  tools_4.5.0                
[71] fs_1.6.6                    XML_3.99-0.18              
[73] fastmatch_1.1-6             cowplot_1.1.3              
[75] grid_4.5.0                  tidyr_1.3.1                
[77] GenomeInfoDbData_1.2.14     cli_3.6.5                  
[79] rappdirs_0.3.3              S4Arrays_1.8.0             
[81] dplyr_1.1.4                 gtable_0.3.6               
[83] R.methodsS3_1.8.2           sass_0.4.10                
[85] digest_0.6.37               SparseArray_1.8.0          
[87] farver_2.1.2                R.oo_1.27.1                
[89] memoise_2.0.1               htmltools_0.5.8.1          
[91] lifecycle_1.0.4             httr_1.4.7                 
[93] statmod_1.5.0               bit64_4.6.0-1              </code></pre>
</div>
</div>
</div>
</div>
</div>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
https://docs.mathjax.org/en/latest/web/configuration.html. This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>




</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
