<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />



<meta name="date" content="2025-12-03" />

<title>Fisher’s Exact Test</title>

<script src="site_libs/header-attrs-2.29/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.13.2/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-6.5.2/css/all.min.css" rel="stylesheet" />
<link href="site_libs/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet" />

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/main/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">My blog</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/davetang/muse">
    <span class="fa fa-github"></span>
     
    Source code
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Fisher’s Exact Test</h1>
<h4 class="date">2025-12-03</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span>
workflowr <span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2025-12-03
</p>
<p>
<strong>Checks:</strong> <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 7
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong> <code>muse/</code> <span
class="glyphicon glyphicon-question-sign" aria-hidden="true"
title="This is the local directory in which the code in this file was executed.">
</span>
</p>
<p>
This reproducible <a href="https://rmarkdown.rstudio.com">R Markdown</a>
analysis was created with <a
  href="https://github.com/workflowr/workflowr">workflowr</a> (version
1.7.1). The <em>Checks</em> tab describes the reproducibility checks
that were applied when the results were created. The <em>Past
versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date
</a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git
repository, you know the exact version of the code that produced these
results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the
global environment can affect the analysis in your R Markdown file in
unknown ways. For reproduciblity it’s best to always run the code in an
empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20200712code">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Seed:</strong>
<code>set.seed(20200712)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20200712code"
class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20200712)</code> was run prior to running
the code in the R Markdown file. Setting a seed ensures that any results
that rely on randomness, e.g. subsampling or permutations, are
reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Session information:</strong>
recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package
versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be
confident that you successfully produced the results during this
run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr
project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomdavetangmusetreeb52ca96f8e5a62856ef9af0a19fadba01612ff74targetblankb52ca96a">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Repository version:</strong>
<a href="https://github.com/davetang/muse/tree/b52ca96f8e5a62856ef9af0a19fadba01612ff74" target="_blank">b52ca96</a>
</a>
</p>
</div>
<div
id="strongRepositoryversionstrongahrefhttpsgithubcomdavetangmusetreeb52ca96f8e5a62856ef9af0a19fadba01612ff74targetblankb52ca96a"
class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development
and connecting the code version to the results is critical for
reproducibility.
</p>
<p>
The results in this page were generated with repository version
<a href="https://github.com/davetang/muse/tree/b52ca96f8e5a62856ef9af0a19fadba01612ff74" target="_blank">b52ca96</a>.
See the <em>Past versions</em> tab to see a history of the changes made
to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for
the analysis have been committed to Git prior to generating the results
(you can use <code>wflow_publish</code> or
<code>wflow_git_commit</code>). workflowr only checks the R Markdown
file, but you know if there are other scripts or data files that it
depends on. Below is the status of the Git repository when the results
were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .Rproj.user/
    Ignored:    data/1M_neurons_filtered_gene_bc_matrices_h5.h5
    Ignored:    data/293t/
    Ignored:    data/293t_3t3_filtered_gene_bc_matrices.tar.gz
    Ignored:    data/293t_filtered_gene_bc_matrices.tar.gz
    Ignored:    data/5k_Human_Donor1_PBMC_3p_gem-x_5k_Human_Donor1_PBMC_3p_gem-x_count_sample_filtered_feature_bc_matrix.h5
    Ignored:    data/5k_Human_Donor2_PBMC_3p_gem-x_5k_Human_Donor2_PBMC_3p_gem-x_count_sample_filtered_feature_bc_matrix.h5
    Ignored:    data/5k_Human_Donor3_PBMC_3p_gem-x_5k_Human_Donor3_PBMC_3p_gem-x_count_sample_filtered_feature_bc_matrix.h5
    Ignored:    data/5k_Human_Donor4_PBMC_3p_gem-x_5k_Human_Donor4_PBMC_3p_gem-x_count_sample_filtered_feature_bc_matrix.h5
    Ignored:    data/97516b79-8d08-46a6-b329-5d0a25b0be98.h5ad
    Ignored:    data/Parent_SC3v3_Human_Glioblastoma_filtered_feature_bc_matrix.tar.gz
    Ignored:    data/brain_counts/
    Ignored:    data/cl.obo
    Ignored:    data/cl.owl
    Ignored:    data/jurkat/
    Ignored:    data/jurkat:293t_50:50_filtered_gene_bc_matrices.tar.gz
    Ignored:    data/jurkat_293t/
    Ignored:    data/jurkat_filtered_gene_bc_matrices.tar.gz
    Ignored:    data/pbmc20k/
    Ignored:    data/pbmc20k_seurat/
    Ignored:    data/pbmc3k.h5ad
    Ignored:    data/pbmc3k/
    Ignored:    data/pbmc3k_bpcells_mat/
    Ignored:    data/pbmc3k_export.mtx
    Ignored:    data/pbmc3k_matrix.mtx
    Ignored:    data/pbmc3k_seurat.rds
    Ignored:    data/pbmc4k_filtered_gene_bc_matrices.tar.gz
    Ignored:    data/pbmc_1k_v3_filtered_feature_bc_matrix.h5
    Ignored:    data/pbmc_1k_v3_raw_feature_bc_matrix.h5
    Ignored:    data/refdata-gex-GRCh38-2020-A.tar.gz
    Ignored:    data/seurat_1m_neuron.rds
    Ignored:    data/t_3k_filtered_gene_bc_matrices.tar.gz
    Ignored:    r_packages_4.4.1/
    Ignored:    r_packages_4.5.0/

Untracked files:
    Untracked:  analysis/bioc.Rmd
    Untracked:  analysis/bioc_scrnaseq.Rmd
    Untracked:  bpcells_matrix/
    Untracked:  data/Caenorhabditis_elegans.WBcel235.113.gtf.gz
    Untracked:  data/GCF_043380555.1-RS_2024_12_gene_ontology.gaf.gz
    Untracked:  data/arab.rds
    Untracked:  data/astronomicalunit.csv
    Untracked:  data/femaleMiceWeights.csv
    Untracked:  data/lung_bcell.rds
    Untracked:  m3/
    Untracked:  women.json

Unstaged changes:
    Modified:   analysis/isoform_switch_analyzer.Rmd

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not
included in this status report because it is ok for generated content to
have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">
<p>
These are the previous versions of the repository in which changes were
made to the R Markdown (<code>analysis/fisher.Rmd</code>) and HTML
(<code>docs/fisher.html</code>) files. If you’ve configured a remote Git
repository (see <code>?wflow_git_remote</code>), click on the hyperlinks
in the table below to view the files as they were in that past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/davetang/muse/blob/b52ca96f8e5a62856ef9af0a19fadba01612ff74/analysis/fisher.Rmd" target="_blank">b52ca96</a>
</td>
<td>
Dave Tang
</td>
<td>
2025-12-03
</td>
<td>
Provide some background information
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/davetang/muse/2f260a0c346250dc3596ae858388048a6356e1fc/docs/fisher.html" target="_blank">2f260a0</a>
</td>
<td>
Dave Tang
</td>
<td>
2025-03-29
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/davetang/muse/blob/a1be60c0835f4d06bc595e05635df1173e18a82e/analysis/fisher.Rmd" target="_blank">a1be60c</a>
</td>
<td>
Dave Tang
</td>
<td>
2025-03-29
</td>
<td>
Left justify by adding colon
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/davetang/muse/6b4d1896dda40cd698ea4cc1e117c1779bc3701e/docs/fisher.html" target="_blank">6b4d189</a>
</td>
<td>
Dave Tang
</td>
<td>
2025-03-29
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/davetang/muse/blob/166142600f4fb589bd1b72b047f197f082845425/analysis/fisher.Rmd" target="_blank">1661426</a>
</td>
<td>
Dave Tang
</td>
<td>
2025-03-29
</td>
<td>
Left justify
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/davetang/muse/3a243513d90158ec87a1287135dae6bb1cff0741/docs/fisher.html" target="_blank">3a24351</a>
</td>
<td>
Dave Tang
</td>
<td>
2025-03-29
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/davetang/muse/blob/cd487f9642fa8160347df6801f48280aa4e374ef/analysis/fisher.Rmd" target="_blank">cd487f9</a>
</td>
<td>
Dave Tang
</td>
<td>
2025-03-29
</td>
<td>
Label significant cluster in volcano plot
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/davetang/muse/c33357f99489b86dbe995e98d15daac458907cfc/docs/fisher.html" target="_blank">c33357f</a>
</td>
<td>
Dave Tang
</td>
<td>
2025-03-29
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/davetang/muse/blob/e9e40c4306ce9ec323be7deeda74d15e31f128df/analysis/fisher.Rmd" target="_blank">e9e40c4</a>
</td>
<td>
Dave Tang
</td>
<td>
2025-03-29
</td>
<td>
Add some notes
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/davetang/muse/4bccd4022ab4588c013a279efa1b8b2a02b1f6ba/docs/fisher.html" target="_blank">4bccd40</a>
</td>
<td>
Dave Tang
</td>
<td>
2025-03-29
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/davetang/muse/blob/77bdfb3d74f5f47969d750e47bb312328f577929/analysis/fisher.Rmd" target="_blank">77bdfb3</a>
</td>
<td>
Dave Tang
</td>
<td>
2025-03-29
</td>
<td>
Volcano plot and cluster proportions
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/davetang/muse/1b5016da96362435160f669e83855b33733bc94f/docs/fisher.html" target="_blank">1b5016d</a>
</td>
<td>
Dave Tang
</td>
<td>
2025-03-29
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/davetang/muse/blob/a86eba56934106bc159badbfe3e34393631cda06/analysis/fisher.Rmd" target="_blank">a86eba5</a>
</td>
<td>
Dave Tang
</td>
<td>
2025-03-29
</td>
<td>
Per-cluster Fisher’s Exact Test
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/davetang/muse/5477eb3ca70d298bf9b527f64df4559de3abc4c7/docs/fisher.html" target="_blank">5477eb3</a>
</td>
<td>
Dave Tang
</td>
<td>
2025-03-29
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/davetang/muse/blob/aa71d484ac6bda06628ca6f67eafdc0a508c086c/analysis/fisher.Rmd" target="_blank">aa71d48</a>
</td>
<td>
Dave Tang
</td>
<td>
2025-03-29
</td>
<td>
Fisher’s Exact Test
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<div id="introduction" class="section level2">
<h2>Introduction</h2>
<p>Fisher’s Exact Test is a statistical method for determining whether
there is a significant association between two categorical variables. It
is particularly useful when:</p>
<ul>
<li>Sample sizes are small</li>
<li>Expected cell frequencies are low (&lt; 5 in any cell)</li>
<li>You need precise p-values, i.e., an exact test, without relying on
large-sample approximations</li>
</ul>
<p>The test examines whether the observed pattern in your data could
reasonably occur by chance alone, or whether it suggests a genuine
relationship between the variables. It works with data organised in
contingency tables (most commonly 2×2, but applicable to larger tables
as well).</p>
<p>Some guidelines for when to use Fisher’s Exact Test:</p>
<ul>
<li>Any cell has an <strong>expected count</strong> less than 5</li>
<li>Total sample size less than 20-40</li>
<li>The test works for any sample size, but may be overkill for large
datasets</li>
</ul>
<p>Regarding expected counts, this is not the same as observed values in
the data.</p>
<pre class="r"><code>eg1 &lt;- matrix(
  c(8, 3, 2, 7), 
  nrow = 2,
  dimnames = list(
    c(&quot;A&quot;, &quot;B&quot;),
    c(&quot;Success&quot;, &quot;Failure&quot;)
  )
)
eg1</code></pre>
<pre><code>  Success Failure
A       8       2
B       3       7</code></pre>
<p>The observed counts are 8, 3, 2, 7 and the expected counts are what
you would expect if there were no association between the groups and
outcomes.</p>
<p>Expected Count = (Row Total * Column Total) / Grand Total.</p>
<pre class="r"><code>row_totals &lt;- rowSums(eg1)
col_totals &lt;- colSums(eg1)
grand_total &lt;- sum(eg1)

expected_counts &lt;- matrix(0, nrow = 2, ncol = 2, dimnames = dimnames(eg1))

for(i in 1:2) {
  for(j in 1:2) {
    expected_counts[i,j] &lt;- row_totals[i] * col_totals[j] / grand_total
  }
}

expected_counts</code></pre>
<pre><code>  Success Failure
A     5.5     4.5
B     5.5     4.5</code></pre>
<ul>
<li>Cell (Group A, Success): (10 × 11) / 20 = 5.5</li>
<li>Cell (Group A, Failure): (10 × 9) / 20 = 4.5</li>
<li>Cell (Group B, Success): (10 × 11) / 20 = 5.5</li>
<li>Cell (Group B, Failure): (10 × 9) / 20 = 4.5</li>
</ul>
<p>Since two cells have expected counts &lt; 5, Fisher’s Exact Test is
appropriate for this data. The chi-square test approximation relies on
expected frequencies being large enough (&gt;= 5). When expected counts
are small, the chi-square approximation becomes unreliable, and Fisher’s
Exact Test should be used instead.</p>
<p>Calculate expected counts with a Chi-square test.</p>
<pre class="r"><code>chisq_test &lt;- chisq.test(eg1)</code></pre>
<pre><code>Warning in chisq.test(eg1): Chi-squared approximation may be incorrect</code></pre>
<pre class="r"><code>chisq_test$expected</code></pre>
<pre><code>  Success Failure
A     5.5     4.5
B     5.5     4.5</code></pre>
</div>
<div id="clinical-trial" class="section level2">
<h2>Clinical trial</h2>
<p>Imagine a small clinical trial to test a new drug for migraines. 20
patients are recruited and randomly assigned to whether they will
receive the new drug or a placebo. After the treatment, whether patient
experienced relief from their migraine was recorded.</p>
<pre class="r"><code>clinical_data &lt;- matrix(
  c(8, 3, 2, 7), 
  nrow = 2,
  dimnames = list(
    c(&quot;New Drug&quot;, &quot;Placebo&quot;),
    c(&quot;Relief&quot;, &quot;No Relief&quot;)
  )
)
clinical_data</code></pre>
<pre><code>         Relief No Relief
New Drug      8         2
Placebo       3         7</code></pre>
<p>We can use Fisher’s Exact Test to test “What is the probability of
seeing a pattern this strong (or stronger) if the drug had no
effect?”</p>
<p>Imagine there are 11 “Relief” cards and 9 “No Relief” cards and you
randomly deal 10 cards to “New Drug” group and 10 to “Placebo” group.
Fisher’s test asks: “If I shuffled these cards randomly many times, how
often would I see a split as extreme as 8-2 vs 3-7?”</p>
<pre class="r"><code>res &lt;- fisher.test(clinical_data)
res</code></pre>
<pre><code>
    Fisher&#39;s Exact Test for Count Data

data:  clinical_data
p-value = 0.06978
alternative hypothesis: true odds ratio is not equal to 1
95 percent confidence interval:
   0.8821175 127.0558418
sample estimates:
odds ratio 
  8.153063 </code></pre>
<p>If the drug had no effect, there’s only about a 0.06978% chance you’d
see a pattern this extreme or more extreme by random chance.</p>
</div>
<div id="exact-test" class="section level2">
<h2>Exact test</h2>
<p>Unlike other statistical tests that use approximations, Fisher’s
Exact test:</p>
<ul>
<li>Counts every single possible outcome</li>
<li>Calculates precise probabilities for each</li>
<li>No approximations</li>
</ul>
</div>
<div id="comparing-with-chi-square-test" class="section level2">
<h2>Comparing with Chi-Square Test</h2>
<p>Let’s see what happens if we use chi-square test (an approximation)
on our small dataset:</p>
<pre class="r"><code>chisq_result &lt;- chisq.test(clinical_data)</code></pre>
<pre><code>Warning in chisq.test(clinical_data): Chi-squared approximation may be
incorrect</code></pre>
<pre class="r"><code>chisq_result</code></pre>
<pre><code>
    Pearson&#39;s Chi-squared test with Yates&#39; continuity correction

data:  clinical_data
X-squared = 3.2323, df = 1, p-value = 0.0722</code></pre>
</div>
<div id="global-test" class="section level2">
<h2>Global test</h2>
<p>If you have more than two categories, Fisher’s Exact Test becomes a
generalised Fisher’s Exact Test (also known as Fisher-Freeman-Halton
test), because the regular Fisher’s test is for 2x2 tables.</p>
<p>As noted above, for larger tables people often use:</p>
<ul>
<li>Chi-squared test (if counts are not too small)</li>
<li>Generalised Fisher’s Exact Test (exact test and have a small
sample)</li>
</ul>
<p>Create cell cluster table.</p>
<pre class="r"><code>treated   &lt;- c(50, 60, 40, 30, 80, 90, 20, 15, 70, 45)
untreated &lt;- c(55, 50, 35, 25, 85, 95, 25, 20, 65, 40)

cluster_table &lt;- rbind(treated, untreated)
colnames(cluster_table) &lt;- paste0(&quot;cluster&quot;, 0:9)
rownames(cluster_table) &lt;- c(&quot;Treated&quot;, &quot;Untreated&quot;)
cluster_table</code></pre>
<pre><code>          cluster0 cluster1 cluster2 cluster3 cluster4 cluster5 cluster6
Treated         50       60       40       30       80       90       20
Untreated       55       50       35       25       85       95       25
          cluster7 cluster8 cluster9
Treated         15       70       45
Untreated       20       65       40</code></pre>
<p>Chi-squared Test.</p>
<pre class="r"><code>chisq_test &lt;- chisq.test(cluster_table)
chisq_test</code></pre>
<pre><code>
    Pearson&#39;s Chi-squared test

data:  cluster_table
X-squared = 3.9458, df = 9, p-value = 0.9149</code></pre>
<p>Generalised Fisher’s Exact Test.</p>
<ul>
<li><code>workspace</code> - an integer specifying the size of the
workspace used in the network algorithm. In units of 4 bytes. Only used
for non-simulated p-values larger than 2×22×2 tables. Since R version
3.5.0, this also increases the internal stack size which allows larger
problems to be solved, however sometimes needing hours. In such cases,
simulate.p.values=TRUE may be more reasonable.</li>
</ul>
<pre class="r"><code>fisher_test &lt;- fisher.test(cluster_table, workspace=2e8)
fisher_test</code></pre>
<pre><code>
    Fisher&#39;s Exact Test for Count Data

data:  cluster_table
p-value = 0.917
alternative hypothesis: two.sided</code></pre>
<p>Another cluster table.</p>
<pre class="r"><code>treated2   &lt;- c(200, 60, 40, 30, 80, 90, 20, 15, 70, 45)
untreated2 &lt;- c(55, 50, 35, 30, 85, 95, 25, 20, 75, 40)

cluster_table2 &lt;- rbind(treated2, untreated2)
colnames(cluster_table2) &lt;- paste0(&quot;cluster&quot;, 0:9)
rownames(cluster_table2) &lt;- c(&quot;Treated&quot;, &quot;Untreated&quot;)
cluster_table2</code></pre>
<pre><code>          cluster0 cluster1 cluster2 cluster3 cluster4 cluster5 cluster6
Treated        200       60       40       30       80       90       20
Untreated       55       50       35       30       85       95       25
          cluster7 cluster8 cluster9
Treated         15       70       45
Untreated       20       75       40</code></pre>
<p>Chi-squared Test; Chi-squared is valid when expected counts are
reasonably large (&gt; 5 cells per condition per cluster).</p>
<pre class="r"><code>chisq_test2 &lt;- chisq.test(cluster_table2)
chisq_test2</code></pre>
<pre><code>
    Pearson&#39;s Chi-squared test

data:  cluster_table2
X-squared = 69.837, df = 9, p-value = 1.639e-11</code></pre>
<p>Generalised Fisher’s Exact Test.</p>
<ul>
<li><code>simulate.p.value</code> - a logical indicating whether to
compute p-values by Monte Carlo simulation, in larger than 2×22×2
tables.</li>
</ul>
<pre class="r"><code>fisher_test2 &lt;- fisher.test(cluster_table2, simulate.p.value = TRUE)
fisher_test2</code></pre>
<pre><code>
    Fisher&#39;s Exact Test for Count Data with simulated p-value (based on
    2000 replicates)

data:  cluster_table2
p-value = 0.0004998
alternative hypothesis: two.sided</code></pre>
<p>Notes:</p>
<ul>
<li>Fisher’s and Chi-squared tests operate on contingency tables of
counts, not proportions.</li>
<li>They intrinsically consider the marginal totals (total treated
cells, total untreated cells, total cells in each cluster) when
computing expected frequencies.</li>
<li>If you use percentages or normalised counts, you actually distort
the expected distribution.</li>
</ul>
<pre class="r"><code>treated3   &lt;- c(50, 60, 40, 30, 80, 90, 20, 15, 70, 45)
untreated3 &lt;- untreated*3

cluster_table3 &lt;- rbind(treated3, untreated3)
colnames(cluster_table3) &lt;- paste0(&quot;cluster&quot;, 0:9)
rownames(cluster_table3) &lt;- c(&quot;Treated&quot;, &quot;Untreated&quot;)

chisq.test(cluster_table3)</code></pre>
<pre><code>
    Pearson&#39;s Chi-squared test

data:  cluster_table3
X-squared = 5.8925, df = 9, p-value = 0.7506</code></pre>
</div>
<div id="per-cluster-fishers-exact-test" class="section level2">
<h2>Per-cluster Fisher’s Exact Test</h2>
<p>Perform one Fisher’s Exact Test per cluster.</p>
<pre class="r"><code>treated   &lt;- c(200, 60, 40, 30, 80, 90, 20, 15, 70, 45)
untreated &lt;- c(55, 50, 35, 30, 85, 95, 25, 20, 75, 40)
clusters &lt;- paste0(&quot;cluster&quot;, 0:9)

total_treated &lt;- sum(treated)
total_untreated &lt;- sum(untreated)

res &lt;- purrr::map(seq_along(clusters), \(i){
  contingency_table &lt;- matrix(
    c(
      treated[i], total_treated - treated[i],
      untreated[i], total_untreated - untreated[i]
    ),
    nrow=2,
    byrow=TRUE
  )
  
  fisher_res &lt;- fisher.test(contingency_table)
  list(
    table = contingency_table,
    stat = fisher_res
  )
})

res[[1]]</code></pre>
<pre><code>$table
     [,1] [,2]
[1,]  200  450
[2,]   55  455

$stat

    Fisher&#39;s Exact Test for Count Data

data:  contingency_table
p-value &lt; 2.2e-16
alternative hypothesis: true odds ratio is not equal to 1
95 percent confidence interval:
 2.631973 5.188538
sample estimates:
odds ratio 
  3.672694 </code></pre>
<p>As a table with multiple testing correction.</p>
<pre class="r"><code>data.frame(
  Cluster = clusters,
  Treated = treated,
  Untreated = untreated,
  pvalue = purrr::map_dbl(res, \(x) x$stat$p.value),
  odds_ratio = purrr::map_dbl(res, \(x) x$stat$estimate)
) -&gt; res_df

res_df$padj &lt;- p.adjust(res_df$pvalue, method = &quot;BH&quot;)
res_df$log2_odds_ratio &lt;- log2(res_df$odds_ratio)

res_df</code></pre>
<pre><code>    Cluster Treated Untreated       pvalue odds_ratio         padj
1  cluster0     200        55 7.654036e-17  3.6726937 7.654036e-16
2  cluster1      60        50 7.625829e-01  0.9356553 7.625829e-01
3  cluster2      40        35 6.324680e-01  0.8900209 7.027423e-01
4  cluster3      30        30 3.519013e-01  0.7743725 5.027161e-01
5  cluster4      80        85 4.189083e-02  0.7019839 1.226881e-01
6  cluster5      90        95 2.924272e-02  0.7022974 1.226881e-01
7  cluster6      20        25 1.259970e-01  0.6161451 2.099949e-01
8  cluster7      15        20 1.215913e-01  0.5790270 2.099949e-01
9  cluster8      70        75 4.907522e-02  0.7002301 1.226881e-01
10 cluster9      45        40 5.716169e-01  0.8740708 7.027423e-01
   log2_odds_ratio
1       1.87683857
2      -0.09595098
3      -0.16808884
4      -0.36890045
5      -0.51049016
6      -0.50984593
7      -0.69865806
8      -0.78829757
9      -0.51409895
10     -0.19417790</code></pre>
<p>Volcano plot.</p>
<pre class="r"><code>library(ggplot2)

alpha &lt;- 0.05
res_df$Significant &lt;- res_df$padj &lt; alpha

ggplot(res_df, aes(x=log2_odds_ratio, y=-log10(pvalue))) +
  geom_point(aes(colour=Significant), size=3) +
  geom_text(
    data=res_df[res_df$Significant,],
    aes(label=Cluster),
    nudge_y=1,
    size=4
  ) +
  geom_hline(yintercept = -log10(0.05), linetype=&quot;dashed&quot;, colour=&quot;red&quot;) +
  scale_colour_manual(values=c(&quot;grey&quot;, &quot;firebrick&quot;)) +
  xlab(&quot;Log2 Odds Ratio&quot;) + 
  ylab(&quot;-log10(p-value)&quot;) +
  ggtitle(&quot;Volcano Plot of Cluster Proportions (Treated vs Untreated)&quot;) +
  theme_minimal()</code></pre>
<p><img src="figure/fisher.Rmd/volcano_plot-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-volcano_plot-1">
Past versions of volcano_plot-1.png
</button>
</p>
<div id="fig-volcano_plot-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/davetang/muse/blob/3a243513d90158ec87a1287135dae6bb1cff0741/docs/figure/fisher.Rmd/volcano_plot-1.png" target="_blank">3a24351</a>
</td>
<td>
Dave Tang
</td>
<td>
2025-03-29
</td>
</tr>
<tr>
<td>
<a href="https://github.com/davetang/muse/blob/4bccd4022ab4588c013a279efa1b8b2a02b1f6ba/docs/figure/fisher.Rmd/volcano_plot-1.png" target="_blank">4bccd40</a>
</td>
<td>
Dave Tang
</td>
<td>
2025-03-29
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>Plot Cluster Proportions.</p>
<pre class="r"><code>prop_df &lt;- data.frame(
  Cluster = rep(clusters, 2),
  Condition = rep(c(&quot;Treated&quot;, &quot;Untreated&quot;), each=length(clusters)),
  Proportion = c(treated / total_treated, untreated / total_untreated)
)

ggplot(prop_df, aes(x=Cluster, y=Proportion, fill=Condition)) +
  geom_bar(stat=&quot;identity&quot;, position=&quot;dodge&quot;) +
  ylab(&quot;Proportion of Cells&quot;) +
  theme_minimal() +
  theme(axis.title.x = element_blank())</code></pre>
<p><img src="figure/fisher.Rmd/cluster_proportions-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-cluster_proportions-1">
Past versions of cluster_proportions-1.png
</button>
</p>
<div id="fig-cluster_proportions-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/davetang/muse/blob/4bccd4022ab4588c013a279efa1b8b2a02b1f6ba/docs/figure/fisher.Rmd/cluster_proportions-1.png" target="_blank">4bccd40</a>
</td>
<td>
Dave Tang
</td>
<td>
2025-03-29
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>If using Seurat, just use
<code>table(Idents(seurat_obj), seurat_obj$condition)</code> to create
the contingency table and carry out the steps as per above.</p>
</div>
<div id="notes" class="section level2">
<h2>Notes</h2>
<p>The <strong>odds ratio</strong> compares the odds of finding a cell
in a specific cluster <strong>in treated cells</strong> versus
<strong>in untreated cells</strong>.</p>
<p>For example:</p>
<table>
<thead>
<tr class="header">
<th align="left"></th>
<th align="left">Cluster A</th>
<th align="left">Not Cluster A</th>
<th align="left">Total</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">Treated</td>
<td align="left">200</td>
<td align="left">450</td>
<td align="left">650</td>
</tr>
<tr class="even">
<td align="left">Untreated</td>
<td align="left">55</td>
<td align="left">455</td>
<td align="left">510</td>
</tr>
</tbody>
</table>
<p>The <strong>odds</strong> of being in Cluster A:</p>
<ul>
<li>For treated cells:</li>
</ul>
<p><span class="math display">\[ {Odds}_{\text{treated}} =
\frac{200}{450} = 0.4444 \]</span></p>
<ul>
<li>For untreated cells:</li>
</ul>
<p><span class="math display">\[ \text{Odds}_{\text{untreated}} =
\frac{55}{455} = 0.1209 \]</span></p>
<p>So the <strong>odds ratio</strong> is:</p>
<p><span class="math display">\[ \text{OR} = \frac{0.4444}{0.1209}
\approx 3.6758 \]</span></p>
<pre class="r"><code>res_df[1, ]</code></pre>
<pre><code>   Cluster Treated Untreated       pvalue odds_ratio         padj
1 cluster0     200        55 7.654036e-17   3.672694 7.654036e-16
  log2_odds_ratio Significant
1        1.876839        TRUE</code></pre>
<table>
<thead>
<tr class="header">
<th align="left">Odds Ratio</th>
<th align="left">Interpretation</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">OR = 1</td>
<td align="left">No difference between treated and untreated</td>
</tr>
<tr class="even">
<td align="left">OR &gt; 1</td>
<td align="left">Treated cells are <strong>enriched</strong> in this
cluster</td>
</tr>
<tr class="odd">
<td align="left">OR &lt; 1</td>
<td align="left">Treated cells are <strong>depleted</strong> in this
cluster</td>
</tr>
</tbody>
</table>
<p>In the example above, OR ~ 3.67 means treated cells are ~3.67 times
more likely to be in Cluster 0 than untreated cells.</p>
<p>Taking the <code>log2</code> of the odds ratio is common because:</p>
<ol style="list-style-type: decimal">
<li>It makes the scale symmetric:
<ul>
<li><code>log2(OR = 2)</code> = <code>1</code><br />
</li>
<li><code>log2(OR = 0.5)</code> = <code>-1</code></li>
</ul></li>
<li>It helps with plotting (like in volcano plots) where you want to
easily spot:
<ul>
<li>Positive values = enrichment</li>
<li>Negative values = depletion</li>
<li>Zero = no change</li>
</ul></li>
</ol>
<ul>
<li><code>log2(OR) = 1</code> means cells are <strong>2x more
likely</strong> to be in that cluster if treated.</li>
<li><code>log2(OR) = -1</code> means cells are <strong>2x less
likely</strong> (i.e., enriched in untreated).</li>
<li><code>log2(OR) = 0</code> means no difference between treated and
untreated.</li>
</ul>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span>
Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 4.5.0 (2025-04-11)
Platform: x86_64-pc-linux-gnu
Running under: Ubuntu 24.04.3 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 
LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.26.so;  LAPACK version 3.12.0

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

time zone: Etc/UTC
tzcode source: system (glibc)

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] lubridate_1.9.4 forcats_1.0.0   stringr_1.5.1   dplyr_1.1.4    
 [5] purrr_1.0.4     readr_2.1.5     tidyr_1.3.1     tibble_3.3.0   
 [9] ggplot2_3.5.2   tidyverse_2.0.0 workflowr_1.7.1

loaded via a namespace (and not attached):
 [1] sass_0.4.10        generics_0.1.4     stringi_1.8.7      hms_1.1.3         
 [5] digest_0.6.37      magrittr_2.0.3     timechange_0.3.0   evaluate_1.0.3    
 [9] grid_4.5.0         RColorBrewer_1.1-3 fastmap_1.2.0      rprojroot_2.0.4   
[13] jsonlite_2.0.0     processx_3.8.6     whisker_0.4.1      ps_1.9.1          
[17] promises_1.3.3     httr_1.4.7         scales_1.4.0       jquerylib_0.1.4   
[21] cli_3.6.5          rlang_1.1.6        withr_3.0.2        cachem_1.1.0      
[25] yaml_2.3.10        tools_4.5.0        tzdb_0.5.0         httpuv_1.6.16     
[29] vctrs_0.6.5        R6_2.6.1           lifecycle_1.0.4    git2r_0.36.2      
[33] fs_1.6.6           pkgconfig_2.0.3    callr_3.7.6        pillar_1.10.2     
[37] bslib_0.9.0        later_1.4.2        gtable_0.3.6       glue_1.8.0        
[41] Rcpp_1.0.14        xfun_0.52          tidyselect_1.2.1   rstudioapi_0.17.1 
[45] knitr_1.50         farver_2.1.2       htmltools_0.5.8.1  labeling_0.4.3    
[49] rmarkdown_2.29     compiler_4.5.0     getPass_0.2-4     </code></pre>
</div>
</div>
</div>
</div>
</div>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
https://docs.mathjax.org/en/latest/web/configuration.html. This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>




</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
