<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />



<meta name="date" content="2026-02-04" />

<title>Fitting</title>

<script src="site_libs/header-attrs-2.29/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.13.2/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-6.5.2/css/all.min.css" rel="stylesheet" />
<link href="site_libs/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet" />

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/main/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">My blog</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/davetang/muse">
    <span class="fa fa-github"></span>
     
    Source code
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Fitting</h1>
<h4 class="date">2026-02-04</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span>
workflowr <span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2026-02-04
</p>
<p>
<strong>Checks:</strong> <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 7
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong> <code>muse/</code> <span
class="glyphicon glyphicon-question-sign" aria-hidden="true"
title="This is the local directory in which the code in this file was executed.">
</span>
</p>
<p>
This reproducible <a href="https://rmarkdown.rstudio.com">R Markdown</a>
analysis was created with <a
  href="https://github.com/workflowr/workflowr">workflowr</a> (version
1.7.1). The <em>Checks</em> tab describes the reproducibility checks
that were applied when the results were created. The <em>Past
versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date
</a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git
repository, you know the exact version of the code that produced these
results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the
global environment can affect the analysis in your R Markdown file in
unknown ways. For reproduciblity it’s best to always run the code in an
empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20200712code">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Seed:</strong>
<code>set.seed(20200712)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20200712code"
class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20200712)</code> was run prior to running
the code in the R Markdown file. Setting a seed ensures that any results
that rely on randomness, e.g. subsampling or permutations, are
reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Session information:</strong>
recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package
versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be
confident that you successfully produced the results during this
run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr
project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomdavetangmusetree2a88aefae8954a5db638a076c9ee26e3ea2aaffetargetblank2a88aefa">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Repository version:</strong>
<a href="https://github.com/davetang/muse/tree/2a88aefae8954a5db638a076c9ee26e3ea2aaffe" target="_blank">2a88aef</a>
</a>
</p>
</div>
<div
id="strongRepositoryversionstrongahrefhttpsgithubcomdavetangmusetree2a88aefae8954a5db638a076c9ee26e3ea2aaffetargetblank2a88aefa"
class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development
and connecting the code version to the results is critical for
reproducibility.
</p>
<p>
The results in this page were generated with repository version
<a href="https://github.com/davetang/muse/tree/2a88aefae8954a5db638a076c9ee26e3ea2aaffe" target="_blank">2a88aef</a>.
See the <em>Past versions</em> tab to see a history of the changes made
to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for
the analysis have been committed to Git prior to generating the results
(you can use <code>wflow_publish</code> or
<code>wflow_git_commit</code>). workflowr only checks the R Markdown
file, but you know if there are other scripts or data files that it
depends on. Below is the status of the Git repository when the results
were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .Rproj.user/
    Ignored:    data/1M_neurons_filtered_gene_bc_matrices_h5.h5
    Ignored:    data/293t/
    Ignored:    data/293t_3t3_filtered_gene_bc_matrices.tar.gz
    Ignored:    data/293t_filtered_gene_bc_matrices.tar.gz
    Ignored:    data/5k_Human_Donor1_PBMC_3p_gem-x_5k_Human_Donor1_PBMC_3p_gem-x_count_sample_filtered_feature_bc_matrix.h5
    Ignored:    data/5k_Human_Donor2_PBMC_3p_gem-x_5k_Human_Donor2_PBMC_3p_gem-x_count_sample_filtered_feature_bc_matrix.h5
    Ignored:    data/5k_Human_Donor3_PBMC_3p_gem-x_5k_Human_Donor3_PBMC_3p_gem-x_count_sample_filtered_feature_bc_matrix.h5
    Ignored:    data/5k_Human_Donor4_PBMC_3p_gem-x_5k_Human_Donor4_PBMC_3p_gem-x_count_sample_filtered_feature_bc_matrix.h5
    Ignored:    data/97516b79-8d08-46a6-b329-5d0a25b0be98.h5ad
    Ignored:    data/Parent_SC3v3_Human_Glioblastoma_filtered_feature_bc_matrix.tar.gz
    Ignored:    data/brain_counts/
    Ignored:    data/cl.obo
    Ignored:    data/cl.owl
    Ignored:    data/jurkat/
    Ignored:    data/jurkat:293t_50:50_filtered_gene_bc_matrices.tar.gz
    Ignored:    data/jurkat_293t/
    Ignored:    data/jurkat_filtered_gene_bc_matrices.tar.gz
    Ignored:    data/pbmc20k/
    Ignored:    data/pbmc20k_seurat/
    Ignored:    data/pbmc3k.csv
    Ignored:    data/pbmc3k.csv.gz
    Ignored:    data/pbmc3k.h5ad
    Ignored:    data/pbmc3k/
    Ignored:    data/pbmc3k_bpcells_mat/
    Ignored:    data/pbmc3k_export.mtx
    Ignored:    data/pbmc3k_matrix.mtx
    Ignored:    data/pbmc3k_seurat.rds
    Ignored:    data/pbmc4k_filtered_gene_bc_matrices.tar.gz
    Ignored:    data/pbmc_1k_v3_filtered_feature_bc_matrix.h5
    Ignored:    data/pbmc_1k_v3_raw_feature_bc_matrix.h5
    Ignored:    data/refdata-gex-GRCh38-2020-A.tar.gz
    Ignored:    data/seurat_1m_neuron.rds
    Ignored:    data/t_3k_filtered_gene_bc_matrices.tar.gz
    Ignored:    r_packages_4.4.1/
    Ignored:    r_packages_4.5.0/

Untracked files:
    Untracked:  .claude/
    Untracked:  CLAUDE.md
    Untracked:  analysis/bioc.Rmd
    Untracked:  analysis/bioc_scrnaseq.Rmd
    Untracked:  analysis/chick_weight.Rmd
    Untracked:  analysis/likelihood.Rmd
    Untracked:  bpcells_matrix/
    Untracked:  data/Caenorhabditis_elegans.WBcel235.113.gtf.gz
    Untracked:  data/GCF_043380555.1-RS_2024_12_gene_ontology.gaf.gz
    Untracked:  data/SeuratObj.rds
    Untracked:  data/arab.rds
    Untracked:  data/astronomicalunit.csv
    Untracked:  data/femaleMiceWeights.csv
    Untracked:  data/lung_bcell.rds
    Untracked:  m3/
    Untracked:  women.json

Unstaged changes:
    Modified:   analysis/isoform_switch_analyzer.Rmd
    Modified:   analysis/linear_models.Rmd

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not
included in this status report because it is ok for generated content to
have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">
<p>
These are the previous versions of the repository in which changes were
made to the R Markdown (<code>analysis/fitting.Rmd</code>) and HTML
(<code>docs/fitting.html</code>) files. If you’ve configured a remote
Git repository (see <code>?wflow_git_remote</code>), click on the
hyperlinks in the table below to view the files as they were in that
past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/davetang/muse/blob/2a88aefae8954a5db638a076c9ee26e3ea2aaffe/analysis/fitting.Rmd" target="_blank">2a88aef</a>
</td>
<td>
Dave Tang
</td>
<td>
2026-02-04
</td>
<td>
Add explanations on fitting
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/davetang/muse/b2fed23ececfc6ec376800e6b6515b2ad5fe2249/docs/fitting.html" target="_blank">b2fed23</a>
</td>
<td>
Dave Tang
</td>
<td>
2025-09-02
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/davetang/muse/blob/15473bae75b26188311b56273753be5455f6ca03/analysis/fitting.Rmd" target="_blank">15473ba</a>
</td>
<td>
Dave Tang
</td>
<td>
2025-09-02
</td>
<td>
Negative binomial
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/davetang/muse/161c3d65b5036190f7351e0cc92564fd7ee2baae/docs/fitting.html" target="_blank">161c3d6</a>
</td>
<td>
Dave Tang
</td>
<td>
2025-09-02
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/davetang/muse/blob/31607923166a9b5d0f958149f3e6a46dd40ef0c3/analysis/fitting.Rmd" target="_blank">3160792</a>
</td>
<td>
Dave Tang
</td>
<td>
2025-09-02
</td>
<td>
Gamma distribution
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/davetang/muse/779b385c4c7734fc7a003aa5855b614d9b552dfb/docs/fitting.html" target="_blank">779b385</a>
</td>
<td>
Dave Tang
</td>
<td>
2025-09-02
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/davetang/muse/blob/5a74914ff0bf97b2d7ba3e3ba5418f92d0be15ea/analysis/fitting.Rmd" target="_blank">5a74914</a>
</td>
<td>
Dave Tang
</td>
<td>
2025-09-02
</td>
<td>
Fitting a distribution
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<div id="introduction" class="section level1">
<h1>Introduction</h1>
<p>When we say a “distribution was fit to” data, it means we assume the
data comes from some probability distribution and we estimate the
distribution’s parameters so the model best represents the observed
values.</p>
<div id="why-fit-distributions" class="section level2">
<h2>Why Fit Distributions?</h2>
<p>Distribution fitting is useful for:</p>
<ul>
<li><strong>Understanding data</strong>: Identifying the underlying
process that generated your data</li>
<li><strong>Simulation</strong>: Generating synthetic data with similar
properties</li>
<li><strong>Prediction</strong>: Calculating probabilities of future
events</li>
<li><strong>Statistical testing</strong>: Many tests assume specific
distributions</li>
<li><strong>Quality control</strong>: Monitoring processes in
manufacturing and other fields</li>
</ul>
</div>
<div id="maximum-likelihood-estimation" class="section level2">
<h2>Maximum Likelihood Estimation</h2>
<p>The most common method for fitting distributions is <strong>Maximum
Likelihood Estimation (MLE)</strong>. The idea is to find parameter
values that make the observed data most probable.</p>
<p>For example, if we have data and assume it comes from a normal
distribution, MLE finds the mean (<span
class="math inline">\(\mu\)</span>) and standard deviation (<span
class="math inline">\(\sigma\)</span>) that maximise the likelihood of
observing that data.</p>
<p>The likelihood function for n independent observations is:</p>
<p><span class="math display">\[L(\theta | x_1, ..., x_n) =
\prod_{i=1}^{n} f(x_i | \theta)\]</span></p>
<p>where <span class="math inline">\(f\)</span> is the probability
density function and <span class="math inline">\(\theta\)</span>
represents the parameters. In practice, we maximise the log-likelihood
for computational convenience.</p>
</div>
</div>
<div id="fitting-distributions-with-mass" class="section level1">
<h1>Fitting Distributions with MASS</h1>
<p>The <code>MASS::fitdistr()</code> function fits univariate
distributions using maximum likelihood estimation. Let’s work through
several examples.</p>
<div id="example-1-normal-distribution" class="section level2">
<h2>Example 1: Normal Distribution</h2>
<p>The normal (Gaussian) distribution is characterised by two
parameters: mean (<span class="math inline">\(\mu\)</span>) and standard
deviation (<span class="math inline">\(\sigma\)</span>).</p>
<pre class="r"><code>set.seed(1984)
eg1 &lt;- rnorm(100, mean = 5, sd = 2)</code></pre>
<p>Fit a normal distribution to the data:</p>
<pre class="r"><code>eg1_fit &lt;- fitdistr(eg1, &quot;normal&quot;)
eg1_fit</code></pre>
<pre><code>     mean        sd   
  5.062545   1.873380 
 (0.187338) (0.132468)</code></pre>
<div id="understanding-the-output" class="section level3">
<h3>Understanding the Output</h3>
<p>The output shows:</p>
<ul>
<li><strong>Estimated parameters</strong>: mean = 5.063, sd = 1.873</li>
<li><strong>Standard errors</strong>: The values in parentheses indicate
uncertainty in the estimates</li>
</ul>
<p>The true parameters were mean = 5 and sd = 2. Our estimates are
close, demonstrating that MLE recovers the true parameters reasonably
well with 100 observations.</p>
</div>
<div id="visualising-the-fit" class="section level3">
<h3>Visualising the Fit</h3>
<p>It’s important to visually verify that the fitted distribution
matches the data:</p>
<pre class="r"><code>ggplot(data.frame(x = eg1), aes(x)) +
  geom_histogram(aes(y = after_stat(density)), bins = 20,
                 fill = &quot;steelblue&quot;, colour = &quot;white&quot;, alpha = 0.7) +
  stat_function(fun = dnorm,
                args = list(mean = eg1_fit$estimate[&#39;mean&#39;],
                            sd = eg1_fit$estimate[&#39;sd&#39;]),
                colour = &quot;red&quot;, linewidth = 1) +
  labs(title = &quot;Normal Distribution Fit&quot;,
       subtitle = paste0(&quot;Estimated: mean = &quot;, round(eg1_fit$estimate[&#39;mean&#39;], 2),
                        &quot;, sd = &quot;, round(eg1_fit$estimate[&#39;sd&#39;], 2)),
       x = &quot;Value&quot;, y = &quot;Density&quot;) +
  theme_minimal()</code></pre>
<p><img src="figure/fitting.Rmd/eg1_plot-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>The red curve shows the fitted normal distribution overlaid on the
histogram of our data.</p>
</div>
</div>
<div id="example-2-poisson-distribution" class="section level2">
<h2>Example 2: Poisson Distribution</h2>
<p>The Poisson distribution models count data and has a single parameter
<span class="math inline">\(\lambda\)</span> (lambda), which represents
both the mean and variance.</p>
<pre class="r"><code>set.seed(1984)
eg2 &lt;- rpois(100, lambda = 3)
eg2_fit &lt;- fitdistr(eg2, &quot;Poisson&quot;)
eg2_fit</code></pre>
<pre><code>    lambda  
  3.1000000 
 (0.1760682)</code></pre>
<p>The estimated <span class="math inline">\(\lambda\)</span> = 3.1 is
close to the true value of 3.</p>
<pre class="r"><code># Create comparison data
x_vals &lt;- 0:max(eg2)
observed &lt;- table(factor(eg2, levels = x_vals)) / length(eg2)
expected &lt;- dpois(x_vals, lambda = eg2_fit$estimate[&#39;lambda&#39;])

plot_data &lt;- data.frame(
  x = rep(x_vals, 2),
  probability = c(as.numeric(observed), expected),
  type = rep(c(&quot;Observed&quot;, &quot;Fitted&quot;), each = length(x_vals))
)

ggplot(plot_data, aes(x, probability, fill = type)) +
  geom_bar(stat = &quot;identity&quot;, position = &quot;dodge&quot;, alpha = 0.7) +
  scale_fill_manual(values = c(&quot;Observed&quot; = &quot;steelblue&quot;, &quot;Fitted&quot; = &quot;red&quot;)) +
  labs(title = &quot;Poisson Distribution Fit&quot;,
       subtitle = paste0(&quot;Estimated: lambda = &quot;, round(eg2_fit$estimate[&#39;lambda&#39;], 2)),
       x = &quot;Count&quot;, y = &quot;Probability&quot;, fill = &quot;&quot;) +
  theme_minimal()</code></pre>
<p><img src="figure/fitting.Rmd/eg2_plot-1.png" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="example-3-gamma-distribution" class="section level2">
<h2>Example 3: Gamma Distribution</h2>
<p>The <a href="https://en.wikipedia.org/wiki/Gamma_distribution">gamma
distribution</a> is useful for modelling positive continuous data, such
as waiting times or rainfall amounts. It has two parameters: shape
(<span class="math inline">\(\alpha\)</span>) and rate (<span
class="math inline">\(\beta\)</span>).</p>
<pre class="r"><code>set.seed(1984)
eg3 &lt;- rgamma(300, shape = 2, rate = 3)

eg3_fit &lt;- fitdistr(eg3, &quot;gamma&quot;)</code></pre>
<pre><code>Warning in densfun(x, parm[1], parm[2], ...): NaNs produced</code></pre>
<pre class="r"><code>eg3_fit</code></pre>
<pre><code>     shape       rate   
  2.0425966   3.1910159 
 (0.1550892) (0.2744437)</code></pre>
<p>True parameters: shape = 2, rate = 3. The estimates are reasonably
close.</p>
<pre class="r"><code>ggplot(data.frame(x = eg3), aes(x)) +
  geom_histogram(aes(y = after_stat(density)), bins = 30,
                 fill = &quot;steelblue&quot;, colour = &quot;white&quot;, alpha = 0.7) +
  stat_function(fun = dgamma,
                args = list(shape = eg3_fit$estimate[&#39;shape&#39;],
                            rate = eg3_fit$estimate[&#39;rate&#39;]),
                colour = &quot;red&quot;, linewidth = 1) +
  labs(title = &quot;Gamma Distribution Fit&quot;,
       subtitle = paste0(&quot;Estimated: shape = &quot;, round(eg3_fit$estimate[&#39;shape&#39;], 2),
                        &quot;, rate = &quot;, round(eg3_fit$estimate[&#39;rate&#39;], 2)),
       x = &quot;Value&quot;, y = &quot;Density&quot;) +
  theme_minimal()</code></pre>
<p><img src="figure/fitting.Rmd/eg3_plot-1.png" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="example-4-negative-binomial-distribution"
class="section level2">
<h2>Example 4: Negative Binomial Distribution</h2>
<p>The negative binomial distribution is useful for overdispersed count
data (when variance exceeds the mean). It has two parameters: size
(<span class="math inline">\(r\)</span>) and mu (<span
class="math inline">\(\mu\)</span>).</p>
<pre class="r"><code>set.seed(1984)
eg4 &lt;- rnbinom(100, mu = 10, size = 5)

eg4_fit &lt;- fitdistr(eg4, &quot;negative binomial&quot;)
eg4_fit</code></pre>
<pre><code>     size         mu    
  4.9764745   9.7999979 
 (1.0492499) (0.5394329)</code></pre>
<pre class="r"><code>x_vals &lt;- 0:max(eg4)
observed &lt;- table(factor(eg4, levels = x_vals)) / length(eg4)
expected &lt;- dnbinom(x_vals, mu = eg4_fit$estimate[&#39;mu&#39;], size = eg4_fit$estimate[&#39;size&#39;])

plot_data &lt;- data.frame(
  x = rep(x_vals, 2),
  probability = c(as.numeric(observed), expected),
  type = rep(c(&quot;Observed&quot;, &quot;Fitted&quot;), each = length(x_vals))
)

ggplot(plot_data, aes(x, probability, fill = type)) +
  geom_bar(stat = &quot;identity&quot;, position = &quot;dodge&quot;, alpha = 0.7) +
  scale_fill_manual(values = c(&quot;Observed&quot; = &quot;steelblue&quot;, &quot;Fitted&quot; = &quot;red&quot;)) +
  labs(title = &quot;Negative Binomial Distribution Fit&quot;,
       subtitle = paste0(&quot;Estimated: size = &quot;, round(eg4_fit$estimate[&#39;size&#39;], 2),
                        &quot;, mu = &quot;, round(eg4_fit$estimate[&#39;mu&#39;], 2)),
       x = &quot;Count&quot;, y = &quot;Probability&quot;, fill = &quot;&quot;) +
  theme_minimal()</code></pre>
<p><img src="figure/fitting.Rmd/eg4_plot-1.png" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="supported-distributions" class="section level2">
<h2>Supported Distributions</h2>
<p><code>MASS::fitdistr()</code> supports the following
distributions:</p>
<ul>
<li>“beta”</li>
<li>“cauchy”</li>
<li>“chi-squared”</li>
<li>“exponential”</li>
<li>“gamma”</li>
<li>“geometric”</li>
<li>“log-normal”</li>
<li>“lognormal”</li>
<li>“logistic”</li>
<li>“negative binomial”</li>
<li>“normal”</li>
<li>“Poisson”</li>
<li>“t”</li>
<li>“weibull”</li>
</ul>
</div>
</div>
<div id="comparing-model-fits" class="section level1">
<h1>Comparing Model Fits</h1>
<p>When multiple distributions could plausibly fit your data, you need
methods to compare them.</p>
<div id="akaike-information-criterion-aic" class="section level2">
<h2>Akaike Information Criterion (AIC)</h2>
<p>The <a
href="https://en.wikipedia.org/wiki/Akaike_information_criterion">Akaike
Information Criterion</a> (AIC) measures how well a model fits the data
while penalising model complexity:</p>
<p><span class="math display">\[AIC = 2k - 2\ln(L)\]</span></p>
<p>where <span class="math inline">\(k\)</span> is the number of
parameters and <span class="math inline">\(L\)</span> is the maximum
likelihood. <strong>Lower AIC values indicate better fit</strong>. The
penalty term (2k) prevents overfitting by discouraging unnecessarily
complex models.</p>
<div id="example-normal-vs-exponential" class="section level3">
<h3>Example: Normal vs Exponential</h3>
<p>Let’s compare normal and exponential fits for our normally
distributed data:</p>
<pre class="r"><code>set.seed(1984)
eg1 &lt;- rnorm(100, mean = 5, sd = 2)

eg1_norm &lt;- fitdistr(eg1, &quot;normal&quot;)
eg1_exp &lt;- fitdistr(eg1, &quot;exponential&quot;)

AIC(eg1_norm, eg1_exp)</code></pre>
<pre><code>         df      AIC
eg1_norm  2 413.3365
eg1_exp   1 526.3739</code></pre>
<p>The normal distribution has a much lower AIC, correctly indicating
it’s a better fit. We can also visualise why:</p>
<pre class="r"><code>ggplot(data.frame(x = eg1), aes(x)) +
  geom_histogram(aes(y = after_stat(density)), bins = 20,
                 fill = &quot;grey70&quot;, colour = &quot;white&quot;) +
  stat_function(fun = dnorm,
                args = list(mean = eg1_norm$estimate[&#39;mean&#39;],
                            sd = eg1_norm$estimate[&#39;sd&#39;]),
                aes(colour = &quot;Normal&quot;), linewidth = 1) +
  stat_function(fun = dexp,
                args = list(rate = eg1_exp$estimate[&#39;rate&#39;]),
                aes(colour = &quot;Exponential&quot;), linewidth = 1) +
  scale_colour_manual(values = c(&quot;Normal&quot; = &quot;blue&quot;, &quot;Exponential&quot; = &quot;red&quot;)) +
  labs(title = &quot;Comparing Normal vs Exponential Fit&quot;,
       x = &quot;Value&quot;, y = &quot;Density&quot;, colour = &quot;Distribution&quot;) +
  theme_minimal()</code></pre>
<p><img src="figure/fitting.Rmd/aic_visual-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>The exponential distribution clearly doesn’t capture the symmetric,
bell-shaped nature of the data.</p>
</div>
</div>
<div id="bayesian-information-criterion-bic" class="section level2">
<h2>Bayesian Information Criterion (BIC)</h2>
<p>BIC is similar to AIC but applies a stronger penalty for model
complexity:</p>
<p><span class="math display">\[BIC = k\ln(n) - 2\ln(L)\]</span></p>
<p>where <span class="math inline">\(n\)</span> is the sample size. BIC
tends to favour simpler models than AIC.</p>
<pre class="r"><code>BIC(eg1_norm, eg1_exp)</code></pre>
<pre><code>         df      BIC
eg1_norm  2 418.5469
eg1_exp   1 528.9791</code></pre>
<p>Both criteria agree that the normal distribution is the better
fit.</p>
</div>
<div id="example-choosing-between-related-distributions"
class="section level2">
<h2>Example: Choosing Between Related Distributions</h2>
<p>Sometimes the choice is less obvious. Consider count data that might
follow either a Poisson or negative binomial distribution:</p>
<pre class="r"><code>set.seed(1984)
# Generate overdispersed count data
overdispersed &lt;- rnbinom(200, mu = 5, size = 2)

# Fit both distributions
pois_fit &lt;- fitdistr(overdispersed, &quot;Poisson&quot;)
nb_fit &lt;- fitdistr(overdispersed, &quot;negative binomial&quot;)

# Compare AIC
AIC(pois_fit, nb_fit)</code></pre>
<pre><code>         df      AIC
pois_fit  1 1371.432
nb_fit    2 1081.379</code></pre>
<pre class="r"><code>x_vals &lt;- 0:max(overdispersed)
observed &lt;- table(factor(overdispersed, levels = x_vals)) / length(overdispersed)
pois_expected &lt;- dpois(x_vals, lambda = pois_fit$estimate[&#39;lambda&#39;])
nb_expected &lt;- dnbinom(x_vals, mu = nb_fit$estimate[&#39;mu&#39;], size = nb_fit$estimate[&#39;size&#39;])

plot_data &lt;- data.frame(
  x = rep(x_vals, 3),
  probability = c(as.numeric(observed), pois_expected, nb_expected),
  type = rep(c(&quot;Observed&quot;, &quot;Poisson&quot;, &quot;Negative Binomial&quot;), each = length(x_vals))
)

ggplot(plot_data, aes(x, probability, fill = type)) +
  geom_bar(stat = &quot;identity&quot;, position = &quot;dodge&quot;, alpha = 0.7) +
  scale_fill_manual(values = c(&quot;Observed&quot; = &quot;grey50&quot;,
                               &quot;Poisson&quot; = &quot;steelblue&quot;,
                               &quot;Negative Binomial&quot; = &quot;darkred&quot;)) +
  labs(title = &quot;Poisson vs Negative Binomial Fit&quot;,
       subtitle = &quot;For overdispersed count data&quot;,
       x = &quot;Count&quot;, y = &quot;Probability&quot;, fill = &quot;&quot;) +
  theme_minimal()</code></pre>
<p><img src="figure/fitting.Rmd/count_comparison_visual-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>The negative binomial has lower AIC because it can accommodate the
overdispersion (variance &gt; mean) in the data, which the Poisson
cannot.</p>
</div>
</div>
<div id="assessing-goodness-of-fit" class="section level1">
<h1>Assessing Goodness of Fit</h1>
<p>Beyond comparing models, you should assess whether any distribution
fits well in absolute terms.</p>
<div id="q-q-plots" class="section level2">
<h2>Q-Q Plots</h2>
<p>Quantile-Quantile (Q-Q) plots compare the quantiles of your data
against theoretical quantiles. If the data follows the assumed
distribution, points should fall along the diagonal line.</p>
<pre class="r"><code># Q-Q plot for normal fit
eg1_sorted &lt;- sort(eg1)
theoretical_quantiles &lt;- qnorm(ppoints(length(eg1)),
                               mean = eg1_fit$estimate[&#39;mean&#39;],
                               sd = eg1_fit$estimate[&#39;sd&#39;])

ggplot(data.frame(theoretical = theoretical_quantiles, observed = eg1_sorted),
       aes(theoretical, observed)) +
  geom_point(alpha = 0.6) +
  geom_abline(slope = 1, intercept = 0, colour = &quot;red&quot;, linetype = &quot;dashed&quot;) +
  labs(title = &quot;Q-Q Plot: Normal Distribution&quot;,
       x = &quot;Theoretical Quantiles&quot;, y = &quot;Observed Quantiles&quot;) +
  theme_minimal()</code></pre>
<p><img src="figure/fitting.Rmd/qq_plot-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>Points closely following the diagonal line indicate a good fit.</p>
</div>
<div id="kolmogorov-smirnov-test" class="section level2">
<h2>Kolmogorov-Smirnov Test</h2>
<p>The Kolmogorov-Smirnov test formally tests whether data comes from a
specified distribution:</p>
<pre class="r"><code>ks.test(eg1, &quot;pnorm&quot;,
        mean = eg1_fit$estimate[&#39;mean&#39;],
        sd = eg1_fit$estimate[&#39;sd&#39;])</code></pre>
<pre><code>
    Asymptotic one-sample Kolmogorov-Smirnov test

data:  eg1
D = 0.086058, p-value = 0.4494
alternative hypothesis: two-sided</code></pre>
<p>A large p-value (&gt; 0.05) suggests we cannot reject the hypothesis
that the data comes from the fitted distribution.</p>
</div>
</div>
<div id="practical-considerations" class="section level1">
<h1>Practical Considerations</h1>
<div id="sample-size" class="section level2">
<h2>Sample Size</h2>
<p>Larger samples produce more reliable parameter estimates:</p>
<pre class="r"><code>set.seed(1984)
true_mean &lt;- 10
true_sd &lt;- 3

# Compare estimates at different sample sizes
sample_sizes &lt;- c(20, 50, 100, 500, 1000)
results &lt;- lapply(sample_sizes, function(n) {
  x &lt;- rnorm(n, mean = true_mean, sd = true_sd)
  fit &lt;- fitdistr(x, &quot;normal&quot;)
  data.frame(
    n = n,
    est_mean = fit$estimate[&#39;mean&#39;],
    est_sd = fit$estimate[&#39;sd&#39;],
    se_mean = fit$sd[&#39;mean&#39;],
    se_sd = fit$sd[&#39;sd&#39;]
  )
})

results_df &lt;- do.call(rbind, results)
results_df</code></pre>
<pre><code>         n  est_mean   est_sd    se_mean      se_sd
mean    20 10.382167 3.026534 0.67675348 0.47853697
mean1   50 10.226157 2.746529 0.38841779 0.27465285
mean2  100  9.604006 2.743902 0.27439024 0.19402320
mean3  500  9.804773 3.020142 0.13506485 0.09550527
mean4 1000  9.954271 3.011147 0.09522084 0.06733130</code></pre>
<p>Notice how standard errors decrease as sample size increases.</p>
</div>
<div id="starting-values" class="section level2">
<h2>Starting Values</h2>
<p>For some distributions, <code>fitdistr()</code> may need starting
values to converge:</p>
<pre class="r"><code>set.seed(1984)
weibull_data &lt;- rweibull(100, shape = 2, scale = 5)

# Provide starting values for Weibull
weibull_fit &lt;- fitdistr(weibull_data, &quot;weibull&quot;,
                        start = list(shape = 1, scale = 1))</code></pre>
<pre><code>Warning in densfun(x, parm[1], parm[2], ...): NaNs produced
Warning in densfun(x, parm[1], parm[2], ...): NaNs produced
Warning in densfun(x, parm[1], parm[2], ...): NaNs produced
Warning in densfun(x, parm[1], parm[2], ...): NaNs produced</code></pre>
<pre class="r"><code>weibull_fit</code></pre>
<pre><code>     shape       scale  
  1.8967375   4.8896060 
 (0.1458478) (0.2720805)</code></pre>
</div>
<div id="accessing-fit-components" class="section level2">
<h2>Accessing Fit Components</h2>
<p>The fitted object contains useful information:</p>
<pre class="r"><code># Parameter estimates
eg1_fit$estimate</code></pre>
<pre><code>    mean       sd 
5.062545 1.873380 </code></pre>
<pre class="r"><code># Standard errors
eg1_fit$sd</code></pre>
<pre><code>    mean       sd 
0.187338 0.132468 </code></pre>
<pre class="r"><code># Log-likelihood
eg1_fit$loglik</code></pre>
<pre><code>[1] -204.6683</code></pre>
<pre class="r"><code># Variance-covariance matrix
eg1_fit$vcov</code></pre>
<pre><code>           mean         sd
mean 0.03509552 0.00000000
sd   0.00000000 0.01754776</code></pre>
</div>
</div>
<div id="summary" class="section level1">
<h1>Summary</h1>
<p>Key points for distribution fitting:</p>
<ol style="list-style-type: decimal">
<li><strong>Maximum Likelihood Estimation</strong> finds parameters that
make observed data most probable</li>
<li><strong>Always visualise</strong> the fitted distribution against
your data</li>
<li><strong>Compare models</strong> using AIC or BIC (lower is
better)</li>
<li><strong>Assess goodness of fit</strong> with Q-Q plots or formal
tests</li>
<li><strong>Larger samples</strong> give more reliable estimates</li>
<li>Consider the <strong>theoretical basis</strong> for choosing a
distribution, not just statistical fit</li>
</ol>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span>
Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 4.5.0 (2025-04-11)
Platform: x86_64-pc-linux-gnu
Running under: Ubuntu 24.04.3 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 
LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.26.so;  LAPACK version 3.12.0

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

time zone: Etc/UTC
tzcode source: system (glibc)

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] ggplot2_3.5.2   MASS_7.3-65     workflowr_1.7.1

loaded via a namespace (and not attached):
 [1] gtable_0.3.6       jsonlite_2.0.0     dplyr_1.1.4        compiler_4.5.0    
 [5] promises_1.3.3     tidyselect_1.2.1   Rcpp_1.0.14        stringr_1.5.1     
 [9] git2r_0.36.2       callr_3.7.6        later_1.4.2        jquerylib_0.1.4   
[13] scales_1.4.0       yaml_2.3.10        fastmap_1.2.0      R6_2.6.1          
[17] labeling_0.4.3     generics_0.1.4     knitr_1.50         tibble_3.3.0      
[21] rprojroot_2.0.4    RColorBrewer_1.1-3 bslib_0.9.0        pillar_1.10.2     
[25] rlang_1.1.6        cachem_1.1.0       stringi_1.8.7      httpuv_1.6.16     
[29] xfun_0.52          getPass_0.2-4      fs_1.6.6           sass_0.4.10       
[33] cli_3.6.5          withr_3.0.2        magrittr_2.0.3     ps_1.9.1          
[37] digest_0.6.37      grid_4.5.0         processx_3.8.6     rstudioapi_0.17.1 
[41] lifecycle_1.0.4    vctrs_0.6.5        evaluate_1.0.3     glue_1.8.0        
[45] farver_2.1.2       whisker_0.4.1      rmarkdown_2.29     httr_1.4.7        
[49] tools_4.5.0        pkgconfig_2.0.3    htmltools_0.5.8.1 </code></pre>
</div>
</div>
</div>
</div>
</div>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
https://docs.mathjax.org/en/latest/web/configuration.html. This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>




</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
