<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />



<meta name="date" content="2025-12-30" />

<title>Understanding R’s Formula Syntax</title>

<script src="site_libs/header-attrs-2.29/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.13.2/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-6.5.2/css/all.min.css" rel="stylesheet" />
<link href="site_libs/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet" />

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/main/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">My blog</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/davetang/muse">
    <span class="fa fa-github"></span>
     
    Source code
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Understanding R’s Formula Syntax</h1>
<h4 class="date">2025-12-30</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span>
workflowr <span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2025-12-30
</p>
<p>
<strong>Checks:</strong> <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 7
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong> <code>muse/</code> <span
class="glyphicon glyphicon-question-sign" aria-hidden="true"
title="This is the local directory in which the code in this file was executed.">
</span>
</p>
<p>
This reproducible <a href="https://rmarkdown.rstudio.com">R Markdown</a>
analysis was created with <a
  href="https://github.com/workflowr/workflowr">workflowr</a> (version
1.7.1). The <em>Checks</em> tab describes the reproducibility checks
that were applied when the results were created. The <em>Past
versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date
</a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git
repository, you know the exact version of the code that produced these
results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the
global environment can affect the analysis in your R Markdown file in
unknown ways. For reproduciblity it’s best to always run the code in an
empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20200712code">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Seed:</strong>
<code>set.seed(20200712)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20200712code"
class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20200712)</code> was run prior to running
the code in the R Markdown file. Setting a seed ensures that any results
that rely on randomness, e.g. subsampling or permutations, are
reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Session information:</strong>
recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package
versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be
confident that you successfully produced the results during this
run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr
project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomdavetangmusetreebc7a77d7d7aaaf3dc2bf68579bf5fc1fbe788ce0targetblankbc7a77da">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Repository version:</strong>
<a href="https://github.com/davetang/muse/tree/bc7a77d7d7aaaf3dc2bf68579bf5fc1fbe788ce0" target="_blank">bc7a77d</a>
</a>
</p>
</div>
<div
id="strongRepositoryversionstrongahrefhttpsgithubcomdavetangmusetreebc7a77d7d7aaaf3dc2bf68579bf5fc1fbe788ce0targetblankbc7a77da"
class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development
and connecting the code version to the results is critical for
reproducibility.
</p>
<p>
The results in this page were generated with repository version
<a href="https://github.com/davetang/muse/tree/bc7a77d7d7aaaf3dc2bf68579bf5fc1fbe788ce0" target="_blank">bc7a77d</a>.
See the <em>Past versions</em> tab to see a history of the changes made
to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for
the analysis have been committed to Git prior to generating the results
(you can use <code>wflow_publish</code> or
<code>wflow_git_commit</code>). workflowr only checks the R Markdown
file, but you know if there are other scripts or data files that it
depends on. Below is the status of the Git repository when the results
were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .Rproj.user/
    Ignored:    data/1M_neurons_filtered_gene_bc_matrices_h5.h5
    Ignored:    data/293t/
    Ignored:    data/293t_3t3_filtered_gene_bc_matrices.tar.gz
    Ignored:    data/293t_filtered_gene_bc_matrices.tar.gz
    Ignored:    data/5k_Human_Donor1_PBMC_3p_gem-x_5k_Human_Donor1_PBMC_3p_gem-x_count_sample_filtered_feature_bc_matrix.h5
    Ignored:    data/5k_Human_Donor2_PBMC_3p_gem-x_5k_Human_Donor2_PBMC_3p_gem-x_count_sample_filtered_feature_bc_matrix.h5
    Ignored:    data/5k_Human_Donor3_PBMC_3p_gem-x_5k_Human_Donor3_PBMC_3p_gem-x_count_sample_filtered_feature_bc_matrix.h5
    Ignored:    data/5k_Human_Donor4_PBMC_3p_gem-x_5k_Human_Donor4_PBMC_3p_gem-x_count_sample_filtered_feature_bc_matrix.h5
    Ignored:    data/97516b79-8d08-46a6-b329-5d0a25b0be98.h5ad
    Ignored:    data/Parent_SC3v3_Human_Glioblastoma_filtered_feature_bc_matrix.tar.gz
    Ignored:    data/brain_counts/
    Ignored:    data/cl.obo
    Ignored:    data/cl.owl
    Ignored:    data/jurkat/
    Ignored:    data/jurkat:293t_50:50_filtered_gene_bc_matrices.tar.gz
    Ignored:    data/jurkat_293t/
    Ignored:    data/jurkat_filtered_gene_bc_matrices.tar.gz
    Ignored:    data/pbmc20k/
    Ignored:    data/pbmc20k_seurat/
    Ignored:    data/pbmc3k.h5ad
    Ignored:    data/pbmc3k/
    Ignored:    data/pbmc3k_bpcells_mat/
    Ignored:    data/pbmc3k_export.mtx
    Ignored:    data/pbmc3k_matrix.mtx
    Ignored:    data/pbmc3k_seurat.rds
    Ignored:    data/pbmc4k_filtered_gene_bc_matrices.tar.gz
    Ignored:    data/pbmc_1k_v3_filtered_feature_bc_matrix.h5
    Ignored:    data/pbmc_1k_v3_raw_feature_bc_matrix.h5
    Ignored:    data/refdata-gex-GRCh38-2020-A.tar.gz
    Ignored:    data/seurat_1m_neuron.rds
    Ignored:    data/t_3k_filtered_gene_bc_matrices.tar.gz
    Ignored:    r_packages_4.4.1/
    Ignored:    r_packages_4.5.0/

Untracked files:
    Untracked:  analysis/bioc.Rmd
    Untracked:  analysis/bioc_scrnaseq.Rmd
    Untracked:  analysis/likelihood.Rmd
    Untracked:  bpcells_matrix/
    Untracked:  data/Caenorhabditis_elegans.WBcel235.113.gtf.gz
    Untracked:  data/GCF_043380555.1-RS_2024_12_gene_ontology.gaf.gz
    Untracked:  data/arab.rds
    Untracked:  data/astronomicalunit.csv
    Untracked:  data/femaleMiceWeights.csv
    Untracked:  data/lung_bcell.rds
    Untracked:  m3/
    Untracked:  women.json

Unstaged changes:
    Modified:   analysis/isoform_switch_analyzer.Rmd

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not
included in this status report because it is ok for generated content to
have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">
<p>
These are the previous versions of the repository in which changes were
made to the R Markdown (<code>analysis/formula.Rmd</code>) and HTML
(<code>docs/formula.html</code>) files. If you’ve configured a remote
Git repository (see <code>?wflow_git_remote</code>), click on the
hyperlinks in the table below to view the files as they were in that
past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/davetang/muse/blob/bc7a77d7d7aaaf3dc2bf68579bf5fc1fbe788ce0/analysis/formula.Rmd" target="_blank">bc7a77d</a>
</td>
<td>
Dave Tang
</td>
<td>
2025-12-30
</td>
<td>
R formula syntax
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<div id="introduction-to-r-formulas" class="section level1">
<h1>Introduction to R Formulas</h1>
<p>R’s formula syntax is a special notation that describes
<strong>relationships between variables</strong>. It’s used extensively
in statistical modeling, data visualisation, and data manipulation.
While it may seem cryptic at first, it’s actually quite intuitive once
you understand the basic components.</p>
<div id="why-formulas-exist" class="section level2">
<h2>Why Formulas Exist</h2>
<p>Formulas provide a concise, readable way to specify:</p>
<ul>
<li>Which variable is the <strong>response</strong> (outcome) and which
are <strong>predictors</strong> (explanatory variables)</li>
<li>How variables interact with each other</li>
<li>Complex model structures without writing lengthy function
arguments</li>
</ul>
</div>
</div>
<div id="basic-formula-structure" class="section level1">
<h1>Basic Formula Structure</h1>
<p>The fundamental structure is:</p>
<pre><code>response ~ predictor</code></pre>
<p>The <code>~</code> (tilde) symbol is read as “is modeled by” or
“depends on”.</p>
<div id="simple-examples" class="section level2">
<h2>Simple Examples</h2>
<pre class="r"><code># Create sample data
set.seed(123)
df &lt;- data.frame(
  height = rnorm(100, 170, 10),
  weight = rnorm(100, 70, 15),
  age = sample(20:60, 100, replace = TRUE),
  gender = sample(c(&quot;M&quot;, &quot;F&quot;), 100, replace = TRUE)
)

# Example 1: Simple linear regression
# &quot;weight depends on height&quot;
model1 &lt;- lm(weight ~ height, data = df)
summary(model1)</code></pre>
<pre><code>
Call:
lm(formula = weight ~ height, data = df)

Residuals:
    Min      1Q  Median      3Q     Max 
-28.610 -10.253  -1.312   8.710  49.356 

Coefficients:
            Estimate Std. Error t value Pr(&gt;|t|)   
(Intercept) 81.83822   27.43765   2.983  0.00361 **
height      -0.07871    0.16032  -0.491  0.62456   
---
Signif. codes:  0 &#39;***&#39; 0.001 &#39;**&#39; 0.01 &#39;*&#39; 0.05 &#39;.&#39; 0.1 &#39; &#39; 1

Residual standard error: 14.56 on 98 degrees of freedom
Multiple R-squared:  0.002453,  Adjusted R-squared:  -0.007726 
F-statistic: 0.241 on 1 and 98 DF,  p-value: 0.6246</code></pre>
</div>
</div>
<div id="formula-operators" class="section level1">
<h1>Formula Operators</h1>
<div id="the-plus-sign-adding-predictors" class="section level2">
<h2>The Plus Sign (+): Adding Predictors</h2>
<p>Use <code>+</code> to include multiple independent predictors:</p>
<pre class="r"><code># Multiple predictors (additive model)
# &quot;weight depends on height AND age&quot;
model2 &lt;- lm(weight ~ height + age, data = df)
summary(model2)</code></pre>
<pre><code>
Call:
lm(formula = weight ~ height + age, data = df)

Residuals:
    Min      1Q  Median      3Q     Max 
-28.025 -10.423  -1.420   8.964  48.563 

Coefficients:
            Estimate Std. Error t value Pr(&gt;|t|)   
(Intercept) 87.81299   28.07134   3.128  0.00232 **
height      -0.08245    0.16035  -0.514  0.60831   
age         -0.13097    0.13018  -1.006  0.31689   
---
Signif. codes:  0 &#39;***&#39; 0.001 &#39;**&#39; 0.01 &#39;*&#39; 0.05 &#39;.&#39; 0.1 &#39; &#39; 1

Residual standard error: 14.56 on 97 degrees of freedom
Multiple R-squared:  0.01275,   Adjusted R-squared:  -0.007601 
F-statistic: 0.6266 on 2 and 97 DF,  p-value: 0.5365</code></pre>
</div>
<div id="the-colon-interactions" class="section level2">
<h2>The Colon (:): Interactions</h2>
<p>Use <code>:</code> to specify interactions between variables:</p>
<pre class="r"><code># Interaction term only
# &quot;weight depends on the interaction between height and age&quot;
model3 &lt;- lm(weight ~ height:age, data = df)
summary(model3)</code></pre>
<pre><code>
Call:
lm(formula = weight ~ height:age, data = df)

Residuals:
    Min      1Q  Median      3Q     Max 
-28.394  -9.711  -1.388   8.990  49.208 

Coefficients:
              Estimate Std. Error t value Pr(&gt;|t|)    
(Intercept) 73.5855405  5.3758323  13.688   &lt;2e-16 ***
height:age  -0.0007469  0.0007437  -1.004    0.318    
---
Signif. codes:  0 &#39;***&#39; 0.001 &#39;**&#39; 0.01 &#39;*&#39; 0.05 &#39;.&#39; 0.1 &#39; &#39; 1

Residual standard error: 14.5 on 98 degrees of freedom
Multiple R-squared:  0.01019,   Adjusted R-squared:  8.709e-05 
F-statistic: 1.009 on 1 and 98 DF,  p-value: 0.3177</code></pre>
</div>
<div id="the-asterisk-main-effects-interaction" class="section level2">
<h2>The Asterisk (*): Main Effects + Interaction</h2>
<p>Use <code>*</code> as shorthand for main effects plus their
interaction:</p>
<pre class="r"><code># Main effects plus interaction
# height * age expands to: height + age + height:age
model4 &lt;- lm(weight ~ height * age, data = df)
summary(model4)</code></pre>
<pre><code>
Call:
lm(formula = weight ~ height * age, data = df)

Residuals:
    Min      1Q  Median      3Q     Max 
-30.132 -10.888  -0.734   9.361  47.093 

Coefficients:
             Estimate Std. Error t value Pr(&gt;|t|)  
(Intercept) 238.16546  107.39400   2.218   0.0289 *
height       -0.95756    0.62430  -1.534   0.1284  
age          -3.84375    2.56409  -1.499   0.1371  
height:age    0.02162    0.01491   1.450   0.1504  
---
Signif. codes:  0 &#39;***&#39; 0.001 &#39;**&#39; 0.01 &#39;*&#39; 0.05 &#39;.&#39; 0.1 &#39; &#39; 1

Residual standard error: 14.48 on 96 degrees of freedom
Multiple R-squared:  0.03391,   Adjusted R-squared:  0.003718 
F-statistic: 1.123 on 3 and 96 DF,  p-value: 0.3437</code></pre>
<pre class="r"><code># This is equivalent to:
model4b &lt;- lm(weight ~ height + age + height:age, data = df)</code></pre>
</div>
<div id="the-minus-sign---removing-terms" class="section level2">
<h2>The Minus Sign (-): Removing Terms</h2>
<p>Use <code>-</code> to exclude specific terms:</p>
<pre class="r"><code># Remove a term
# Include all but remove age
model5 &lt;- lm(weight ~ height + age + gender - age, data = df)
# Equivalent to: lm(weight ~ height + gender, data = df)</code></pre>
</div>
<div id="the-dot-.-all-other-variables" class="section level2">
<h2>The Dot (.): All Other Variables</h2>
<p>Use <code>.</code> to include all variables except the response:</p>
<pre class="r"><code># Use all predictors in the dataset
model6 &lt;- lm(weight ~ ., data = df)
summary(model6)</code></pre>
<pre><code>
Call:
lm(formula = weight ~ ., data = df)

Residuals:
    Min      1Q  Median      3Q     Max 
-27.274 -10.122  -0.648   9.480  47.619 

Coefficients:
            Estimate Std. Error t value Pr(&gt;|t|)   
(Intercept) 85.73544   28.35303   3.024   0.0032 **
height      -0.07606    0.16117  -0.472   0.6380   
age         -0.12677    0.13076  -0.969   0.3347   
genderM      1.85382    2.95186   0.628   0.5315   
---
Signif. codes:  0 &#39;***&#39; 0.001 &#39;**&#39; 0.01 &#39;*&#39; 0.05 &#39;.&#39; 0.1 &#39; &#39; 1

Residual standard error: 14.61 on 96 degrees of freedom
Multiple R-squared:  0.01679,   Adjusted R-squared:  -0.01393 
F-statistic: 0.5466 on 3 and 96 DF,  p-value: 0.6516</code></pre>
<pre class="r"><code># Remove specific variables
model7 &lt;- lm(weight ~ . - gender, data = df)</code></pre>
</div>
<div id="the-caret-interactions-up-to-a-level" class="section level2">
<h2>The Caret (^): Interactions Up to a Level</h2>
<p>Use <code>^</code> to include interactions up to a specified
level:</p>
<pre class="r"><code># All main effects and 2-way interactions
model8 &lt;- lm(weight ~ (height + age + gender)^2, data = df)
# Expands to: height + age + gender + height:age + height:gender + age:gender

summary(model8)</code></pre>
<pre><code>
Call:
lm(formula = weight ~ (height + age + gender)^2, data = df)

Residuals:
    Min      1Q  Median      3Q     Max 
-28.950 -10.750  -1.079   7.571  43.103 

Coefficients:
                Estimate Std. Error t value Pr(&gt;|t|)  
(Intercept)    202.72867  111.36163   1.820   0.0719 .
height          -0.77940    0.64475  -1.209   0.2298  
age             -3.46884    2.60273  -1.333   0.1859  
genderM         68.60697   59.99128   1.144   0.2557  
height:age       0.02000    0.01506   1.328   0.1875  
height:genderM  -0.33507    0.34259  -0.978   0.3306  
age:genderM     -0.23413    0.26561  -0.881   0.3803  
---
Signif. codes:  0 &#39;***&#39; 0.001 &#39;**&#39; 0.01 &#39;*&#39; 0.05 &#39;.&#39; 0.1 &#39; &#39; 1

Residual standard error: 14.55 on 93 degrees of freedom
Multiple R-squared:  0.0552,    Adjusted R-squared:  -0.005753 
F-statistic: 0.9056 on 6 and 93 DF,  p-value: 0.4945</code></pre>
</div>
<div id="the-vertical-bar-conditioning-advanced" class="section level2">
<h2>The Vertical Bar (|): Conditioning (Advanced)</h2>
<p>Used in packages like <code>lattice</code> and <code>lme4</code> for
grouping:</p>
<pre class="r"><code># In lattice graphics: plot y by x, separately for each level of z
xyplot(weight ~ height | gender, data = df)

# In mixed models: random effects grouped by subject
library(lme4)
lmer(weight ~ height + (1|subject), data = df)</code></pre>
</div>
</div>
<div id="special-functions-in-formulas" class="section level1">
<h1>Special Functions in Formulas</h1>
<div id="i-identity-function" class="section level2">
<h2>I(): Identity Function</h2>
<p>Use <code>I()</code> to perform arithmetic operations literally:</p>
<pre class="r"><code># Square a variable
model9 &lt;- lm(weight ~ height + I(height^2), data = df)

# Without I(), the ^ operator has special meaning in formulas
# With I(), it&#39;s treated as arithmetic exponentiation</code></pre>
</div>
<div id="poly-polynomial-terms" class="section level2">
<h2>poly(): Polynomial Terms</h2>
<p>Create orthogonal polynomial terms:</p>
<pre class="r"><code># Second-degree polynomial
model10 &lt;- lm(weight ~ poly(height, 2), data = df)
summary(model10)</code></pre>
<pre><code>
Call:
lm(formula = weight ~ poly(height, 2), data = df)

Residuals:
    Min      1Q  Median      3Q     Max 
-29.052 -10.937  -1.786   9.816  49.979 

Coefficients:
                 Estimate Std. Error t value Pr(&gt;|t|)    
(Intercept)        68.387      1.456  46.980   &lt;2e-16 ***
poly(height, 2)1   -7.149     14.557  -0.491    0.624    
poly(height, 2)2  -14.952     14.557  -1.027    0.307    
---
Signif. codes:  0 &#39;***&#39; 0.001 &#39;**&#39; 0.01 &#39;*&#39; 0.05 &#39;.&#39; 0.1 &#39; &#39; 1

Residual standard error: 14.56 on 97 degrees of freedom
Multiple R-squared:  0.01319,   Adjusted R-squared:  -0.007159 
F-statistic: 0.6481 on 2 and 97 DF,  p-value: 0.5253</code></pre>
</div>
<div id="transformations" class="section level2">
<h2>Transformations</h2>
<p>Apply functions directly in formulas:</p>
<pre class="r"><code># Log transformation
model11 &lt;- lm(log(weight) ~ height, data = df)

# Multiple transformations
model12 &lt;- lm(weight ~ log(height) + sqrt(age), data = df)</code></pre>
</div>
</div>
<div id="practical-examples" class="section level1">
<h1>Practical Examples</h1>
<div id="example-1-data-visualization-with-formulas"
class="section level2">
<h2>Example 1: Data Visualization with Formulas</h2>
<pre class="r"><code># Scatter plot with formula notation
plot(weight ~ height, data = df,
     main = &quot;Weight vs Height&quot;,
     xlab = &quot;Height (cm)&quot;,
     ylab = &quot;Weight (kg)&quot;,
     pch = 19, col = &quot;blue&quot;)
abline(lm(weight ~ height, data = df), col = &quot;red&quot;, lwd = 2)</code></pre>
<p><img src="figure/formula.Rmd/visualization-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code># Boxplot with formula
boxplot(weight ~ gender, data = df,
        main = &quot;Weight by Gender&quot;,
        xlab = &quot;Gender&quot;,
        ylab = &quot;Weight (kg)&quot;,
        col = c(&quot;pink&quot;, &quot;lightblue&quot;))</code></pre>
<p><img src="figure/formula.Rmd/visualization-2.png" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="example-2-anova-with-formulas" class="section level2">
<h2>Example 2: ANOVA with Formulas</h2>
<pre class="r"><code># Create categorical data
df$age_group &lt;- cut(df$age, breaks = c(0, 30, 45, 100),
                     labels = c(&quot;Young&quot;, &quot;Middle&quot;, &quot;Senior&quot;))

# One-way ANOVA
aov1 &lt;- aov(weight ~ age_group, data = df)
summary(aov1)</code></pre>
<pre><code>            Df Sum Sq Mean Sq F value Pr(&gt;F)
age_group    2    305   152.5   0.721  0.489
Residuals   97  20523   211.6               </code></pre>
<pre class="r"><code># Two-way ANOVA with interaction
aov2 &lt;- aov(weight ~ age_group * gender, data = df)
summary(aov2)</code></pre>
<pre><code>                 Df Sum Sq Mean Sq F value Pr(&gt;F)
age_group         2    305  152.52   0.718  0.490
gender            1     92   91.89   0.433  0.512
age_group:gender  2    471  235.51   1.109  0.334
Residuals        94  19961  212.35               </code></pre>
</div>
<div id="example-3-data-manipulation-with-formulas"
class="section level2">
<h2>Example 3: Data Manipulation with Formulas</h2>
<pre class="r"><code># Aggregate function uses formulas
# Calculate mean weight by gender
aggregate(weight ~ gender, data = df, FUN = mean)</code></pre>
<pre><code>  gender   weight
1      F 67.47088
2      M 69.55251</code></pre>
<pre class="r"><code># Multiple variables
aggregate(cbind(weight, height) ~ gender + age_group, data = df, FUN = mean)</code></pre>
<pre><code>  gender age_group   weight   height
1      F     Young 70.60824 172.8042
2      M     Young 71.67341 172.5615
3      F    Middle 65.43550 169.9501
4      M    Middle 73.15206 168.0989
5      F    Senior 67.69240 171.9951
6      M    Senior 65.33892 170.4329</code></pre>
</div>
</div>
<div id="common-patterns-and-shortcuts" class="section level1">
<h1>Common Patterns and Shortcuts</h1>
<div id="no-intercept-models" class="section level2">
<h2>No Intercept Models</h2>
<pre class="r"><code># Remove intercept with -1 or +0
model13 &lt;- lm(weight ~ height - 1, data = df)
model14 &lt;- lm(weight ~ height + 0, data = df)</code></pre>
</div>
<div id="update-formula" class="section level2">
<h2>Update Formula</h2>
<pre class="r"><code># Update an existing model
model_updated &lt;- update(model2, . ~ . + gender)
# Adds gender to the existing formula

# Remove a term
model_updated2 &lt;- update(model2, . ~ . - age)</code></pre>
</div>
<div id="complex-interactions" class="section level2">
<h2>Complex Interactions</h2>
<pre class="r"><code># Three-way interaction
model15 &lt;- lm(weight ~ height * age * gender, data = df)

# Specific interaction structure
model16 &lt;- lm(weight ~ height + age + height:age + gender, data = df)</code></pre>
</div>
</div>
<div id="formula-cheat-sheet" class="section level1">
<h1>Formula Cheat Sheet</h1>
<table>
<colgroup>
<col width="35%" />
<col width="32%" />
<col width="32%" />
</colgroup>
<thead>
<tr class="header">
<th>Operator</th>
<th>Meaning</th>
<th>Example</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><code>~</code></td>
<td>“modeled by”</td>
<td><code>y ~ x</code></td>
</tr>
<tr class="even">
<td><code>+</code></td>
<td>Include term</td>
<td><code>y ~ x + z</code></td>
</tr>
<tr class="odd">
<td><code>-</code></td>
<td>Exclude term</td>
<td><code>y ~ . - x</code></td>
</tr>
<tr class="even">
<td><code>:</code></td>
<td>Interaction only</td>
<td><code>y ~ x:z</code></td>
</tr>
<tr class="odd">
<td><code>*</code></td>
<td>Main effects + interaction</td>
<td><code>y ~ x * z</code> equals <code>y ~ x + z + x:z</code></td>
</tr>
<tr class="even">
<td><code>.</code></td>
<td>All other variables</td>
<td><code>y ~ .</code></td>
</tr>
<tr class="odd">
<td><code>^</code></td>
<td>Interactions to degree</td>
<td><code>y ~ (x + z)^2</code></td>
</tr>
<tr class="even">
<td><code>I()</code></td>
<td>As-is arithmetic</td>
<td><code>y ~ I(x^2)</code></td>
</tr>
<tr class="odd">
<td><code>-1</code> or <code>+0</code></td>
<td>Remove intercept</td>
<td><code>y ~ x - 1</code></td>
</tr>
<tr class="even">
<td><code>1</code></td>
<td>Intercept only</td>
<td><code>y ~ 1</code></td>
</tr>
</tbody>
</table>
</div>
<div id="tips-for-success" class="section level1">
<h1>Tips for Success</h1>
<ol style="list-style-type: decimal">
<li><strong>Start simple</strong>: Begin with basic formulas and add
complexity as needed</li>
<li><strong>Check expansions</strong>: Use <code>model.matrix()</code>
to see how your formula expands</li>
<li><strong>Name clearly</strong>: Use descriptive variable names to
make formulas readable</li>
<li><strong>Test interactions</strong>: Always examine whether
interactions improve your model</li>
<li><strong>Read the docs</strong>: Different packages may interpret
formulas slightly differently</li>
</ol>
<pre class="r"><code># See how formula expands to design matrix
head(model.matrix(weight ~ height * gender, data = df))</code></pre>
<pre><code>  (Intercept)   height genderM height:genderM
1           1 164.3952       0         0.0000
2           1 167.6982       0         0.0000
3           1 185.5871       1       185.5871
4           1 170.7051       0         0.0000
5           1 171.2929       1       171.2929
6           1 187.1506       0         0.0000</code></pre>
</div>
<div id="conclusion" class="section level1">
<h1>Conclusion</h1>
<p>R’s formula syntax is a powerful and elegant way to specify
statistical models and data relationships. While it has a learning
curve, mastering formulas will make your R code more concise and
expressive. Practice with real data, and soon the notation will become
second nature!</p>
</div>
<div id="additional-resources" class="section level1">
<h1>Additional Resources</h1>
<ul>
<li>R Documentation: <code>?formula</code></li>
<li>Model specification: <code>?lm</code>, <code>?glm</code>,
<code>?aov</code></li>
<li>Formula tutorial: <code>vignette("formula")</code> in various
packages</li>
<li>Advanced modeling: Check packages like <code>lme4</code>,
<code>mgcv</code>, and <code>nlme</code> for more complex formula
applications</li>
</ul>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span>
Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 4.5.0 (2025-04-11)
Platform: x86_64-pc-linux-gnu
Running under: Ubuntu 24.04.3 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 
LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.26.so;  LAPACK version 3.12.0

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

time zone: Etc/UTC
tzcode source: system (glibc)

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] lubridate_1.9.4 forcats_1.0.0   stringr_1.5.1   dplyr_1.1.4    
 [5] purrr_1.0.4     readr_2.1.5     tidyr_1.3.1     tibble_3.3.0   
 [9] ggplot2_3.5.2   tidyverse_2.0.0 workflowr_1.7.1

loaded via a namespace (and not attached):
 [1] sass_0.4.10        generics_0.1.4     stringi_1.8.7      hms_1.1.3         
 [5] digest_0.6.37      magrittr_2.0.3     timechange_0.3.0   evaluate_1.0.3    
 [9] grid_4.5.0         RColorBrewer_1.1-3 fastmap_1.2.0      rprojroot_2.0.4   
[13] jsonlite_2.0.0     processx_3.8.6     whisker_0.4.1      ps_1.9.1          
[17] promises_1.3.3     httr_1.4.7         scales_1.4.0       jquerylib_0.1.4   
[21] cli_3.6.5          rlang_1.1.6        withr_3.0.2        cachem_1.1.0      
[25] yaml_2.3.10        tools_4.5.0        tzdb_0.5.0         httpuv_1.6.16     
[29] vctrs_0.6.5        R6_2.6.1           lifecycle_1.0.4    git2r_0.36.2      
[33] fs_1.6.6           pkgconfig_2.0.3    callr_3.7.6        pillar_1.10.2     
[37] bslib_0.9.0        later_1.4.2        gtable_0.3.6       glue_1.8.0        
[41] Rcpp_1.0.14        xfun_0.52          tidyselect_1.2.1   rstudioapi_0.17.1 
[45] knitr_1.50         farver_2.1.2       htmltools_0.5.8.1  rmarkdown_2.29    
[49] compiler_4.5.0     getPass_0.2-4     </code></pre>
</div>
</div>
</div>
</div>
</div>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
https://docs.mathjax.org/en/latest/web/configuration.html. This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>




</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
