<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />



<meta name="date" content="2024-08-03" />

<title>Plot ARCHS4 gene expression data</title>

<script src="site_libs/header-attrs-2.27/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/htmltools-fill-0.5.8.1/fill.css" rel="stylesheet" />
<script src="site_libs/htmlwidgets-1.6.4/htmlwidgets.js"></script>
<script src="site_libs/plotly-binding-4.10.4/plotly.js"></script>
<script src="site_libs/typedarray-0.1/typedarray.min.js"></script>
<link href="site_libs/crosstalk-1.2.1/css/crosstalk.min.css" rel="stylesheet" />
<script src="site_libs/crosstalk-1.2.1/js/crosstalk.min.js"></script>
<link href="site_libs/plotly-htmlwidgets-css-2.11.1/plotly-htmlwidgets.css" rel="stylesheet" />
<script src="site_libs/plotly-main-2.11.1/plotly-latest.min.js"></script>
<link href="site_libs/font-awesome-6.4.2/css/all.min.css" rel="stylesheet" />
<link href="site_libs/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet" />

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/main/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->




</head>

<body>


<div class="container-fluid main-container">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">My blog</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/davetang/muse">
    <span class="fa fa-github"></span>
     
    Source code
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Plot ARCHS4 gene expression data</h1>
<h4 class="date">2024-08-03</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span>
workflowr <span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2024-08-03
</p>
<p>
<strong>Checks:</strong> <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 7
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong> <code>muse/</code> <span
class="glyphicon glyphicon-question-sign" aria-hidden="true"
title="This is the local directory in which the code in this file was executed.">
</span>
</p>
<p>
This reproducible <a href="https://rmarkdown.rstudio.com">R Markdown</a>
analysis was created with <a
  href="https://github.com/workflowr/workflowr">workflowr</a> (version
1.7.1). The <em>Checks</em> tab describes the reproducibility checks
that were applied when the results were created. The <em>Past
versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date
</a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git
repository, you know the exact version of the code that produced these
results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the
global environment can affect the analysis in your R Markdown file in
unknown ways. For reproduciblity it’s best to always run the code in an
empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20200712code">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Seed:</strong>
<code>set.seed(20200712)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20200712code"
class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20200712)</code> was run prior to running
the code in the R Markdown file. Setting a seed ensures that any results
that rely on randomness, e.g. subsampling or permutations, are
reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Session information:</strong>
recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package
versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be
confident that you successfully produced the results during this
run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr
project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomdavetangmusetreeee406736a2459af1413aa4f1540d1fe86448566atargetblankee40673a">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Repository version:</strong>
<a href="https://github.com/davetang/muse/tree/ee406736a2459af1413aa4f1540d1fe86448566a" target="_blank">ee40673</a>
</a>
</p>
</div>
<div
id="strongRepositoryversionstrongahrefhttpsgithubcomdavetangmusetreeee406736a2459af1413aa4f1540d1fe86448566atargetblankee40673a"
class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development
and connecting the code version to the results is critical for
reproducibility.
</p>
<p>
The results in this page were generated with repository version
<a href="https://github.com/davetang/muse/tree/ee406736a2459af1413aa4f1540d1fe86448566a" target="_blank">ee40673</a>.
See the <em>Past versions</em> tab to see a history of the changes made
to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for
the analysis have been committed to Git prior to generating the results
(you can use <code>wflow_publish</code> or
<code>wflow_git_commit</code>). workflowr only checks the R Markdown
file, but you know if there are other scripts or data files that it
depends on. Below is the status of the Git repository when the results
were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .Rhistory
    Ignored:    .Rproj.user/
    Ignored:    r_packages_4.3.3/
    Ignored:    r_packages_4.4.0/

Untracked files:
    Untracked:  data/archs4/CCL4_correlation.csv
    Untracked:  data/archs4/TNF_correlation.csv

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not
included in this status report because it is ok for generated content to
have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">
<p>
These are the previous versions of the repository in which changes were
made to the R Markdown (<code>analysis/gene_heatmap.Rmd</code>) and HTML
(<code>docs/gene_heatmap.html</code>) files. If you’ve configured a
remote Git repository (see <code>?wflow_git_remote</code>), click on the
hyperlinks in the table below to view the files as they were in that
past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/davetang/muse/blob/ee406736a2459af1413aa4f1540d1fe86448566a/analysis/gene_heatmap.Rmd" target="_blank">ee40673</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-08-03
</td>
<td>
Top 100 most correlated
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/davetang/muse/42c3d3910fb14b1e835279adea7f9e4a87976b1b/docs/gene_heatmap.html" target="_blank">42c3d39</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-08-03
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/davetang/muse/blob/b5aa7d60142cc36560309eafaf20eb682662c0ae/analysis/gene_heatmap.Rmd" target="_blank">b5aa7d6</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-08-03
</td>
<td>
Plot TNF expression as a boxplot
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/davetang/muse/df009693c18b4da612aa0abda9ac58ef2d30d02d/docs/gene_heatmap.html" target="_blank">df00969</a>
</td>
<td>
Dave Tang
</td>
<td>
2023-07-14
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/davetang/muse/blob/7188c559ea817aea231f610a0e3f308d3523b60f/analysis/gene_heatmap.Rmd" target="_blank">7188c55</a>
</td>
<td>
Dave Tang
</td>
<td>
2023-07-14
</td>
<td>
pheatmap resolution
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/davetang/muse/e2d73ebbdf709e5653ea5fb98c76b1970a92d070/docs/gene_heatmap.html" target="_blank">e2d73eb</a>
</td>
<td>
Dave Tang
</td>
<td>
2023-07-14
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/davetang/muse/blob/3a5ec3ff1a669a95a52a9c9883b126961952b5d4/analysis/gene_heatmap.Rmd" target="_blank">3a5ec3f</a>
</td>
<td>
Dave Tang
</td>
<td>
2023-07-14
</td>
<td>
Interactive heatmap
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/davetang/muse/7fa8b413e65d0b6e323896aee31c32eb05e0c2dd/docs/gene_heatmap.html" target="_blank">7fa8b41</a>
</td>
<td>
Dave Tang
</td>
<td>
2023-07-13
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/davetang/muse/blob/922aa566e9e1e05b14a042278bd82434d6ae9071/analysis/gene_heatmap.Rmd" target="_blank">922aa56</a>
</td>
<td>
Dave Tang
</td>
<td>
2023-07-13
</td>
<td>
ARCHS4 heatmap
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<div id="archs4" class="section level2">
<h2>ARCHS4</h2>
<p>Use <a href="https://github.com/pachterlab/gget">gget</a> and the <a
href="https://pachterlab.github.io/gget/en/archs4.html">archs4</a>
subtool to find the most correlated genes to a gene of interest or find
the gene’s tissue expression atlas using <a
href="https://maayanlab.cloud/archs4/">ARCHS4</a>.</p>
<p><a
href="https://pachterlab.github.io/gget/en/installation.html">Install</a>
using pip:</p>
<pre><code>pip install --upgrade gget</code></pre>
<p>Usage.</p>
<pre><code>gget archs4 -h</code></pre>
<pre><code>usage: gget archs4 [-h] [-e] [-w {correlation,tissue}] [-gc GENE_COUNT] [-s {human,mouse}] [-csv] [-o OUT] [-q] [-g GENE_DEPRECATED] [-j] gene

Find the most correlated genes or the tissue expression atlas of a gene using data from the human and mouse RNA-seq database ARCHS4 (https://maayanlab.cloud/archs4/).

positional arguments:
  gene                  Gene symbol or Ensembl gene ID of gene of interest, e.g. &#39;STAT4&#39;.

options:
  -h, --help            show this help message and exit
  -e, --ensembl         Add this flag if gene is given as an Ensembl gene ID. (default: False)
  -w {correlation,tissue}, --which {correlation,tissue}
                        &#39;correlation&#39; (default) or &#39;tissue&#39;.
                        - &#39;correlation&#39; returns a gene correlation table that contains the 100 most correlated genes to the gene of interest. The Pearson correlation is calculated over all samples and tissues in ARCHS4.
                        - &#39;tissue&#39; returns a tissue expression atlas calculated from human or mouse samples (as defined by &#39;species&#39;) in ARCHS4. (default: correlation)
  -gc GENE_COUNT, --gene_count GENE_COUNT
                        Number of correlated genes to return (default: 100).
                        (Only for gene correlation.) (default: 100)
  -s {human,mouse}, --species {human,mouse}
                        &#39;human&#39; (default) or &#39;mouse&#39;. (Only for tissue expression atlas.) (default: human)
  -csv, --csv           Returns results in csv format instead of json. (default: True)
  -o OUT, --out OUT     Path to the csv file the results will be saved in, e.g. path/to/directory/results.csv.
                        Default: Standard out.
  -q, --quiet           Does not print progress information. (default: True)
  -g GENE_DEPRECATED, --gene GENE_DEPRECATED
                        DEPRECATED - use positional argument instead. Gene symbol or Ensembl gene ID of gene of interest (str), e.g. &#39;STAT4&#39;.
  -j, --json            DEPRECATED - json is now the default output format (convert to csv using flag [--csv]). (default: False)</code></pre>
<p>Get tissue expression for TNF in CSV.</p>
<pre><code>gget archs4 --which tissue --csv --out TNF_tissue.csv TNF</code></pre>
<p>Get the 100 most correlated genes to TNF from ARCHS4.</p>
<pre><code>gget archs4 --which correlation --csv --out TNF_correlation.csv TNF</code></pre>
</div>
<div id="tnf-tissue-expression" class="section level2">
<h2>TNF tissue expression</h2>
<p>Load data.</p>
<pre class="r"><code>tnf_exp &lt;- readr::read_csv(file = &quot;data/archs4/TNF_tissue.csv&quot;, show_col_types = FALSE)
head(tnf_exp)</code></pre>
<pre><code># A tibble: 6 × 6
  id                                              min    q1 median    q3   max
  &lt;chr&gt;                                         &lt;dbl&gt; &lt;dbl&gt;  &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;
1 System.Immune System.Myeloid.MACROPHAGE       0.114  6.54   9.01 11.2  14.1 
2 System.Immune System.Myeloid.DENDRITIC CELL   4.76   6.81   8.85 11.5  15.2 
3 System.Immune System.Spleen.SPLEEN            0.114  6.63   8.82 10.7  13.2 
4 System.Immune System.Lymphoid.TLYMPHOCYTE     0.114  6.22   8.75 10.4  12.6 
5 System.Immune System.Granulocytic.NEUTROPHIL  5.71   6.94   7.63  9.62 11.2 
6 System.Immune System.Granulocytic.GRANULOCYTE 0.114  5.04   7.22  8.44  9.50</code></pre>
<p>Split the ID into multiple columns.</p>
<pre class="r"><code>tnf_exp |&gt;
  tidyr::separate(col = id, sep = &quot;\\.&quot;, into = c(&quot;junk&quot;, &quot;system&quot;, &quot;organ&quot;, &quot;tissue&quot;)) |&gt;
  dplyr::mutate(tissue = stringr::str_to_title(tissue)) |&gt;
  dplyr::arrange(system) |&gt;
  dplyr::mutate(tissue = factor(tissue, levels = tissue)) -&gt; tnf_exp

head(tnf_exp)</code></pre>
<pre><code># A tibble: 6 × 9
  junk   system                organ       tissue   min    q1 median    q3   max
  &lt;chr&gt;  &lt;chr&gt;                 &lt;chr&gt;       &lt;fct&gt;  &lt;dbl&gt; &lt;dbl&gt;  &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;
1 System Cardiovascular System Heart       Valve  0.114 1.21   3.23   4.30  5.16
2 System Cardiovascular System Heart       Ventr… 0.114 2.16   2.78   3.39  4.31
3 System Cardiovascular System Heart       Atrium 0.114 0.114  1.21   2.78  3.35
4 System Cardiovascular System Heart       Heart  0.114 0.114  0.114  2.42  8.60
5 System Connective Tissue     Adipose ti… Adipo… 0.114 0.114  2.42   3.29  4.84
6 System Connective Tissue     Bone marrow Chond… 0.114 0.114  0.114  1.81  9.13</code></pre>
<p>Plot tissue expression as box plots.</p>
<pre class="r"><code>library(ggplot2)
ggplot(
  tnf_exp,
  aes(
    x = tissue,
    ymin = min,
    lower = q1,
    middle = median,
    upper = q3,
    ymax = max,
    fill = system
  )
) +
  geom_boxplot(stat = &#39;identity&#39;) +
  theme_minimal() +
  theme(
    legend.position = &#39;right&#39;,
    axis.title.x = element_blank(),
    axis.text.x = element_text(angle = 70, hjust = 1)
  ) +
  labs(y = &quot;Expression (TPM)&quot;)</code></pre>
<p><img src="figure/gene_heatmap.Rmd/tnf_box_plot-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-tnf_box_plot-1">
Past versions of tnf_box_plot-1.png
</button>
</p>
<div id="fig-tnf_box_plot-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/davetang/muse/blob/42c3d3910fb14b1e835279adea7f9e4a87976b1b/docs/figure/gene_heatmap.Rmd/tnf_box_plot-1.png" target="_blank">42c3d39</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-08-03
</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<div id="top-100-most-correlated-genes-to-tnf" class="section level2">
<h2>Top 100 most correlated genes to TNF</h2>
<p>Load data.</p>
<pre class="r"><code>tnf_cor &lt;- readr::read_csv(file = &quot;data/archs4/TNF_correlation.csv&quot;, show_col_types = FALSE)
head(tnf_cor)</code></pre>
<pre><code># A tibble: 6 × 2
  gene_symbol pearson_correlation
  &lt;chr&gt;                     &lt;dbl&gt;
1 CCL4                      0.779
2 GPR132                    0.765
3 SLAMF1                    0.763
4 CCL3                      0.758
5 CD48                      0.754
6 OSM                       0.745</code></pre>
<p>CCL4 is the most correlated to TNF. Get the 100 most correlated genes
to CCL4 from ARCHS4 to check if this is reciprocated.</p>
<pre><code>gget archs4 --which correlation --csv --out CCL4_correlation.csv CCL4</code></pre>
<p>Load CCL4 correlation data.</p>
<pre class="r"><code>ccl4_cor &lt;- readr::read_csv(file = &quot;data/archs4/CCL4_correlation.csv&quot;, show_col_types = FALSE)
head(ccl4_cor)</code></pre>
<pre><code># A tibble: 6 × 2
  gene_symbol pearson_correlation
  &lt;chr&gt;                     &lt;dbl&gt;
1 CCL4L2                    0.923
2 CCL3                      0.913
3 CCL3L3                    0.854
4 CD48                      0.843
5 PLEK                      0.833
6 CYTIP                     0.820</code></pre>
<p>Check rank of TNF.</p>
<pre class="r"><code>which(ccl4_cor$gene_symbol == &quot;TNF&quot;)</code></pre>
<pre><code>[1] 36</code></pre>
<p>Check correlation as a sanity check.</p>
<pre class="r"><code>ccl4_cor[which(ccl4_cor$gene_symbol == &quot;TNF&quot;), ]</code></pre>
<pre><code># A tibble: 1 × 2
  gene_symbol pearson_correlation
  &lt;chr&gt;                     &lt;dbl&gt;
1 TNF                       0.779</code></pre>
</div>
<div id="load-data" class="section level2">
<h2>Load data</h2>
<p>Prepare data using base R.</p>
<pre class="r"><code>lapply(
  list.files(&quot;data/archs4/cancer&quot;, pattern = &quot;.csv$&quot;, full.names = TRUE),
  function(x){
    cbind(gene = sub(&quot;\\.\\w+$&quot;, &quot;&quot;, basename(x)), read.csv(x))
  }
) |&gt;
  do.call(&quot;rbind&quot;, args = _) -&gt; my_df

# Split `id` column.
do.call(&quot;rbind&quot;, strsplit(x = my_df$id, split = &quot;\\.&quot;)) |&gt;
  as.data.frame() -&gt; id_split

colnames(id_split) &lt;- c(&#39;root&#39;, &#39;system&#39;, &#39;organ&#39;, &#39;tissue&#39;)

# Rename tissues.
cap_first &lt;- function(x){
  s &lt;- strsplit(x, &quot;&quot;)[[1]][1]
  return(sub(s, toupper(s), x))
}

id_split$tissue &lt;- tolower(id_split$tissue)
id_split$tissue &lt;- sapply(id_split$tissue, cap_first)

my_df &lt;- cbind(my_df, id_split)

# Order `my_df` by system.
my_df &lt;- my_df[order(my_df$gene, my_df$system), ]
my_df$tissue &lt;- factor(my_df$tissue, levels = unique(my_df$tissue))

head(my_df)</code></pre>
<pre><code>    gene                                                id      min       q1
12 CCND1          System.Cardiovascular System.Heart.VALVE 10.62560 11.68490
28 CCND1          System.Cardiovascular System.Heart.HEART  5.87724 10.15820
30 CCND1      System.Cardiovascular System.Heart.VENTRICLE  9.54469 10.37180
36 CCND1         System.Cardiovascular System.Heart.ATRIUM  8.44515  9.67321
5  CCND1          System.Connective Tissue.Bone.OSTEOBLAST 11.30840 12.09570
18 CCND1 System.Connective Tissue.Adipose tissue.ADIPOCYTE  8.38312 10.48580
    median      q3     max   root                system          organ
12 12.0648 12.5311 13.7986 System Cardiovascular System          Heart
28 10.9207 11.5210 12.8617 System Cardiovascular System          Heart
30 10.8446 11.2841 11.9118 System Cardiovascular System          Heart
36 10.5234 11.0560 11.4873 System Cardiovascular System          Heart
5  12.6214 13.2789 14.0211 System     Connective Tissue           Bone
18 11.7684 12.7769 14.1867 System     Connective Tissue Adipose tissue
       tissue
12      Valve
28      Heart
30  Ventricle
36     Atrium
5  Osteoblast
18  Adipocyte</code></pre>
<p>Back to wide format.</p>
<pre class="r"><code>my_df |&gt;
  dplyr::select(gene, median, tissue) |&gt;
  tidyr::pivot_wider(names_from = tissue, values_from = median) -&gt; my_df_wide</code></pre>
<p>Convert to matrix and plot.</p>
<pre class="r"><code>my_mat &lt;- as.matrix(my_df_wide[, -1])
row.names(my_mat) &lt;- my_df_wide$gene

pheatmap(my_mat)</code></pre>
<p><img src="figure/gene_heatmap.Rmd/pheatmap-1.png" width="1152" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-pheatmap-1">
Past versions of pheatmap-1.png
</button>
</p>
<div id="fig-pheatmap-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/davetang/muse/blob/df009693c18b4da612aa0abda9ac58ef2d30d02d/docs/figure/gene_heatmap.Rmd/pheatmap-1.png" target="_blank">df00969</a>
</td>
<td>
Dave Tang
</td>
<td>
2023-07-14
</td>
</tr>
<tr>
<td>
<a href="https://github.com/davetang/muse/blob/e2d73ebbdf709e5653ea5fb98c76b1970a92d070/docs/figure/gene_heatmap.Rmd/pheatmap-1.png" target="_blank">e2d73eb</a>
</td>
<td>
Dave Tang
</td>
<td>
2023-07-14
</td>
</tr>
<tr>
<td>
<a href="https://github.com/davetang/muse/blob/7fa8b413e65d0b6e323896aee31c32eb05e0c2dd/docs/figure/gene_heatmap.Rmd/pheatmap-1.png" target="_blank">7fa8b41</a>
</td>
<td>
Dave Tang
</td>
<td>
2023-07-13
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>Create sample annotation.</p>
<pre class="r"><code>my_order &lt;- colnames(my_mat)

my_df |&gt;
  dplyr::select(system, tissue) |&gt;
  dplyr::distinct() |&gt;
  dplyr::arrange(match(tissue, my_order)) |&gt;
  dplyr::select(-tissue) -&gt; sample_anno

row.names(sample_anno) &lt;- my_order
head(sample_anno)</code></pre>
<pre><code>                          system
Valve      Cardiovascular System
Heart      Cardiovascular System
Ventricle  Cardiovascular System
Atrium     Cardiovascular System
Osteoblast     Connective Tissue
Adipocyte      Connective Tissue</code></pre>
<p>Heatmap with system annotation.</p>
<pre class="r"><code>pheatmap(my_mat, annotation_col = sample_anno)</code></pre>
<p><img src="figure/gene_heatmap.Rmd/pheatmap_with_anno-1.png" width="1152" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-pheatmap_with_anno-1">
Past versions of pheatmap_with_anno-1.png
</button>
</p>
<div id="fig-pheatmap_with_anno-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/davetang/muse/blob/df009693c18b4da612aa0abda9ac58ef2d30d02d/docs/figure/gene_heatmap.Rmd/pheatmap_with_anno-1.png" target="_blank">df00969</a>
</td>
<td>
Dave Tang
</td>
<td>
2023-07-14
</td>
</tr>
<tr>
<td>
<a href="https://github.com/davetang/muse/blob/e2d73ebbdf709e5653ea5fb98c76b1970a92d070/docs/figure/gene_heatmap.Rmd/pheatmap_with_anno-1.png" target="_blank">e2d73eb</a>
</td>
<td>
Dave Tang
</td>
<td>
2023-07-14
</td>
</tr>
<tr>
<td>
<a href="https://github.com/davetang/muse/blob/7fa8b413e65d0b6e323896aee31c32eb05e0c2dd/docs/figure/gene_heatmap.Rmd/pheatmap_with_anno-1.png" target="_blank">7fa8b41</a>
</td>
<td>
Dave Tang
</td>
<td>
2023-07-13
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>Interactive heatmap.</p>
<pre class="r"><code>plot_ly(
  x=colnames(my_mat),
  y=rownames(my_mat),
  z = my_mat,
  colors = colorRamp(c(&quot;green&quot;, &quot;red&quot;)),
  type = &quot;heatmap&quot;
)</code></pre>
<div class="plotly html-widget html-fill-item" id="htmlwidget-dd063ddf2a01a55af4fb" style="width:1152px;height:576px;"></div>
<script type="application/json" data-for="htmlwidget-dd063ddf2a01a55af4fb">{"x":{"visdat":{"2a951b19f083":["function () ","plotlyVisDat"]},"cur_data":"2a951b19f083","attrs":{"2a951b19f083":{"x":["Valve","Heart","Ventricle","Atrium","Osteoblast","Adipocyte","Stromal cell","Adipose","Chondrocyte","Bone marrow","Hepatic stellate cell","Colon mucosa","Esophagus","Gastric epithelial cell","Colon","Liver","Hepatocyte","Small intestine","Beta cell","Intestinal epithelial cell","Pancreatic islet","Alpha cell","Kupffer cell","Microglia","Macrophage","Dendritic cell","Blymphocyte","Spleen","Tlymphocyte","Neutrophil","Thymus","Plasmacytoid dendritic cell","Thymocyte","Granulocyte","Plasma cell","Melanocyte","Hair follicle","Fibroblast","Skin","Keratinocyte","Basal cell","Airway smooth muscle","Vascular smooth muscle","Myofibroblast","Myoblast","Skeletal muscle","Medulla","Astrocyte","Retina","Oligodendrocyte","Spinal cord","Thalamus","Hypothalamus","Neuron","Motor neuron","Pons","Midbrain","Cerebellum","Sensory neuron","Cerebral cortex","Lung epithelial cell","Trachea","Lung","Podocyte","Mammary gland","Breast","Renal cortex","Kidney","Ovary","Testis","Granulosa","Oocyte"],"y":["CCND1","ERBB2","FGFR1","GATA3","MYC","PIK3CA","PTEN","TP53"],"z":[[12.0648,10.9207,10.8446,10.523400000000001,12.6214,11.7684,11.521000000000001,10.6913,10.6685,2.4196800000000001,12.7034,11.9095,11.4069,11.085699999999999,10.856299999999999,10.753500000000001,9.9708000000000006,9.0804500000000008,8.8523200000000006,8.4126200000000004,6.48238,6.3763100000000001,11.437200000000001,10.6508,9.4730000000000008,7.5060700000000002,5.6535700000000002,5.2597100000000001,4.5991099999999996,4.47438,4.3693600000000004,4.3131199999999996,4.2694700000000001,3.0561799999999999,0.113644,13.4353,12.513400000000001,12.416600000000001,11.8024,11.711,11.421099999999999,13.055300000000001,12.920299999999999,12.1812,11.925599999999999,7.5241100000000003,11.4955,11.344200000000001,10.5601,10.4015,9.7479899999999997,9.4297000000000004,9.4297000000000004,9.4048400000000001,8.9287500000000009,8.0427199999999992,7.7495399999999997,6.9254800000000003,6.5832100000000002,6.4102899999999998,12.165800000000001,11.787599999999999,11.462199999999999,12.5558,12.2423,11.914199999999999,10.2394,9.85717,9.4542699999999993,7.5683999999999996,4.6540299999999997,4.3868200000000002],[10.2692,9.3984699999999997,10.273899999999999,10.6142,10.146599999999999,9.2744099999999996,9.46706,8.4478799999999996,9.8323499999999999,7.3108500000000003,10.124499999999999,10.5663,9.4771800000000006,10.4719,10.841699999999999,9.7904400000000003,10.4771,9.1651600000000002,7.4385199999999996,10.248200000000001,8.1997699999999991,7.3108500000000003,10.5198,8.4512400000000003,5.98306,5.9674300000000002,7.3704799999999997,6.6467400000000003,6.5713100000000004,7.1483600000000003,7.0423799999999996,6.3141800000000003,7.13293,6.8102799999999997,7.1829599999999996,10.136900000000001,10.578799999999999,9.7924399999999991,10.190799999999999,10.2029,10.0199,10.432600000000001,10.073499999999999,10.323399999999999,9.6507799999999992,9.2034800000000008,9.3817599999999999,9.9847999999999999,8.7984799999999996,10.097099999999999,9.5164399999999993,9.5203000000000007,9.9596900000000002,8.7323199999999996,7.9858399999999996,8.1150000000000002,8.3779299999999992,7.0922700000000001,6.2248999999999999,8.7713199999999993,9.3227899999999995,10.2011,9.6198700000000006,10.0374,10.7765,10.5875,11.8561,9.7066700000000008,10.0396,8.5101800000000001,8.3609399999999994,6.4520200000000001],[11.494899999999999,10.283799999999999,11.083399999999999,11.279,11.2865,11.1631,10.811500000000001,11.8192,11.940200000000001,8.0816400000000002,11.635899999999999,10.649699999999999,8.5249100000000002,8.2421100000000003,8.6755300000000002,7.8329300000000002,7.2387100000000002,9.8526600000000002,10.4473,5.9979899999999997,10.0191,10.540699999999999,8.3689599999999995,9.29955,8.8127700000000004,8.1465300000000003,6.8853900000000001,8.0115800000000004,7.9008799999999999,6.61965,8.6386599999999998,4.0296099999999999,9.6008300000000002,7.2214600000000004,7.1003299999999996,10.7204,11.8367,11.446300000000001,10.6363,5.9359099999999998,7.2675799999999997,13.1328,11.0885,11.650499999999999,9.8328600000000002,9.46706,9.9665999999999997,11.184100000000001,9.8614899999999999,11.5328,10.9184,10.8344,10.967499999999999,10.5382,10.2463,7.4642600000000003,9.7124299999999995,11.4183,10.5512,10.453099999999999,9.0888500000000008,6.7873599999999996,9.7532300000000003,10.324999999999999,8.8013300000000001,9.1505700000000001,11.198399999999999,9.7830300000000001,10.7105,9.8774999999999995,10.014099999999999,9.6029800000000005],[5.5238399999999999,3.5682499999999999,4.0038799999999997,3.2851699999999999,1.2096800000000001,2.7759499999999999,0.113644,6.3010299999999999,3.2294299999999998,0.113644,2.9245700000000001,3.1149399999999998,4.4311100000000003,2.7759499999999999,3.7696100000000001,3.48123,3.0561799999999999,2.6169600000000002,0.113644,2.4196800000000001,0.113644,0.113644,3.2294299999999998,0.113644,2.4196800000000001,2.7759499999999999,4.0553400000000002,9.2616599999999991,9.4509299999999996,7.6995300000000002,9.0554299999999994,3.7700200000000001,11.226100000000001,4.0572499999999998,0.113644,0.113644,9.1523400000000006,1.81287,5.3428199999999997,8.2046600000000005,3.2236699999999998,1.2096800000000001,3.2851699999999999,1.81287,1.2096800000000001,3.8958400000000002,4.8804299999999996,2.4196800000000001,0.113644,5.91845,1.2096800000000001,4.63537,4.63537,1.2096800000000001,0.113644,7.3633100000000002,1.2096800000000001,0.113644,1.81287,0.113644,6.1461399999999999,1.81287,4.5603400000000001,3.2851699999999999,11.2203,10.764699999999999,7.5320200000000002,6.9902899999999999,3.8321800000000001,1.2096800000000001,3.52474,3.6447799999999999],[9.5263299999999997,8.2964599999999997,7.5097399999999999,8.1248199999999997,9.6937599999999993,9.6198700000000006,9.2215600000000002,7.0465799999999996,9.3426200000000001,9.1293699999999998,9.8964800000000004,8.7035400000000003,9.59131,10.5909,10.0093,9.4591899999999995,9.0884,7.82186,3.48123,6.8385499999999997,0.113644,0.113644,8.7527200000000001,6.9753999999999996,8.2757900000000006,8.53721,11.872299999999999,9.9686599999999999,10.7232,9.2448099999999993,9.5158400000000007,6.7123400000000002,8.7956500000000002,6.4439099999999998,9.6108499999999992,10.7722,9.3195399999999999,9.2270099999999999,9.5696100000000008,11.5671,10.775499999999999,8.80701,10.001200000000001,9.6650799999999997,9.3877600000000001,6.8452999999999999,8.0750200000000003,7.7964799999999999,5.3010799999999998,5.7600600000000002,6.49925,5.4302000000000001,5.7092400000000003,6.9753999999999996,6.8994900000000001,9.44557,7.2561600000000004,2.4196800000000001,8.2473799999999997,7.9199599999999997,10.478,10.671799999999999,9.9193599999999993,9.7766400000000004,10.1158,10.3813,8.1209799999999994,9.3474799999999991,9.4467599999999994,6.1387499999999999,9.7781300000000009,8.4035700000000002],[9.6330799999999996,8.9930099999999999,9.4407599999999992,8.0389199999999992,9.3847000000000005,10.108000000000001,9.2443799999999996,10.034000000000001,9.2190999999999992,8.4560600000000008,9.2154100000000003,9.1133299999999995,8.7901199999999999,8.8259100000000004,8.2606400000000004,8.8217499999999998,8.7294199999999993,8.6069499999999994,8.4888399999999997,8.5842399999999994,7.0553499999999998,7.9278000000000004,9.2331900000000005,9.2426200000000005,9.2034800000000008,9.2928300000000004,9.1458100000000009,9.5156299999999998,9.7224299999999992,9.3351799999999994,9.6424599999999998,7.9747599999999998,10.183,9.7064000000000004,8.5101800000000001,8.4462700000000002,9.2782800000000005,9.0671300000000006,8.9559999999999995,8.8924699999999994,9.1840399999999995,9.2950900000000001,9.1427600000000009,8.9765999999999995,9.2342399999999998,9.32958,9.2635699999999996,9.1790800000000008,9.1760099999999998,9.93459,9.3514099999999996,8.2503399999999996,8.2503399999999996,9.10398,8.9685699999999997,9.2175999999999991,8.4344000000000001,10.491400000000001,8.4243799999999993,9.5988500000000005,7.5343600000000004,8.9287500000000009,8.5424399999999991,8.7597799999999992,9.1519700000000004,8.9163399999999999,9.0857100000000006,8.3763799999999993,8.3716399999999993,8.4327100000000002,8.1715199999999992,8.6125299999999996],[10.757,10.9533,10.7224,10.427899999999999,10.545,11.4962,11.278700000000001,11.5793,10.9876,11.671799999999999,10.7044,10.6333,10.407500000000001,10.032500000000001,10.5238,10.8185,10.4681,10.925800000000001,10.7807,10.7614,11.765700000000001,11.4183,10.7713,10.8155,10.411199999999999,10.786799999999999,11.210000000000001,10.8254,10.5311,11.671799999999999,11.819699999999999,11.0747,11.826700000000001,12.3726,11.2209,10.4216,10.9674,10.804500000000001,10.867800000000001,9.9908400000000004,10.6044,10.9421,10.5678,10.1953,11.288600000000001,10.203799999999999,11.062200000000001,10.5945,10.8607,11.15,10.432499999999999,10.097099999999999,10.154299999999999,10.700799999999999,10.8833,11.269399999999999,10.5154,12.360300000000001,9.9449400000000008,10.2043,10.3035,10.101800000000001,10.4476,9.8508499999999994,10.8614,10.2399,10.797499999999999,10.2509,9.9159100000000002,10.828799999999999,10.1767,10.1387],[9.3990200000000002,9.0623400000000007,6.9304100000000002,7.5760699999999996,9.6022499999999997,8.6650100000000005,9.5024499999999996,8.4279499999999992,9.3241999999999994,9.8517499999999991,9.9527000000000001,9.2112300000000005,9.2383900000000008,9.9471000000000007,9.4560099999999991,8.6330299999999998,8.8100299999999994,8.1777300000000004,8.0035399999999992,8.4321599999999997,8.4008699999999994,7.67462,9.1559100000000004,9.2761899999999997,8.6006300000000007,9.1427600000000009,10.230700000000001,9.3877600000000001,9.5356699999999996,8.8242999999999991,10.0527,9.6650799999999997,9.9465299999999992,7.6972300000000002,9.2525399999999998,9.5987200000000001,9.5754300000000008,9.5108899999999998,9.4951000000000008,9.6947299999999998,8.8680699999999995,9.1427600000000009,9.8591099999999994,9.6818600000000004,9.5455400000000008,7.2994500000000002,9.1116600000000005,9.4018800000000002,8.3892399999999991,9.3094999999999999,7.6451200000000004,7.5889600000000002,7.5889600000000002,8.5399399999999996,8.6698599999999999,9.3273700000000002,7.3604700000000003,6.1613300000000004,7.4111000000000002,10.204800000000001,10.206799999999999,8.8720700000000008,9.3587799999999994,10.324999999999999,10.0382,10.108000000000001,7.82172,9.4591899999999995,9.0280500000000004,9.5074100000000001,8.7022600000000008,8.15151]],"colors":["function (x) ","roundcolor(cbind(palette[[1L]](x), palette[[2L]](x), palette[[3L]](x), ","    if (alpha) palette[[4L]](x))) * 255"],"alpha_stroke":1,"sizes":[10,100],"spans":[1,20],"type":"heatmap"}},"layout":{"margin":{"b":40,"l":60,"t":25,"r":10},"xaxis":{"domain":[0,1],"automargin":true,"title":[]},"yaxis":{"domain":[0,1],"automargin":true,"title":[]},"scene":{"zaxis":{"title":[]}},"hovermode":"closest","showlegend":false,"legend":{"yanchor":"top","y":0.5}},"source":"A","config":{"modeBarButtonsToAdd":["hoverclosest","hovercompare"],"showSendToCloud":false},"data":[{"colorbar":{"title":"","ticklen":2,"len":0.5,"lenmode":"fraction","y":1,"yanchor":"top"},"colorscale":[["0","rgba(0,255,0,1)"],["0.0416666666666667","rgba(11,244,0,1)"],["0.0833333333333333","rgba(21,234,0,1)"],["0.125","rgba(32,223,0,1)"],["0.166666666666667","rgba(42,212,0,1)"],["0.208333333333333","rgba(53,202,0,1)"],["0.25","rgba(64,191,0,1)"],["0.291666666666667","rgba(74,181,0,1)"],["0.333333333333333","rgba(85,170,0,1)"],["0.375","rgba(96,159,0,1)"],["0.416666666666667","rgba(106,149,0,1)"],["0.458333333333333","rgba(117,138,0,1)"],["0.5","rgba(127,128,0,1)"],["0.541666666666667","rgba(138,117,0,1)"],["0.583333333333333","rgba(149,106,0,1)"],["0.625","rgba(159,96,0,1)"],["0.666666666666667","rgba(170,85,0,1)"],["0.708333333333333","rgba(181,74,0,1)"],["0.75","rgba(191,64,0,1)"],["0.791666666666667","rgba(202,53,0,1)"],["0.833333333333333","rgba(212,43,0,1)"],["0.875","rgba(223,32,0,1)"],["0.916666666666667","rgba(234,21,0,1)"],["0.958333333333333","rgba(244,11,0,1)"],["1","rgba(255,0,0,1)"]],"showscale":true,"x":["Valve","Heart","Ventricle","Atrium","Osteoblast","Adipocyte","Stromal cell","Adipose","Chondrocyte","Bone marrow","Hepatic stellate cell","Colon mucosa","Esophagus","Gastric epithelial cell","Colon","Liver","Hepatocyte","Small intestine","Beta cell","Intestinal epithelial cell","Pancreatic islet","Alpha cell","Kupffer cell","Microglia","Macrophage","Dendritic cell","Blymphocyte","Spleen","Tlymphocyte","Neutrophil","Thymus","Plasmacytoid dendritic cell","Thymocyte","Granulocyte","Plasma cell","Melanocyte","Hair follicle","Fibroblast","Skin","Keratinocyte","Basal cell","Airway smooth muscle","Vascular smooth muscle","Myofibroblast","Myoblast","Skeletal muscle","Medulla","Astrocyte","Retina","Oligodendrocyte","Spinal cord","Thalamus","Hypothalamus","Neuron","Motor neuron","Pons","Midbrain","Cerebellum","Sensory neuron","Cerebral cortex","Lung epithelial cell","Trachea","Lung","Podocyte","Mammary gland","Breast","Renal cortex","Kidney","Ovary","Testis","Granulosa","Oocyte"],"y":["CCND1","ERBB2","FGFR1","GATA3","MYC","PIK3CA","PTEN","TP53"],"z":[[12.0648,10.9207,10.8446,10.523400000000001,12.6214,11.7684,11.521000000000001,10.6913,10.6685,2.4196800000000001,12.7034,11.9095,11.4069,11.085699999999999,10.856299999999999,10.753500000000001,9.9708000000000006,9.0804500000000008,8.8523200000000006,8.4126200000000004,6.48238,6.3763100000000001,11.437200000000001,10.6508,9.4730000000000008,7.5060700000000002,5.6535700000000002,5.2597100000000001,4.5991099999999996,4.47438,4.3693600000000004,4.3131199999999996,4.2694700000000001,3.0561799999999999,0.113644,13.4353,12.513400000000001,12.416600000000001,11.8024,11.711,11.421099999999999,13.055300000000001,12.920299999999999,12.1812,11.925599999999999,7.5241100000000003,11.4955,11.344200000000001,10.5601,10.4015,9.7479899999999997,9.4297000000000004,9.4297000000000004,9.4048400000000001,8.9287500000000009,8.0427199999999992,7.7495399999999997,6.9254800000000003,6.5832100000000002,6.4102899999999998,12.165800000000001,11.787599999999999,11.462199999999999,12.5558,12.2423,11.914199999999999,10.2394,9.85717,9.4542699999999993,7.5683999999999996,4.6540299999999997,4.3868200000000002],[10.2692,9.3984699999999997,10.273899999999999,10.6142,10.146599999999999,9.2744099999999996,9.46706,8.4478799999999996,9.8323499999999999,7.3108500000000003,10.124499999999999,10.5663,9.4771800000000006,10.4719,10.841699999999999,9.7904400000000003,10.4771,9.1651600000000002,7.4385199999999996,10.248200000000001,8.1997699999999991,7.3108500000000003,10.5198,8.4512400000000003,5.98306,5.9674300000000002,7.3704799999999997,6.6467400000000003,6.5713100000000004,7.1483600000000003,7.0423799999999996,6.3141800000000003,7.13293,6.8102799999999997,7.1829599999999996,10.136900000000001,10.578799999999999,9.7924399999999991,10.190799999999999,10.2029,10.0199,10.432600000000001,10.073499999999999,10.323399999999999,9.6507799999999992,9.2034800000000008,9.3817599999999999,9.9847999999999999,8.7984799999999996,10.097099999999999,9.5164399999999993,9.5203000000000007,9.9596900000000002,8.7323199999999996,7.9858399999999996,8.1150000000000002,8.3779299999999992,7.0922700000000001,6.2248999999999999,8.7713199999999993,9.3227899999999995,10.2011,9.6198700000000006,10.0374,10.7765,10.5875,11.8561,9.7066700000000008,10.0396,8.5101800000000001,8.3609399999999994,6.4520200000000001],[11.494899999999999,10.283799999999999,11.083399999999999,11.279,11.2865,11.1631,10.811500000000001,11.8192,11.940200000000001,8.0816400000000002,11.635899999999999,10.649699999999999,8.5249100000000002,8.2421100000000003,8.6755300000000002,7.8329300000000002,7.2387100000000002,9.8526600000000002,10.4473,5.9979899999999997,10.0191,10.540699999999999,8.3689599999999995,9.29955,8.8127700000000004,8.1465300000000003,6.8853900000000001,8.0115800000000004,7.9008799999999999,6.61965,8.6386599999999998,4.0296099999999999,9.6008300000000002,7.2214600000000004,7.1003299999999996,10.7204,11.8367,11.446300000000001,10.6363,5.9359099999999998,7.2675799999999997,13.1328,11.0885,11.650499999999999,9.8328600000000002,9.46706,9.9665999999999997,11.184100000000001,9.8614899999999999,11.5328,10.9184,10.8344,10.967499999999999,10.5382,10.2463,7.4642600000000003,9.7124299999999995,11.4183,10.5512,10.453099999999999,9.0888500000000008,6.7873599999999996,9.7532300000000003,10.324999999999999,8.8013300000000001,9.1505700000000001,11.198399999999999,9.7830300000000001,10.7105,9.8774999999999995,10.014099999999999,9.6029800000000005],[5.5238399999999999,3.5682499999999999,4.0038799999999997,3.2851699999999999,1.2096800000000001,2.7759499999999999,0.113644,6.3010299999999999,3.2294299999999998,0.113644,2.9245700000000001,3.1149399999999998,4.4311100000000003,2.7759499999999999,3.7696100000000001,3.48123,3.0561799999999999,2.6169600000000002,0.113644,2.4196800000000001,0.113644,0.113644,3.2294299999999998,0.113644,2.4196800000000001,2.7759499999999999,4.0553400000000002,9.2616599999999991,9.4509299999999996,7.6995300000000002,9.0554299999999994,3.7700200000000001,11.226100000000001,4.0572499999999998,0.113644,0.113644,9.1523400000000006,1.81287,5.3428199999999997,8.2046600000000005,3.2236699999999998,1.2096800000000001,3.2851699999999999,1.81287,1.2096800000000001,3.8958400000000002,4.8804299999999996,2.4196800000000001,0.113644,5.91845,1.2096800000000001,4.63537,4.63537,1.2096800000000001,0.113644,7.3633100000000002,1.2096800000000001,0.113644,1.81287,0.113644,6.1461399999999999,1.81287,4.5603400000000001,3.2851699999999999,11.2203,10.764699999999999,7.5320200000000002,6.9902899999999999,3.8321800000000001,1.2096800000000001,3.52474,3.6447799999999999],[9.5263299999999997,8.2964599999999997,7.5097399999999999,8.1248199999999997,9.6937599999999993,9.6198700000000006,9.2215600000000002,7.0465799999999996,9.3426200000000001,9.1293699999999998,9.8964800000000004,8.7035400000000003,9.59131,10.5909,10.0093,9.4591899999999995,9.0884,7.82186,3.48123,6.8385499999999997,0.113644,0.113644,8.7527200000000001,6.9753999999999996,8.2757900000000006,8.53721,11.872299999999999,9.9686599999999999,10.7232,9.2448099999999993,9.5158400000000007,6.7123400000000002,8.7956500000000002,6.4439099999999998,9.6108499999999992,10.7722,9.3195399999999999,9.2270099999999999,9.5696100000000008,11.5671,10.775499999999999,8.80701,10.001200000000001,9.6650799999999997,9.3877600000000001,6.8452999999999999,8.0750200000000003,7.7964799999999999,5.3010799999999998,5.7600600000000002,6.49925,5.4302000000000001,5.7092400000000003,6.9753999999999996,6.8994900000000001,9.44557,7.2561600000000004,2.4196800000000001,8.2473799999999997,7.9199599999999997,10.478,10.671799999999999,9.9193599999999993,9.7766400000000004,10.1158,10.3813,8.1209799999999994,9.3474799999999991,9.4467599999999994,6.1387499999999999,9.7781300000000009,8.4035700000000002],[9.6330799999999996,8.9930099999999999,9.4407599999999992,8.0389199999999992,9.3847000000000005,10.108000000000001,9.2443799999999996,10.034000000000001,9.2190999999999992,8.4560600000000008,9.2154100000000003,9.1133299999999995,8.7901199999999999,8.8259100000000004,8.2606400000000004,8.8217499999999998,8.7294199999999993,8.6069499999999994,8.4888399999999997,8.5842399999999994,7.0553499999999998,7.9278000000000004,9.2331900000000005,9.2426200000000005,9.2034800000000008,9.2928300000000004,9.1458100000000009,9.5156299999999998,9.7224299999999992,9.3351799999999994,9.6424599999999998,7.9747599999999998,10.183,9.7064000000000004,8.5101800000000001,8.4462700000000002,9.2782800000000005,9.0671300000000006,8.9559999999999995,8.8924699999999994,9.1840399999999995,9.2950900000000001,9.1427600000000009,8.9765999999999995,9.2342399999999998,9.32958,9.2635699999999996,9.1790800000000008,9.1760099999999998,9.93459,9.3514099999999996,8.2503399999999996,8.2503399999999996,9.10398,8.9685699999999997,9.2175999999999991,8.4344000000000001,10.491400000000001,8.4243799999999993,9.5988500000000005,7.5343600000000004,8.9287500000000009,8.5424399999999991,8.7597799999999992,9.1519700000000004,8.9163399999999999,9.0857100000000006,8.3763799999999993,8.3716399999999993,8.4327100000000002,8.1715199999999992,8.6125299999999996],[10.757,10.9533,10.7224,10.427899999999999,10.545,11.4962,11.278700000000001,11.5793,10.9876,11.671799999999999,10.7044,10.6333,10.407500000000001,10.032500000000001,10.5238,10.8185,10.4681,10.925800000000001,10.7807,10.7614,11.765700000000001,11.4183,10.7713,10.8155,10.411199999999999,10.786799999999999,11.210000000000001,10.8254,10.5311,11.671799999999999,11.819699999999999,11.0747,11.826700000000001,12.3726,11.2209,10.4216,10.9674,10.804500000000001,10.867800000000001,9.9908400000000004,10.6044,10.9421,10.5678,10.1953,11.288600000000001,10.203799999999999,11.062200000000001,10.5945,10.8607,11.15,10.432499999999999,10.097099999999999,10.154299999999999,10.700799999999999,10.8833,11.269399999999999,10.5154,12.360300000000001,9.9449400000000008,10.2043,10.3035,10.101800000000001,10.4476,9.8508499999999994,10.8614,10.2399,10.797499999999999,10.2509,9.9159100000000002,10.828799999999999,10.1767,10.1387],[9.3990200000000002,9.0623400000000007,6.9304100000000002,7.5760699999999996,9.6022499999999997,8.6650100000000005,9.5024499999999996,8.4279499999999992,9.3241999999999994,9.8517499999999991,9.9527000000000001,9.2112300000000005,9.2383900000000008,9.9471000000000007,9.4560099999999991,8.6330299999999998,8.8100299999999994,8.1777300000000004,8.0035399999999992,8.4321599999999997,8.4008699999999994,7.67462,9.1559100000000004,9.2761899999999997,8.6006300000000007,9.1427600000000009,10.230700000000001,9.3877600000000001,9.5356699999999996,8.8242999999999991,10.0527,9.6650799999999997,9.9465299999999992,7.6972300000000002,9.2525399999999998,9.5987200000000001,9.5754300000000008,9.5108899999999998,9.4951000000000008,9.6947299999999998,8.8680699999999995,9.1427600000000009,9.8591099999999994,9.6818600000000004,9.5455400000000008,7.2994500000000002,9.1116600000000005,9.4018800000000002,8.3892399999999991,9.3094999999999999,7.6451200000000004,7.5889600000000002,7.5889600000000002,8.5399399999999996,8.6698599999999999,9.3273700000000002,7.3604700000000003,6.1613300000000004,7.4111000000000002,10.204800000000001,10.206799999999999,8.8720700000000008,9.3587799999999994,10.324999999999999,10.0382,10.108000000000001,7.82172,9.4591899999999995,9.0280500000000004,9.5074100000000001,8.7022600000000008,8.15151]],"type":"heatmap","xaxis":"x","yaxis":"y","frame":null}],"highlight":{"on":"plotly_click","persistent":false,"dynamic":false,"selectize":false,"opacityDim":0.20000000000000001,"selected":{"opacity":1},"debounce":0},"shinyEvents":["plotly_hover","plotly_click","plotly_selected","plotly_relayout","plotly_brushed","plotly_brushing","plotly_clickannotation","plotly_doubleclick","plotly_deselect","plotly_afterplot","plotly_sunburstclick"],"base_url":"https://plot.ly"},"evals":[],"jsHooks":[]}</script>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span>
Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 4.4.0 (2024-04-24)
Platform: x86_64-pc-linux-gnu
Running under: Ubuntu 22.04.4 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 
LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

time zone: Etc/UTC
tzcode source: system (glibc)

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] plotly_4.10.4   ggplot2_3.5.1   pheatmap_1.0.12 workflowr_1.7.1

loaded via a namespace (and not attached):
 [1] gtable_0.3.5       xfun_0.44          bslib_0.7.0        htmlwidgets_1.6.4 
 [5] processx_3.8.4     callr_3.7.6        tzdb_0.4.0         crosstalk_1.2.1   
 [9] vctrs_0.6.5        tools_4.4.0        ps_1.7.6           generics_0.1.3    
[13] parallel_4.4.0     tibble_3.2.1       fansi_1.0.6        highr_0.11        
[17] pkgconfig_2.0.3    data.table_1.15.4  RColorBrewer_1.1-3 lifecycle_1.0.4   
[21] compiler_4.4.0     farver_2.1.2       stringr_1.5.1      git2r_0.33.0      
[25] munsell_0.5.1      getPass_0.2-4      httpuv_1.6.15      htmltools_0.5.8.1 
[29] sass_0.4.9         yaml_2.3.8         lazyeval_0.2.2     later_1.3.2       
[33] pillar_1.9.0       crayon_1.5.2       jquerylib_0.1.4    whisker_0.4.1     
[37] tidyr_1.3.1        cachem_1.1.0       tidyselect_1.2.1   digest_0.6.35     
[41] stringi_1.8.4      dplyr_1.1.4        purrr_1.0.2        labeling_0.4.3    
[45] rprojroot_2.0.4    fastmap_1.2.0      grid_4.4.0         colorspace_2.1-0  
[49] cli_3.6.2          magrittr_2.0.3     utf8_1.2.4         readr_2.1.5       
[53] withr_3.0.0        scales_1.3.0       promises_1.3.0     bit64_4.0.5       
[57] rmarkdown_2.27     httr_1.4.7         bit_4.0.5          hms_1.1.3         
[61] evaluate_0.24.0    knitr_1.47         viridisLite_0.4.2  rlang_1.1.4       
[65] Rcpp_1.0.12        glue_1.7.0         rstudioapi_0.16.0  vroom_1.6.5       
[69] jsonlite_1.8.8     R6_2.5.1           fs_1.6.4          </code></pre>
</div>
</div>
</div>
</div>
</div>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
https://docs.mathjax.org/en/latest/web/configuration.html. This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>





</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
