<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />



<meta name="date" content="2024-08-07" />

<title>Generative models</title>

<script src="site_libs/header-attrs-2.27/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.13.2/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-6.4.2/css/all.min.css" rel="stylesheet" />
<link href="site_libs/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet" />

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/main/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">My blog</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/davetang/muse">
    <span class="fa fa-github"></span>
     
    Source code
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Generative models</h1>
<h4 class="date">2024-08-07</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span>
workflowr <span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2024-08-07
</p>
<p>
<strong>Checks:</strong> <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 7
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong> <code>muse/</code> <span
class="glyphicon glyphicon-question-sign" aria-hidden="true"
title="This is the local directory in which the code in this file was executed.">
</span>
</p>
<p>
This reproducible <a href="https://rmarkdown.rstudio.com">R Markdown</a>
analysis was created with <a
  href="https://github.com/workflowr/workflowr">workflowr</a> (version
1.7.1). The <em>Checks</em> tab describes the reproducibility checks
that were applied when the results were created. The <em>Past
versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date
</a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git
repository, you know the exact version of the code that produced these
results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the
global environment can affect the analysis in your R Markdown file in
unknown ways. For reproduciblity it’s best to always run the code in an
empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20200712code">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Seed:</strong>
<code>set.seed(20200712)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20200712code"
class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20200712)</code> was run prior to running
the code in the R Markdown file. Setting a seed ensures that any results
that rely on randomness, e.g. subsampling or permutations, are
reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Session information:</strong>
recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package
versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be
confident that you successfully produced the results during this
run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr
project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomdavetangmusetree66b58259988948e290217f7b20275af259c916fdtargetblank66b5825a">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Repository version:</strong>
<a href="https://github.com/davetang/muse/tree/66b58259988948e290217f7b20275af259c916fd" target="_blank">66b5825</a>
</a>
</p>
</div>
<div
id="strongRepositoryversionstrongahrefhttpsgithubcomdavetangmusetree66b58259988948e290217f7b20275af259c916fdtargetblank66b5825a"
class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development
and connecting the code version to the results is critical for
reproducibility.
</p>
<p>
The results in this page were generated with repository version
<a href="https://github.com/davetang/muse/tree/66b58259988948e290217f7b20275af259c916fd" target="_blank">66b5825</a>.
See the <em>Past versions</em> tab to see a history of the changes made
to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for
the analysis have been committed to Git prior to generating the results
(you can use <code>wflow_publish</code> or
<code>wflow_git_commit</code>). workflowr only checks the R Markdown
file, but you know if there are other scripts or data files that it
depends on. Below is the status of the Git repository when the results
were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .Rhistory
    Ignored:    .Rproj.user/
    Ignored:    analysis/figure/
    Ignored:    r_packages_4.3.3/
    Ignored:    r_packages_4.4.0/

Untracked files:
    Untracked:  data/e100.RData

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not
included in this status report because it is ok for generated content to
have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">
<p>
These are the previous versions of the repository in which changes were
made to the R Markdown (<code>analysis/generative_models.Rmd</code>) and
HTML (<code>docs/generative_models.html</code>) files. If you’ve
configured a remote Git repository (see <code>?wflow_git_remote</code>),
click on the hyperlinks in the table below to view the files as they
were in that past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/davetang/muse/blob/66b58259988948e290217f7b20275af259c916fd/analysis/generative_models.Rmd" target="_blank">66b5825</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-08-07
</td>
<td>
Epitope detection
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/davetang/muse/738aa36d5df7b41a6ab0824fae770678fd58a69a/docs/generative_models.html" target="_blank">738aa36</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-08-07
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/davetang/muse/blob/5cd88105e2b2568443a5b08377786ec50d5b6bc6/analysis/generative_models.Rmd" target="_blank">5cd8810</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-08-07
</td>
<td>
Binomial success counts
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/davetang/muse/4ec2479562fb0a2588e7ec42efb914478873cf6f/docs/generative_models.html" target="_blank">4ec2479</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-08-07
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/davetang/muse/blob/7f9058a5b42353dceb05074dbc2d6275d29952a9/analysis/generative_models.Rmd" target="_blank">7f9058a</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-08-07
</td>
<td>
Generative models
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<div id="discrete-example" class="section level2">
<h2>Discrete example</h2>
<p>If we know the rules (parameters) of a mechanism, then even if the
outcomes are random, we can generate probabilities of any event by using
computations and standard probability laws.</p>
<p>Consider that mutations along the genome of Human Immunodeficiency
Virus (HIV) occur at random with a rate of <span class="math inline">\(5
\times 10^{-4}\)</span> per nucleotide per replication cycle. This means
that after one cycle, the number of mutations in a genome of about <span
class="math inline">\(10^4\)</span> = 10,000 nucleotides will follow a
<a href="poisson.html">Poisson distribution</a> with rate 5.</p>
<p>This <strong>probability model</strong> predicts that the number of
mutations over one replication cycle will be close to 5 and that the
variability of this estimate is <span
class="math inline">\(\sqrt{5}\)</span> (the standard error). We now
have baseline reference values for both the number of mutations we
expect to see in a typical HIV strain and its variability.</p>
<p>If we want to know how often 3 mutations could occur under the
Poisson(5) model, we can use the <code>dpois()</code> function to
generate the probability of seeing x = 3 events, taking the value of the
<em>rate parameter</em> of the Poisson distribution, called lambda
(<span class="math inline">\(\lambda\)</span>), to be 5.</p>
<pre class="r"><code>dpois(x = 3, lambda = 5)</code></pre>
<pre><code>[1] 0.1403739</code></pre>
<p>The output above says that the chance of seeing exactly three events
is around 0.14 or about 1 in 7.</p>
<p>Probabilities of a range of values.</p>
<pre class="r"><code>dpois(x = 0:12, lambda = 5) |&gt;
  barplot(names.arg = 0:12, col = 2)</code></pre>
<p><img src="figure/generative_models.Rmd/dpois_0_to_12-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-dpois_0_to_12-1">
Past versions of dpois_0_to_12-1.png
</button>
</p>
<div id="fig-dpois_0_to_12-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/davetang/muse/blob/4ec2479562fb0a2588e7ec42efb914478873cf6f/docs/figure/generative_models.Rmd/dpois_0_to_12-1.png" target="_blank">4ec2479</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-08-07
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>Mathematical theory tells us that the Poisson probability of seeing
<span class="math inline">\(x\)</span> is given by:</p>
<p><span class="math display">\[
p = \frac{e^{-\lambda} \lambda^x}{x!}
\]</span></p>
<pre class="r"><code>my_dpois &lt;- function(x, lambda){
  e &lt;- exp(1)
  ((e^-lambda)*(lambda^x))/factorial(x)
}

my_dpois(3, 5)</code></pre>
<pre><code>[1] 0.1403739</code></pre>
<pre class="r"><code>dpois(3, 5)</code></pre>
<pre><code>[1] 0.1403739</code></pre>
<p>The Poisson distribution is a good model for rare events such as
mutations. Other useful probability models for discrete events are the
Bernoulli, binomial, and multinomial distributions.</p>
<div id="using-discrete-probability-models" class="section level3">
<h3>Using discrete probability models</h3>
<p>A point mutation can either occur or not; it is a binary event. The
two possible outcomes (yes, no) are called the <strong>levels</strong>
of the categorical variable. However, not all events are binary such as
the genotypes in a diploid organism, which can take three levels:
<code>AA</code>, <code>Aa</code>, and <code>aa</code>.</p>
<p>Sometimes the number of levels in a categorical variable is very
large; examples include the number of different types of bacteria in a
biological sample (hundreds or thousands) and the number of codons
formed of three nucleotides (64 levels).</p>
<p>Tossing a coin has two possible outcomes and this simple experiment
is called a Bernoulli trial; this is modeled using a so-called Bernoulli
random variable. Bernoulli trials can be used to build more complex
models.</p>
<p>We can use the <code>rbinom()</code> function (<code>r</code> for
random and <code>binom</code> for binomial) to generate some random
events that follow a binomial distribution. Below we simulate a sequence
of 15 fair coin tosses. For <code>rbinom()</code> we have specified 15
trials (<code>n = 15</code>), where each individual trial consists of
just one single toss (<code>size = 1</code>), and the probability of
success is 50/50 (<code>prob = 0.5</code>).</p>
<pre class="r"><code>set.seed(1984)
rbinom(n = 15, size = 1, prob = 0.5)</code></pre>
<pre><code> [1] 1 0 0 0 1 1 0 0 1 0 1 0 1 1 0</code></pre>
<p>Success and failure can have unequal probabilities in a Bernoulli
trial, as long as the probabilities sum to one, i.e., complementary
events. To simulate 12 trials with unequal probabilities, we just use a
different <code>prob</code>. The <code>1</code>’s indicate success and
<code>0</code>’s failure.</p>
<pre class="r"><code>set.seed(1984)
rbinom(n = 12, size = 1, prob = 2/3)</code></pre>
<pre><code> [1] 1 1 1 1 0 0 1 1 0 1 0 1</code></pre>
</div>
<div id="binomial-success-counts" class="section level3">
<h3>Binomial success counts</h3>
<p>If we only care about successes, then the order doesn’t matter and we
can just sum the <code>1</code>’s. We can get just the successes by
setting <code>n = 1</code> and <code>size</code> to the number of
trials. The number of successes below is close to the specified
probability.</p>
<pre class="r"><code>set.seed(1984)
rbinom(n = 1, size = 100, prob = 2/3)</code></pre>
<pre><code>[1] 62</code></pre>
<p>When there are only two possible outcomes, such as heads or tails, we
only need to specify the probability, <span
class="math inline">\(p\)</span>, of “success” since the probability of
“failure” is <span class="math inline">\(1 - p\)</span>.</p>
<p>The number of successes in 15 Bernoulli trials with a probability of
success of 0.3 is called a <strong>binomial</strong> random variable or
a random variable that follows the <span
class="math inline">\(B\)</span>(15,0.3) distribution. If we replicate
trial 100 times, we will see that the most frequent value is 4.</p>
<pre class="r"><code>set.seed(1984)
replicate(
  n = 100,
  rbinom(n = 1, prob = 0.3, size = 15)
) |&gt;
  table()</code></pre>
<pre><code>
 0  1  2  3  4  5  6  7  8  9 
 1  4  6 16 22 19 17 10  3  2 </code></pre>
<p>The complete <strong>probability mass distribution</strong> is
outputted using the <code>dbinom()</code> function:</p>
<pre class="r"><code>dbinom(0:15, prob = 0.3, size = 15) |&gt;
  barplot(names.arg = 0:15, col = 2)</code></pre>
<p><img src="figure/generative_models.Rmd/plot_dbinom-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-plot_dbinom-1">
Past versions of plot_dbinom-1.png
</button>
</p>
<div id="fig-plot_dbinom-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/davetang/muse/blob/738aa36d5df7b41a6ab0824fae770678fd58a69a/docs/figure/generative_models.Rmd/plot_dbinom-1.png" target="_blank">738aa36</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-08-07
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>The number of trials is the number we input to R as <code>size</code>
and is often written <span class="math inline">\(n\)</span>, while the
probability of success is <span class="math inline">\(p\)</span>.
Mathematical theory tells us that for <span
class="math inline">\(X\)</span> distributed as a binomial distribution
with parameters <span class="math inline">\((n,p)\)</span>, written
<span class="math inline">\(X \sim B(n,p)\)</span>, the probability of
seeing <span class="math inline">\(X = k\)</span> successes is</p>
<p><span class="math display">\[
P(X = k) = \binom{n}{k} p^k (1-p)^{n-k}
\]</span></p>
<pre class="r"><code>my_dbinom &lt;- function(k, n, p){
  choose(n, k) * p^k * (1-p)^(n-k)
}

my_dbinom(4, 15, 0.3)</code></pre>
<pre><code>[1] 0.2186231</code></pre>
<pre class="r"><code>dbinom(x = 4, size = 15, prob = 0.3)</code></pre>
<pre><code>[1] 0.2186231</code></pre>
</div>
<div id="epitope-detection" class="section level3">
<h3>Epitope detection</h3>
<p>When testing certain pharmaceutical compounds, it is important to
detect proteins that provoke an allergic reaction. The molecular sites
that are responsible for such reactions are called
<strong>epitopes</strong>:</p>
<blockquote>
<p>A specific portion of a macromolecular antigen to which an antibody
binds. In the case of a protein antigen recognised by a T cell, the
epitope or determinant is the peptide portion or site that binds to a
major histocompatibility complex (MHC) molecule for recognition by the
T-cell receptor (TCR)</p>
</blockquote>
<p>An <a href="antibody.html">antibody</a> is a type of protein made by
certain white blood cells in response to a foreign substance in the
body, which is called an antigen. An antibody binds (with more or less
specificity) to its antigen. The purpose of the binding is to help
destroy the antigen.</p>
<p>Antibodies can work in several ways, depending on the nature of the
antigen. Some antibodies destroy antigens directly while others help
recruit white blood cells to destroy the antigen. An epitope, also known
as antigenic determinant, is the part of an antigen that is recognised
by the immune system, specifically by antibodies, B cells or T
cells.</p>
<p>ELISA assays are used to detect specific epitopes at different
positions along a protein. Suppose the following facts hold for an ELISA
assay:</p>
<ul>
<li>The baseline noise level per position, or more precisely the false
positive rate, is 1%. This is the probability of declaring a hit - we
think we have an epitope - when there is none. We write this as <span
class="math inline">\(P\)</span>(declare epitope | no epitope).</li>
<li>The protein is tested at 100 different positions, supposed to be
independent.</li>
</ul>
<p>Below is data for 50 patients tallied at each of the 100 positions.
If there are no allergic reactions, the false positive rate of 1% means
that for a single patient, each individual position has a probability of
1 in 100 of being a 1. After tallying 50 patients, we expect at any
given position the sum of the 50 observed 0/1 variables to have a
Poisson distribution with paramater 0.5 (50 * 1/100).</p>
<pre class="r"><code>load(&quot;data/e100.RData&quot;)
barplot(e100, names.arg = 1:100, col = 2)</code></pre>
<p><img src="figure/generative_models.Rmd/e100-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>What are the chances of seeing a value as large as 7, if no epitope
is present?</p>
<p>If we look for the probability of seeing a number as big as 7 (or
larger) when considering one Poisson(0.5) random variable, the answer
can be calculated in closed form as:</p>
<p><span class="math display">\[
P(X \ge 7) = \sum^\infty_{k=7} P(X = k)
\]</span></p>
<p>This is the same as <span class="math inline">\(1 - P(X \le
6)\)</span> and the probability <span class="math inline">\(P(X \le
6)\)</span> is the so-called cumulative distribution function at 6 and
the <code>ppois()</code> function is used this calculate this.</p>
<pre class="r"><code>1 - ppois(6, 0.5)</code></pre>
<pre><code>[1] 1.00238e-06</code></pre>
<p>We denote this number by <span
class="math inline">\(\epsilon\)</span> and have shown that the
probability of seeing a count as large as 7, assuming no epitope
reactions, is:</p>
<p><span class="math display">\[
\epsilon = P(X \ge 7) = 1 - P(X \le 6) \approx 10^{-6}
\]</span></p>
<p>However, the above calculation is not the correct computation.</p>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span>
Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 4.4.0 (2024-04-24)
Platform: x86_64-pc-linux-gnu
Running under: Ubuntu 22.04.4 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 
LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

time zone: Etc/UTC
tzcode source: system (glibc)

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] lubridate_1.9.3 forcats_1.0.0   stringr_1.5.1   dplyr_1.1.4    
 [5] purrr_1.0.2     readr_2.1.5     tidyr_1.3.1     tibble_3.2.1   
 [9] ggplot2_3.5.1   tidyverse_2.0.0 workflowr_1.7.1

loaded via a namespace (and not attached):
 [1] sass_0.4.9        utf8_1.2.4        generics_0.1.3    stringi_1.8.4    
 [5] hms_1.1.3         digest_0.6.35     magrittr_2.0.3    timechange_0.3.0 
 [9] evaluate_0.24.0   grid_4.4.0        fastmap_1.2.0     rprojroot_2.0.4  
[13] jsonlite_1.8.8    processx_3.8.4    whisker_0.4.1     ps_1.7.6         
[17] promises_1.3.0    httr_1.4.7        fansi_1.0.6       scales_1.3.0     
[21] jquerylib_0.1.4   cli_3.6.2         rlang_1.1.4       munsell_0.5.1    
[25] withr_3.0.0       cachem_1.1.0      yaml_2.3.8        tools_4.4.0      
[29] tzdb_0.4.0        colorspace_2.1-0  httpuv_1.6.15     vctrs_0.6.5      
[33] R6_2.5.1          lifecycle_1.0.4   git2r_0.33.0      fs_1.6.4         
[37] pkgconfig_2.0.3   callr_3.7.6       pillar_1.9.0      bslib_0.7.0      
[41] later_1.3.2       gtable_0.3.5      glue_1.7.0        Rcpp_1.0.12      
[45] highr_0.11        xfun_0.44         tidyselect_1.2.1  rstudioapi_0.16.0
[49] knitr_1.47        htmltools_0.5.8.1 rmarkdown_2.27    compiler_4.4.0   
[53] getPass_0.2-4    </code></pre>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
https://docs.mathjax.org/en/latest/web/configuration.html. This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>




</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
