<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />



<meta name="date" content="2025-02-17" />

<title>Getting started with harmony</title>

<script src="site_libs/header-attrs-2.28/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.13.2/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-6.4.2/css/all.min.css" rel="stylesheet" />
<link href="site_libs/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet" />

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/main/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">My blog</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/davetang/muse">
    <span class="fa fa-github"></span>
     
    Source code
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Getting started with harmony</h1>
<h4 class="date">2025-02-17</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span>
workflowr <span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2025-02-17
</p>
<p>
<strong>Checks:</strong> <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 7
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong> <code>muse/</code> <span
class="glyphicon glyphicon-question-sign" aria-hidden="true"
title="This is the local directory in which the code in this file was executed.">
</span>
</p>
<p>
This reproducible <a href="https://rmarkdown.rstudio.com">R Markdown</a>
analysis was created with <a
  href="https://github.com/workflowr/workflowr">workflowr</a> (version
1.7.1). The <em>Checks</em> tab describes the reproducibility checks
that were applied when the results were created. The <em>Past
versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date
</a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git
repository, you know the exact version of the code that produced these
results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the
global environment can affect the analysis in your R Markdown file in
unknown ways. For reproduciblity it’s best to always run the code in an
empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20200712code">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Seed:</strong>
<code>set.seed(20200712)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20200712code"
class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20200712)</code> was run prior to running
the code in the R Markdown file. Setting a seed ensures that any results
that rely on randomness, e.g. subsampling or permutations, are
reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Session information:</strong>
recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package
versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be
confident that you successfully produced the results during this
run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr
project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomdavetangmusetreeabc7ef459384e79a70bb29a7f0b148381f824c9ctargetblankabc7ef4a">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Repository version:</strong>
<a href="https://github.com/davetang/muse/tree/abc7ef459384e79a70bb29a7f0b148381f824c9c" target="_blank">abc7ef4</a>
</a>
</p>
</div>
<div
id="strongRepositoryversionstrongahrefhttpsgithubcomdavetangmusetreeabc7ef459384e79a70bb29a7f0b148381f824c9ctargetblankabc7ef4a"
class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development
and connecting the code version to the results is critical for
reproducibility.
</p>
<p>
The results in this page were generated with repository version
<a href="https://github.com/davetang/muse/tree/abc7ef459384e79a70bb29a7f0b148381f824c9c" target="_blank">abc7ef4</a>.
See the <em>Past versions</em> tab to see a history of the changes made
to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for
the analysis have been committed to Git prior to generating the results
(you can use <code>wflow_publish</code> or
<code>wflow_git_commit</code>). workflowr only checks the R Markdown
file, but you know if there are other scripts or data files that it
depends on. Below is the status of the Git repository when the results
were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .Rproj.user/
    Ignored:    data/1M_neurons_filtered_gene_bc_matrices_h5.h5
    Ignored:    data/293t_3t3_filtered_gene_bc_matrices.tar.gz
    Ignored:    data/293t_filtered_gene_bc_matrices.tar.gz
    Ignored:    data/5k_Human_Donor1_PBMC_3p_gem-x_5k_Human_Donor1_PBMC_3p_gem-x_count_sample_filtered_feature_bc_matrix.h5
    Ignored:    data/5k_Human_Donor2_PBMC_3p_gem-x_5k_Human_Donor2_PBMC_3p_gem-x_count_sample_filtered_feature_bc_matrix.h5
    Ignored:    data/5k_Human_Donor3_PBMC_3p_gem-x_5k_Human_Donor3_PBMC_3p_gem-x_count_sample_filtered_feature_bc_matrix.h5
    Ignored:    data/5k_Human_Donor4_PBMC_3p_gem-x_5k_Human_Donor4_PBMC_3p_gem-x_count_sample_filtered_feature_bc_matrix.h5
    Ignored:    data/Parent_SC3v3_Human_Glioblastoma_filtered_feature_bc_matrix.tar.gz
    Ignored:    data/brain_counts/
    Ignored:    data/cl.obo
    Ignored:    data/cl.owl
    Ignored:    data/jurkat:293t_50:50_filtered_gene_bc_matrices.tar.gz
    Ignored:    data/jurkat_filtered_gene_bc_matrices.tar.gz
    Ignored:    data/pbmc4k_filtered_gene_bc_matrices.tar.gz
    Ignored:    data/refdata-gex-GRCh38-2020-A.tar.gz
    Ignored:    data/seurat_1m_neuron.rds
    Ignored:    data/t_3k_filtered_gene_bc_matrices.tar.gz
    Ignored:    r_packages_4.4.1/

Untracked files:
    Untracked:  data/293t/
    Untracked:  data/jurkat/
    Untracked:  data/jurkat_293t/

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not
included in this status report because it is ok for generated content to
have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">
<p>
These are the previous versions of the repository in which changes were
made to the R Markdown (<code>analysis/harmony.Rmd</code>) and HTML
(<code>docs/harmony.html</code>) files. If you’ve configured a remote
Git repository (see <code>?wflow_git_remote</code>), click on the
hyperlinks in the table below to view the files as they were in that
past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/davetang/muse/blob/abc7ef459384e79a70bb29a7f0b148381f824c9c/analysis/harmony.Rmd" target="_blank">abc7ef4</a>
</td>
<td>
Dave Tang
</td>
<td>
2025-02-17
</td>
<td>
Increase memory for future
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/davetang/muse/4db2bdfea444a1eb980089c4808a54e461d5f4e5/docs/harmony.html" target="_blank">4db2bdf</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-04-15
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/davetang/muse/blob/9a4bcb8fd5157fd2656b4cfa6e73639185716978/analysis/harmony.Rmd" target="_blank">9a4bcb8</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-04-15
</td>
<td>
Using 10x filtered matrices
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/davetang/muse/a940a7e352aacdb1cf2283d0518d59c58bed8bc9/docs/harmony.html" target="_blank">a940a7e</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-04-15
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/davetang/muse/blob/03f3f3b183f3b7cc7db4cff53bb43f35707fd3a1/analysis/harmony.Rmd" target="_blank">03f3f3b</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-04-15
</td>
<td>
Quieter
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/davetang/muse/b174605ed41929903234d582b174888e56ed5de6/docs/harmony.html" target="_blank">b174605</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-04-15
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/davetang/muse/blob/6c16dce83b8942f9d10c15a28de506f8eca86bd4/analysis/harmony.Rmd" target="_blank">6c16dce</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-04-15
</td>
<td>
Seurat SCTransform workflow with harmony
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/davetang/muse/5466912f6b18ca31f06c0fdc52983a37419b5bc2/docs/harmony.html" target="_blank">5466912</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-04-15
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/davetang/muse/blob/3ec33679dd70ebc29333a10f971660b918a0420e/analysis/harmony.Rmd" target="_blank">3ec3367</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-04-15
</td>
<td>
Seurat SCTransform workflow
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/davetang/muse/c3d7314fea2e83d5bf86f0932c6be6bee6ecbd13/docs/harmony.html" target="_blank">c3d7314</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-04-15
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/davetang/muse/blob/708f1ab53a05e899ce6a1512929dfa5ac9ac346b/analysis/harmony.Rmd" target="_blank">708f1ab</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-04-15
</td>
<td>
Using harmony with Seurat
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/davetang/muse/085be81db7542032f44afdd8690c1a38cf5cc2c9/docs/harmony.html" target="_blank">085be81</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-04-14
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/davetang/muse/blob/95fa6fdcfe7dd694a9c5ac6192b6a92b46101950/analysis/harmony.Rmd" target="_blank">95fa6fd</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-04-14
</td>
<td>
Create some plots
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/davetang/muse/9abb7b6548fe687ab4143fed5109ff9b47ac0d2c/docs/harmony.html" target="_blank">9abb7b6</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-04-14
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/davetang/muse/blob/72ffea914f6fbdb9d1648bb7adc93b0a5b7e0fa1/analysis/harmony.Rmd" target="_blank">72ffea9</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-04-14
</td>
<td>
Getting started with harmony
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<div id="quickstart" class="section level2">
<h2>Quickstart</h2>
<p>Follow the <a
href="https://htmlpreview.github.io/?https://github.com/immunogenomics/harmony/blob/master/doc/quickstart.html">quickstart
tutorial</a></p>
<pre class="r"><code>install.packages(&quot;harmony&quot;)</code></pre>
<p>{harmony} version.</p>
<pre class="r"><code>packageVersion(&quot;harmony&quot;)</code></pre>
<pre><code>[1] &#39;1.2.1&#39;</code></pre>
<div id="data" class="section level3">
<h3>Data</h3>
<p>We library normalized the cells, log transformed the counts, and
scaled the genes. Then we performed PCA and kept the top 20 PCs. The PCA
embeddings and meta data are available as part of this package.</p>
<pre class="r"><code>data(cell_lines)
V &lt;- cell_lines$scaled_pcs
meta_data &lt;- cell_lines$meta_data

str(cell_lines)</code></pre>
<pre><code>List of 2
 $ meta_data : tibble [2,370 × 5] (S3: tbl_df/tbl/data.frame)
  ..$ cell_id     : chr [1:2370] &quot;half_GTACGAACCACCAA&quot; &quot;t293_AGGTCATGCACTTT&quot; &quot;half_ATAGTTGACTTCTA&quot; &quot;half_GAGCGGCTTGCTTT&quot; ...
  ..$ dataset     : chr [1:2370] &quot;half&quot; &quot;t293&quot; &quot;half&quot; &quot;half&quot; ...
  ..$ nGene       : int [1:2370] 1508 4009 3545 2450 2388 3762 3792 4089 3374 3023 ...
  ..$ percent_mito: num [1:2370] 0.0148 0.0232 0.0153 0.017 0.0601 ...
  ..$ cell_type   : chr [1:2370] &quot;jurkat&quot; &quot;t293&quot; &quot;jurkat&quot; &quot;jurkat&quot; ...
  ..- attr(*, &quot;.internal.selfref&quot;)=&lt;externalptr&gt; 
 $ scaled_pcs:Classes &#39;data.table&#39; and &#39;data.frame&#39;:    2370 obs. of  20 variables:
  ..$ X1 : num [1:2370] 0.00281 -0.01167 0.00933 0.00634 0.00855 ...
  ..$ X2 : num [1:2370] -0.00145 0.000877 -0.006972 -0.002518 0.007087 ...
  ..$ X3 : num [1:2370] -0.00639 0.000897 -0.002599 -0.00439 -0.002254 ...
  ..$ X4 : num [1:2370] 0.000282 0.001324 0.001882 0.000274 0.001679 ...
  ..$ X5 : num [1:2370] 0.00144 -0.00329 -0.0038 -0.0025 0.00455 ...
  ..$ X6 : num [1:2370] 0.000752 0.001303 -0.000347 0.000435 0.0003 ...
  ..$ X7 : num [1:2370] -0.00283 -0.00198 -0.00157 0.00136 -0.0016 ...
  ..$ X8 : num [1:2370] -0.000653 0.001625 -0.003272 -0.00263 -0.000263 ...
  ..$ X9 : num [1:2370] 0.001411 -0.000913 -0.001031 -0.001876 0.001389 ...
  ..$ X10: num [1:2370] -0.000417 -0.000175 -0.001623 -0.000425 0.000391 ...
  ..$ X11: num [1:2370] 0.001652 -0.000034 0.001241 -0.000458 -0.001444 ...
  ..$ X12: num [1:2370] 6.71e-05 3.76e-04 -7.61e-04 -6.52e-04 -2.44e-03 ...
  ..$ X13: num [1:2370] 0.000542 0.000219 -0.001502 -0.002067 -0.000907 ...
  ..$ X14: num [1:2370] 0.001223 0.001688 -0.000279 -0.000927 -0.000135 ...
  ..$ X15: num [1:2370] 0.002081 0.000386 -0.001141 0.001114 0.001015 ...
  ..$ X16: num [1:2370] 1.87e-03 -1.50e-03 5.99e-04 -1.98e-05 -1.25e-03 ...
  ..$ X17: num [1:2370] 0.000429 0.000259 0.001224 -0.001069 -0.001165 ...
  ..$ X18: num [1:2370] 0.00115 -0.00106 0.00145 0.00028 0.00111 ...
  ..$ X19: num [1:2370] -1.09e-03 4.11e-04 7.41e-05 9.33e-04 -1.76e-04 ...
  ..$ X20: num [1:2370] 0.000265 -0.00171 -0.000662 0.000365 0.000477 ...
  ..- attr(*, &quot;.internal.selfref&quot;)=&lt;externalptr&gt; </code></pre>
<p>Table of cell types.</p>
<pre class="r"><code>table(cell_lines$meta_data$cell_type)</code></pre>
<pre><code>
jurkat   t293 
  1266   1104 </code></pre>
<p>Table of datasets.</p>
<pre class="r"><code>table(cell_lines$meta_data$dataset)</code></pre>
<pre><code>
  half jurkat   t293 
   846    824    700 </code></pre>
</div>
<div id="analysis" class="section level3">
<h3>Analysis</h3>
<p>Initially, the cells cluster by both dataset (left) and cell type
(right). The quickstart guide uses the <code>do_scatter()</code>
function, which is <a
href="https://github.com/immunogenomics/harmony/issues/239">missing</a>.
We can simply plot the first two PCs using {ggplot2}.</p>
<p>Plot PC1 versus PC2.</p>
<pre class="r"><code>my_df &lt;- data.frame(PC1 = V$X1, PC2 = V$X2, dataset = meta_data$dataset, cell_type = meta_data$cell_type)

ggplot(my_df, aes(PC1, PC2, colour = dataset)) +
  geom_point() +
  theme_minimal() +
  ggtitle(&quot;Before harmony&quot;) -&gt; p1

ggplot(my_df, aes(PC1, PC2, colour = cell_type)) +
  geom_point() +
  theme_minimal() -&gt; p2

p1 + p2</code></pre>
<p><img src="figure/harmony.Rmd/plot_pc1_pc2-1.png" width="960" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-plot_pc1_pc2-1">
Past versions of plot_pc1_pc2-1.png
</button>
</p>
<div id="fig-plot_pc1_pc2-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/davetang/muse/blob/085be81db7542032f44afdd8690c1a38cf5cc2c9/docs/figure/harmony.Rmd/plot_pc1_pc2-1.png" target="_blank">085be81</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-04-14
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>Let’s run Harmony to remove the influence of dataset-of-origin from
the cell embeddings.</p>
<pre class="r"><code>harmony_embeddings &lt;- harmony::RunHarmony(
    V, meta_data, &#39;dataset&#39;, verbose=FALSE
)

my_df2 &lt;- data.frame(PC1 = harmony_embeddings[, 1], PC2 = harmony_embeddings[, 2], dataset = meta_data$dataset, cell_type = meta_data$cell_type)

ggplot(my_df2, aes(PC1, PC2, colour = dataset)) +
  geom_point() +
  theme_minimal() +
  ggtitle(&quot;After harmony&quot;) -&gt; p1

ggplot(my_df2, aes(PC1, PC2, colour = cell_type)) +
  geom_point() +
  theme_minimal() -&gt; p2

p1 + p2</code></pre>
<p><img src="figure/harmony.Rmd/run_harmony_quickstart-1.png" width="960" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-run_harmony_quickstart-1">
Past versions of run_harmony_quickstart-1.png
</button>
</p>
<div id="fig-run_harmony_quickstart-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/davetang/muse/blob/c3d7314fea2e83d5bf86f0932c6be6bee6ecbd13/docs/figure/harmony.Rmd/run_harmony_quickstart-1.png" target="_blank">c3d7314</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-04-15
</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
</div>
<div id="using-harmony-with-seurat" class="section level2">
<h2>Using harmony with Seurat</h2>
<p>Following the <a
href="https://github.com/immunogenomics/harmony/blob/master/doc/Seurat.Rmd">Using
harmony with Seurat</a> tutorial, which describes how to use harmony in
Seurat v5 single-cell analysis workflows. Also, it will provide some
basic downstream analyses demonstrating the properties of harmonised
cell embeddings and a brief explanation of the exposed algorithm
parameters.</p>
<div id="data-1" class="section level3">
<h3>Data</h3>
<p>For this demo, we will be aligning two groups of PBMCs <a
href="https://doi.org/10.1038/nbt.4042">Kang et al., 2017</a>:</p>
<ul>
<li>Control PBMCs</li>
<li>Stimulated PBMCs treated with interferon beta.</li>
</ul>
<pre class="r"><code>data(&quot;pbmc_stim&quot;)
str(pbmc.ctrl)</code></pre>
<pre><code>Formal class &#39;dgCMatrix&#39; [package &quot;Matrix&quot;] with 6 slots
  ..@ i       : int [1:558407] 1 3 8 19 33 40 51 53 89 100 ...
  ..@ p       : int [1:1001] 0 812 1248 1692 2463 2810 3314 4127 4660 5229 ...
  ..@ Dim     : int [1:2] 9015 1000
  ..@ Dimnames:List of 2
  .. ..$ : chr [1:9015] &quot;LINC00115&quot; &quot;NOC2L&quot; &quot;HES4&quot; &quot;ISG15&quot; ...
  .. ..$ : chr [1:1000] &quot;TCGCAAGAGCGATT-1&quot; &quot;GCAAACTGAACTGC-1&quot; &quot;TGATAAACTGGTAC-1&quot; &quot;GTTAAATGACTGTG-1&quot; ...
  ..@ x       : num [1:558407] 1 1 1 1 1 2 1 3 1 2 ...
  ..@ factors : list()</code></pre>
<pre class="r"><code>str(pbmc.stim)</code></pre>
<pre><code>Formal class &#39;dgCMatrix&#39; [package &quot;Matrix&quot;] with 6 slots
  ..@ i       : int [1:571399] 3 33 53 118 128 138 144 154 171 208 ...
  ..@ p       : int [1:1001] 0 425 505 852 1403 2010 2325 2859 3435 3955 ...
  ..@ Dim     : int [1:2] 9015 1000
  ..@ Dimnames:List of 2
  .. ..$ : chr [1:9015] &quot;LINC00115&quot; &quot;NOC2L&quot; &quot;HES4&quot; &quot;ISG15&quot; ...
  .. ..$ : chr [1:1000] &quot;ATCACTTGCTCGAA-1&quot; &quot;CCGGAGACTGTGAC-1&quot; &quot;CAAGCCCTGTTAGC-1&quot; &quot;GAGGTACTAACGGG-1&quot; ...
  ..@ x       : num [1:571399] 1 2 3 1 8 1 1 1 3 1 ...
  ..@ factors : list()</code></pre>
<p>The full dataset used for this vignette have been upload to <a
href="https://zenodo.org/record/8164711">Zenodo</a> but currently <a
href="https://github.com/immunogenomics/harmony/issues/248#issuecomment-2051755519">does
not work</a> with newer versions of R.</p>
</div>
<div id="create-seurat-object" class="section level3">
<h3>Create Seurat object</h3>
<p>Create a Seurat object with treatment conditions in the metadata.</p>
<pre class="r"><code>pbmc &lt;- CreateSeuratObject(
  counts = cbind(pbmc.stim, pbmc.ctrl),
  project = &quot;Kang&quot;,
  min.cells = 5
)

pbmc@meta.data$stim &lt;- c(rep(&quot;STIM&quot;, ncol(pbmc.stim)), rep(&quot;CTRL&quot;, ncol(pbmc.ctrl)))
pbmc</code></pre>
<pre><code>An object of class Seurat 
9015 features across 2000 samples within 1 assay 
Active assay: RNA (9015 features, 0 variable features)
 1 layer present: counts</code></pre>
</div>
<div id="seurat-sctransform-workflow" class="section level3">
<h3>Seurat SCTransform workflow</h3>
<p><a
href="https://satijalab.org/seurat/articles/sctransform_vignette.html">Using
sctransform in Seurat</a>.</p>
<pre class="r"><code>pbmc_sct &lt;- SCTransform(pbmc) |&gt;
  RunPCA() |&gt;
  FindNeighbors() |&gt;
  RunUMAP(dims = 1:20) |&gt;
  FindClusters()</code></pre>
<pre><code>Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck

Number of nodes: 2000
Number of edges: 64411

Running Louvain algorithm...
Maximum modularity in 10 random starts: 0.8846
Number of communities: 15
Elapsed time: 0 seconds</code></pre>
<pre class="r"><code>DimPlot(pbmc_sct, reduction = &quot;umap&quot;, group.by = &quot;stim&quot;, pt.size = .1) + ggtitle(&quot;Seurat SCTransform workflow&quot;)</code></pre>
<p><img src="figure/harmony.Rmd/sctransform_workflow-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-sctransform_workflow-1">
Past versions of sctransform_workflow-1.png
</button>
</p>
<div id="fig-sctransform_workflow-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/davetang/muse/blob/5466912f6b18ca31f06c0fdc52983a37419b5bc2/docs/figure/harmony.Rmd/sctransform_workflow-1.png" target="_blank">5466912</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-04-15
</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<div id="seurat-workflow-with-harmony" class="section level3">
<h3>Seurat workflow with harmony</h3>
<p>Harmony works on an existing matrix with cell embeddings and outputs
its transformed version with the datasets aligned according to some
user-defined experimental conditions. By default, harmony will look up
the <code>pca</code> cell embeddings and use these to run harmony.
Therefore, it assumes that the Seurat object has these embeddings
already precomputed.</p>
<p>We will run the Seurat workflow to generate the embeddings.</p>
<p>Here, using <code>Seurat::NormalizeData()</code>, we will be
generating a union of highly variable genes using each condition (the
control and stimulated cells). These features are going to be
subsequently used to generate the 20 PCs with
<code>Seurat::RunPCA()</code>.</p>
<p>Note that the defaults for <code>NormalizeData</code> are:</p>
<ul>
<li><code>normalization.method</code> = “LogNormalize”</li>
<li><code>scale.factor</code> = 10000</li>
</ul>
<pre class="r"><code>pbmc &lt;- NormalizeData(pbmc, verbose = FALSE)

pbmc &lt;- FindVariableFeatures(object = pbmc, selection.method = &quot;vst&quot;, nfeatures = 2000)</code></pre>
<pre><code>Finding variable features for layer counts</code></pre>
<pre class="r"><code>cell_by_cond &lt;- split(row.names(pbmc@meta.data), pbmc@meta.data$stim)

vfeatures &lt;- lapply(cell_by_cond, function(cells){
  FindVariableFeatures(object = pbmc[, cells], selection.method = &quot;vst&quot;, nfeatures = 2000) |&gt;
    VariableFeatures()
})</code></pre>
<pre><code>Finding variable features for layer counts
Finding variable features for layer counts</code></pre>
<pre class="r"><code>VariableFeatures(pbmc) &lt;- unique(unlist(vfeatures))
length(VariableFeatures(pbmc))</code></pre>
<pre><code>[1] 3237</code></pre>
<p>Scale and perform PCA.</p>
<pre class="r"><code>pbmc &lt;- ScaleData(pbmc, verbose = FALSE) |&gt;
  RunPCA(features = VariableFeatures(pbmc), npcs = 20, verbose = FALSE)</code></pre>
<p><code>RunHarmony()</code> is a generic function designed to interact
with Seurat objects. To run harmony on a Seurat object after it has been
normalised, only one argument needs to be specified which contains the
batch covariate located in the metadata. For this vignette, further
parameters are specified to align the dataset but the minimum parameters
are shown in the snippet below and is <strong>not run</strong>.</p>
<pre class="r"><code>pbmc &lt;- RunHarmony(pbmc, &quot;stim&quot;)</code></pre>
<p>Here, we will be running harmony with some indicative parameters and
plotting the convergence plot to illustrate some of the under the hood
functionality. By setting <code>plot_converge=TRUE</code>, harmony will
generate a plot with its objective showing the flow of the integration.
Each point represents the cost measured after a clustering round.
Different colors represent different Harmony iterations which is
controlled by <code>max_iter</code> (assuming that
<code>early_stop=FALSE</code>). Here <code>max_iter=10</code> and up to
10 correction steps are expected. However, <code>early_stop=TRUE</code>
so harmony will stop after the cost plateaus.</p>
<pre class="r"><code>pbmc &lt;- RunHarmony(pbmc, &quot;stim&quot;, plot_convergence = TRUE, nclust = 50, max_iter = 10, early_stop = TRUE)</code></pre>
<pre><code>Transposing data matrix</code></pre>
<pre><code>Initializing state using k-means centroids initialization</code></pre>
<pre><code>Harmony 1/10</code></pre>
<pre><code>Harmony 2/10</code></pre>
<pre><code>Harmony 3/10</code></pre>
<pre><code>Harmony 4/10</code></pre>
<pre><code>Harmony 5/10</code></pre>
<pre><code>Harmony converged after 5 iterations</code></pre>
<p><img src="figure/harmony.Rmd/run_harmony_pbmc-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-run_harmony_pbmc-1">
Past versions of run_harmony_pbmc-1.png
</button>
</p>
<div id="fig-run_harmony_pbmc-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/davetang/muse/blob/c3d7314fea2e83d5bf86f0932c6be6bee6ecbd13/docs/figure/harmony.Rmd/run_harmony_pbmc-1.png" target="_blank">c3d7314</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-04-15
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p><code>RunHarmony</code> has several parameters accessible to users
which are outlined below.</p>
<ul>
<li><code>object</code> (required) - The Seurat object. This vignette
assumes Seurat objects are version 5.</li>
<li><code>group.by.vars</code> (required) - A character vector that
specifies all the experimental covariates to be corrected/harmonized by
the algorithm.</li>
</ul>
<p>When using <code>RunHarmony()</code> with Seurat, harmony will look
up the <code>group.by.vars</code> metadata fields in the Seurat Object
metadata. For example, given the <code>pbmc[["stim"]]</code> exists as
the stim condition, setting <code>group.by.vars="stim"</code> will
perform integration of these samples accordingly. If you want to
integrate on another variable, it needs to be present in Seurat object’s
meta.data. To correct for several covariates, specify them in a vector:
<code>group.by.vars = c("stim", "new_covariate")</code>.</p>
<ul>
<li><code>reduction.use</code> - The cell embeddings to be used for the
batch alignment. This parameter assumes that a reduced dimension already
exists in the reduction slot of the Seurat object. By default, the
<code>pca</code> reduction is used.</li>
<li><code>dims.use</code> - Optional parameter which can use a name
vector to select specific dimensions to be harmonised.</li>
<li><code>nclust</code> - is a positive integer. Under the hood, harmony
applies k-means soft-clustering. For this task, <code>k</code> needs to
be determined. <code>nclust</code> corresponds to <code>k</code>. The
harmonisation results and performance are not particularly sensitive for
a reasonable range of this parameter value. If this parameter is not
set, harmony will autodetermine this based on the dataset size with a
maximum cap of 200. For dataset with a vast amount of different cell
types and batches this pamameter may need to be determined
manually.</li>
<li><code>sigma</code> - a positive scalar that controls the soft
clustering probability assignment of single-cells to different clusters.
Larger values will assign a larger probability to distant clusters of
cells resulting in a different correction profile. Single-cells are
assigned to clusters by their euclidean distance <span
class="math inline">\(d\)</span> to some cluster center <span
class="math inline">\(Y\)</span> after cosine normalisation which is
defined in the range [0,4]. The clustering probability of each cell is
calculated as <span class="math inline">\(e^{-\frac{d}{\sigma}}\)</span>
where <span class="math inline">\(\sigma\)</span> is controlled by the
<code>sigma</code> parameter. Default value of <code>sigma</code> is 0.1
and it generally works well since it defines probability assignment of a
cell in the range <span class="math inline">\([e^{-40}, e^0]\)</span>.
Larger values of <code>sigma</code> restrict the dynamic range of
probabilities that can be assigned to cells. For example,
<code>sigma=1</code> will yield a probabilities in the range of <span
class="math inline">\([e^{-4}, e^0]\)</span>.</li>
<li><code>theta</code> - <code>theta</code> is a positive scalar vector
that determines the coefficient of harmony’s diversity penalty for each
corrected experimental covariate. In challenging experimental
conditions, increasing theta may result in better integration results.
Theta is an expontential parameter of the diversity penalty, thus
setting <code>theta=0</code> disables this penalty while increasing it
to greater values than 1 will perform more aggressive corrections in an
expontential manner. By default, it will set <code>theta=2</code> for
each experimental covariate.</li>
<li><code>max_iter</code> - The number of correction steps harmony will
perform before completing the data set integration. In general, more
iterations than necessary increases computational runtime especially
which becomes evident in bigger datasets. Setting
<code>early_stop=TRUE</code> may reduce the actual number of correction
steps which will be smaller than <code>max_iter</code>.</li>
<li><code>early_stop</code> - Under the hood, harmony minimizes its
objective function through a series of clustering and integration tests.
By setting <code>early_stop=TRUE</code>, when the objective function is
less than <code>1e-4</code> after a correction step harmony exits before
reaching the <code>max_iter</code> correction steps. This parameter can
drastically reduce run-time in bigger datasets.</li>
<li><code>.options</code> - A set of internal algorithm parameters that
can be overriden. For advanced users only.</li>
</ul>
<p>These parameters are Seurat-specific and do not affect the flow of
the algorithm.</p>
<ul>
<li><code>project_dim</code> - Toggle-like parameter, by default
<code>project_dim=TRUE</code>. When enabled, <code>RunHarmony()</code>
calculates genomic feature loadings using Seurat’s
<code>ProjectDim()</code> that correspond to the harmonized cell
embeddings.</li>
<li><code>reduction.save</code> - The new Reduced Dimension slot
identifier. By default, <code>reduction.save=TRUE</code>. This option
allows several independent runs of harmony to be retained in the
appropriate slots in the SeuratObjects. It is useful if you want to try
Harmony with multiple parameters and save them as e.g. ‘harmony_theta0’,
‘harmony_theta1’, ‘harmony_theta2’.</li>
</ul>
<p>Miscellaneous parameters</p>
<p>These parameters help users troubleshoot harmony.</p>
<ul>
<li><code>plot_convergence</code> - Option that plots the convergence
plot after the execution of the algorithm. By default
<code>FALSE</code>. Setting it to <code>TRUE</code> will collect
harmony’s objective value and plot it allowing the user to troubleshoot
the flow of the algorithm and fine-tune the parameters of the dataset
integration procedure.</li>
</ul>
</div>
<div id="results" class="section level3">
<h3>Results</h3>
<p><code>RunHarmony()</code> returns the Seurat object which contains
the harmonised cell embeddings in a slot named <code>harmony</code>.
This entry can be accessed via <code>pbmc@reductions$harmony</code>. To
access the values of the cell embeddings we can also use
<code>Embeddings</code>.</p>
<pre class="r"><code>harmony.embeddings &lt;- Embeddings(pbmc, reduction = &quot;harmony&quot;)
head(harmony.embeddings)</code></pre>
<pre><code>                 harmony_1    harmony_2   harmony_3 harmony_4    harmony_5
ATCACTTGCTCGAA-1 -6.479702  0.008171644 -3.37515156 -4.099469 -0.008402849
CCGGAGACTGTGAC-1 -6.899788 -1.576543680 -6.21883818 -4.600855  5.431930567
CAAGCCCTGTTAGC-1 -6.778387  1.168303812 -6.51744885  9.119283  0.221378456
GAGGTACTAACGGG-1 -7.457892  0.230626098  0.04633372 -2.056185  2.202061677
CGCGGATGCCACAA-1 14.727748 -4.628065895 -4.46893682 -1.576657  1.511387377
ACATGGTGCCTAAG-1 -7.610653  0.169061161 -3.56685757 -2.899556  1.763453269
                   harmony_6  harmony_7  harmony_8  harmony_9   harmony_10
ATCACTTGCTCGAA-1 -0.50324290  0.7064309  1.3559831  0.6606093  0.896258741
CCGGAGACTGTGAC-1  2.62606436 -3.8302384  4.4752736  2.4579002  2.165149731
CAAGCCCTGTTAGC-1  0.02790648 -1.6405949  1.1675994  1.3743442 -0.161641981
GAGGTACTAACGGG-1  0.76979146  4.0367549 -0.9729053  2.6217274 -2.654879912
CGCGGATGCCACAA-1  0.50237890 -0.7808236 -1.7185065 -2.1407409  2.190081503
ACATGGTGCCTAAG-1  0.81828382 -0.1700942  0.6456826 -0.7853121  0.002772687
                 harmony_11 harmony_12 harmony_13   harmony_14 harmony_15
ATCACTTGCTCGAA-1 -0.4417489  1.4743115 -0.4759258 -0.212537776  0.2676670
CCGGAGACTGTGAC-1  2.7730307  3.3926242  0.7897797  4.203443728  4.3141192
CAAGCCCTGTTAGC-1 -0.6121661  1.3246926  0.3312554 -1.127996550 -0.5355179
GAGGTACTAACGGG-1  0.9421837  3.7047669 -0.3159213 -0.707279006 -1.3044352
CGCGGATGCCACAA-1  0.1525977  0.1711006  4.2970980  3.255725383 -0.7619538
ACATGGTGCCTAAG-1  0.3060353  2.1010580 -0.5419756  0.004886988  0.6353239
                 harmony_16  harmony_17 harmony_18 harmony_19 harmony_20
ATCACTTGCTCGAA-1  0.6435108  0.62385140  2.3402714 -0.3539086  0.1408163
CCGGAGACTGTGAC-1  3.4670592  9.67068155 -1.5557773  1.5395016  4.0110442
CAAGCCCTGTTAGC-1  3.9306648 -2.37229434 -0.4681119  1.1673524  0.0574584
GAGGTACTAACGGG-1 -2.4202051 -0.09903653  0.3197214  1.0333426 -3.4470187
CGCGGATGCCACAA-1 -1.0867898  0.55686789  3.0173422  4.8504994  1.5093119
ACATGGTGCCTAAG-1 -0.9257885  0.08646308 -0.6890350 -0.4465735 -0.3449852</code></pre>
<p>After Harmony integration, we should inspect the quality of the
harmonisation and contrast it with the unharmonised algorithm input.
Ideally, cells from different conditions will align along the Harmonized
PCs. If they are not, you could increase the <em>theta</em> value above
to force a more aggressive fit of the dataset and rerun the
workflow.</p>
<pre class="r"><code>p1 &lt;- DimPlot(object = pbmc, reduction = &quot;harmony&quot;, pt.size = .1, group.by = &quot;stim&quot;)
p2 &lt;- VlnPlot(object = pbmc, features = &quot;harmony_1&quot;, group.by = &quot;stim&quot;,  pt.size = .1)
p1 + p2</code></pre>
<p><img src="figure/harmony.Rmd/contrast_harmony-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-contrast_harmony-1">
Past versions of contrast_harmony-1.png
</button>
</p>
<div id="fig-contrast_harmony-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/davetang/muse/blob/c3d7314fea2e83d5bf86f0932c6be6bee6ecbd13/docs/figure/harmony.Rmd/contrast_harmony-1.png" target="_blank">c3d7314</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-04-15
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>Plot Genes correlated with the Harmonized PCs</p>
<pre class="r"><code>DimHeatmap(object = pbmc, reduction = &quot;harmony&quot;, cells = 500, dims = 1:3)</code></pre>
<p><img src="figure/harmony.Rmd/heatmap_harmony-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-heatmap_harmony-1">
Past versions of heatmap_harmony-1.png
</button>
</p>
<div id="fig-heatmap_harmony-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/davetang/muse/blob/c3d7314fea2e83d5bf86f0932c6be6bee6ecbd13/docs/figure/harmony.Rmd/heatmap_harmony-1.png" target="_blank">c3d7314</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-04-15
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>The harmonised cell embeddings generated by harmony can be used for
further integrated analyses. In this workflow, the Seurat object
contains the harmony <code>reduction</code> modality name in the method
that requires it.</p>
<p>Perform clustering using the harmonized vectors of cells</p>
<pre class="r"><code>pbmc &lt;- FindNeighbors(pbmc, reduction = &quot;harmony&quot;) |&gt;
  FindClusters(resolution = 0.5) </code></pre>
<pre><code>Computing nearest neighbor graph</code></pre>
<pre><code>Computing SNN</code></pre>
<pre><code>Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck

Number of nodes: 2000
Number of edges: 71873

Running Louvain algorithm...
Maximum modularity in 10 random starts: 0.8714
Number of communities: 10
Elapsed time: 0 seconds</code></pre>
<p>TSNE visualisation of harmony embeddings.</p>
<pre class="r"><code>pbmc &lt;- RunTSNE(pbmc, reduction = &quot;harmony&quot;)

p1 &lt;- DimPlot(pbmc, reduction = &quot;tsne&quot;, group.by = &quot;stim&quot;, pt.size = .1) + ggtitle(&quot;tSNE using harmony&quot;)
p2 &lt;- DimPlot(pbmc, reduction = &quot;tsne&quot;, label = TRUE, pt.size = .1) + ggtitle(&quot;Cell clusters&quot;)
p1 + p2</code></pre>
<p><img src="figure/harmony.Rmd/tsne-1.png" width="768" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-tsne-1">
Past versions of tsne-1.png
</button>
</p>
<div id="fig-tsne-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/davetang/muse/blob/a940a7e352aacdb1cf2283d0518d59c58bed8bc9/docs/figure/harmony.Rmd/tsne-1.png" target="_blank">a940a7e</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-04-15
</td>
</tr>
<tr>
<td>
<a href="https://github.com/davetang/muse/blob/c3d7314fea2e83d5bf86f0932c6be6bee6ecbd13/docs/figure/harmony.Rmd/tsne-1.png" target="_blank">c3d7314</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-04-15
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>One important observation is to assess that the harmonised data
contain biological states of the cells. Therefore by checking the
following genes we can see that biological cell states are preserved
after harmonisation.</p>
<p>Expression of gene panel heatmap in the harmonized PBMC dataset.</p>
<pre class="r"><code>FeaturePlot(
  object = pbmc,
  features= c(&quot;CD3D&quot;, &quot;SELL&quot;, &quot;CREM&quot;, &quot;CD8A&quot;, &quot;GNLY&quot;, &quot;CD79A&quot;, &quot;FCGR3A&quot;, &quot;CCL2&quot;, &quot;PPBP&quot;), 
  min.cutoff = &quot;q9&quot;,
  cols = c(&quot;lightgrey&quot;, &quot;blue&quot;),
  pt.size = 0.5
)</code></pre>
<p><img src="figure/harmony.Rmd/feature_plot-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-feature_plot-1">
Past versions of feature_plot-1.png
</button>
</p>
<div id="fig-feature_plot-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/davetang/muse/blob/c3d7314fea2e83d5bf86f0932c6be6bee6ecbd13/docs/figure/harmony.Rmd/feature_plot-1.png" target="_blank">c3d7314</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-04-15
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>Similar to TSNE, we can run UMAP by passing the harmony reduction in
the function.</p>
<pre class="r"><code>pbmc &lt;- RunUMAP(pbmc, reduction = &quot;harmony&quot;,  dims = 1:20)

p1 &lt;- DimPlot(pbmc, reduction = &quot;umap&quot;, group.by = &quot;stim&quot;, pt.size = .1)
p2 &lt;- DimPlot(pbmc, reduction = &quot;umap&quot;, label = TRUE,  pt.size = .1)
p1 + p2</code></pre>
<p><img src="figure/harmony.Rmd/umap_harmony-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-umap_harmony-1">
Past versions of umap_harmony-1.png
</button>
</p>
<div id="fig-umap_harmony-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/davetang/muse/blob/c3d7314fea2e83d5bf86f0932c6be6bee6ecbd13/docs/figure/harmony.Rmd/umap_harmony-1.png" target="_blank">c3d7314</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-04-15
</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<div id="seurat-sctransform-workflow-with-harmony"
class="section level3">
<h3>Seurat SCTransform workflow with harmony</h3>
<p>Use <code>SCTransform()</code> instead of
<code>NormalizeData()</code>, <code>ScaleData()</code>, and
<code>FindVariableFeatures()</code>.</p>
<pre class="r"><code>pbmc &lt;- CreateSeuratObject(
  counts = cbind(pbmc.stim, pbmc.ctrl),
  project = &quot;Kang&quot;,
  min.cells = 5
)

pbmc@meta.data$stim &lt;- c(rep(&quot;STIM&quot;, ncol(pbmc.stim)), rep(&quot;CTRL&quot;, ncol(pbmc.ctrl)))

pbmc &lt;- SCTransform(pbmc) |&gt;
  RunPCA(npcs = 20, verbose = FALSE)

pbmc &lt;- RunHarmony(pbmc, &quot;stim&quot;)

pbmc &lt;- RunUMAP(pbmc, reduction = &quot;harmony&quot;,  dims = 1:20)

DimPlot(pbmc_sct, reduction = &quot;umap&quot;, group.by = &quot;stim&quot;, pt.size = .1) + ggtitle(&quot;SCTransform without harmony&quot;) -&gt; p1
DimPlot(pbmc, reduction = &quot;umap&quot;, group.by = &quot;stim&quot;, pt.size = .1) + ggtitle(&quot;SCTransform with harmony&quot;) -&gt; p2

p1 + p2</code></pre>
<p><img src="figure/harmony.Rmd/sctransform_with_harmony-1.png" width="768" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-sctransform_with_harmony-1">
Past versions of sctransform_with_harmony-1.png
</button>
</p>
<div id="fig-sctransform_with_harmony-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/davetang/muse/blob/b174605ed41929903234d582b174888e56ed5de6/docs/figure/harmony.Rmd/sctransform_with_harmony-1.png" target="_blank">b174605</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-04-15
</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
</div>
<div id="from-10x-genomics-filtered-matrices" class="section level2">
<h2>From 10x Genomics filtered matrices</h2>
<p><a
href="https://github.com/davetang/benchmark/tree/main/data/10x">Downloaded</a>
Jurkat, 293t, and mixture cells.</p>
<pre class="r"><code>jurkat &lt;- Seurat::Read10X(&quot;data/jurkat/filtered_matrices_mex/hg19/&quot;)
t293 &lt;- Seurat::Read10X(&quot;data/293t/filtered_matrices_mex/hg19/&quot;)
half &lt;- Seurat::Read10X(&quot;data/jurkat_293t/filtered_matrices_mex/hg19/&quot;)

colnames(jurkat) &lt;- paste0(&#39;jurkat_&#39;, colnames(jurkat))
colnames(t293) &lt;- paste0(&#39;t293_&#39;, colnames(t293))
colnames(half) &lt;- paste0(&#39;half_&#39;, colnames(half))

seurat_obj &lt;- CreateSeuratObject(
  counts = cbind(jurkat, t293, half),
  project = &quot;Mix&quot;,
  min.cells = 5
)

seurat_obj@meta.data$dataset &lt;- c(
  rep(&quot;jurkat&quot;, ncol(jurkat)),
  rep(&quot;t293&quot;, ncol(t293)),
  rep(&quot;half&quot;, ncol(half))
)

options(future.globals.maxSize = 1.5 * 1024^3)

seurat_obj &lt;- SCTransform(seurat_obj) |&gt;
  RunPCA(npcs = 20, verbose = FALSE)

seurat_obj &lt;- RunUMAP(seurat_obj, dims = 1:20)

DimPlot(seurat_obj, reduction = &quot;umap&quot;, group.by = &quot;dataset&quot;, pt.size = .1) + ggtitle(&quot;Without harmony&quot;)</code></pre>
<p><img src="figure/harmony.Rmd/from_10x_filtered-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-from_10x_filtered-1">
Past versions of from_10x_filtered-1.png
</button>
</p>
<div id="fig-from_10x_filtered-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/davetang/muse/blob/4db2bdfea444a1eb980089c4808a54e461d5f4e5/docs/figure/harmony.Rmd/from_10x_filtered-1.png" target="_blank">4db2bdf</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-04-15
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>With harmony.</p>
<pre class="r"><code>seurat_obj_harmony &lt;- RunHarmony(seurat_obj, &quot;dataset&quot;)
seurat_obj_harmony &lt;- RunUMAP(seurat_obj_harmony, reduction = &quot;harmony&quot;,  dims = 1:20)

DimPlot(seurat_obj_harmony, reduction = &quot;umap&quot;, group.by = &quot;dataset&quot;, pt.size = .1) + ggtitle(&quot;With harmony&quot;)</code></pre>
<p><img src="figure/harmony.Rmd/from_10x_filtered_with_harmony-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-from_10x_filtered_with_harmony-1">
Past versions of from_10x_filtered_with_harmony-1.png
</button>
</p>
<div id="fig-from_10x_filtered_with_harmony-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/davetang/muse/blob/4db2bdfea444a1eb980089c4808a54e461d5f4e5/docs/figure/harmony.Rmd/from_10x_filtered_with_harmony-1.png" target="_blank">4db2bdf</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-04-15
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>Plot XIST.</p>
<ul>
<li><a href="https://en.wikipedia.org/wiki/HEK_293_cells">Human
embryonic kidney 293 cells</a>, also often referred to as HEK 293,
HEK-293, 293 cells, are an immortalised cell line derived from HEK cells
isolated from a <strong>female</strong> fetus in the 1970s.</li>
<li>The <a href="https://en.wikipedia.org/wiki/Jurkat_cells">Jurkat cell
line</a> (originally called JM) was established in the mid-1970s from
the peripheral blood of a 14-year-old boy with T cell leukemia.</li>
</ul>
<pre class="r"><code>FeaturePlot(
  object = seurat_obj_harmony,
  features= &quot;XIST&quot;,
  cols = c(&quot;lightgrey&quot;, &quot;blue&quot;),
  pt.size = 0.5
)</code></pre>
<p><img src="figure/harmony.Rmd/feature_plot_xist-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-feature_plot_xist-1">
Past versions of feature_plot_xist-1.png
</button>
</p>
<div id="fig-feature_plot_xist-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/davetang/muse/blob/4db2bdfea444a1eb980089c4808a54e461d5f4e5/docs/figure/harmony.Rmd/feature_plot_xist-1.png" target="_blank">4db2bdf</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-04-15
</td>
</tr>
</tbody>
</table>
</div>
</div>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span>
Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 4.4.1 (2024-06-14)
Platform: x86_64-pc-linux-gnu
Running under: Ubuntu 22.04.5 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 
LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

time zone: Etc/UTC
tzcode source: system (glibc)

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] Seurat_5.1.0       SeuratObject_5.0.2 sp_2.1-4           harmony_1.2.1     
 [5] Rcpp_1.0.13        patchwork_1.3.0    lubridate_1.9.3    forcats_1.0.0     
 [9] stringr_1.5.1      dplyr_1.1.4        purrr_1.0.2        readr_2.1.5       
[13] tidyr_1.3.1        tibble_3.2.1       ggplot2_3.5.1      tidyverse_2.0.0   
[17] workflowr_1.7.1   

loaded via a namespace (and not attached):
  [1] RColorBrewer_1.1-3     rstudioapi_0.17.1      jsonlite_1.8.9        
  [4] magrittr_2.0.3         spatstat.utils_3.1-0   farver_2.1.2          
  [7] rmarkdown_2.28         fs_1.6.4               vctrs_0.6.5           
 [10] ROCR_1.0-11            spatstat.explore_3.3-3 htmltools_0.5.8.1     
 [13] sass_0.4.9             sctransform_0.4.1      parallelly_1.38.0     
 [16] KernSmooth_2.23-24     bslib_0.8.0            htmlwidgets_1.6.4     
 [19] ica_1.0-3              plyr_1.8.9             plotly_4.10.4         
 [22] zoo_1.8-12             cachem_1.1.0           whisker_0.4.1         
 [25] igraph_2.1.1           mime_0.12              lifecycle_1.0.4       
 [28] pkgconfig_2.0.3        Matrix_1.7-0           R6_2.5.1              
 [31] fastmap_1.2.0          fitdistrplus_1.2-1     future_1.34.0         
 [34] shiny_1.9.1            digest_0.6.37          colorspace_2.1-1      
 [37] ps_1.8.1               rprojroot_2.0.4        tensor_1.5            
 [40] RSpectra_0.16-2        irlba_2.3.5.1          labeling_0.4.3        
 [43] progressr_0.15.0       fansi_1.0.6            spatstat.sparse_3.1-0 
 [46] timechange_0.3.0       httr_1.4.7             polyclip_1.10-7       
 [49] abind_1.4-8            compiler_4.4.1         withr_3.0.2           
 [52] fastDummies_1.7.4      highr_0.11             MASS_7.3-60.2         
 [55] tools_4.4.1            lmtest_0.9-40          httpuv_1.6.15         
 [58] future.apply_1.11.3    goftest_1.2-3          glue_1.8.0            
 [61] callr_3.7.6            nlme_3.1-164           promises_1.3.0        
 [64] grid_4.4.1             Rtsne_0.17             getPass_0.2-4         
 [67] cluster_2.1.6          reshape2_1.4.4         generics_0.1.3        
 [70] gtable_0.3.6           spatstat.data_3.1-2    tzdb_0.4.0            
 [73] data.table_1.16.2      hms_1.1.3              utf8_1.2.4            
 [76] spatstat.geom_3.3-3    RcppAnnoy_0.0.22       ggrepel_0.9.6         
 [79] RANN_2.6.2             pillar_1.9.0           spam_2.11-0           
 [82] RcppHNSW_0.6.0         later_1.3.2            splines_4.4.1         
 [85] lattice_0.22-6         deldir_2.0-4           survival_3.6-4        
 [88] tidyselect_1.2.1       miniUI_0.1.1.1         pbapply_1.7-2         
 [91] knitr_1.48             git2r_0.35.0           gridExtra_2.3         
 [94] scattermore_1.2        RhpcBLASctl_0.23-42    xfun_0.48             
 [97] matrixStats_1.4.1      stringi_1.8.4          lazyeval_0.2.2        
[100] yaml_2.3.10            evaluate_1.0.1         codetools_0.2-20      
[103] cli_3.6.3              uwot_0.2.2             xtable_1.8-4          
[106] reticulate_1.39.0      munsell_0.5.1          processx_3.8.4        
[109] jquerylib_0.1.4        spatstat.random_3.3-2  globals_0.16.3        
[112] png_0.1-8              spatstat.univar_3.0-1  parallel_4.4.1        
[115] dotCall64_1.2          listenv_0.9.1          viridisLite_0.4.2     
[118] scales_1.3.0           ggridges_0.5.6         leiden_0.4.3.1        
[121] rlang_1.1.4            cowplot_1.1.3         </code></pre>
</div>
</div>
</div>
</div>
</div>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
https://docs.mathjax.org/en/latest/web/configuration.html. This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>




</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
