<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />



<meta name="date" content="2025-03-05" />

<title>Working with HDF5 data in R</title>

<script src="site_libs/header-attrs-2.28/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-6.4.2/css/all.min.css" rel="stylesheet" />
<link href="site_libs/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet" />

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/main/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->




</head>

<body>


<div class="container-fluid main-container">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">My blog</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/davetang/muse">
    <span class="fa fa-github"></span>
     
    Source code
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Working with HDF5 data in R</h1>
<h4 class="date">2025-03-05</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span>
workflowr <span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2025-03-05
</p>
<p>
<strong>Checks:</strong> <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 7
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong> <code>muse/</code> <span
class="glyphicon glyphicon-question-sign" aria-hidden="true"
title="This is the local directory in which the code in this file was executed.">
</span>
</p>
<p>
This reproducible <a href="https://rmarkdown.rstudio.com">R Markdown</a>
analysis was created with <a
  href="https://github.com/workflowr/workflowr">workflowr</a> (version
1.7.1). The <em>Checks</em> tab describes the reproducibility checks
that were applied when the results were created. The <em>Past
versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date
</a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git
repository, you know the exact version of the code that produced these
results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the
global environment can affect the analysis in your R Markdown file in
unknown ways. For reproduciblity it’s best to always run the code in an
empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20200712code">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Seed:</strong>
<code>set.seed(20200712)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20200712code"
class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20200712)</code> was run prior to running
the code in the R Markdown file. Setting a seed ensures that any results
that rely on randomness, e.g. subsampling or permutations, are
reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Session information:</strong>
recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package
versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be
confident that you successfully produced the results during this
run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr
project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomdavetangmusetree413b2a08a2ffb4a984005fc945527bfc8b0ac860targetblank413b2a0a">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Repository version:</strong>
<a href="https://github.com/davetang/muse/tree/413b2a08a2ffb4a984005fc945527bfc8b0ac860" target="_blank">413b2a0</a>
</a>
</p>
</div>
<div
id="strongRepositoryversionstrongahrefhttpsgithubcomdavetangmusetree413b2a08a2ffb4a984005fc945527bfc8b0ac860targetblank413b2a0a"
class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development
and connecting the code version to the results is critical for
reproducibility.
</p>
<p>
The results in this page were generated with repository version
<a href="https://github.com/davetang/muse/tree/413b2a08a2ffb4a984005fc945527bfc8b0ac860" target="_blank">413b2a0</a>.
See the <em>Past versions</em> tab to see a history of the changes made
to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for
the analysis have been committed to Git prior to generating the results
(you can use <code>wflow_publish</code> or
<code>wflow_git_commit</code>). workflowr only checks the R Markdown
file, but you know if there are other scripts or data files that it
depends on. Below is the status of the Git repository when the results
were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .Rproj.user/
    Ignored:    data/1M_neurons_filtered_gene_bc_matrices_h5.h5
    Ignored:    data/293t/
    Ignored:    data/293t_3t3_filtered_gene_bc_matrices.tar.gz
    Ignored:    data/293t_filtered_gene_bc_matrices.tar.gz
    Ignored:    data/5k_Human_Donor1_PBMC_3p_gem-x_5k_Human_Donor1_PBMC_3p_gem-x_count_sample_filtered_feature_bc_matrix.h5
    Ignored:    data/5k_Human_Donor2_PBMC_3p_gem-x_5k_Human_Donor2_PBMC_3p_gem-x_count_sample_filtered_feature_bc_matrix.h5
    Ignored:    data/5k_Human_Donor3_PBMC_3p_gem-x_5k_Human_Donor3_PBMC_3p_gem-x_count_sample_filtered_feature_bc_matrix.h5
    Ignored:    data/5k_Human_Donor4_PBMC_3p_gem-x_5k_Human_Donor4_PBMC_3p_gem-x_count_sample_filtered_feature_bc_matrix.h5
    Ignored:    data/Parent_SC3v3_Human_Glioblastoma_filtered_feature_bc_matrix.tar.gz
    Ignored:    data/brain_counts/
    Ignored:    data/cl.obo
    Ignored:    data/cl.owl
    Ignored:    data/jurkat/
    Ignored:    data/jurkat:293t_50:50_filtered_gene_bc_matrices.tar.gz
    Ignored:    data/jurkat_293t/
    Ignored:    data/jurkat_filtered_gene_bc_matrices.tar.gz
    Ignored:    data/pbmc20k/
    Ignored:    data/pbmc20k_seurat/
    Ignored:    data/pbmc3k/
    Ignored:    data/pbmc4k_filtered_gene_bc_matrices.tar.gz
    Ignored:    data/pbmc_1k_v3_raw_feature_bc_matrix.h5
    Ignored:    data/refdata-gex-GRCh38-2020-A.tar.gz
    Ignored:    data/seurat_1m_neuron.rds
    Ignored:    data/t_3k_filtered_gene_bc_matrices.tar.gz
    Ignored:    r_packages_4.4.1/

Untracked files:
    Untracked:  analysis/bioc_scrnaseq.Rmd
    Untracked:  data/97516b79-8d08-46a6-b329-5d0a25b0be98.h5ad
    Untracked:  data/pbmc_1k_v3_filtered_feature_bc_matrix.h5

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not
included in this status report because it is ok for generated content to
have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">
<p>
These are the previous versions of the repository in which changes were
made to the R Markdown (<code>analysis/hdf5.Rmd</code>) and HTML
(<code>docs/hdf5.html</code>) files. If you’ve configured a remote Git
repository (see <code>?wflow_git_remote</code>), click on the hyperlinks
in the table below to view the files as they were in that past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/davetang/muse/blob/413b2a08a2ffb4a984005fc945527bfc8b0ac860/analysis/hdf5.Rmd" target="_blank">413b2a0</a>
</td>
<td>
Dave Tang
</td>
<td>
2025-03-05
</td>
<td>
Tabula Sapiens h5ad
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/davetang/muse/4c0da5462729815b0acc7b35a4419ab417b2df58/docs/hdf5.html" target="_blank">4c0da54</a>
</td>
<td>
Dave Tang
</td>
<td>
2025-03-01
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/davetang/muse/blob/f3c325ea5d27f13806703cec404cf530ea49880d/analysis/hdf5.Rmd" target="_blank">f3c325e</a>
</td>
<td>
Dave Tang
</td>
<td>
2025-03-01
</td>
<td>
10x gene expression HDF5 file
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/davetang/muse/f706b109b93ff78ad7a60d8d226ffb65531a2e38/docs/hdf5.html" target="_blank">f706b10</a>
</td>
<td>
Dave Tang
</td>
<td>
2023-07-21
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/davetang/muse/blob/e5a16afc4824b70109ce517b478ddb0214aab07a/analysis/hdf5.Rmd" target="_blank">e5a16af</a>
</td>
<td>
Dave Tang
</td>
<td>
2023-07-21
</td>
<td>
HDF5
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<div id="introduction" class="section level1">
<h1>Introduction</h1>
<p>From <a
href="https://en.wikipedia.org/wiki/Hierarchical_Data_Format">Wikipedia</a>:</p>
<blockquote>
<p>Hierarchical Data Format (HDF) is a set of file formats (HDF4, HDF5)
designed to store and organize large amounts of data. Originally
developed at the U.S. National Center for Supercomputing Applications,
it is supported by The HDF Group, a non-profit corporation whose mission
is to ensure continued development of HDF5 technologies and the
continued accessibility of data stored in HDF.</p>
<p>In keeping with this goal, the HDF libraries and associated tools are
available under a liberal, BSD-like license for general use. HDF is
supported by many commercial and non-commercial software platforms and
programming languages. The freely available HDF distribution consists of
the library, command-line utilities, test suite source, Java interface,
and the Java-based HDF Viewer (HDFView).</p>
<p>The current version, HDF5, differs significantly in design and API
from the major legacy version HDF4.</p>
</blockquote>
</div>
<div id="h5ad" class="section level1">
<h1>h5ad</h1>
<p><a
href="https://cellxgene.cziscience.com/collections/e5f58829-1a66-40b5-a624-9046778e74f5">Download</a>
from Tabula Sapiens:</p>
<blockquote>
<p>Tabula Sapiens is a benchmark, first-draft human cell atlas of over
1.1M cells from 28 organs of 24 normal human subjects. This work is the
product of the Tabula Sapiens Consortium. Taking the organs from the
same individual controls for genetic background, age, environment, and
epigenetic effects, and allows detailed analysis and comparison of cell
types that are shared between tissues.</p>
</blockquote>
<pre class="r"><code>ts_heart_url &lt;- &quot;https://datasets.cellxgene.cziscience.com/97516b79-8d08-46a6-b329-5d0a25b0be98.h5ad&quot;
ts_heart_file &lt;- paste0(&#39;data/&#39;, basename(ts_heart_url))
if(file.exists(ts_heart_file)){
  message(paste0(ts_heart_file, &quot; already exists&quot;))
} else {
  options(timeout = 1000)
  download.file(url = ts_heart_url, destfile = ts_heart_file)
}</code></pre>
<pre><code>data/97516b79-8d08-46a6-b329-5d0a25b0be98.h5ad already exists</code></pre>
<p>Creat new <code>H5File</code>.</p>
<pre class="r"><code>ts_heart_h5 &lt;- H5File$new(filename = ts_heart_file, mode = &quot;r&quot;)
ts_heart_h5</code></pre>
<pre><code>Class: H5File
Filename: /home/rstudio/muse/data/97516b79-8d08-46a6-b329-5d0a25b0be98.h5ad
Access type: H5F_ACC_RDONLY
Attributes: encoding-type, encoding-version
Listing:
   name  obj_type dataset.dims dataset.type_class
      X H5I_GROUP         &lt;NA&gt;               &lt;NA&gt;
 layers H5I_GROUP         &lt;NA&gt;               &lt;NA&gt;
    obs H5I_GROUP         &lt;NA&gt;               &lt;NA&gt;
   obsm H5I_GROUP         &lt;NA&gt;               &lt;NA&gt;
   obsp H5I_GROUP         &lt;NA&gt;               &lt;NA&gt;
    raw H5I_GROUP         &lt;NA&gt;               &lt;NA&gt;
    uns H5I_GROUP         &lt;NA&gt;               &lt;NA&gt;
    var H5I_GROUP         &lt;NA&gt;               &lt;NA&gt;
   varm H5I_GROUP         &lt;NA&gt;               &lt;NA&gt;
   varp H5I_GROUP         &lt;NA&gt;               &lt;NA&gt;</code></pre>
<p>List datasets.</p>
<pre class="r"><code>hdf5r::list.datasets(ts_heart_h5) |&gt;
  head()</code></pre>
<pre><code>[1] &quot;X/data&quot;                       &quot;X/indices&quot;                   
[3] &quot;X/indptr&quot;                     &quot;layers/decontXcounts/data&quot;   
[5] &quot;layers/decontXcounts/indices&quot; &quot;layers/decontXcounts/indptr&quot; </code></pre>
<p>UMAP.</p>
<pre class="r"><code>ts_heart_umap_h5 &lt;- ts_heart_h5[[&quot;obsm/X_umap&quot;]]
ts_heart_umap &lt;- ts_heart_umap_h5[,]
plot(t(ts_heart_umap), pch = &#39;.&#39;, xlab = &quot;UMAP1&quot;, ylab = &quot;UMAP2&quot;, main = &quot;Tabula Sapiens Heart&quot;)</code></pre>
<p><img src="figure/hdf5.Rmd/heart_h5_umap-1.png" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="x-hdf5" class="section level1">
<h1>10x HDF5</h1>
<div id="gene-expression-data" class="section level2">
<h2>Gene expression data</h2>
<p>Download <a
href="https://support.10xgenomics.com/single-cell-gene-expression/datasets/1.3.0/1M_neurons?">1.3
Million Brain Cells from E18 Mice</a> (3.93 GB).</p>
<pre class="r"><code>my_url &lt;- &#39;https://cf.10xgenomics.com/samples/cell-exp/1.3.0/1M_neurons/1M_neurons_filtered_gene_bc_matrices_h5.h5&#39;
my_file &lt;- paste0(&quot;data/&quot;, basename(my_url))

if(!file.exists(my_file)){
  options(timeout = 10000)
  download.file(url = my_url, destfile = my_file)
}</code></pre>
<p>New <code>H5File</code> object.</p>
<pre class="r"><code>gex_feat_mat &lt;- H5File$new(filename = my_file, mode = &quot;r&quot;)
gex_feat_mat</code></pre>
<pre><code>Class: H5File
Filename: /home/rstudio/muse/data/1M_neurons_filtered_gene_bc_matrices_h5.h5
Access type: H5F_ACC_RDONLY
Attributes: TITLE, CLASS, VERSION, FILTERS, PYTABLES_FORMAT_VERSION, filetype, chemistry_description, library_ids, original_gem_groups
Listing:
 name  obj_type dataset.dims dataset.type_class
 mm10 H5I_GROUP         &lt;NA&gt;               &lt;NA&gt;</code></pre>
<p>List datasets.</p>
<pre class="r"><code>hdf5r::list.datasets(gex_feat_mat)</code></pre>
<pre><code>[1] &quot;mm10/barcodes&quot;   &quot;mm10/data&quot;       &quot;mm10/gene_names&quot; &quot;mm10/genes&quot;     
[5] &quot;mm10/indices&quot;    &quot;mm10/indptr&quot;     &quot;mm10/shape&quot;     </code></pre>
<p>Get gene names.</p>
<pre class="r"><code>gene_names_h5 &lt;- gex_feat_mat[[&quot;mm10/gene_names&quot;]]
gene_names &lt;- gene_names_h5[]
class(gene_names)</code></pre>
<pre><code>[1] &quot;character&quot;</code></pre>
<pre class="r"><code>length(gene_names)</code></pre>
<pre><code>[1] 27998</code></pre>
<pre class="r"><code>head(gene_names)</code></pre>
<pre><code>[1] &quot;Xkr4&quot;    &quot;Gm1992&quot;  &quot;Gm37381&quot; &quot;Rp1&quot;     &quot;Rp1&quot;     &quot;Sox17&quot;  </code></pre>
</div>
</div>
<div id="archs4" class="section level1">
<h1>ARCHS4</h1>
<p><a href="https://maayanlab.cloud/archs4/download.html">Download
ARCHS4 HDF5 data</a> and load using the <a
href="https://github.com/hhoeflin/hdf5r">hdf5r</a> package.</p>
<pre class="r"><code>human &lt;- H5File$new(filename = &quot;/data2/archs4/human_gene_v2.5.h5&quot;, mode = &quot;r&quot;)
human</code></pre>
<pre><code>Class: H5File
Filename: /data2/archs4/human_gene_v2.5.h5
Access type: H5F_ACC_RDONLY
Listing:
 name  obj_type dataset.dims dataset.type_class
 data H5I_GROUP         &lt;NA&gt;               &lt;NA&gt;
 meta H5I_GROUP         &lt;NA&gt;               &lt;NA&gt;</code></pre>
<p>Different groups.</p>
<pre class="r"><code>human[[&quot;data&quot;]]</code></pre>
<pre><code>Class: H5Group
Filename: /data2/archs4/human_gene_v2.5.h5
Group: /data
Listing:
       name    obj_type   dataset.dims dataset.type_class
 expression H5I_DATASET 888821 x 67186        H5T_INTEGER</code></pre>
<pre class="r"><code>human[[&quot;meta&quot;]]</code></pre>
<pre><code>Class: H5Group
Filename: /data2/archs4/human_gene_v2.5.h5
Group: /meta
Listing:
    name  obj_type dataset.dims dataset.type_class
   genes H5I_GROUP         &lt;NA&gt;               &lt;NA&gt;
    info H5I_GROUP         &lt;NA&gt;               &lt;NA&gt;
 samples H5I_GROUP         &lt;NA&gt;               &lt;NA&gt;</code></pre>
<div id="getting-information-about-objects" class="section level2">
<h2>Getting information about objects</h2>
<p>Use the <code>names()</code> function to get all names of
objects.</p>
<pre class="r"><code>names(human)</code></pre>
<pre><code>[1] &quot;data&quot; &quot;meta&quot;</code></pre>
<p>The <code>ls()</code> function provides more information.</p>
<pre class="r"><code>human$ls()</code></pre>
<pre><code>  name     link.type  obj_type num_attrs group.nlinks group.mounted
1 data H5L_TYPE_HARD H5I_GROUP         0            1             0
2 meta H5L_TYPE_HARD H5I_GROUP         0            3             0
  dataset.rank dataset.dims dataset.maxdims dataset.type_class
1           NA         &lt;NA&gt;            &lt;NA&gt;               &lt;NA&gt;
2           NA         &lt;NA&gt;            &lt;NA&gt;               &lt;NA&gt;
  dataset.space_class committed_type
1                &lt;NA&gt;           &lt;NA&gt;
2                &lt;NA&gt;           &lt;NA&gt;</code></pre>
<p>List functions.</p>
<pre class="r"><code>hdf5r::list.attributes(human)</code></pre>
<pre><code>character(0)</code></pre>
<pre class="r"><code>hdf5r::list.datasets(human)</code></pre>
<pre><code> [1] &quot;data/expression&quot;                    &quot;meta/genes/biotype&quot;                
 [3] &quot;meta/genes/ensembl_gene&quot;            &quot;meta/genes/symbol&quot;                 
 [5] &quot;meta/info/author&quot;                   &quot;meta/info/contact&quot;                 
 [7] &quot;meta/info/creation-date&quot;            &quot;meta/info/laboratory&quot;              
 [9] &quot;meta/info/version&quot;                  &quot;meta/samples/channel_count&quot;        
[11] &quot;meta/samples/characteristics_ch1&quot;   &quot;meta/samples/contact_address&quot;      
[13] &quot;meta/samples/contact_city&quot;          &quot;meta/samples/contact_country&quot;      
[15] &quot;meta/samples/contact_institute&quot;     &quot;meta/samples/contact_name&quot;         
[17] &quot;meta/samples/contact_zip&quot;           &quot;meta/samples/data_processing&quot;      
[19] &quot;meta/samples/extract_protocol_ch1&quot;  &quot;meta/samples/geo_accession&quot;        
[21] &quot;meta/samples/instrument_model&quot;      &quot;meta/samples/last_update_date&quot;     
[23] &quot;meta/samples/library_selection&quot;     &quot;meta/samples/library_source&quot;       
[25] &quot;meta/samples/library_strategy&quot;      &quot;meta/samples/molecule_ch1&quot;         
[27] &quot;meta/samples/organism_ch1&quot;          &quot;meta/samples/platform_id&quot;          
[29] &quot;meta/samples/relation&quot;              &quot;meta/samples/sample&quot;               
[31] &quot;meta/samples/series_id&quot;             &quot;meta/samples/singlecellprobability&quot;
[33] &quot;meta/samples/source_name_ch1&quot;       &quot;meta/samples/status&quot;               
[35] &quot;meta/samples/submission_date&quot;       &quot;meta/samples/taxid_ch1&quot;            
[37] &quot;meta/samples/title&quot;                 &quot;meta/samples/type&quot;                 </code></pre>
<pre class="r"><code>hdf5r::list.groups(human)</code></pre>
<pre><code>[1] &quot;data&quot;         &quot;meta&quot;         &quot;meta/genes&quot;   &quot;meta/info&quot;    &quot;meta/samples&quot;</code></pre>
<pre class="r"><code>hdf5r::list.objects(human)</code></pre>
<pre><code> [1] &quot;data&quot;                               &quot;data/expression&quot;                   
 [3] &quot;meta&quot;                               &quot;meta/genes&quot;                        
 [5] &quot;meta/genes/biotype&quot;                 &quot;meta/genes/ensembl_gene&quot;           
 [7] &quot;meta/genes/symbol&quot;                  &quot;meta/info&quot;                         
 [9] &quot;meta/info/author&quot;                   &quot;meta/info/contact&quot;                 
[11] &quot;meta/info/creation-date&quot;            &quot;meta/info/laboratory&quot;              
[13] &quot;meta/info/version&quot;                  &quot;meta/samples&quot;                      
[15] &quot;meta/samples/channel_count&quot;         &quot;meta/samples/characteristics_ch1&quot;  
[17] &quot;meta/samples/contact_address&quot;       &quot;meta/samples/contact_city&quot;         
[19] &quot;meta/samples/contact_country&quot;       &quot;meta/samples/contact_institute&quot;    
[21] &quot;meta/samples/contact_name&quot;          &quot;meta/samples/contact_zip&quot;          
[23] &quot;meta/samples/data_processing&quot;       &quot;meta/samples/extract_protocol_ch1&quot; 
[25] &quot;meta/samples/geo_accession&quot;         &quot;meta/samples/instrument_model&quot;     
[27] &quot;meta/samples/last_update_date&quot;      &quot;meta/samples/library_selection&quot;    
[29] &quot;meta/samples/library_source&quot;        &quot;meta/samples/library_strategy&quot;     
[31] &quot;meta/samples/molecule_ch1&quot;          &quot;meta/samples/organism_ch1&quot;         
[33] &quot;meta/samples/platform_id&quot;           &quot;meta/samples/relation&quot;             
[35] &quot;meta/samples/sample&quot;                &quot;meta/samples/series_id&quot;            
[37] &quot;meta/samples/singlecellprobability&quot; &quot;meta/samples/source_name_ch1&quot;      
[39] &quot;meta/samples/status&quot;                &quot;meta/samples/submission_date&quot;      
[41] &quot;meta/samples/taxid_ch1&quot;             &quot;meta/samples/title&quot;                
[43] &quot;meta/samples/type&quot;                 </code></pre>
</div>
<div id="get-data" class="section level2">
<h2>Get data</h2>
<p>Function to get data.</p>
<pre class="r"><code>wrap &lt;- function(x){
  x &lt;- sub(&quot;^&quot;, &quot;[[&#39;&quot;, x)
  sub(&quot;$&quot;, &quot;&#39;]]&quot;, x)
}

get_data &lt;- function(obj, ds, ...){
  s &lt;- unlist(strsplit(x = ds, split = &quot;/&quot;))
  txt &lt;- paste0(obj, paste(sapply(s, wrap), collapse = &quot;&quot;, sep = &quot;&quot;), ...)
  eval(parse(text = txt))
}</code></pre>
<p>Check out some of the data.</p>
<pre class="r"><code>get_data(&#39;human&#39;, &#39;meta/samples/title&#39;, &#39;[1:5]&#39;)</code></pre>
<pre><code>[1] &quot;OCI-LY1_48hrs_mRNAseq_3x_siNT_R1&quot;   &quot;OCI-LY1_48hrs_mRNAseq_3x_siNT_R2&quot;  
[3] &quot;OCI-LY1_48hrs_mRNAseq_3x_siNT_R3&quot;   &quot;OCI-LY1_48hrs_mRNAseq_3x_siBCL6_R1&quot;
[5] &quot;OCI-LY1_48hrs_mRNAseq_3x_siBCL6_R2&quot;</code></pre>
<pre class="r"><code>get_data(&#39;human&#39;, &#39;meta/samples/sample&#39;, &#39;[1:5]&#39;)</code></pre>
<pre><code>[1] &quot;GSM1000981&quot; &quot;GSM1000982&quot; &quot;GSM1000983&quot; &quot;GSM1000984&quot; &quot;GSM1000985&quot;</code></pre>
<pre class="r"><code>get_data(&#39;human&#39;, &#39;meta/samples/geo_accession&#39;, &#39;[1:5]&#39;)</code></pre>
<pre><code>[1] &quot;GSM1000981&quot; &quot;GSM1000982&quot; &quot;GSM1000983&quot; &quot;GSM1000984&quot; &quot;GSM1000985&quot;</code></pre>
<pre class="r"><code>get_data(&#39;human&#39;, &#39;meta/samples/series_id&#39;, &#39;[1:5]&#39;)</code></pre>
<pre><code>[1] &quot;GSE29282&quot; &quot;GSE29282&quot; &quot;GSE29282&quot; &quot;GSE29282&quot; &quot;GSE29282&quot;</code></pre>
<pre class="r"><code>get_data(&#39;human&#39;, &#39;meta/samples/characteristics_ch1&#39;, &#39;[1:5]&#39;)</code></pre>
<pre><code>[1] &quot;treatment: siNT,cell line: OCI-LY1&quot;  
[2] &quot;treatment: siNT,cell line: OCI-LY1&quot;  
[3] &quot;treatment: siNT,cell line: OCI-LY1&quot;  
[4] &quot;treatment: siBCL6,cell line: OCI-LY1&quot;
[5] &quot;treatment: siBCL6,cell line: OCI-LY1&quot;</code></pre>
<pre class="r"><code>get_data(&#39;human&#39;, &#39;meta/samples/extract_protocol_ch1&#39;, &#39;[1]&#39;)</code></pre>
<pre><code>[1] &quot;ChIP-seq libraries were prepared using the Illumina ChIP-seq Library preparation Kit following the manufactureâ\u0080\u0099s instructions with minor modifications. Briefly 10ng of purified ChIP DNA was end repaired by conversion of overhangs to phosphorylated blunt ends. Aâ\u0080\u0099 bases were added to the 3â\u0080\u0099ends of the DNA fragments and Illumina adapters (1:30 dilution) were ligated to the ends of ChIP fragments. After adapter ligation DNA was separated by electrophoresis and size selected by isolating a gel band of 250Â±25bp. This fragment range corresponds to a ChIP fragment range of about 158Â±25bp. Size selected fragments were PCR amplified for 15cycles using Illumina genomic DNA primers 1.1 and 1.2 with the following program (30s at 98oC, 15cycles of 10 at 98oC, 30s at 65oC,30s at 72oC and 5min extension at 75oC). Libraries were quantified and validated using Agilent Technologies 2100 Bioanalyser for size, concentration and purity. Q-PCR was repeated to confirm retention of relative enrichment.                               RNAseq: Three ug of total RNA was isolated from OCI-Ly1 cells transfected using Nucleofector 96-well Shuttle system (Lonza) with siBCL6 (HSS100968) or siNT (46-2001) (Stealth RNAi, Invitrogen) at 24hrs and 48hrs after nucleofection. RNAeasy Plus Kit (Qiagen) that included a gDNA elimination step was used for RNA isolation. RNA concentration and purity were determined using Nanodrop (Thermo Scientific) and integrity was verified using Agilent 2100 Bioanalyzer (Agilent Technologies, Santa Clara, CA). Libraries were generated using mRNA-seq sample prep kit by Illumina. Briefly, mRNA was selected by two rounds of purification using magnetic polydT beads and then fragmented. First strand synthesis was performed using random oligos and SupersciptIII (Invitrogen). After second strand synthesis a 200bp paired-end library was prepared following the Illumina paired-end library preparation protocol.&quot;</code></pre>
<pre class="r"><code>get_data(&#39;human&#39;, &#39;meta/samples/source_name_ch1&#39;, &#39;[1:5]&#39;)</code></pre>
<pre><code>[1] &quot;Human DLBCL cel line&quot; &quot;Human DLBCL cel line&quot; &quot;Human DLBCL cel line&quot;
[4] &quot;Human DLBCL cel line&quot; &quot;Human DLBCL cel line&quot;</code></pre>
</div>
<div id="store-gene-symbols" class="section level2">
<h2>Store gene symbols</h2>
<p>Gene symbols.</p>
<pre class="r"><code>human[[&#39;meta&#39;]][[&#39;genes&#39;]][[&#39;symbol&#39;]]</code></pre>
<pre><code>Class: H5D
Dataset: /meta/genes/symbol
Filename: /data2/archs4/human_gene_v2.5.h5
Access type: H5F_ACC_RDONLY
Datatype: H5T_STRING {
      STRSIZE H5T_VARIABLE;
      STRPAD H5T_STR_NULLTERM;
      CSET H5T_CSET_UTF8;
      CTYPE H5T_C_S1;
   }
Space: Type=Simple     Dims=67186     Maxdims=67186
Chunk: 2100</code></pre>
<p>Store gene symbols.</p>
<pre class="r"><code>my_idx &lt;- 1:human[[&#39;meta&#39;]][[&#39;genes&#39;]][[&#39;symbol&#39;]]$maxdims
my_gene_sym &lt;- human[[&#39;meta&#39;]][[&#39;genes&#39;]][[&#39;symbol&#39;]][my_idx]
head(my_gene_sym)</code></pre>
<pre><code>[1] &quot;TSPAN6&quot;   &quot;TNMD&quot;     &quot;DPM1&quot;     &quot;SCYL3&quot;    &quot;C1orf112&quot; &quot;FGR&quot;     </code></pre>
</div>
<div id="source" class="section level2">
<h2>Source</h2>
<p>Check out some sources.</p>
<pre class="r"><code>unique(get_data(&#39;human&#39;, &#39;meta/samples/source_name_ch1&#39;, &#39;[1:100]&#39;))</code></pre>
<pre><code> [1] &quot;Human DLBCL cel line&quot;                          
 [2] &quot;placenta&quot;                                      
 [3] &quot;HeLa&quot;                                          
 [4] &quot;RNA-seq H1-hESC&quot;                               
 [5] &quot;RNA-seq H1-hESC-48hr-Activin&quot;                  
 [6] &quot;Primary Human Umbilical Vein Endothelial Cells&quot;
 [7] &quot;Myeloid Leukemia&quot;                              
 [8] &quot;HT29 treated with 0 Î¼M of 5-Aza&quot;              
 [9] &quot;HT29 treated with 5 Î¼M of 5-Aza&quot;              
[10] &quot;HT29 treated with 10 Î¼M of 5-Aza&quot;             
[11] &quot;corneal endothelial cells&quot;                     
[12] &quot;fibroblast&quot;                                    
[13] &quot;iPS&quot;                                           
[14] &quot;Human Stem Derived Motor Neurons GFP high&quot;     
[15] &quot;Human Stem Derived Motor Neurons no GFP&quot;       
[16] &quot;Immortalized primary fibroblasts&quot;              
[17] &quot;NTera2/D1 cells&quot;                               
[18] &quot;gastric adenocarcinoma cells&quot;                  
[19] &quot;CNE1-MOCK_RNA-seq&quot;                             
[20] &quot;CNE1-BART1_RNA-seq&quot;                            
[21] &quot;CNE1-BART3_RNA-seq&quot;                            
[22] &quot;CNE1-BART7_RNA-seq&quot;                            
[23] &quot;iPSC&quot;                                          
[24] &quot;NPC&quot;                                           
[25] &quot;cells in culture&quot;                              
[26] &quot;mammary tumor xenograft&quot;                       
[27] &quot;Bone Marrow&quot;                                   
[28] &quot;Foreskin fibroblasts&quot;                          
[29] &quot;Human Fib X Human Fib Homokaryons&quot;             
[30] &quot;Fertile mature sperm&quot;                          </code></pre>
<p>Store all sources.</p>
<pre class="r"><code>my_idx &lt;- 1:human[[&#39;meta&#39;]][[&#39;samples&#39;]][[&#39;source_name_ch1&#39;]]$maxdims
my_source &lt;- human[[&#39;meta&#39;]][[&#39;samples&#39;]][[&#39;source_name_ch1&#39;]][my_idx]
head(my_source)</code></pre>
<pre><code>[1] &quot;Human DLBCL cel line&quot; &quot;Human DLBCL cel line&quot; &quot;Human DLBCL cel line&quot;
[4] &quot;Human DLBCL cel line&quot; &quot;Human DLBCL cel line&quot; &quot;Human DLBCL cel line&quot;</code></pre>
<p>Search.</p>
<pre class="r"><code>length(my_source[grepl(&quot;immune&quot;, my_source, ignore.case = TRUE)])</code></pre>
<pre><code>[1] 234</code></pre>
<pre class="r"><code>length(my_source[grepl(&quot;dendritic&quot;, my_source, ignore.case = TRUE)])</code></pre>
<pre><code>[1] 1799</code></pre>
</div>
<div id="expression" class="section level2">
<h2>Expression</h2>
<p>Genes are columns and rows are samples.</p>
<pre class="r"><code>exp_data &lt;- human[[&#39;data&#39;]][[&#39;expression&#39;]]
exp_data</code></pre>
<pre><code>Class: H5D
Dataset: /data/expression
Filename: /data2/archs4/human_gene_v2.5.h5
Access type: H5F_ACC_RDONLY
Datatype: H5T_STD_U32LE
Space: Type=Simple     Dims=888821 x 67186     Maxdims=Inf x 67186
Chunk: 1 x 2000</code></pre>
<p>Sum across first ten samples.</p>
<pre class="r"><code>apply(exp_data[1:10, ], 1, sum)</code></pre>
<pre><code> [1] 70714473 58300222 65226925 72034813 66846773 59968330   468037   519638
 [9]   599381   621721</code></pre>
<p>TP53 position.</p>
<pre class="r"><code>tp53_idx &lt;- match(&quot;TP53&quot;, my_gene_sym)
tp53_idx</code></pre>
<pre><code>[1] 8139</code></pre>
</div>
<div id="clean-up" class="section level2">
<h2>Clean up</h2>
<p>Close.</p>
<pre class="r"><code>human$close_all()</code></pre>
</div>
<div id="related" class="section level2">
<h2>Related</h2>
<ul>
<li><a href="https://github.com/MaayanLab/archs4py">Official ARCHS4
companion package</a>.</li>
<li><a href="https://maayanlab.cloud/archs4/help.html">ARCHS4 help
page</a></li>
</ul>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span>
Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 4.4.1 (2024-06-14)
Platform: x86_64-pc-linux-gnu
Running under: Ubuntu 22.04.5 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 
LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

time zone: Etc/UTC
tzcode source: system (glibc)

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] hdf5r_1.3.11    lubridate_1.9.3 forcats_1.0.0   stringr_1.5.1  
 [5] dplyr_1.1.4     purrr_1.0.2     readr_2.1.5     tidyr_1.3.1    
 [9] tibble_3.2.1    ggplot2_3.5.1   tidyverse_2.0.0 workflowr_1.7.1

loaded via a namespace (and not attached):
 [1] sass_0.4.9        utf8_1.2.4        generics_0.1.3    stringi_1.8.4    
 [5] hms_1.1.3         digest_0.6.37     magrittr_2.0.3    timechange_0.3.0 
 [9] evaluate_1.0.1    grid_4.4.1        fastmap_1.2.0     rprojroot_2.0.4  
[13] jsonlite_1.8.9    processx_3.8.4    whisker_0.4.1     ps_1.8.1         
[17] promises_1.3.0    httr_1.4.7        fansi_1.0.6       scales_1.3.0     
[21] jquerylib_0.1.4   cli_3.6.3         rlang_1.1.4       bit64_4.5.2      
[25] munsell_0.5.1     withr_3.0.2       cachem_1.1.0      yaml_2.3.10      
[29] tools_4.4.1       tzdb_0.4.0        colorspace_2.1-1  httpuv_1.6.15    
[33] vctrs_0.6.5       R6_2.5.1          lifecycle_1.0.4   git2r_0.35.0     
[37] bit_4.5.0         fs_1.6.4          pkgconfig_2.0.3   callr_3.7.6      
[41] pillar_1.9.0      bslib_0.8.0       later_1.3.2       gtable_0.3.6     
[45] glue_1.8.0        Rcpp_1.0.13       highr_0.11        xfun_0.48        
[49] tidyselect_1.2.1  rstudioapi_0.17.1 knitr_1.48        htmltools_0.5.8.1
[53] rmarkdown_2.28    compiler_4.4.1    getPass_0.2-4    </code></pre>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
https://docs.mathjax.org/en/latest/web/configuration.html. This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>





</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
