<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />



<meta name="date" content="2025-11-06" />

<title>JSON versus YAML</title>

<script src="site_libs/header-attrs-2.29/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.13.2/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-6.5.2/css/all.min.css" rel="stylesheet" />
<link href="site_libs/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet" />

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/main/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">My blog</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/davetang/muse">
    <span class="fa fa-github"></span>
     
    Source code
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">JSON versus YAML</h1>
<h4 class="date">2025-11-06</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span>
workflowr <span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2025-11-06
</p>
<p>
<strong>Checks:</strong> <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 7
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong> <code>muse/</code> <span
class="glyphicon glyphicon-question-sign" aria-hidden="true"
title="This is the local directory in which the code in this file was executed.">
</span>
</p>
<p>
This reproducible <a href="https://rmarkdown.rstudio.com">R Markdown</a>
analysis was created with <a
  href="https://github.com/workflowr/workflowr">workflowr</a> (version
1.7.1). The <em>Checks</em> tab describes the reproducibility checks
that were applied when the results were created. The <em>Past
versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date
</a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git
repository, you know the exact version of the code that produced these
results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the
global environment can affect the analysis in your R Markdown file in
unknown ways. For reproduciblity it’s best to always run the code in an
empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20200712code">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Seed:</strong>
<code>set.seed(20200712)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20200712code"
class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20200712)</code> was run prior to running
the code in the R Markdown file. Setting a seed ensures that any results
that rely on randomness, e.g. subsampling or permutations, are
reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Session information:</strong>
recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package
versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be
confident that you successfully produced the results during this
run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr
project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomdavetangmusetree4ca8b1aa9fbdbd134934acd2a8fae75285f6e9f8targetblank4ca8b1aa">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Repository version:</strong>
<a href="https://github.com/davetang/muse/tree/4ca8b1aa9fbdbd134934acd2a8fae75285f6e9f8" target="_blank">4ca8b1a</a>
</a>
</p>
</div>
<div
id="strongRepositoryversionstrongahrefhttpsgithubcomdavetangmusetree4ca8b1aa9fbdbd134934acd2a8fae75285f6e9f8targetblank4ca8b1aa"
class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development
and connecting the code version to the results is critical for
reproducibility.
</p>
<p>
The results in this page were generated with repository version
<a href="https://github.com/davetang/muse/tree/4ca8b1aa9fbdbd134934acd2a8fae75285f6e9f8" target="_blank">4ca8b1a</a>.
See the <em>Past versions</em> tab to see a history of the changes made
to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for
the analysis have been committed to Git prior to generating the results
(you can use <code>wflow_publish</code> or
<code>wflow_git_commit</code>). workflowr only checks the R Markdown
file, but you know if there are other scripts or data files that it
depends on. Below is the status of the Git repository when the results
were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .Rproj.user/
    Ignored:    data/1M_neurons_filtered_gene_bc_matrices_h5.h5
    Ignored:    data/293t/
    Ignored:    data/293t_3t3_filtered_gene_bc_matrices.tar.gz
    Ignored:    data/293t_filtered_gene_bc_matrices.tar.gz
    Ignored:    data/5k_Human_Donor1_PBMC_3p_gem-x_5k_Human_Donor1_PBMC_3p_gem-x_count_sample_filtered_feature_bc_matrix.h5
    Ignored:    data/5k_Human_Donor2_PBMC_3p_gem-x_5k_Human_Donor2_PBMC_3p_gem-x_count_sample_filtered_feature_bc_matrix.h5
    Ignored:    data/5k_Human_Donor3_PBMC_3p_gem-x_5k_Human_Donor3_PBMC_3p_gem-x_count_sample_filtered_feature_bc_matrix.h5
    Ignored:    data/5k_Human_Donor4_PBMC_3p_gem-x_5k_Human_Donor4_PBMC_3p_gem-x_count_sample_filtered_feature_bc_matrix.h5
    Ignored:    data/97516b79-8d08-46a6-b329-5d0a25b0be98.h5ad
    Ignored:    data/Parent_SC3v3_Human_Glioblastoma_filtered_feature_bc_matrix.tar.gz
    Ignored:    data/brain_counts/
    Ignored:    data/cl.obo
    Ignored:    data/cl.owl
    Ignored:    data/jurkat/
    Ignored:    data/jurkat:293t_50:50_filtered_gene_bc_matrices.tar.gz
    Ignored:    data/jurkat_293t/
    Ignored:    data/jurkat_filtered_gene_bc_matrices.tar.gz
    Ignored:    data/pbmc20k/
    Ignored:    data/pbmc20k_seurat/
    Ignored:    data/pbmc3k.h5ad
    Ignored:    data/pbmc3k/
    Ignored:    data/pbmc3k_bpcells_mat/
    Ignored:    data/pbmc3k_export.mtx
    Ignored:    data/pbmc3k_matrix.mtx
    Ignored:    data/pbmc3k_seurat.rds
    Ignored:    data/pbmc4k_filtered_gene_bc_matrices.tar.gz
    Ignored:    data/pbmc_1k_v3_filtered_feature_bc_matrix.h5
    Ignored:    data/pbmc_1k_v3_raw_feature_bc_matrix.h5
    Ignored:    data/refdata-gex-GRCh38-2020-A.tar.gz
    Ignored:    data/seurat_1m_neuron.rds
    Ignored:    data/t_3k_filtered_gene_bc_matrices.tar.gz
    Ignored:    r_packages_4.4.1/
    Ignored:    r_packages_4.5.0/

Untracked files:
    Untracked:  analysis/bioc_scrnaseq.Rmd
    Untracked:  bpcells_matrix/
    Untracked:  data/Caenorhabditis_elegans.WBcel235.113.gtf.gz
    Untracked:  data/GCF_043380555.1-RS_2024_12_gene_ontology.gaf.gz
    Untracked:  data/arab.rds
    Untracked:  data/astronomicalunit.csv
    Untracked:  data/femaleMiceWeights.csv
    Untracked:  m3/
    Untracked:  women.json

Unstaged changes:
    Modified:   analysis/isoform_switch_analyzer.Rmd

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not
included in this status report because it is ok for generated content to
have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">
<p>
These are the previous versions of the repository in which changes were
made to the R Markdown (<code>analysis/json_vs_yaml.Rmd</code>) and HTML
(<code>docs/json_vs_yaml.html</code>) files. If you’ve configured a
remote Git repository (see <code>?wflow_git_remote</code>), click on the
hyperlinks in the table below to view the files as they were in that
past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/davetang/muse/blob/4ca8b1aa9fbdbd134934acd2a8fae75285f6e9f8/analysis/json_vs_yaml.Rmd" target="_blank">4ca8b1a</a>
</td>
<td>
Dave Tang
</td>
<td>
2025-11-06
</td>
<td>
Tweak list_flatten_all()
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/davetang/muse/3f3aa2213889b8afa1fa935d101bbe8521b6ae87/docs/json_vs_yaml.html" target="_blank">3f3aa22</a>
</td>
<td>
Dave Tang
</td>
<td>
2025-11-06
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/davetang/muse/blob/431f1abb3aa8c5fb9af59e97e42f481bbc29b32f/analysis/json_vs_yaml.Rmd" target="_blank">431f1ab</a>
</td>
<td>
Dave Tang
</td>
<td>
2025-11-06
</td>
<td>
Recursively flatten a list using purrr
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/davetang/muse/c5ae3bc03f190e288da54b8a225cf6a2e996eec8/docs/json_vs_yaml.html" target="_blank">c5ae3bc</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-06-10
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/davetang/muse/blob/eaa2acae7e13fcc8b4ef2d09f6df9a2f60524f45/analysis/json_vs_yaml.Rmd" target="_blank">eaa2aca</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-06-10
</td>
<td>
Load nested lists
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/davetang/muse/097bb550125ccf5789deb13c5892ff6cda2d6d4d/docs/json_vs_yaml.html" target="_blank">097bb55</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-06-10
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/davetang/muse/blob/38bfda109c467a51485f2ff899eafaa8d99b1a85/analysis/json_vs_yaml.Rmd" target="_blank">38bfda1</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-06-10
</td>
<td>
YAML defined
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/davetang/muse/97e60dd6b77c722436311e3b7dfd20d4b5d5f70d/docs/json_vs_yaml.html" target="_blank">97e60dd</a>
</td>
<td>
Dave Tang
</td>
<td>
2023-06-19
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/davetang/muse/blob/032e05867ea3c9d694d484b30ea0be4ee23b5d86/analysis/json_vs_yaml.Rmd" target="_blank">032e058</a>
</td>
<td>
Dave Tang
</td>
<td>
2023-06-19
</td>
<td>
More details on JSON
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/davetang/muse/643530fd4fe102b150f9e4da6e9538a1895a8ebd/docs/json_vs_yaml.html" target="_blank">643530f</a>
</td>
<td>
Dave Tang
</td>
<td>
2023-06-14
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/davetang/muse/blob/c7dcdf2fbdb48445e1909a69edb7a308c1004128/analysis/json_vs_yaml.Rmd" target="_blank">c7dcdf2</a>
</td>
<td>
Dave Tang
</td>
<td>
2023-06-14
</td>
<td>
Parsing JSON
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/davetang/muse/ca51b55320e809449923f5dc79ddfe9ae961b90a/docs/json_vs_yaml.html" target="_blank">ca51b55</a>
</td>
<td>
Dave Tang
</td>
<td>
2023-05-24
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/davetang/muse/blob/b72b5b38c67a74d70b76ff0f246e466ae4427abb/analysis/json_vs_yaml.Rmd" target="_blank">b72b5b3</a>
</td>
<td>
Dave Tang
</td>
<td>
2023-05-24
</td>
<td>
Check out additional packages
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/davetang/muse/3bb32451d34101733fe138231c3f76d09b3cacab/docs/json_vs_yaml.html" target="_blank">3bb3245</a>
</td>
<td>
Dave Tang
</td>
<td>
2023-05-19
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/davetang/muse/blob/50c92d7216f8e09adcaa390b558e4647b5d6fcbd/analysis/json_vs_yaml.Rmd" target="_blank">50c92d7</a>
</td>
<td>
Dave Tang
</td>
<td>
2023-05-19
</td>
<td>
JSON and YAML formats
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<p>JSON and YAML are popular <a
href="https://en.wikipedia.org/wiki/Serialization">serialisation</a>
formats.</p>
<blockquote>
<p>In computing, serialization (or serialisation) is the process of
translating a data structure or object state into a format that can be
stored (e.g. files in secondary storage devices, data buffers in primary
storage devices) or transmitted (e.g. data streams over computer
networks) and reconstructed later (possibly in a different computer
environment).</p>
</blockquote>
<p>Install the following packages:</p>
<ul>
<li><a href="https://github.com/jeroen/jsonlite">jsonlite</a> for
parsing JSON</li>
<li><a href="https://github.com/vubiostat/r-yaml">yaml</a> for parsing
YAML</li>
<li><a href="https://github.com/colearendt/tidyjson">tidyjson</a> for
converting JSON to tidy data frames</li>
<li><a href="https://github.com/alexcb/rjson">rjson</a> for parsing
JSON</li>
</ul>
<pre class="r"><code>install.packages(c(&quot;jsonlite&quot;, &quot;yaml&quot;, &quot;tidyjson&quot;, &quot;rjson&quot;))</code></pre>
<p>Load libraries.</p>
<pre class="r"><code>library(jsonlite)
library(yaml)
library(tidyjson)</code></pre>
<pre><code>
Attaching package: &#39;tidyjson&#39;</code></pre>
<pre><code>The following object is masked from &#39;package:jsonlite&#39;:

    read_json</code></pre>
<pre><code>The following object is masked from &#39;package:stats&#39;:

    filter</code></pre>
<pre class="r"><code>library(rjson)</code></pre>
<pre><code>
Attaching package: &#39;rjson&#39;</code></pre>
<pre><code>The following objects are masked from &#39;package:jsonlite&#39;:

    fromJSON, toJSON</code></pre>
<p>As a first example, we will convert the <code>women</code> data set,
which is a small data set with 15 observations for 2 variables.</p>
<pre class="r"><code>women</code></pre>
<pre><code>   height weight
1      58    115
2      59    117
3      60    120
4      61    123
5      62    126
6      63    129
7      64    132
8      65    135
9      66    139
10     67    142
11     68    146
12     69    150
13     70    154
14     71    159
15     72    164</code></pre>
<p>Convert <code>women</code> to JSON using <code>jsonlite</code>.</p>
<pre class="r"><code>women_json &lt;- jsonlite::toJSON(women, pretty = TRUE)
women_json</code></pre>
<pre><code>[
  {
    &quot;height&quot;: 58,
    &quot;weight&quot;: 115
  },
  {
    &quot;height&quot;: 59,
    &quot;weight&quot;: 117
  },
  {
    &quot;height&quot;: 60,
    &quot;weight&quot;: 120
  },
  {
    &quot;height&quot;: 61,
    &quot;weight&quot;: 123
  },
  {
    &quot;height&quot;: 62,
    &quot;weight&quot;: 126
  },
  {
    &quot;height&quot;: 63,
    &quot;weight&quot;: 129
  },
  {
    &quot;height&quot;: 64,
    &quot;weight&quot;: 132
  },
  {
    &quot;height&quot;: 65,
    &quot;weight&quot;: 135
  },
  {
    &quot;height&quot;: 66,
    &quot;weight&quot;: 139
  },
  {
    &quot;height&quot;: 67,
    &quot;weight&quot;: 142
  },
  {
    &quot;height&quot;: 68,
    &quot;weight&quot;: 146
  },
  {
    &quot;height&quot;: 69,
    &quot;weight&quot;: 150
  },
  {
    &quot;height&quot;: 70,
    &quot;weight&quot;: 154
  },
  {
    &quot;height&quot;: 71,
    &quot;weight&quot;: 159
  },
  {
    &quot;height&quot;: 72,
    &quot;weight&quot;: 164
  }
] </code></pre>
<p><code>read_json</code> does not parse the output of
<code>toJSON</code>.</p>
<pre class="r"><code>jsonlite::write_json(x = women_json, path = &quot;women.json&quot;)
tidyjson::read_json(path = &quot;women.json&quot;)</code></pre>
<pre><code># A tbl_json: 1 x 2 tibble with a &quot;JSON&quot; attribute
  ..JSON                   document.id
  &lt;chr&gt;                          &lt;int&gt;
1 &quot;[\&quot;[\\n  {\\n    \\...&quot;           1</code></pre>
<p>Converts into list.</p>
<pre class="r"><code>str(rjson::fromJSON(women_json))</code></pre>
<pre><code>List of 15
 $ :List of 2
  ..$ height: num 58
  ..$ weight: num 115
 $ :List of 2
  ..$ height: num 59
  ..$ weight: num 117
 $ :List of 2
  ..$ height: num 60
  ..$ weight: num 120
 $ :List of 2
  ..$ height: num 61
  ..$ weight: num 123
 $ :List of 2
  ..$ height: num 62
  ..$ weight: num 126
 $ :List of 2
  ..$ height: num 63
  ..$ weight: num 129
 $ :List of 2
  ..$ height: num 64
  ..$ weight: num 132
 $ :List of 2
  ..$ height: num 65
  ..$ weight: num 135
 $ :List of 2
  ..$ height: num 66
  ..$ weight: num 139
 $ :List of 2
  ..$ height: num 67
  ..$ weight: num 142
 $ :List of 2
  ..$ height: num 68
  ..$ weight: num 146
 $ :List of 2
  ..$ height: num 69
  ..$ weight: num 150
 $ :List of 2
  ..$ height: num 70
  ..$ weight: num 154
 $ :List of 2
  ..$ height: num 71
  ..$ weight: num 159
 $ :List of 2
  ..$ height: num 72
  ..$ weight: num 164</code></pre>
<p>Convert <code>women</code> to YAML.</p>
<pre class="r"><code>women_yaml &lt;- as.yaml(women, indent = 3)
writeLines(women_yaml)</code></pre>
<pre><code>height:
- 58.0
- 59.0
- 60.0
- 61.0
- 62.0
- 63.0
- 64.0
- 65.0
- 66.0
- 67.0
- 68.0
- 69.0
- 70.0
- 71.0
- 72.0
weight:
- 115.0
- 117.0
- 120.0
- 123.0
- 126.0
- 129.0
- 132.0
- 135.0
- 139.0
- 142.0
- 146.0
- 150.0
- 154.0
- 159.0
- 164.0</code></pre>
<div id="to-data-frame" class="section level2">
<h2>To data frame</h2>
<p>JSON to data frame.</p>
<pre class="r"><code>jsonlite::fromJSON(women_json)</code></pre>
<pre><code>   height weight
1      58    115
2      59    117
3      60    120
4      61    123
5      62    126
6      63    129
7      64    132
8      65    135
9      66    139
10     67    142
11     68    146
12     69    150
13     70    154
14     71    159
15     72    164</code></pre>
<p>YAML to data frame. This does not work for more complex data
structures (see below).</p>
<pre class="r"><code>yaml.load(women_yaml, handlers = list(map = function(x) as.data.frame(x) ))</code></pre>
<pre><code>   height weight
1      58    115
2      59    117
3      60    120
4      61    123
5      62    126
6      63    129
7      64    132
8      65    135
9      66    139
10     67    142
11     68    146
12     69    150
13     70    154
14     71    159
15     72    164</code></pre>
</div>
<div id="non-tidy-data-frame" class="section level2">
<h2>Non-tidy data frame</h2>
<p>A data frame containing lists.</p>
<pre class="r"><code>my_df &lt;- data.frame(
  id = 1:3,
  title = letters[1:3]
)
my_df$keywords = list(
    c(&#39;aa&#39;, &#39;aaa&#39;, &#39;aaaa&#39;),
    c(&#39;bb&#39;, &#39;bbb&#39;),
    c(&#39;cc&#39;, &#39;ccc&#39;, &#39;cccc&#39;, &#39;ccccc&#39;)
)

my_df</code></pre>
<pre><code>  id title             keywords
1  1     a        aa, aaa, aaaa
2  2     b              bb, bbb
3  3     c cc, ccc, cccc, ccccc</code></pre>
<p>Convert <code>my_df</code> to JSON.</p>
<pre class="r"><code>my_df_json &lt;- jsonlite::toJSON(my_df, pretty = TRUE)
my_df_json</code></pre>
<pre><code>[
  {
    &quot;id&quot;: 1,
    &quot;title&quot;: &quot;a&quot;,
    &quot;keywords&quot;: [&quot;aa&quot;, &quot;aaa&quot;, &quot;aaaa&quot;]
  },
  {
    &quot;id&quot;: 2,
    &quot;title&quot;: &quot;b&quot;,
    &quot;keywords&quot;: [&quot;bb&quot;, &quot;bbb&quot;]
  },
  {
    &quot;id&quot;: 3,
    &quot;title&quot;: &quot;c&quot;,
    &quot;keywords&quot;: [&quot;cc&quot;, &quot;ccc&quot;, &quot;cccc&quot;, &quot;ccccc&quot;]
  }
] </code></pre>
<p>Convert <code>my_df</code> to YAML.</p>
<pre class="r"><code>my_df_yaml &lt;- as.yaml(my_df, indent = 3)
writeLines(my_df_yaml)</code></pre>
<pre><code>id:
- 1
- 2
- 3
title:
- a
- b
- c
keywords:
-  - aa
   - aaa
   - aaaa
-  - bb
   - bbb
-  - cc
   - ccc
   - cccc
   - ccccc</code></pre>
</div>
<div id="json-to-yaml-and-vice-versa" class="section level2">
<h2>JSON to YAML and vice versa</h2>
<p>Converting from JSON to YAML is easy.</p>
<pre class="r"><code>identical(writeLines(as.yaml(jsonlite::fromJSON(my_df_json))), writeLines(my_df_yaml))</code></pre>
<pre><code>id:
- 1
- 2
- 3
title:
- a
- b
- c
keywords:
- - aa
  - aaa
  - aaaa
- - bb
  - bbb
- - cc
  - ccc
  - cccc
  - ccccc

id:
- 1
- 2
- 3
title:
- a
- b
- c
keywords:
-  - aa
   - aaa
   - aaaa
-  - bb
   - bbb
-  - cc
   - ccc
   - cccc
   - ccccc</code></pre>
<pre><code>[1] TRUE</code></pre>
<p>Converting from YAML to JSON for <code>my_df</code> is not as
straight-forward because of the different number of keywords.</p>
<pre class="r"><code>my_df_list &lt;- yaml.load(my_df_yaml)
my_df_list</code></pre>
<pre><code>$id
[1] 1 2 3

$title
[1] &quot;a&quot; &quot;b&quot; &quot;c&quot;

$keywords
$keywords[[1]]
[1] &quot;aa&quot;   &quot;aaa&quot;  &quot;aaaa&quot;

$keywords[[2]]
[1] &quot;bb&quot;  &quot;bbb&quot;

$keywords[[3]]
[1] &quot;cc&quot;    &quot;ccc&quot;   &quot;cccc&quot;  &quot;ccccc&quot;</code></pre>
<p>This conversion is different from the original data frame to JSON
conversion because this creates a single object, where as the original
conversion creates an array with three objects.</p>
<pre class="r"><code>jsonlite::toJSON(my_df_list, pretty = TRUE)</code></pre>
<pre><code>{
  &quot;id&quot;: [1, 2, 3],
  &quot;title&quot;: [&quot;a&quot;, &quot;b&quot;, &quot;c&quot;],
  &quot;keywords&quot;: [
    [&quot;aa&quot;, &quot;aaa&quot;, &quot;aaaa&quot;],
    [&quot;bb&quot;, &quot;bbb&quot;],
    [&quot;cc&quot;, &quot;ccc&quot;, &quot;cccc&quot;, &quot;ccccc&quot;]
  ]
} </code></pre>
<pre class="r"><code>my_df_json</code></pre>
<pre><code>[
  {
    &quot;id&quot;: 1,
    &quot;title&quot;: &quot;a&quot;,
    &quot;keywords&quot;: [&quot;aa&quot;, &quot;aaa&quot;, &quot;aaaa&quot;]
  },
  {
    &quot;id&quot;: 2,
    &quot;title&quot;: &quot;b&quot;,
    &quot;keywords&quot;: [&quot;bb&quot;, &quot;bbb&quot;]
  },
  {
    &quot;id&quot;: 3,
    &quot;title&quot;: &quot;c&quot;,
    &quot;keywords&quot;: [&quot;cc&quot;, &quot;ccc&quot;, &quot;cccc&quot;, &quot;ccccc&quot;]
  }
] </code></pre>
<p>I could probably write a hacky function to make the conversion but I
won’t.</p>
</div>
<div id="parsing-json" class="section level2">
<h2>Parsing JSON</h2>
<p>The <a href="https://github.com/pachterlab/ffq">ffq</a> tool
generates metadata in JSON:</p>
<pre><code>ffq SRX079566 &gt; data/SRX079566.json</code></pre>
<pre class="r"><code>ffq_json &lt;- jsonlite::read_json(path = &quot;data/SRX079566.json&quot;, simplifyVector = TRUE)
str(ffq_json)</code></pre>
<pre><code>List of 1
 $ SRX079566:List of 5
  ..$ accession : chr &quot;SRX079566&quot;
  ..$ title     : chr &quot;Illumina Genome Analyzer IIx paired end sequencing; RNA-Seq (polyA+) analysis of DLBCL cell line HS0798&quot;
  ..$ platform  : chr &quot;ILLUMINA&quot;
  ..$ instrument: chr &quot;Illumina Genome Analyzer IIx&quot;
  ..$ runs      :List of 2
  .. ..$ SRR292241:List of 7
  .. .. ..$ accession : chr &quot;SRR292241&quot;
  .. .. ..$ experiment: chr &quot;SRX079566&quot;
  .. .. ..$ study     : chr &quot;SRP020237&quot;
  .. .. ..$ sample    : chr &quot;SRS212581&quot;
  .. .. ..$ title     : chr &quot;Illumina Genome Analyzer IIx paired end sequencing; RNA-Seq (polyA+) analysis of DLBCL cell line HS0798&quot;
  .. .. ..$ attributes:List of 6
  .. .. .. ..$ RUN             : chr &quot;94367&quot;
  .. .. .. ..$ instrument_model: chr &quot;Illumina Genome Analyzer II&quot;
  .. .. .. ..$ ENA-SPOT-COUNT  : int 9721384
  .. .. .. ..$ ENA-BASE-COUNT  : int 699939648
  .. .. .. ..$ ENA-FIRST-PUBLIC: chr &quot;2011-07-05&quot;
  .. .. .. ..$ ENA-LAST-UPDATE : chr &quot;2019-10-07&quot;
  .. .. ..$ files     :List of 4
  .. .. .. ..$ ftp :&#39;data.frame&#39;:   2 obs. of  8 variables:
  .. .. .. .. ..$ accession : chr [1:2] &quot;SRR292241&quot; &quot;SRR292241&quot;
  .. .. .. .. ..$ filename  : chr [1:2] &quot;SRR292241_1.fastq.gz&quot; &quot;SRR292241_2.fastq.gz&quot;
  .. .. .. .. ..$ filetype  : chr [1:2] &quot;fastq&quot; &quot;fastq&quot;
  .. .. .. .. ..$ filesize  : int [1:2] 387227151 395115704
  .. .. .. .. ..$ filenumber: int [1:2] 1 2
  .. .. .. .. ..$ md5       : chr [1:2] &quot;a5e0d2d51550127ea9ce3a0219deb375&quot; &quot;e9ce7abd3bce9d5ff194d6e045a36c1c&quot;
  .. .. .. .. ..$ urltype   : chr [1:2] &quot;ftp&quot; &quot;ftp&quot;
  .. .. .. .. ..$ url       : chr [1:2] &quot;ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR292/SRR292241/SRR292241_1.fastq.gz&quot; &quot;ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR292/SRR292241/SRR292241_2.fastq.gz&quot;
  .. .. .. ..$ aws :&#39;data.frame&#39;:   2 obs. of  8 variables:
  .. .. .. .. ..$ accession : chr [1:2] &quot;SRR292241&quot; &quot;SRR292241&quot;
  .. .. .. .. ..$ filename  : chr [1:2] &quot;Run94367Lane6.srf&quot; &quot;SRR292241&quot;
  .. .. .. .. ..$ filetype  : chr [1:2] &quot;sra&quot; &quot;sra&quot;
  .. .. .. .. ..$ filesize  : logi [1:2] NA NA
  .. .. .. .. ..$ filenumber: int [1:2] 1 1
  .. .. .. .. ..$ md5       : logi [1:2] NA NA
  .. .. .. .. ..$ urltype   : chr [1:2] &quot;aws&quot; &quot;aws&quot;
  .. .. .. .. ..$ url       : chr [1:2] &quot;s3://sra-pub-src-13/SRR292241/Run94367Lane6.srf&quot; &quot;https://sra-pub-run-odp.s3.amazonaws.com/sra/SRR292241/SRR292241&quot;
  .. .. .. ..$ gcp :&#39;data.frame&#39;:   1 obs. of  8 variables:
  .. .. .. .. ..$ accession : chr &quot;SRR292241&quot;
  .. .. .. .. ..$ filename  : chr &quot;SRR292241.3&quot;
  .. .. .. .. ..$ filetype  : chr &quot;sra&quot;
  .. .. .. .. ..$ filesize  : logi NA
  .. .. .. .. ..$ filenumber: int 1
  .. .. .. .. ..$ md5       : logi NA
  .. .. .. .. ..$ urltype   : chr &quot;gcp&quot;
  .. .. .. .. ..$ url       : chr &quot;gs://sra-pub-crun-3/SRR292241/SRR292241.3&quot;
  .. .. .. ..$ ncbi:&#39;data.frame&#39;:   1 obs. of  8 variables:
  .. .. .. .. ..$ accession : chr &quot;SRR292241&quot;
  .. .. .. .. ..$ filename  : chr &quot;SRR292241.3&quot;
  .. .. .. .. ..$ filetype  : chr &quot;sra&quot;
  .. .. .. .. ..$ filesize  : logi NA
  .. .. .. .. ..$ filenumber: int 1
  .. .. .. .. ..$ md5       : logi NA
  .. .. .. .. ..$ urltype   : chr &quot;ncbi&quot;
  .. .. .. .. ..$ url       : chr &quot;https://sra-downloadb.be-md.ncbi.nlm.nih.gov/sos5/sra-pub-run-32/SRR000/292/SRR292241/SRR292241.3&quot;
  .. ..$ SRR390728:List of 7
  .. .. ..$ accession : chr &quot;SRR390728&quot;
  .. .. ..$ experiment: chr &quot;SRX079566&quot;
  .. .. ..$ study     : chr &quot;SRP020237&quot;
  .. .. ..$ sample    : chr &quot;SRS212581&quot;
  .. .. ..$ title     : chr &quot;Illumina Genome Analyzer IIx paired end sequencing; RNA-Seq (polyA+) analysis of DLBCL cell line HS0798&quot;
  .. .. ..$ attributes:List of 6
  .. .. .. ..$ RUN             : chr &quot;94367&quot;
  .. .. .. ..$ assembly        : chr &quot;NCBI36_BCCAGSC_variant&quot;
  .. .. .. ..$ ENA-SPOT-COUNT  : int 7178576
  .. .. .. ..$ ENA-BASE-COUNT  : int 516857472
  .. .. .. ..$ ENA-FIRST-PUBLIC: chr &quot;2011-12-23&quot;
  .. .. .. ..$ ENA-LAST-UPDATE : chr &quot;2016-06-28&quot;
  .. .. ..$ files     :List of 4
  .. .. .. ..$ ftp :&#39;data.frame&#39;:   2 obs. of  8 variables:
  .. .. .. .. ..$ accession : chr [1:2] &quot;SRR390728&quot; &quot;SRR390728&quot;
  .. .. .. .. ..$ filename  : chr [1:2] &quot;SRR390728_1.fastq.gz&quot; &quot;SRR390728_2.fastq.gz&quot;
  .. .. .. .. ..$ filetype  : chr [1:2] &quot;fastq&quot; &quot;fastq&quot;
  .. .. .. .. ..$ filesize  : int [1:2] 170346275 168836179
  .. .. .. .. ..$ filenumber: int [1:2] 1 2
  .. .. .. .. ..$ md5       : chr [1:2] &quot;9a3d37cbb3e47cf8930ed2ba6c8d2cef&quot; &quot;bc4e6304170876186522f4175ee39a8f&quot;
  .. .. .. .. ..$ urltype   : chr [1:2] &quot;ftp&quot; &quot;ftp&quot;
  .. .. .. .. ..$ url       : chr [1:2] &quot;ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR390/SRR390728/SRR390728_1.fastq.gz&quot; &quot;ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR390/SRR390728/SRR390728_2.fastq.gz&quot;
  .. .. .. ..$ aws :&#39;data.frame&#39;:   2 obs. of  8 variables:
  .. .. .. .. ..$ accession : chr [1:2] &quot;SRR390728&quot; &quot;SRR390728&quot;
  .. .. .. .. ..$ filename  : chr [1:2] &quot;30KWMAAXX_6.sorted_withJunctionsOnGenome_dupsFlagged.bam.1&quot; &quot;SRR390728&quot;
  .. .. .. .. ..$ filetype  : chr [1:2] &quot;bam&quot; &quot;sra&quot;
  .. .. .. .. ..$ filesize  : logi [1:2] NA NA
  .. .. .. .. ..$ filenumber: int [1:2] 1 1
  .. .. .. .. ..$ md5       : logi [1:2] NA NA
  .. .. .. .. ..$ urltype   : chr [1:2] &quot;aws&quot; &quot;aws&quot;
  .. .. .. .. ..$ url       : chr [1:2] &quot;s3://sra-pub-src-15/SRR390728/30KWMAAXX_6.sorted_withJunctionsOnGenome_dupsFlagged.bam.1&quot; &quot;https://sra-pub-run-odp.s3.amazonaws.com/sra/SRR390728/SRR390728&quot;
  .. .. .. ..$ gcp :&#39;data.frame&#39;:   1 obs. of  8 variables:
  .. .. .. .. ..$ accession : chr &quot;SRR390728&quot;
  .. .. .. .. ..$ filename  : chr &quot;SRR390728.lite.2&quot;
  .. .. .. .. ..$ filetype  : chr &quot;sra&quot;
  .. .. .. .. ..$ filesize  : logi NA
  .. .. .. .. ..$ filenumber: int 1
  .. .. .. .. ..$ md5       : logi NA
  .. .. .. .. ..$ urltype   : chr &quot;gcp&quot;
  .. .. .. .. ..$ url       : chr &quot;gs://sra-pub-zq-5/SRR390728/SRR390728.lite.2&quot;
  .. .. .. ..$ ncbi:&#39;data.frame&#39;:   1 obs. of  8 variables:
  .. .. .. .. ..$ accession : chr &quot;SRR390728&quot;
  .. .. .. .. ..$ filename  : chr &quot;SRR390728.lite.2&quot;
  .. .. .. .. ..$ filetype  : chr &quot;sra&quot;
  .. .. .. .. ..$ filesize  : logi NA
  .. .. .. .. ..$ filenumber: int 1
  .. .. .. .. ..$ md5       : logi NA
  .. .. .. .. ..$ urltype   : chr &quot;ncbi&quot;
  .. .. .. .. ..$ url       : chr &quot;https://sra-downloadb.be-md.ncbi.nlm.nih.gov/sos2/sra-pub-zq-20/SRR000/390/SRR390728/SRR390728.lite.2&quot;</code></pre>
<pre class="r"><code>str(ffq_json, max.level = 3)</code></pre>
<pre><code>List of 1
 $ SRX079566:List of 5
  ..$ accession : chr &quot;SRX079566&quot;
  ..$ title     : chr &quot;Illumina Genome Analyzer IIx paired end sequencing; RNA-Seq (polyA+) analysis of DLBCL cell line HS0798&quot;
  ..$ platform  : chr &quot;ILLUMINA&quot;
  ..$ instrument: chr &quot;Illumina Genome Analyzer IIx&quot;
  ..$ runs      :List of 2
  .. ..$ SRR292241:List of 7
  .. ..$ SRR390728:List of 7</code></pre>
<p>Use a recursive apply to create a named character vector, which is
convenient for plucking values.</p>
<pre class="r"><code>test &lt;- rapply(object = ffq_json, f = function(x) x)
class(test)</code></pre>
<pre><code>[1] &quot;character&quot;</code></pre>
<p>Subset the FTP links.</p>
<pre class="r"><code>test[grepl(&quot;ftp.url\\d+$&quot;, names(test))]</code></pre>
<pre><code>                                   SRX079566.runs.SRR292241.files.ftp.url1 
&quot;ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR292/SRR292241/SRR292241_1.fastq.gz&quot; 
                                   SRX079566.runs.SRR292241.files.ftp.url2 
&quot;ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR292/SRR292241/SRR292241_2.fastq.gz&quot; 
                                   SRX079566.runs.SRR390728.files.ftp.url1 
&quot;ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR390/SRR390728/SRR390728_1.fastq.gz&quot; 
                                   SRX079566.runs.SRR390728.files.ftp.url2 
&quot;ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR390/SRR390728/SRR390728_2.fastq.gz&quot; </code></pre>
<p>Using {purrr} to recursively flatten.</p>
<pre class="r"><code>list_flatten_all &lt;- function(x) {
  if (!is.list(x)) return(x)
  purrr::map(x, list_flatten_all) |&gt; purrr::list_flatten()
}

ffq_json_flat &lt;- list_flatten_all(ffq_json)
ffq_json_flat[grep(&quot;ftp_url$&quot;, names(ffq_json_flat))] |&gt;
  unlist()</code></pre>
<pre><code>                                   SRX079566_runs_SRR292241_files_ftp_url1 
&quot;ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR292/SRR292241/SRR292241_1.fastq.gz&quot; 
                                   SRX079566_runs_SRR292241_files_ftp_url2 
&quot;ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR292/SRR292241/SRR292241_2.fastq.gz&quot; 
                                   SRX079566_runs_SRR390728_files_ftp_url1 
&quot;ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR390/SRR390728/SRR390728_1.fastq.gz&quot; 
                                   SRX079566_runs_SRR390728_files_ftp_url2 
&quot;ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR390/SRR390728/SRR390728_2.fastq.gz&quot; </code></pre>
</div>
<div id="appendix" class="section level2">
<h2>Appendix</h2>
<div id="yaml-defined" class="section level3">
<h3>YAML defined</h3>
<p><a
href="https://www.freecodecamp.org/news/what-is-yaml-the-yml-file-format/">What
is YAML? The YML File Format</a>.</p>
<ul>
<li>YAML created in 2001 stands for YAML Ain’t Markup Language, but it
originally stood for Yet Another Markup Language.</li>
<li>YAML is a human-readable data serialisation language, just like XML
and JSON.
<ul>
<li>YAML can do everything that JSON can and more. A valid YAML file can
contain JSON, and JSON can transform into YAML.</li>
</ul></li>
<li>Serialisation is a process where one application or service, which
has different data structures and is written in a different set of
technologies, can transfer data to another application using a standard
format.</li>
<li>In other words, serialisation is about translating, converting, and
wrapping up a data structure in another format.</li>
<li>The data in the new format can be stored in a file or transmitted to
another application or service over a network.</li>
<li>YAML is a widely used format for writing configuration files for
different DevOps tools, programs, and applications because of its
human-readable and intuitive syntax.</li>
<li>YAML uses indentation to define structure in the file, which is
helpful if you are used to writing Python code and are familiar with the
indentation style the language uses.</li>
</ul>
<p>To create a YAML file, use either the <code>.yaml</code> or
<code>.yml</code> file extension. Before writing any YAML code, you can
add three dashes (<code>---</code>) at the start of the file to allow
having multiple YAML documents in a single YAML file, making file
organisation much easier; separate each document with three dashes
(<code>---</code>). You can use three dots (<code>...</code>) to mark
the end of the document.</p>
<ul>
<li>In YAML, there is an emphasis on indentation and line separation to
denote levels and structure in data. The indentation system is quite
similar to the one Python uses.</li>
<li>YAML doesn’t use symbols such as curly braces, square brackets, or
opening or closing tags - just indentation.</li>
<li>YAML doesn’t allow you to use any tabs when creating indentation -
use spaces instead.</li>
<li>Whitespace doesn’t matter as long as child elements are indented
inside the parent element.</li>
<li>To add a comment to comment out a line of code, use the
<code>#</code> character</li>
</ul>
<p>Although YAML auto-detects the data types in a file, you can specify
the type of data you want to use. To explicitly specify the type of
data, use the <code>!!</code> symbol and the name of the data type
before the value:</p>
<pre><code># parse this value as a string
date: !!str 2022-11-11

## parse this value as a float (it will be 1.0 instead of 1)
fave_number: !!float 1</code></pre>
<p>In YAML, strings in some cases can be left unquoted, but you can also
wrap them in single (<code>'</code>) or double (<code>"</code>)
quotation marks.</p>
<p>If you want to write a string that spans across multiple lines and
you want to preserve the line breaks, use the pipe symbol
(<code>|</code>) and make sure that the message is indented!</p>
<pre><code>|
 I am message that spans multiple lines
 I go on and on across lines
 and lines
 and more lines</code></pre>
<p>If you have a string in a YAML file that spans across multiple lines
for readability, but you want the parser to interpret it as a single
line string, you can use the <code>&gt;</code> character instead of
<code>|</code>, which will replace each line break with a space.</p>
<p>Numbers express numerical data, and in YAML, these include integers
(whole numbers), floats (numbers with a decimal point), exponentials,
octals, and hexadecimals.</p>
<pre><code># integer
19

# float 
8.7

# exponential
4.5e+13

# octal 
0o23

# hexadecimal
0xFF</code></pre>
<p>Booleans in YAML, and other programming languages, have one of two
states and are expressed with either <code>true</code> or
<code>false</code>. Words like <code>true</code> and <code>false</code>
are keywords in YAML, so don’t surround them with quotation marks if you
want them interpreted as Booleans. Null values are expressed with the
keyword null or the tilde character, <code>~</code>.</p>
<p>Collections in YAML can be:</p>
<ul>
<li>Sequences (lists/arrays)</li>
<li>Mappings (dictionaries/hashes)</li>
</ul>
<p>To write a sequence, use a dash (<code>-</code>) followed by a
space:</p>
<pre><code>- HTML
- CSS
- JavaScript</code></pre>
<p>Each item in the sequence (list) is placed on a separate line, with a
dash in front of the value and each item in the list is on the same
level.</p>
<p>You can create a nested sequence (remember, use spaces - not tabs -
to create the levels of indentation):</p>
<pre><code>- HTML
- CSS
- JavaScript
 - React
 - Angular
 - Vue</code></pre>
<p>Mappings allow you to list keys with values. Key/value pairs are the
building blocks of YAML documents. Use a colon (<code>:</code>) followed
by a space to create key/value pairs:</p>
<pre><code>Employees:
 name: John Doe
 age: 23
 country: USA</code></pre>
<p>Combining the two to create a list of objects:</p>
<pre><code>Employees:
- name: John Doe
  department: Engineering
  country: USA
- name: Kate Kateson
  department: IT support
  country: United Kingdom</code></pre>
</div>
<div id="read-yaml" class="section level3">
<h3>Read YAML</h3>
<p>Read a test YAML file.</p>
<pre class="r"><code>dat &lt;- yaml::read_yaml(&quot;data/nested_list.yml&quot;)
dat$root</code></pre>
<pre><code>$simple_list
[1] &quot;one&quot;   &quot;two&quot;   &quot;three&quot;

$simple_list_2
[1] &quot;four&quot; &quot;five&quot; &quot;six&quot; 

$nested_list
$nested_list$one
[1] &quot;one&quot;   &quot;two&quot;   &quot;three&quot;

$nested_list$two
[1] &quot;four&quot; &quot;five&quot; &quot;six&quot; 


$nested_list_2
$nested_list_2$one
$nested_list_2$one$two
[1] &quot;one&quot;   &quot;two&quot;   &quot;three&quot;

$nested_list_2$one$three
[1] &quot;four&quot; &quot;five&quot; &quot;six&quot; </code></pre>
<p>Return only specific “level” by using <code>names()</code>.</p>
<pre class="r"><code>names(dat$root)</code></pre>
<pre><code>[1] &quot;simple_list&quot;   &quot;simple_list_2&quot; &quot;nested_list&quot;   &quot;nested_list_2&quot;</code></pre>
<pre class="r"><code>names(dat$root$nested_list)</code></pre>
<pre><code>[1] &quot;one&quot; &quot;two&quot;</code></pre>
</div>
<div id="json-defined" class="section level3">
<h3>JSON defined</h3>
<p>Notes from <a
href="https://www.oracle.com/database/what-is-json/">JSON
Defined</a>.</p>
<p>JavaScript Object Notation (JSON) is a data-exchange format that
makes it possible to transfer populated data structures between
different languages/tools.</p>
<p>JSON can be used in JavaScript programs without any need for parsing
or serialising. It is a text-based way of representing JavaScript object
literals, arrays, and scalar data.</p>
<p>JSON is relatively easy to read and write, while also easy for
software to parse and generate. It is often used for serialising
structured data and exchanging it over a network, typically between a
server and web applications.</p>
<p>At the granular level, JSON consists of data types.</p>
<ol style="list-style-type: decimal">
<li><p>String - composed of Unicode characters, with backslash
(<code>\</code>) escaping.</p>
<p>{ “name” : “Bob” }</p></li>
<li><p>Number - a JSON number follows JavaScript’s double-precision
floating-point format.</p></li>
</ol>
<pre class="json"><code>{
   &quot;number_1&quot; : 210,
   &quot;number_2&quot; : 215,
   &quot;number_3&quot; : 21.05,
   &quot;number_4&quot; : 10.05
}</code></pre>
<ol start="3" style="list-style-type: decimal">
<li><p>Boolean - either <code>true</code> or <code>false</code>, not
surrounded with quotes, and are treated as string values.</p>
<p>{ “AllowPartialShipment” : false }</p></li>
<li><p>Null - empty value and can be used when there is no value
assigned to a key.</p>
<p>{ “Special Instructions” : null }</p></li>
<li><p>Object - a set of name or value pairs inserted between curly
braces (<code>{}</code>). The keys must be strings and should be unique,
and separated by comma/s.</p></li>
</ol>
<pre class="json"><code>{
  &quot;Influencer&quot; : { &quot;name&quot; : &quot;Jaxon&quot; , &quot;age&quot; : &quot;42&quot; , &quot;city&quot; , &quot;New York&quot; }
}</code></pre>
<ol start="6" style="list-style-type: decimal">
<li>Array - an ordered collection of values. Array values must be of
type string, number, object, array, Boolean, or null.</li>
</ol>
<pre class="json"><code>{
   &quot;Influencers&quot; : [ 
      {
         &quot;name&quot; : &quot;Jaxon&quot;, 
         &quot;age&quot; : 42, 
         &quot;Works At&quot; : &quot;Tech News&quot;
      }
      {
         &quot;name&quot; : &quot;Miller&quot;, 
         &quot;age&quot; : 35
         &quot;Works At&quot; : &quot;IT Day&quot;
      }
   ] 
}</code></pre>
</div>
<div id="use-cases" class="section level3">
<h3>Use cases</h3>
<ol style="list-style-type: decimal">
<li>Generating a JSON object from user-generated data</li>
</ol>
<p>JSON is perfect for storing temporary data. For example, temporary
data can be user-generated data, such as a submitted form on a website.
JSON can also be used as a data format for any programming language to
provide a high level of interoperability.</p>
<ol start="2" style="list-style-type: decimal">
<li>Transferring data between systems</li>
</ol>
<p>A website database has a customer’s mailing address, but the address
needs to be verified via an API to make sure it is valid. Send the
address data in JSON format to the address validation service API.</p>
<ol start="3" style="list-style-type: decimal">
<li>Configuring data for applications</li>
</ol>
<p>When developing applications, each application needs the credentials
to connect to a database as well as a log file path. The credentials and
the file path can be specified in a JSON file.</p>
<ol start="4" style="list-style-type: decimal">
<li>Simplifying complex data models</li>
</ol>
<p>JSON simplifies complex documents down to the components that have
been identified as being meaningful by converting the process of data
extraction to a predictable and human readable JSON file.</p>
</div>
<div id="why-json-is-popular" class="section level3">
<h3>Why JSON is popular</h3>
<p>JSON has gained momentum in API code programming and web services
because it helps in faster data interchange and web service results.</p>
<p>It is text-based, lightweight, and has an easy-to-parse data format
requiring no additional code for parsing. For web services, the need to
return and display a lot of data makes JSON the ideal choice.</p>
</div>
<div id="what-is-a-document-database" class="section level3">
<h3>What is a document database?</h3>
<p>A document database is a type of nonrelational database designed to
store, retrieve, and manage document-oriented information. Rather than
having a schema defined upfront, document databases allow for storing
data in collections consisting of documents. NoSQL databases and JSON
databases are types of document databases.</p>
<p>Document databases are often popular among developers because they
store data in a document-model format (semi-structured) rather than
relational (structured).</p>
<p>Document databases offer more flexibility, because developers do not
have to plan out the schemas ahead of time and they can use the same
format they are using in their application code. This means the careful
planning of a SQL database is not as necessary, which makes document
databases useful for rapidly evolving schemas, which can be common in
software development. However, this can come at the cost of speed, size,
and specificity.</p>
</div>
<div id="what-is-a-json-document-database" class="section level3">
<h3>What is a JSON document database?</h3>
<p>Applications that use different JSON data types and JSON-oriented
query language can interact with data stored in a JSON document
database. The JSON document database also provides native support for
JSON.</p>
<p>Characterisitics that define a JSON document database:</p>
<ul>
<li>Non-relational database designed to store and query as a JSON
document.</li>
<li>Data is in plain text but can be stored using the data types
<code>BLOB</code>, <code>VARCHAR2</code>, <code>CLOB</code>, or binary
JSON in 21c.</li>
<li>Accessing JSON data stored in the database is similar to accessing
other database data, including using OCI, .NET, and JDBC.</li>
<li>A schema is not necessary</li>
</ul>
<p>Storing JSON data in a JSON document database makes use of columns
whose data types are <code>VARCHAR2</code>, <code>CLOB</code>,
<code>BLOB</code>, or binary JSON in 21c. The choice of which to use is
usually determined by the size of the JSON documents. Storing JSON data
in the database using standard SQL data types means that JSON data can
be manipulated like any other data type.</p>
<p>JSON data can be managed and manipulated with tables in a JSON
document database, regardless of the data type. The choice of which
table to use is typically motivated by the size of the JSON
documents.</p>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span>
Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 4.5.0 (2025-04-11)
Platform: x86_64-pc-linux-gnu
Running under: Ubuntu 24.04.3 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 
LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.26.so;  LAPACK version 3.12.0

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

time zone: Etc/UTC
tzcode source: system (glibc)

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] rjson_0.2.23    tidyjson_0.3.2  yaml_2.3.10     jsonlite_2.0.0 
[5] workflowr_1.7.1

loaded via a namespace (and not attached):
 [1] dplyr_1.1.4       compiler_4.5.0    promises_1.3.3    tidyselect_1.2.1 
 [5] Rcpp_1.0.14       stringr_1.5.1     git2r_0.36.2      assertthat_0.2.1 
 [9] tidyr_1.3.1       callr_3.7.6       later_1.4.2       jquerylib_0.1.4  
[13] fastmap_1.2.0     R6_2.6.1          generics_0.1.4    knitr_1.50       
[17] tibble_3.3.0      rprojroot_2.0.4   bslib_0.9.0       pillar_1.10.2    
[21] rlang_1.1.6       utf8_1.2.6        cachem_1.1.0      stringi_1.8.7    
[25] httpuv_1.6.16     xfun_0.52         getPass_0.2-4     fs_1.6.6         
[29] sass_0.4.10       cli_3.6.5         magrittr_2.0.3    ps_1.9.1         
[33] digest_0.6.37     processx_3.8.6    rstudioapi_0.17.1 lifecycle_1.0.4  
[37] vctrs_0.6.5       evaluate_1.0.3    glue_1.8.0        whisker_0.4.1    
[41] purrr_1.0.4       rmarkdown_2.29    httr_1.4.7        tools_4.5.0      
[45] pkgconfig_2.0.3   htmltools_0.5.8.1</code></pre>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
https://docs.mathjax.org/en/latest/web/configuration.html. This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>




</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
