<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />



<meta name="date" content="2025-11-14" />

<title>Likelihood Ratio Tests</title>

<script src="site_libs/header-attrs-2.29/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.13.2/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-6.5.2/css/all.min.css" rel="stylesheet" />
<link href="site_libs/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet" />

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/main/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">My blog</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/davetang/muse">
    <span class="fa fa-github"></span>
     
    Source code
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Likelihood Ratio Tests</h1>
<h4 class="date">2025-11-14</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span>
workflowr <span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2025-11-14
</p>
<p>
<strong>Checks:</strong> <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 7
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong> <code>muse/</code> <span
class="glyphicon glyphicon-question-sign" aria-hidden="true"
title="This is the local directory in which the code in this file was executed.">
</span>
</p>
<p>
This reproducible <a href="https://rmarkdown.rstudio.com">R Markdown</a>
analysis was created with <a
  href="https://github.com/workflowr/workflowr">workflowr</a> (version
1.7.1). The <em>Checks</em> tab describes the reproducibility checks
that were applied when the results were created. The <em>Past
versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date
</a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git
repository, you know the exact version of the code that produced these
results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the
global environment can affect the analysis in your R Markdown file in
unknown ways. For reproduciblity it’s best to always run the code in an
empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20200712code">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Seed:</strong>
<code>set.seed(20200712)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20200712code"
class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20200712)</code> was run prior to running
the code in the R Markdown file. Setting a seed ensures that any results
that rely on randomness, e.g. subsampling or permutations, are
reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Session information:</strong>
recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package
versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be
confident that you successfully produced the results during this
run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr
project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomdavetangmusetreeac8e9ff25095addb9919cf6973f2e25f939d8407targetblankac8e9ffa">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Repository version:</strong>
<a href="https://github.com/davetang/muse/tree/ac8e9ff25095addb9919cf6973f2e25f939d8407" target="_blank">ac8e9ff</a>
</a>
</p>
</div>
<div
id="strongRepositoryversionstrongahrefhttpsgithubcomdavetangmusetreeac8e9ff25095addb9919cf6973f2e25f939d8407targetblankac8e9ffa"
class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development
and connecting the code version to the results is critical for
reproducibility.
</p>
<p>
The results in this page were generated with repository version
<a href="https://github.com/davetang/muse/tree/ac8e9ff25095addb9919cf6973f2e25f939d8407" target="_blank">ac8e9ff</a>.
See the <em>Past versions</em> tab to see a history of the changes made
to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for
the analysis have been committed to Git prior to generating the results
(you can use <code>wflow_publish</code> or
<code>wflow_git_commit</code>). workflowr only checks the R Markdown
file, but you know if there are other scripts or data files that it
depends on. Below is the status of the Git repository when the results
were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .Rproj.user/
    Ignored:    data/1M_neurons_filtered_gene_bc_matrices_h5.h5
    Ignored:    data/293t/
    Ignored:    data/293t_3t3_filtered_gene_bc_matrices.tar.gz
    Ignored:    data/293t_filtered_gene_bc_matrices.tar.gz
    Ignored:    data/5k_Human_Donor1_PBMC_3p_gem-x_5k_Human_Donor1_PBMC_3p_gem-x_count_sample_filtered_feature_bc_matrix.h5
    Ignored:    data/5k_Human_Donor2_PBMC_3p_gem-x_5k_Human_Donor2_PBMC_3p_gem-x_count_sample_filtered_feature_bc_matrix.h5
    Ignored:    data/5k_Human_Donor3_PBMC_3p_gem-x_5k_Human_Donor3_PBMC_3p_gem-x_count_sample_filtered_feature_bc_matrix.h5
    Ignored:    data/5k_Human_Donor4_PBMC_3p_gem-x_5k_Human_Donor4_PBMC_3p_gem-x_count_sample_filtered_feature_bc_matrix.h5
    Ignored:    data/97516b79-8d08-46a6-b329-5d0a25b0be98.h5ad
    Ignored:    data/Parent_SC3v3_Human_Glioblastoma_filtered_feature_bc_matrix.tar.gz
    Ignored:    data/brain_counts/
    Ignored:    data/cl.obo
    Ignored:    data/cl.owl
    Ignored:    data/jurkat/
    Ignored:    data/jurkat:293t_50:50_filtered_gene_bc_matrices.tar.gz
    Ignored:    data/jurkat_293t/
    Ignored:    data/jurkat_filtered_gene_bc_matrices.tar.gz
    Ignored:    data/pbmc20k/
    Ignored:    data/pbmc20k_seurat/
    Ignored:    data/pbmc3k.h5ad
    Ignored:    data/pbmc3k/
    Ignored:    data/pbmc3k_bpcells_mat/
    Ignored:    data/pbmc3k_export.mtx
    Ignored:    data/pbmc3k_matrix.mtx
    Ignored:    data/pbmc3k_seurat.rds
    Ignored:    data/pbmc4k_filtered_gene_bc_matrices.tar.gz
    Ignored:    data/pbmc_1k_v3_filtered_feature_bc_matrix.h5
    Ignored:    data/pbmc_1k_v3_raw_feature_bc_matrix.h5
    Ignored:    data/refdata-gex-GRCh38-2020-A.tar.gz
    Ignored:    data/seurat_1m_neuron.rds
    Ignored:    data/t_3k_filtered_gene_bc_matrices.tar.gz
    Ignored:    r_packages_4.4.1/
    Ignored:    r_packages_4.5.0/

Untracked files:
    Untracked:  analysis/bioc_scrnaseq.Rmd
    Untracked:  bpcells_matrix/
    Untracked:  data/Caenorhabditis_elegans.WBcel235.113.gtf.gz
    Untracked:  data/GCF_043380555.1-RS_2024_12_gene_ontology.gaf.gz
    Untracked:  data/arab.rds
    Untracked:  data/astronomicalunit.csv
    Untracked:  data/femaleMiceWeights.csv
    Untracked:  data/lung_bcell.rds
    Untracked:  m3/
    Untracked:  women.json

Unstaged changes:
    Modified:   analysis/isoform_switch_analyzer.Rmd

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not
included in this status report because it is ok for generated content to
have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">
<p>
These are the previous versions of the repository in which changes were
made to the R Markdown (<code>analysis/lrt.Rmd</code>) and HTML
(<code>docs/lrt.html</code>) files. If you’ve configured a remote Git
repository (see <code>?wflow_git_remote</code>), click on the hyperlinks
in the table below to view the files as they were in that past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/davetang/muse/blob/ac8e9ff25095addb9919cf6973f2e25f939d8407/analysis/lrt.Rmd" target="_blank">ac8e9ff</a>
</td>
<td>
Dave Tang
</td>
<td>
2025-11-14
</td>
<td>
Likelihood Ratio Tests
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<div id="introduction" class="section level2">
<h2>Introduction</h2>
<p>This notebook demonstrates three fundamental concepts in statistical
hypothesis testing:</p>
<ol style="list-style-type: decimal">
<li><strong>Test Statistics</strong>: Numbers calculated from data to
make decisions.</li>
<li><strong>Null Distributions</strong>: The expected distribution of
test statistics under the null hypothesis.</li>
<li><strong>Likelihood Ratio Tests (LRT)</strong>: A principled approach
to comparing models.</li>
</ol>
<p>We’ll use a simple, concrete example throughout: testing whether a
coin is fair.</p>
</div>
<div id="test-statistics" class="section level2">
<h2>Test Statistics</h2>
<div id="the-scenario" class="section level3">
<h3>The Scenario</h3>
<p>Suppose we flip a coin 100 times and observe 65 heads. Is this coin
fair (<span class="math inline">\(p = 0.5\)</span>) or biased?</p>
<pre class="r"><code>n_flips &lt;- 100
observed_heads &lt;- 65
observed_proportion &lt;- observed_heads / n_flips</code></pre>
<p>A <strong>test statistic</strong> is a function of our data. For this
problem, we’ll use the number of heads as our test statistic.</p>
<pre class="r"><code>test_statistic &lt;- observed_heads
test_statistic</code></pre>
<pre><code>[1] 65</code></pre>
<p>The test statistic alone doesn’t tell us much. We need to know:
<em>is 65 heads unusual if the coin is fair?</em></p>
</div>
</div>
<div id="null-distributions" class="section level2">
<h2>Null Distributions</h2>
<div id="defining-the-null-hypothesis" class="section level3">
<h3>Defining the Null Hypothesis</h3>
<p>Our null hypothesis (<span class="math inline">\(H_0\)</span>) is
that the coin is fair: <span class="math inline">\(p = 0.5\)</span>.</p>
<p>Under this hypothesis, the number of heads follows a
<strong>Binomial(100, 0.5)</strong> distribution. This is our
<strong>null distribution</strong>.</p>
<pre class="r"><code>possible_heads &lt;- 0:100
null_probabilities &lt;- dbinom(possible_heads, size = n_flips, prob = 0.5)

null_dist_df &lt;- data.frame(
  heads = possible_heads,
  probability = null_probabilities
)

ggplot(null_dist_df, aes(x = heads, y = probability)) +
  geom_col(fill = &quot;skyblue&quot;, alpha = 0.7) +
  geom_vline(xintercept = observed_heads, color = &quot;red&quot;, lty = 2) +
  annotate(
    &quot;text&quot;,
    x = observed_heads + 10,
    y = max(null_probabilities) * 0.9,
    label = paste(&quot;Observed:&quot;, observed_heads),
    color = &quot;red&quot;,
    size = 5
  ) +
  labs(
    title = &quot;Null Distribution: Binomial(100, 0.5)&quot;,
    subtitle = &quot;Distribution of heads if coin is fair&quot;,
    x = &quot;Number of Heads&quot;,
    y = &quot;Probability&quot;
  ) +
  theme_minimal()</code></pre>
<p><img src="figure/lrt.Rmd/null-distribution-1.png" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="calculating-the-p-value" class="section level3">
<h3>Calculating the p-value</h3>
<p><strong>The p-value is the probability of observing a test statistic
as extreme or more extreme than what we observed, assuming the null
hypothesis is true</strong>.</p>
<pre class="r"><code>p_value &lt;- 2 * pbinom(
  observed_heads - 1,
  size = n_flips, 
  prob = 0.5,
  lower.tail = FALSE
)

p_value</code></pre>
<pre><code>[1] 0.003517642</code></pre>
<p>If the coin were fair, we’d observe 65 or more heads only 0.35% of
the time.</p>
</div>
<div id="simulation-approach-to-null-distribution"
class="section level3">
<h3>Simulation Approach to Null Distribution</h3>
<p>We can also build the null distribution empirically through
simulation.</p>
<pre class="r"><code>set.seed(1984)
n_simulations &lt;- 10000
simulated_heads &lt;- rbinom(n_simulations, size = n_flips, prob = 0.5)

ggplot(data.frame(heads = simulated_heads), aes(x = heads)) +
  geom_histogram(
    aes(y = after_stat(density)),
    bins = 20, 
    fill = &quot;lightblue&quot;,
    alpha = 0.7,
    color = &quot;black&quot;
  ) +
  geom_vline(
    xintercept = observed_heads,
    color = &quot;red&quot;,
    lty = 2
  ) +
  labs(
    title = &quot;Simulated Null Distribution (10,000 simulations)&quot;,
    subtitle = &quot;Each simulation: 100 flips of a fair coin&quot;,
    x = &quot;Number of Heads&quot;,
    y = &quot;Density&quot;
  ) +
  theme_minimal()</code></pre>
<p><img src="figure/lrt.Rmd/simulate-null-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>empirical_p_value &lt;- mean(abs(simulated_heads - 50) &gt;= abs(observed_heads - 50))
round(empirical_p_value, 4)</code></pre>
<pre><code>[1] 0.0029</code></pre>
</div>
</div>
<div id="likelihood-ratio-tests" class="section level2">
<h2>Likelihood Ratio Tests</h2>
<p>The <strong>Likelihood Ratio Test</strong> compares two models:</p>
<ul>
<li>Null model (<span class="math inline">\(H_0\)</span>): Coin is fair,
p = 0.5 (1 parameter fixed)</li>
<li>Alternative model (<span class="math inline">\(H_1\)</span>): Coin
has some bias, p = ? (1 parameter estimated from data)</li>
</ul>
<p><strong>The likelihood is the probability of observing our data given
a model</strong>.</p>
<pre class="r"><code>p_null &lt;- 0.5
likelihood_null &lt;- dbinom(observed_heads, size = n_flips, prob = p_null)
likelihood_null</code></pre>
<pre><code>[1] 0.0008638557</code></pre>
<pre class="r"><code># Maximum likelihood estimate
p_alternative &lt;- observed_heads / n_flips
likelihood_alternative &lt;- dbinom(observed_heads, size = n_flips, prob = p_alternative)
likelihood_alternative</code></pre>
<pre><code>[1] 0.08340469</code></pre>
<pre class="r"><code>my_ratio &lt;- round(likelihood_alternative / likelihood_null, 2)
my_ratio</code></pre>
<pre><code>[1] 96.55</code></pre>
<p>The data is 96.55 times more likely under <span
class="math inline">\(H_1\)</span> than <span
class="math inline">\(H_0\)</span>.</p>
<div id="log-likelihoods" class="section level3">
<h3>Log-Likelihoods</h3>
<p>We typically work with log-likelihoods because likelihoods can be
very small.</p>
<pre class="r"><code>log_lik_null &lt;- dbinom(observed_heads, size = n_flips, prob = p_null, log = TRUE)
log_lik_null</code></pre>
<pre><code>[1] -7.054105</code></pre>
<pre class="r"><code>log_lik_alternative &lt;- dbinom(observed_heads, size = n_flips, prob = p_alternative, log = TRUE)
log_lik_alternative</code></pre>
<pre><code>[1] -2.484051</code></pre>
</div>
<div id="the-lrt-statistic" class="section level3">
<h3>The LRT Statistic</h3>
<p>The likelihood ratio test statistic is:</p>
<p><span class="math display">\[\Lambda = -2
\log\left(\frac{L(\text{H}_0)}{L(\text{H}_1)}\right) = -2[\log
L(\text{H}_0) - \log L(\text{H}_1)]\]</span></p>
<pre class="r"><code>lrt_statistic &lt;- -2 * (log_lik_null - log_lik_alternative)
lrt_statistic</code></pre>
<pre><code>[1] 9.140108</code></pre>
</div>
<div id="null-distribution-of-the-lrt-statistic" class="section level3">
<h3>Null Distribution of the LRT Statistic</h3>
<p>Under certain conditions (Wilks’ theorem), the LRT statistic follows
a <strong>chi-squared distribution</strong> with degrees of freedom
equal to the difference in number of parameters between models. In this
case: df = 1 (alternative has 1 more parameter than null).</p>
<pre class="r"><code># Degrees of freedom
df &lt;- 1

# P-value from chi-squared distribution
lrt_p_value &lt;- pchisq(lrt_statistic, df = df, lower.tail = FALSE)

chi_sq_values &lt;- seq(0, 15, length.out = 1000)
chi_sq_density &lt;- dchisq(chi_sq_values, df = df)

ggplot(data.frame(x = chi_sq_values, y = chi_sq_density), aes(x, y)) +
  geom_line(linewidth = 1, color = &quot;blue&quot;) +
  geom_area(
    data = subset(
      data.frame(
        x = chi_sq_values,
        y = chi_sq_density
      ), 
      x &gt;= lrt_statistic
    ),
    aes(x, y),
    fill = &quot;red&quot;,
    alpha = 0.3
  ) +
  geom_vline(xintercept = lrt_statistic, color = &quot;red&quot;, lty = 2) +
  annotate(
    &quot;text&quot;,
    x = lrt_statistic + 2,
    y = 3,
    label = paste(&quot;Observed LRT:&quot;, round(lrt_statistic, 2)), 
    color = &quot;red&quot;, size = 5) +
  labs(
    title = &quot;Null Distribution of LRT Statistic&quot;,
    subtitle = &quot;Chi-squared distribution with 1 degree of freedom&quot;,
    x = &quot;LRT Statistic Value&quot;,
    y = &quot;Density&quot;
  ) +
  theme_minimal()</code></pre>
<p><img src="figure/lrt.Rmd/lrt-null-distribution-1.png" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="comparing-to-standard-test" class="section level3">
<h3>Comparing to Standard Test</h3>
<p>Let’s verify our LRT matches the standard proportion test:</p>
<pre class="r"><code># Standard proportion test
prop_test &lt;- prop.test(observed_heads, n_flips, p = 0.5, correct = FALSE)

cat(&quot;=== Comparison of Methods ===\n\n&quot;)</code></pre>
<pre><code>=== Comparison of Methods ===</code></pre>
<pre class="r"><code>cat(&quot;Exact binomial test p-value:&quot;, round(p_value, 4), &quot;\n&quot;)</code></pre>
<pre><code>Exact binomial test p-value: 0.0035 </code></pre>
<pre class="r"><code>cat(&quot;Simulated p-value:&quot;, round(empirical_p_value, 4), &quot;\n&quot;)</code></pre>
<pre><code>Simulated p-value: 0.0029 </code></pre>
<pre class="r"><code>cat(&quot;LRT p-value:&quot;, round(lrt_p_value, 4), &quot;\n&quot;)</code></pre>
<pre><code>LRT p-value: 0.0025 </code></pre>
<pre class="r"><code>cat(&quot;prop.test p-value:&quot;, round(prop_test$p.value, 4), &quot;\n&quot;)</code></pre>
<pre><code>prop.test p-value: 0.0027 </code></pre>
</div>
</div>
<div id="summary" class="section level2">
<h2>Summary</h2>
<ul>
<li>The <strong>test statistic</strong> summarises data in a way that
helps detect departures from the null hypothesis.</li>
<li>The <strong>null distribution</strong> shows what values we’d expect
by chance alone.</li>
<li>The <strong>LRT</strong> is a principled way to compare models by
looking at their relative likelihoods.</li>
<li>Under regularity conditions, the LRT statistic has a known null
distribution (chi-squared), making it easy to calculate p-values.</li>
</ul>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span>
Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 4.5.0 (2025-04-11)
Platform: x86_64-pc-linux-gnu
Running under: Ubuntu 24.04.3 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 
LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.26.so;  LAPACK version 3.12.0

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

time zone: Etc/UTC
tzcode source: system (glibc)

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] lubridate_1.9.4 forcats_1.0.0   stringr_1.5.1   dplyr_1.1.4    
 [5] purrr_1.0.4     readr_2.1.5     tidyr_1.3.1     tibble_3.3.0   
 [9] ggplot2_3.5.2   tidyverse_2.0.0 workflowr_1.7.1

loaded via a namespace (and not attached):
 [1] sass_0.4.10        generics_0.1.4     stringi_1.8.7      hms_1.1.3         
 [5] digest_0.6.37      magrittr_2.0.3     timechange_0.3.0   evaluate_1.0.3    
 [9] grid_4.5.0         RColorBrewer_1.1-3 fastmap_1.2.0      rprojroot_2.0.4   
[13] jsonlite_2.0.0     processx_3.8.6     whisker_0.4.1      ps_1.9.1          
[17] promises_1.3.3     httr_1.4.7         scales_1.4.0       jquerylib_0.1.4   
[21] cli_3.6.5          rlang_1.1.6        withr_3.0.2        cachem_1.1.0      
[25] yaml_2.3.10        tools_4.5.0        tzdb_0.5.0         httpuv_1.6.16     
[29] vctrs_0.6.5        R6_2.6.1           lifecycle_1.0.4    git2r_0.36.2      
[33] fs_1.6.6           pkgconfig_2.0.3    callr_3.7.6        pillar_1.10.2     
[37] bslib_0.9.0        later_1.4.2        gtable_0.3.6       glue_1.8.0        
[41] Rcpp_1.0.14        xfun_0.52          tidyselect_1.2.1   rstudioapi_0.17.1 
[45] knitr_1.50         farver_2.1.2       htmltools_0.5.8.1  labeling_0.4.3    
[49] rmarkdown_2.29     compiler_4.5.0     getPass_0.2-4     </code></pre>
</div>
</div>
</div>
</div>
</div>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
https://docs.mathjax.org/en/latest/web/configuration.html. This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>




</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
