<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />



<meta name="date" content="2025-01-10" />

<title>Differential abundance testing with Milo</title>

<script src="site_libs/header-attrs-2.28/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.13.2/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-6.4.2/css/all.min.css" rel="stylesheet" />
<link href="site_libs/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet" />

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/main/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">My blog</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/davetang/muse">
    <span class="fa fa-github"></span>
     
    Source code
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Differential abundance testing with
Milo</h1>
<h4 class="date">2025-01-10</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span>
workflowr <span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2025-01-10
</p>
<p>
<strong>Checks:</strong> <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 7
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong> <code>muse/</code> <span
class="glyphicon glyphicon-question-sign" aria-hidden="true"
title="This is the local directory in which the code in this file was executed.">
</span>
</p>
<p>
This reproducible <a href="https://rmarkdown.rstudio.com">R Markdown</a>
analysis was created with <a
  href="https://github.com/workflowr/workflowr">workflowr</a> (version
1.7.1). The <em>Checks</em> tab describes the reproducibility checks
that were applied when the results were created. The <em>Past
versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date
</a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git
repository, you know the exact version of the code that produced these
results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the
global environment can affect the analysis in your R Markdown file in
unknown ways. For reproduciblity it’s best to always run the code in an
empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20200712code">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Seed:</strong>
<code>set.seed(20200712)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20200712code"
class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20200712)</code> was run prior to running
the code in the R Markdown file. Setting a seed ensures that any results
that rely on randomness, e.g. subsampling or permutations, are
reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Session information:</strong>
recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package
versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be
confident that you successfully produced the results during this
run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr
project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomdavetangmusetree2dd7facf185bc8cbd90a16df1ed43582c38a8f0btargetblank2dd7faca">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Repository version:</strong>
<a href="https://github.com/davetang/muse/tree/2dd7facf185bc8cbd90a16df1ed43582c38a8f0b" target="_blank">2dd7fac</a>
</a>
</p>
</div>
<div
id="strongRepositoryversionstrongahrefhttpsgithubcomdavetangmusetree2dd7facf185bc8cbd90a16df1ed43582c38a8f0btargetblank2dd7faca"
class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development
and connecting the code version to the results is critical for
reproducibility.
</p>
<p>
The results in this page were generated with repository version
<a href="https://github.com/davetang/muse/tree/2dd7facf185bc8cbd90a16df1ed43582c38a8f0b" target="_blank">2dd7fac</a>.
See the <em>Past versions</em> tab to see a history of the changes made
to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for
the analysis have been committed to Git prior to generating the results
(you can use <code>wflow_publish</code> or
<code>wflow_git_commit</code>). workflowr only checks the R Markdown
file, but you know if there are other scripts or data files that it
depends on. Below is the status of the Git repository when the results
were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .Rhistory
    Ignored:    .Rproj.user/
    Ignored:    data/pbmc3k.csv
    Ignored:    data/pbmc3k.csv.gz
    Ignored:    data/pbmc3k/
    Ignored:    r_packages_4.4.0/
    Ignored:    r_packages_4.4.1/

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not
included in this status report because it is ok for generated content to
have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">
<p>
These are the previous versions of the repository in which changes were
made to the R Markdown (<code>analysis/miloR.Rmd</code>) and HTML
(<code>docs/miloR.html</code>) files. If you’ve configured a remote Git
repository (see <code>?wflow_git_remote</code>), click on the hyperlinks
in the table below to view the files as they were in that past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/davetang/muse/blob/2dd7facf185bc8cbd90a16df1ed43582c38a8f0b/analysis/miloR.Rmd" target="_blank">2dd7fac</a>
</td>
<td>
Dave Tang
</td>
<td>
2025-01-10
</td>
<td>
DA testing with miloR
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<p><a
href="https://www.bioconductor.org/packages/release/bioc/html/miloR.html">miloR</a>:</p>
<blockquote>
<p>Milo is a tool for analysis of complex single cell datasets generated
from replicated multi-condition experiments, which detects changes in
composition between conditions. While differential abundance (DA) is
commonly quantified in discrete cell clusters, Milo uses partially
overlapping neighbourhoods of cells on a KNN graph. Starting from a
graph that faithfully recapitulates the biology of the cell population,
Milo analysis consists of 3 steps:</p>
<p>Sampling of representative neighbourhoods Testing for differential
abundance of conditions in all neighbourhoods Accounting for multiple
hypothesis testing using a weighted FDR procedure that accounts for the
overlap of neighbourhoods</p>
</blockquote>
<div id="dependencies" class="section level2">
<h2>Dependencies</h2>
<p>Install Bioconductor packages using
<code>BiocManager::install()</code>.</p>
<pre class="r"><code>if (!require(&quot;BiocManager&quot;, quietly = TRUE))
    install.packages(&quot;BiocManager&quot;)

BiocManager::install(&quot;SingleCellExperiment&quot;)
BiocManager::install(&quot;scran&quot;)
BiocManager::install(&quot;scater&quot;)
BiocManager::install(&quot;miloR&quot;)

install.packages(&#39;dplyr&#39;)
install.packages(&#39;patchwork&#39;)</code></pre>
<p>Load libraries.</p>
<pre class="r"><code>library(miloR)</code></pre>
<pre><code>Loading required package: edgeR</code></pre>
<pre><code>Loading required package: limma</code></pre>
<pre class="r"><code>library(SingleCellExperiment)</code></pre>
<pre><code>Loading required package: SummarizedExperiment</code></pre>
<pre><code>Loading required package: MatrixGenerics</code></pre>
<pre><code>Loading required package: matrixStats</code></pre>
<pre><code>
Attaching package: &#39;matrixStats&#39;</code></pre>
<pre><code>The following object is masked from &#39;package:dplyr&#39;:

    count</code></pre>
<pre><code>
Attaching package: &#39;MatrixGenerics&#39;</code></pre>
<pre><code>The following objects are masked from &#39;package:matrixStats&#39;:

    colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,
    colCounts, colCummaxs, colCummins, colCumprods, colCumsums,
    colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,
    colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,
    colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,
    colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,
    colWeightedMeans, colWeightedMedians, colWeightedSds,
    colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,
    rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,
    rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,
    rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,
    rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,
    rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,
    rowWeightedMads, rowWeightedMeans, rowWeightedMedians,
    rowWeightedSds, rowWeightedVars</code></pre>
<pre><code>Loading required package: GenomicRanges</code></pre>
<pre><code>Loading required package: stats4</code></pre>
<pre><code>Loading required package: BiocGenerics</code></pre>
<pre><code>
Attaching package: &#39;BiocGenerics&#39;</code></pre>
<pre><code>The following object is masked from &#39;package:limma&#39;:

    plotMA</code></pre>
<pre><code>The following objects are masked from &#39;package:lubridate&#39;:

    intersect, setdiff, union</code></pre>
<pre><code>The following objects are masked from &#39;package:dplyr&#39;:

    combine, intersect, setdiff, union</code></pre>
<pre><code>The following objects are masked from &#39;package:stats&#39;:

    IQR, mad, sd, var, xtabs</code></pre>
<pre><code>The following objects are masked from &#39;package:base&#39;:

    anyDuplicated, aperm, append, as.data.frame, basename, cbind,
    colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,
    get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,
    match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,
    Position, rank, rbind, Reduce, rownames, sapply, saveRDS, setdiff,
    table, tapply, union, unique, unsplit, which.max, which.min</code></pre>
<pre><code>Loading required package: S4Vectors</code></pre>
<pre><code>
Attaching package: &#39;S4Vectors&#39;</code></pre>
<pre><code>The following objects are masked from &#39;package:lubridate&#39;:

    second, second&lt;-</code></pre>
<pre><code>The following objects are masked from &#39;package:dplyr&#39;:

    first, rename</code></pre>
<pre><code>The following object is masked from &#39;package:tidyr&#39;:

    expand</code></pre>
<pre><code>The following object is masked from &#39;package:utils&#39;:

    findMatches</code></pre>
<pre><code>The following objects are masked from &#39;package:base&#39;:

    expand.grid, I, unname</code></pre>
<pre><code>Loading required package: IRanges</code></pre>
<pre><code>
Attaching package: &#39;IRanges&#39;</code></pre>
<pre><code>The following object is masked from &#39;package:lubridate&#39;:

    %within%</code></pre>
<pre><code>The following objects are masked from &#39;package:dplyr&#39;:

    collapse, desc, slice</code></pre>
<pre><code>The following object is masked from &#39;package:purrr&#39;:

    reduce</code></pre>
<pre><code>Loading required package: GenomeInfoDb</code></pre>
<pre><code>Loading required package: Biobase</code></pre>
<pre><code>Welcome to Bioconductor

    Vignettes contain introductory material; view with
    &#39;browseVignettes()&#39;. To cite Bioconductor, see
    &#39;citation(&quot;Biobase&quot;)&#39;, and for packages &#39;citation(&quot;pkgname&quot;)&#39;.</code></pre>
<pre><code>
Attaching package: &#39;Biobase&#39;</code></pre>
<pre><code>The following object is masked from &#39;package:MatrixGenerics&#39;:

    rowMedians</code></pre>
<pre><code>The following objects are masked from &#39;package:matrixStats&#39;:

    anyMissing, rowMedians</code></pre>
<pre><code>
Attaching package: &#39;SingleCellExperiment&#39;</code></pre>
<pre><code>The following object is masked from &#39;package:edgeR&#39;:

    cpm</code></pre>
<pre class="r"><code>library(scater)</code></pre>
<pre><code>Loading required package: scuttle</code></pre>
<pre><code>
Attaching package: &#39;scater&#39;</code></pre>
<pre><code>The following object is masked from &#39;package:limma&#39;:

    plotMDS</code></pre>
<pre class="r"><code>library(scran)
library(dplyr)
library(patchwork)</code></pre>
</div>
<div id="data" class="section level2">
<h2>Data</h2>
<p>Load testing data.</p>
<pre class="r"><code>data(&quot;sim_trajectory&quot;, package = &quot;miloR&quot;)

## Extract SingleCellExperiment object
traj_sce &lt;- sim_trajectory[[&#39;SCE&#39;]]

## Extract sample metadata to use for testing
traj_meta &lt;- sim_trajectory[[&quot;meta&quot;]]

## Add metadata to colData slot
colData(traj_sce) &lt;- DataFrame(traj_meta)
colnames(traj_sce) &lt;- colData(traj_sce)$cell_id

redim &lt;- reducedDim(traj_sce, &quot;PCA&quot;)
dimnames(redim) &lt;- list(colnames(traj_sce), paste0(&quot;PC&quot;, c(1:50)))
reducedDim(traj_sce, &quot;PCA&quot;) &lt;- redim </code></pre>
</div>
<div id="pre-processing" class="section level2">
<h2>Pre-processing</h2>
<pre class="r"><code>logcounts(traj_sce) &lt;- log(counts(traj_sce) + 1)
traj_sce &lt;- runPCA(traj_sce, ncomponents=30)
traj_sce &lt;- runUMAP(traj_sce)

plotUMAP(traj_sce)</code></pre>
<p><img src="figure/miloR.Rmd/preprocessing-1.png" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="milo-object" class="section level2">
<h2>Milo object</h2>
<pre class="r"><code>traj_milo &lt;- Milo(traj_sce)
reducedDim(traj_milo, &quot;UMAP&quot;) &lt;- reducedDim(traj_sce, &quot;UMAP&quot;)

traj_milo</code></pre>
<pre><code>class: Milo 
dim: 500 500 
metadata(0):
assays(2): counts logcounts
rownames(500): G1 G2 ... G499 G500
rowData names(0):
colnames(500): C1 C2 ... C499 C500
colData names(5): cell_id group_id Condition Replicate Sample
reducedDimNames(2): PCA UMAP
mainExpName: NULL
altExpNames(0):
nhoods dimensions(2): 1 1
nhoodCounts dimensions(2): 1 1
nhoodDistances dimension(1): 0
graph names(0):
nhoodIndex names(1): 0
nhoodExpression dimension(2): 1 1
nhoodReducedDim names(0):
nhoodGraph names(0):
nhoodAdjacency dimension(2): 1 1</code></pre>
</div>
<div id="construct-knn-graph" class="section level2">
<h2>Construct KNN graph</h2>
<pre class="r"><code>traj_milo &lt;- buildGraph(traj_milo, k = 10, d = 30)</code></pre>
<pre><code>Constructing kNN graph with k:10</code></pre>
</div>
<div id="defining-representative-neighbourhoods" class="section level2">
<h2>Defining representative neighbourhoods</h2>
<pre class="r"><code>traj_milo &lt;- makeNhoods(traj_milo, prop = 0.1, k = 10, d=30, refined = TRUE)</code></pre>
<pre><code>Checking valid object</code></pre>
<pre><code>Running refined sampling with reduced_dim</code></pre>
<pre class="r"><code>plotNhoodSizeHist(traj_milo)</code></pre>
<p><img src="figure/miloR.Rmd/make_neighbourhoods-1.png" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="counting-cells-in-neighbourhoods" class="section level2">
<h2>Counting cells in neighbourhoods</h2>
<pre class="r"><code>traj_milo &lt;- countCells(traj_milo, meta.data = data.frame(colData(traj_milo)), samples=&quot;Sample&quot;)</code></pre>
<pre><code>Checking meta.data validity</code></pre>
<pre><code>Counting cells in neighbourhoods</code></pre>
<pre class="r"><code>head(nhoodCounts(traj_milo))</code></pre>
<pre><code>6 x 6 sparse Matrix of class &quot;dgCMatrix&quot;
  B_R1 A_R1 A_R2 B_R2 B_R3 A_R3
1   13    2    1   19   21    1
2    8    5    3    4    5    8
3   12    1    1    9   12    .
4    9    3    3   11   11    8
5    8    .    .   14   28    1
6    9    6    7    7    6   10</code></pre>
</div>
<div id="differential-abundance-testing" class="section level2">
<h2>Differential abundance testing</h2>
<pre class="r"><code>traj_design &lt;- data.frame(colData(traj_milo))[,c(&quot;Sample&quot;, &quot;Condition&quot;)]
traj_design &lt;- distinct(traj_design)
rownames(traj_design) &lt;- traj_design$Sample
## Reorder rownames to match columns of nhoodCounts(milo)
traj_design &lt;- traj_design[colnames(nhoodCounts(traj_milo)), , drop=FALSE]

traj_design</code></pre>
<pre><code>     Sample Condition
B_R1   B_R1         B
A_R1   A_R1         A
A_R2   A_R2         A
B_R2   B_R2         B
B_R3   B_R3         B
A_R3   A_R3         A</code></pre>
<pre class="r"><code>traj_milo &lt;- calcNhoodDistance(traj_milo, d=30)</code></pre>
<pre><code>&#39;as(&lt;dgTMatrix&gt;, &quot;dgCMatrix&quot;)&#39; is deprecated.
Use &#39;as(., &quot;CsparseMatrix&quot;)&#39; instead.
See help(&quot;Deprecated&quot;) and help(&quot;Matrix-deprecated&quot;).</code></pre>
<pre class="r"><code>da_results &lt;- testNhoods(traj_milo, design = ~ Condition, design.df = traj_design)</code></pre>
<pre><code>Using TMM normalisation</code></pre>
<pre><code>Running with model contrasts</code></pre>
<pre><code>Performing spatial FDR correction with k-distance weighting</code></pre>
<pre class="r"><code>da_results %&gt;%
  dplyr::arrange(SpatialFDR) %&gt;%
  head() </code></pre>
<pre><code>       logFC   logCPM        F       PValue          FDR Nhood   SpatialFDR
12  3.780446 16.31466 29.17226 3.761988e-07 1.053357e-05    12 1.025632e-05
5   4.238085 15.97004 25.42427 1.786476e-06 2.501067e-05     5 2.450793e-05
21  3.100395 15.77424 32.86600 9.631834e-06 8.989712e-05    21 8.890275e-05
1   2.545269 16.10138 16.53118 8.925256e-05 6.247679e-04     1 6.193743e-04
3   2.791546 15.54753 19.27492 1.518024e-04 8.500936e-04     3 8.447641e-04
17 -1.640569 15.74817 22.08324 2.019242e-04 9.423129e-04    17 9.412823e-04</code></pre>
</div>
<div id="visualise-neighbourhoods-displaying-da" class="section level2">
<h2>Visualise neighbourhoods displaying DA</h2>
<pre class="r"><code>traj_milo &lt;- buildNhoodGraph(traj_milo)

plotUMAP(traj_milo) + plotNhoodGraphDA(traj_milo, da_results, alpha=0.05) +
  plot_layout(guides=&quot;collect&quot;)</code></pre>
<pre><code>Adding nhood effect sizes to neighbourhood graph attributes</code></pre>
<p><img src="figure/miloR.Rmd/plot_neighbourhood-1.png" width="672" style="display: block; margin: auto;" /></p>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span>
Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 4.4.1 (2024-06-14)
Platform: x86_64-pc-linux-gnu
Running under: Ubuntu 22.04.5 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 
LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

time zone: Etc/UTC
tzcode source: system (glibc)

attached base packages:
[1] stats4    stats     graphics  grDevices utils     datasets  methods  
[8] base     

other attached packages:
 [1] patchwork_1.3.0             scran_1.34.0               
 [3] scater_1.34.0               scuttle_1.16.0             
 [5] SingleCellExperiment_1.28.1 SummarizedExperiment_1.36.0
 [7] Biobase_2.66.0              GenomicRanges_1.58.0       
 [9] GenomeInfoDb_1.42.1         IRanges_2.40.1             
[11] S4Vectors_0.44.0            BiocGenerics_0.52.0        
[13] MatrixGenerics_1.18.0       matrixStats_1.4.1          
[15] miloR_2.2.0                 edgeR_4.4.1                
[17] limma_3.62.1                lubridate_1.9.3            
[19] forcats_1.0.0               stringr_1.5.1              
[21] dplyr_1.1.4                 purrr_1.0.2                
[23] readr_2.1.5                 tidyr_1.3.1                
[25] tibble_3.2.1                ggplot2_3.5.1              
[27] tidyverse_2.0.0             workflowr_1.7.1            

loaded via a namespace (and not attached):
  [1] RColorBrewer_1.1-3      rstudioapi_0.17.1       jsonlite_1.8.9         
  [4] magrittr_2.0.3          ggbeeswarm_0.7.2        farver_2.1.2           
  [7] rmarkdown_2.28          fs_1.6.4                zlibbioc_1.52.0        
 [10] vctrs_0.6.5             memoise_2.0.1           htmltools_0.5.8.1      
 [13] S4Arrays_1.6.0          BiocNeighbors_2.0.1     SparseArray_1.6.0      
 [16] sass_0.4.9              pracma_2.4.4            bslib_0.8.0            
 [19] cachem_1.1.0            whisker_0.4.1           igraph_2.1.2           
 [22] lifecycle_1.0.4         pkgconfig_2.0.3         rsvd_1.0.5             
 [25] Matrix_1.7-0            R6_2.5.1                fastmap_1.2.0          
 [28] GenomeInfoDbData_1.2.13 digest_0.6.37           numDeriv_2016.8-1.1    
 [31] colorspace_2.1-1        ps_1.8.1                rprojroot_2.0.4        
 [34] dqrng_0.4.1             irlba_2.3.5.1           beachmat_2.22.0        
 [37] labeling_0.4.3          fansi_1.0.6             timechange_0.3.0       
 [40] httr_1.4.7              polyclip_1.10-7         abind_1.4-8            
 [43] compiler_4.4.1          withr_3.0.2             BiocParallel_1.40.0    
 [46] viridis_0.6.5           highr_0.11              ggforce_0.4.2          
 [49] MASS_7.3-60.2           DelayedArray_0.32.0     bluster_1.16.0         
 [52] gtools_3.9.5            tools_4.4.1             vipor_0.4.7            
 [55] beeswarm_0.4.0          httpuv_1.6.15           glue_1.8.0             
 [58] callr_3.7.6             promises_1.3.0          grid_4.4.1             
 [61] getPass_0.2-4           cluster_2.1.6           generics_0.1.3         
 [64] gtable_0.3.6            tzdb_0.4.0              hms_1.1.3              
 [67] metapod_1.14.0          BiocSingular_1.22.0     tidygraph_1.3.1        
 [70] ScaledMatrix_1.14.0     utf8_1.2.4              XVector_0.46.0         
 [73] ggrepel_0.9.6           pillar_1.9.0            later_1.3.2            
 [76] splines_4.4.1           tweenr_2.0.3            lattice_0.22-6         
 [79] FNN_1.1.4.1             tidyselect_1.2.1        locfit_1.5-9.10        
 [82] knitr_1.48              git2r_0.35.0            gridExtra_2.3          
 [85] xfun_0.48               graphlayouts_1.2.0      statmod_1.5.0          
 [88] stringi_1.8.4           UCSC.utils_1.2.0        yaml_2.3.10            
 [91] evaluate_1.0.1          codetools_0.2-20        ggraph_2.2.1           
 [94] cli_3.6.3               uwot_0.2.2              munsell_0.5.1          
 [97] processx_3.8.4          jquerylib_0.1.4         Rcpp_1.0.13            
[100] parallel_4.4.1          viridisLite_0.4.2       scales_1.3.0           
[103] crayon_1.5.3            rlang_1.1.4             cowplot_1.1.3          </code></pre>
</div>
</div>
</div>
</div>
</div>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
https://docs.mathjax.org/en/latest/web/configuration.html. This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>




</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
