<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />



<meta name="date" content="2025-10-24" />

<title>Mixed-effects models</title>

<script src="site_libs/header-attrs-2.29/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.13.2/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-6.5.2/css/all.min.css" rel="stylesheet" />
<link href="site_libs/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet" />

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/main/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">My blog</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/davetang/muse">
    <span class="fa fa-github"></span>
     
    Source code
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Mixed-effects models</h1>
<h4 class="date">2025-10-24</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span>
workflowr <span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2025-10-24
</p>
<p>
<strong>Checks:</strong> <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 7
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong> <code>muse/</code> <span
class="glyphicon glyphicon-question-sign" aria-hidden="true"
title="This is the local directory in which the code in this file was executed.">
</span>
</p>
<p>
This reproducible <a href="https://rmarkdown.rstudio.com">R Markdown</a>
analysis was created with <a
  href="https://github.com/workflowr/workflowr">workflowr</a> (version
1.7.1). The <em>Checks</em> tab describes the reproducibility checks
that were applied when the results were created. The <em>Past
versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date
</a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git
repository, you know the exact version of the code that produced these
results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the
global environment can affect the analysis in your R Markdown file in
unknown ways. For reproduciblity it’s best to always run the code in an
empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20200712code">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Seed:</strong>
<code>set.seed(20200712)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20200712code"
class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20200712)</code> was run prior to running
the code in the R Markdown file. Setting a seed ensures that any results
that rely on randomness, e.g. subsampling or permutations, are
reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Session information:</strong>
recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package
versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be
confident that you successfully produced the results during this
run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr
project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomdavetangmusetree638126d31118ca8ce0b521bf83435131e332d213targetblank638126da">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Repository version:</strong>
<a href="https://github.com/davetang/muse/tree/638126d31118ca8ce0b521bf83435131e332d213" target="_blank">638126d</a>
</a>
</p>
</div>
<div
id="strongRepositoryversionstrongahrefhttpsgithubcomdavetangmusetree638126d31118ca8ce0b521bf83435131e332d213targetblank638126da"
class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development
and connecting the code version to the results is critical for
reproducibility.
</p>
<p>
The results in this page were generated with repository version
<a href="https://github.com/davetang/muse/tree/638126d31118ca8ce0b521bf83435131e332d213" target="_blank">638126d</a>.
See the <em>Past versions</em> tab to see a history of the changes made
to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for
the analysis have been committed to Git prior to generating the results
(you can use <code>wflow_publish</code> or
<code>wflow_git_commit</code>). workflowr only checks the R Markdown
file, but you know if there are other scripts or data files that it
depends on. Below is the status of the Git repository when the results
were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .Rproj.user/
    Ignored:    data/1M_neurons_filtered_gene_bc_matrices_h5.h5
    Ignored:    data/293t/
    Ignored:    data/293t_3t3_filtered_gene_bc_matrices.tar.gz
    Ignored:    data/293t_filtered_gene_bc_matrices.tar.gz
    Ignored:    data/5k_Human_Donor1_PBMC_3p_gem-x_5k_Human_Donor1_PBMC_3p_gem-x_count_sample_filtered_feature_bc_matrix.h5
    Ignored:    data/5k_Human_Donor2_PBMC_3p_gem-x_5k_Human_Donor2_PBMC_3p_gem-x_count_sample_filtered_feature_bc_matrix.h5
    Ignored:    data/5k_Human_Donor3_PBMC_3p_gem-x_5k_Human_Donor3_PBMC_3p_gem-x_count_sample_filtered_feature_bc_matrix.h5
    Ignored:    data/5k_Human_Donor4_PBMC_3p_gem-x_5k_Human_Donor4_PBMC_3p_gem-x_count_sample_filtered_feature_bc_matrix.h5
    Ignored:    data/97516b79-8d08-46a6-b329-5d0a25b0be98.h5ad
    Ignored:    data/Parent_SC3v3_Human_Glioblastoma_filtered_feature_bc_matrix.tar.gz
    Ignored:    data/brain_counts/
    Ignored:    data/cl.obo
    Ignored:    data/cl.owl
    Ignored:    data/jurkat/
    Ignored:    data/jurkat:293t_50:50_filtered_gene_bc_matrices.tar.gz
    Ignored:    data/jurkat_293t/
    Ignored:    data/jurkat_filtered_gene_bc_matrices.tar.gz
    Ignored:    data/pbmc20k/
    Ignored:    data/pbmc20k_seurat/
    Ignored:    data/pbmc3k.h5ad
    Ignored:    data/pbmc3k/
    Ignored:    data/pbmc3k_bpcells_mat/
    Ignored:    data/pbmc3k_export.mtx
    Ignored:    data/pbmc3k_matrix.mtx
    Ignored:    data/pbmc3k_seurat.rds
    Ignored:    data/pbmc4k_filtered_gene_bc_matrices.tar.gz
    Ignored:    data/pbmc_1k_v3_filtered_feature_bc_matrix.h5
    Ignored:    data/pbmc_1k_v3_raw_feature_bc_matrix.h5
    Ignored:    data/refdata-gex-GRCh38-2020-A.tar.gz
    Ignored:    data/seurat_1m_neuron.rds
    Ignored:    data/t_3k_filtered_gene_bc_matrices.tar.gz
    Ignored:    r_packages_4.4.1/
    Ignored:    r_packages_4.5.0/

Untracked files:
    Untracked:  analysis/bioc_scrnaseq.Rmd
    Untracked:  bpcells_matrix/
    Untracked:  data/Caenorhabditis_elegans.WBcel235.113.gtf.gz
    Untracked:  data/GCF_043380555.1-RS_2024_12_gene_ontology.gaf.gz
    Untracked:  data/arab.rds
    Untracked:  data/astronomicalunit.csv
    Untracked:  data/femaleMiceWeights.csv
    Untracked:  m3/

Unstaged changes:
    Modified:   analysis/isoform_switch_analyzer.Rmd

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not
included in this status report because it is ok for generated content to
have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">
<p>
These are the previous versions of the repository in which changes were
made to the R Markdown (<code>analysis/mixed_effects_models.Rmd</code>)
and HTML (<code>docs/mixed_effects_models.html</code>) files. If you’ve
configured a remote Git repository (see <code>?wflow_git_remote</code>),
click on the hyperlinks in the table below to view the files as they
were in that past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/davetang/muse/blob/638126d31118ca8ce0b521bf83435131e332d213/analysis/mixed_effects_models.Rmd" target="_blank">638126d</a>
</td>
<td>
Dave Tang
</td>
<td>
2025-10-24
</td>
<td>
AIC and BIC
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/davetang/muse/0400650595bc25d3437d877795980e9c385d55c3/docs/mixed_effects_models.html" target="_blank">0400650</a>
</td>
<td>
Dave Tang
</td>
<td>
2025-10-21
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/davetang/muse/blob/c20a3880b4cac31444dcf80cca0340b741e4e941/analysis/mixed_effects_models.Rmd" target="_blank">c20a388</a>
</td>
<td>
Dave Tang
</td>
<td>
2025-10-21
</td>
<td>
Random effects specification in R’s formula syntax
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/davetang/muse/238e4062e173a8ce7473f174344c033f676cda0a/docs/mixed_effects_models.html" target="_blank">238e406</a>
</td>
<td>
Dave Tang
</td>
<td>
2025-10-21
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/davetang/muse/blob/96acdc584f3a8e3b924741d32078a74740ff7329/analysis/mixed_effects_models.Rmd" target="_blank">96acdc5</a>
</td>
<td>
Dave Tang
</td>
<td>
2025-10-21
</td>
<td>
Mixed-effects models
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<div id="introduction" class="section level2">
<h2>Introduction</h2>
<p>Mixed-effects models (also known as multilevel models, hierarchical
models, or random-effects models) are extensions of linear regression
that account for both fixed and random effects.</p>
<p>Given students’ test scores across multiple schools:</p>
<ul>
<li>Fixed effects are predictors that apply to everyone such as the
effect of study hours.</li>
<li>Random effects are effects that vary across groups such as
school-to-school differences.</li>
</ul>
<p>Each school might have a slightly different average score (intercept)
or respond differently to study hours (slope).</p>
<p>Mixed-effects models handle this by modeling:</p>
<p><span class="math display">\[
\text{Score}*{ij} = \beta_0 + \beta_1 \text{Hours}*{ij} + u_{0j} +
\epsilon_{ij}
\]</span></p>
<p>where:</p>
<ul>
<li><span class="math inline">\(( j )\)</span>: the group index, i.e.,
the school.</li>
<li><span class="math inline">\(( i )\)</span>: the individual
(observation) index within each group, i.e., the student within school
<span class="math inline">\(( j )\)</span>.</li>
</ul>
<p>Each school <span class="math inline">\(( j )\)</span> has its own
intercept <span class="math inline">\(( u_{0j} )\)</span> that
represents how that school’s mean score differs from the overall
population mean <span class="math inline">\(( \beta_0 )\)</span>.</p>
<p>Each student <span class="math inline">\(( i )\)</span> within that
school has their own hours studied (<span class="math inline">\((
\text{Hours}*{ij} )\)</span>) and residual error (<span
class="math inline">\(( \epsilon_{ij} )\)</span>).</p>
</div>
<div id="random-effects" class="section level2">
<h2>Random effects</h2>
<p>A random effect is a variable whose levels are drawn from a larger
population, not specifically fixed.</p>
<p>For instance:</p>
<ul>
<li>If we measured 10 schools, they are a sample from all possible
schools.</li>
<li>We assume their effects (intercepts/slopes) come from a distribution
(often Normal).</li>
</ul>
<p><strong>Random effects account for correlation among observations
within the same group and share information across groups</strong>.</p>
</div>
<div id="example-in-r" class="section level2">
<h2>Example in R</h2>
<p>Install dependencies.</p>
<pre class="r"><code>install.packages(&quot;lme4&quot;)</code></pre>
<p>Example data: Reaction times in a sleep deprivation study.</p>
<blockquote>
<p>These data are from the study described in Belenky et al. (2003), for
the most sleep-deprived group (3 hours time-in-bed) and for the first 10
days of the study, up to the recovery period. The original study
analyzed speed (1/(reaction time)) and treated day as a categorical
rather than a continuous predictor.</p>
</blockquote>
<pre class="r"><code>library(lme4)</code></pre>
<pre><code>Loading required package: Matrix</code></pre>
<pre><code>
Attaching package: &#39;Matrix&#39;</code></pre>
<pre><code>The following objects are masked from &#39;package:tidyr&#39;:

    expand, pack, unpack</code></pre>
<pre class="r"><code>data(&quot;sleepstudy&quot;)
head(sleepstudy)</code></pre>
<pre><code>  Reaction Days Subject
1 249.5600    0     308
2 258.7047    1     308
3 250.8006    2     308
4 321.4398    3     308
5 356.8519    4     308
6 414.6901    5     308</code></pre>
<p>Visualise the data.</p>
<pre class="r"><code>ggplot(sleepstudy, aes(x = Days, y = Reaction, group = Subject, colour = Subject)) +
  geom_line() +
  geom_point() +
  theme_minimal() +
  labs(
    title = &quot;Reaction Time over Days of Sleep Deprivation&quot;,
    y = &quot;Average reaction time (ms)&quot;,
    x = &quot;Number of days of sleep deprivation&quot;
  ) +
  theme(legend.position = &quot;none&quot;)</code></pre>
<p><img src="figure/mixed_effects_models.Rmd/sleepstudy_by_subject-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-sleepstudy_by_subject-1">
Past versions of sleepstudy_by_subject-1.png
</button>
</p>
<div id="fig-sleepstudy_by_subject-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/davetang/muse/blob/238e4062e173a8ce7473f174344c033f676cda0a/docs/figure/mixed_effects_models.Rmd/sleepstudy_by_subject-1.png" target="_blank">238e406</a>
</td>
<td>
Dave Tang
</td>
<td>
2025-10-21
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>Fit a simple linear model (no random effects).</p>
<pre class="r"><code>lm_model &lt;- lm(Reaction ~ Days, data = sleepstudy)
summary(lm_model)</code></pre>
<pre><code>
Call:
lm(formula = Reaction ~ Days, data = sleepstudy)

Residuals:
     Min       1Q   Median       3Q      Max 
-110.848  -27.483    1.546   26.142  139.953 

Coefficients:
            Estimate Std. Error t value Pr(&gt;|t|)    
(Intercept)  251.405      6.610  38.033  &lt; 2e-16 ***
Days          10.467      1.238   8.454 9.89e-15 ***
---
Signif. codes:  0 &#39;***&#39; 0.001 &#39;**&#39; 0.01 &#39;*&#39; 0.05 &#39;.&#39; 0.1 &#39; &#39; 1

Residual standard error: 47.71 on 178 degrees of freedom
Multiple R-squared:  0.2865,    Adjusted R-squared:  0.2825 
F-statistic: 71.46 on 1 and 178 DF,  p-value: 9.894e-15</code></pre>
<p>This assumes everyone has the same intercept and slope but each
subject behaves differently.</p>
<p>Fit a mixed-effects model (random intercepts).</p>
<pre class="r"><code>m1 &lt;- lmer(Reaction ~ Days + (1 | Subject), data = sleepstudy)
summary(m1)</code></pre>
<pre><code>Linear mixed model fit by REML [&#39;lmerMod&#39;]
Formula: Reaction ~ Days + (1 | Subject)
   Data: sleepstudy

REML criterion at convergence: 1786.5

Scaled residuals: 
    Min      1Q  Median      3Q     Max 
-3.2257 -0.5529  0.0109  0.5188  4.2506 

Random effects:
 Groups   Name        Variance Std.Dev.
 Subject  (Intercept) 1378.2   37.12   
 Residual              960.5   30.99   
Number of obs: 180, groups:  Subject, 18

Fixed effects:
            Estimate Std. Error t value
(Intercept) 251.4051     9.7467   25.79
Days         10.4673     0.8042   13.02

Correlation of Fixed Effects:
     (Intr)
Days -0.371</code></pre>
<ul>
<li><code>Reaction</code> - The <strong>response variable</strong>
(dependent variable).</li>
<li><code>Days</code> - A <strong>fixed effect</strong> (predictor that
applies to everyone).</li>
<li><code>(1 | Subject)</code> - A <strong>random effect</strong> where
each Subject has their own intercept.
<ul>
<li><code>lme4::lmer()</code> (and similar functions), random effects
always have the structure:
<code>(random effects terms | grouping factor)</code></li>
<li><code>(1 | Subject)</code> means allow the intercept (1) to vary by
<code>Subject</code>; in a model formula, 1 represents the intercept
term, which is the baseline level of the response variable when all
predictors = 0.</li>
</ul></li>
<li><code>+</code> - Combine fixed and random effects.</li>
</ul>
<p>The model says:</p>
<p><span class="math display">\[
\text{Reaction}*{ij} = \beta_0 + \beta_1 \text{Days}*{ij} + u_{0j} +
\epsilon_{ij}
\]</span></p>
<p>where:</p>
<ul>
<li><span class="math inline">\(( i )\)</span> = measurement within
subject ( j ).</li>
<li><span class="math inline">\(( j )\)</span> = subject.</li>
<li><span class="math inline">\(( \beta_0 )\)</span>: population-level
intercept (average reaction time at Day 0).</li>
<li><span class="math inline">\(( \beta_1 )\)</span>: population-level
slope (effect of each additional day).</li>
<li><span class="math inline">\(( u_{0j} )\)</span>: random intercept
(how much each subject deviates from the average intercept).</li>
<li><span class="math inline">\(( \epsilon_{ij} )\)</span>: residual
error.</li>
</ul>
<p>Each subject has their <em>own baseline</em> reaction time, but we
assume the <em>same slope</em> (same rate of slowing down).</p>
<p>Fit multiple models with increasing complexity and calculate AIC
(Akaike Information Criterion) and BIC (Bayesian Information Criterion),
which are information criteria used for model selection. They help you
balance:</p>
<ul>
<li>Model fit (how well the model explains the data)</li>
<li>Model complexity (how many parameters the model uses)</li>
</ul>
<p>The actual numbers are meaningless by themselves and what matters is
the relative comparison between models:</p>
<ul>
<li>Lower AIC/BIC = better model fit</li>
<li>Delta AIC/BIC shows difference from best model</li>
<li>Delta &lt; 2: substantial support</li>
<li>Delta 4-7: considerably less support</li>
<li>Delta &gt; 10: essentially no support</li>
</ul>
<pre class="r"><code>m0 &lt;- lm(Reaction ~ Days, data = sleepstudy)

# Model 1: Random intercept only
m1 &lt;- lmer(Reaction ~ Days + (1 | Subject), data = sleepstudy)

# Model 2: Random intercept and random slope
m2 &lt;- lmer(Reaction ~ Days + (Days | Subject), data = sleepstudy)

# Model 3: Random intercept and random slope (uncorrelated)
m3 &lt;- lmer(Reaction ~ Days + (Days || Subject), data = sleepstudy)

aic_values &lt;- c(
  M0_lm = AIC(m0),
  M1_random_intercept = AIC(m1),
  M2_random_slope_corr = AIC(m2),
  M3_random_slope_uncorr = AIC(m3)
)

bic_values &lt;- c(
  M0_lm = BIC(m0),
  M1_random_intercept = BIC(m1),
  M2_random_slope_corr = BIC(m2),
  M3_random_slope_uncorr = BIC(m3)
)

comparison_df &lt;- data.frame(
  Model = c(&quot;M0: Linear model&quot;, 
            &quot;M1: Random intercept&quot;, 
            &quot;M2: Random intercept + slope (correlated)&quot;,
            &quot;M3: Random intercept + slope (uncorrelated)&quot;),
  AIC = round(aic_values, 2),
  BIC = round(bic_values, 2),
  Delta_AIC = round(aic_values - min(aic_values), 2),
  Delta_BIC = round(bic_values - min(bic_values), 2)
)

comparison_df</code></pre>
<pre><code>                                                             Model     AIC
M0_lm                                             M0: Linear model 1906.29
M1_random_intercept                           M1: Random intercept 1794.47
M2_random_slope_corr     M2: Random intercept + slope (correlated) 1755.63
M3_random_slope_uncorr M3: Random intercept + slope (uncorrelated) 1753.67
                           BIC Delta_AIC Delta_BIC
M0_lm                  1915.87    152.62    146.24
M1_random_intercept    1807.24     40.80     37.60
M2_random_slope_corr   1774.79      1.96      5.15
M3_random_slope_uncorr 1769.63      0.00      0.00</code></pre>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span>
Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 4.5.0 (2025-04-11)
Platform: x86_64-pc-linux-gnu
Running under: Ubuntu 24.04.3 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 
LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.26.so;  LAPACK version 3.12.0

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

time zone: Etc/UTC
tzcode source: system (glibc)

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] lme4_1.1-37     Matrix_1.7-3    lubridate_1.9.4 forcats_1.0.0  
 [5] stringr_1.5.1   dplyr_1.1.4     purrr_1.0.4     readr_2.1.5    
 [9] tidyr_1.3.1     tibble_3.3.0    ggplot2_3.5.2   tidyverse_2.0.0
[13] workflowr_1.7.1

loaded via a namespace (and not attached):
 [1] gtable_0.3.6       xfun_0.52          bslib_0.9.0        processx_3.8.6    
 [5] lattice_0.22-6     callr_3.7.6        tzdb_0.5.0         vctrs_0.6.5       
 [9] tools_4.5.0        Rdpack_2.6.4       ps_1.9.1           generics_0.1.4    
[13] pkgconfig_2.0.3    RColorBrewer_1.1-3 lifecycle_1.0.4    compiler_4.5.0    
[17] farver_2.1.2       git2r_0.36.2       getPass_0.2-4      httpuv_1.6.16     
[21] htmltools_0.5.8.1  sass_0.4.10        yaml_2.3.10        later_1.4.2       
[25] pillar_1.10.2      nloptr_2.2.1       jquerylib_0.1.4    whisker_0.4.1     
[29] MASS_7.3-65        cachem_1.1.0       reformulas_0.4.1   boot_1.3-31       
[33] nlme_3.1-168       tidyselect_1.2.1   digest_0.6.37      stringi_1.8.7     
[37] labeling_0.4.3     splines_4.5.0      rprojroot_2.0.4    fastmap_1.2.0     
[41] grid_4.5.0         cli_3.6.5          magrittr_2.0.3     withr_3.0.2       
[45] scales_1.4.0       promises_1.3.3     timechange_0.3.0   rmarkdown_2.29    
[49] httr_1.4.7         hms_1.1.3          evaluate_1.0.3     knitr_1.50        
[53] rbibutils_2.3      rlang_1.1.6        Rcpp_1.0.14        glue_1.8.0        
[57] rstudioapi_0.17.1  minqa_1.2.8        jsonlite_2.0.0     R6_2.6.1          
[61] fs_1.6.6          </code></pre>
</div>
</div>
</div>
</div>
</div>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
https://docs.mathjax.org/en/latest/web/configuration.html. This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>




</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
