<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />



<meta name="date" content="2026-02-04" />

<title>Odds Ratio</title>

<script src="site_libs/header-attrs-2.29/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.13.2/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-6.5.2/css/all.min.css" rel="stylesheet" />
<link href="site_libs/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet" />

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/main/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">My blog</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/davetang/muse">
    <span class="fa fa-github"></span>
     
    Source code
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Odds Ratio</h1>
<h4 class="date">2026-02-04</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span>
workflowr <span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2026-02-04
</p>
<p>
<strong>Checks:</strong> <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 7
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong> <code>muse/</code> <span
class="glyphicon glyphicon-question-sign" aria-hidden="true"
title="This is the local directory in which the code in this file was executed.">
</span>
</p>
<p>
This reproducible <a href="https://rmarkdown.rstudio.com">R Markdown</a>
analysis was created with <a
  href="https://github.com/workflowr/workflowr">workflowr</a> (version
1.7.1). The <em>Checks</em> tab describes the reproducibility checks
that were applied when the results were created. The <em>Past
versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date
</a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git
repository, you know the exact version of the code that produced these
results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the
global environment can affect the analysis in your R Markdown file in
unknown ways. For reproduciblity it’s best to always run the code in an
empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20200712code">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Seed:</strong>
<code>set.seed(20200712)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20200712code"
class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20200712)</code> was run prior to running
the code in the R Markdown file. Setting a seed ensures that any results
that rely on randomness, e.g. subsampling or permutations, are
reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Session information:</strong>
recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package
versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be
confident that you successfully produced the results during this
run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr
project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomdavetangmusetreebfee866b0645257af6ea5a9de2c0ea555aa81e45targetblankbfee866a">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Repository version:</strong>
<a href="https://github.com/davetang/muse/tree/bfee866b0645257af6ea5a9de2c0ea555aa81e45" target="_blank">bfee866</a>
</a>
</p>
</div>
<div
id="strongRepositoryversionstrongahrefhttpsgithubcomdavetangmusetreebfee866b0645257af6ea5a9de2c0ea555aa81e45targetblankbfee866a"
class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development
and connecting the code version to the results is critical for
reproducibility.
</p>
<p>
The results in this page were generated with repository version
<a href="https://github.com/davetang/muse/tree/bfee866b0645257af6ea5a9de2c0ea555aa81e45" target="_blank">bfee866</a>.
See the <em>Past versions</em> tab to see a history of the changes made
to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for
the analysis have been committed to Git prior to generating the results
(you can use <code>wflow_publish</code> or
<code>wflow_git_commit</code>). workflowr only checks the R Markdown
file, but you know if there are other scripts or data files that it
depends on. Below is the status of the Git repository when the results
were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .Rproj.user/
    Ignored:    data/1M_neurons_filtered_gene_bc_matrices_h5.h5
    Ignored:    data/293t/
    Ignored:    data/293t_3t3_filtered_gene_bc_matrices.tar.gz
    Ignored:    data/293t_filtered_gene_bc_matrices.tar.gz
    Ignored:    data/5k_Human_Donor1_PBMC_3p_gem-x_5k_Human_Donor1_PBMC_3p_gem-x_count_sample_filtered_feature_bc_matrix.h5
    Ignored:    data/5k_Human_Donor2_PBMC_3p_gem-x_5k_Human_Donor2_PBMC_3p_gem-x_count_sample_filtered_feature_bc_matrix.h5
    Ignored:    data/5k_Human_Donor3_PBMC_3p_gem-x_5k_Human_Donor3_PBMC_3p_gem-x_count_sample_filtered_feature_bc_matrix.h5
    Ignored:    data/5k_Human_Donor4_PBMC_3p_gem-x_5k_Human_Donor4_PBMC_3p_gem-x_count_sample_filtered_feature_bc_matrix.h5
    Ignored:    data/97516b79-8d08-46a6-b329-5d0a25b0be98.h5ad
    Ignored:    data/Parent_SC3v3_Human_Glioblastoma_filtered_feature_bc_matrix.tar.gz
    Ignored:    data/brain_counts/
    Ignored:    data/cl.obo
    Ignored:    data/cl.owl
    Ignored:    data/jurkat/
    Ignored:    data/jurkat:293t_50:50_filtered_gene_bc_matrices.tar.gz
    Ignored:    data/jurkat_293t/
    Ignored:    data/jurkat_filtered_gene_bc_matrices.tar.gz
    Ignored:    data/pbmc20k/
    Ignored:    data/pbmc20k_seurat/
    Ignored:    data/pbmc3k.csv
    Ignored:    data/pbmc3k.csv.gz
    Ignored:    data/pbmc3k.h5ad
    Ignored:    data/pbmc3k/
    Ignored:    data/pbmc3k_bpcells_mat/
    Ignored:    data/pbmc3k_export.mtx
    Ignored:    data/pbmc3k_matrix.mtx
    Ignored:    data/pbmc3k_seurat.rds
    Ignored:    data/pbmc4k_filtered_gene_bc_matrices.tar.gz
    Ignored:    data/pbmc_1k_v3_filtered_feature_bc_matrix.h5
    Ignored:    data/pbmc_1k_v3_raw_feature_bc_matrix.h5
    Ignored:    data/refdata-gex-GRCh38-2020-A.tar.gz
    Ignored:    data/seurat_1m_neuron.rds
    Ignored:    data/t_3k_filtered_gene_bc_matrices.tar.gz
    Ignored:    r_packages_4.4.1/
    Ignored:    r_packages_4.5.0/

Untracked files:
    Untracked:  .claude/
    Untracked:  CLAUDE.md
    Untracked:  analysis/bioc.Rmd
    Untracked:  analysis/bioc_scrnaseq.Rmd
    Untracked:  analysis/chick_weight.Rmd
    Untracked:  analysis/likelihood.Rmd
    Untracked:  bpcells_matrix/
    Untracked:  data/Caenorhabditis_elegans.WBcel235.113.gtf.gz
    Untracked:  data/GCF_043380555.1-RS_2024_12_gene_ontology.gaf.gz
    Untracked:  data/SeuratObj.rds
    Untracked:  data/arab.rds
    Untracked:  data/astronomicalunit.csv
    Untracked:  data/femaleMiceWeights.csv
    Untracked:  data/lung_bcell.rds
    Untracked:  m3/
    Untracked:  women.json

Unstaged changes:
    Modified:   analysis/isoform_switch_analyzer.Rmd
    Modified:   analysis/linear_models.Rmd

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not
included in this status report because it is ok for generated content to
have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">
<p>
These are the previous versions of the repository in which changes were
made to the R Markdown (<code>analysis/odds_ratio.Rmd</code>) and HTML
(<code>docs/odds_ratio.html</code>) files. If you’ve configured a remote
Git repository (see <code>?wflow_git_remote</code>), click on the
hyperlinks in the table below to view the files as they were in that
past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/davetang/muse/blob/bfee866b0645257af6ea5a9de2c0ea555aa81e45/analysis/odds_ratio.Rmd" target="_blank">bfee866</a>
</td>
<td>
Dave Tang
</td>
<td>
2026-02-04
</td>
<td>
Add more background material
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/davetang/muse/34a78edc1d1ec21c5eec77601acdfb6468a3158d/docs/odds_ratio.html" target="_blank">34a78ed</a>
</td>
<td>
Dave Tang
</td>
<td>
2025-12-05
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/davetang/muse/blob/222d2c42fb202f10164793526700f625341e9512/analysis/odds_ratio.Rmd" target="_blank">222d2c4</a>
</td>
<td>
Dave Tang
</td>
<td>
2025-12-05
</td>
<td>
Odds ratio
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<div id="introduction" class="section level2">
<h2>Introduction</h2>
<p>Odds ratios are widely used to measure <strong>the strength of
association between two variables</strong>, particularly in medical
research, epidemiology, and case-control studies. Before understanding
odds ratios, we need to understand the difference between probability
and odds.</p>
</div>
<div id="probability-vs-odds" class="section level2">
<h2>Probability vs Odds</h2>
<p>Probability and odds are related but different:</p>
<ul>
<li><strong>Probability</strong>: The chance of something happening out
of all possibilities (ranges from 0 to 1)</li>
<li><strong>Odds</strong>: The ratio of something happening to it NOT
happening (ranges from 0 to infinity)</li>
</ul>
<pre class="r"><code>sick &lt;- 2
healthy &lt;- 8
total &lt;- sick + healthy

probability &lt;- sick / total
probability</code></pre>
<pre><code>[1] 0.2</code></pre>
<pre class="r"><code>odds &lt;- sick / healthy
odds</code></pre>
<pre><code>[1] 0.25</code></pre>
<p>In this example:</p>
<ul>
<li>The probability of being sick is 2/10 = 0.2 (or 20%)</li>
<li>The odds of being sick are 2:8 = 0.25 (read as “1 to 4” or “0.25 to
1”)</li>
</ul>
<div id="converting-between-probability-and-odds"
class="section level3">
<h3>Converting Between Probability and Odds</h3>
<p>You can convert between probability (P) and odds using these
formulas:</p>
<p><span class="math display">\[\text{Odds} = \frac{P}{1 -
P}\]</span></p>
<p><span class="math display">\[P = \frac{\text{Odds}}{1 +
\text{Odds}}\]</span></p>
<p>Convert probability to odds.</p>
<pre class="r"><code>prob &lt;- 0.2
odds_from_prob &lt;- prob / (1 - prob)
odds_from_prob</code></pre>
<pre><code>[1] 0.25</code></pre>
<pre class="r"><code># Convert odds back to probability
prob_from_odds &lt;- odds_from_prob / (1 + odds_from_prob)
prob_from_odds</code></pre>
<pre><code>[1] 0.2</code></pre>
</div>
<div id="example-different-probability-values" class="section level3">
<h3>Example: Different Probability Values</h3>
<p>To build intuition, let’s see how odds change across different
probabilities:</p>
<pre class="r"><code>probs &lt;- c(0.1, 0.2, 0.5, 0.8, 0.9)
odds_values &lt;- probs / (1 - probs)

data.frame(
  Probability = probs,
  Odds = round(odds_values, 3),
  Interpretation = c(&quot;1 to 9&quot;, &quot;1 to 4&quot;, &quot;1 to 1 (even)&quot;, &quot;4 to 1&quot;, &quot;9 to 1&quot;)
)</code></pre>
<pre><code>  Probability  Odds Interpretation
1         0.1 0.111         1 to 9
2         0.2 0.250         1 to 4
3         0.5 1.000  1 to 1 (even)
4         0.8 4.000         4 to 1
5         0.9 9.000         9 to 1</code></pre>
<p>Notice that when P = 0.5, the odds are 1 (even odds). When P &gt;
0.5, odds &gt; 1; when P &lt; 0.5, odds &lt; 1.</p>
</div>
</div>
<div id="what-is-an-odds-ratio" class="section level2">
<h2>What is an Odds Ratio?</h2>
<p><strong>An odds ratio (OR) compares the odds of an event between two
groups</strong>. It answers the question: “How many times higher (or
lower) are the odds of the outcome in one group compared to
another?”</p>
<p>Interpreting odds ratio values:</p>
<ul>
<li><strong>OR = 1</strong>: No association between exposure and
outcome</li>
<li><strong>OR &gt; 1</strong>: Exposure is associated with higher odds
of the outcome</li>
<li><strong>OR &lt; 1</strong>: Exposure is associated with lower odds
of the outcome</li>
</ul>
</div>
<div id="smoking-and-lung-cancer-example" class="section level2">
<h2>Smoking and Lung Cancer Example</h2>
<p>Let’s work through a classic example examining the association
between smoking and lung cancer. We have a sample of 200 people broken
down into a 2x2 contingency table:</p>
<pre class="r"><code>data &lt;- matrix(
  c(20, 80, 2, 98),
  nrow = 2,
  byrow = TRUE,
  dimnames = list(
    c(&quot;Smokers&quot;, &quot;Non-smokers&quot;),
    c(&quot;Cancer&quot;, &quot;No Cancer&quot;)
  )
)

data</code></pre>
<pre><code>            Cancer No Cancer
Smokers         20        80
Non-smokers      2        98</code></pre>
<p>The general structure of a 2x2 table is:</p>
<table>
<thead>
<tr class="header">
<th></th>
<th>Disease+</th>
<th>Disease-</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>Exposed</td>
<td>a</td>
<td>b</td>
</tr>
<tr class="even">
<td>Not Exposed</td>
<td>c</td>
<td>d</td>
</tr>
</tbody>
</table>
<p>In our example: a=20, b=80, c=2, d=98.</p>
<div id="step-1-calculate-odds-for-each-group" class="section level3">
<h3>Step 1: Calculate odds for each group</h3>
<pre class="r"><code># Odds of cancer for smokers = a/b
odds_smokers &lt;- data[1, 1] / data[1, 2]
odds_smokers</code></pre>
<pre><code>[1] 0.25</code></pre>
<pre class="r"><code># Odds of cancer for non-smokers = c/d
odds_nonsmokers &lt;- data[2, 1] / data[2, 2]
odds_nonsmokers</code></pre>
<pre><code>[1] 0.02040816</code></pre>
<ul>
<li>Smokers: 20 have cancer, 80 don’t → odds = 20/80 = 0.25 (or
1:4)</li>
<li>Non-smokers: 2 have cancer, 98 don’t → odds = 2/98 ≈ 0.02 (or about
1:49)</li>
</ul>
</div>
<div id="step-2-calculate-odds-ratio" class="section level3">
<h3>Step 2: Calculate odds ratio</h3>
<p>The odds ratio is simply the ratio of these two odds:</p>
<pre class="r"><code>odds_ratio &lt;- odds_smokers / odds_nonsmokers
odds_ratio</code></pre>
<pre><code>[1] 12.25</code></pre>
<p><strong>Interpretation</strong>: Smokers have 12.3 times the odds of
developing lung cancer compared to non-smokers. This is a strong
positive association.</p>
</div>
<div id="cross-product-formula" class="section level3">
<h3>Cross-Product Formula</h3>
<p>For a 2x2 table, the odds ratio can also be calculated directly using
the cross-product formula:</p>
<p><span class="math display">\[OR = \frac{a \times d}{b \times
c}\]</span></p>
<p>Using the cross-product formula.</p>
<pre class="r"><code>or_cross &lt;- (data[1, 1] * data[2, 2]) / (data[1, 2] * data[2, 1])
or_cross</code></pre>
<pre><code>[1] 12.25</code></pre>
<p>This gives the same result and is often more convenient for quick
calculations.</p>
</div>
</div>
<div id="confidence-intervals" class="section level2">
<h2>Confidence Intervals</h2>
<p>A point estimate of the odds ratio alone doesn’t tell us how precise
our estimate is. We need <strong>confidence intervals</strong> to
understand the range of plausible values for the true odds ratio.</p>
<div id="using-fishers-exact-test" class="section level3">
<h3>Using Fisher’s Exact Test</h3>
<p>The <code>fisher.test()</code> function calculates the odds ratio
with a 95% confidence interval:</p>
<pre class="r"><code>result &lt;- fisher.test(data)
result</code></pre>
<pre><code>
    Fisher&#39;s Exact Test for Count Data

data:  data
p-value = 5.091e-05
alternative hypothesis: true odds ratio is not equal to 1
95 percent confidence interval:
   2.809877 110.218221
sample estimates:
odds ratio 
  12.12786 </code></pre>
<p>Key outputs:</p>
<ul>
<li><strong>Odds ratio</strong>: 12.13 (similar to our manual
calculation; slight difference due to conditional maximum likelihood
estimation)</li>
<li><strong>95% CI</strong>: [2.81, 110.22]</li>
<li><strong>p-value</strong>: 5.09e-05</li>
</ul>
<p><strong>Interpreting the confidence interval</strong>: Since the 95%
CI does not include 1, we can conclude there is a statistically
significant association between smoking and lung cancer at the 0.05
significance level.</p>
</div>
<div id="rule-of-thumb-for-confidence-intervals" class="section level3">
<h3>Rule of Thumb for Confidence Intervals</h3>
<ul>
<li>If the 95% CI <strong>excludes 1</strong>: The association is
statistically significant (p &lt; 0.05)</li>
<li>If the 95% CI <strong>includes 1</strong>: The association is not
statistically significant</li>
</ul>
</div>
</div>
<div id="logistic-regression-and-odds-ratios" class="section level2">
<h2>Logistic Regression and Odds Ratios</h2>
<p>Logistic regression is commonly used to model binary outcomes
(yes/no, disease/no disease). The coefficients from logistic regression
are <strong>log odds ratios</strong>, which can be converted to odds
ratios by exponentiation.</p>
<div id="why-log-odds" class="section level3">
<h3>Why Log Odds?</h3>
<p>Logistic regression models the <strong>log odds</strong> (also called
logit) of the outcome:</p>
<p><span class="math display">\[\log\left(\frac{P}{1-P}\right) = \beta_0
+ \beta_1 X_1 + \beta_2 X_2 + ...\]</span></p>
<p>The coefficients (<span class="math inline">\(\beta\)</span>)
represent the change in log odds for a one-unit increase in the
predictor.</p>
</div>
<div id="example-disease-prediction" class="section level3">
<h3>Example: Disease Prediction</h3>
<p>Let’s simulate data where disease risk depends on age and smoking
status:</p>
<pre class="r"><code>set.seed(1984)
n &lt;- 200
age &lt;- rnorm(n, mean = 50, sd = 15)
smoker &lt;- rbinom(n, 1, 0.3)
# True model: log odds = -5 + 0.05*age + 2*smoker
disease &lt;- rbinom(n, 1, plogis(-5 + 0.05*age + 2*smoker))

model &lt;- glm(disease ~ age + smoker, family = binomial)
summary(model)</code></pre>
<pre><code>
Call:
glm(formula = disease ~ age + smoker, family = binomial)

Coefficients:
            Estimate Std. Error z value Pr(&gt;|z|)    
(Intercept) -4.53700    0.87994  -5.156 2.52e-07 ***
age          0.04302    0.01493   2.881  0.00396 ** 
smoker       1.83998    0.40800   4.510 6.49e-06 ***
---
Signif. codes:  0 &#39;***&#39; 0.001 &#39;**&#39; 0.01 &#39;*&#39; 0.05 &#39;.&#39; 0.1 &#39; &#39; 1

(Dispersion parameter for binomial family taken to be 1)

    Null deviance: 188.56  on 199  degrees of freedom
Residual deviance: 157.34  on 197  degrees of freedom
AIC: 163.34

Number of Fisher Scoring iterations: 5</code></pre>
</div>
<div id="converting-coefficients-to-odds-ratios" class="section level3">
<h3>Converting Coefficients to Odds Ratios</h3>
<p>The model coefficients are log odds ratios. To get odds ratios, we
exponentiate them:</p>
<pre class="r"><code>or_coef &lt;- exp(coef(model))
or_coef</code></pre>
<pre><code>(Intercept)         age      smoker 
 0.01070543  1.04395510  6.29641114 </code></pre>
<p><strong>Interpretation</strong>:</p>
<ul>
<li><strong>Age</strong>: OR = 1.044. For each additional year of age,
the odds of disease increase by about 4.4%</li>
<li><strong>Smoker</strong>: OR = 6.3. Smokers have about 6.3 times the
odds of disease compared to non-smokers</li>
</ul>
</div>
<div id="odds-ratios-with-confidence-intervals" class="section level3">
<h3>Odds Ratios with Confidence Intervals</h3>
<p>For a complete picture, we should report odds ratios with confidence
intervals:</p>
<pre class="r"><code># Get confidence intervals for coefficients
ci &lt;- confint(model)</code></pre>
<pre><code>Waiting for profiling to be done...</code></pre>
<pre class="r"><code># Combine into a nice table
or_table &lt;- data.frame(
  OR = exp(coef(model)),
  Lower_CI = exp(ci[, 1]),
  Upper_CI = exp(ci[, 2])
)
round(or_table, 3)</code></pre>
<pre><code>               OR Lower_CI Upper_CI
(Intercept) 0.011    0.002    0.054
age         1.044    1.015    1.076
smoker      6.296    2.878   14.392</code></pre>
</div>
</div>
<div id="summary" class="section level2">
<h2>Summary</h2>
<p>Key points about odds ratios:</p>
<ol style="list-style-type: decimal">
<li><strong>Odds</strong> are the ratio of an event occurring to it not
occurring</li>
<li><strong>Odds ratios</strong> compare odds between two groups</li>
<li>OR = 1 means no association; OR &gt; 1 means positive association;
OR &lt; 1 means negative association</li>
<li>Always report confidence intervals alongside point estimates</li>
<li>In logistic regression, exponentiate coefficients to obtain odds
ratios</li>
<li>The <strong>cross-product formula</strong> (ad/bc) provides a quick
way to calculate OR from 2x2 tables</li>
</ol>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span>
Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 4.5.0 (2025-04-11)
Platform: x86_64-pc-linux-gnu
Running under: Ubuntu 24.04.3 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 
LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.26.so;  LAPACK version 3.12.0

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

time zone: Etc/UTC
tzcode source: system (glibc)

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] lubridate_1.9.4 forcats_1.0.0   stringr_1.5.1   dplyr_1.1.4    
 [5] purrr_1.0.4     readr_2.1.5     tidyr_1.3.1     tibble_3.3.0   
 [9] ggplot2_3.5.2   tidyverse_2.0.0 workflowr_1.7.1

loaded via a namespace (and not attached):
 [1] sass_0.4.10        generics_0.1.4     stringi_1.8.7      hms_1.1.3         
 [5] digest_0.6.37      magrittr_2.0.3     timechange_0.3.0   evaluate_1.0.3    
 [9] grid_4.5.0         RColorBrewer_1.1-3 fastmap_1.2.0      rprojroot_2.0.4   
[13] jsonlite_2.0.0     processx_3.8.6     whisker_0.4.1      ps_1.9.1          
[17] promises_1.3.3     httr_1.4.7         scales_1.4.0       jquerylib_0.1.4   
[21] cli_3.6.5          rlang_1.1.6        withr_3.0.2        cachem_1.1.0      
[25] yaml_2.3.10        tools_4.5.0        tzdb_0.5.0         httpuv_1.6.16     
[29] vctrs_0.6.5        R6_2.6.1           lifecycle_1.0.4    git2r_0.36.2      
[33] fs_1.6.6           pkgconfig_2.0.3    callr_3.7.6        pillar_1.10.2     
[37] bslib_0.9.0        later_1.4.2        gtable_0.3.6       glue_1.8.0        
[41] Rcpp_1.0.14        xfun_0.52          tidyselect_1.2.1   rstudioapi_0.17.1 
[45] knitr_1.50         farver_2.1.2       htmltools_0.5.8.1  rmarkdown_2.29    
[49] compiler_4.5.0     getPass_0.2-4     </code></pre>
</div>
</div>
</div>
</div>
</div>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
https://docs.mathjax.org/en/latest/web/configuration.html. This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>




</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
