<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />



<meta name="date" content="2025-03-05" />

<title>Doublet prediction using {scDblFinder}</title>

<script src="site_libs/header-attrs-2.28/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.13.2/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-6.4.2/css/all.min.css" rel="stylesheet" />
<link href="site_libs/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet" />

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/main/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">My blog</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/davetang/muse">
    <span class="fa fa-github"></span>
     
    Source code
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Doublet prediction using {scDblFinder}</h1>
<h4 class="date">2025-03-05</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span>
workflowr <span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2025-03-05
</p>
<p>
<strong>Checks:</strong> <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 7
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong> <code>muse/</code> <span
class="glyphicon glyphicon-question-sign" aria-hidden="true"
title="This is the local directory in which the code in this file was executed.">
</span>
</p>
<p>
This reproducible <a href="https://rmarkdown.rstudio.com">R Markdown</a>
analysis was created with <a
  href="https://github.com/workflowr/workflowr">workflowr</a> (version
1.7.1). The <em>Checks</em> tab describes the reproducibility checks
that were applied when the results were created. The <em>Past
versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date
</a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git
repository, you know the exact version of the code that produced these
results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the
global environment can affect the analysis in your R Markdown file in
unknown ways. For reproduciblity it’s best to always run the code in an
empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20200712code">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Seed:</strong>
<code>set.seed(20200712)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20200712code"
class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20200712)</code> was run prior to running
the code in the R Markdown file. Setting a seed ensures that any results
that rely on randomness, e.g. subsampling or permutations, are
reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Session information:</strong>
recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package
versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be
confident that you successfully produced the results during this
run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr
project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomdavetangmusetreecc8428f053fe26cb050434e535af9acc024b8344targetblankcc8428fa">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Repository version:</strong>
<a href="https://github.com/davetang/muse/tree/cc8428f053fe26cb050434e535af9acc024b8344" target="_blank">cc8428f</a>
</a>
</p>
</div>
<div
id="strongRepositoryversionstrongahrefhttpsgithubcomdavetangmusetreecc8428f053fe26cb050434e535af9acc024b8344targetblankcc8428fa"
class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development
and connecting the code version to the results is critical for
reproducibility.
</p>
<p>
The results in this page were generated with repository version
<a href="https://github.com/davetang/muse/tree/cc8428f053fe26cb050434e535af9acc024b8344" target="_blank">cc8428f</a>.
See the <em>Past versions</em> tab to see a history of the changes made
to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for
the analysis have been committed to Git prior to generating the results
(you can use <code>wflow_publish</code> or
<code>wflow_git_commit</code>). workflowr only checks the R Markdown
file, but you know if there are other scripts or data files that it
depends on. Below is the status of the Git repository when the results
were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .Rproj.user/
    Ignored:    data/1M_neurons_filtered_gene_bc_matrices_h5.h5
    Ignored:    data/293t/
    Ignored:    data/293t_3t3_filtered_gene_bc_matrices.tar.gz
    Ignored:    data/293t_filtered_gene_bc_matrices.tar.gz
    Ignored:    data/5k_Human_Donor1_PBMC_3p_gem-x_5k_Human_Donor1_PBMC_3p_gem-x_count_sample_filtered_feature_bc_matrix.h5
    Ignored:    data/5k_Human_Donor2_PBMC_3p_gem-x_5k_Human_Donor2_PBMC_3p_gem-x_count_sample_filtered_feature_bc_matrix.h5
    Ignored:    data/5k_Human_Donor3_PBMC_3p_gem-x_5k_Human_Donor3_PBMC_3p_gem-x_count_sample_filtered_feature_bc_matrix.h5
    Ignored:    data/5k_Human_Donor4_PBMC_3p_gem-x_5k_Human_Donor4_PBMC_3p_gem-x_count_sample_filtered_feature_bc_matrix.h5
    Ignored:    data/Parent_SC3v3_Human_Glioblastoma_filtered_feature_bc_matrix.tar.gz
    Ignored:    data/brain_counts/
    Ignored:    data/cl.obo
    Ignored:    data/cl.owl
    Ignored:    data/jurkat/
    Ignored:    data/jurkat:293t_50:50_filtered_gene_bc_matrices.tar.gz
    Ignored:    data/jurkat_293t/
    Ignored:    data/jurkat_filtered_gene_bc_matrices.tar.gz
    Ignored:    data/pbmc20k/
    Ignored:    data/pbmc20k_seurat/
    Ignored:    data/pbmc3k/
    Ignored:    data/pbmc4k_filtered_gene_bc_matrices.tar.gz
    Ignored:    data/refdata-gex-GRCh38-2020-A.tar.gz
    Ignored:    data/seurat_1m_neuron.rds
    Ignored:    data/t_3k_filtered_gene_bc_matrices.tar.gz
    Ignored:    r_packages_4.4.1/

Untracked files:
    Untracked:  Nothobranchius_furzeri.Nfu_20140520.113.gtf.gz
    Untracked:  analysis/bioc_scrnaseq.Rmd
    Untracked:  data/pbmc_1k_v3_filtered_feature_bc_matrix.h5
    Untracked:  data/pbmc_1k_v3_raw_feature_bc_matrix.h5

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not
included in this status report because it is ok for generated content to
have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">
<p>
These are the previous versions of the repository in which changes were
made to the R Markdown (<code>analysis/scdblfinder.Rmd</code>) and HTML
(<code>docs/scdblfinder.html</code>) files. If you’ve configured a
remote Git repository (see <code>?wflow_git_remote</code>), click on the
hyperlinks in the table below to view the files as they were in that
past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/davetang/muse/blob/cc8428f053fe26cb050434e535af9acc024b8344/analysis/scdblfinder.Rmd" target="_blank">cc8428f</a>
</td>
<td>
Dave Tang
</td>
<td>
2025-03-05
</td>
<td>
Combined batch run versus isolated run
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/davetang/muse/e6e32688e00a20a4eb999959b4aa89d124a49c47/docs/scdblfinder.html" target="_blank">e6e3268</a>
</td>
<td>
Dave Tang
</td>
<td>
2025-03-05
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/davetang/muse/blob/0f0153124814105e600db474e41bb42dbd8add43/analysis/scdblfinder.Rmd" target="_blank">0f01531</a>
</td>
<td>
Dave Tang
</td>
<td>
2025-03-05
</td>
<td>
Predicting doublets using scDblFinder
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<p><a
href="https://bioconductor.org/packages/release/bioc/html/scDblFinder.html">scDblFinder</a>:</p>
<blockquote>
<p>The scDblFinder package gathers various methods for the detection and
handling of doublets/multiplets in single-cell sequencing data
(i.e. multiple cells captured within the same droplet or reaction
volume). It includes methods formerly found in the scran package, the
new fast and comprehensive scDblFinder method, and a reimplementation of
the Amulet detection method for single-cell ATAC-seq.</p>
</blockquote>
<div id="dependencies" class="section level2">
<h2>Dependencies</h2>
<p>Install Bioconductor packages using
<code>BiocManager::install()</code>.</p>
<pre class="r"><code>if (!require(&quot;BiocManager&quot;, quietly = TRUE))
  install.packages(&quot;BiocManager&quot;)

BiocManager::install(&quot;scDblFinder&quot;)
BiocManager::install(&quot;TENxIO&quot;)</code></pre>
<p>Load libraries.</p>
<pre class="r"><code>suppressPackageStartupMessages(library(scDblFinder))
suppressPackageStartupMessages(library(SingleCellExperiment))
suppressPackageStartupMessages(library(TENxIO))</code></pre>
</div>
<div id="data" class="section level2">
<h2>Data</h2>
<p>Download data.</p>
<pre class="r"><code>outdir &lt;- &#39;data/&#39;
filtered_h5_url &lt;- &#39;https://cf.10xgenomics.com/samples/cell-exp/3.0.0/pbmc_1k_v3/pbmc_1k_v3_filtered_feature_bc_matrix.h5&#39;
filtered_h5 &lt;- paste0(outdir, basename(filtered_h5_url))

pbmc5k_1_url &lt;- &quot;https://cf.10xgenomics.com/samples/cell-exp/9.0.0/5k_Human_Donor1_PBMC_3p_gem-x_5k_Human_Donor1_PBMC_3p_gem-x/5k_Human_Donor1_PBMC_3p_gem-x_5k_Human_Donor1_PBMC_3p_gem-x_count_sample_filtered_feature_bc_matrix.h5&quot;
pbmc5k_2_url &lt;- &quot;https://cf.10xgenomics.com/samples/cell-exp/9.0.0/5k_Human_Donor2_PBMC_3p_gem-x_5k_Human_Donor2_PBMC_3p_gem-x/5k_Human_Donor2_PBMC_3p_gem-x_5k_Human_Donor2_PBMC_3p_gem-x_count_sample_filtered_feature_bc_matrix.h5&quot;
pbmc5k_1_h5 &lt;- paste0(outdir, basename(pbmc5k_1_url))
pbmc5k_2_h5 &lt;- paste0(outdir, basename(pbmc5k_2_url))

download_file &lt;- function(url, outfile){
  fn &lt;- basename(url)
  if(file.exists(outfile) == FALSE){
    download.file(url, destfile = outfile)
  } else {
    message(paste0(outfile, &quot; already exists&quot;))
  }
}

download_file(filtered_h5_url, filtered_h5)</code></pre>
<pre><code>data/pbmc_1k_v3_filtered_feature_bc_matrix.h5 already exists</code></pre>
<pre class="r"><code>download_file(pbmc5k_1_url, pbmc5k_1_h5)</code></pre>
<pre><code>data/5k_Human_Donor1_PBMC_3p_gem-x_5k_Human_Donor1_PBMC_3p_gem-x_count_sample_filtered_feature_bc_matrix.h5 already exists</code></pre>
<pre class="r"><code>download_file(pbmc5k_2_url, pbmc5k_2_h5)</code></pre>
<pre><code>data/5k_Human_Donor2_PBMC_3p_gem-x_5k_Human_Donor2_PBMC_3p_gem-x_count_sample_filtered_feature_bc_matrix.h5 already exists</code></pre>
</div>
<div id="create-objects" class="section level2">
<h2>Create objects</h2>
<p>Create <code>SingleCellExperiment</code> files.</p>
<pre class="r"><code>create_sce_obj &lt;- function(h5){
  con &lt;- TENxH5(h5)
  import(con)
}

pbmc1k &lt;- create_sce_obj(filtered_h5)</code></pre>
<pre><code>Warning in rhdf5::h5read(file, name = paste0(group, ranges), index = list(1L),
: Object &#39;matrix/features/interval&#39; does not exist in this HDF5 file.</code></pre>
<pre class="r"><code>pbmc1k</code></pre>
<pre><code>class: SingleCellExperiment 
dim: 33538 1222 
metadata(1): TENxFile
assays(1): counts
rownames(33538): ENSG00000243485 ENSG00000237613 ... ENSG00000277475
  ENSG00000268674
rowData names(3): ID Symbol Type
colnames(1222): AAACCCAAGGAGAGTA-1 AAACGCTTCAGCCCAG-1 ...
  TTTGGTTGTAGAATAC-1 TTTGTTGCAATTAGGA-1
colData names(0):
reducedDimNames(0):
mainExpName: Gene Expression
altExpNames(0):</code></pre>
</div>
<div id="predict-doublets" class="section level2">
<h2>Predict doublets</h2>
<p>Run <code>scDblFinder()</code>.</p>
<pre class="r"><code>pbmc1k.pred &lt;- scDblFinder(pbmc1k)</code></pre>
<pre><code>Creating ~1500 artificial doublets...</code></pre>
<pre><code>Dimensional reduction</code></pre>
<pre><code>Evaluating kNN...</code></pre>
<pre><code>Training model...</code></pre>
<pre><code>iter=0, 155 cells excluded from training.</code></pre>
<pre><code>iter=1, 142 cells excluded from training.</code></pre>
<pre><code>iter=2, 135 cells excluded from training.</code></pre>
<pre><code>Threshold found:0.294</code></pre>
<pre><code>26 (2.1%) doublets called</code></pre>
<pre class="r"><code>colData(pbmc1k.pred) |&gt;
  as.data.frame() -&gt; pbmc1k.pred

head(pbmc1k.pred)</code></pre>
<pre><code>                   scDblFinder.class scDblFinder.score scDblFinder.weighted
AAACCCAAGGAGAGTA-1           singlet      0.0169715248           0.40543264
AAACGCTTCAGCCCAG-1           singlet      0.0008992357           0.29487536
AAAGAACAGACGACTG-1           singlet      0.0003129665           0.21450453
AAAGAACCAATGGCAG-1           singlet      0.0005328555           0.22582183
AAAGAACGTCTGCAAT-1           singlet      0.0127416188           0.37730614
AAAGGATAGTAGACAT-1           singlet      0.0020924010           0.07760762
                   scDblFinder.cxds_score
AAACCCAAGGAGAGTA-1            0.006469792
AAACGCTTCAGCCCAG-1            0.049250538
AAAGAACAGACGACTG-1            0.074387442
AAAGAACCAATGGCAG-1            0.050214668
AAAGAACGTCTGCAAT-1            0.103131556
AAAGGATAGTAGACAT-1            0.075818423</code></pre>
<p>Plot scores per label.</p>
<pre class="r"><code>boxplot(scDblFinder.score ~ scDblFinder.class, data = pbmc1k.pred, pch = 16)</code></pre>
<p><img src="figure/scdblfinder.Rmd/boxplot-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-boxplot-1">
Past versions of boxplot-1.png
</button>
</p>
<div id="fig-boxplot-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/davetang/muse/blob/e6e32688e00a20a4eb999959b4aa89d124a49c47/docs/figure/scdblfinder.Rmd/boxplot-1.png" target="_blank">e6e3268</a>
</td>
<td>
Dave Tang
</td>
<td>
2025-03-05
</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<div id="data-subset" class="section level2">
<h2>Data subset</h2>
<p>Since doublets are artifically generated using the available data,
providing a different input should affect the scoring and
prediction.</p>
<pre class="r"><code>set.seed(1984)

n &lt;- ceiling(ncol(pbmc1k)*0.9)
bcs &lt;- sample(x = colnames(pbmc1k), size = n)

pbmc1k_subset &lt;- pbmc1k[, bcs]
pbmc1k_subset</code></pre>
<pre><code>class: SingleCellExperiment 
dim: 33538 1100 
metadata(1): TENxFile
assays(1): counts
rownames(33538): ENSG00000243485 ENSG00000237613 ... ENSG00000277475
  ENSG00000268674
rowData names(3): ID Symbol Type
colnames(1100): AGGACGAAGACCTTTG-1 TTTATGCGTTGATCGT-1 ...
  GTCACTCTCTTGAACG-1 TTCAGGAGTCTACAAC-1
colData names(0):
reducedDimNames(0):
mainExpName: Gene Expression
altExpNames(0):</code></pre>
<p>Predict doublets on the data subset.</p>
<pre class="r"><code>pbmc1k_subset.pred &lt;- scDblFinder(pbmc1k_subset)</code></pre>
<pre><code>Creating ~1500 artificial doublets...</code></pre>
<pre><code>Dimensional reduction</code></pre>
<pre><code>Evaluating kNN...</code></pre>
<pre><code>Training model...</code></pre>
<pre><code>iter=0, 136 cells excluded from training.</code></pre>
<pre><code>iter=1, 120 cells excluded from training.</code></pre>
<pre><code>iter=2, 118 cells excluded from training.</code></pre>
<pre><code>Threshold found:0.46</code></pre>
<pre><code>33 (3%) doublets called</code></pre>
<pre class="r"><code>colData(pbmc1k_subset.pred) |&gt;
  as.data.frame() -&gt; pbmc1k_subset.pred

idx &lt;- match(
  row.names(pbmc1k_subset.pred),
  row.names(pbmc1k.pred)
)

x &lt;- pbmc1k.pred[idx, ]$scDblFinder.score
y &lt;- pbmc1k_subset.pred$scDblFinder.score

plot(x, y, pch = 16, xlab = &#39;Full dataset&#39;, ylab = &#39;Subset&#39;)</code></pre>
<p><img src="figure/scdblfinder.Rmd/pbmc1k_subset_pred-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-pbmc1k_subset_pred-1">
Past versions of pbmc1k_subset_pred-1.png
</button>
</p>
<div id="fig-pbmc1k_subset_pred-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/davetang/muse/blob/e6e32688e00a20a4eb999959b4aa89d124a49c47/docs/figure/scdblfinder.Rmd/pbmc1k_subset_pred-1.png" target="_blank">e6e3268</a>
</td>
<td>
Dave Tang
</td>
<td>
2025-03-05
</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<div id="running-in-batches" class="section level2">
<h2>Running in batches</h2>
<p>Create single <code>SingleCellExperiment</code> object.</p>
<pre class="r"><code>pbmc5k_1 &lt;- create_sce_obj(pbmc5k_1_h5)</code></pre>
<pre><code>Warning in rhdf5::h5read(file, name = paste0(group, ranges), index = list(1L),
: Object &#39;matrix/features/interval&#39; does not exist in this HDF5 file.</code></pre>
<pre class="r"><code>pbmc5k_2 &lt;- create_sce_obj(pbmc5k_2_h5)</code></pre>
<pre><code>Warning in rhdf5::h5read(file, name = paste0(group, ranges), index = list(1L),
: Object &#39;matrix/features/interval&#39; does not exist in this HDF5 file.</code></pre>
<pre class="r"><code>colnames(pbmc5k_1) &lt;- paste0(colnames(pbmc5k_1), &quot;_1&quot;)
colnames(pbmc5k_2) &lt;- paste0(colnames(pbmc5k_2), &quot;_2&quot;)

if(any(colnames(pbmc5k_1) %in% colnames(pbmc5k_2)) == TRUE){
  stop(&quot;Cell barcodes are not unique&quot;)
}

cbind(
  pbmc5k_1,
  pbmc5k_2
) -&gt; pbmc5k

coldata &lt;- DataFrame(
  donor = paste0(&#39;donor&#39;, sub(pattern = &quot;.*_&quot;, replacement = &quot;&quot;, x = colnames(pbmc5k)))
)
row.names(coldata) &lt;- colnames(pbmc5k)
colData(pbmc5k) &lt;- coldata

pbmc5k</code></pre>
<pre><code>class: SingleCellExperiment 
dim: 38606 11699 
metadata(2): TENxFile TENxFile
assays(1): counts
rownames(38606): ENSG00000290825 ENSG00000243485 ... ENSG00000278817
  ENSG00000277196
rowData names(3): ID Symbol Type
colnames(11699): AAACCAAAGGTGACGA-1_1 AAACCCTGTGACGAGT-1_1 ...
  TGTGTTAGTTAAGTGG-1_2 TGTGTTGAGCCAATGC-1_2
colData names(1): donor
reducedDimNames(0):
mainExpName: Gene Expression
altExpNames(0):</code></pre>
<p>Predict on combined object but on different samples.</p>
<blockquote>
<p>A vector of the same length as cells (or the name of a column of
colData(x)), indicating to which sample each cell belongs. Here, a
sample is understood as being processed independently. If omitted,
doublets will be searched for with all cells together. If given,
doublets will be searched for independently for each sample, which is
preferable if they represent different captures. If your samples were
multiplexed using cell hashes, what you want to give here are the
different batches/wells (i.e. independent captures, since doublets
cannot arise across them) rather than biological samples.</p>
</blockquote>
<pre class="r"><code>pbmc5k.pred &lt;- scDblFinder(pbmc5k, samples=&quot;donor&quot;)</code></pre>
<pre><code>  |                                                                              |                                                                      |   0%  |                                                                              |===================================                                   |  50%  |                                                                              |======================================================================| 100%</code></pre>
<p>Predict on single batch.</p>
<pre class="r"><code>pbmc5k_1.pred &lt;- scDblFinder(pbmc5k_1)</code></pre>
<pre><code>Creating ~4568 artificial doublets...</code></pre>
<pre><code>Dimensional reduction</code></pre>
<pre><code>Evaluating kNN...</code></pre>
<pre><code>Training model...</code></pre>
<pre><code>iter=0, 405 cells excluded from training.</code></pre>
<pre><code>iter=1, 397 cells excluded from training.</code></pre>
<pre><code>iter=2, 384 cells excluded from training.</code></pre>
<pre><code>Threshold found:0.381</code></pre>
<pre><code>223 (3.9%) doublets called</code></pre>
<p>Compare results.</p>
<pre class="r"><code>idx &lt;- match(colnames(pbmc5k_1.pred), colnames(pbmc5k.pred))

x &lt;- pbmc5k.pred[, idx]$scDblFinder.score
y &lt;- pbmc5k_1.pred$scDblFinder.score

plot(x, y, pch = 16, xlab = &quot;Combined but using batch&quot;, ylab = &quot;Independent batch&quot;)</code></pre>
<p><img src="figure/scdblfinder.Rmd/compare_batch_run-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>Numbers.</p>
<pre class="r"><code>table(pbmc5k.pred[, idx]$scDblFinder.class)</code></pre>
<pre><code>
singlet doublet 
   5453     257 </code></pre>
<pre class="r"><code>table(pbmc5k_1.pred$scDblFinder.class)</code></pre>
<pre><code>
singlet doublet 
   5487     223 </code></pre>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span>
Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 4.4.1 (2024-06-14)
Platform: x86_64-pc-linux-gnu
Running under: Ubuntu 22.04.5 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 
LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

time zone: Etc/UTC
tzcode source: system (glibc)

attached base packages:
[1] stats4    stats     graphics  grDevices utils     datasets  methods  
[8] base     

other attached packages:
 [1] TENxIO_1.8.2                scDblFinder_1.20.2         
 [3] SingleCellExperiment_1.28.1 SummarizedExperiment_1.36.0
 [5] Biobase_2.66.0              GenomicRanges_1.58.0       
 [7] GenomeInfoDb_1.42.3         IRanges_2.40.1             
 [9] S4Vectors_0.44.0            BiocGenerics_0.52.0        
[11] MatrixGenerics_1.18.1       matrixStats_1.4.1          
[13] workflowr_1.7.1            

loaded via a namespace (and not attached):
  [1] rstudioapi_0.17.1        jsonlite_1.8.9           magrittr_2.0.3          
  [4] ggbeeswarm_0.7.2         rmarkdown_2.28           fs_1.6.4                
  [7] BiocIO_1.16.0            zlibbioc_1.52.0          vctrs_0.6.5             
 [10] Rsamtools_2.22.0         RCurl_1.98-1.16          htmltools_0.5.8.1       
 [13] S4Arrays_1.6.0           BiocBaseUtils_1.8.0      curl_5.2.3              
 [16] BiocNeighbors_2.0.1      Rhdf5lib_1.28.0          xgboost_1.7.8.1         
 [19] SparseArray_1.6.1        rhdf5_2.50.2             sass_0.4.9              
 [22] bslib_0.8.0              cachem_1.1.0             GenomicAlignments_1.42.0
 [25] whisker_0.4.1            igraph_2.1.1             lifecycle_1.0.4         
 [28] pkgconfig_2.0.3          rsvd_1.0.5               Matrix_1.7-0            
 [31] R6_2.5.1                 fastmap_1.2.0            GenomeInfoDbData_1.2.13 
 [34] digest_0.6.37            colorspace_2.1-1         ps_1.8.1                
 [37] rprojroot_2.0.4          scater_1.34.0            dqrng_0.4.1             
 [40] irlba_2.3.5.1            beachmat_2.22.0          fansi_1.0.6             
 [43] httr_1.4.7               abind_1.4-8              compiler_4.4.1          
 [46] BiocParallel_1.40.0      viridis_0.6.5            highr_0.11              
 [49] HDF5Array_1.34.0         R.utils_2.12.3           MASS_7.3-60.2           
 [52] DelayedArray_0.32.0      rjson_0.2.23             bluster_1.16.0          
 [55] tools_4.4.1              vipor_0.4.7              beeswarm_0.4.0          
 [58] httpuv_1.6.15            R.oo_1.26.0              glue_1.8.0              
 [61] restfulr_0.0.15          callr_3.7.6              rhdf5filters_1.18.0     
 [64] promises_1.3.0           grid_4.4.1               getPass_0.2-4           
 [67] cluster_2.1.6            generics_0.1.3           gtable_0.3.6            
 [70] tzdb_0.4.0               R.methodsS3_1.8.2        data.table_1.16.2       
 [73] hms_1.1.3                BiocSingular_1.22.0      ScaledMatrix_1.14.0     
 [76] metapod_1.14.0           utf8_1.2.4               XVector_0.46.0          
 [79] ggrepel_0.9.6            pillar_1.9.0             stringr_1.5.1           
 [82] limma_3.62.2             later_1.3.2              dplyr_1.1.4             
 [85] lattice_0.22-6           rtracklayer_1.66.0       tidyselect_1.2.1        
 [88] locfit_1.5-9.10          Biostrings_2.74.1        scuttle_1.16.0          
 [91] knitr_1.48               git2r_0.35.0             gridExtra_2.3           
 [94] edgeR_4.4.2              xfun_0.48                statmod_1.5.0           
 [97] stringi_1.8.4            UCSC.utils_1.2.0         yaml_2.3.10             
[100] evaluate_1.0.1           codetools_0.2-20         tibble_3.2.1            
[103] cli_3.6.3                munsell_0.5.1            processx_3.8.4          
[106] jquerylib_0.1.4          Rcpp_1.0.13              XML_3.99-0.17           
[109] parallel_4.4.1           ggplot2_3.5.1            readr_2.1.5             
[112] scran_1.34.0             bitops_1.0-9             viridisLite_0.4.2       
[115] scales_1.3.0             crayon_1.5.3             rlang_1.1.4             </code></pre>
</div>
</div>
</div>
</div>
</div>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
https://docs.mathjax.org/en/latest/web/configuration.html. This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>




</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
