<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />



<meta name="date" content="2025-03-05" />

<title>The SingleCellExperiment class</title>

<script src="site_libs/header-attrs-2.28/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.13.2/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-6.4.2/css/all.min.css" rel="stylesheet" />
<link href="site_libs/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet" />

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/main/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">My blog</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/davetang/muse">
    <span class="fa fa-github"></span>
     
    Source code
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">The SingleCellExperiment class</h1>
<h4 class="date">2025-03-05</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span>
workflowr <span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2025-03-05
</p>
<p>
<strong>Checks:</strong> <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 7
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong> <code>muse/</code> <span
class="glyphicon glyphicon-question-sign" aria-hidden="true"
title="This is the local directory in which the code in this file was executed.">
</span>
</p>
<p>
This reproducible <a href="https://rmarkdown.rstudio.com">R Markdown</a>
analysis was created with <a
  href="https://github.com/workflowr/workflowr">workflowr</a> (version
1.7.1). The <em>Checks</em> tab describes the reproducibility checks
that were applied when the results were created. The <em>Past
versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date
</a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git
repository, you know the exact version of the code that produced these
results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the
global environment can affect the analysis in your R Markdown file in
unknown ways. For reproduciblity it’s best to always run the code in an
empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20200712code">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Seed:</strong>
<code>set.seed(20200712)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20200712code"
class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20200712)</code> was run prior to running
the code in the R Markdown file. Setting a seed ensures that any results
that rely on randomness, e.g. subsampling or permutations, are
reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Session information:</strong>
recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package
versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be
confident that you successfully produced the results during this
run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr
project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomdavetangmusetree78c20ac9fa712206a894a03d28e6c473986592d5targetblank78c20aca">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Repository version:</strong>
<a href="https://github.com/davetang/muse/tree/78c20ac9fa712206a894a03d28e6c473986592d5" target="_blank">78c20ac</a>
</a>
</p>
</div>
<div
id="strongRepositoryversionstrongahrefhttpsgithubcomdavetangmusetree78c20ac9fa712206a894a03d28e6c473986592d5targetblank78c20aca"
class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development
and connecting the code version to the results is critical for
reproducibility.
</p>
<p>
The results in this page were generated with repository version
<a href="https://github.com/davetang/muse/tree/78c20ac9fa712206a894a03d28e6c473986592d5" target="_blank">78c20ac</a>.
See the <em>Past versions</em> tab to see a history of the changes made
to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for
the analysis have been committed to Git prior to generating the results
(you can use <code>wflow_publish</code> or
<code>wflow_git_commit</code>). workflowr only checks the R Markdown
file, but you know if there are other scripts or data files that it
depends on. Below is the status of the Git repository when the results
were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .Rproj.user/
    Ignored:    data/1M_neurons_filtered_gene_bc_matrices_h5.h5
    Ignored:    data/293t/
    Ignored:    data/293t_3t3_filtered_gene_bc_matrices.tar.gz
    Ignored:    data/293t_filtered_gene_bc_matrices.tar.gz
    Ignored:    data/5k_Human_Donor1_PBMC_3p_gem-x_5k_Human_Donor1_PBMC_3p_gem-x_count_sample_filtered_feature_bc_matrix.h5
    Ignored:    data/5k_Human_Donor2_PBMC_3p_gem-x_5k_Human_Donor2_PBMC_3p_gem-x_count_sample_filtered_feature_bc_matrix.h5
    Ignored:    data/5k_Human_Donor3_PBMC_3p_gem-x_5k_Human_Donor3_PBMC_3p_gem-x_count_sample_filtered_feature_bc_matrix.h5
    Ignored:    data/5k_Human_Donor4_PBMC_3p_gem-x_5k_Human_Donor4_PBMC_3p_gem-x_count_sample_filtered_feature_bc_matrix.h5
    Ignored:    data/Parent_SC3v3_Human_Glioblastoma_filtered_feature_bc_matrix.tar.gz
    Ignored:    data/brain_counts/
    Ignored:    data/cl.obo
    Ignored:    data/cl.owl
    Ignored:    data/jurkat/
    Ignored:    data/jurkat:293t_50:50_filtered_gene_bc_matrices.tar.gz
    Ignored:    data/jurkat_293t/
    Ignored:    data/jurkat_filtered_gene_bc_matrices.tar.gz
    Ignored:    data/pbmc20k/
    Ignored:    data/pbmc20k_seurat/
    Ignored:    data/pbmc3k/
    Ignored:    data/pbmc4k_filtered_gene_bc_matrices.tar.gz
    Ignored:    data/pbmc_1k_v3_raw_feature_bc_matrix.h5
    Ignored:    data/refdata-gex-GRCh38-2020-A.tar.gz
    Ignored:    data/seurat_1m_neuron.rds
    Ignored:    data/t_3k_filtered_gene_bc_matrices.tar.gz
    Ignored:    r_packages_4.4.1/

Untracked files:
    Untracked:  analysis/bioc_scrnaseq.Rmd

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not
included in this status report because it is ok for generated content to
have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">
<p>
These are the previous versions of the repository in which changes were
made to the R Markdown (<code>analysis/sce.Rmd</code>) and HTML
(<code>docs/sce.html</code>) files. If you’ve configured a remote Git
repository (see <code>?wflow_git_remote</code>), click on the hyperlinks
in the table below to view the files as they were in that past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/davetang/muse/blob/78c20ac9fa712206a894a03d28e6c473986592d5/analysis/sce.Rmd" target="_blank">78c20ac</a>
</td>
<td>
Dave Tang
</td>
<td>
2025-03-05
</td>
<td>
The SingleCellExperiment class
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<p>From <a
href="https://bioconductor.org/books/3.13/OSCA.intro/the-singlecellexperiment-class.html">Introduction
to Single-Cell Analysis with Bioconductor</a>:</p>
<blockquote>
<p>… the SingleCellExperiment class (from the SingleCellExperiment
package) serves as the common currency for data exchange across 70+
single-cell-related Bioconductor packages. This class implements a data
structure that stores all aspects of our single-cell data - gene-by-cell
expression data, per-cell metadata and per-gene annotation (Figure 4.1)
- and manipulate them in a synchronized manner.</p>
</blockquote>
<div id="dependencies" class="section level2">
<h2>Dependencies</h2>
<p>Install Bioconductor packages using
<code>BiocManager::install()</code>.</p>
<pre class="r"><code>if (!require(&quot;BiocManager&quot;, quietly = TRUE))
  install.packages(&quot;BiocManager&quot;)

BiocManager::install(&quot;SingleCellExperiment&quot;)
BiocManager::install(&quot;TENxIO&quot;)
install.packages(&quot;hdf5r&quot;)
install.packages(&quot;Seurat&quot;)</code></pre>
<p>Load libraries.</p>
<pre class="r"><code>suppressPackageStartupMessages(library(SingleCellExperiment))
suppressPackageStartupMessages(library(TENxIO))
suppressPackageStartupMessages(library(hdf5r))
suppressPackageStartupMessages(library(Seurat))</code></pre>
</div>
<div id="data" class="section level2">
<h2>Data</h2>
<p>Download data.</p>
<pre class="r"><code>outdir &lt;- &#39;data/&#39;

pbmc5k_1_url &lt;- &quot;https://cf.10xgenomics.com/samples/cell-exp/9.0.0/5k_Human_Donor1_PBMC_3p_gem-x_5k_Human_Donor1_PBMC_3p_gem-x/5k_Human_Donor1_PBMC_3p_gem-x_5k_Human_Donor1_PBMC_3p_gem-x_count_sample_filtered_feature_bc_matrix.h5&quot;
pbmc5k_2_url &lt;- &quot;https://cf.10xgenomics.com/samples/cell-exp/9.0.0/5k_Human_Donor2_PBMC_3p_gem-x_5k_Human_Donor2_PBMC_3p_gem-x/5k_Human_Donor2_PBMC_3p_gem-x_5k_Human_Donor2_PBMC_3p_gem-x_count_sample_filtered_feature_bc_matrix.h5&quot;
pbmc5k_1_h5 &lt;- paste0(outdir, basename(pbmc5k_1_url))
pbmc5k_2_h5 &lt;- paste0(outdir, basename(pbmc5k_2_url))

download_file &lt;- function(url, outfile){
  fn &lt;- basename(url)
  if(file.exists(outfile) == FALSE){
    download.file(url, destfile = outfile)
  } else {
    message(paste0(outfile, &quot; already exists&quot;))
  }
}

download_file(pbmc5k_1_url, pbmc5k_1_h5)</code></pre>
<pre><code>data/5k_Human_Donor1_PBMC_3p_gem-x_5k_Human_Donor1_PBMC_3p_gem-x_count_sample_filtered_feature_bc_matrix.h5 already exists</code></pre>
<pre class="r"><code>download_file(pbmc5k_2_url, pbmc5k_2_h5)</code></pre>
<pre><code>data/5k_Human_Donor2_PBMC_3p_gem-x_5k_Human_Donor2_PBMC_3p_gem-x_count_sample_filtered_feature_bc_matrix.h5 already exists</code></pre>
</div>
<div id="create-objects" class="section level2">
<h2>Create objects</h2>
<p>Create <code>SingleCellExperiment</code> files.</p>
<pre class="r"><code>create_sce_obj &lt;- function(h5){
  con &lt;- TENxH5(h5)
  import(con)
}

pbmc5k_1 &lt;- create_sce_obj(pbmc5k_1_h5)</code></pre>
<pre><code>Warning in rhdf5::h5read(file, name = paste0(group, ranges), index = list(1L),
: Object &#39;matrix/features/interval&#39; does not exist in this HDF5 file.</code></pre>
<pre class="r"><code>pbmc5k_1</code></pre>
<pre><code>class: SingleCellExperiment 
dim: 38606 5710 
metadata(1): TENxFile
assays(1): counts
rownames(38606): ENSG00000290825 ENSG00000243485 ... ENSG00000278817
  ENSG00000277196
rowData names(3): ID Symbol Type
colnames(5710): AAACCAAAGGTGACGA-1 AAACCCTGTGACGAGT-1 ...
  TGTGTTAGTCAATCGT-1 TGTGTTGAGGATCTCA-1
colData names(0):
reducedDimNames(0):
mainExpName: Gene Expression
altExpNames(0):</code></pre>
<p>Seurat object.</p>
<pre class="r"><code>min_cells &lt;- 10
min_features &lt;- 500

pbmc5k_1_seurat &lt;- Seurat::CreateSeuratObject(
  counts = Seurat::Read10X_h5(pbmc5k_1_h5),
  min.cells = min_cells,
  min.features = min_features
)
pbmc5k_1_seurat</code></pre>
<pre><code>An object of class Seurat 
20935 features across 5672 samples within 1 assay 
Active assay: RNA (20935 features, 0 variable features)
 1 layer present: counts</code></pre>
<p>Filter <code>SingleCellExperiment</code> object in the same way; the
order and manner of filtering is important or else you can not reproduce
Seurat’s behaviour.</p>
<pre class="r"><code>filter_sce &lt;- function(sce, min.features, min.cells){
  detected_genes_in_bcs &lt;- colSums(counts(sce) &gt; 0)
  sce &lt;- sce[, detected_genes_in_bcs &gt;= min.features]
  detected_genes &lt;- rowSums(counts(sce) &gt; 0)
  sce &lt;- sce[detected_genes &gt;= min.cells, ]
  sce
}

pbmc5k_1_filtered &lt;- filter_sce(pbmc5k_1, min_features, min_cells)
pbmc5k_1_filtered</code></pre>
<pre><code>class: SingleCellExperiment 
dim: 20935 5672 
metadata(1): TENxFile
assays(1): counts
rownames(20935): ENSG00000238009 ENSG00000241860 ... ENSG00000271254
  ENSG00000278817
rowData names(3): ID Symbol Type
colnames(5672): AAACCAAAGGTGACGA-1 AAACCCTGTGACGAGT-1 ...
  TGTGTTAGTCAATCGT-1 TGTGTTGAGGATCTCA-1
colData names(0):
reducedDimNames(0):
mainExpName: Gene Expression
altExpNames(0):</code></pre>
<p>Compare barcodes.</p>
<pre class="r"><code>all(colnames(pbmc5k_1_seurat) == colnames(pbmc5k_1_filtered))</code></pre>
<pre><code>[1] TRUE</code></pre>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span>
Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 4.4.1 (2024-06-14)
Platform: x86_64-pc-linux-gnu
Running under: Ubuntu 22.04.5 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 
LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

time zone: Etc/UTC
tzcode source: system (glibc)

attached base packages:
[1] stats4    stats     graphics  grDevices utils     datasets  methods  
[8] base     

other attached packages:
 [1] Seurat_5.1.0                SeuratObject_5.0.2         
 [3] sp_2.1-4                    hdf5r_1.3.11               
 [5] TENxIO_1.8.2                SingleCellExperiment_1.28.1
 [7] SummarizedExperiment_1.36.0 Biobase_2.66.0             
 [9] GenomicRanges_1.58.0        GenomeInfoDb_1.42.3        
[11] IRanges_2.40.1              S4Vectors_0.44.0           
[13] BiocGenerics_0.52.0         MatrixGenerics_1.18.1      
[15] matrixStats_1.4.1           workflowr_1.7.1            

loaded via a namespace (and not attached):
  [1] RcppAnnoy_0.0.22        splines_4.4.1           later_1.3.2            
  [4] BiocIO_1.16.0           tibble_3.2.1            R.oo_1.26.0            
  [7] polyclip_1.10-7         fastDummies_1.7.4       lifecycle_1.0.4        
 [10] rprojroot_2.0.4         globals_0.16.3          processx_3.8.4         
 [13] lattice_0.22-6          MASS_7.3-60.2           magrittr_2.0.3         
 [16] plotly_4.10.4           sass_0.4.9              rmarkdown_2.28         
 [19] jquerylib_0.1.4         yaml_2.3.10             httpuv_1.6.15          
 [22] sctransform_0.4.1       spam_2.11-0             spatstat.sparse_3.1-0  
 [25] reticulate_1.39.0       cowplot_1.1.3           pbapply_1.7-2          
 [28] RColorBrewer_1.1-3      abind_1.4-8             zlibbioc_1.52.0        
 [31] Rtsne_0.17              purrr_1.0.2             R.utils_2.12.3         
 [34] git2r_0.35.0            GenomeInfoDbData_1.2.13 ggrepel_0.9.6          
 [37] irlba_2.3.5.1           listenv_0.9.1           spatstat.utils_3.1-0   
 [40] goftest_1.2-3           RSpectra_0.16-2         spatstat.random_3.3-2  
 [43] fitdistrplus_1.2-1      parallelly_1.38.0       leiden_0.4.3.1         
 [46] codetools_0.2-20        DelayedArray_0.32.0     tidyselect_1.2.1       
 [49] UCSC.utils_1.2.0        farver_2.1.2            spatstat.explore_3.3-3 
 [52] jsonlite_1.8.9          progressr_0.15.0        ggridges_0.5.6         
 [55] survival_3.6-4          tools_4.4.1             ica_1.0-3              
 [58] Rcpp_1.0.13             glue_1.8.0              gridExtra_2.3          
 [61] SparseArray_1.6.1       BiocBaseUtils_1.8.0     xfun_0.48              
 [64] HDF5Array_1.34.0        dplyr_1.1.4             fastmap_1.2.0          
 [67] rhdf5filters_1.18.0     fansi_1.0.6             callr_3.7.6            
 [70] digest_0.6.37           R6_2.5.1                mime_0.12              
 [73] colorspace_2.1-1        scattermore_1.2         tensor_1.5             
 [76] spatstat.data_3.1-2     R.methodsS3_1.8.2       utf8_1.2.4             
 [79] tidyr_1.3.1             generics_0.1.3          data.table_1.16.2      
 [82] httr_1.4.7              htmlwidgets_1.6.4       S4Arrays_1.6.0         
 [85] whisker_0.4.1           uwot_0.2.2              pkgconfig_2.0.3        
 [88] gtable_0.3.6            lmtest_0.9-40           XVector_0.46.0         
 [91] htmltools_0.5.8.1       dotCall64_1.2           scales_1.3.0           
 [94] png_0.1-8               spatstat.univar_3.0-1   knitr_1.48             
 [97] rstudioapi_0.17.1       tzdb_0.4.0              reshape2_1.4.4         
[100] nlme_3.1-164            cachem_1.1.0            zoo_1.8-12             
[103] rhdf5_2.50.2            stringr_1.5.1           KernSmooth_2.23-24     
[106] parallel_4.4.1          miniUI_0.1.1.1          pillar_1.9.0           
[109] grid_4.4.1              vctrs_0.6.5             RANN_2.6.2             
[112] promises_1.3.0          xtable_1.8-4            cluster_2.1.6          
[115] evaluate_1.0.1          readr_2.1.5             cli_3.6.3              
[118] compiler_4.4.1          rlang_1.1.4             crayon_1.5.3           
[121] future.apply_1.11.3     ps_1.8.1                getPass_0.2-4          
[124] plyr_1.8.9              fs_1.6.4                stringi_1.8.4          
[127] viridisLite_0.4.2       deldir_2.0-4            munsell_0.5.1          
[130] lazyeval_0.2.2          spatstat.geom_3.3-3     Matrix_1.7-0           
[133] RcppHNSW_0.6.0          hms_1.1.3               patchwork_1.3.0        
[136] bit64_4.5.2             future_1.34.0           Rhdf5lib_1.28.0        
[139] ggplot2_3.5.1           shiny_1.9.1             ROCR_1.0-11            
[142] igraph_2.1.1            bslib_0.8.0             bit_4.5.0              </code></pre>
</div>
</div>
</div>
</div>
</div>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
https://docs.mathjax.org/en/latest/web/configuration.html. This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>




</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
