<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />



<meta name="date" content="2025-03-11" />

<title>Seurat v4 versus v5</title>

<script src="site_libs/header-attrs-2.29/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.13.2/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-6.5.2/css/all.min.css" rel="stylesheet" />
<link href="site_libs/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet" />

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/main/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">My blog</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/davetang/muse">
    <span class="fa fa-github"></span>
     
    Source code
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Seurat v4 versus v5</h1>
<h4 class="date">2025-03-11</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span>
workflowr <span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2025-03-11
</p>
<p>
<strong>Checks:</strong> <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 7
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong> <code>muse/</code> <span
class="glyphicon glyphicon-question-sign" aria-hidden="true"
title="This is the local directory in which the code in this file was executed.">
</span>
</p>
<p>
This reproducible <a href="https://rmarkdown.rstudio.com">R Markdown</a>
analysis was created with <a
  href="https://github.com/workflowr/workflowr">workflowr</a> (version
1.7.1). The <em>Checks</em> tab describes the reproducibility checks
that were applied when the results were created. The <em>Past
versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date
</a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git
repository, you know the exact version of the code that produced these
results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the
global environment can affect the analysis in your R Markdown file in
unknown ways. For reproduciblity it’s best to always run the code in an
empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20200712code">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Seed:</strong>
<code>set.seed(20200712)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20200712code"
class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20200712)</code> was run prior to running
the code in the R Markdown file. Setting a seed ensures that any results
that rely on randomness, e.g. subsampling or permutations, are
reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Session information:</strong>
recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package
versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be
confident that you successfully produced the results during this
run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr
project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomdavetangmusetree4688c4091f3496e67130a46ca05149d12336efd1targetblank4688c40a">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Repository version:</strong>
<a href="https://github.com/davetang/muse/tree/4688c4091f3496e67130a46ca05149d12336efd1" target="_blank">4688c40</a>
</a>
</p>
</div>
<div
id="strongRepositoryversionstrongahrefhttpsgithubcomdavetangmusetree4688c4091f3496e67130a46ca05149d12336efd1targetblank4688c40a"
class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development
and connecting the code version to the results is critical for
reproducibility.
</p>
<p>
The results in this page were generated with repository version
<a href="https://github.com/davetang/muse/tree/4688c4091f3496e67130a46ca05149d12336efd1" target="_blank">4688c40</a>.
See the <em>Past versions</em> tab to see a history of the changes made
to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for
the analysis have been committed to Git prior to generating the results
(you can use <code>wflow_publish</code> or
<code>wflow_git_commit</code>). workflowr only checks the R Markdown
file, but you know if there are other scripts or data files that it
depends on. Below is the status of the Git repository when the results
were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .Rproj.user/
    Ignored:    data/1M_neurons_filtered_gene_bc_matrices_h5.h5
    Ignored:    data/293t/
    Ignored:    data/293t_3t3_filtered_gene_bc_matrices.tar.gz
    Ignored:    data/293t_filtered_gene_bc_matrices.tar.gz
    Ignored:    data/5k_Human_Donor1_PBMC_3p_gem-x_5k_Human_Donor1_PBMC_3p_gem-x_count_sample_filtered_feature_bc_matrix.h5
    Ignored:    data/5k_Human_Donor2_PBMC_3p_gem-x_5k_Human_Donor2_PBMC_3p_gem-x_count_sample_filtered_feature_bc_matrix.h5
    Ignored:    data/5k_Human_Donor3_PBMC_3p_gem-x_5k_Human_Donor3_PBMC_3p_gem-x_count_sample_filtered_feature_bc_matrix.h5
    Ignored:    data/5k_Human_Donor4_PBMC_3p_gem-x_5k_Human_Donor4_PBMC_3p_gem-x_count_sample_filtered_feature_bc_matrix.h5
    Ignored:    data/97516b79-8d08-46a6-b329-5d0a25b0be98.h5ad
    Ignored:    data/Parent_SC3v3_Human_Glioblastoma_filtered_feature_bc_matrix.tar.gz
    Ignored:    data/brain_counts/
    Ignored:    data/cl.obo
    Ignored:    data/cl.owl
    Ignored:    data/jurkat/
    Ignored:    data/jurkat:293t_50:50_filtered_gene_bc_matrices.tar.gz
    Ignored:    data/jurkat_293t/
    Ignored:    data/jurkat_filtered_gene_bc_matrices.tar.gz
    Ignored:    data/pbmc20k/
    Ignored:    data/pbmc20k_seurat/
    Ignored:    data/pbmc3k/
    Ignored:    data/pbmc3k_seurat.rds
    Ignored:    data/pbmc4k_filtered_gene_bc_matrices.tar.gz
    Ignored:    data/pbmc_1k_v3_filtered_feature_bc_matrix.h5
    Ignored:    data/pbmc_1k_v3_raw_feature_bc_matrix.h5
    Ignored:    data/refdata-gex-GRCh38-2020-A.tar.gz
    Ignored:    data/seurat_1m_neuron.rds
    Ignored:    data/t_3k_filtered_gene_bc_matrices.tar.gz
    Ignored:    r_packages_4.4.1/

Untracked files:
    Untracked:  analysis/bioc_scrnaseq.Rmd

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not
included in this status report because it is ok for generated content to
have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">
<p>
These are the previous versions of the repository in which changes were
made to the R Markdown (<code>analysis/seurat_v4_vs_v5.Rmd</code>) and
HTML (<code>docs/seurat_v4_vs_v5.html</code>) files. If you’ve
configured a remote Git repository (see <code>?wflow_git_remote</code>),
click on the hyperlinks in the table below to view the files as they
were in that past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/davetang/muse/blob/4688c4091f3496e67130a46ca05149d12336efd1/analysis/seurat_v4_vs_v5.Rmd" target="_blank">4688c40</a>
</td>
<td>
Dave Tang
</td>
<td>
2025-03-11
</td>
<td>
Overdispersion exists in all scRNA-seq datasets if sufficiently
sequenced
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/davetang/muse/6248d96391af91d09a54ae7766378710f1670fe5/docs/seurat_v4_vs_v5.html" target="_blank">6248d96</a>
</td>
<td>
Dave Tang
</td>
<td>
2025-03-11
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/davetang/muse/blob/d47858c27395b410916e68f1c4fee611c9ba13da/analysis/seurat_v4_vs_v5.Rmd" target="_blank">d47858c</a>
</td>
<td>
Dave Tang
</td>
<td>
2025-03-11
</td>
<td>
Pearson residuals
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/davetang/muse/747fe3f384dde4c6a24a3051b69e5f6590e7af62/docs/seurat_v4_vs_v5.html" target="_blank">747fe3f</a>
</td>
<td>
Dave Tang
</td>
<td>
2025-03-11
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/davetang/muse/blob/4aacb58f49523299d8a1340ee80af0c058916da9/analysis/seurat_v4_vs_v5.Rmd" target="_blank">4aacb58</a>
</td>
<td>
Dave Tang
</td>
<td>
2025-03-11
</td>
<td>
Error versus variance
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/davetang/muse/fc7e28db6d2b4db4731c8e91af1d3ddd5b551524/docs/seurat_v4_vs_v5.html" target="_blank">fc7e28d</a>
</td>
<td>
Dave Tang
</td>
<td>
2025-03-11
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/davetang/muse/blob/613993535f63f46ec1a68472f976de64d5d14888/analysis/seurat_v4_vs_v5.Rmd" target="_blank">6139935</a>
</td>
<td>
Dave Tang
</td>
<td>
2025-03-11
</td>
<td>
Introduction to modelling variance
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/davetang/muse/94116b19e07cb243d9a27be3d7522ecbcc4d6e94/docs/seurat_v4_vs_v5.html" target="_blank">94116b1</a>
</td>
<td>
Dave Tang
</td>
<td>
2025-03-10
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/davetang/muse/blob/4c187b3ffae504a683192c5eb899efeb338c09f8/analysis/seurat_v4_vs_v5.Rmd" target="_blank">4c187b3</a>
</td>
<td>
Dave Tang
</td>
<td>
2025-03-10
</td>
<td>
Data layer
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/davetang/muse/0d51a6921cae77a7fd87026f177935bd63bb4f12/docs/seurat_v4_vs_v5.html" target="_blank">0d51a69</a>
</td>
<td>
Dave Tang
</td>
<td>
2025-03-10
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/davetang/muse/blob/48661b3c394279d8ad104456c084c1cbda9e1892/analysis/seurat_v4_vs_v5.Rmd" target="_blank">48661b3</a>
</td>
<td>
Dave Tang
</td>
<td>
2025-03-10
</td>
<td>
Seurat version 4 vs. 5
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<div id="sctransform" class="section level2">
<h2>sctransform</h2>
<p>The paper <a
href="https://link.springer.com/article/10.1186/s13059-021-02584-9">Comparison
and evaluation of statistical error models for scRNA-seq</a> is the
basis for the default approach used in Seurat version 5. The following
is text from the paper:</p>
<ul>
<li>Heterogeneity in single-cell RNA-seq (scRNA-seq) data is driven by
multiple sources, including biological variation in cellular state as
well as technical variation introduced during experimental processing.
Deconvolving these effects is a key challenge for preprocessing
workflows.
<ul>
<li>Separating biological heterogeneity across cells that corresponds to
differences in cell type and state from alternative sources of variation
represents a key analytical challenge in the normalization and
preprocessing of single-cell RNA-seq data.</li>
</ul></li>
<li>Data normalization aims to adjust for differences in cellular
sequencing depth, which collectively arise from fluctuations in cellular
RNA content, efficiency in lysis and reverse transcription, and
stochastic sampling during next-generation sequencing.</li>
<li><strong>Variance stabilization aims to address the confounding
relationship between gene abundance and gene variance, and to ensure
that both lowly and highly expressed genes can contribute to the
downstream definition of cellular state</strong>.</li>
</ul>
<p>Using statistical models like Generalised Linear Models:</p>
<ul>
<li>Two recent studies proposed to use generalized linear models (GLMs),
where cellular sequencing depth was included as a covariate, as part of
scRNA-seq preprocessing workflows.</li>
<li>The sctransform approach utilizes the <strong>Pearson residuals from
negative binomial regression as input to standard dimensional reduction
techniques</strong>, while GLM-PCA focuses on a generalized version of
principal component analysis (PCA) for data with Poisson-distributed
errors.</li>
<li>More broadly, multiple techniques aim to learn a latent state that
captures biologically relevant cellular heterogeneity using either
matrix factorization or neural networks, alongside a defined error model
that describes the variation that is not captured by the latent
space.</li>
</ul>
<p>If a regression model doesn’t fully explain variability, the
residuals might contain structure that another technique can capture to
uncover hidden patterns. For example, if a regression model captures
main trends, applying Principal Component Analysis (PCA) on residuals
can find underlying structures in the unexplained variance. Another use
case can be clustering on residuals to group data points based on
deviations from a model.</p>
<p>Parameterising statistical models:</p>
<ul>
<li>Likelihood-based approaches require an explicit definition of a
statistical error model for scRNA-seq, and there is little consensus on
how to define or parameterize this model.</li>
<li>Multiple groups have utilized a Poisson error model but others argue
that the data exhibit evidence of overdispersion, requiring the use of a
negative-binomial (NB) distribution.</li>
<li>Methods that assume a NB distribution have different methods to
parameterize their model.
<ul>
<li>A recent study argued that fixing the NB inverse overdispersion
parameter <span class="math inline">\(\theta\)</span> to a single value
is an appropriate estimate of technical overdispersion for all genes in
all scRNA-seq datasets, while others propose learning unique parameter
values for each gene in each dataset.</li>
</ul></li>
<li>This lack of consensus is further exemplified by the scvi-tools
suite, which supports nine different methods for parameterizing error
models.</li>
<li>The purpose of this error model is to describe and quantify
heterogeneity that is not captured by biologically relevant differences
in cell state, and highlights a specific question: <strong>How can we
model the observed variation in gene expression for an scRNA-seq
experiment conducted on a biologically “homogeneous”
population?</strong></li>
</ul>
</div>
<div id="error-and-variance" class="section level2">
<h2>Error and variance</h2>
<ul>
<li>Error modeling refers to capturing and understanding the
<strong>uncertainty, randomness, and deviations</strong> in data or
predictions. Error is not exactly synonymous with variance but they are
related.</li>
<li>Errors arise from:
<ul>
<li>Randomness: unavoidable variability in data, e.g., measurement noise
and natural fluctuations</li>
<li>Model Imperfections: due to missing information or incorrect model
assumptions.</li>
</ul></li>
<li>Statistical models, like regression, often include an error term
<span class="math inline">\(\epsilon\)</span> to account for these
uncertainties:</li>
</ul>
<p><span class="math display">\[
Y = f(X) + \epsilon
\]</span></p>
<p>where <span class="math inline">\(\epsilon\)</span> captures random
fluctuations or unknown influences.</p>
<p>While error contributes to variance, they are distinct:</p>
<ul>
<li><strong>Error</strong>: The deviation of an observation from the
model’s predicted value.</li>
<li><strong>Variance</strong>: A statistical measure of how much values
deviate from their mean (spread of data).</li>
</ul>
<p>Errors can be random (causing variability) or systematic (bias), but
variance is a quantification of dispersion.</p>
<ul>
<li>In regression, we separate variance into <strong>explained
variance</strong> (by the model) and <strong>unexplained
variance</strong> (error).</li>
<li>Error variance <span class="math inline">\(\sigma^2\)</span> is
modeled to improve predictions and uncertainty estimation.</li>
<li>For a negative binomial regression model, the <strong>Pearson
residual</strong> <span class="math inline">\(r_i\)</span> for
observation <span class="math inline">\(i\)</span> is given by:</li>
</ul>
<p><span class="math display">\[
r_i = \frac{y_i - \hat{y}_i}{\sqrt{\text{Var}(y_i)}}
\]</span></p>
<p>where:<br />
* <span class="math inline">\(y_i\)</span> = observed count<br />
* <span class="math inline">\(\hat{y}_i\)</span> = predicted mean
(expected value under the model)<br />
* <span class="math inline">\(\text{Var}(y_i)\)</span> = model-estimated
variance of <span class="math inline">\(y_i\)</span></p>
<ul>
<li>Unlike Poisson regression (where <span
class="math inline">\(\text{Var}(y) = \hat{y}\)</span>), the
<strong>negative binomial</strong> model accounts for
<strong>overdispersion</strong> using a dispersion parameter <span
class="math inline">\(\theta\)</span>, and the variance is:</li>
</ul>
<p><span class="math display">\[
\text{Var}(y_i) = \hat{y}_i + \frac{\hat{y}_i^2}{\theta}
\]</span></p>
<p>This means the variance <strong>grows faster than the mean</strong>,
making negative binomial regression suitable when count data has extra
variability.</p>
<ul>
<li>Pearson residuals can be used for:
<ul>
<li><strong>Standardized measure</strong> - Large residuals (<span
class="math inline">\(|r_i| &gt; 2\)</span>) may indicate poor model
fit.</li>
<li><strong>Overdispersion check</strong> - If Pearson residuals
systematically increase with predicted values, it suggests the model may
not fully account for overdispersion.</li>
<li><strong>Model diagnostics</strong> - Residual plots help assess
whether assumptions (e.g., correct functional form) hold.</li>
</ul></li>
</ul>
<p>Pearson residuals focus on <strong>variance-adjusted
differences</strong> and deviance residuals come from likelihood-based
goodness-of-fit measures. They both help diagnose model fit, but
deviance residuals tend to emphasise extreme deviations more. Pearson
residuals in negative binomial regression are useful for model
diagnostics, particularly for checking overdispersion and assessing
fit.</p>
</div>
<div id="seurat-object" class="section level2">
<h2>Seurat object</h2>
<p>Import <a href="pbmc3k.html">raw pbmc3k dataset</a> from my
server.</p>
<pre class="r"><code>seurat_obj &lt;- readRDS(url(&quot;https://davetang.org/file/pbmc3k_seurat.rds&quot;, &quot;rb&quot;))
seurat_obj</code></pre>
<pre><code>An object of class Seurat 
32738 features across 2700 samples within 1 assay 
Active assay: RNA (32738 features, 0 variable features)
 1 layer present: counts</code></pre>
<p>Filter.</p>
<pre class="r"><code>pbmc3k &lt;- CreateSeuratObject(
  counts = seurat_obj@assays$RNA$counts,
  min.cells = 3,
  min.features = 200,
  project = &quot;pbmc3k&quot;
)
pbmc3k</code></pre>
<pre><code>An object of class Seurat 
13714 features across 2700 samples within 1 assay 
Active assay: RNA (13714 features, 0 variable features)
 1 layer present: counts</code></pre>
</div>
<div id="seurat-workflows" class="section level2">
<h2>Seurat workflows</h2>
<p>Process with the Seurat 4 workflow.</p>
<pre class="r"><code>seurat_wf_v4 &lt;- function(seurat_obj, scale_factor = 1e4, num_features = 2000, num_pcs = 30, cluster_res = 0.5, debug_flag = FALSE){
  
  seurat_obj &lt;- NormalizeData(seurat_obj, normalization.method = &quot;LogNormalize&quot;, scale.factor = scale_factor, verbose = debug_flag)
  seurat_obj &lt;- FindVariableFeatures(seurat_obj, selection.method = &#39;vst&#39;, nfeatures = num_features, verbose = debug_flag)
  seurat_obj &lt;- ScaleData(seurat_obj, verbose = debug_flag)
  seurat_obj &lt;- RunPCA(seurat_obj, verbose = debug_flag)
  seurat_obj &lt;- RunUMAP(seurat_obj, dims = 1:num_pcs, verbose = debug_flag)
  seurat_obj &lt;- FindNeighbors(seurat_obj, dims = 1:num_pcs, verbose = debug_flag)
  seurat_obj &lt;- FindClusters(seurat_obj, resolution = cluster_res, verbose = debug_flag)
  
  seurat_obj
}

pbmc3k_v4 &lt;- seurat_wf_v4(pbmc3k)</code></pre>
<pre><code>Warning: The default method for RunUMAP has changed from calling Python UMAP via reticulate to the R-native UWOT using the cosine metric
To use Python UMAP via reticulate, set umap.method to &#39;umap-learn&#39; and metric to &#39;correlation&#39;
This message will be shown once per session</code></pre>
<pre class="r"><code>pbmc3k_v4</code></pre>
<pre><code>An object of class Seurat 
13714 features across 2700 samples within 1 assay 
Active assay: RNA (13714 features, 2000 variable features)
 3 layers present: counts, data, scale.data
 2 dimensional reductions calculated: pca, umap</code></pre>
<p>UMAP.</p>
<pre class="r"><code>DimPlot(pbmc3k_v4, reduction = &quot;umap&quot;)</code></pre>
<p><img src="figure/seurat_v4_vs_v5.Rmd/umap_v4-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-umap_v4-1">
Past versions of umap_v4-1.png
</button>
</p>
<div id="fig-umap_v4-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/davetang/muse/blob/0d51a6921cae77a7fd87026f177935bd63bb4f12/docs/figure/seurat_v4_vs_v5.Rmd/umap_v4-1.png" target="_blank">0d51a69</a>
</td>
<td>
Dave Tang
</td>
<td>
2025-03-10
</td>
</tr>
</tbody>
</table>
</div>
</div>
<pre class="r"><code>seurat_wf_v5 &lt;- function(seurat_obj, scale_factor = 1e4, num_features = 2000, num_pcs = 30, cluster_res = 0.5, debug_flag = FALSE){
  
  seurat_obj &lt;- SCTransform(seurat_obj, verbose = debug_flag)
  seurat_obj &lt;- RunPCA(seurat_obj, verbose = debug_flag)
  seurat_obj &lt;- RunUMAP(seurat_obj, dims = 1:num_pcs, verbose = debug_flag)
  seurat_obj &lt;- FindNeighbors(seurat_obj, dims = 1:num_pcs, verbose = debug_flag)
  seurat_obj &lt;- FindClusters(seurat_obj, resolution = cluster_res, verbose = debug_flag)
  
  seurat_obj
}

pbmc3k_v5 &lt;- seurat_wf_v5(pbmc3k)
pbmc3k_v5</code></pre>
<pre><code>An object of class Seurat 
26286 features across 2700 samples within 2 assays 
Active assay: SCT (12572 features, 3000 variable features)
 3 layers present: counts, data, scale.data
 1 other assay present: RNA
 2 dimensional reductions calculated: pca, umap</code></pre>
<p>UMAP.</p>
<pre class="r"><code>DimPlot(pbmc3k_v5, reduction = &quot;umap&quot;)</code></pre>
<p><img src="figure/seurat_v4_vs_v5.Rmd/umap_v5-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-umap_v5-1">
Past versions of umap_v5-1.png
</button>
</p>
<div id="fig-umap_v5-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/davetang/muse/blob/0d51a6921cae77a7fd87026f177935bd63bb4f12/docs/figure/seurat_v4_vs_v5.Rmd/umap_v5-1.png" target="_blank">0d51a69</a>
</td>
<td>
Dave Tang
</td>
<td>
2025-03-10
</td>
</tr>
</tbody>
</table>
</div>
</div>
<div id="data-layer" class="section level3">
<h3>Data layer</h3>
<p>Version 4 store log normalised data.</p>
<pre class="r"><code>colSums(pbmc3k_v4@assays$RNA$data)[1:6]</code></pre>
<pre><code>AAACATACAACCAC-1 AAACATTGAGCTAC-1 AAACATTGATCAGC-1 AAACCGTGCTTCCG-1 
        1605.823         2027.859         2040.169         1902.960 
AAACCGTGTATGCG-1 AAACGCACTGGTAC-1 
        1388.125         1653.061 </code></pre>
<p>The data layer is in the SCT assay.</p>
<pre class="r"><code>colSums(pbmc3k_v5@assays$SCT$data)[1:6]</code></pre>
<pre><code>AAACATACAACCAC-1 AAACATTGAGCTAC-1 AAACATTGATCAGC-1 AAACCGTGCTTCCG-1 
        786.2686        1024.4731        1029.3032         934.4454 
AAACCGTGTATGCG-1 AAACGCACTGGTAC-1 
        666.1142         764.8101 </code></pre>
</div>
<div id="compare-clustering" class="section level3">
<h3>Compare clustering</h3>
<p>More granular clustering of version 4’s cluster 0 in version 5.</p>
<pre class="r"><code>stopifnot(all(row.names(pbmc3k_v4@meta.data) == row.names(pbmc3k_v5@meta.data)))

table(
  pbmc3k_v4@meta.data$seurat_clusters,
  pbmc3k_v5@meta.data$seurat_clusters
)</code></pre>
<pre><code>   
      0   1   2   3   4   5   6   7   8   9
  0 970   0  71   2   0   0 100  44   0   0
  1   0 479   0   0   0   9   0   0   3   0
  2   1   0   0 349   0   0   0   1   0   0
  3   4   0 290   1   5   0   0   1   0   0
  4   0   0   5   6 152   0   0   0   0   0
  5   0  16   0   0   0 145   0   0   0   0
  6   0   1   0   0   0   0   0   0  31   0
  7   0   1   0   1   0   0   0   0   0  12</code></pre>
</div>
</div>
<div id="appendix" class="section level2">
<h2>Appendix</h2>
<p>More information text from the paper <a
href="https://link.springer.com/article/10.1186/s13059-021-02584-9">Comparison
and evaluation of statistical error models for scRNA-seq</a>.</p>
<div id="shallow-sequencing-masks-overdispersion-in-scrna-seq-data"
class="section level3">
<h3>Shallow sequencing masks overdispersion in scRNA-seq data</h3>
<ul>
<li>The rationale behind a Poisson model assumes that homogeneous cells
express mRNA molecules for a given gene at a <strong>fixed underlying
rate</strong>, and the variation in scRNA-seq results specifically from
a <strong>stochastic sampling</strong> of mRNA molecules, for example
due to inefficiencies in reverse transcription and PCR, combined with
incomplete molecular sampling during DNA sequencing</li>
<li>While the Poisson distribution is well suited to capture variation
driven by stochastic technical loss and sampling noise, it cannot
capture other sources of biological heterogeneity between cells that are
not driven by changes in cell state, for example, intrinsic variation
caused by stochastic transcriptional bursts.
<ul>
<li>These fluctuations would cause scRNA-seq data to deviate from
Poisson statistics, exhibiting overdispersion.</li>
</ul></li>
<li>To assess whether scRNA-seq datasets follow the Poisson
distribution, the authors performed a goodness-of-fit test,
independently modeling the observed counts for each gene to be Poisson
distributed, while accounting for differences in sequencing depth
between individual cells.</li>
<li>In each of the 59 datasets analysed, genes exhibiting Poisson
variation were overwhelmingly lowly expressed compared to genes that
were overdispersed.
<ul>
<li>Moreover, when comparing results for cell-line datasets where we
expect low levels of variation in cell state, we found that the global
fraction of genes deviating from a Poisson distribution was correlated
with the average sequencing depth of the dataset.</li>
</ul></li>
<li>The results suggest that scRNA-seq datasets commonly exhibit
biological variation that exceeds Poisson sampling, but that the
statistical power to detect these fluctuations requires sufficient
sequencing depth.</li>
<li>After downsampling, only 0.5% genes failed the Poisson
goodness-of-fit test, demonstrating that reducing cellular sequencing
depth can artificially create the appearance of Poisson variation.</li>
<li><strong>The authors conclude that the Poisson error model may
represent an acceptable approximation for scRNA-seq datasets with
shallow sequencing, but as the sensitivity of molecular profiling
continues to increase, error models that allow for overdispersion are
required for scRNA-seq analysis</strong>.</li>
</ul>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span>
Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 4.4.1 (2024-06-14)
Platform: x86_64-pc-linux-gnu
Running under: Ubuntu 22.04.5 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 
LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

time zone: Etc/UTC
tzcode source: system (glibc)

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] Seurat_5.2.1       SeuratObject_5.0.2 sp_2.2-0           lubridate_1.9.4   
 [5] forcats_1.0.0      stringr_1.5.1      dplyr_1.1.4        purrr_1.0.4       
 [9] readr_2.1.5        tidyr_1.3.1        tibble_3.2.1       ggplot2_3.5.1     
[13] tidyverse_2.0.0    workflowr_1.7.1   

loaded via a namespace (and not attached):
  [1] RColorBrewer_1.1-3          rstudioapi_0.17.1          
  [3] jsonlite_1.9.1              magrittr_2.0.3             
  [5] spatstat.utils_3.1-2        farver_2.1.2               
  [7] rmarkdown_2.29              zlibbioc_1.52.0            
  [9] fs_1.6.5                    vctrs_0.6.5                
 [11] ROCR_1.0-11                 DelayedMatrixStats_1.28.1  
 [13] spatstat.explore_3.3-4      S4Arrays_1.6.0             
 [15] htmltools_0.5.8.1           SparseArray_1.6.2          
 [17] sass_0.4.9                  sctransform_0.4.1          
 [19] parallelly_1.42.0           KernSmooth_2.23-24         
 [21] bslib_0.9.0                 htmlwidgets_1.6.4          
 [23] ica_1.0-3                   plyr_1.8.9                 
 [25] plotly_4.10.4               zoo_1.8-13                 
 [27] cachem_1.1.0                whisker_0.4.1              
 [29] igraph_2.1.4                mime_0.12                  
 [31] lifecycle_1.0.4             pkgconfig_2.0.3            
 [33] Matrix_1.7-0                R6_2.6.1                   
 [35] fastmap_1.2.0               GenomeInfoDbData_1.2.13    
 [37] MatrixGenerics_1.18.1       fitdistrplus_1.2-2         
 [39] future_1.34.0               shiny_1.10.0               
 [41] digest_0.6.37               colorspace_2.1-1           
 [43] S4Vectors_0.44.0            patchwork_1.3.0            
 [45] ps_1.9.0                    rprojroot_2.0.4            
 [47] tensor_1.5                  RSpectra_0.16-2            
 [49] irlba_2.3.5.1               GenomicRanges_1.58.0       
 [51] labeling_0.4.3              progressr_0.15.1           
 [53] spatstat.sparse_3.1-0       timechange_0.3.0           
 [55] httr_1.4.7                  polyclip_1.10-7            
 [57] abind_1.4-8                 compiler_4.4.1             
 [59] withr_3.0.2                 fastDummies_1.7.5          
 [61] MASS_7.3-60.2               DelayedArray_0.32.0        
 [63] tools_4.4.1                 lmtest_0.9-40              
 [65] httpuv_1.6.15               future.apply_1.11.3        
 [67] goftest_1.2-3               glmGamPoi_1.18.0           
 [69] glue_1.8.0                  callr_3.7.6                
 [71] nlme_3.1-164                promises_1.3.2             
 [73] grid_4.4.1                  Rtsne_0.17                 
 [75] getPass_0.2-4               cluster_2.1.6              
 [77] reshape2_1.4.4              generics_0.1.3             
 [79] gtable_0.3.6                spatstat.data_3.1-4        
 [81] tzdb_0.4.0                  data.table_1.17.0          
 [83] hms_1.1.3                   XVector_0.46.0             
 [85] BiocGenerics_0.52.0         spatstat.geom_3.3-5        
 [87] RcppAnnoy_0.0.22            ggrepel_0.9.6              
 [89] RANN_2.6.2                  pillar_1.10.1              
 [91] spam_2.11-1                 RcppHNSW_0.6.0             
 [93] later_1.4.1                 splines_4.4.1              
 [95] lattice_0.22-6              survival_3.6-4             
 [97] deldir_2.0-4                tidyselect_1.2.1           
 [99] miniUI_0.1.1.1              pbapply_1.7-2              
[101] knitr_1.49                  git2r_0.35.0               
[103] gridExtra_2.3               IRanges_2.40.1             
[105] SummarizedExperiment_1.36.0 scattermore_1.2            
[107] stats4_4.4.1                xfun_0.51                  
[109] Biobase_2.66.0              matrixStats_1.5.0          
[111] UCSC.utils_1.2.0            stringi_1.8.4              
[113] lazyeval_0.2.2              yaml_2.3.10                
[115] evaluate_1.0.3              codetools_0.2-20           
[117] cli_3.6.4                   uwot_0.2.3                 
[119] xtable_1.8-4                reticulate_1.41.0          
[121] munsell_0.5.1               processx_3.8.6             
[123] jquerylib_0.1.4             GenomeInfoDb_1.42.3        
[125] Rcpp_1.0.14                 globals_0.16.3             
[127] spatstat.random_3.3-2       png_0.1-8                  
[129] spatstat.univar_3.1-2       parallel_4.4.1             
[131] dotCall64_1.2               sparseMatrixStats_1.18.0   
[133] listenv_0.9.1               viridisLite_0.4.2          
[135] scales_1.3.0                ggridges_0.5.6             
[137] crayon_1.5.3                rlang_1.1.5                
[139] cowplot_1.1.3              </code></pre>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
https://docs.mathjax.org/en/latest/web/configuration.html. This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>




</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
