<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />



<meta name="date" content="2024-08-29" />

<title>Getting started with Seurat</title>

<script src="site_libs/header-attrs-2.27/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.13.2/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-6.4.2/css/all.min.css" rel="stylesheet" />
<link href="site_libs/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet" />

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/main/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">My blog</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/davetang/muse">
    <span class="fa fa-github"></span>
     
    Source code
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Getting started with Seurat</h1>
<h4 class="date">2024-08-29</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span>
workflowr <span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2024-08-29
</p>
<p>
<strong>Checks:</strong> <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 7
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong> <code>muse/</code> <span
class="glyphicon glyphicon-question-sign" aria-hidden="true"
title="This is the local directory in which the code in this file was executed.">
</span>
</p>
<p>
This reproducible <a href="https://rmarkdown.rstudio.com">R Markdown</a>
analysis was created with <a
  href="https://github.com/workflowr/workflowr">workflowr</a> (version
1.7.1). The <em>Checks</em> tab describes the reproducibility checks
that were applied when the results were created. The <em>Past
versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date
</a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git
repository, you know the exact version of the code that produced these
results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the
global environment can affect the analysis in your R Markdown file in
unknown ways. For reproduciblity it’s best to always run the code in an
empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20200712code">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Seed:</strong>
<code>set.seed(20200712)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20200712code"
class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20200712)</code> was run prior to running
the code in the R Markdown file. Setting a seed ensures that any results
that rely on randomness, e.g. subsampling or permutations, are
reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Session information:</strong>
recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package
versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be
confident that you successfully produced the results during this
run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr
project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomdavetangmusetree717d440cc434d3600b98502c3de91aa2d9112e5ftargetblank717d440a">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Repository version:</strong>
<a href="https://github.com/davetang/muse/tree/717d440cc434d3600b98502c3de91aa2d9112e5f" target="_blank">717d440</a>
</a>
</p>
</div>
<div
id="strongRepositoryversionstrongahrefhttpsgithubcomdavetangmusetree717d440cc434d3600b98502c3de91aa2d9112e5ftargetblank717d440a"
class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development
and connecting the code version to the results is critical for
reproducibility.
</p>
<p>
The results in this page were generated with repository version
<a href="https://github.com/davetang/muse/tree/717d440cc434d3600b98502c3de91aa2d9112e5f" target="_blank">717d440</a>.
See the <em>Past versions</em> tab to see a history of the changes made
to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for
the analysis have been committed to Git prior to generating the results
(you can use <code>wflow_publish</code> or
<code>wflow_git_commit</code>). workflowr only checks the R Markdown
file, but you know if there are other scripts or data files that it
depends on. Below is the status of the Git repository when the results
were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .Rhistory
    Ignored:    .Rproj.user/
    Ignored:    analysis/figure/
    Ignored:    r_packages_4.3.3/
    Ignored:    r_packages_4.4.0/

Untracked files:
    Untracked:  data/pbmc3k.csv.gz
    Untracked:  data/pbmc3k/

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not
included in this status report because it is ok for generated content to
have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">
<p>
These are the previous versions of the repository in which changes were
made to the R Markdown (<code>analysis/seurat.Rmd</code>) and HTML
(<code>docs/seurat.html</code>) files. If you’ve configured a remote Git
repository (see <code>?wflow_git_remote</code>), click on the hyperlinks
in the table below to view the files as they were in that past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/davetang/muse/blob/717d440cc434d3600b98502c3de91aa2d9112e5f/analysis/seurat.Rmd" target="_blank">717d440</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-08-29
</td>
<td>
Re-run using updated Seurat version
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/davetang/muse/8188fefc3d1003fd5abbe964b502e6ff138cbb37/docs/seurat.html" target="_blank">8188fef</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-04-15
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/davetang/muse/blob/d10317047924abe45011362f723a76cee5584658/analysis/seurat.Rmd" target="_blank">d103170</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-04-15
</td>
<td>
Variance explained
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/davetang/muse/b1853704502f2dbd42a59d77f9aef121ab4fab46/docs/seurat.html" target="_blank">b185370</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-04-15
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/davetang/muse/blob/30c319dfbe945dc04793063fa25c1d291aad990e/analysis/seurat.Rmd" target="_blank">30c319d</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-04-15
</td>
<td>
Download example AnnData file
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/davetang/muse/237607ad3a5552ae89c7cfde592723ad4c043c61/docs/seurat.html" target="_blank">237607a</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-04-07
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/davetang/muse/blob/2ba0e6da90de4dbd71aa78bf458f90ce3a0a03bb/analysis/seurat.Rmd" target="_blank">2ba0e6d</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-04-07
</td>
<td>
Fix zip conditional
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/davetang/muse/5f6b5a75f1dfa606244f52e6a9963e2bdfef2cc3/docs/seurat.html" target="_blank">5f6b5a7</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-04-07
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/davetang/muse/blob/f5d29f970deab5f59ae91fd86dcb9cb3d4f9deeb/analysis/seurat.Rmd" target="_blank">f5d29f9</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-04-07
</td>
<td>
Export as CSV
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/davetang/muse/261b9997754bdabca3acce69f971000c40717909/docs/seurat.html" target="_blank">261b999</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-03-20
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/davetang/muse/blob/de7593195062d3844a6de1b0f46c9657d6c10a5a/analysis/seurat.Rmd" target="_blank">de75931</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-03-20
</td>
<td>
Filtering on mitochondrial percent
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/davetang/muse/03bb7a1bd974a1557d1f99682950668bc9d23b48/docs/seurat.html" target="_blank">03bb7a1</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-03-20
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/davetang/muse/blob/088718440c6b969bff3e4c7980f98fc533117206/analysis/seurat.Rmd" target="_blank">0887184</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-03-20
</td>
<td>
Complete official tutorial
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/davetang/muse/23b21ad1f7602cb64f89100f5b18a774e609aad1/docs/seurat.html" target="_blank">23b21ad</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-03-20
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/davetang/muse/blob/64ac12fd2f21d398fc957a7d6de9abeab3cdbebc/analysis/seurat.Rmd" target="_blank">64ac12f</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-03-20
</td>
<td>
Normalisation, variable genes, and scaling
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/davetang/muse/3c41e26bf6b718001ae68b125ecab024a53d86b3/docs/seurat.html" target="_blank">3c41e26</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-03-20
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/davetang/muse/blob/d5ab5a8c8fa26bd2edd6a2632ce93935120ed49d/analysis/seurat.Rmd" target="_blank">d5ab5a8</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-03-20
</td>
<td>
Basic filtering
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/davetang/muse/5de3e6ce917c55d8600f4435d3952f90603e5801/docs/seurat.html" target="_blank">5de3e6c</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-03-19
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/davetang/muse/blob/0e1dfa9172cb55b17447f3a621fc9b61cf175124/analysis/seurat.Rmd" target="_blank">0e1dfa9</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-03-19
</td>
<td>
Sparse matrix
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/davetang/muse/944e5f2fc3dcce3e4b63a2b724876f092d25f4ab/docs/seurat.html" target="_blank">944e5f2</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-03-19
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/davetang/muse/blob/ae0ea429baa8faa1df208288d885bfa5b4d6043d/analysis/seurat.Rmd" target="_blank">ae0ea42</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-03-19
</td>
<td>
Getting started with Seurat
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<p>This post follows the Peripheral Blood Mononuclear Cells (PBMCs) <a
href="https://satijalab.org/seurat/articles/pbmc3k_tutorial">tutorial</a>
for 2,700 single cells. It was written while I was going through the
tutorial and contains my notes. The dataset for this tutorial can be
downloaded from the <a
href="https://support.10xgenomics.com/single-cell-gene-expression/datasets/1.1.0/pbmc3k">10X
Genomics dataset page</a> but it is also hosted on Amazon (see below).
The PBMCs, which are primary cells with relatively small amounts of RNA
(around 1pg RNA/cell), come from a healthy donor. There were 2,700 cells
detected and sequencing was performed on an Illumina NextSeq 500 with
around 69,000 reads per cell. To get started <a
href="https://satijalab.org/seurat/articles/install.html">install
Seurat</a> by using install.packages().</p>
<pre class="r"><code>install.packages(&quot;Seurat&quot;)
remotes::install_github(&quot;immunogenomics/presto&quot;)</code></pre>
<p>If you get the warning:</p>
<blockquote>
<p>‘SeuratObject’ was built under R 4.3.0 but the current version is
4.3.2; it is recomended that you reinstall ‘SeuratObject’ as the ABI for
R may have changed</p>
</blockquote>
<p>re-install the <code>SeuratObject</code> package using a repository
that has an updated copy. The same goes for the <code>htmltools</code>
package.</p>
<pre class="r"><code>install.packages(&quot;SeuratObject&quot;, repos = &quot;https://cloud.r-project.org/&quot;)
install.packages(&quot;htmltools&quot;, repos = &quot;https://cloud.r-project.org/&quot;)
packageVersion(&quot;SeuratObject&quot;)
packageVersion(&quot;htmltools&quot;)</code></pre>
<p>Load <code>Seurat</code> and <code>bench</code> for some
benchmarking.</p>
<pre class="r"><code>library(&quot;Seurat&quot;)</code></pre>
<pre><code>Loading required package: SeuratObject</code></pre>
<pre><code>Loading required package: sp</code></pre>
<pre><code>
Attaching package: &#39;SeuratObject&#39;</code></pre>
<pre><code>The following objects are masked from &#39;package:base&#39;:

    intersect, t</code></pre>
<pre class="r"><code>library(&quot;bench&quot;)
library(&quot;presto&quot;)</code></pre>
<pre><code>Loading required package: Rcpp</code></pre>
<pre><code>Loading required package: data.table</code></pre>
<pre class="r"><code>packageVersion(&quot;Seurat&quot;)</code></pre>
<pre><code>[1] &#39;5.1.0&#39;</code></pre>
<div id="data" class="section level2">
<h2>Data</h2>
<p>To follow the tutorial, you need the 10X data.</p>
<pre class="bash"><code>mkdir -p data/pbmc3k &amp;&amp; cd data/pbmc3k
wget -c https://s3-us-west-2.amazonaws.com/10x.files/samples/cell/pbmc3k/pbmc3k_filtered_gene_bc_matrices.tar.gz
tar -xzf pbmc3k_filtered_gene_bc_matrices.tar.gz</code></pre>
<p>The extracted files.</p>
<pre class="bash"><code>ls -1 data/pbmc3k/filtered_gene_bc_matrices/hg19</code></pre>
<pre><code>barcodes.tsv
genes.tsv
matrix.mtx</code></pre>
<p><code>matrix.mtx</code> is a <a
href="https://math.nist.gov/MatrixMarket/formats.html">MatrixMarket</a>
file. It has the following properties:</p>
<ul>
<li>Only non-zero entries are stored in the file</li>
<li>Comments start with a <code>%</code>, like LaTeX</li>
<li>The first line indicates the total number of rows, columns, and
entries</li>
<li>The following lines after the first provide a row and column number
and the value at that coordinate</li>
</ul>
<pre class="bash"><code>head data/pbmc3k/filtered_gene_bc_matrices/hg19/matrix.mtx</code></pre>
<pre><code>%%MatrixMarket matrix coordinate real general
%
32738 2700 2286884
32709 1 4
32707 1 1
32706 1 10
32704 1 1
32703 1 5
32702 1 6
32700 1 10</code></pre>
</div>
<div id="seurat-object" class="section level2">
<h2>Seurat object</h2>
<p>Load 10x data into a matrix.</p>
<pre class="r"><code>bench::mark(
  pbmc.data &lt;- Read10X(data.dir = &quot;data/pbmc3k/filtered_gene_bc_matrices/hg19/&quot;)
)</code></pre>
<pre><code>Warning: Some expressions had a GC in every iteration; so filtering is
disabled.</code></pre>
<pre><code># A tibble: 1 × 6
  expression                             min median `itr/sec` mem_alloc `gc/sec`
  &lt;bch:expr&gt;                           &lt;bch&gt; &lt;bch:&gt;     &lt;dbl&gt; &lt;bch:byt&gt;    &lt;dbl&gt;
1 &quot;pbmc.data &lt;- Read10X(data.dir = \&quot;… 866ms  866ms      1.16     206MB     4.62</code></pre>
<p><a
href="https://stat.ethz.ch/R-manual/R-devel/library/Matrix/html/dgTMatrix-class.html">dgTMatrix-class</a>.</p>
<pre class="r"><code>class(pbmc.data)</code></pre>
<pre><code>[1] &quot;dgCMatrix&quot;
attr(,&quot;package&quot;)
[1] &quot;Matrix&quot;</code></pre>
<p>32,738 genes and 2,700 cells.</p>
<pre class="r"><code>dim(pbmc.data)</code></pre>
<pre><code>[1] 32738  2700</code></pre>
<p>Save as CSV file.</p>
<pre class="r"><code>system.time(
  write.csv(x = pbmc.data, file = &quot;data/pbmc3k.csv&quot;)
)</code></pre>
<pre><code>   user  system elapsed 
 51.866   0.467  52.346 </code></pre>
<p>Gzip.</p>
<pre class="bash"><code>if [[ ! -f data/pbmc3k.csv.gz ]]; then
   gzip data/pbmc3k.csv
fi
ls -lh data/pbmc3k.csv.gz</code></pre>
<pre><code>-rw-r--r-- 1 rstudio rstudio 3.4M Aug 29 09:22 data/pbmc3k.csv.gz</code></pre>
<p>Check out the first six genes and cells</p>
<pre class="r"><code>pbmc.data[1:6, 1:6]</code></pre>
<pre><code>6 x 6 sparse Matrix of class &quot;dgCMatrix&quot;
             AAACATACAACCAC-1 AAACATTGAGCTAC-1 AAACATTGATCAGC-1
MIR1302-10                  .                .                .
FAM138A                     .                .                .
OR4F5                       .                .                .
RP11-34P13.7                .                .                .
RP11-34P13.8                .                .                .
AL627309.1                  .                .                .
             AAACCGTGCTTCCG-1 AAACCGTGTATGCG-1 AAACGCACTGGTAC-1
MIR1302-10                  .                .                .
FAM138A                     .                .                .
OR4F5                       .                .                .
RP11-34P13.7                .                .                .
RP11-34P13.8                .                .                .
AL627309.1                  .                .                .</code></pre>
<p>Summary of total expression per single cell.</p>
<pre class="r"><code>summary(colSums(pbmc.data))</code></pre>
<pre><code>   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
    548    1758    2197    2367    2763   15844 </code></pre>
<p>Check how many genes have at least one transcript in each cell.</p>
<p>The median number of detected genes among the single cells is
817.</p>
<pre class="r"><code>at_least_one &lt;- apply(pbmc.data, 2, function(x) sum(x&gt;0))
hist(
  at_least_one,
  breaks = 100,
  main = &quot;Distribution of detected genes&quot;,
  xlab = &quot;Genes with at least one tag&quot;
)
abline(v = median(at_least_one), col = 2, lty = 3)</code></pre>
<p><img src="figure/seurat.Rmd/unnamed-chunk-10-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-10-1">
Past versions of unnamed-chunk-10-1.png
</button>
</p>
<div id="fig-unnamed-chunk-10-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/davetang/muse/blob/5f6b5a75f1dfa606244f52e6a9963e2bdfef2cc3/docs/figure/seurat.Rmd/unnamed-chunk-10-1.png" target="_blank">5f6b5a7</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-04-07
</td>
</tr>
<tr>
<td>
<a href="https://github.com/davetang/muse/blob/3c41e26bf6b718001ae68b125ecab024a53d86b3/docs/figure/seurat.Rmd/unnamed-chunk-10-1.png" target="_blank">3c41e26</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-03-20
</td>
</tr>
<tr>
<td>
<a href="https://github.com/davetang/muse/blob/5de3e6ce917c55d8600f4435d3952f90603e5801/docs/figure/seurat.Rmd/unnamed-chunk-10-1.png" target="_blank">5de3e6c</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-03-19
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>Total expression per cell. The median sum of expression among the
single cells is 2,197. This distribution is very similar to the
distribution of detected genes shown above.</p>
<pre class="r"><code>hist(
  colSums(pbmc.data),
  breaks = 100,
  main = &quot;Expression sum per cell&quot;,
  xlab = &quot;Sum expression&quot;
)
abline(v = median(colSums(pbmc.data)), col = 2, lty = 3)</code></pre>
<p><img src="figure/seurat.Rmd/unnamed-chunk-11-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-11-1">
Past versions of unnamed-chunk-11-1.png
</button>
</p>
<div id="fig-unnamed-chunk-11-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/davetang/muse/blob/5f6b5a75f1dfa606244f52e6a9963e2bdfef2cc3/docs/figure/seurat.Rmd/unnamed-chunk-11-1.png" target="_blank">5f6b5a7</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-04-07
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>We will filter out genes and single cells before we continue with the
analysis. The tutorial has arbitrary values of keeping genes expressed
in three or more cells and keeping cells with at least 200 detected
genes.</p>
<p>Manually check the number of genes detected in three or more cells; a
lot of genes are not detected in 3 or more cells.</p>
<pre class="r"><code>tmp &lt;- apply(pbmc.data, 1, function(x) sum(x&gt;0))
table(tmp&gt;=3)</code></pre>
<pre><code>
FALSE  TRUE 
19024 13714 </code></pre>
<p>All cells have at least 200 detected genes</p>
<pre class="r"><code>keep &lt;- tmp&gt;=3
tmp &lt;- pbmc.data[keep,]
at_least_one &lt;- apply(tmp, 2, function(x) sum(x&gt;0))
summary(at_least_one)</code></pre>
<pre><code>   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
  212.0   690.0   816.0   845.5   952.0  3400.0 </code></pre>
<pre class="r"><code>dim(tmp)</code></pre>
<pre><code>[1] 13714  2700</code></pre>
<p>See <code>?SeuratObject</code> for more information on the class.</p>
<pre class="r"><code>pbmc &lt;- CreateSeuratObject(
  counts = pbmc.data,
  min.cells = 3,
  min.features = 200,
  project = &quot;pbmc3k&quot;
)</code></pre>
<pre><code>Warning: Feature names cannot have underscores (&#39;_&#39;), replacing with dashes
(&#39;-&#39;)</code></pre>
<pre class="r"><code>class(pbmc)</code></pre>
<pre><code>[1] &quot;Seurat&quot;
attr(,&quot;package&quot;)
[1] &quot;SeuratObject&quot;</code></pre>
<p>Same numbers as above</p>
<pre class="r"><code>pbmc</code></pre>
<pre><code>An object of class Seurat 
13714 features across 2700 samples within 1 assay 
Active assay: RNA (13714 features, 0 variable features)
 1 layer present: counts</code></pre>
<p>Slots in Seurat object.</p>
<blockquote>
<p>SeuratObject: Data Structures for Single Cell Data</p>
<p>Defines S4 classes for single-cell genomic data and associated
information, such as dimensionality reduction embeddings,
nearest-neighbor graphs, and spatially-resolved coordinates. Provides
data access methods and R-native hooks to ensure the Seurat object is
familiar to other R users</p>
</blockquote>
<p>Read more about the <a href="https://adv-r.hadley.nz/s4.html">S4
class</a> in the Advanced R book.</p>
<pre class="r"><code>slotNames(pbmc)</code></pre>
<pre><code> [1] &quot;assays&quot;       &quot;meta.data&quot;    &quot;active.assay&quot; &quot;active.ident&quot; &quot;graphs&quot;      
 [6] &quot;neighbors&quot;    &quot;reductions&quot;   &quot;images&quot;       &quot;project.name&quot; &quot;misc&quot;        
[11] &quot;version&quot;      &quot;commands&quot;     &quot;tools&quot;       </code></pre>
</div>
<div id="basic-filtering" class="section level2">
<h2>Basic filtering</h2>
<p>The tutorial states that “The number of genes and UMIs (nGene and
nUMI) are automatically calculated for every object by Seurat.” The nUMI
is calculated as <code>num.mol &lt;- colSums(object.raw.data)</code>,
i.e. each transcript is a unique molecule. The number of genes is simply
the tally of genes with at least 1 transcript;
<code>num.genes &lt;- colSums(object.raw.data &gt; is.expr</code>) where
<code>is.expr</code> is zero.</p>
<p>A common quality control metric is the percentage of transcripts from
the mitochondrial genome. According to the paper <a
href="https://pubmed.ncbi.nlm.nih.gov/26887813/">Classification of low
quality cells from single-cell RNA-seq data</a> the reason this is a
quality control metric is because if a single cell is lysed, cytoplasmic
RNA will be lost apart from the RNA that is enclosed in the
mitochondria, which will be retained and sequenced.</p>
<p>Human mitochondria genes conveniently start with MT.</p>
<pre class="r"><code>mito.genes &lt;- grep(pattern = &quot;^MT-&quot;, x = rownames(x = pbmc@assays$RNA), value = TRUE)
length(mito.genes)</code></pre>
<pre><code>[1] 13</code></pre>
<pre class="r"><code>percent.mito &lt;- Matrix::colSums(pbmc[[&#39;RNA&#39;]]$counts[mito.genes, ]) / Matrix::colSums(pbmc[[&#39;RNA&#39;]]$counts) * 100
head(percent.mito)</code></pre>
<pre><code>AAACATACAACCAC-1 AAACATTGAGCTAC-1 AAACATTGATCAGC-1 AAACCGTGCTTCCG-1 
       3.0177759        3.7935958        0.8897363        1.7430845 
AAACCGTGTATGCG-1 AAACGCACTGGTAC-1 
       1.2244898        1.6643551 </code></pre>
<p>Check out the meta data.</p>
<pre class="r"><code>head(pbmc@meta.data)</code></pre>
<pre><code>                 orig.ident nCount_RNA nFeature_RNA
AAACATACAACCAC-1     pbmc3k       2419          779
AAACATTGAGCTAC-1     pbmc3k       4903         1352
AAACATTGATCAGC-1     pbmc3k       3147         1129
AAACCGTGCTTCCG-1     pbmc3k       2639          960
AAACCGTGTATGCG-1     pbmc3k        980          521
AAACGCACTGGTAC-1     pbmc3k       2163          781</code></pre>
<p>Add mitochondrial percent to the meta using
<code>AddMetaData</code>.</p>
<pre class="r"><code>pbmc &lt;- AddMetaData(
  object = pbmc,
  metadata = percent.mito,
  col.name = &quot;percent.mito&quot;
)
head(pbmc@meta.data)</code></pre>
<pre><code>                 orig.ident nCount_RNA nFeature_RNA percent.mito
AAACATACAACCAC-1     pbmc3k       2419          779    3.0177759
AAACATTGAGCTAC-1     pbmc3k       4903         1352    3.7935958
AAACATTGATCAGC-1     pbmc3k       3147         1129    0.8897363
AAACCGTGCTTCCG-1     pbmc3k       2639          960    1.7430845
AAACCGTGTATGCG-1     pbmc3k        980          521    1.2244898
AAACGCACTGGTAC-1     pbmc3k       2163          781    1.6643551</code></pre>
<p>Use <code>PercentageFeatureSet</code> to calculate the percentage of
a set of features, which saves us from some typing. The <code>[[</code>
operator can add columns to object metadata, which is a great place to
stash QC stats.</p>
<pre class="r"><code>pbmc[[&quot;percent.mt&quot;]] &lt;- PercentageFeatureSet(pbmc, pattern = &quot;^MT-&quot;)
head(pbmc@meta.data[, c(&#39;percent.mito&#39;, &#39;percent.mt&#39;)])</code></pre>
<pre><code>                 percent.mito percent.mt
AAACATACAACCAC-1    3.0177759  3.0177759
AAACATTGAGCTAC-1    3.7935958  3.7935958
AAACATTGATCAGC-1    0.8897363  0.8897363
AAACCGTGCTTCCG-1    1.7430845  1.7430845
AAACCGTGTATGCG-1    1.2244898  1.2244898
AAACGCACTGGTAC-1    1.6643551  1.6643551</code></pre>
<p>Instead of setting a hard filtering threshold, one can use the 3 * <a
href="https://www.sc-best-practices.org/preprocessing_visualization/quality_control.html#filtering-low-quality-reads">Median
Absolute Deviation</a> to determine threshold limits for filtering out
cells.</p>
<pre class="r"><code>median(pbmc@meta.data$percent.mt) - 3 * mad(pbmc@meta.data$percent.mt)</code></pre>
<pre><code>[1] -0.3751742</code></pre>
<pre class="r"><code>median(pbmc@meta.data$percent.mt) + 3 * mad(pbmc@meta.data$percent.mt)</code></pre>
<pre><code>[1] 4.436775</code></pre>
<p>Plot number of genes, UMIs, and % mitochondria</p>
<p>Visualize QC metrics as a violin plot</p>
<pre class="r"><code>VlnPlot(pbmc, features = c(&quot;nFeature_RNA&quot;, &quot;nCount_RNA&quot;, &quot;percent.mt&quot;), ncol = 3)</code></pre>
<pre><code>Warning: Default search for &quot;data&quot; layer in &quot;RNA&quot; assay yielded no results;
utilizing &quot;counts&quot; layer instead.</code></pre>
<p><img src="figure/seurat.Rmd/unnamed-chunk-22-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-22-1">
Past versions of unnamed-chunk-22-1.png
</button>
</p>
<div id="fig-unnamed-chunk-22-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/davetang/muse/blob/5f6b5a75f1dfa606244f52e6a9963e2bdfef2cc3/docs/figure/seurat.Rmd/unnamed-chunk-22-1.png" target="_blank">5f6b5a7</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-04-07
</td>
</tr>
<tr>
<td>
<a href="https://github.com/davetang/muse/blob/261b9997754bdabca3acce69f971000c40717909/docs/figure/seurat.Rmd/unnamed-chunk-22-1.png" target="_blank">261b999</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-03-20
</td>
</tr>
<tr>
<td>
<a href="https://github.com/davetang/muse/blob/3c41e26bf6b718001ae68b125ecab024a53d86b3/docs/figure/seurat.Rmd/unnamed-chunk-22-1.png" target="_blank">3c41e26</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-03-20
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>A couple of cells have high mitochondrial percentage which may
indicate lost of cytoplasmic RNA.</p>
<p>The GenePlot() function can be used to visualise gene-gene
relationships as well as any columns in the seurat object. Below we use
the plotting function to spot cells that have a high percentage of
mitochondrial RNA and to plot the relationship between the number of
unique molecules and the number of genes captured.</p>
<p>FeatureScatter is typically used to visualize feature-feature
relationships, but can be used for anything calculated by the object,
i.e. columns in object metadata, PC scores etc.</p>
<pre class="r"><code>plot1 &lt;- FeatureScatter(pbmc, feature1 = &quot;nCount_RNA&quot;, feature2 = &quot;percent.mt&quot;)
plot2 &lt;- FeatureScatter(pbmc, feature1 = &quot;nCount_RNA&quot;, feature2 = &quot;nFeature_RNA&quot;)
plot1 + plot2</code></pre>
<p><img src="figure/seurat.Rmd/unnamed-chunk-23-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-23-1">
Past versions of unnamed-chunk-23-1.png
</button>
</p>
<div id="fig-unnamed-chunk-23-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/davetang/muse/blob/5f6b5a75f1dfa606244f52e6a9963e2bdfef2cc3/docs/figure/seurat.Rmd/unnamed-chunk-23-1.png" target="_blank">5f6b5a7</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-04-07
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>Manual check; I already know all cells have &gt;200 genes.</p>
<pre class="r"><code>table(pbmc@meta.data$percent.mito &lt; 5 &amp; pbmc@meta.data$nFeature_RNA&lt;2500)</code></pre>
<pre><code>
FALSE  TRUE 
   62  2638 </code></pre>
<pre class="r"><code>pbmc &lt;- subset(pbmc, subset = nFeature_RNA &gt; 200 &amp; nFeature_RNA &lt; 2500 &amp; percent.mt &lt; 5)
pbmc</code></pre>
<pre><code>An object of class Seurat 
13714 features across 2638 samples within 1 assay 
Active assay: RNA (13714 features, 0 variable features)
 1 layer present: counts</code></pre>
</div>
<div id="normalisation" class="section level2">
<h2>Normalisation</h2>
<p>The next step is to normalise the data, so that each cell can be
compared against each other. At the time of writing, the only
normalisation method implemented in Seurat is by log normalisation. Gene
expression measurements for each cell are normalised by its total
expression, scaled by 10,000, and log-transformed.</p>
<pre class="r"><code>hist(
  colSums(pbmc[[&#39;RNA&#39;]]$counts),
  breaks = 100,
  main = &quot;Total expression before normalisation&quot;,
  xlab = &quot;Sum of expression&quot;
)</code></pre>
<p><img src="figure/seurat.Rmd/unnamed-chunk-26-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-26-1">
Past versions of unnamed-chunk-26-1.png
</button>
</p>
<div id="fig-unnamed-chunk-26-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/davetang/muse/blob/5f6b5a75f1dfa606244f52e6a9963e2bdfef2cc3/docs/figure/seurat.Rmd/unnamed-chunk-26-1.png" target="_blank">5f6b5a7</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-04-07
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>After removing unwanted cells from the dataset, the next step is to
normalise the data. By default, we employ a global-scaling normalization
method “LogNormalize” that normalises the feature expression
measurements for each cell by the total expression, multiplies this by a
scale factor (10,000 by default), and log-transforms the result. In
Seurat v5, Normalized values are stored in pbmc[[“RNA”]]$data.</p>
<pre class="r"><code>pbmc &lt;- NormalizeData(pbmc, normalization.method = &quot;LogNormalize&quot;, scale.factor = 10000)</code></pre>
<pre><code>Normalizing layer: counts</code></pre>
<p>For clarity, in this previous line of code (and in future commands),
we provide the default values for certain parameters in the function
call. However, this isn’t required and the same behavior can be achieved
with:</p>
<pre class="r"><code>pbmc &lt;- NormalizeData(pbmc)</code></pre>
<pre><code>Normalizing layer: counts</code></pre>
<p>While this method of normalization is standard and widely used in
scRNA-seq analysis, global-scaling relies on an assumption that each
cell originally contains the same number of RNA molecules. We and others
have developed alternative workflows for the single cell preprocessing
that do not make these assumptions. For users who are interested, please
check out our SCTransform() normalization workflow. The method is
described in ourpaper, with a separate vignette using Seurat here. The
use of SCTransform replaces the need to run NormalizeData,
FindVariableFeatures, or ScaleData (described below.)</p>
<pre class="r"><code>hist(
  colSums(pbmc[[&#39;RNA&#39;]]$data),
  breaks = 100,
  main = &quot;Total expression after normalisation&quot;,
  xlab = &quot;Sum of expression&quot;
)</code></pre>
<p><img src="figure/seurat.Rmd/unnamed-chunk-29-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-29-1">
Past versions of unnamed-chunk-29-1.png
</button>
</p>
<div id="fig-unnamed-chunk-29-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/davetang/muse/blob/5f6b5a75f1dfa606244f52e6a9963e2bdfef2cc3/docs/figure/seurat.Rmd/unnamed-chunk-29-1.png" target="_blank">5f6b5a7</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-04-07
</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<div id="identification-of-highly-variable-features-feature-selection"
class="section level2">
<h2>Identification of highly variable features (feature selection)</h2>
<p>Once the data is normalised, the next step is to find genes are vary
between single cells; genes that are constant among all cells have no
distinguishing power. The <code>FindVariableFeatures()</code> function
calculates the average expression and dispersion for each gene, places
these genes into bins, and then calculates a z-score for dispersion
within each bin. I interpret that as take each gene, get the average
expression and variance of the gene across the 2,638 cells, categorise
genes into bins (default is 20) based on their expression and variance,
and finally normalise the variance in each bin. This was the same
approach in <a
href="https://www.ncbi.nlm.nih.gov/pubmed/26000488">Macosko et al.</a>
and new methods for detecting genes with variable expression patterns
will be implemented in Seurat soon (according to the tutorial). The
parameters used below are typical settings for UMI data that is
normalised to a total of 10,000 molecules and will identify around 2,000
variable genes. The tutorial recommends that users should explore the
parameters themselves since each dataset is different.</p>
<p>We next calculate a subset of features that exhibit high cell-to-cell
variation in the dataset (i.e, they are highly expressed in some cells,
and lowly expressed in others). We and others have found that focusing
on these genes in downstream analysis helps to highlight biological
signal in single-cell datasets.</p>
<p>Our procedure in Seurat is described in detail here, and improves on
previous versions by directly modeling the mean-variance relationship
inherent in single-cell data, and is implemented in the
FindVariableFeatures() function. By default, we return 2,000 features
per dataset. These will be used in downstream analysis, like PCA.</p>
<pre class="r"><code>pbmc &lt;- FindVariableFeatures(pbmc, selection.method = &quot;vst&quot;, nfeatures = 2000)</code></pre>
<pre><code>Finding variable features for layer counts</code></pre>
<pre class="r"><code>length(VariableFeatures(pbmc))</code></pre>
<pre><code>[1] 2000</code></pre>
<p>Identify the 10 most highly variable genes</p>
<pre class="r"><code>top10 &lt;- head(VariableFeatures(pbmc), 10)
top10</code></pre>
<pre><code> [1] &quot;PPBP&quot;   &quot;LYZ&quot;    &quot;S100A9&quot; &quot;IGLL5&quot;  &quot;GNLY&quot;   &quot;FTL&quot;    &quot;PF4&quot;    &quot;FTH1&quot;  
 [9] &quot;GNG11&quot;  &quot;S100A8&quot;</code></pre>
<p>Plot variable features with and without labels</p>
<pre class="r"><code>plot1 &lt;- VariableFeaturePlot(pbmc)
plot2 &lt;- LabelPoints(plot = plot1, points = top10, repel = TRUE)</code></pre>
<pre><code>When using repel, set xnudge and ynudge to 0 for optimal results</code></pre>
<pre class="r"><code>plot1 + plot2</code></pre>
<pre><code>Warning in scale_x_log10(): log-10 transformation introduced infinite values.
log-10 transformation introduced infinite values.</code></pre>
<p><img src="figure/seurat.Rmd/unnamed-chunk-33-1.png" width="960" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-33-1">
Past versions of unnamed-chunk-33-1.png
</button>
</p>
<div id="fig-unnamed-chunk-33-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/davetang/muse/blob/b1853704502f2dbd42a59d77f9aef121ab4fab46/docs/figure/seurat.Rmd/unnamed-chunk-33-1.png" target="_blank">b185370</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-04-15
</td>
</tr>
<tr>
<td>
<a href="https://github.com/davetang/muse/blob/5f6b5a75f1dfa606244f52e6a9963e2bdfef2cc3/docs/figure/seurat.Rmd/unnamed-chunk-33-1.png" target="_blank">5f6b5a7</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-04-07
</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<div id="scaling-the-data" class="section level2">
<h2>Scaling the data</h2>
<p>Next, we apply a linear transformation (“scaling”) that is a standard
pre-processing step prior to dimensional reduction techniques like PCA.
The <code>ScaleData()</code> function:</p>
<ul>
<li>Shifts the expression of each gene, so that the mean expression
across cells is 0</li>
<li>Scales the expression of each gene, so that the variance across
cells is 1
<ul>
<li>This step gives equal weight in downstream analyses, so that
highly-expressed genes do not dominate</li>
</ul></li>
<li>The results of this are stored in pbmc[[“RNA”]]$scale.data</li>
<li>By default, only variable features are scaled.</li>
<li>You can specify the features argument to scale additional
features</li>
</ul>
<pre class="r"><code>all.genes &lt;- rownames(pbmc)
pbmc &lt;- ScaleData(pbmc, features = all.genes)</code></pre>
<pre><code>Centering and scaling data matrix</code></pre>
<pre class="r"><code>dim(pbmc[[&quot;RNA&quot;]]$scale.data)</code></pre>
<pre><code>[1] 13714  2638</code></pre>
<pre class="r"><code>hist(
  colSums(pbmc[[&#39;RNA&#39;]]$scale.data),
  breaks = 100,
  main = &quot;Total expression after scaling&quot;,
  xlab = &quot;Sum of expression&quot;
)</code></pre>
<p><img src="figure/seurat.Rmd/unnamed-chunk-35-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-35-1">
Past versions of unnamed-chunk-35-1.png
</button>
</p>
<div id="fig-unnamed-chunk-35-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/davetang/muse/blob/5f6b5a75f1dfa606244f52e6a9963e2bdfef2cc3/docs/figure/seurat.Rmd/unnamed-chunk-35-1.png" target="_blank">5f6b5a7</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-04-07
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>How can I remove unwanted sources of variation?</p>
<p>In Seurat, we also use the <code>ScaleData()</code> function to
remove unwanted sources of variation from a single-cell dataset. For
example, we could “regress out” heterogeneity associated with (for
example) cell cycle stage, or mitochondrial contamination i.e.:</p>
<pre class="r"><code>pbmc &lt;- ScaleData(pbmc, features = all.genes, vars.to.regress = &quot;percent.mt&quot;)</code></pre>
<p>However, particularly for advanced users who would like to use this
functionality, we strongly recommend the use of our new normalization
workflow, <code>SCTransform()</code>. The method is described in <a
href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-021-02584-9">this
paper</a>, with a separate <a
href="https://satijalab.org/seurat/articles/sctransform_vignette">vignette
using Seurat</a>. As with <code>ScaleData()</code>, the function
<code>SCTransform()</code> also includes a <code>vars.to.regress</code>
parameter.</p>
</div>
<div id="perform-linear-dimensional-reduction" class="section level2">
<h2>Perform linear dimensional reduction</h2>
<p>Next we perform PCA on the scaled data. By default, only the
previously determined variable features are used as input, but can be
defined using features argument if you wish to choose a different subset
(if you do want to use a custom subset of features, make sure you pass
these to <code>ScaleData</code> first).</p>
<p>For the first principal components, Seurat outputs a list of genes
with the most positive and negative loadings, representing modules of
genes that exhibit either correlation (or anti-correlation) across
single-cells in the dataset.</p>
<pre class="r"><code>pbmc &lt;- RunPCA(pbmc, features = VariableFeatures(object = pbmc))</code></pre>
<pre><code>PC_ 1 
Positive:  CST3, TYROBP, LST1, AIF1, FTL, FTH1, LYZ, FCN1, S100A9, TYMP 
       FCER1G, CFD, LGALS1, S100A8, CTSS, LGALS2, SERPINA1, IFITM3, SPI1, CFP 
       PSAP, IFI30, SAT1, COTL1, S100A11, NPC2, GRN, LGALS3, GSTP1, PYCARD 
Negative:  MALAT1, LTB, IL32, IL7R, CD2, B2M, ACAP1, CD27, STK17A, CTSW 
       CD247, GIMAP5, AQP3, CCL5, SELL, TRAF3IP3, GZMA, MAL, CST7, ITM2A 
       MYC, GIMAP7, HOPX, BEX2, LDLRAP1, GZMK, ETS1, ZAP70, TNFAIP8, RIC3 
PC_ 2 
Positive:  CD79A, MS4A1, TCL1A, HLA-DQA1, HLA-DQB1, HLA-DRA, LINC00926, CD79B, HLA-DRB1, CD74 
       HLA-DMA, HLA-DPB1, HLA-DQA2, CD37, HLA-DRB5, HLA-DMB, HLA-DPA1, FCRLA, HVCN1, LTB 
       BLNK, P2RX5, IGLL5, IRF8, SWAP70, ARHGAP24, FCGR2B, SMIM14, PPP1R14A, C16orf74 
Negative:  NKG7, PRF1, CST7, GZMB, GZMA, FGFBP2, CTSW, GNLY, B2M, SPON2 
       CCL4, GZMH, FCGR3A, CCL5, CD247, XCL2, CLIC3, AKR1C3, SRGN, HOPX 
       TTC38, APMAP, CTSC, S100A4, IGFBP7, ANXA1, ID2, IL32, XCL1, RHOC 
PC_ 3 
Positive:  HLA-DQA1, CD79A, CD79B, HLA-DQB1, HLA-DPB1, HLA-DPA1, CD74, MS4A1, HLA-DRB1, HLA-DRA 
       HLA-DRB5, HLA-DQA2, TCL1A, LINC00926, HLA-DMB, HLA-DMA, CD37, HVCN1, FCRLA, IRF8 
       PLAC8, BLNK, MALAT1, SMIM14, PLD4, LAT2, IGLL5, P2RX5, SWAP70, FCGR2B 
Negative:  PPBP, PF4, SDPR, SPARC, GNG11, NRGN, GP9, RGS18, TUBB1, CLU 
       HIST1H2AC, AP001189.4, ITGA2B, CD9, TMEM40, PTCRA, CA2, ACRBP, MMD, TREML1 
       NGFRAP1, F13A1, SEPT5, RUFY1, TSC22D1, MPP1, CMTM5, RP11-367G6.3, MYL9, GP1BA 
PC_ 4 
Positive:  HLA-DQA1, CD79B, CD79A, MS4A1, HLA-DQB1, CD74, HLA-DPB1, HIST1H2AC, PF4, TCL1A 
       SDPR, HLA-DPA1, HLA-DRB1, HLA-DQA2, HLA-DRA, PPBP, LINC00926, GNG11, HLA-DRB5, SPARC 
       GP9, AP001189.4, CA2, PTCRA, CD9, NRGN, RGS18, GZMB, CLU, TUBB1 
Negative:  VIM, IL7R, S100A6, IL32, S100A8, S100A4, GIMAP7, S100A10, S100A9, MAL 
       AQP3, CD2, CD14, FYB, LGALS2, GIMAP4, ANXA1, CD27, FCN1, RBP7 
       LYZ, S100A11, GIMAP5, MS4A6A, S100A12, FOLR3, TRABD2A, AIF1, IL8, IFI6 
PC_ 5 
Positive:  GZMB, NKG7, S100A8, FGFBP2, GNLY, CCL4, CST7, PRF1, GZMA, SPON2 
       GZMH, S100A9, LGALS2, CCL3, CTSW, XCL2, CD14, CLIC3, S100A12, CCL5 
       RBP7, MS4A6A, GSTP1, FOLR3, IGFBP7, TYROBP, TTC38, AKR1C3, XCL1, HOPX 
Negative:  LTB, IL7R, CKB, VIM, MS4A7, AQP3, CYTIP, RP11-290F20.3, SIGLEC10, HMOX1 
       PTGES3, LILRB2, MAL, CD27, HN1, CD2, GDI2, ANXA5, CORO1B, TUBA1B 
       FAM110A, ATP1A1, TRADD, PPA1, CCDC109B, ABRACL, CTD-2006K23.1, WARS, VMO1, FYB </code></pre>
<p>Seurat provides several useful ways of visualizing both cells and
features that define the PCA, including <code>VizDimReduction()</code>,
<code>DimPlot()</code>, and <code>DimHeatmap()</code>.</p>
<p>Examine and visualize PCA results a few different ways</p>
<pre class="r"><code>print(pbmc[[&quot;pca&quot;]], dims = 1:5, nfeatures = 5)</code></pre>
<pre><code>PC_ 1 
Positive:  CST3, TYROBP, LST1, AIF1, FTL 
Negative:  MALAT1, LTB, IL32, IL7R, CD2 
PC_ 2 
Positive:  CD79A, MS4A1, TCL1A, HLA-DQA1, HLA-DQB1 
Negative:  NKG7, PRF1, CST7, GZMB, GZMA 
PC_ 3 
Positive:  HLA-DQA1, CD79A, CD79B, HLA-DQB1, HLA-DPB1 
Negative:  PPBP, PF4, SDPR, SPARC, GNG11 
PC_ 4 
Positive:  HLA-DQA1, CD79B, CD79A, MS4A1, HLA-DQB1 
Negative:  VIM, IL7R, S100A6, IL32, S100A8 
PC_ 5 
Positive:  GZMB, NKG7, S100A8, FGFBP2, GNLY 
Negative:  LTB, IL7R, CKB, VIM, MS4A7 </code></pre>
<pre class="r"><code>VizDimLoadings(pbmc, dims = 1:2, reduction = &quot;pca&quot;)</code></pre>
<p><img src="figure/seurat.Rmd/unnamed-chunk-39-1.png" width="768" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-39-1">
Past versions of unnamed-chunk-39-1.png
</button>
</p>
<div id="fig-unnamed-chunk-39-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/davetang/muse/blob/5f6b5a75f1dfa606244f52e6a9963e2bdfef2cc3/docs/figure/seurat.Rmd/unnamed-chunk-39-1.png" target="_blank">5f6b5a7</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-04-07
</td>
</tr>
<tr>
<td>
<a href="https://github.com/davetang/muse/blob/03bb7a1bd974a1557d1f99682950668bc9d23b48/docs/figure/seurat.Rmd/unnamed-chunk-39-1.png" target="_blank">03bb7a1</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-03-20
</td>
</tr>
</tbody>
</table>
</div>
</div>
<pre class="r"><code>DimPlot(pbmc, reduction = &quot;pca&quot;) + NoLegend()</code></pre>
<p><img src="figure/seurat.Rmd/unnamed-chunk-40-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-40-1">
Past versions of unnamed-chunk-40-1.png
</button>
</p>
<div id="fig-unnamed-chunk-40-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/davetang/muse/blob/5f6b5a75f1dfa606244f52e6a9963e2bdfef2cc3/docs/figure/seurat.Rmd/unnamed-chunk-40-1.png" target="_blank">5f6b5a7</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-04-07
</td>
</tr>
<tr>
<td>
<a href="https://github.com/davetang/muse/blob/03bb7a1bd974a1557d1f99682950668bc9d23b48/docs/figure/seurat.Rmd/unnamed-chunk-40-1.png" target="_blank">03bb7a1</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-03-20
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>In particular <code>DimHeatmap()</code> allows for easy exploration
of the primary sources of heterogeneity in a dataset, and can be useful
when trying to decide which PCs to include for further downstream
analyses. Both cells and features are ordered according to their PCA
scores. Setting cells to a number plots the ‘extreme’ cells on both ends
of the spectrum, which dramatically speeds plotting for large datasets.
Though clearly a supervised analysis, we find this to be a valuable tool
for exploring correlated feature sets.</p>
<pre class="r"><code>DimHeatmap(pbmc, dims = 1, cells = 500, balanced = TRUE)</code></pre>
<p><img src="figure/seurat.Rmd/unnamed-chunk-41-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-41-1">
Past versions of unnamed-chunk-41-1.png
</button>
</p>
<div id="fig-unnamed-chunk-41-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/davetang/muse/blob/5f6b5a75f1dfa606244f52e6a9963e2bdfef2cc3/docs/figure/seurat.Rmd/unnamed-chunk-41-1.png" target="_blank">5f6b5a7</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-04-07
</td>
</tr>
<tr>
<td>
<a href="https://github.com/davetang/muse/blob/03bb7a1bd974a1557d1f99682950668bc9d23b48/docs/figure/seurat.Rmd/unnamed-chunk-41-1.png" target="_blank">03bb7a1</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-03-20
</td>
</tr>
</tbody>
</table>
</div>
</div>
<pre class="r"><code>DimHeatmap(pbmc, dims = 1:15, cells = 500, balanced = TRUE)</code></pre>
<p><img src="figure/seurat.Rmd/unnamed-chunk-42-1.png" width="768" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-42-1">
Past versions of unnamed-chunk-42-1.png
</button>
</p>
<div id="fig-unnamed-chunk-42-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/davetang/muse/blob/5f6b5a75f1dfa606244f52e6a9963e2bdfef2cc3/docs/figure/seurat.Rmd/unnamed-chunk-42-1.png" target="_blank">5f6b5a7</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-04-07
</td>
</tr>
<tr>
<td>
<a href="https://github.com/davetang/muse/blob/03bb7a1bd974a1557d1f99682950668bc9d23b48/docs/figure/seurat.Rmd/unnamed-chunk-42-1.png" target="_blank">03bb7a1</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-03-20
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p><a href="https://github.com/satijalab/seurat/issues/982">Variance
explained</a>.</p>
<pre class="r"><code># manually calculate
total_variance &lt;- sum(
  matrixStats::rowVars(pbmc[[&#39;RNA&#39;]]$scale.data[VariableFeatures(pbmc), ])
)

stopifnot(all.equal(total_variance, pbmc@reductions$pca@misc$total.variance))

eig_values &lt;- (pbmc@reductions$pca@stdev)^2
eig_values / total_variance</code></pre>
<pre><code> [1] 0.029064316 0.011657124 0.008650478 0.008106527 0.005802778 0.003736905
 [7] 0.002467149 0.002182405 0.001968552 0.001941607 0.001911565 0.001844931
[13] 0.001825969 0.001781458 0.001769644 0.001759204 0.001730575 0.001725452
[19] 0.001712370 0.001707524 0.001701826 0.001696903 0.001687343 0.001678263
[25] 0.001677217 0.001671899 0.001668394 0.001658801 0.001657135 0.001649345
[31] 0.001644300 0.001630762 0.001627221 0.001625400 0.001617703 0.001617336
[37] 0.001608842 0.001604178 0.001602133 0.001597393 0.001590078 0.001585467
[43] 0.001583661 0.001578873 0.001578360 0.001569454 0.001565028 0.001561961
[49] 0.001556993 0.001553569</code></pre>
</div>
<div id="determine-the-dimensionality-of-the-dataset"
class="section level2">
<h2>Determine the dimensionality of the dataset</h2>
<p>To overcome the extensive technical noise in any single feature for
scRNA-seq data, Seurat clusters cells based on their PCA scores, with
each PC essentially representing a ‘metafeature’ that combines
information across a correlated feature set. The top principal
components therefore represent a robust compression of the dataset.
However, how many components should we choose to include? 10? 20?
100?</p>
<p>In Macosko et al, we implemented a resampling test inspired by the
JackStraw procedure. While still available in Seurat (see previous
vignette), this is a slow and computationally expensive procedure, and
we is no longer routinely used in single cell analysis.</p>
<p>An alternative heuristic method generates an ‘Elbow plot’: a ranking
of principle components based on the percentage of variance explained by
each one (ElbowPlot() function). In this example, we can observe an
‘elbow’ around PC9-10, suggesting that the majority of true signal is
captured in the first 10 PCs.</p>
<pre class="r"><code># plots pbmc@reductions$pca@stdev
ElbowPlot(pbmc)</code></pre>
<p><img src="figure/seurat.Rmd/elbow_plot-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-elbow_plot-1">
Past versions of elbow_plot-1.png
</button>
</p>
<div id="fig-elbow_plot-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/davetang/muse/blob/8188fefc3d1003fd5abbe964b502e6ff138cbb37/docs/figure/seurat.Rmd/elbow_plot-1.png" target="_blank">8188fef</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-04-15
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>Identifying the true dimensionality of a dataset – can be
challenging/uncertain for the user. We therefore suggest these multiple
approaches for users. The first is more supervised, exploring PCs to
determine relevant sources of heterogeneity, and could be used in
conjunction with GSEA for example. The second (ElbowPlot) The third is a
heuristic that is commonly used, and can be calculated instantly. In
this example, we might have been justified in choosing anything between
PC 7-12 as a cutoff.</p>
<p>We chose 10 here, but encourage users to consider the following:</p>
<ul>
<li>Dendritic cell and NK aficionados may recognize that genes strongly
associated with PCs 12 and 13 define rare immune subsets (i.e. MZB1 is a
marker for plasmacytoid DCs). However, these groups are so rare, they
are difficult to distinguish from background noise for a dataset of this
size without prior knowledge.</li>
<li>We encourage users to repeat downstream analyses with a different
number of PCs (10, 15, or even 50!). As you will observe, the results
often do not differ dramatically.</li>
<li>We advise users to err on the higher side when choosing this
parameter. For example, performing downstream analyses with only 5 PCs
does significantly and adversely affect results.</li>
</ul>
</div>
<div id="cluster-the-cells" class="section level2">
<h2>Cluster the cells</h2>
<p>Seurat applies a graph-based clustering approach, building upon
initial strategies in (Macosko et al). Importantly, the distance metric
which drives the clustering analysis (based on previously identified
PCs) remains the same. However, our approach to partitioning the
cellular distance matrix into clusters has dramatically improved. Our
approach was heavily inspired by recent manuscripts which applied
graph-based clustering approaches to scRNA-seq data [SNN-Cliq, Xu and
Su, Bioinformatics, 2015] and CyTOF data [PhenoGraph, Levine et al.,
Cell, 2015]. Briefly, these methods embed cells in a graph structure -
for example a K-nearest neighbor (KNN) graph, with edges drawn between
cells with similar feature expression patterns, and then attempt to
partition this graph into highly interconnected ‘quasi-cliques’ or
‘communities’.</p>
<p>As in PhenoGraph, we first construct a KNN graph based on the
euclidean distance in PCA space, and refine the edge weights between any
two cells based on the shared overlap in their local neighborhoods
(Jaccard similarity). This step is performed using the FindNeighbors()
function, and takes as input the previously defined dimensionality of
the dataset (first 10 PCs).</p>
<p>To cluster the cells, we next apply modularity optimization
techniques such as the Louvain algorithm (default) or SLM [SLM, Blondel
et al., Journal of Statistical Mechanics], to iteratively group cells
together, with the goal of optimizing the standard modularity function.
The FindClusters() function implements this procedure, and contains a
resolution parameter that sets the ‘granularity’ of the downstream
clustering, with increased values leading to a greater number of
clusters. We find that setting this parameter between 0.4-1.2 typically
returns good results for single-cell datasets of around 3K cells.
Optimal resolution often increases for larger datasets. The clusters can
be found using the Idents() function.</p>
<pre class="r"><code>pbmc &lt;- FindNeighbors(pbmc, dims = 1:10)</code></pre>
<pre><code>Computing nearest neighbor graph</code></pre>
<pre><code>Computing SNN</code></pre>
<pre class="r"><code>pbmc &lt;- FindClusters(pbmc, resolution = 0.5)</code></pre>
<pre><code>Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck

Number of nodes: 2638
Number of edges: 95965

Running Louvain algorithm...
Maximum modularity in 10 random starts: 0.8723
Number of communities: 9
Elapsed time: 0 seconds</code></pre>
<p>Look at cluster IDs of the first 5 cells</p>
<pre class="r"><code>head(Idents(pbmc), 5)</code></pre>
<pre><code>AAACATACAACCAC-1 AAACATTGAGCTAC-1 AAACATTGATCAGC-1 AAACCGTGCTTCCG-1 
               2                3                2                1 
AAACCGTGTATGCG-1 
               6 
Levels: 0 1 2 3 4 5 6 7 8</code></pre>
</div>
<div id="non-linear-dimensional-reduction" class="section level2">
<h2>Non-linear dimensional reduction</h2>
<p>Seurat offers several non-linear dimensional reduction techniques,
such as tSNE and UMAP, to visualize and explore these datasets. The goal
of these algorithms is to learn underlying structure in the dataset, in
order to place similar cells together in low-dimensional space.
Therefore, cells that are grouped together within graph-based clusters
determined above should co-localize on these dimension reduction
plots.</p>
<p>While we and others have routinely found 2D visualization techniques
like tSNE and UMAP to be valuable tools for exploring datasets, all
visualization techniques have limitations, and cannot fully represent
the complexity of the underlying data. In particular, these methods aim
to preserve local distances in the dataset (i.e. ensuring that cells
with very similar gene expression profiles co-localize), but often do
not preserve more global relationships. We encourage users to leverage
techniques like UMAP for visualization, but to avoid drawing biological
conclusions solely on the basis of visualization techniques.</p>
<pre class="r"><code>pbmc &lt;- RunUMAP(pbmc, dims = 1:10)</code></pre>
<pre><code>Warning: The default method for RunUMAP has changed from calling Python UMAP via reticulate to the R-native UWOT using the cosine metric
To use Python UMAP via reticulate, set umap.method to &#39;umap-learn&#39; and metric to &#39;correlation&#39;
This message will be shown once per session</code></pre>
<pre><code>09:25:55 UMAP embedding parameters a = 0.9922 b = 1.112</code></pre>
<pre><code>09:25:55 Read 2638 rows and found 10 numeric columns</code></pre>
<pre><code>09:25:55 Using Annoy for neighbor search, n_neighbors = 30</code></pre>
<pre><code>09:25:55 Building Annoy index with metric = cosine, n_trees = 50</code></pre>
<pre><code>0%   10   20   30   40   50   60   70   80   90   100%</code></pre>
<pre><code>[----|----|----|----|----|----|----|----|----|----|</code></pre>
<pre><code>**************************************************|
09:25:55 Writing NN index file to temp file /tmp/RtmpUkyYeG/file13c750c4354b
09:25:55 Searching Annoy index using 1 thread, search_k = 3000
09:25:56 Annoy recall = 100%
09:25:56 Commencing smooth kNN distance calibration using 1 thread with target n_neighbors = 30
09:25:57 Initializing from normalized Laplacian + noise (using RSpectra)
09:25:57 Commencing optimization for 500 epochs, with 105124 positive edges
09:25:59 Optimization finished</code></pre>
<p>Note that you can set <code>label = TRUE</code> or use the
LabelClusters function to help label individual clusters</p>
<pre class="r"><code>DimPlot(pbmc, reduction = &quot;umap&quot;)</code></pre>
<p><img src="figure/seurat.Rmd/unnamed-chunk-46-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-46-1">
Past versions of unnamed-chunk-46-1.png
</button>
</p>
<div id="fig-unnamed-chunk-46-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/davetang/muse/blob/8188fefc3d1003fd5abbe964b502e6ff138cbb37/docs/figure/seurat.Rmd/unnamed-chunk-46-1.png" target="_blank">8188fef</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-04-15
</td>
</tr>
<tr>
<td>
<a href="https://github.com/davetang/muse/blob/5f6b5a75f1dfa606244f52e6a9963e2bdfef2cc3/docs/figure/seurat.Rmd/unnamed-chunk-46-1.png" target="_blank">5f6b5a7</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-04-07
</td>
</tr>
<tr>
<td>
<a href="https://github.com/davetang/muse/blob/03bb7a1bd974a1557d1f99682950668bc9d23b48/docs/figure/seurat.Rmd/unnamed-chunk-46-1.png" target="_blank">03bb7a1</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-03-20
</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<div id="finding-differentially-expressed-features"
class="section level2">
<h2>Finding differentially expressed features</h2>
<p>Seurat can help you find markers that define clusters via
differential expression (DE). By default, it identifies positive and
negative markers of a single cluster (specified in ident.1), compared to
all other cells. FindAllMarkers() automates this process for all
clusters, but you can also test groups of clusters vs. each other, or
against all cells.</p>
<p>In Seurat v5, we use the presto package (as described here and
available for installation here), to dramatically improve the speed of
DE analysis, particularly for large datasets. For users who are not
using presto, you can examine the documentation for this function
(?FindMarkers) to explore the min.pct and logfc.threshold parameters,
which can be increased in order to increase the speed of DE testing.</p>
<p>For a (much!) faster implementation of the Wilcoxon Rank Sum Test,
(default method for FindMarkers) please install the {presto} package.
After installation of {presto}, Seurat will automatically use the more
efficient implementation (no further action necessary).</p>
<pre class="r"><code>install.packages(&quot;remotes&quot;)
remotes::install_github(&#39;immunogenomics/presto&#39;)</code></pre>
<p>Load {presto}.</p>
<pre class="r"><code>library(&#39;presto&#39;)
packageVersion(&#39;presto&#39;)</code></pre>
<pre><code>[1] &#39;1.0.0&#39;</code></pre>
<p>Find all markers of cluster 2.</p>
<pre class="r"><code>cluster2.markers &lt;- FindMarkers(pbmc, ident.1 = 2)
head(cluster2.markers, n = 5)</code></pre>
<pre><code>            p_val avg_log2FC pct.1 pct.2    p_val_adj
IL32 2.593535e-91  1.3221171 0.949 0.466 3.556774e-87
LTB  7.994465e-87  1.3450377 0.981 0.644 1.096361e-82
CD3D 3.922451e-70  1.0562099 0.922 0.433 5.379250e-66
IL7R 1.130870e-66  1.4256944 0.748 0.327 1.550876e-62
LDHB 4.082189e-65  0.9765875 0.953 0.614 5.598314e-61</code></pre>
<p>Find all markers distinguishing cluster 5 from clusters 0 and 3</p>
<pre class="r"><code>cluster5.markers &lt;- FindMarkers(pbmc, ident.1 = 5, ident.2 = c(0, 3))
head(cluster5.markers, n = 5)</code></pre>
<pre><code>                      p_val avg_log2FC pct.1 pct.2     p_val_adj
FCGR3A        2.150929e-209   6.832372 0.975 0.039 2.949784e-205
IFITM3        6.103366e-199   6.181000 0.975 0.048 8.370156e-195
CFD           8.891428e-198   6.052575 0.938 0.037 1.219370e-193
CD68          2.374425e-194   5.493138 0.926 0.035 3.256286e-190
RP11-290F20.3 9.308287e-191   6.335402 0.840 0.016 1.276538e-186</code></pre>
<p>Find markers for every cluster compared to all remaining cells,
report only the positive ones.</p>
<pre class="r"><code>pbmc.markers &lt;- FindAllMarkers(pbmc, only.pos = TRUE)
pbmc.markers %&gt;%
    dplyr::group_by(cluster) %&gt;%
    dplyr::filter(avg_log2FC &gt; 1)</code></pre>
<pre><code># A tibble: 7,046 × 7
# Groups:   cluster [9]
       p_val avg_log2FC pct.1 pct.2 p_val_adj cluster gene     
       &lt;dbl&gt;      &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;     &lt;dbl&gt; &lt;fct&gt;   &lt;chr&gt;    
 1 1.74e-109       1.19 0.897 0.593 2.39e-105 0       LDHB     
 2 1.17e- 83       2.37 0.435 0.108 1.60e- 79 0       CCR7     
 3 8.94e- 79       1.09 0.838 0.403 1.23e- 74 0       CD3D     
 4 3.05e- 53       1.02 0.722 0.399 4.19e- 49 0       CD3E     
 5 3.28e- 49       2.10 0.333 0.103 4.50e- 45 0       LEF1     
 6 6.66e- 49       1.25 0.623 0.358 9.13e- 45 0       NOSIP    
 7 9.31e- 44       2.02 0.328 0.11  1.28e- 39 0       PRKCQ-AS1
 8 4.69e- 43       1.53 0.435 0.184 6.43e- 39 0       PIK3IP1  
 9 1.47e- 39       2.70 0.195 0.04  2.01e- 35 0       FHIT     
10 2.44e- 33       1.94 0.262 0.087 3.34e- 29 0       MAL      
# ℹ 7,036 more rows</code></pre>
<p>Seurat has several tests for differential expression which can be set
with the test.use parameter (see our DE vignette for details). For
example, the ROC test returns the ‘classification power’ for any
individual marker (ranging from 0 - random, to 1 - perfect).</p>
<pre class="r"><code>cluster0.markers &lt;- FindMarkers(pbmc, ident.1 = 0, logfc.threshold = 0.25, test.use = &quot;roc&quot;, only.pos = TRUE)</code></pre>
<p>We include several tools for visualizing marker expression.
<code>VlnPlot()</code> (shows expression probability distributions
across clusters), and <code>FeaturePlot()</code> (visualizes feature
expression on a tSNE or PCA plot) are our most commonly used
visualizations. We also suggest exploring <code>RidgePlot()</code>,
<code>CellScatter()</code>, and <code>DotPlot()</code> as additional
methods to view your dataset.</p>
<pre class="r"><code>VlnPlot(pbmc, features = c(&quot;MS4A1&quot;, &quot;CD79A&quot;))</code></pre>
<p><img src="figure/seurat.Rmd/unnamed-chunk-52-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-52-1">
Past versions of unnamed-chunk-52-1.png
</button>
</p>
<div id="fig-unnamed-chunk-52-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/davetang/muse/blob/8188fefc3d1003fd5abbe964b502e6ff138cbb37/docs/figure/seurat.Rmd/unnamed-chunk-52-1.png" target="_blank">8188fef</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-04-15
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>You can plot raw counts as well.</p>
<pre class="r"><code>VlnPlot(pbmc, features = c(&quot;NKG7&quot;, &quot;PF4&quot;), layer = &quot;counts&quot;, log = TRUE)</code></pre>
<p><img src="figure/seurat.Rmd/unnamed-chunk-53-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-53-1">
Past versions of unnamed-chunk-53-1.png
</button>
</p>
<div id="fig-unnamed-chunk-53-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/davetang/muse/blob/8188fefc3d1003fd5abbe964b502e6ff138cbb37/docs/figure/seurat.Rmd/unnamed-chunk-53-1.png" target="_blank">8188fef</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-04-15
</td>
</tr>
<tr>
<td>
<a href="https://github.com/davetang/muse/blob/b1853704502f2dbd42a59d77f9aef121ab4fab46/docs/figure/seurat.Rmd/unnamed-chunk-53-1.png" target="_blank">b185370</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-04-15
</td>
</tr>
<tr>
<td>
<a href="https://github.com/davetang/muse/blob/5f6b5a75f1dfa606244f52e6a9963e2bdfef2cc3/docs/figure/seurat.Rmd/unnamed-chunk-53-1.png" target="_blank">5f6b5a7</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-04-07
</td>
</tr>
<tr>
<td>
<a href="https://github.com/davetang/muse/blob/03bb7a1bd974a1557d1f99682950668bc9d23b48/docs/figure/seurat.Rmd/unnamed-chunk-53-1.png" target="_blank">03bb7a1</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-03-20
</td>
</tr>
</tbody>
</table>
</div>
</div>
<pre class="r"><code>FeaturePlot(pbmc, features = c(&quot;MS4A1&quot;, &quot;GNLY&quot;, &quot;CD3E&quot;, &quot;CD14&quot;, &quot;FCER1A&quot;, &quot;FCGR3A&quot;, &quot;LYZ&quot;, &quot;PPBP&quot;,
    &quot;CD8A&quot;))</code></pre>
<p><img src="figure/seurat.Rmd/unnamed-chunk-54-1.png" width="960" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-54-1">
Past versions of unnamed-chunk-54-1.png
</button>
</p>
<div id="fig-unnamed-chunk-54-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/davetang/muse/blob/8188fefc3d1003fd5abbe964b502e6ff138cbb37/docs/figure/seurat.Rmd/unnamed-chunk-54-1.png" target="_blank">8188fef</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-04-15
</td>
</tr>
<tr>
<td>
<a href="https://github.com/davetang/muse/blob/b1853704502f2dbd42a59d77f9aef121ab4fab46/docs/figure/seurat.Rmd/unnamed-chunk-54-1.png" target="_blank">b185370</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-04-15
</td>
</tr>
<tr>
<td>
<a href="https://github.com/davetang/muse/blob/5f6b5a75f1dfa606244f52e6a9963e2bdfef2cc3/docs/figure/seurat.Rmd/unnamed-chunk-54-1.png" target="_blank">5f6b5a7</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-04-07
</td>
</tr>
<tr>
<td>
<a href="https://github.com/davetang/muse/blob/03bb7a1bd974a1557d1f99682950668bc9d23b48/docs/figure/seurat.Rmd/unnamed-chunk-54-1.png" target="_blank">03bb7a1</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-03-20
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p><code>DoHeatmap()</code> generates an expression heatmap for given
cells and features. In this case, we are plotting the top 20 markers (or
all markers if less than 20) for each cluster.</p>
<pre class="r"><code>pbmc.markers %&gt;%
    dplyr::group_by(cluster) %&gt;%
    dplyr::filter(avg_log2FC &gt; 1) %&gt;%
    dplyr::slice_head(n = 10) %&gt;%
    dplyr::ungroup() -&gt; top10

DoHeatmap(pbmc, features = top10$gene) + NoLegend()</code></pre>
<p><img src="figure/seurat.Rmd/unnamed-chunk-55-1.png" width="1152" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-55-1">
Past versions of unnamed-chunk-55-1.png
</button>
</p>
<div id="fig-unnamed-chunk-55-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/davetang/muse/blob/8188fefc3d1003fd5abbe964b502e6ff138cbb37/docs/figure/seurat.Rmd/unnamed-chunk-55-1.png" target="_blank">8188fef</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-04-15
</td>
</tr>
<tr>
<td>
<a href="https://github.com/davetang/muse/blob/b1853704502f2dbd42a59d77f9aef121ab4fab46/docs/figure/seurat.Rmd/unnamed-chunk-55-1.png" target="_blank">b185370</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-04-15
</td>
</tr>
<tr>
<td>
<a href="https://github.com/davetang/muse/blob/5f6b5a75f1dfa606244f52e6a9963e2bdfef2cc3/docs/figure/seurat.Rmd/unnamed-chunk-55-1.png" target="_blank">5f6b5a7</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-04-07
</td>
</tr>
<tr>
<td>
<a href="https://github.com/davetang/muse/blob/03bb7a1bd974a1557d1f99682950668bc9d23b48/docs/figure/seurat.Rmd/unnamed-chunk-55-1.png" target="_blank">03bb7a1</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-03-20
</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<div id="assigning-cell-type-identity-to-clusters"
class="section level2">
<h2>Assigning cell type identity to clusters</h2>
<p>Fortunately in the case of this dataset, we can use canonical markers
to easily match the unbiased clustering to known cell types:</p>
<table>
<thead>
<tr class="header">
<th>Cluster ID</th>
<th>Markers</th>
<th>Cell Type</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>0</td>
<td>IL7R, CCR7</td>
<td>Naive CD4+ T</td>
</tr>
<tr class="even">
<td>1</td>
<td>CD14, LYZ</td>
<td>CD14+ Mono</td>
</tr>
<tr class="odd">
<td>2</td>
<td>IL7R, S100A4</td>
<td>Memory CD4+</td>
</tr>
<tr class="even">
<td>3</td>
<td>MS4A1</td>
<td>B</td>
</tr>
<tr class="odd">
<td>4</td>
<td>CD8A</td>
<td>CD8+ T</td>
</tr>
<tr class="even">
<td>5</td>
<td>FCGR3A, MS4A7</td>
<td>FCGR3A+ Mono</td>
</tr>
<tr class="odd">
<td>6</td>
<td>GNLY, NKG7</td>
<td>NK</td>
</tr>
<tr class="even">
<td>7</td>
<td>FCER1A, CST3</td>
<td>DC</td>
</tr>
<tr class="odd">
<td>8</td>
<td>PPBP</td>
<td>Platelet</td>
</tr>
</tbody>
</table>
<pre class="r"><code>new.cluster.ids &lt;- c(&quot;Naive CD4 T&quot;, &quot;CD14+ Mono&quot;, &quot;Memory CD4 T&quot;, &quot;B&quot;, &quot;CD8 T&quot;, &quot;FCGR3A+ Mono&quot;,
    &quot;NK&quot;, &quot;DC&quot;, &quot;Platelet&quot;)
names(new.cluster.ids) &lt;- levels(pbmc)
pbmc &lt;- RenameIdents(pbmc, new.cluster.ids)
DimPlot(pbmc, reduction = &quot;umap&quot;, label = TRUE, pt.size = 0.5) + NoLegend()</code></pre>
<p><img src="figure/seurat.Rmd/unnamed-chunk-56-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-56-1">
Past versions of unnamed-chunk-56-1.png
</button>
</p>
<div id="fig-unnamed-chunk-56-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/davetang/muse/blob/8188fefc3d1003fd5abbe964b502e6ff138cbb37/docs/figure/seurat.Rmd/unnamed-chunk-56-1.png" target="_blank">8188fef</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-04-15
</td>
</tr>
<tr>
<td>
<a href="https://github.com/davetang/muse/blob/b1853704502f2dbd42a59d77f9aef121ab4fab46/docs/figure/seurat.Rmd/unnamed-chunk-56-1.png" target="_blank">b185370</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-04-15
</td>
</tr>
<tr>
<td>
<a href="https://github.com/davetang/muse/blob/5f6b5a75f1dfa606244f52e6a9963e2bdfef2cc3/docs/figure/seurat.Rmd/unnamed-chunk-56-1.png" target="_blank">5f6b5a7</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-04-07
</td>
</tr>
<tr>
<td>
<a href="https://github.com/davetang/muse/blob/03bb7a1bd974a1557d1f99682950668bc9d23b48/docs/figure/seurat.Rmd/unnamed-chunk-56-1.png" target="_blank">03bb7a1</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-03-20
</td>
</tr>
</tbody>
</table>
</div>
</div>
<pre class="r"><code>library(ggplot2)
DimPlot(pbmc, reduction = &quot;umap&quot;, label = TRUE, label.size = 4.5) + xlab(&quot;UMAP 1&quot;) +
  ylab(&quot;UMAP 2&quot;) +
  theme(axis.title = element_text(size = 18), legend.text = element_text(size = 18)) +
  guides(colour = guide_legend(override.aes = list(size = 10)))</code></pre>
<p><img src="figure/seurat.Rmd/unnamed-chunk-57-1.png" width="1152" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-57-1">
Past versions of unnamed-chunk-57-1.png
</button>
</p>
<div id="fig-unnamed-chunk-57-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/davetang/muse/blob/8188fefc3d1003fd5abbe964b502e6ff138cbb37/docs/figure/seurat.Rmd/unnamed-chunk-57-1.png" target="_blank">8188fef</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-04-15
</td>
</tr>
<tr>
<td>
<a href="https://github.com/davetang/muse/blob/b1853704502f2dbd42a59d77f9aef121ab4fab46/docs/figure/seurat.Rmd/unnamed-chunk-57-1.png" target="_blank">b185370</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-04-15
</td>
</tr>
<tr>
<td>
<a href="https://github.com/davetang/muse/blob/5f6b5a75f1dfa606244f52e6a9963e2bdfef2cc3/docs/figure/seurat.Rmd/unnamed-chunk-57-1.png" target="_blank">5f6b5a7</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-04-07
</td>
</tr>
<tr>
<td>
<a href="https://github.com/davetang/muse/blob/03bb7a1bd974a1557d1f99682950668bc9d23b48/docs/figure/seurat.Rmd/unnamed-chunk-57-1.png" target="_blank">03bb7a1</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-03-20
</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<div id="from-anndata" class="section level2">
<h2>From AnnData</h2>
<p>Download <a
href="https://openproblems.bio/datasets/openproblems_v1/tenx_1k_pbmc/">1k
PBMCs</a> in AnnData format from <a
href="https://openproblems.bio/">Open Problems in Single-Cell
Analysis</a>.</p>
<pre><code>aws s3 cp --no-sign-request s3://openproblems-data/resources/datasets/openproblems_v1/tenx_1k_pbmc/l1_sqrt/dataset.h5ad .</code></pre>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span>
Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 4.4.0 (2024-04-24)
Platform: x86_64-pc-linux-gnu
Running under: Ubuntu 22.04.4 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 
LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

time zone: Etc/UTC
tzcode source: system (glibc)

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] ggplot2_3.5.1      presto_1.0.0       data.table_1.15.4  Rcpp_1.0.12       
[5] bench_1.1.3        Seurat_5.1.0       SeuratObject_5.0.2 sp_2.1-4          
[9] workflowr_1.7.1   

loaded via a namespace (and not attached):
  [1] RColorBrewer_1.1-3     rstudioapi_0.16.0      jsonlite_1.8.8        
  [4] magrittr_2.0.3         spatstat.utils_3.0-4   farver_2.1.2          
  [7] rmarkdown_2.27         fs_1.6.4               vctrs_0.6.5           
 [10] ROCR_1.0-11            spatstat.explore_3.2-7 htmltools_0.5.8.1     
 [13] sass_0.4.9             sctransform_0.4.1      parallelly_1.37.1     
 [16] KernSmooth_2.23-22     bslib_0.7.0            htmlwidgets_1.6.4     
 [19] ica_1.0-3              plyr_1.8.9             plotly_4.10.4         
 [22] zoo_1.8-12             cachem_1.1.0           whisker_0.4.1         
 [25] igraph_2.0.3           mime_0.12              lifecycle_1.0.4       
 [28] pkgconfig_2.0.3        Matrix_1.7-0           R6_2.5.1              
 [31] fastmap_1.2.0          fitdistrplus_1.1-11    future_1.33.2         
 [34] shiny_1.8.1.1          digest_0.6.37          colorspace_2.1-0      
 [37] patchwork_1.2.0        ps_1.7.6               rprojroot_2.0.4       
 [40] tensor_1.5             RSpectra_0.16-1        irlba_2.3.5.1         
 [43] labeling_0.4.3         progressr_0.14.0       fansi_1.0.6           
 [46] spatstat.sparse_3.0-3  httr_1.4.7             polyclip_1.10-6       
 [49] abind_1.4-5            compiler_4.4.0         withr_3.0.1           
 [52] fastDummies_1.7.3      highr_0.11             R.utils_2.12.3        
 [55] MASS_7.3-60.2          tools_4.4.0            lmtest_0.9-40         
 [58] httpuv_1.6.15          future.apply_1.11.2    goftest_1.2-3         
 [61] R.oo_1.26.0            glue_1.7.0             callr_3.7.6           
 [64] profmem_0.6.0          nlme_3.1-164           promises_1.3.0        
 [67] grid_4.4.0             Rtsne_0.17             getPass_0.2-4         
 [70] cluster_2.1.6          reshape2_1.4.4         generics_0.1.3        
 [73] gtable_0.3.5           spatstat.data_3.0-4    R.methodsS3_1.8.2     
 [76] tidyr_1.3.1            utf8_1.2.4             spatstat.geom_3.2-9   
 [79] RcppAnnoy_0.0.22       ggrepel_0.9.5          RANN_2.6.1            
 [82] pillar_1.9.0           stringr_1.5.1          limma_3.60.4          
 [85] spam_2.10-0            RcppHNSW_0.6.0         later_1.3.2           
 [88] splines_4.4.0          dplyr_1.1.4            lattice_0.22-6        
 [91] survival_3.5-8         deldir_2.0-4           tidyselect_1.2.1      
 [94] miniUI_0.1.1.1         pbapply_1.7-2          knitr_1.47            
 [97] git2r_0.33.0           gridExtra_2.3          scattermore_1.2       
[100] xfun_0.44              statmod_1.5.0          matrixStats_1.3.0     
[103] stringi_1.8.4          lazyeval_0.2.2         yaml_2.3.8            
[106] evaluate_0.24.0        codetools_0.2-20       tibble_3.2.1          
[109] cli_3.6.3              uwot_0.2.2             xtable_1.8-4          
[112] reticulate_1.37.0      munsell_0.5.1          processx_3.8.4        
[115] jquerylib_0.1.4        globals_0.16.3         spatstat.random_3.2-3 
[118] png_0.1-8              parallel_4.4.0         dotCall64_1.1-1       
[121] listenv_0.9.1          viridisLite_0.4.2      scales_1.3.0          
[124] ggridges_0.5.6         leiden_0.4.3.1         purrr_1.0.2           
[127] rlang_1.1.4            cowplot_1.1.3         </code></pre>
</div>
</div>
</div>
</div>
</div>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
https://docs.mathjax.org/en/latest/web/configuration.html. This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>




</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
