<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />



<meta name="date" content="2025-02-12" />

<title>Using BPCells with Seurat</title>

<script src="site_libs/header-attrs-2.28/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.13.2/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-6.4.2/css/all.min.css" rel="stylesheet" />
<link href="site_libs/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet" />

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/main/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">My blog</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/davetang/muse">
    <span class="fa fa-github"></span>
     
    Source code
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Using BPCells with Seurat</h1>
<h4 class="date">2025-02-12</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span>
workflowr <span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2025-02-12
</p>
<p>
<strong>Checks:</strong> <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 7
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong> <code>muse/</code> <span
class="glyphicon glyphicon-question-sign" aria-hidden="true"
title="This is the local directory in which the code in this file was executed.">
</span>
</p>
<p>
This reproducible <a href="https://rmarkdown.rstudio.com">R Markdown</a>
analysis was created with <a
  href="https://github.com/workflowr/workflowr">workflowr</a> (version
1.7.1). The <em>Checks</em> tab describes the reproducibility checks
that were applied when the results were created. The <em>Past
versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date
</a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git
repository, you know the exact version of the code that produced these
results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the
global environment can affect the analysis in your R Markdown file in
unknown ways. For reproduciblity it’s best to always run the code in an
empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20200712code">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Seed:</strong>
<code>set.seed(20200712)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20200712code"
class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20200712)</code> was run prior to running
the code in the R Markdown file. Setting a seed ensures that any results
that rely on randomness, e.g. subsampling or permutations, are
reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Session information:</strong>
recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package
versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be
confident that you successfully produced the results during this
run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr
project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomdavetangmusetreefa1564d72fc18e9b1b8337dedec5f291bd57faaftargetblankfa1564da">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Repository version:</strong>
<a href="https://github.com/davetang/muse/tree/fa1564d72fc18e9b1b8337dedec5f291bd57faaf" target="_blank">fa1564d</a>
</a>
</p>
</div>
<div
id="strongRepositoryversionstrongahrefhttpsgithubcomdavetangmusetreefa1564d72fc18e9b1b8337dedec5f291bd57faaftargetblankfa1564da"
class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development
and connecting the code version to the results is critical for
reproducibility.
</p>
<p>
The results in this page were generated with repository version
<a href="https://github.com/davetang/muse/tree/fa1564d72fc18e9b1b8337dedec5f291bd57faaf" target="_blank">fa1564d</a>.
See the <em>Past versions</em> tab to see a history of the changes made
to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for
the analysis have been committed to Git prior to generating the results
(you can use <code>wflow_publish</code> or
<code>wflow_git_commit</code>). workflowr only checks the R Markdown
file, but you know if there are other scripts or data files that it
depends on. Below is the status of the Git repository when the results
were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .Rproj.user/
    Ignored:    r_packages_4.4.1/

Untracked files:
    Untracked:  data/1M_neurons_filtered_gene_bc_matrices_h5.h5
    Untracked:  data/brain_counts/
    Untracked:  data/seurat_1m_neuron.rds

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not
included in this status report because it is ok for generated content to
have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">
<p>
These are the previous versions of the repository in which changes were
made to the R Markdown (<code>analysis/seurat_bpcells.Rmd</code>) and
HTML (<code>docs/seurat_bpcells.html</code>) files. If you’ve configured
a remote Git repository (see <code>?wflow_git_remote</code>), click on
the hyperlinks in the table below to view the files as they were in that
past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/davetang/muse/blob/fa1564d72fc18e9b1b8337dedec5f291bd57faaf/analysis/seurat_bpcells.Rmd" target="_blank">fa1564d</a>
</td>
<td>
Dave Tang
</td>
<td>
2025-02-12
</td>
<td>
BPCells with Seurat
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<p><a
href="https://github.com/satijalab/seurat/blob/9755c164d99828dbc5dd9c8364389766cd4ff7fd/vignettes/seurat5_bpcells_interaction_vignette.Rmd"
class="uri">https://github.com/satijalab/seurat/blob/9755c164d99828dbc5dd9c8364389766cd4ff7fd/vignettes/seurat5_bpcells_interaction_vignette.Rmd</a></p>
<p>BPCells is an <a href="https://github.com/bnprks/BPCells">R
package</a> that allows for computationally efficient single-cell
analysis. It utilizes bit-packing compression to store counts matrices
on disk and C++ code to cache operations.</p>
<p>We leverage the high performance capabilities of BPCells to work with
Seurat objects in memory while accessing the counts on disk. In this
vignette, we show how to use BPCells to load data, work with a Seurat
objects in a more memory-efficient way, and write out Seurat objects
with BPCells matrices.</p>
<p>We will show the methods for interacting with both a single dataset
in one file or multiple datasets across multiple files using BPCells.
BPCells allows us to easily analyze these large datasets in memory, and
we encourage users to check out some of our other vignettes <a
href="">here</a> and <a href="">here</a> to see further
applications.</p>
<pre class="r"><code>remotes::install_github(&quot;bnprks/BPCells/r&quot;)</code></pre>
<pre class="r"><code>suppressPackageStartupMessages(library(BPCells))
suppressPackageStartupMessages(library(Seurat))</code></pre>
<p>We use BPCells functionality to both load in our data and write the
counts layers to bitpacked compressed binary files on disk to improve
computation speeds. BPCells has multiple functions for reading in
files.</p>
<div id="load-data" class="section level1">
<h1>Load Data</h1>
<p>Download <a
href="https://support.10xgenomics.com/single-cell-gene-expression/datasets/1.3.0/1M_neurons?">1.3
Million Brain Cells from E18 Mice</a> (3.93 GB).</p>
<pre class="r"><code>my_url &lt;- &#39;https://cf.10xgenomics.com/samples/cell-exp/1.3.0/1M_neurons/1M_neurons_filtered_gene_bc_matrices_h5.h5&#39;
my_file &lt;- paste0(&quot;data/&quot;, basename(my_url))

if(!file.exists(my_file)){
  options(timeout = 10000)
  download.file(url = my_url, destfile = my_file)
}</code></pre>
<div id="load-data-from-one-h5-file" class="section level2">
<h2>Load Data from one h5 file</h2>
<p>In this section, we will load the 1.3 million brain cells dataset. We
will use <code>BPCells::open_matrix_10x_hdf5()</code> that reads in
feature matrices from 10x data. We then write a matrix directory, load
the matrix, and create a Seurat object.</p>
<pre class="r"><code>brain.data &lt;- BPCells::open_matrix_10x_hdf5(path = my_file)
brain.data</code></pre>
<pre><code>27998 x 1306127 IterableMatrix object with class 10xMatrixH5

Row names: ENSMUSG00000051951, ENSMUSG00000089699 ... ENSMUSG00000095742
Col names: AAACCTGAGATAGGAG-1, AAACCTGAGCGGCTTC-1 ... TTTGTCATCTGAAAGA-133

Data type: uint32_t
Storage order: column major

Queued Operations:
1. 10x HDF5 feature matrix in file /home/rstudio/muse/data/1M_neurons_filtered_gene_bc_matrices_h5.h5</code></pre>
<pre class="r"><code># Write the matrix to a directory
my_outdir &lt;- &quot;data/brain_counts&quot;
if(!dir.exists(my_outdir)){
  BPCells::write_matrix_dir(
    mat = brain.data,
    dir = my_outdir
  )
}

# Now that we have the matrix on disk, we can load it
brain.mat &lt;- open_matrix_dir(dir = my_outdir)
brain.mat</code></pre>
<pre><code>27998 x 1306127 IterableMatrix object with class MatrixDir

Row names: ENSMUSG00000051951, ENSMUSG00000089699 ... ENSMUSG00000095742
Col names: AAACCTGAGATAGGAG-1, AAACCTGAGCGGCTTC-1 ... TTTGTCATCTGAAAGA-133

Data type: uint32_t
Storage order: column major

Queued Operations:
1. Load compressed matrix from directory /home/rstudio/muse/data/brain_counts</code></pre>
<pre class="r"><code># Create Seurat Object
brain &lt;- CreateSeuratObject(
  counts = brain.mat,
  project = &#39;1m_neurons&#39;
)
brain</code></pre>
<pre><code>An object of class Seurat 
27998 features across 1306127 samples within 1 assay 
Active assay: RNA (27998 features, 0 variable features)
 1 layer present: counts</code></pre>
<p><strong>What if I already have a Seurat Object?</strong></p>
<p>You can use BPCells to convert the matrices in your already created
Seurat objects to on-disk matrices. Note, that this is only possible for
V5 assays. As an example, if you’d like to convert the counts matrix of
your RNA assay to a BPCells matrix, you can use the following:</p>
<pre class="r"><code>obj &lt;- readRDS(&quot;/path/to/reference.rds&quot;)

# Write the counts layer to a directory
write_matrix_dir(mat = obj[[&quot;RNA&quot;]]$counts, dir = &#39;/brahms/hartmana/vignette_data/bpcells/brain_counts&#39;)
counts.mat &lt;- open_matrix_dir(dir = &quot;/brahms/hartmana/vignette_data/bpcells/brain_counts&quot;)

obj[[&quot;RNA&quot;]]$counts &lt;- counts.mat</code></pre>
<div id="example-analysis" class="section level3">
<h3>Example Analysis</h3>
<p>Once this conversion is done, you can perform typical Seurat
functions on the object. For example, we can normalize data and
visualize features by automatically accessing the on-disk counts.</p>
<pre class="r"><code>brain &lt;- NormalizeData(brain, normalization.method = &quot;LogNormalize&quot;)</code></pre>
<pre><code>Normalizing layer: counts</code></pre>
<pre class="r"><code>brain</code></pre>
<pre><code>An object of class Seurat 
27998 features across 1306127 samples within 1 assay 
Active assay: RNA (27998 features, 0 variable features)
 2 layers present: counts, data</code></pre>
</div>
<div id="saving-seurat-objects-with-on-disk-layers"
class="section level3">
<h3>Saving Seurat objects with on-disk layers</h3>
<p>If you save your object and load it in in the future, Seurat will
access the on-disk matrices by their path, which is stored in the assay
level data. To make it easy to ensure these are saved in the same place,
we provide new functionality to the <code>SaveSeuratRds()</code>
function. In this function, you specify your filename. The pointer to
the path in the Seurat object will change to the current directory.</p>
<p>This also makes it easy to share your Seurat objects with BPCells
matrices by sharing a folder that contains both the object and the
BPCells directory.</p>
<pre class="r"><code>SaveSeuratRds(
  object = brain,
  file = &quot;data/seurat_1m_neuron.rds&quot;
)</code></pre>
<pre><code>Warning: Trying to move &#39;/home/rstudio/muse/data/brain_counts&#39; to itself,
skipping
Trying to move &#39;/home/rstudio/muse/data/brain_counts&#39; to itself,
skipping</code></pre>
<p>If needed, a layer with an on-disk matrix can be converted to an
in-memory matrix using the <code>as()</code> function. For the purposes
of this demo, we’ll subset the object so that it takes up less space in
memory.</p>
<pre class="r"><code>brain_subset &lt;- subset(brain, downsample = 1000)
brain_subset[[&quot;RNA&quot;]]$counts &lt;- as(object = brain_subset[[&quot;RNA&quot;]]$counts, Class = &quot;dgCMatrix&quot;)
brain_subset</code></pre>
<pre><code>An object of class Seurat 
27998 features across 1000 samples within 1 assay 
Active assay: RNA (27998 features, 0 variable features)
 2 layers present: counts, data</code></pre>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span>
Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 4.4.1 (2024-06-14)
Platform: x86_64-pc-linux-gnu
Running under: Ubuntu 22.04.5 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 
LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

time zone: Etc/UTC
tzcode source: system (glibc)

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] Seurat_5.1.0       SeuratObject_5.0.2 sp_2.1-4           BPCells_0.3.0     
[5] workflowr_1.7.1   

loaded via a namespace (and not attached):
  [1] deldir_2.0-4           pbapply_1.7-2          gridExtra_2.3         
  [4] rlang_1.1.4            magrittr_2.0.3         git2r_0.35.0          
  [7] RcppAnnoy_0.0.22       spatstat.geom_3.3-3    matrixStats_1.4.1     
 [10] ggridges_0.5.6         compiler_4.4.1         getPass_0.2-4         
 [13] reshape2_1.4.4         png_0.1-8              callr_3.7.6           
 [16] vctrs_0.6.5            stringr_1.5.1          pkgconfig_2.0.3       
 [19] fastmap_1.2.0          utf8_1.2.4             promises_1.3.0        
 [22] rmarkdown_2.28         ps_1.8.1               purrr_1.0.2           
 [25] xfun_0.48              cachem_1.1.0           jsonlite_1.8.9        
 [28] goftest_1.2-3          later_1.3.2            spatstat.utils_3.1-0  
 [31] irlba_2.3.5.1          parallel_4.4.1         cluster_2.1.6         
 [34] R6_2.5.1               ica_1.0-3              spatstat.data_3.1-2   
 [37] bslib_0.8.0            stringi_1.8.4          RColorBrewer_1.1-3    
 [40] reticulate_1.39.0      spatstat.univar_3.0-1  parallelly_1.38.0     
 [43] scattermore_1.2        lmtest_0.9-40          jquerylib_0.1.4       
 [46] Rcpp_1.0.13            knitr_1.48             tensor_1.5            
 [49] future.apply_1.11.3    zoo_1.8-12             sctransform_0.4.1     
 [52] httpuv_1.6.15          Matrix_1.7-0           splines_4.4.1         
 [55] igraph_2.1.1           tidyselect_1.2.1       abind_1.4-8           
 [58] rstudioapi_0.17.1      yaml_2.3.10            spatstat.random_3.3-2 
 [61] spatstat.explore_3.3-3 codetools_0.2-20       miniUI_0.1.1.1        
 [64] processx_3.8.4         listenv_0.9.1          plyr_1.8.9            
 [67] lattice_0.22-6         tibble_3.2.1           shiny_1.9.1           
 [70] ROCR_1.0-11            evaluate_1.0.1         Rtsne_0.17            
 [73] future_1.34.0          fastDummies_1.7.4      survival_3.6-4        
 [76] polyclip_1.10-7        fitdistrplus_1.2-1     pillar_1.9.0          
 [79] whisker_0.4.1          KernSmooth_2.23-24     plotly_4.10.4         
 [82] generics_0.1.3         rprojroot_2.0.4        RcppHNSW_0.6.0        
 [85] ggplot2_3.5.1          munsell_0.5.1          scales_1.3.0          
 [88] globals_0.16.3         xtable_1.8-4           glue_1.8.0            
 [91] lazyeval_0.2.2         tools_4.4.1            data.table_1.16.2     
 [94] RSpectra_0.16-2        RANN_2.6.2             fs_1.6.4              
 [97] leiden_0.4.3.1         dotCall64_1.2          cowplot_1.1.3         
[100] grid_4.4.1             tidyr_1.3.1            colorspace_2.1-1      
[103] nlme_3.1-164           patchwork_1.3.0        cli_3.6.3             
[106] spatstat.sparse_3.1-0  spam_2.11-0            fansi_1.0.6           
[109] viridisLite_0.4.2      dplyr_1.1.4            uwot_0.2.2            
[112] gtable_0.3.6           sass_0.4.9             digest_0.6.37         
[115] progressr_0.15.0       ggrepel_0.9.6          htmlwidgets_1.6.4     
[118] farver_2.1.2           htmltools_0.5.8.1      lifecycle_1.0.4       
[121] httr_1.4.7             mime_0.12              MASS_7.3-60.2         </code></pre>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
https://docs.mathjax.org/en/latest/web/configuration.html. This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>




</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
