<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />



<meta name="date" content="2024-03-20" />

<title>Getting started with Seurat</title>

<script src="site_libs/header-attrs-2.26/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.13.2/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-6.4.2/css/all.min.css" rel="stylesheet" />
<link href="site_libs/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet" />

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/main/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">My blog</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/davetang/muse">
    <span class="fa fa-github"></span>
     
    Source code
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Getting started with Seurat</h1>
<h4 class="date">2024-03-20</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span>
workflowr <span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2024-03-20
</p>
<p>
<strong>Checks:</strong> <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 7
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong> <code>muse/</code> <span
class="glyphicon glyphicon-question-sign" aria-hidden="true"
title="This is the local directory in which the code in this file was executed.">
</span>
</p>
<p>
This reproducible <a href="https://rmarkdown.rstudio.com">R Markdown</a>
analysis was created with <a
  href="https://github.com/workflowr/workflowr">workflowr</a> (version
1.7.1). The <em>Checks</em> tab describes the reproducibility checks
that were applied when the results were created. The <em>Past
versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date
</a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git
repository, you know the exact version of the code that produced these
results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the
global environment can affect the analysis in your R Markdown file in
unknown ways. For reproduciblity it’s best to always run the code in an
empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20200712code">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Seed:</strong>
<code>set.seed(20200712)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20200712code"
class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20200712)</code> was run prior to running
the code in the R Markdown file. Setting a seed ensures that any results
that rely on randomness, e.g. subsampling or permutations, are
reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Session information:</strong>
recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package
versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be
confident that you successfully produced the results during this
run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr
project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomdavetangmusetree64ac12fd2f21d398fc957a7d6de9abeab3cdbebctargetblank64ac12fa">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Repository version:</strong>
<a href="https://github.com/davetang/muse/tree/64ac12fd2f21d398fc957a7d6de9abeab3cdbebc" target="_blank">64ac12f</a>
</a>
</p>
</div>
<div
id="strongRepositoryversionstrongahrefhttpsgithubcomdavetangmusetree64ac12fd2f21d398fc957a7d6de9abeab3cdbebctargetblank64ac12fa"
class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development
and connecting the code version to the results is critical for
reproducibility.
</p>
<p>
The results in this page were generated with repository version
<a href="https://github.com/davetang/muse/tree/64ac12fd2f21d398fc957a7d6de9abeab3cdbebc" target="_blank">64ac12f</a>.
See the <em>Past versions</em> tab to see a history of the changes made
to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for
the analysis have been committed to Git prior to generating the results
(you can use <code>wflow_publish</code> or
<code>wflow_git_commit</code>). workflowr only checks the R Markdown
file, but you know if there are other scripts or data files that it
depends on. Below is the status of the Git repository when the results
were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .Rproj.user/
    Ignored:    r_packages_4.3.0/
    Ignored:    r_packages_4.3.2/

Untracked files:
    Untracked:  analysis/breast_cancer.Rmd
    Untracked:  code/multiz100way/
    Untracked:  data/pbmc3k/

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not
included in this status report because it is ok for generated content to
have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">

<p>
These are the previous versions of the repository in which changes were
made to the R Markdown (<code>analysis/seurat.Rmd</code>) and HTML
(<code>docs/seurat.html</code>) files. If you’ve configured a remote Git
repository (see <code>?wflow_git_remote</code>), click on the hyperlinks
in the table below to view the files as they were in that past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/davetang/muse/blob/64ac12fd2f21d398fc957a7d6de9abeab3cdbebc/analysis/seurat.Rmd" target="_blank">64ac12f</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-03-20
</td>
<td>
Normalisation, variable genes, and scaling
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/davetang/muse/3c41e26bf6b718001ae68b125ecab024a53d86b3/docs/seurat.html" target="_blank">3c41e26</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-03-20
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/davetang/muse/blob/d5ab5a8c8fa26bd2edd6a2632ce93935120ed49d/analysis/seurat.Rmd" target="_blank">d5ab5a8</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-03-20
</td>
<td>
Basic filtering
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/davetang/muse/5de3e6ce917c55d8600f4435d3952f90603e5801/docs/seurat.html" target="_blank">5de3e6c</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-03-19
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/davetang/muse/blob/0e1dfa9172cb55b17447f3a621fc9b61cf175124/analysis/seurat.Rmd" target="_blank">0e1dfa9</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-03-19
</td>
<td>
Sparse matrix
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/davetang/muse/944e5f2fc3dcce3e4b63a2b724876f092d25f4ab/docs/seurat.html" target="_blank">944e5f2</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-03-19
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/davetang/muse/blob/ae0ea429baa8faa1df208288d885bfa5b4d6043d/analysis/seurat.Rmd" target="_blank">ae0ea42</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-03-19
</td>
<td>
Getting started with Seurat
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<p>This post follows the Peripheral Blood Mononuclear Cells (PBMCs) <a
href="https://satijalab.org/seurat/articles/pbmc3k_tutorial">tutorial</a>
for 2,700 single cells. It was written while I was going through the
tutorial and contains my notes. The dataset for this tutorial can be
downloaded from the <a
href="https://support.10xgenomics.com/single-cell-gene-expression/datasets/1.1.0/pbmc3k">10X
Genomics dataset page</a> but it is also hosted on Amazon (see below).
The PBMCs, which are primary cells with relatively small amounts of RNA
(around 1pg RNA/cell), come from a healthy donor. There were 2,700 cells
detected and sequencing was performed on an Illumina NextSeq 500 with
around 69,000 reads per cell. To get started <a
href="https://satijalab.org/seurat/articles/install.html">install
Seurat</a> by using install.packages().</p>
<pre class="r"><code>install.packages(&quot;Seurat&quot;)</code></pre>
<p>If you get the warning:</p>
<blockquote>
<p>‘SeuratObject’ was built under R 4.3.0 but the current version is
4.3.2; it is recomended that you reinstall ‘SeuratObject’ as the ABI for
R may have changed</p>
</blockquote>
<p>re-install the <code>SeuratObject</code> package using a repository
that has an updated copy. The same goes for the <code>htmltools</code>
package.</p>
<pre class="r"><code>install.packages(&quot;SeuratObject&quot;, repos = &quot;https://cran.ism.ac.jp/&quot;)
install.packages(&quot;htmltools&quot;, repos = &quot;https://cran.ism.ac.jp/&quot;)
packageVersion(&quot;SeuratObject&quot;)
packageVersion(&quot;htmltools&quot;)</code></pre>
<p>Load <code>Seurat</code>.</p>
<pre class="r"><code>library(&quot;Seurat&quot;)</code></pre>
<pre><code>Loading required package: SeuratObject</code></pre>
<pre><code>Loading required package: sp</code></pre>
<pre><code>
Attaching package: &#39;SeuratObject&#39;</code></pre>
<pre><code>The following object is masked from &#39;package:base&#39;:

    intersect</code></pre>
<pre class="r"><code>packageVersion(&quot;Seurat&quot;)</code></pre>
<pre><code>[1] &#39;5.0.3&#39;</code></pre>
<div id="data" class="section level2">
<h2>Data</h2>
<p>To follow the tutorial, you need the 10X data.</p>
<pre class="bash"><code>mkdir -p data/pbmc3k &amp;&amp; cd data/pbmc3k
wget -c https://s3-us-west-2.amazonaws.com/10x.files/samples/cell/pbmc3k/pbmc3k_filtered_gene_bc_matrices.tar.gz
tar -xzf pbmc3k_filtered_gene_bc_matrices.tar.gz</code></pre>
<p>The extracted files.</p>
<pre class="bash"><code>ls -1 data/pbmc3k/filtered_gene_bc_matrices/hg19</code></pre>
<pre><code>barcodes.tsv
genes.tsv
matrix.mtx</code></pre>
<p><code>matrix.mtx</code> is a <a
href="https://math.nist.gov/MatrixMarket/formats.html">MatrixMarket</a>
file. It has the following properties:</p>
<ul>
<li>Only non-zero entries are stored in the file</li>
<li>Comments start with a <code>%</code>, like LaTeX</li>
<li>The first line indicates the total number of rows, columns, and
entries</li>
<li>The following lines after the first provide a row and column number
and the value at that coordinate</li>
</ul>
<pre class="bash"><code>head data/pbmc3k/filtered_gene_bc_matrices/hg19/matrix.mtx</code></pre>
<pre><code>%%MatrixMarket matrix coordinate real general
%
32738 2700 2286884
32709 1 4
32707 1 1
32706 1 10
32704 1 1
32703 1 5
32702 1 6
32700 1 10</code></pre>
</div>
<div id="seurat-object" class="section level2">
<h2>Seurat object</h2>
<p>Load 10x data into a matrix.</p>
<pre class="r"><code>pbmc.data &lt;- Read10X(data.dir = &quot;data/pbmc3k/filtered_gene_bc_matrices/hg19/&quot;)</code></pre>
<p><a
href="https://stat.ethz.ch/R-manual/R-devel/library/Matrix/html/dgTMatrix-class.html">dgTMatrix-class</a>.</p>
<pre class="r"><code>class(pbmc.data)</code></pre>
<pre><code>[1] &quot;dgCMatrix&quot;
attr(,&quot;package&quot;)
[1] &quot;Matrix&quot;</code></pre>
<p>32,738 genes and 2,700 cells.</p>
<pre class="r"><code>dim(pbmc.data)</code></pre>
<pre><code>[1] 32738  2700</code></pre>
<p>Check out the first six genes and cells</p>
<pre class="r"><code>pbmc.data[1:6, 1:6]</code></pre>
<pre><code>6 x 6 sparse Matrix of class &quot;dgCMatrix&quot;
             AAACATACAACCAC-1 AAACATTGAGCTAC-1 AAACATTGATCAGC-1
MIR1302-10                  .                .                .
FAM138A                     .                .                .
OR4F5                       .                .                .
RP11-34P13.7                .                .                .
RP11-34P13.8                .                .                .
AL627309.1                  .                .                .
             AAACCGTGCTTCCG-1 AAACCGTGTATGCG-1 AAACGCACTGGTAC-1
MIR1302-10                  .                .                .
FAM138A                     .                .                .
OR4F5                       .                .                .
RP11-34P13.7                .                .                .
RP11-34P13.8                .                .                .
AL627309.1                  .                .                .</code></pre>
<p>Summary of total expression per single cell.</p>
<pre class="r"><code>summary(colSums(pbmc.data))</code></pre>
<pre><code>   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
    548    1758    2197    2367    2763   15844 </code></pre>
<p>Check how many genes have at least one transcript in each cell.</p>
<p>The median number of detected genes among the single cells is
817.</p>
<pre class="r"><code>at_least_one &lt;- apply(pbmc.data, 2, function(x) sum(x&gt;0))
hist(
  at_least_one,
  breaks = 100,
  main = &quot;Distribution of detected genes&quot;,
  xlab = &quot;Genes with at least one tag&quot;
)
abline(v = median(at_least_one), col = 2, lty = 3)</code></pre>
<p><img src="figure/seurat.Rmd/unnamed-chunk-9-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-9-1">
Past versions of unnamed-chunk-9-1.png
</button>
</p>
<div id="fig-unnamed-chunk-9-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/davetang/muse/blob/3c41e26bf6b718001ae68b125ecab024a53d86b3/docs/figure/seurat.Rmd/unnamed-chunk-9-1.png" target="_blank">3c41e26</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-03-20
</td>
</tr>
<tr>
<td>
<a href="https://github.com/davetang/muse/blob/5de3e6ce917c55d8600f4435d3952f90603e5801/docs/figure/seurat.Rmd/unnamed-chunk-9-1.png" target="_blank">5de3e6c</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-03-19
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>Total expression per cell. The median sum of expression among the
single cells is 2,197. This distribution is very similar to the
distribution of detected genes shown above.</p>
<pre class="r"><code>hist(
  colSums(pbmc.data),
  breaks = 100,
  main = &quot;Expression sum per cell&quot;,
  xlab = &quot;Sum expression&quot;
)
abline(v = median(colSums(pbmc.data)), col = 2, lty = 3)</code></pre>
<p><img src="figure/seurat.Rmd/unnamed-chunk-10-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-10-1">
Past versions of unnamed-chunk-10-1.png
</button>
</p>
<div id="fig-unnamed-chunk-10-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/davetang/muse/blob/3c41e26bf6b718001ae68b125ecab024a53d86b3/docs/figure/seurat.Rmd/unnamed-chunk-10-1.png" target="_blank">3c41e26</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-03-20
</td>
</tr>
<tr>
<td>
<a href="https://github.com/davetang/muse/blob/5de3e6ce917c55d8600f4435d3952f90603e5801/docs/figure/seurat.Rmd/unnamed-chunk-10-1.png" target="_blank">5de3e6c</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-03-19
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>We will filter out genes and single cells before we continue with the
analysis. The tutorial has arbitrary values of keeping genes expressed
in three or more cells and keeping cells with at least 200 detected
genes.</p>
<p>Manually check the number of genes detected in three or more cells; a
lot of genes are not detected in 3 or more cells.</p>
<pre class="r"><code>tmp &lt;- apply(pbmc.data, 1, function(x) sum(x&gt;0))
table(tmp&gt;=3)</code></pre>
<pre><code>
FALSE  TRUE 
19024 13714 </code></pre>
<p>All cells have at least 200 detected genes</p>
<pre class="r"><code>keep &lt;- tmp&gt;=3
tmp &lt;- pbmc.data[keep,]
at_least_one &lt;- apply(tmp, 2, function(x) sum(x&gt;0))
summary(at_least_one)</code></pre>
<pre><code>   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
  212.0   690.0   816.0   845.5   952.0  3400.0 </code></pre>
<pre class="r"><code>dim(tmp)</code></pre>
<pre><code>[1] 13714  2700</code></pre>
<p>See <code>?SeuratObject</code> for more information on the class.</p>
<pre class="r"><code>pbmc &lt;- CreateSeuratObject(
  counts = pbmc.data,
  min.cells = 3,
  min.features = 200,
  project = &quot;pbmc3k&quot;
)</code></pre>
<pre><code>Warning: Feature names cannot have underscores (&#39;_&#39;), replacing with dashes
(&#39;-&#39;)</code></pre>
<pre class="r"><code>class(pbmc)</code></pre>
<pre><code>[1] &quot;Seurat&quot;
attr(,&quot;package&quot;)
[1] &quot;SeuratObject&quot;</code></pre>
<p>Same numbers as above</p>
<pre class="r"><code>pbmc</code></pre>
<pre><code>An object of class Seurat 
13714 features across 2700 samples within 1 assay 
Active assay: RNA (13714 features, 0 variable features)
 1 layer present: counts</code></pre>
<p>Slots in Seurat object.</p>
<blockquote>
<p>SeuratObject: Data Structures for Single Cell Data</p>
<p>Defines S4 classes for single-cell genomic data and associated
information, such as dimensionality reduction embeddings,
nearest-neighbor graphs, and spatially-resolved coordinates. Provides
data access methods and R-native hooks to ensure the Seurat object is
familiar to other R users</p>
</blockquote>
<p>Read more about the <a href="https://adv-r.hadley.nz/s4.html">S4
class</a> in the Advanced R book.</p>
<pre class="r"><code>slotNames(pbmc)</code></pre>
<pre><code> [1] &quot;assays&quot;       &quot;meta.data&quot;    &quot;active.assay&quot; &quot;active.ident&quot; &quot;graphs&quot;      
 [6] &quot;neighbors&quot;    &quot;reductions&quot;   &quot;images&quot;       &quot;project.name&quot; &quot;misc&quot;        
[11] &quot;version&quot;      &quot;commands&quot;     &quot;tools&quot;       </code></pre>
</div>
<div id="basic-filtering" class="section level2">
<h2>Basic filtering</h2>
<p>The tutorial states that “The number of genes and UMIs (nGene and
nUMI) are automatically calculated for every object by Seurat.” The nUMI
is calculated as num.mol &lt;- colSums(object.raw.data), i.e. each
transcript is a unique molecule. The number of genes is simply the tally
of genes with at least 1 transcript; num.genes &lt;-
colSums(object.raw.data &gt; is.expr) where is.expr is zero.</p>
<p>A common quality control metric is the percentage of transcripts from
the mitochondrial genome. According to the paper <a
href="https://pubmed.ncbi.nlm.nih.gov/26887813/">Classification of low
quality cells from single-cell RNA-seq data</a> the reason this is a
quality control metric is because if a single cell is lysed, cytoplasmic
RNA will be lost apart from the RNA that is enclosed in the
mitochondria, which will be retained and sequenced.</p>
<p>Mitochondria genes conveniently start with MT</p>
<pre class="r"><code>mito.genes &lt;- grep(pattern = &quot;^MT-&quot;, x = rownames(x = pbmc@assays$RNA), value = TRUE)
length(mito.genes)</code></pre>
<pre><code>[1] 13</code></pre>
<pre class="r"><code>percent.mito &lt;- Matrix::colSums(pbmc[[&#39;RNA&#39;]]$counts[mito.genes, ]) / Matrix::colSums(pbmc[[&#39;RNA&#39;]]$counts)
head(percent.mito)</code></pre>
<pre><code>AAACATACAACCAC-1 AAACATTGAGCTAC-1 AAACATTGATCAGC-1 AAACCGTGCTTCCG-1 
     0.030177759      0.037935958      0.008897363      0.017430845 
AAACCGTGTATGCG-1 AAACGCACTGGTAC-1 
     0.012244898      0.016643551 </code></pre>
<p>Check out the meta data</p>
<pre class="r"><code>head(pbmc@meta.data)</code></pre>
<pre><code>                 orig.ident nCount_RNA nFeature_RNA
AAACATACAACCAC-1     pbmc3k       2419          779
AAACATTGAGCTAC-1     pbmc3k       4903         1352
AAACATTGATCAGC-1     pbmc3k       3147         1129
AAACCGTGCTTCCG-1     pbmc3k       2639          960
AAACCGTGTATGCG-1     pbmc3k        980          521
AAACGCACTGGTAC-1     pbmc3k       2163          781</code></pre>
<p>add some more meta data</p>
<pre class="r"><code>pbmc &lt;- AddMetaData(object = pbmc,
                    metadata = percent.mito,
                    col.name = &quot;percent.mito&quot;)
head(pbmc@meta.data)</code></pre>
<pre><code>                 orig.ident nCount_RNA nFeature_RNA percent.mito
AAACATACAACCAC-1     pbmc3k       2419          779  0.030177759
AAACATTGAGCTAC-1     pbmc3k       4903         1352  0.037935958
AAACATTGATCAGC-1     pbmc3k       3147         1129  0.008897363
AAACCGTGCTTCCG-1     pbmc3k       2639          960  0.017430845
AAACCGTGTATGCG-1     pbmc3k        980          521  0.012244898
AAACGCACTGGTAC-1     pbmc3k       2163          781  0.016643551</code></pre>
<p>Plot number of genes, UMIs, and % mitochondria</p>
<p>Visualize QC metrics as a violin plot</p>
<pre class="r"><code>VlnPlot(pbmc, features = c(&quot;nFeature_RNA&quot;, &quot;nCount_RNA&quot;, &quot;percent.mito&quot;), ncol = 3)</code></pre>
<pre><code>Warning: Default search for &quot;data&quot; layer in &quot;RNA&quot; assay yielded no results;
utilizing &quot;counts&quot; layer instead.</code></pre>
<p><img src="figure/seurat.Rmd/unnamed-chunk-21-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-21-1">
Past versions of unnamed-chunk-21-1.png
</button>
</p>
<div id="fig-unnamed-chunk-21-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/davetang/muse/blob/3c41e26bf6b718001ae68b125ecab024a53d86b3/docs/figure/seurat.Rmd/unnamed-chunk-21-1.png" target="_blank">3c41e26</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-03-20
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>A couple of cells have high mitochondrial percentage which may
indicate lost of cytoplasmic RNA.</p>
<p>The GenePlot() function can be used to visualise gene-gene
relationships as well as any columns in the seurat object. Below we use
the plotting function to spot cells that have a high percentage of
mitochondrial RNA and to plot the relationship between the number of
unique molecules and the number of genes captured.</p>
<p>FeatureScatter is typically used to visualize feature-feature
relationships, but can be used for anything calculated by the object,
i.e. columns in object metadata, PC scores etc.</p>
<pre class="r"><code>plot1 &lt;- FeatureScatter(pbmc, feature1 = &quot;nCount_RNA&quot;, feature2 = &quot;percent.mito&quot;)
plot2 &lt;- FeatureScatter(pbmc, feature1 = &quot;nCount_RNA&quot;, feature2 = &quot;nFeature_RNA&quot;)
plot1 + plot2</code></pre>
<p><img src="figure/seurat.Rmd/unnamed-chunk-22-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-22-1">
Past versions of unnamed-chunk-22-1.png
</button>
</p>
<div id="fig-unnamed-chunk-22-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/davetang/muse/blob/3c41e26bf6b718001ae68b125ecab024a53d86b3/docs/figure/seurat.Rmd/unnamed-chunk-22-1.png" target="_blank">3c41e26</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-03-20
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>Manual check; I already know all cells have &gt;200 genes.</p>
<pre class="r"><code>table(pbmc@meta.data$percent.mito &lt; 0.05 &amp; pbmc@meta.data$nFeature_RNA&lt;2500)</code></pre>
<pre><code>
FALSE  TRUE 
   62  2638 </code></pre>
<pre class="r"><code>pbmc &lt;- subset(pbmc, subset = nFeature_RNA &gt; 200 &amp; nFeature_RNA &lt; 2500 &amp; percent.mito &lt; 0.05)
pbmc</code></pre>
<pre><code>An object of class Seurat 
13714 features across 2638 samples within 1 assay 
Active assay: RNA (13714 features, 0 variable features)
 1 layer present: counts</code></pre>
</div>
<div id="normalisation" class="section level2">
<h2>Normalisation</h2>
<p>The next step is to normalise the data, so that each cell can be
compared against each other. At the time of writing, the only
normalisation method implemented in Seurat is by log normalisation. Gene
expression measurements for each cell are normalised by its total
expression, scaled by 10,000, and log-transformed.</p>
<pre class="r"><code>hist(
  colSums(pbmc[[&#39;RNA&#39;]]$counts),
  breaks = 100,
  main = &quot;Total expression before normalisation&quot;,
  xlab = &quot;Sum of expression&quot;
)</code></pre>
<p><img src="figure/seurat.Rmd/unnamed-chunk-25-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>After removing unwanted cells from the dataset, the next step is to
normalise the data. By default, we employ a global-scaling normalization
method “LogNormalize” that normalises the feature expression
measurements for each cell by the total expression, multiplies this by a
scale factor (10,000 by default), and log-transforms the result. In
Seurat v5, Normalized values are stored in pbmc[[“RNA”]]$data.</p>
<pre class="r"><code>pbmc &lt;- NormalizeData(pbmc, normalization.method = &quot;LogNormalize&quot;, scale.factor = 10000)</code></pre>
<pre><code>Normalizing layer: counts</code></pre>
<p>For clarity, in this previous line of code (and in future commands),
we provide the default values for certain parameters in the function
call. However, this isn’t required and the same behavior can be achieved
with:</p>
<pre class="r"><code>pbmc &lt;- NormalizeData(pbmc)</code></pre>
<pre><code>Normalizing layer: counts</code></pre>
<p>While this method of normalization is standard and widely used in
scRNA-seq analysis, global-scaling relies on an assumption that each
cell originally contains the same number of RNA molecules. We and others
have developed alternative workflows for the single cell preprocessing
that do not make these assumptions. For users who are interested, please
check out our SCTransform() normalization workflow. The method is
described in ourpaper, with a separate vignette using Seurat here. The
use of SCTransform replaces the need to run NormalizeData,
FindVariableFeatures, or ScaleData (described below.)</p>
<pre class="r"><code>hist(
  colSums(pbmc[[&#39;RNA&#39;]]$data),
  breaks = 100,
  main = &quot;Total expression after normalisation&quot;,
  xlab = &quot;Sum of expression&quot;
)</code></pre>
<p><img src="figure/seurat.Rmd/unnamed-chunk-28-1.png" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="identification-of-highly-variable-features-feature-selection"
class="section level2">
<h2>Identification of highly variable features (feature selection)</h2>
<p>Once the data is normalised, the next step is to find genes are vary
between single cells; genes that are constant among all cells have no
distinguishing power. The <code>FindVariableFeatures()</code> function
calculates the average expression and dispersion for each gene, places
these genes into bins, and then calculates a z-score for dispersion
within each bin. I interpret that as take each gene, get the average
expression and variance of the gene across the 2,638 cells, categorise
genes into bins (default is 20) based on their expression and variance,
and finally normalise the variance in each bin. This was the same
approach in <a
href="https://www.ncbi.nlm.nih.gov/pubmed/26000488">Macosko et al.</a>
and new methods for detecting genes with variable expression patterns
will be implemented in Seurat soon (according to the tutorial). The
parameters used below are typical settings for UMI data that is
normalised to a total of 10,000 molecules and will identify around 2,000
variable genes. The tutorial recommends that users should explore the
parameters themselves since each dataset is different.</p>
<p>We next calculate a subset of features that exhibit high cell-to-cell
variation in the dataset (i.e, they are highly expressed in some cells,
and lowly expressed in others). We and others have found that focusing
on these genes in downstream analysis helps to highlight biological
signal in single-cell datasets.</p>
<p>Our procedure in Seurat is described in detail here, and improves on
previous versions by directly modeling the mean-variance relationship
inherent in single-cell data, and is implemented in the
FindVariableFeatures() function. By default, we return 2,000 features
per dataset. These will be used in downstream analysis, like PCA.</p>
<pre class="r"><code>pbmc &lt;- FindVariableFeatures(pbmc, selection.method = &quot;vst&quot;, nfeatures = 2000)</code></pre>
<pre><code>Finding variable features for layer counts</code></pre>
<pre class="r"><code>length(VariableFeatures(pbmc))</code></pre>
<pre><code>[1] 2000</code></pre>
<p>Identify the 10 most highly variable genes</p>
<pre class="r"><code>top10 &lt;- head(VariableFeatures(pbmc), 10)
top10</code></pre>
<pre><code> [1] &quot;PPBP&quot;   &quot;LYZ&quot;    &quot;S100A9&quot; &quot;IGLL5&quot;  &quot;GNLY&quot;   &quot;FTL&quot;    &quot;PF4&quot;    &quot;FTH1&quot;  
 [9] &quot;GNG11&quot;  &quot;S100A8&quot;</code></pre>
<p>Plot variable features with and without labels</p>
<pre class="r"><code>plot1 &lt;- VariableFeaturePlot(pbmc)
plot2 &lt;- LabelPoints(plot = plot1, points = top10, repel = TRUE)</code></pre>
<pre><code>When using repel, set xnudge and ynudge to 0 for optimal results</code></pre>
<pre class="r"><code>plot1 + plot2</code></pre>
<pre><code>Warning in scale_x_log10(): log-10 transformation introduced infinite values.
log-10 transformation introduced infinite values.</code></pre>
<p><img src="figure/seurat.Rmd/unnamed-chunk-32-1.png" width="960" style="display: block; margin: auto;" /></p>
</div>
<div id="scaling-the-data" class="section level2">
<h2>Scaling the data</h2>
<p>Next, we apply a linear transformation (“scaling”) that is a standard
pre-processing step prior to dimensional reduction techniques like PCA.
The <code>ScaleData()</code> function:</p>
<ul>
<li>Shifts the expression of each gene, so that the mean expression
across cells is 0</li>
<li>Scales the expression of each gene, so that the variance across
cells is 1
<ul>
<li>This step gives equal weight in downstream analyses, so that
highly-expressed genes do not dominate</li>
</ul></li>
<li>The results of this are stored in pbmc[[“RNA”]]$scale.data</li>
<li>By default, only variable features are scaled.</li>
<li>You can specify the features argument to scale additional
features</li>
</ul>
<pre class="r"><code>all.genes &lt;- rownames(pbmc)
pbmc &lt;- ScaleData(pbmc, features = all.genes)</code></pre>
<pre><code>Centering and scaling data matrix</code></pre>
<pre class="r"><code>dim(pbmc[[&quot;RNA&quot;]]$scale.data)</code></pre>
<pre><code>[1] 13714  2638</code></pre>
<pre class="r"><code>hist(
  colSums(pbmc[[&#39;RNA&#39;]]$scale.data),
  breaks = 100,
  main = &quot;Total expression after scaling&quot;,
  xlab = &quot;Sum of expression&quot;
)</code></pre>
<p><img src="figure/seurat.Rmd/unnamed-chunk-34-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>How can I remove unwanted sources of variation?</p>
<p>In Seurat, we also use the <code>ScaleData()</code> function to
remove unwanted sources of variation from a single-cell dataset. For
example, we could “regress out” heterogeneity associated with (for
example) cell cycle stage, or mitochondrial contamination i.e.:</p>
<pre class="r"><code>pbmc &lt;- ScaleData(pbmc, features = all.genes, vars.to.regress = &quot;percent.mito&quot;)</code></pre>
<p>However, particularly for advanced users who would like to use this
functionality, we strongly recommend the use of our new normalization
workflow, <code>SCTransform()</code>. The method is described in <a
href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-021-02584-9">this
paper</a>, with a separate <a
href="https://satijalab.org/seurat/articles/sctransform_vignette">vignette
using Seurat</a>. As with <code>ScaleData()</code>, the function
<code>SCTransform()</code> also includes a <code>vars.to.regress</code>
parameter.</p>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span>
Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 4.3.2 (2023-10-31)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 22.04.3 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 
LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

time zone: Etc/UTC
tzcode source: system (glibc)

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] Seurat_5.0.3       SeuratObject_5.0.1 sp_2.1-3           workflowr_1.7.1   

loaded via a namespace (and not attached):
  [1] RColorBrewer_1.1-3     rstudioapi_0.15.0      jsonlite_1.8.8        
  [4] magrittr_2.0.3         spatstat.utils_3.0-4   farver_2.1.1          
  [7] rmarkdown_2.26         fs_1.6.3               vctrs_0.6.5           
 [10] ROCR_1.0-11            spatstat.explore_3.2-6 htmltools_0.5.7       
 [13] sass_0.4.9             sctransform_0.4.1      parallelly_1.37.1     
 [16] KernSmooth_2.23-22     bslib_0.6.1            htmlwidgets_1.6.4     
 [19] ica_1.0-3              plyr_1.8.9             plotly_4.10.4         
 [22] zoo_1.8-12             cachem_1.0.8           whisker_0.4.1         
 [25] igraph_2.0.3           mime_0.12              lifecycle_1.0.4       
 [28] pkgconfig_2.0.3        Matrix_1.6-5           R6_2.5.1              
 [31] fastmap_1.1.1          fitdistrplus_1.1-11    future_1.33.1         
 [34] shiny_1.8.0            digest_0.6.35          colorspace_2.1-0      
 [37] patchwork_1.2.0        ps_1.7.6               rprojroot_2.0.4       
 [40] tensor_1.5             RSpectra_0.16-1        irlba_2.3.5.1         
 [43] labeling_0.4.3         progressr_0.14.0       fansi_1.0.6           
 [46] spatstat.sparse_3.0-3  httr_1.4.7             polyclip_1.10-6       
 [49] abind_1.4-5            compiler_4.3.2         withr_3.0.0           
 [52] fastDummies_1.7.3      highr_0.10             MASS_7.3-60           
 [55] tools_4.3.2            lmtest_0.9-40          httpuv_1.6.14         
 [58] future.apply_1.11.1    goftest_1.2-3          glue_1.7.0            
 [61] callr_3.7.5            nlme_3.1-163           promises_1.2.1        
 [64] grid_4.3.2             Rtsne_0.17             getPass_0.2-4         
 [67] cluster_2.1.4          reshape2_1.4.4         generics_0.1.3        
 [70] gtable_0.3.4           spatstat.data_3.0-4    tidyr_1.3.1           
 [73] data.table_1.15.2      utf8_1.2.4             spatstat.geom_3.2-9   
 [76] RcppAnnoy_0.0.22       ggrepel_0.9.5          RANN_2.6.1            
 [79] pillar_1.9.0           stringr_1.5.1          spam_2.10-0           
 [82] RcppHNSW_0.6.0         later_1.3.2            splines_4.3.2         
 [85] dplyr_1.1.4            lattice_0.21-9         survival_3.5-7        
 [88] deldir_2.0-4           tidyselect_1.2.1       miniUI_0.1.1.1        
 [91] pbapply_1.7-2          knitr_1.45             git2r_0.33.0          
 [94] gridExtra_2.3          scattermore_1.2        xfun_0.42             
 [97] matrixStats_1.2.0      stringi_1.8.3          lazyeval_0.2.2        
[100] yaml_2.3.8             evaluate_0.23          codetools_0.2-19      
[103] tibble_3.2.1           cli_3.6.2              uwot_0.1.16           
[106] xtable_1.8-4           reticulate_1.35.0      munsell_0.5.0         
[109] processx_3.8.4         jquerylib_0.1.4        Rcpp_1.0.12           
[112] globals_0.16.3         spatstat.random_3.2-3  png_0.1-8             
[115] parallel_4.3.2         ellipsis_0.3.2         ggplot2_3.5.0         
[118] dotCall64_1.1-1        listenv_0.9.1          viridisLite_0.4.2     
[121] scales_1.3.0           ggridges_0.5.6         leiden_0.4.3.1        
[124] purrr_1.0.2            rlang_1.1.3            cowplot_1.1.3         </code></pre>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
https://docs.mathjax.org/en/latest/web/configuration.html. This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>




</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
