<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />



<meta name="date" content="2025-03-03" />

<title>Reference-Based Single-Cell RNA-Seq Annotation</title>

<script src="site_libs/header-attrs-2.28/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.13.2/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-6.4.2/css/all.min.css" rel="stylesheet" />
<link href="site_libs/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet" />

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/main/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">My blog</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/davetang/muse">
    <span class="fa fa-github"></span>
     
    Source code
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Reference-Based Single-Cell RNA-Seq
Annotation</h1>
<h4 class="date">2025-03-03</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span>
workflowr <span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2025-03-03
</p>
<p>
<strong>Checks:</strong> <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 7
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong> <code>muse/</code> <span
class="glyphicon glyphicon-question-sign" aria-hidden="true"
title="This is the local directory in which the code in this file was executed.">
</span>
</p>
<p>
This reproducible <a href="https://rmarkdown.rstudio.com">R Markdown</a>
analysis was created with <a
  href="https://github.com/workflowr/workflowr">workflowr</a> (version
1.7.1). The <em>Checks</em> tab describes the reproducibility checks
that were applied when the results were created. The <em>Past
versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date
</a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git
repository, you know the exact version of the code that produced these
results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the
global environment can affect the analysis in your R Markdown file in
unknown ways. For reproduciblity it’s best to always run the code in an
empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20200712code">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Seed:</strong>
<code>set.seed(20200712)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20200712code"
class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20200712)</code> was run prior to running
the code in the R Markdown file. Setting a seed ensures that any results
that rely on randomness, e.g. subsampling or permutations, are
reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Session information:</strong>
recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package
versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be
confident that you successfully produced the results during this
run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr
project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomdavetangmusetree35706ef6d6253c3d2d56003478f69ee5e52c1334targetblank35706efa">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Repository version:</strong>
<a href="https://github.com/davetang/muse/tree/35706ef6d6253c3d2d56003478f69ee5e52c1334" target="_blank">35706ef</a>
</a>
</p>
</div>
<div
id="strongRepositoryversionstrongahrefhttpsgithubcomdavetangmusetree35706ef6d6253c3d2d56003478f69ee5e52c1334targetblank35706efa"
class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development
and connecting the code version to the results is critical for
reproducibility.
</p>
<p>
The results in this page were generated with repository version
<a href="https://github.com/davetang/muse/tree/35706ef6d6253c3d2d56003478f69ee5e52c1334" target="_blank">35706ef</a>.
See the <em>Past versions</em> tab to see a history of the changes made
to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for
the analysis have been committed to Git prior to generating the results
(you can use <code>wflow_publish</code> or
<code>wflow_git_commit</code>). workflowr only checks the R Markdown
file, but you know if there are other scripts or data files that it
depends on. Below is the status of the Git repository when the results
were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .Rproj.user/
    Ignored:    data/1M_neurons_filtered_gene_bc_matrices_h5.h5
    Ignored:    data/293t/
    Ignored:    data/293t_3t3_filtered_gene_bc_matrices.tar.gz
    Ignored:    data/293t_filtered_gene_bc_matrices.tar.gz
    Ignored:    data/5k_Human_Donor1_PBMC_3p_gem-x_5k_Human_Donor1_PBMC_3p_gem-x_count_sample_filtered_feature_bc_matrix.h5
    Ignored:    data/5k_Human_Donor2_PBMC_3p_gem-x_5k_Human_Donor2_PBMC_3p_gem-x_count_sample_filtered_feature_bc_matrix.h5
    Ignored:    data/5k_Human_Donor3_PBMC_3p_gem-x_5k_Human_Donor3_PBMC_3p_gem-x_count_sample_filtered_feature_bc_matrix.h5
    Ignored:    data/5k_Human_Donor4_PBMC_3p_gem-x_5k_Human_Donor4_PBMC_3p_gem-x_count_sample_filtered_feature_bc_matrix.h5
    Ignored:    data/Parent_SC3v3_Human_Glioblastoma_filtered_feature_bc_matrix.tar.gz
    Ignored:    data/brain_counts/
    Ignored:    data/cl.obo
    Ignored:    data/cl.owl
    Ignored:    data/jurkat/
    Ignored:    data/jurkat:293t_50:50_filtered_gene_bc_matrices.tar.gz
    Ignored:    data/jurkat_293t/
    Ignored:    data/jurkat_filtered_gene_bc_matrices.tar.gz
    Ignored:    data/pbmc20k/
    Ignored:    data/pbmc20k_seurat/
    Ignored:    data/pbmc3k/
    Ignored:    data/pbmc4k_filtered_gene_bc_matrices.tar.gz
    Ignored:    data/refdata-gex-GRCh38-2020-A.tar.gz
    Ignored:    data/seurat_1m_neuron.rds
    Ignored:    data/t_3k_filtered_gene_bc_matrices.tar.gz
    Ignored:    r_packages_4.4.1/

Untracked files:
    Untracked:  analysis/bioc_scrnaseq.Rmd

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not
included in this status report because it is ok for generated content to
have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">
<p>
These are the previous versions of the repository in which changes were
made to the R Markdown (<code>analysis/singler.Rmd</code>) and HTML
(<code>docs/singler.html</code>) files. If you’ve configured a remote
Git repository (see <code>?wflow_git_remote</code>), click on the
hyperlinks in the table below to view the files as they were in that
past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/davetang/muse/blob/35706ef6d6253c3d2d56003478f69ee5e52c1334/analysis/singler.Rmd" target="_blank">35706ef</a>
</td>
<td>
Dave Tang
</td>
<td>
2025-03-03
</td>
<td>
Self annotate pbmc3k set
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/davetang/muse/e624270559d4e07ea403430b8d9fa8c2622b8965/docs/singler.html" target="_blank">e624270</a>
</td>
<td>
Dave Tang
</td>
<td>
2025-03-03
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/davetang/muse/blob/49a802331549cc37ae97f1d3e797fe7a2e3e9477/analysis/singler.Rmd" target="_blank">49a8023</a>
</td>
<td>
Dave Tang
</td>
<td>
2025-03-03
</td>
<td>
Fix subsetting
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/davetang/muse/359326fa3c447f2decc91d8a787dd92e5487c773/docs/singler.html" target="_blank">359326f</a>
</td>
<td>
Dave Tang
</td>
<td>
2025-02-26
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/davetang/muse/blob/8a3ad8c0f280f9e0b88f80d70989e7ce1430785b/analysis/singler.Rmd" target="_blank">8a3ad8c</a>
</td>
<td>
Dave Tang
</td>
<td>
2025-02-26
</td>
<td>
pbmc3k
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/davetang/muse/5c7a2bac70dc999545105e753a4ae784475afd4a/docs/singler.html" target="_blank">5c7a2ba</a>
</td>
<td>
Dave Tang
</td>
<td>
2025-02-26
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/davetang/muse/blob/690145ce2a5bf1f4ee9c1dc43e65d00fbfa9e0cd/analysis/singler.Rmd" target="_blank">690145c</a>
</td>
<td>
Dave Tang
</td>
<td>
2025-02-26
</td>
<td>
Using single-cell references
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/davetang/muse/aeeb29460c40ebb46aff3b322e73d63a5d2e19e3/docs/singler.html" target="_blank">aeeb294</a>
</td>
<td>
Dave Tang
</td>
<td>
2025-02-14
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/davetang/muse/blob/e6f0a058a0a001ec3b1f572e3872f6715d3ed2a7/analysis/singler.Rmd" target="_blank">e6f0a05</a>
</td>
<td>
Dave Tang
</td>
<td>
2025-02-14
</td>
<td>
Using SingleR
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<blockquote>
<p>Performs unbiased cell type recognition from single-cell RNA
sequencing data, by leveraging reference transcriptomic datasets of pure
cell types to infer the cell of origin of each single cell
independently.</p>
</blockquote>
<div id="installation" class="section level2">
<h2>Installation</h2>
<p>Install <a
href="https://www.bioconductor.org/packages/release/bioc/html/SingleR.html">SingleR</a>.</p>
<pre class="r"><code>if (!require(&quot;BiocManager&quot;, quietly = TRUE))
    install.packages(&quot;BiocManager&quot;)

BiocManager::install(&quot;SingleR&quot;)
BiocManager::install(&quot;scRNAseq&quot;)
BiocManager::install(&quot;scuttle&quot;)
BiocManager::install(&quot;scran&quot;)
install.packages(&quot;viridis&quot;)
install.packages(&quot;pheatmap&quot;)</code></pre>
</div>
<div id="vignette" class="section level2">
<h2>Vignette</h2>
<p>Following <a
href="https://www.bioconductor.org/packages/release/bioc/vignettes/SingleR/inst/doc/SingleR.html">Using
SingleR to annotate single-cell RNA-seq data</a>.</p>
<blockquote>
<p>SingleR is an automatic annotation method for single-cell RNA
sequencing (scRNAseq) data (Aran et al. 2019). Given a reference dataset
of samples (single-cell or bulk) with known labels, it labels new cells
from a test dataset based on similarity to the reference. Thus, the
burden of manually interpreting clusters and defining marker genes only
has to be done once, for the reference dataset, and this biological
knowledge can be propagated to new datasets in an automated manner.</p>
</blockquote>
<blockquote>
<p>The easiest way to use SingleR is to annotate cells against built-in
references. In particular, the celldex package provides access to
several reference datasets (mostly derived from bulk RNA-seq or
microarray data) through dedicated retrieval functions. Here, we will
use the Human Primary Cell Atlas (Mabbott et al. 2013), represented as a
SummarizedExperiment object containing a matrix of log-expression values
with sample-level labels.</p>
</blockquote>
<pre class="r"><code>suppressPackageStartupMessages(library(celldex))
hpca.se &lt;- HumanPrimaryCellAtlasData()
hpca.se</code></pre>
<pre><code>class: SummarizedExperiment 
dim: 19363 713 
metadata(0):
assays(1): logcounts
rownames(19363): A1BG A1BG-AS1 ... ZZEF1 ZZZ3
rowData names(0):
colnames(713): GSM112490 GSM112491 ... GSM92233 GSM92234
colData names(3): label.main label.fine label.ont</code></pre>
</div>
<div id="human-embryonic-stem-cells" class="section level2">
<h2>Human Embryonic Stem Cells</h2>
<p>Our test dataset consists of some human embryonic stem cells (La
Manno et al. 2016) from the scRNAseq package. For the sake of speed, we
will only label the first 100 cells from this dataset.</p>
<pre class="r"><code>suppressPackageStartupMessages(library(scRNAseq))
hESCs &lt;- LaMannoBrainData(&#39;human-es&#39;)
hESCs &lt;- hESCs[,1:100]</code></pre>
<p>We use our hpca.se reference to annotate each cell in hESCs via the
SingleR() function. This identifies marker genes from the reference and
uses them to compute assignment scores (based on the Spearman
correlation across markers) for each cell in the test dataset against
each label in the reference. The label with the highest score is the
assigned to the test cell, possibly with further fine-tuning to resolve
closely related labels.</p>
<pre class="r"><code>suppressPackageStartupMessages(library(SingleR))
pred.hesc &lt;- SingleR(
  test = hESCs,
  ref = hpca.se,
  assay.type.test=1,
  labels = hpca.se$label.main
)</code></pre>
<p>Each row of the output DataFrame contains prediction results for a
single cell. Labels are shown before (labels) and after pruning
(pruned.labels), along with the associated scores.</p>
<pre class="r"><code>pred.hesc</code></pre>
<pre><code>DataFrame with 100 rows and 4 columns
                                        scores               labels delta.next
                                      &lt;matrix&gt;          &lt;character&gt;  &lt;numeric&gt;
1772122_301_C02 0.347652:0.139036:0.109547:... Neuroepithelial_cell 0.08332864
1772122_180_E05 0.361187:0.155395:0.134934:...              Neurons 0.07283500
1772122_300_H02 0.446411:0.218052:0.190084:... Neuroepithelial_cell 0.13882912
1772122_180_B09 0.373512:0.172438:0.143537:... Neuroepithelial_cell 0.00317443
1772122_180_G04 0.357341:0.157275:0.126511:... Neuroepithelial_cell 0.09717938
...                                        ...                  ...        ...
1772122_299_E07 0.371989:0.202363:0.169379:... Neuroepithelial_cell  0.0837521
1772122_180_D02 0.353314:0.146049:0.115864:... Neuroepithelial_cell  0.0842804
1772122_300_D09 0.348789:0.129193:0.136732:... Neuroepithelial_cell  0.0595056
1772122_298_F09 0.332361:0.173357:0.141439:... Neuroepithelial_cell  0.1200606
1772122_302_A11 0.324928:0.127518:0.101609:...            Astrocyte  0.0509478
                       pruned.labels
                         &lt;character&gt;
1772122_301_C02 Neuroepithelial_cell
1772122_180_E05              Neurons
1772122_300_H02 Neuroepithelial_cell
1772122_180_B09 Neuroepithelial_cell
1772122_180_G04 Neuroepithelial_cell
...                              ...
1772122_299_E07 Neuroepithelial_cell
1772122_180_D02 Neuroepithelial_cell
1772122_300_D09 Neuroepithelial_cell
1772122_298_F09 Neuroepithelial_cell
1772122_302_A11            Astrocyte</code></pre>
<p>SingleR is workflow/package agnostic. The above example uses
<code>SummarizedExperiment</code> objects, but the same functions will
accept any (log-)normalized expression matrix.</p>
</div>
<div id="using-single-cell-references" class="section level2">
<h2>Using single-cell references</h2>
<p>Here, we will use two human pancreas datasets from the scRNAseq
package. The aim is to use one pre-labelled dataset to annotate the
other unlabelled dataset. First, we set up the Muraro et al. (2016)
dataset to be our reference.</p>
<pre class="r"><code>suppressPackageStartupMessages(library(scuttle))
sceM &lt;- MuraroPancreasData()

# One should normally do cell-based quality control at this point, but for
# brevity&#39;s sake, we will just remove the unlabelled libraries here.
sceM &lt;- sceM[,!is.na(sceM$label)]

# SingleR() expects reference datasets to be normalized and log-transformed.
sceM &lt;- logNormCounts(sceM)
sceM</code></pre>
<pre><code>class: SingleCellExperiment 
dim: 19059 2126 
metadata(0):
assays(2): counts logcounts
rownames(19059): A1BG-AS1__chr19 A1BG__chr19 ... ZZEF1__chr17
  ZZZ3__chr1
rowData names(2): symbol chr
colnames(2126): D28-1_1 D28-1_2 ... D30-8_93 D30-8_94
colData names(4): label donor plate sizeFactor
reducedDimNames(0):
mainExpName: endogenous
altExpNames(1): ERCC</code></pre>
<p>Label tally.</p>
<pre class="r"><code>table(colData(sceM)$label)</code></pre>
<pre><code>
     acinar       alpha        beta       delta        duct endothelial 
        219         812         448         193         245          21 
    epsilon mesenchymal          pp     unclear 
          3          80         101           4 </code></pre>
<p>We then set up our test dataset from Grun et al. (2016).</p>
<pre class="r"><code>sceG &lt;- GrunPancreasData()
sceG &lt;- sceG[,colSums(counts(sceG)) &gt; 0] # Remove libraries with no counts.
sceG &lt;- logNormCounts(sceG) 
sceG</code></pre>
<pre><code>class: SingleCellExperiment 
dim: 20064 1718 
metadata(0):
assays(2): counts logcounts
rownames(20064): A1BG-AS1__chr19 A1BG__chr19 ... ZZEF1__chr17
  ZZZ3__chr1
rowData names(2): symbol chr
colnames(1718): D2ex_1 D2ex_2 ... D17TGFB_95 D17TGFB_96
colData names(3): donor sample sizeFactor
reducedDimNames(0):
mainExpName: endogenous
altExpNames(1): ERCC</code></pre>
<p>We then run SingleR() as described previously but with a marker
detection mode that considers the variance of expression across cells.
Here, we will use the Wilcoxon ranked sum test to identify the top
markers for each pairwise comparison between labels. This is slower but
more appropriate for single-cell data compared to the default marker
detection algorithm (which may fail for low-coverage data where the
median is frequently zero).</p>
<pre class="r"><code>pred.grun &lt;- SingleR(
  test=sceG,
  ref=sceM,
  labels=sceM$label,
  de.method=&quot;wilcox&quot;
)

table(pred.grun$labels)</code></pre>
<pre><code>
     acinar       alpha        beta       delta        duct endothelial 
        657         245         276          57         367          34 
    epsilon mesenchymal          pp     unclear 
          1          41          35           5 </code></pre>
</div>
<div id="annotation-diagnostics" class="section level2">
<h2>Annotation diagnostics</h2>
<p><code>plotScoreHeatmap()</code> displays the scores for all cells
across all reference labels, which allows users to inspect the
confidence of the predicted labels across the dataset. Ideally, each
cell (i.e., column of the heatmap) should have one score that is
obviously larger than the rest, indicating that it is unambiguously
assigned to a single label. A spread of similar scores for a given cell
indicates that the assignment is uncertain, though this may be
acceptable if the uncertainty is distributed across similar cell types
that cannot be easily resolved.</p>
<pre class="r"><code>plotScoreHeatmap(pred.grun)</code></pre>
<p><img src="figure/singler.Rmd/plot_score_heatmap-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-plot_score_heatmap-1">
Past versions of plot_score_heatmap-1.png
</button>
</p>
<div id="fig-plot_score_heatmap-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/davetang/muse/blob/aeeb29460c40ebb46aff3b322e73d63a5d2e19e3/docs/figure/singler.Rmd/plot_score_heatmap-1.png" target="_blank">aeeb294</a>
</td>
<td>
Dave Tang
</td>
<td>
2025-02-14
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>Another diagnostic is based on the per-cell “deltas”, i.e., the
difference between the score for the assigned label and the median
across all labels for each cell. Low deltas indicate that the assignment
is uncertain, which is especially relevant if the cell’s true label does
not exist in the reference. We can inspect these deltas across cells for
each label using the plotDeltaDistribution() function.</p>
<pre class="r"><code>plotDeltaDistribution(pred.grun, ncol = 3)</code></pre>
<pre><code>Warning: Groups with fewer than two datapoints have been dropped.
ℹ Set `drop = FALSE` to consider such groups for position adjustment purposes.</code></pre>
<pre><code>Warning in max(data$density, na.rm = TRUE): no non-missing arguments to max;
returning -Inf</code></pre>
<pre><code>Warning: Computation failed in `stat_ydensity()`.
Caused by error in `$&lt;-.data.frame`:
! replacement has 1 row, data has 0</code></pre>
<p><img src="figure/singler.Rmd/plot_delta_dist-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-plot_delta_dist-1">
Past versions of plot_delta_dist-1.png
</button>
</p>
<div id="fig-plot_delta_dist-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/davetang/muse/blob/aeeb29460c40ebb46aff3b322e73d63a5d2e19e3/docs/figure/singler.Rmd/plot_delta_dist-1.png" target="_blank">aeeb294</a>
</td>
<td>
Dave Tang
</td>
<td>
2025-02-14
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>The <code>pruneScores()</code> function will remove potentially
poor-quality or ambiguous assignments based on the deltas. The minimum
threshold on the deltas is defined using an outlier-based approach that
accounts for differences in the scale of the correlations in various
contexts - see ?pruneScores for more details. SingleR() will also report
the pruned scores automatically in the pruned.labels field where
low-quality assignments are replaced with NA.</p>
<pre class="r"><code>summary(is.na(pred.grun$pruned.labels))</code></pre>
<pre><code>   Mode   FALSE    TRUE 
logical    1651      67 </code></pre>
<p>Finally, a simple yet effective diagnostic is to examine the
expression of the marker genes for each label in the test dataset. We
extract the identity of the markers from the metadata of the SingleR()
results and use them in the plotMarkerHeatmap() function, as shown below
for beta cell markers. If a cell in the test dataset is confidently
assigned to a particular label, we would expect it to have strong
expression of that label’s markers. At the very least, it should exhibit
upregulation of those markers relative to cells assigned to other
labels.</p>
<pre class="r"><code>plotMarkerHeatmap(pred.grun, sceG, label=&quot;beta&quot;)</code></pre>
<p><img src="figure/singler.Rmd/plot_marker_heatmap-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-plot_marker_heatmap-1">
Past versions of plot_marker_heatmap-1.png
</button>
</p>
<div id="fig-plot_marker_heatmap-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/davetang/muse/blob/aeeb29460c40ebb46aff3b322e73d63a5d2e19e3/docs/figure/singler.Rmd/plot_marker_heatmap-1.png" target="_blank">aeeb294</a>
</td>
<td>
Dave Tang
</td>
<td>
2025-02-14
</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<div id="pbmc3k" class="section level2">
<h2>pbmc3k</h2>
<p>pbmc3k.</p>
<pre class="r"><code>suppressPackageStartupMessages(library(Seurat))
pbmc.data &lt;- Read10X(data.dir = &quot;data/pbmc3k/filtered_gene_bc_matrices/hg19/&quot;)

pbmc3k &lt;- CreateSeuratObject(
  counts = pbmc.data,
  min.cells = 3,
  min.features = 200,
  project = &quot;pbmc3k&quot;
)</code></pre>
<pre><code>Warning: Feature names cannot have underscores (&#39;_&#39;), replacing with dashes
(&#39;-&#39;)</code></pre>
<pre class="r"><code>pbmc3k &lt;- NormalizeData(pbmc3k)</code></pre>
<pre><code>Normalizing layer: counts</code></pre>
<p>Use Monaco.</p>
<pre class="r"><code>suppressPackageStartupMessages(library(celldex))
monaco_immune &lt;- fetchReference(&quot;monaco_immune&quot;, &quot;2024-02-26&quot;)
monaco_immune</code></pre>
<pre><code>class: SummarizedExperiment 
dim: 46077 114 
metadata(0):
assays(1): logcounts
rownames(46077): A1BG A1BG-AS1 ... ZYX ZZEF1
rowData names(0):
colnames(114): DZQV_CD8_naive DZQV_CD8_CM ... G4YW_Neutrophils
  G4YW_Basophils
colData names(3): label.main label.fine label.ont</code></pre>
<p>Annotate.</p>
<pre class="r"><code>pbmc3k.anno &lt;- SingleR(
  test=pbmc3k@assays$RNA$data,
  ref=monaco_immune,
  labels=colData(monaco_immune)$label.main
)

cbind(
  pbmc3k@meta.data,
  as.data.frame(pbmc3k.anno)
) -&gt; pbmc3k@meta.data

head(pbmc3k@meta.data)</code></pre>
<pre><code>                 orig.ident nCount_RNA nFeature_RNA scores.B.cells
AAACATACAACCAC-1     pbmc3k       2419          779      0.2180441
AAACATTGAGCTAC-1     pbmc3k       4903         1352      0.3757485
AAACATTGATCAGC-1     pbmc3k       3147         1129      0.2021054
AAACCGTGCTTCCG-1     pbmc3k       2639          960      0.2612174
AAACCGTGTATGCG-1     pbmc3k        980          521      0.1354756
AAACGCACTGGTAC-1     pbmc3k       2163          781      0.2428544
                 scores.Basophils scores.CD4..T.cells scores.CD8..T.cells
AAACATACAACCAC-1        0.1383833           0.3191626           0.3271229
AAACATTGAGCTAC-1        0.1595869           0.2226758           0.2314824
AAACATTGATCAGC-1        0.1398993           0.3603814           0.3381419
AAACCGTGCTTCCG-1        0.1925995           0.1409225           0.1433708
AAACCGTGTATGCG-1        0.1062113           0.1834073           0.2328461
AAACGCACTGGTAC-1        0.1773750           0.3236777           0.3062542
                 scores.Dendritic.cells scores.Monocytes scores.Neutrophils
AAACATACAACCAC-1              0.1813302        0.1823273         0.10993201
AAACATTGAGCTAC-1              0.2636019        0.2165097         0.12052022
AAACATTGATCAGC-1              0.1654294        0.1732526         0.12294480
AAACCGTGCTTCCG-1              0.3438019        0.3694793         0.23698599
AAACCGTGTATGCG-1              0.1082860        0.1210741         0.07931607
AAACGCACTGGTAC-1              0.2057659        0.2115375         0.10881833
                 scores.NK.cells scores.Progenitors scores.T.cells       labels
AAACATACAACCAC-1       0.2818949          0.2027839      0.3311532      T cells
AAACATTGAGCTAC-1       0.2137517          0.2441976      0.2270059      B cells
AAACATTGATCAGC-1       0.2701806          0.1943829      0.3374040 CD4+ T cells
AAACCGTGCTTCCG-1       0.1803712          0.2258147      0.1466077    Monocytes
AAACCGTGTATGCG-1       0.2725055          0.1577020      0.2524348     NK cells
AAACGCACTGGTAC-1       0.2690438          0.2098472      0.3002346 CD4+ T cells
                   delta.next pruned.labels
AAACATACAACCAC-1 0.0009944445       T cells
AAACATTGAGCTAC-1 0.1121465835       B cells
AAACATTGATCAGC-1 0.0850874976  CD4+ T cells
AAACCGTGCTTCCG-1 0.1135955003     Monocytes
AAACCGTGTATGCG-1 0.0623907247      NK cells
AAACGCACTGGTAC-1 0.0574108543  CD4+ T cells</code></pre>
<div id="reference" class="section level3">
<h3>Reference</h3>
<p>Split in half.</p>
<pre class="r"><code>total &lt;- nrow(pbmc3k@meta.data)
half &lt;- floor(total / 2)
first_half &lt;- row.names(pbmc3k@meta.data)[1:half]
second_half &lt;- row.names(pbmc3k@meta.data)[(half+1):total]</code></pre>
<p>Create reference.</p>
<pre class="r"><code>sce_ref &lt;- SingleCellExperiment(
  assays = list(counts = pbmc3k@assays$RNA$counts[, first_half])
)
colLabels(sce_ref) &lt;- pbmc3k@meta.data[first_half, &#39;labels&#39;]
sce_ref &lt;- logNormCounts(sce_ref)
sce_ref</code></pre>
<pre><code>class: SingleCellExperiment 
dim: 13714 1350 
metadata(0):
assays(2): counts logcounts
rownames(13714): AL627309.1 AP006222.2 ... PNRC2.1 SRSF10.1
rowData names(0):
colnames(1350): AAACATACAACCAC-1 AAACATTGAGCTAC-1 ... CTATAGCTTCGCTC-1
  CTATAGCTTGCCTC-1
colData names(2): label sizeFactor
reducedDimNames(0):
mainExpName: NULL
altExpNames(0):</code></pre>
<p>Create query.</p>
<pre class="r"><code>sce_query &lt;- SingleCellExperiment(
  assays = list(counts = pbmc3k@assays$RNA$counts[, second_half])
)
sce_query &lt;- logNormCounts(sce_query)
sce_query</code></pre>
<pre><code>class: SingleCellExperiment 
dim: 13714 1350 
metadata(0):
assays(2): counts logcounts
rownames(13714): AL627309.1 AP006222.2 ... PNRC2.1 SRSF10.1
rowData names(0):
colnames(1350): CTATCAACGAACTC-1 CTATCAACGCAGAG-1 ... TTTGCATGAGAGGC-1
  TTTGCATGCCTCAC-1
colData names(1): sizeFactor
reducedDimNames(0):
mainExpName: NULL
altExpNames(0):</code></pre>
<p>Annotate.</p>
<pre class="r"><code>sce.pred &lt;- SingleR(
  test=sce_query,
  ref=sce_ref,
  labels=sce_ref$label,
  de.method=&quot;wilcox&quot;
)

table(sce_ref$label)</code></pre>
<pre><code>
        B cells    CD4+ T cells    CD8+ T cells Dendritic cells       Monocytes 
            179             455             160              23             317 
       NK cells     Progenitors         T cells 
             87               6             123 </code></pre>
<pre class="r"><code>table(sce.pred$labels)</code></pre>
<pre><code>
        B cells    CD4+ T cells    CD8+ T cells Dendritic cells       Monocytes 
            175             508             126              34             312 
       NK cells     Progenitors         T cells 
             72               9             114 </code></pre>
<p>Self annotate.</p>
<pre class="r"><code>pbmc3k_ref &lt;- SingleCellExperiment(
  assays = list(counts = pbmc3k@assays$RNA$counts)
)
colLabels(pbmc3k_ref) &lt;- pbmc3k@meta.data$labels
pbmc3k_ref &lt;- logNormCounts(pbmc3k_ref)

pbmc3k_query &lt;- SingleCellExperiment(
  assays = list(counts = pbmc3k@assays$RNA$counts)
)
pbmc3k_query &lt;- logNormCounts(pbmc3k_query)

pbmc3k.self.pred &lt;- SingleR(
  test=pbmc3k_query,
  ref=pbmc3k_ref,
  labels=pbmc3k_ref$label,
  de.method=&quot;wilcox&quot;
)

stopifnot(all(row.names(pbmc3k.self.pred) == row.names(pbmc3k@meta.data)))
table(
  pbmc3k@meta.data$labels,
  pbmc3k.self.pred$labels
)</code></pre>
<pre><code>                 
                  B cells CD4+ T cells CD8+ T cells Dendritic cells Monocytes
  B cells             351            1            0               0         3
  CD4+ T cells          1          892           12               0         0
  CD8+ T cells          0           62          171               0         0
  Dendritic cells       1            0            0              31        13
  Monocytes             0            1            0               2       638
  NK cells              0            1            3               0         0
  Progenitors           0            2            0               1         0
  T cells               1           54           17               0         0
                 
                  NK cells Progenitors T cells
  B cells                0           0       0
  CD4+ T cells           0           0       6
  CD8+ T cells           1           0      80
  Dendritic cells        0           0       0
  Monocytes              0           0       0
  NK cells             147           0      14
  Progenitors            0          12       0
  T cells                2           0     180</code></pre>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span>
Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 4.4.1 (2024-06-14)
Platform: x86_64-pc-linux-gnu
Running under: Ubuntu 22.04.5 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 
LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

time zone: Etc/UTC
tzcode source: system (glibc)

attached base packages:
[1] stats4    stats     graphics  grDevices utils     datasets  methods  
[8] base     

other attached packages:
 [1] Seurat_5.1.0                SeuratObject_5.0.2         
 [3] sp_2.1-4                    scuttle_1.16.0             
 [5] SingleR_2.8.0               scRNAseq_2.20.0            
 [7] SingleCellExperiment_1.28.1 celldex_1.16.0             
 [9] SummarizedExperiment_1.36.0 Biobase_2.66.0             
[11] GenomicRanges_1.58.0        GenomeInfoDb_1.42.3        
[13] IRanges_2.40.1              S4Vectors_0.44.0           
[15] BiocGenerics_0.52.0         MatrixGenerics_1.18.1      
[17] matrixStats_1.4.1           workflowr_1.7.1            

loaded via a namespace (and not attached):
  [1] spatstat.sparse_3.1-0     fs_1.6.4                 
  [3] ProtGenerics_1.38.0       bitops_1.0-9             
  [5] httr_1.4.7                RColorBrewer_1.1-3       
  [7] sctransform_0.4.1         tools_4.4.1              
  [9] alabaster.base_1.6.1      utf8_1.2.4               
 [11] R6_2.5.1                  HDF5Array_1.34.0         
 [13] uwot_0.2.2                lazyeval_0.2.2           
 [15] rhdf5filters_1.18.0       withr_3.0.2              
 [17] gridExtra_2.3             progressr_0.15.0         
 [19] cli_3.6.3                 spatstat.explore_3.3-3   
 [21] fastDummies_1.7.4         alabaster.se_1.6.0       
 [23] labeling_0.4.3            sass_0.4.9               
 [25] spatstat.data_3.1-2       ggridges_0.5.6           
 [27] pbapply_1.7-2             Rsamtools_2.22.0         
 [29] R.utils_2.12.3            parallelly_1.38.0        
 [31] limma_3.62.2              rstudioapi_0.17.1        
 [33] RSQLite_2.3.7             generics_0.1.3           
 [35] BiocIO_1.16.0             spatstat.random_3.3-2    
 [37] ica_1.0-3                 dplyr_1.1.4              
 [39] Matrix_1.7-0              fansi_1.0.6              
 [41] abind_1.4-8               R.methodsS3_1.8.2        
 [43] lifecycle_1.0.4           whisker_0.4.1            
 [45] yaml_2.3.10               edgeR_4.4.2              
 [47] rhdf5_2.50.2              SparseArray_1.6.1        
 [49] BiocFileCache_2.14.0      Rtsne_0.17               
 [51] grid_4.4.1                blob_1.2.4               
 [53] promises_1.3.0            dqrng_0.4.1              
 [55] ExperimentHub_2.14.0      crayon_1.5.3             
 [57] miniUI_0.1.1.1            lattice_0.22-6           
 [59] beachmat_2.22.0           cowplot_1.1.3            
 [61] GenomicFeatures_1.58.0    KEGGREST_1.46.0          
 [63] pillar_1.9.0              knitr_1.48               
 [65] metapod_1.14.0            rjson_0.2.23             
 [67] future.apply_1.11.3       codetools_0.2-20         
 [69] leiden_0.4.3.1            glue_1.8.0               
 [71] getPass_0.2-4             spatstat.univar_3.0-1    
 [73] data.table_1.16.2         vctrs_0.6.5              
 [75] png_0.1-8                 gypsum_1.2.0             
 [77] spam_2.11-0               gtable_0.3.6             
 [79] cachem_1.1.0              xfun_0.48                
 [81] S4Arrays_1.6.0            mime_0.12                
 [83] survival_3.6-4            pheatmap_1.0.12          
 [85] statmod_1.5.0             bluster_1.16.0           
 [87] fitdistrplus_1.2-1        ROCR_1.0-11              
 [89] nlme_3.1-164              bit64_4.5.2              
 [91] alabaster.ranges_1.6.0    filelock_1.0.3           
 [93] RcppAnnoy_0.0.22          rprojroot_2.0.4          
 [95] bslib_0.8.0               irlba_2.3.5.1            
 [97] KernSmooth_2.23-24        colorspace_2.1-1         
 [99] DBI_1.2.3                 tidyselect_1.2.1         
[101] processx_3.8.4            bit_4.5.0                
[103] compiler_4.4.1            curl_5.2.3               
[105] git2r_0.35.0              httr2_1.0.5              
[107] BiocNeighbors_2.0.1       DelayedArray_0.32.0      
[109] plotly_4.10.4             rtracklayer_1.66.0       
[111] scales_1.3.0              lmtest_0.9-40            
[113] callr_3.7.6               rappdirs_0.3.3           
[115] goftest_1.2-3             stringr_1.5.1            
[117] digest_0.6.37             spatstat.utils_3.1-0     
[119] alabaster.matrix_1.6.1    rmarkdown_2.28           
[121] XVector_0.46.0            htmltools_0.5.8.1        
[123] pkgconfig_2.0.3           sparseMatrixStats_1.18.0 
[125] highr_0.11                dbplyr_2.5.0             
[127] fastmap_1.2.0             ensembldb_2.30.0         
[129] htmlwidgets_1.6.4         rlang_1.1.4              
[131] UCSC.utils_1.2.0          shiny_1.9.1              
[133] DelayedMatrixStats_1.28.1 farver_2.1.2             
[135] jquerylib_0.1.4           zoo_1.8-12               
[137] jsonlite_1.8.9            BiocParallel_1.40.0      
[139] R.oo_1.26.0               BiocSingular_1.22.0      
[141] RCurl_1.98-1.16           magrittr_2.0.3           
[143] GenomeInfoDbData_1.2.13   dotCall64_1.2            
[145] patchwork_1.3.0           Rhdf5lib_1.28.0          
[147] munsell_0.5.1             Rcpp_1.0.13              
[149] viridis_0.6.5             reticulate_1.39.0        
[151] stringi_1.8.4             alabaster.schemas_1.6.0  
[153] zlibbioc_1.52.0           MASS_7.3-60.2            
[155] plyr_1.8.9                AnnotationHub_3.14.0     
[157] parallel_4.4.1            listenv_0.9.1            
[159] ggrepel_0.9.6             deldir_2.0-4             
[161] Biostrings_2.74.1         splines_4.4.1            
[163] tensor_1.5                locfit_1.5-9.10          
[165] ps_1.8.1                  igraph_2.1.1             
[167] spatstat.geom_3.3-3       RcppHNSW_0.6.0           
[169] reshape2_1.4.4            ScaledMatrix_1.14.0      
[171] BiocVersion_3.20.0        XML_3.99-0.17            
[173] evaluate_1.0.1            scran_1.34.0             
[175] BiocManager_1.30.25       httpuv_1.6.15            
[177] polyclip_1.10-7           tidyr_1.3.1              
[179] purrr_1.0.2               RANN_2.6.2               
[181] scattermore_1.2           future_1.34.0            
[183] alabaster.sce_1.6.0       ggplot2_3.5.1            
[185] rsvd_1.0.5                xtable_1.8-4             
[187] restfulr_0.0.15           AnnotationFilter_1.30.0  
[189] RSpectra_0.16-2           later_1.3.2              
[191] viridisLite_0.4.2         tibble_3.2.1             
[193] memoise_2.0.1             AnnotationDbi_1.68.0     
[195] GenomicAlignments_1.42.0  cluster_2.1.6            
[197] globals_0.16.3           </code></pre>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
https://docs.mathjax.org/en/latest/web/configuration.html. This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>




</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
