<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />



<meta name="date" content="2024-08-02" />

<title>Split single column of annotations into multiple columns in R</title>

<script src="site_libs/header-attrs-2.27/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-6.4.2/css/all.min.css" rel="stylesheet" />
<link href="site_libs/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet" />

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/main/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->




</head>

<body>


<div class="container-fluid main-container">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">My blog</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/davetang/muse">
    <span class="fa fa-github"></span>
     
    Source code
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Split single column of annotations into
multiple columns in R</h1>
<h4 class="date">2024-08-02</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span>
workflowr <span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2024-08-02
</p>
<p>
<strong>Checks:</strong> <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 7
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong> <code>muse/</code> <span
class="glyphicon glyphicon-question-sign" aria-hidden="true"
title="This is the local directory in which the code in this file was executed.">
</span>
</p>
<p>
This reproducible <a href="https://rmarkdown.rstudio.com">R Markdown</a>
analysis was created with <a
  href="https://github.com/workflowr/workflowr">workflowr</a> (version
1.7.1). The <em>Checks</em> tab describes the reproducibility checks
that were applied when the results were created. The <em>Past
versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date
</a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git
repository, you know the exact version of the code that produced these
results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the
global environment can affect the analysis in your R Markdown file in
unknown ways. For reproduciblity it’s best to always run the code in an
empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20200712code">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Seed:</strong>
<code>set.seed(20200712)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20200712code"
class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20200712)</code> was run prior to running
the code in the R Markdown file. Setting a seed ensures that any results
that rely on randomness, e.g. subsampling or permutations, are
reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Session information:</strong>
recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package
versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be
confident that you successfully produced the results during this
run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr
project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomdavetangmusetreeb0eadf1b37408fcf8c607f24895b2f7ab5730869targetblankb0eadf1a">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Repository version:</strong>
<a href="https://github.com/davetang/muse/tree/b0eadf1b37408fcf8c607f24895b2f7ab5730869" target="_blank">b0eadf1</a>
</a>
</p>
</div>
<div
id="strongRepositoryversionstrongahrefhttpsgithubcomdavetangmusetreeb0eadf1b37408fcf8c607f24895b2f7ab5730869targetblankb0eadf1a"
class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development
and connecting the code version to the results is critical for
reproducibility.
</p>
<p>
The results in this page were generated with repository version
<a href="https://github.com/davetang/muse/tree/b0eadf1b37408fcf8c607f24895b2f7ab5730869" target="_blank">b0eadf1</a>.
See the <em>Past versions</em> tab to see a history of the changes made
to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for
the analysis have been committed to Git prior to generating the results
(you can use <code>wflow_publish</code> or
<code>wflow_git_commit</code>). workflowr only checks the R Markdown
file, but you know if there are other scripts or data files that it
depends on. Below is the status of the Git repository when the results
were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .Rhistory
    Ignored:    .Rproj.user/
    Ignored:    r_packages_4.3.3/
    Ignored:    r_packages_4.4.0/

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not
included in this status report because it is ok for generated content to
have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">
<p>
These are the previous versions of the repository in which changes were
made to the R Markdown (<code>analysis/split_column.Rmd</code>) and HTML
(<code>docs/split_column.html</code>) files. If you’ve configured a
remote Git repository (see <code>?wflow_git_remote</code>), click on the
hyperlinks in the table below to view the files as they were in that
past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/davetang/muse/blob/b0eadf1b37408fcf8c607f24895b2f7ab5730869/analysis/split_column.Rmd" target="_blank">b0eadf1</a>
</td>
<td>
Dave Tang
</td>
<td>
2024-08-02
</td>
<td>
Update VCF example
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/davetang/muse/5fa80f2a968bf218621dc00beccce106685b374b/docs/split_column.html" target="_blank">5fa80f2</a>
</td>
<td>
Dave Tang
</td>
<td>
2022-09-01
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/davetang/muse/blob/c8588b7211da908a96779e6e88c053b112e6ea7e/analysis/split_column.Rmd" target="_blank">c8588b7</a>
</td>
<td>
Dave Tang
</td>
<td>
2022-09-01
</td>
<td>
Split GTF group column
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/davetang/muse/d6644d22c3cb83eda1a8cf0e25dda89805a5dac3/docs/split_column.html" target="_blank">d6644d2</a>
</td>
<td>
Dave Tang
</td>
<td>
2021-06-22
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/davetang/muse/blob/8ae8e21a9a830ad4293cc7ac59dd03763d6945d8/analysis/split_column.Rmd" target="_blank">8ae8e21</a>
</td>
<td>
Dave Tang
</td>
<td>
2021-06-22
</td>
<td>
Post on splitting a single column of annotations
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<div id="introduction" class="section level2">
<h2>Introduction</h2>
<p>Two widely used file formats in bioinformatics, VCF and GTF, have
single columns that are packed with annotation information. This makes
them a bit inconvenient to work with in R when using data frames because
the values need to be unpacked, i.e. split. In addition, this violates
one of the conditions for tidy data, which is that every cell is a
single value. In this post, we will use tools from the
<code>tidyverse</code> to split the values into multiple columns to make
the data easier to work with in R.</p>
<p>To get started, install the <code>tidyverse</code> if you haven’t
already.</p>
<pre class="r"><code>if(!require(&quot;tidyverse&quot;)){
  install.packages(&quot;tidyverse&quot;)
}
library(tidyverse)</code></pre>
</div>
<div id="splitting-the-info-column-in-a-vcf-file"
class="section level2">
<h2>Splitting the info column in a VCF file</h2>
<p>We will load a small portion of a VCF file using
<code>read_tsv</code>; in addition we will rename <code>#CHROM</code> to
<code>CHROM</code> and then change all the column names to lower
case.</p>
<pre class="r"><code>vcf_url &lt;- &quot;https://raw.githubusercontent.com/davetang/learning_vcf_file/master/eg/Pfeiffer.vcf&quot;
read_tsv(vcf_url, comment = &quot;##&quot;, show_col_types = FALSE, n_max = 1000) |&gt;
  dplyr::rename(CHROM = `#CHROM`) |&gt;
  dplyr::rename_with(tolower) -&gt; my_vcf

head(my_vcf)</code></pre>
<pre><code># A tibble: 6 × 10
  chrom    pos id         ref   alt    qual filter info            format manuel
  &lt;dbl&gt;  &lt;dbl&gt; &lt;chr&gt;      &lt;chr&gt; &lt;chr&gt; &lt;dbl&gt; &lt;chr&gt;  &lt;chr&gt;           &lt;chr&gt;  &lt;chr&gt; 
1     1 866511 rs60722469 C     CCCCT 259.  PASS   AC=2;AF=1.00;A… GT:AD… 1/1:6…
2     1 879317 rs7523549  C     T     151.  PASS   AC=1;AF=0.50;A… GT:AD… 0/1:1…
3     1 879482 .          G     C     485.  PASS   AC=1;AF=0.50;A… GT:AD… 0/1:2…
4     1 880390 rs3748593  C     A     288.  PASS   AC=1;AF=0.50;A… GT:AD… 0/1:1…
5     1 881627 rs2272757  G     A     486.  PASS   AC=1;AF=0.50;A… GT:AD… 0/1:1…
6     1 884091 rs7522415  C     G      65.5 PASS   AC=1;AF=0.50;A… GT:AD… 0/1:6…</code></pre>
<p>Note that the <code>info</code> column is packed with all sorts of
information for each variant. Also note the consistent format of the
<code>info</code> column: each annotation is separated by a semi-colon
(<code>;</code>) and annotations are stored as key-value pairs with an
equal sign in between.</p>
<pre class="r"><code>my_vcf |&gt;
  select(info) |&gt;
  head()</code></pre>
<pre><code># A tibble: 6 × 1
  info                                                                          
  &lt;chr&gt;                                                                         
1 AC=2;AF=1.00;AN=2;DB;DP=11;FS=0.000;HRun=0;HaplotypeScore=41.3338;MQ0=0;MQ=61…
2 AC=1;AF=0.50;AN=2;BaseQRankSum=1.455;DB;DP=21;Dels=0.00;FS=1.984;HRun=0;Haplo…
3 AC=1;AF=0.50;AN=2;BaseQRankSum=1.934;DP=48;Dels=0.00;FS=4.452;HRun=0;Haplotyp…
4 AC=1;AF=0.50;AN=2;BaseQRankSum=-4.517;DB;DP=29;Dels=0.00;FS=1.485;HRun=0;Hapl…
5 AC=1;AF=0.50;AN=2;BaseQRankSum=0.199;DB;DP=33;Dels=0.00;FS=0.000;HRun=1;Haplo…
6 AC=1;AF=0.50;AN=2;BaseQRankSum=-0.259;DB;DP=12;Dels=0.00;FS=0.000;HRun=1;Hapl…</code></pre>
<p>Firstly, we will use <code>separate_rows</code> to create a new row
for each annotation by using <code>;</code> as the separator/delimiter;
note that I have included <code>\\s*</code> after <code>;</code>, which
is a regex for specifying a single whitespace occurring 0 or more times.
By including the regex, whitespace after <code>;</code> will be removed,
which is good because we do not want whitespaces in our data. In
addition, a <code>mutate</code> call is used prior to calling
<code>separate_rows</code> and it is used to remove a trailing
semicolon, if it exists.</p>
<pre class="r"><code>my_vcf |&gt;
  mutate(info = sub(pattern = &quot;;$&quot;, replacement = &quot;&quot;, x = .data$info)) |&gt;
  separate_rows(info, sep = &quot;;\\s*&quot;) |&gt;
  head()</code></pre>
<pre><code># A tibble: 6 × 10
  chrom    pos id         ref   alt    qual filter info     format        manuel
  &lt;dbl&gt;  &lt;dbl&gt; &lt;chr&gt;      &lt;chr&gt; &lt;chr&gt; &lt;dbl&gt; &lt;chr&gt;  &lt;chr&gt;    &lt;chr&gt;         &lt;chr&gt; 
1     1 866511 rs60722469 C     CCCCT  259. PASS   AC=2     GT:AD:DP:GQ:… 1/1:6…
2     1 866511 rs60722469 C     CCCCT  259. PASS   AF=1.00  GT:AD:DP:GQ:… 1/1:6…
3     1 866511 rs60722469 C     CCCCT  259. PASS   AN=2     GT:AD:DP:GQ:… 1/1:6…
4     1 866511 rs60722469 C     CCCCT  259. PASS   DB       GT:AD:DP:GQ:… 1/1:6…
5     1 866511 rs60722469 C     CCCCT  259. PASS   DP=11    GT:AD:DP:GQ:… 1/1:6…
6     1 866511 rs60722469 C     CCCCT  259. PASS   FS=0.000 GT:AD:DP:GQ:… 1/1:6…</code></pre>
<p>The next step is to split the key-value pairs and we will use the
<code>separate</code> function to separate the pairs into two columns,
which we will name <code>key</code> and <code>value</code>, using the
equal sign as the separator/delimiter. Sometimes a key is missing a
value and in these cases, the value will be <code>NA</code>.</p>
<pre class="r"><code>my_vcf |&gt;
  mutate(info = sub(pattern = &quot;;$&quot;, replacement = &quot;&quot;, x = .data$info)) |&gt;
  separate_rows(info, sep = &quot;;\\s*&quot;) |&gt;
  separate(info, c(&#39;key&#39;, &#39;value&#39;), sep = &quot;=&quot;) |&gt;
  head(10)</code></pre>
<pre><code># A tibble: 10 × 11
   chrom    pos id         ref   alt    qual filter key      value format manuel
   &lt;dbl&gt;  &lt;dbl&gt; &lt;chr&gt;      &lt;chr&gt; &lt;chr&gt; &lt;dbl&gt; &lt;chr&gt;  &lt;chr&gt;    &lt;chr&gt; &lt;chr&gt;  &lt;chr&gt; 
 1     1 866511 rs60722469 C     CCCCT  259. PASS   AC       2     GT:AD… 1/1:6…
 2     1 866511 rs60722469 C     CCCCT  259. PASS   AF       1.00  GT:AD… 1/1:6…
 3     1 866511 rs60722469 C     CCCCT  259. PASS   AN       2     GT:AD… 1/1:6…
 4     1 866511 rs60722469 C     CCCCT  259. PASS   DB       &lt;NA&gt;  GT:AD… 1/1:6…
 5     1 866511 rs60722469 C     CCCCT  259. PASS   DP       11    GT:AD… 1/1:6…
 6     1 866511 rs60722469 C     CCCCT  259. PASS   FS       0.000 GT:AD… 1/1:6…
 7     1 866511 rs60722469 C     CCCCT  259. PASS   HRun     0     GT:AD… 1/1:6…
 8     1 866511 rs60722469 C     CCCCT  259. PASS   Haploty… 41.3… GT:AD… 1/1:6…
 9     1 866511 rs60722469 C     CCCCT  259. PASS   MQ0      0     GT:AD… 1/1:6…
10     1 866511 rs60722469 C     CCCCT  259. PASS   MQ       61.94 GT:AD… 1/1:6…</code></pre>
<p>The current state of the transformation produces a new row for each
variant annotation and two columns containing the key and value. If we
want our data in wide format where each annotation is a column, we can
use the <code>pivot_wider</code> function.</p>
<p>In the code below, I have used the first seven columns
(<code>id_cols = chrom:filter</code>) to specify the columns that
uniquely identifies each variant, i.e. the same variant will have the
same values in these columns. We specify our column names from the
<code>key</code> column and the values for these cells will come from
the <code>value</code> column.</p>
<pre class="r"><code>my_vcf |&gt;
  mutate(info = sub(pattern = &quot;;$&quot;, replacement = &quot;&quot;, x = .data$info)) |&gt;
  separate_rows(info, sep = &quot;;\\s*&quot;) |&gt;
  separate(info, c(&#39;key&#39;, &#39;value&#39;), sep = &quot;=&quot;) |&gt;
  distinct() |&gt; # remove duplicated annotations, if any
  pivot_wider(id_cols = chrom:filter, names_from = key, values_from = value) -&gt; my_vcf_tidy

head(my_vcf_tidy, 10)</code></pre>
<pre><code># A tibble: 10 × 24
   chrom    pos id       ref   alt     qual filter AC    AF    AN    DB    DP   
   &lt;dbl&gt;  &lt;dbl&gt; &lt;chr&gt;    &lt;chr&gt; &lt;chr&gt;  &lt;dbl&gt; &lt;chr&gt;  &lt;chr&gt; &lt;chr&gt; &lt;chr&gt; &lt;chr&gt; &lt;chr&gt;
 1     1 866511 rs60722… C     CCCCT  259.  PASS   2     1.00  2     &lt;NA&gt;  11   
 2     1 879317 rs75235… C     T      151.  PASS   1     0.50  2     &lt;NA&gt;  21   
 3     1 879482 .        G     C      485.  PASS   1     0.50  2     &lt;NA&gt;  48   
 4     1 880390 rs37485… C     A      288.  PASS   1     0.50  2     &lt;NA&gt;  29   
 5     1 881627 rs22727… G     A      486.  PASS   1     0.50  2     &lt;NA&gt;  33   
 6     1 884091 rs75224… C     G       65.5 PASS   1     0.50  2     &lt;NA&gt;  12   
 7     1 884101 rs49704… A     C       85.8 PASS   1     0.50  2     &lt;NA&gt;  12   
 8     1 892460 rs41285… G     C     1737.  PASS   1     0.50  2     &lt;NA&gt;  152  
 9     1 897730 rs75496… C     T      225.  PASS   1     0.50  2     &lt;NA&gt;  21   
10     1 909238 rs38297… G     C      248.  PASS   1     0.50  2     &lt;NA&gt;  19   
# ℹ 12 more variables: FS &lt;chr&gt;, HRun &lt;chr&gt;, HaplotypeScore &lt;chr&gt;, MQ0 &lt;chr&gt;,
#   MQ &lt;chr&gt;, QD &lt;chr&gt;, set &lt;chr&gt;, BaseQRankSum &lt;chr&gt;, Dels &lt;chr&gt;,
#   MQRankSum &lt;chr&gt;, ReadPosRankSum &lt;chr&gt;, DS &lt;chr&gt;</code></pre>
<p>Now each row is a single variant and each column is a variable,
making it much easier to work with! We can easily subset variants with a
<code>QD &gt; 50</code> (after transforming the type of numeric).</p>
<pre class="r"><code>my_vcf_tidy |&gt;
  dplyr::mutate(QD = as.numeric(QD)) |&gt;
  dplyr::filter(QD &gt; 50)</code></pre>
<pre><code># A tibble: 6 × 24
  chrom      pos id       ref   alt    qual filter AC    AF    AN    DB    DP   
  &lt;dbl&gt;    &lt;dbl&gt; &lt;chr&gt;    &lt;chr&gt; &lt;chr&gt; &lt;dbl&gt; &lt;chr&gt;  &lt;chr&gt; &lt;chr&gt; &lt;chr&gt; &lt;chr&gt; &lt;chr&gt;
1     1  1276973 rs70949… G     GACAC  459. PASS   2     1.00  2     &lt;NA&gt;  7    
2     1  3396278 .        T     TGCC…  447. PASS   1     0.50  2     &lt;NA&gt;  8    
3     1 19208145 rs11254… G     GGGT… 2735. PASS   2     1.00  2     &lt;NA&gt;  25   
4     1 22207804 rs66803… ACCC… A     1769. PASS   2     1.00  2     &lt;NA&gt;  27   
5     1 31764713 rs11166… C     CAAG  1018. PASS   2     1.00  2     &lt;NA&gt;  16   
6     1 31905889 rs30504… A     ACAG  2172. PASS   2     1.00  2     &lt;NA&gt;  39   
# ℹ 12 more variables: FS &lt;chr&gt;, HRun &lt;chr&gt;, HaplotypeScore &lt;chr&gt;, MQ0 &lt;chr&gt;,
#   MQ &lt;chr&gt;, QD &lt;dbl&gt;, set &lt;chr&gt;, BaseQRankSum &lt;chr&gt;, Dels &lt;chr&gt;,
#   MQRankSum &lt;chr&gt;, ReadPosRankSum &lt;chr&gt;, DS &lt;chr&gt;</code></pre>
</div>
<div id="splitting-the-group-column-in-a-gtf-file"
class="section level2">
<h2>Splitting the group column in a GTF file</h2>
<p>The GTF also has a column packed with key-value pairs.</p>
<pre class="r"><code>my_gtf &lt;- read_tsv(
  file = &quot;https://github.com/davetang/importbio/raw/master/inst/extdata/gencode.v38.annotation.sample.gtf.gz&quot;,
  comment = &quot;#&quot;,
  show_col_types = FALSE,
  col_names = c(&#39;chr&#39;, &#39;src&#39;, &#39;feat&#39;, &#39;start&#39;, &#39;end&#39;, &#39;score&#39;, &#39;strand&#39;, &#39;frame&#39;, &#39;group&#39;)
)

my_gtf |&gt;
  select(group) |&gt;
  head()</code></pre>
<pre><code># A tibble: 6 × 1
  group                                                                         
  &lt;chr&gt;                                                                         
1 &quot;gene_id \&quot;ENSG00000223972.5\&quot;; gene_type \&quot;transcribed_unprocessed_pseudogen…
2 &quot;gene_id \&quot;ENSG00000223972.5\&quot;; transcript_id \&quot;ENST00000456328.2\&quot;; gene_typ…
3 &quot;gene_id \&quot;ENSG00000223972.5\&quot;; transcript_id \&quot;ENST00000456328.2\&quot;; gene_typ…
4 &quot;gene_id \&quot;ENSG00000223972.5\&quot;; transcript_id \&quot;ENST00000456328.2\&quot;; gene_typ…
5 &quot;gene_id \&quot;ENSG00000223972.5\&quot;; transcript_id \&quot;ENST00000456328.2\&quot;; gene_typ…
6 &quot;gene_id \&quot;ENSG00000223972.5\&quot;; transcript_id \&quot;ENST00000450305.2\&quot;; gene_typ…</code></pre>
<p>We can use the same strategy (but with some additional formatting
steps) to split the column up.</p>
<pre class="r"><code>my_gtf |&gt;
  mutate(group = sub(pattern = &quot;;$&quot;, replacement = &quot;&quot;, x = .data$group)) |&gt;
  mutate(group = gsub(pattern = &#39;&quot;&#39;, replacement = &quot;&quot;, x = .data$group)) |&gt;
  separate_rows(group, sep = &quot;;\\s*&quot;) |&gt;
  separate(group, c(&#39;key&#39;, &#39;value&#39;), sep = &quot;\\s&quot;) |&gt;
  distinct() |&gt; # remove duplicated annotations, if any
  pivot_wider(id_cols = chr:frame, names_from = key, values_from = value) -&gt; my_gtf_split</code></pre>
<pre><code>Warning: Values from `value` are not uniquely identified; output will contain list-cols.
• Use `values_fn = list` to suppress this warning.
• Use `values_fn = {summary_fun}` to summarise duplicates.
• Use the following dplyr code to identify duplicates.
  {data} |&gt;
  dplyr::summarise(n = dplyr::n(), .by = c(chr, src, feat, start, end, score,
  strand, frame, key)) |&gt;
  dplyr::filter(n &gt; 1L)</code></pre>
<pre class="r"><code>head(my_gtf_split, 10)</code></pre>
<pre><code># A tibble: 10 × 24
   chr   src    feat  start   end score strand frame gene_id gene_type gene_name
   &lt;chr&gt; &lt;chr&gt;  &lt;chr&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;chr&gt; &lt;chr&gt;  &lt;chr&gt; &lt;list&gt;  &lt;list&gt;    &lt;list&gt;   
 1 chr1  HAVANA gene  11869 14409 .     +      .     &lt;chr&gt;   &lt;chr [1]&gt; &lt;chr [1]&gt;
 2 chr1  HAVANA tran… 11869 14409 .     +      .     &lt;chr&gt;   &lt;chr [1]&gt; &lt;chr [1]&gt;
 3 chr1  HAVANA exon  11869 12227 .     +      .     &lt;chr&gt;   &lt;chr [1]&gt; &lt;chr [1]&gt;
 4 chr1  HAVANA exon  12613 12721 .     +      .     &lt;chr&gt;   &lt;chr [1]&gt; &lt;chr [1]&gt;
 5 chr1  HAVANA exon  13221 14409 .     +      .     &lt;chr&gt;   &lt;chr [1]&gt; &lt;chr [1]&gt;
 6 chr1  HAVANA tran… 12010 13670 .     +      .     &lt;chr&gt;   &lt;chr [1]&gt; &lt;chr [1]&gt;
 7 chr1  HAVANA exon  12010 12057 .     +      .     &lt;chr&gt;   &lt;chr [1]&gt; &lt;chr [1]&gt;
 8 chr1  HAVANA exon  12179 12227 .     +      .     &lt;chr&gt;   &lt;chr [1]&gt; &lt;chr [1]&gt;
 9 chr1  HAVANA exon  12613 12697 .     +      .     &lt;chr&gt;   &lt;chr [1]&gt; &lt;chr [1]&gt;
10 chr1  HAVANA exon  12975 13052 .     +      .     &lt;chr&gt;   &lt;chr [1]&gt; &lt;chr [1]&gt;
# ℹ 13 more variables: level &lt;list&gt;, hgnc_id &lt;list&gt;, havana_gene &lt;list&gt;,
#   transcript_id &lt;list&gt;, transcript_type &lt;list&gt;, transcript_name &lt;list&gt;,
#   transcript_support_level &lt;list&gt;, tag &lt;list&gt;, havana_transcript &lt;list&gt;,
#   exon_number &lt;list&gt;, exon_id &lt;list&gt;, ont &lt;list&gt;, protein_id &lt;list&gt;</code></pre>
<p>However, the split columns are lists because there were some cases
where there were multiple annotations with the same key and a list is
needed to store multiple values (which was what the warning above was
about). For example the <code>tag</code> key was repeated more than once
with different unique values for some annotations.</p>
<pre class="r"><code>map_lgl(my_gtf_split$tag, function(x) length(x) &gt; 1)</code></pre>
<pre><code> [1] FALSE FALSE FALSE FALSE FALSE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE
[13] FALSE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE
[25]  TRUE FALSE  TRUE  TRUE FALSE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE
[37] FALSE  TRUE  TRUE FALSE  TRUE  TRUE  TRUE  TRUE FALSE FALSE FALSE FALSE
[49]  TRUE  TRUE FALSE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE FALSE  TRUE  TRUE
[61]  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE FALSE  TRUE  TRUE
[73]  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE FALSE FALSE
[85] FALSE FALSE FALSE FALSE FALSE FALSE FALSE  TRUE  TRUE  TRUE</code></pre>
<p>We can check which columns have multiple values.</p>
<pre class="r"><code>check_column &lt;- function(x){
  any(map_lgl(x, function(y) length(y) &gt; 1))
}

my_check &lt;- map_lgl(my_gtf_split, check_column)
my_check[my_check]</code></pre>
<pre><code>    transcript_id   transcript_name               tag havana_transcript 
             TRUE              TRUE              TRUE              TRUE 
      exon_number               ont 
             TRUE              TRUE </code></pre>
<p>Therefore despite only a subset of the columns containing multiple
values, all the pivoted columns were turned into lists. However we can
turn the columns back into characters, which makes sense for the
<code>gene_id</code> column which only contained single unique character
values in the first place.</p>
<pre class="r"><code>my_gtf_split |&gt;
  mutate(gene_id = as.character(gene_id)) |&gt;
  head()</code></pre>
<pre><code># A tibble: 6 × 24
  chr   src    feat   start   end score strand frame gene_id gene_type gene_name
  &lt;chr&gt; &lt;chr&gt;  &lt;chr&gt;  &lt;dbl&gt; &lt;dbl&gt; &lt;chr&gt; &lt;chr&gt;  &lt;chr&gt; &lt;chr&gt;   &lt;list&gt;    &lt;list&gt;   
1 chr1  HAVANA gene   11869 14409 .     +      .     ENSG00… &lt;chr [1]&gt; &lt;chr [1]&gt;
2 chr1  HAVANA trans… 11869 14409 .     +      .     ENSG00… &lt;chr [1]&gt; &lt;chr [1]&gt;
3 chr1  HAVANA exon   11869 12227 .     +      .     ENSG00… &lt;chr [1]&gt; &lt;chr [1]&gt;
4 chr1  HAVANA exon   12613 12721 .     +      .     ENSG00… &lt;chr [1]&gt; &lt;chr [1]&gt;
5 chr1  HAVANA exon   13221 14409 .     +      .     ENSG00… &lt;chr [1]&gt; &lt;chr [1]&gt;
6 chr1  HAVANA trans… 12010 13670 .     +      .     ENSG00… &lt;chr [1]&gt; &lt;chr [1]&gt;
# ℹ 13 more variables: level &lt;list&gt;, hgnc_id &lt;list&gt;, havana_gene &lt;list&gt;,
#   transcript_id &lt;list&gt;, transcript_type &lt;list&gt;, transcript_name &lt;list&gt;,
#   transcript_support_level &lt;list&gt;, tag &lt;list&gt;, havana_transcript &lt;list&gt;,
#   exon_number &lt;list&gt;, exon_id &lt;list&gt;, ont &lt;list&gt;, protein_id &lt;list&gt;</code></pre>
<p>But we can also do this to the <code>tag</code> column (even if it
needed a list to store the multiple values) and entries with multiple
values get turned into R (character) code!</p>
<pre class="r"><code>my_gtf_split |&gt;
  mutate(tag = as.character(tag)) |&gt;
  select(tag) |&gt;
  head()</code></pre>
<pre><code># A tibble: 6 × 1
  tag                                  
  &lt;chr&gt;                                
1 &quot;NULL&quot;                               
2 &quot;basic&quot;                              
3 &quot;basic&quot;                              
4 &quot;basic&quot;                              
5 &quot;basic&quot;                              
6 &quot;c(\&quot;basic\&quot;, \&quot;Ensembl_canonical\&quot;)&quot;</code></pre>
<p>If you don’t mind having R (character) code in your data, you can
perform this transformation across all pivoted columns.</p>
<pre class="r"><code>my_gtf_split |&gt;
  mutate(across(gene_id:protein_id, as.character))</code></pre>
<pre><code># A tibble: 94 × 24
   chr   src    feat  start   end score strand frame gene_id gene_type gene_name
   &lt;chr&gt; &lt;chr&gt;  &lt;chr&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;chr&gt; &lt;chr&gt;  &lt;chr&gt; &lt;chr&gt;   &lt;chr&gt;     &lt;chr&gt;    
 1 chr1  HAVANA gene  11869 14409 .     +      .     ENSG00… transcri… DDX11L1  
 2 chr1  HAVANA tran… 11869 14409 .     +      .     ENSG00… transcri… DDX11L1  
 3 chr1  HAVANA exon  11869 12227 .     +      .     ENSG00… transcri… DDX11L1  
 4 chr1  HAVANA exon  12613 12721 .     +      .     ENSG00… transcri… DDX11L1  
 5 chr1  HAVANA exon  13221 14409 .     +      .     ENSG00… transcri… DDX11L1  
 6 chr1  HAVANA tran… 12010 13670 .     +      .     ENSG00… transcri… DDX11L1  
 7 chr1  HAVANA exon  12010 12057 .     +      .     ENSG00… transcri… DDX11L1  
 8 chr1  HAVANA exon  12179 12227 .     +      .     ENSG00… transcri… DDX11L1  
 9 chr1  HAVANA exon  12613 12697 .     +      .     ENSG00… transcri… DDX11L1  
10 chr1  HAVANA exon  12975 13052 .     +      .     ENSG00… transcri… DDX11L1  
# ℹ 84 more rows
# ℹ 13 more variables: level &lt;chr&gt;, hgnc_id &lt;chr&gt;, havana_gene &lt;chr&gt;,
#   transcript_id &lt;chr&gt;, transcript_type &lt;chr&gt;, transcript_name &lt;chr&gt;,
#   transcript_support_level &lt;chr&gt;, tag &lt;chr&gt;, havana_transcript &lt;chr&gt;,
#   exon_number &lt;chr&gt;, exon_id &lt;chr&gt;, ont &lt;chr&gt;, protein_id &lt;chr&gt;</code></pre>
</div>
<div id="summary" class="section level2">
<h2>Summary</h2>
<p>The following steps can be used to split a column containing
key-value pairs into separate columns:</p>
<ol style="list-style-type: decimal">
<li>Use <code>separate_rows</code> to split a single column into
rows</li>
<li>Use <code>separate</code> to split a key-value pair into two
columns</li>
<li>Use <code>pivot_wider</code> to convert the long format table back
to wide format</li>
</ol>
<p>However, sometimes data is packed into a single column because it
cannot be nicely formatted in the first place.</p>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span>
Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 4.4.0 (2024-04-24)
Platform: x86_64-pc-linux-gnu
Running under: Ubuntu 22.04.4 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 
LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

time zone: Etc/UTC
tzcode source: system (glibc)

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] lubridate_1.9.3 forcats_1.0.0   stringr_1.5.1   dplyr_1.1.4    
 [5] purrr_1.0.2     readr_2.1.5     tidyr_1.3.1     tibble_3.2.1   
 [9] ggplot2_3.5.1   tidyverse_2.0.0 workflowr_1.7.1

loaded via a namespace (and not attached):
 [1] sass_0.4.9        utf8_1.2.4        generics_0.1.3    stringi_1.8.4    
 [5] hms_1.1.3         digest_0.6.35     magrittr_2.0.3    timechange_0.3.0 
 [9] evaluate_0.24.0   grid_4.4.0        fastmap_1.2.0     rprojroot_2.0.4  
[13] jsonlite_1.8.8    processx_3.8.4    whisker_0.4.1     ps_1.7.6         
[17] promises_1.3.0    httr_1.4.7        fansi_1.0.6       scales_1.3.0     
[21] jquerylib_0.1.4   cli_3.6.2         crayon_1.5.2      rlang_1.1.4      
[25] bit64_4.0.5       munsell_0.5.1     withr_3.0.0       cachem_1.1.0     
[29] yaml_2.3.8        parallel_4.4.0    tools_4.4.0       tzdb_0.4.0       
[33] colorspace_2.1-0  httpuv_1.6.15     curl_5.2.1        vctrs_0.6.5      
[37] R6_2.5.1          lifecycle_1.0.4   git2r_0.33.0      bit_4.0.5        
[41] fs_1.6.4          vroom_1.6.5       pkgconfig_2.0.3   callr_3.7.6      
[45] pillar_1.9.0      bslib_0.7.0       later_1.3.2       gtable_0.3.5     
[49] glue_1.7.0        Rcpp_1.0.12       xfun_0.44         tidyselect_1.2.1 
[53] rstudioapi_0.16.0 knitr_1.47        htmltools_0.5.8.1 rmarkdown_2.27   
[57] compiler_4.4.0    getPass_0.2-4    </code></pre>
</div>
</div>
</div>
</div>
</div>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
https://docs.mathjax.org/en/latest/web/configuration.html. This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>





</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
