<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />



<meta name="date" content="2025-02-24" />

<title>Trajectory Analysis</title>

<script src="site_libs/header-attrs-2.28/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.13.2/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-6.4.2/css/all.min.css" rel="stylesheet" />
<link href="site_libs/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet" />

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/main/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">My blog</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/davetang/muse">
    <span class="fa fa-github"></span>
     
    Source code
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Trajectory Analysis</h1>
<h4 class="date">2025-02-24</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span>
workflowr <span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2025-02-24
</p>
<p>
<strong>Checks:</strong> <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 7
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong> <code>muse/</code> <span
class="glyphicon glyphicon-question-sign" aria-hidden="true"
title="This is the local directory in which the code in this file was executed.">
</span>
</p>
<p>
This reproducible <a href="https://rmarkdown.rstudio.com">R Markdown</a>
analysis was created with <a
  href="https://github.com/workflowr/workflowr">workflowr</a> (version
1.7.1). The <em>Checks</em> tab describes the reproducibility checks
that were applied when the results were created. The <em>Past
versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date
</a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git
repository, you know the exact version of the code that produced these
results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the
global environment can affect the analysis in your R Markdown file in
unknown ways. For reproduciblity it’s best to always run the code in an
empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20200712code">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Seed:</strong>
<code>set.seed(20200712)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20200712code"
class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20200712)</code> was run prior to running
the code in the R Markdown file. Setting a seed ensures that any results
that rely on randomness, e.g. subsampling or permutations, are
reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Session information:</strong>
recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package
versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be
confident that you successfully produced the results during this
run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr
project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomdavetangmusetree8c9397f8f2208c4bfa28655e7834441c49e9fcc6targetblank8c9397fa">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Repository version:</strong>
<a href="https://github.com/davetang/muse/tree/8c9397f8f2208c4bfa28655e7834441c49e9fcc6" target="_blank">8c9397f</a>
</a>
</p>
</div>
<div
id="strongRepositoryversionstrongahrefhttpsgithubcomdavetangmusetree8c9397f8f2208c4bfa28655e7834441c49e9fcc6targetblank8c9397fa"
class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development
and connecting the code version to the results is critical for
reproducibility.
</p>
<p>
The results in this page were generated with repository version
<a href="https://github.com/davetang/muse/tree/8c9397f8f2208c4bfa28655e7834441c49e9fcc6" target="_blank">8c9397f</a>.
See the <em>Past versions</em> tab to see a history of the changes made
to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for
the analysis have been committed to Git prior to generating the results
(you can use <code>wflow_publish</code> or
<code>wflow_git_commit</code>). workflowr only checks the R Markdown
file, but you know if there are other scripts or data files that it
depends on. Below is the status of the Git repository when the results
were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .Rproj.user/
    Ignored:    data/1M_neurons_filtered_gene_bc_matrices_h5.h5
    Ignored:    data/293t/
    Ignored:    data/293t_3t3_filtered_gene_bc_matrices.tar.gz
    Ignored:    data/293t_filtered_gene_bc_matrices.tar.gz
    Ignored:    data/5k_Human_Donor1_PBMC_3p_gem-x_5k_Human_Donor1_PBMC_3p_gem-x_count_sample_filtered_feature_bc_matrix.h5
    Ignored:    data/5k_Human_Donor2_PBMC_3p_gem-x_5k_Human_Donor2_PBMC_3p_gem-x_count_sample_filtered_feature_bc_matrix.h5
    Ignored:    data/5k_Human_Donor3_PBMC_3p_gem-x_5k_Human_Donor3_PBMC_3p_gem-x_count_sample_filtered_feature_bc_matrix.h5
    Ignored:    data/5k_Human_Donor4_PBMC_3p_gem-x_5k_Human_Donor4_PBMC_3p_gem-x_count_sample_filtered_feature_bc_matrix.h5
    Ignored:    data/Parent_SC3v3_Human_Glioblastoma_filtered_feature_bc_matrix.tar.gz
    Ignored:    data/brain_counts/
    Ignored:    data/cl.obo
    Ignored:    data/cl.owl
    Ignored:    data/jurkat/
    Ignored:    data/jurkat:293t_50:50_filtered_gene_bc_matrices.tar.gz
    Ignored:    data/jurkat_293t/
    Ignored:    data/jurkat_filtered_gene_bc_matrices.tar.gz
    Ignored:    data/pbmc20k/
    Ignored:    data/pbmc20k_seurat/
    Ignored:    data/pbmc3k/
    Ignored:    data/pbmc4k_filtered_gene_bc_matrices.tar.gz
    Ignored:    data/refdata-gex-GRCh38-2020-A.tar.gz
    Ignored:    data/seurat_1m_neuron.rds
    Ignored:    data/t_3k_filtered_gene_bc_matrices.tar.gz
    Ignored:    r_packages_4.4.1/

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not
included in this status report because it is ok for generated content to
have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">
<p>
These are the previous versions of the repository in which changes were
made to the R Markdown (<code>analysis/trajectory.Rmd</code>) and HTML
(<code>docs/trajectory.html</code>) files. If you’ve configured a remote
Git repository (see <code>?wflow_git_remote</code>), click on the
hyperlinks in the table below to view the files as they were in that
past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/davetang/muse/blob/8c9397f8f2208c4bfa28655e7834441c49e9fcc6/analysis/trajectory.Rmd" target="_blank">8c9397f</a>
</td>
<td>
Dave Tang
</td>
<td>
2025-02-24
</td>
<td>
Downstream processing
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/davetang/muse/fcd8cb2633d8962a5f0328760178508f542865fa/docs/trajectory.html" target="_blank">fcd8cb2</a>
</td>
<td>
Dave Tang
</td>
<td>
2025-02-23
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/davetang/muse/blob/4a2d928173449b59b376fb377e0aa21f92b8a855/analysis/trajectory.Rmd" target="_blank">4a2d928</a>
</td>
<td>
Dave Tang
</td>
<td>
2025-02-23
</td>
<td>
Trajectory analysis
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<div id="installation" class="section level2">
<h2>Installation</h2>
<p>Install dependencies.</p>
<pre class="r"><code>if (!require(&quot;BiocManager&quot;, quietly = TRUE))
    install.packages(&quot;BiocManager&quot;)

BiocManager::install(&quot;scRNAseq&quot;)
BiocManager::install(&quot;AnnotationHub&quot;)
BiocManager::install(&quot;scater&quot;)</code></pre>
</div>
<div id="data" class="section level2">
<h2>Data</h2>
<p><a
href="https://bioconductor.org/books/3.14/OSCA.workflows/nestorowa-mouse-hsc-smart-seq2.html">Mouse
haematopoietic stem cell</a> (HSC) dataset generated with Smart-seq2
(Nestorowa et al. 2016).</p>
<pre class="r"><code>suppressPackageStartupMessages(library(scRNAseq))
sce.nest &lt;- NestorowaHSCData()</code></pre>
<pre><code>loading from cache</code></pre>
<pre><code>require(&quot;ensembldb&quot;)</code></pre>
<pre class="r"><code>suppressPackageStartupMessages(library(AnnotationHub))
ens.mm.v97 &lt;- AnnotationHub()[[&quot;AH73905&quot;]]</code></pre>
<pre><code>loading from cache</code></pre>
<pre class="r"><code>anno &lt;- select(ens.mm.v97, keys=rownames(sce.nest), 
    keytype=&quot;GENEID&quot;, columns=c(&quot;SYMBOL&quot;, &quot;SEQNAME&quot;))
rowData(sce.nest) &lt;- anno[match(rownames(sce.nest), anno$GENEID),]

sce.nest</code></pre>
<pre><code>class: SingleCellExperiment 
dim: 46078 1920 
metadata(0):
assays(1): counts
rownames(46078): ENSMUSG00000000001 ENSMUSG00000000003 ...
  ENSMUSG00000107391 ENSMUSG00000107392
rowData names(3): GENEID SYMBOL SEQNAME
colnames(1920): HSPC_007 HSPC_013 ... Prog_852 Prog_810
colData names(9): gate broad ... projected metrics
reducedDimNames(1): diffusion
mainExpName: endogenous
altExpNames(2): ERCC FACS</code></pre>
<div id="quality-control" class="section level3">
<h3>Quality control</h3>
<pre class="r"><code>unfiltered &lt;- sce.nest</code></pre>
<p>For some reason, no mitochondrial transcripts are available, so we
will perform quality control using the spike-in proportions only.</p>
<pre class="r"><code>suppressPackageStartupMessages(library(scater))

stats &lt;- perCellQCMetrics(sce.nest)
qc &lt;- quickPerCellQC(stats, percent_subsets=&quot;altexps_ERCC_percent&quot;)
sce.nest &lt;- sce.nest[,!qc$discard]</code></pre>
<p>We examine the number of cells discarded for each reason.</p>
<pre class="r"><code>colSums(as.matrix(qc))</code></pre>
<pre><code>             low_lib_size            low_n_features high_altexps_ERCC_percent 
                      146                        28                       241 
                  discard 
                      264 </code></pre>
<p>We create some diagnostic plots for each metric.</p>
<pre class="r"><code>colData(unfiltered) &lt;- cbind(colData(unfiltered), stats)
unfiltered$discard &lt;- qc$discard

gridExtra::grid.arrange(
    plotColData(unfiltered, y=&quot;sum&quot;, colour_by=&quot;discard&quot;) +
        scale_y_log10() + ggtitle(&quot;Total count&quot;),
    plotColData(unfiltered, y=&quot;detected&quot;, colour_by=&quot;discard&quot;) +
        scale_y_log10() + ggtitle(&quot;Detected features&quot;),
    plotColData(unfiltered, y=&quot;altexps_ERCC_percent&quot;,
        colour_by=&quot;discard&quot;) + ggtitle(&quot;ERCC percent&quot;),
    ncol=2
)</code></pre>
<p><img src="figure/trajectory.Rmd/unnamed-chunk-1-1.png" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="normalisation" class="section level3">
<h3>Normalisation</h3>
<pre class="r"><code>suppressPackageStartupMessages(library(scran))
set.seed(101000110)
clusters &lt;- quickCluster(sce.nest)
sce.nest &lt;- computeSumFactors(sce.nest, clusters=clusters)
sce.nest &lt;- logNormCounts(sce.nest)</code></pre>
<p>We examine some key metrics for the distribution of size factors, and
compare it to the library sizes as a sanity check.</p>
<pre class="r"><code>summary(sizeFactors(sce.nest))</code></pre>
<pre><code>    Min.  1st Qu.   Median     Mean  3rd Qu.     Max. 
 0.03876  0.42023  0.74297  1.00000  1.24900 16.78896 </code></pre>
<pre class="r"><code>plot(librarySizeFactors(sce.nest), sizeFactors(sce.nest), pch=16,
    xlab=&quot;Library size factors&quot;, ylab=&quot;Deconvolution factors&quot;, log=&quot;xy&quot;)</code></pre>
<p><img src="figure/trajectory.Rmd/size_factors-1.png" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="variance-modelling" class="section level3">
<h3>Variance modelling</h3>
<p>We use the spike-in transcripts to model the technical noise as a
function of the mean.</p>
<pre class="r"><code>set.seed(00010101)
dec.nest &lt;- modelGeneVarWithSpikes(sce.nest, &quot;ERCC&quot;)
top.nest &lt;- getTopHVGs(dec.nest, prop=0.1)</code></pre>
<p>Per-gene variance as a function of the mean for the log-expression
values in the Nestorowa HSC dataset. Each point represents a gene
(black) with the mean-variance trend (blue) fitted to the spike-ins
(red).</p>
<pre class="r"><code>plot(dec.nest$mean, dec.nest$total, pch=16, cex=0.5,
    xlab=&quot;Mean of log-expression&quot;, ylab=&quot;Variance of log-expression&quot;)
curfit &lt;- metadata(dec.nest)
curve(curfit$trend(x), col=&#39;dodgerblue&#39;, add=TRUE, lwd=2)
points(curfit$mean, curfit$var, col=&quot;red&quot;)</code></pre>
<p><img src="figure/trajectory.Rmd/per_gene_variance_plot-1.png" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="dimensionality-reduction" class="section level3">
<h3>Dimensionality reduction</h3>
<pre class="r"><code>set.seed(101010011)
sce.nest &lt;- denoisePCA(sce.nest, technical=dec.nest, subset.row=top.nest)
sce.nest &lt;- runTSNE(sce.nest, dimred=&quot;PCA&quot;)</code></pre>
<p>We check that the number of retained PCs is sensible.</p>
<pre class="r"><code>ncol(reducedDim(sce.nest, &quot;PCA&quot;))</code></pre>
<pre><code>[1] 9</code></pre>
</div>
<div id="clustering" class="section level3">
<h3>Clustering</h3>
<pre class="r"><code>snn.gr &lt;- buildSNNGraph(sce.nest, use.dimred=&quot;PCA&quot;)
colLabels(sce.nest) &lt;- factor(igraph::cluster_walktrap(snn.gr)$membership)
table(colLabels(sce.nest))</code></pre>
<pre><code>
  1   2   3   4   5   6   7   8   9  10 
198 319 208 147 221 182  21 209  74  77 </code></pre>
<pre class="r"><code>plotTSNE(sce.nest, colour_by=&quot;label&quot;)</code></pre>
<p><img src="figure/trajectory.Rmd/snn_graph-1.png" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="marker-gene-detection" class="section level3">
<h3>Marker gene detection</h3>
<pre class="r"><code>markers &lt;- findMarkers(sce.nest, colLabels(sce.nest), 
    test.type=&quot;wilcox&quot;, direction=&quot;up&quot;, lfc=0.5,
    row.data=rowData(sce.nest)[,&quot;SYMBOL&quot;,drop=FALSE])</code></pre>
<p>To illustrate the manual annotation process, we examine the marker
genes for one of the clusters. Upregulation of Car2, Hebp1 amd
hemoglobins indicates that cluster 8 contains erythroid precursors.</p>
<pre class="r"><code>chosen &lt;- markers[[&#39;8&#39;]]
best &lt;- chosen[chosen$Top &lt;= 10,]
aucs &lt;- getMarkerEffects(best, prefix=&quot;AUC&quot;)
rownames(aucs) &lt;- best$SYMBOL

suppressPackageStartupMessages(library(pheatmap))
pheatmap(aucs, color=viridis::plasma(100))</code></pre>
<p><img src="figure/trajectory.Rmd/heatmap-1.png" width="672" style="display: block; margin: auto;" /></p>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span>
Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 4.4.1 (2024-06-14)
Platform: x86_64-pc-linux-gnu
Running under: Ubuntu 22.04.5 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 
LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

time zone: Etc/UTC
tzcode source: system (glibc)

attached base packages:
[1] stats4    stats     graphics  grDevices utils     datasets  methods  
[8] base     

other attached packages:
 [1] pheatmap_1.0.12             scran_1.34.0               
 [3] scater_1.34.0               ggplot2_3.5.1              
 [5] scuttle_1.16.0              AnnotationHub_3.14.0       
 [7] BiocFileCache_2.14.0        dbplyr_2.5.0               
 [9] ensembldb_2.30.0            AnnotationFilter_1.30.0    
[11] GenomicFeatures_1.58.0      AnnotationDbi_1.68.0       
[13] scRNAseq_2.20.0             SingleCellExperiment_1.28.1
[15] SummarizedExperiment_1.36.0 Biobase_2.66.0             
[17] GenomicRanges_1.58.0        GenomeInfoDb_1.42.3        
[19] IRanges_2.40.1              S4Vectors_0.44.0           
[21] BiocGenerics_0.52.0         MatrixGenerics_1.18.1      
[23] matrixStats_1.4.1           workflowr_1.7.1            

loaded via a namespace (and not attached):
  [1] later_1.3.2              BiocIO_1.16.0            bitops_1.0-9            
  [4] filelock_1.0.3           tibble_3.2.1             XML_3.99-0.17           
  [7] lifecycle_1.0.4          httr2_1.0.5              edgeR_4.4.2             
 [10] rprojroot_2.0.4          processx_3.8.4           lattice_0.22-6          
 [13] alabaster.base_1.6.1     magrittr_2.0.3           limma_3.62.2            
 [16] sass_0.4.9               rmarkdown_2.28           jquerylib_0.1.4         
 [19] yaml_2.3.10              metapod_1.14.0           httpuv_1.6.15           
 [22] RColorBrewer_1.1-3       cowplot_1.1.3            DBI_1.2.3               
 [25] abind_1.4-8              zlibbioc_1.52.0          Rtsne_0.17              
 [28] purrr_1.0.2              RCurl_1.98-1.16          rappdirs_0.3.3          
 [31] git2r_0.35.0             GenomeInfoDbData_1.2.13  ggrepel_0.9.6           
 [34] irlba_2.3.5.1            alabaster.sce_1.6.0      dqrng_0.4.1             
 [37] codetools_0.2-20         DelayedArray_0.32.0      tidyselect_1.2.1        
 [40] UCSC.utils_1.2.0         farver_2.1.2             ScaledMatrix_1.14.0     
 [43] viridis_0.6.5            GenomicAlignments_1.42.0 jsonlite_1.8.9          
 [46] BiocNeighbors_2.0.1      tools_4.4.1              Rcpp_1.0.13             
 [49] glue_1.8.0               gridExtra_2.3            SparseArray_1.6.1       
 [52] xfun_0.48                dplyr_1.1.4              HDF5Array_1.34.0        
 [55] gypsum_1.2.0             withr_3.0.2              BiocManager_1.30.25     
 [58] fastmap_1.2.0            rhdf5filters_1.18.0      bluster_1.16.0          
 [61] fansi_1.0.6              callr_3.7.6              digest_0.6.37           
 [64] rsvd_1.0.5               R6_2.5.1                 mime_0.12               
 [67] colorspace_2.1-1         RSQLite_2.3.7            utf8_1.2.4              
 [70] generics_0.1.3           rtracklayer_1.66.0       httr_1.4.7              
 [73] S4Arrays_1.6.0           whisker_0.4.1            pkgconfig_2.0.3         
 [76] gtable_0.3.6             blob_1.2.4               XVector_0.46.0          
 [79] htmltools_0.5.8.1        ProtGenerics_1.38.0      scales_1.3.0            
 [82] alabaster.matrix_1.6.1   png_0.1-8                knitr_1.48              
 [85] rstudioapi_0.17.1        rjson_0.2.23             curl_5.2.3              
 [88] cachem_1.1.0             rhdf5_2.50.2             stringr_1.5.1           
 [91] BiocVersion_3.20.0       parallel_4.4.1           vipor_0.4.7             
 [94] restfulr_0.0.15          pillar_1.9.0             grid_4.4.1              
 [97] alabaster.schemas_1.6.0  vctrs_0.6.5              promises_1.3.0          
[100] BiocSingular_1.22.0      beachmat_2.22.0          cluster_2.1.6           
[103] beeswarm_0.4.0           evaluate_1.0.1           cli_3.6.3               
[106] locfit_1.5-9.10          compiler_4.4.1           Rsamtools_2.22.0        
[109] rlang_1.1.4              crayon_1.5.3             labeling_0.4.3          
[112] ps_1.8.1                 getPass_0.2-4            fs_1.6.4                
[115] ggbeeswarm_0.7.2         stringi_1.8.4            alabaster.se_1.6.0      
[118] viridisLite_0.4.2        BiocParallel_1.40.0      munsell_0.5.1           
[121] Biostrings_2.74.1        lazyeval_0.2.2           Matrix_1.7-0            
[124] ExperimentHub_2.14.0     bit64_4.5.2              Rhdf5lib_1.28.0         
[127] KEGGREST_1.46.0          statmod_1.5.0            alabaster.ranges_1.6.0  
[130] highr_0.11               igraph_2.1.1             memoise_2.0.1           
[133] bslib_0.8.0              bit_4.5.0               </code></pre>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
https://docs.mathjax.org/en/latest/web/configuration.html. This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>




</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
