<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />



<meta name="date" content="2025-05-07" />

<title>Nothobranchius furzeri</title>

<script src="site_libs/header-attrs-2.28/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.13.2/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-6.4.2/css/all.min.css" rel="stylesheet" />
<link href="site_libs/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet" />

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/main/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">My blog</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/davetang/muse">
    <span class="fa fa-github"></span>
     
    Source code
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Nothobranchius furzeri</h1>
<h4 class="date">2025-05-07</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span>
workflowr <span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2025-05-07
</p>
<p>
<strong>Checks:</strong> <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 7
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong> <code>muse/</code> <span
class="glyphicon glyphicon-question-sign" aria-hidden="true"
title="This is the local directory in which the code in this file was executed.">
</span>
</p>
<p>
This reproducible <a href="https://rmarkdown.rstudio.com">R Markdown</a>
analysis was created with <a
  href="https://github.com/workflowr/workflowr">workflowr</a> (version
1.7.1). The <em>Checks</em> tab describes the reproducibility checks
that were applied when the results were created. The <em>Past
versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date
</a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git
repository, you know the exact version of the code that produced these
results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the
global environment can affect the analysis in your R Markdown file in
unknown ways. For reproduciblity it’s best to always run the code in an
empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20200712code">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Seed:</strong>
<code>set.seed(20200712)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20200712code"
class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20200712)</code> was run prior to running
the code in the R Markdown file. Setting a seed ensures that any results
that rely on randomness, e.g. subsampling or permutations, are
reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Session information:</strong>
recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package
versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be
confident that you successfully produced the results during this
run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr
project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomdavetangmusetree990a827b55f22a6571cc9ec6bf10b787bd08b843targetblank990a827a">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Repository version:</strong>
<a href="https://github.com/davetang/muse/tree/990a827b55f22a6571cc9ec6bf10b787bd08b843" target="_blank">990a827</a>
</a>
</p>
</div>
<div
id="strongRepositoryversionstrongahrefhttpsgithubcomdavetangmusetree990a827b55f22a6571cc9ec6bf10b787bd08b843targetblank990a827a"
class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development
and connecting the code version to the results is critical for
reproducibility.
</p>
<p>
The results in this page were generated with repository version
<a href="https://github.com/davetang/muse/tree/990a827b55f22a6571cc9ec6bf10b787bd08b843" target="_blank">990a827</a>.
See the <em>Past versions</em> tab to see a history of the changes made
to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for
the analysis have been committed to Git prior to generating the results
(you can use <code>wflow_publish</code> or
<code>wflow_git_commit</code>). workflowr only checks the R Markdown
file, but you know if there are other scripts or data files that it
depends on. Below is the status of the Git repository when the results
were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .Rproj.user/
    Ignored:    data/1M_neurons_filtered_gene_bc_matrices_h5.h5
    Ignored:    data/293t/
    Ignored:    data/293t_3t3_filtered_gene_bc_matrices.tar.gz
    Ignored:    data/293t_filtered_gene_bc_matrices.tar.gz
    Ignored:    data/5k_Human_Donor1_PBMC_3p_gem-x_5k_Human_Donor1_PBMC_3p_gem-x_count_sample_filtered_feature_bc_matrix.h5
    Ignored:    data/5k_Human_Donor2_PBMC_3p_gem-x_5k_Human_Donor2_PBMC_3p_gem-x_count_sample_filtered_feature_bc_matrix.h5
    Ignored:    data/5k_Human_Donor3_PBMC_3p_gem-x_5k_Human_Donor3_PBMC_3p_gem-x_count_sample_filtered_feature_bc_matrix.h5
    Ignored:    data/5k_Human_Donor4_PBMC_3p_gem-x_5k_Human_Donor4_PBMC_3p_gem-x_count_sample_filtered_feature_bc_matrix.h5
    Ignored:    data/97516b79-8d08-46a6-b329-5d0a25b0be98.h5ad
    Ignored:    data/Parent_SC3v3_Human_Glioblastoma_filtered_feature_bc_matrix.tar.gz
    Ignored:    data/brain_counts/
    Ignored:    data/cl.obo
    Ignored:    data/cl.owl
    Ignored:    data/jurkat/
    Ignored:    data/jurkat:293t_50:50_filtered_gene_bc_matrices.tar.gz
    Ignored:    data/jurkat_293t/
    Ignored:    data/jurkat_filtered_gene_bc_matrices.tar.gz
    Ignored:    data/pbmc20k/
    Ignored:    data/pbmc20k_seurat/
    Ignored:    data/pbmc3k/
    Ignored:    data/pbmc3k_bpcells_mat/
    Ignored:    data/pbmc3k_seurat.rds
    Ignored:    data/pbmc4k_filtered_gene_bc_matrices.tar.gz
    Ignored:    data/pbmc_1k_v3_filtered_feature_bc_matrix.h5
    Ignored:    data/pbmc_1k_v3_raw_feature_bc_matrix.h5
    Ignored:    data/refdata-gex-GRCh38-2020-A.tar.gz
    Ignored:    data/seurat_1m_neuron.rds
    Ignored:    data/t_3k_filtered_gene_bc_matrices.tar.gz
    Ignored:    r_packages_4.4.1/

Untracked files:
    Untracked:  Nothobranchius_furzeri.Nfu_20140520.113.gtf.gz
    Untracked:  analysis/bioc_scrnaseq.Rmd
    Untracked:  bpcells_matrix/
    Untracked:  data/GCF_043380555.1-RS_2024_12_gene_ontology.gaf.gz
    Untracked:  m3/
    Untracked:  pbmc3k_before_filtering.rds
    Untracked:  pbmc3k_save_rds.rds
    Untracked:  rsem.merged.gene_counts.tsv

Unstaged changes:
    Modified:   data/nfurzeri_gene_id_to_go_id.csv.gz
    Modified:   data/nfurzeri_gene_info.csv.gz

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not
included in this status report because it is ok for generated content to
have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">
<p>
These are the previous versions of the repository in which changes were
made to the R Markdown (<code>analysis/turquoise_killifish.Rmd</code>)
and HTML (<code>docs/turquoise_killifish.html</code>) files. If you’ve
configured a remote Git repository (see <code>?wflow_git_remote</code>),
click on the hyperlinks in the table below to view the files as they
were in that past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/davetang/muse/blob/990a827b55f22a6571cc9ec6bf10b787bd08b843/analysis/turquoise_killifish.Rmd" target="_blank">990a827</a>
</td>
<td>
Dave Tang
</td>
<td>
2025-05-07
</td>
<td>
Lookup table
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/davetang/muse/074ec500694cb4f8832f84349147d0608d728cc5/docs/turquoise_killifish.html" target="_blank">074ec50</a>
</td>
<td>
Dave Tang
</td>
<td>
2025-05-07
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/davetang/muse/blob/331a019bc36b4387c8114cfe1add2ff966ccaf09/analysis/turquoise_killifish.Rmd" target="_blank">331a019</a>
</td>
<td>
Dave Tang
</td>
<td>
2025-05-07
</td>
<td>
GAF file
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/davetang/muse/9228995a2b9e839b1cf2f6aa7060e1f9baa2001c/docs/turquoise_killifish.html" target="_blank">9228995</a>
</td>
<td>
Dave Tang
</td>
<td>
2025-03-04
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/davetang/muse/blob/225a1d1868b53b769c60405789020a122d233b75/analysis/turquoise_killifish.Rmd" target="_blank">225a1d1</a>
</td>
<td>
Dave Tang
</td>
<td>
2025-03-04
</td>
<td>
nothobranchius_furzeri
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<p>The <a
href="https://bioconductor.org/packages/release/bioc/html/biomaRt.html">biomaRt
package</a> provides an interface to BioMart databases provided by
Ensembl.</p>
<blockquote>
<p>biomaRt provides an interface to a growing collection of databases
implementing the BioMart software suite. The package enables retrieval
of large amounts of data in a uniform way without the need to know the
underlying database schemas or write complex SQL queries. The most
prominent examples of BioMart databases are maintain by Ensembl, which
provides biomaRt users direct access to a diverse set of data and
enables a wide range of powerful online queries from gene annotation to
database mining.</p>
</blockquote>
<p>For more information, check out the <a
href="https://bioconductor.org/packages/release/bioc/vignettes/biomaRt/inst/doc/accessing_ensembl.html">Accessing
Ensembl annotation with biomaRt</a> guide.</p>
<div id="installation" class="section level2">
<h2>Installation</h2>
<p>To begin, install the {biomaRt} package.</p>
<pre class="r"><code>if (!require(&quot;BiocManager&quot;, quietly = TRUE))
    install.packages(&quot;BiocManager&quot;)

BiocManager::install(&quot;biomaRt&quot;)</code></pre>
</div>
<div id="package" class="section level2">
<h2>Package</h2>
<p>Load package.</p>
<pre class="r"><code>packageVersion(&quot;biomaRt&quot;)</code></pre>
<pre><code>[1] &#39;2.62.1&#39;</code></pre>
<pre class="r"><code>suppressPackageStartupMessages(library(biomaRt))</code></pre>
</div>
<div id="getting-started" class="section level2">
<h2>Getting started</h2>
<p>Use mirro.</p>
<pre class="r"><code>ensembl &lt;- useEnsembl(biomart = &quot;ensembl&quot;, mirror = &quot;asia&quot;)
ensembl</code></pre>
<pre><code>Object of class &#39;Mart&#39;:
  Using the ENSEMBL_MART_ENSEMBL BioMart database
  No dataset selected.</code></pre>
<p>Connect to the selected BioMart database by using
<code>useMart()</code>.</p>
<pre class="r"><code>avail_datasets &lt;- listDatasets(ensembl)
head(avail_datasets)</code></pre>
<pre><code>                       dataset                           description
1 abrachyrhynchus_gene_ensembl Pink-footed goose genes (ASM259213v1)
2     acalliptera_gene_ensembl      Eastern happy genes (fAstCal1.3)
3   acarolinensis_gene_ensembl       Green anole genes (AnoCar2.0v2)
4    acchrysaetos_gene_ensembl       Golden eagle genes (bAquChr1.2)
5    acitrinellus_gene_ensembl        Midas cichlid genes (Midas_v5)
6    amelanoleuca_gene_ensembl       Giant panda genes (ASM200744v2)
      version
1 ASM259213v1
2  fAstCal1.3
3 AnoCar2.0v2
4  bAquChr1.2
5    Midas_v5
6 ASM200744v2</code></pre>
<p>Look for human datasets by searching the description column.</p>
<pre class="r"><code>idx &lt;- grep(pattern = &quot;furzeri&quot;, avail_datasets$dataset, ignore.case = TRUE)
avail_datasets[idx, ]</code></pre>
<pre><code>                  dataset                              description      version
117 nfurzeri_gene_ensembl Turquoise killifish genes (Nfu_20140520) Nfu_20140520</code></pre>
<p>Connect to the selected BioMart database and turquoise killifish
dataset.</p>
<pre class="r"><code>ensembl &lt;- useEnsembl(biomart = &quot;ensembl&quot;, mirror = &quot;asia&quot;, dataset=avail_datasets[idx, &#39;dataset&#39;])
ensembl</code></pre>
<pre><code>Object of class &#39;Mart&#39;:
  Using the ENSEMBL_MART_ENSEMBL BioMart database
  Using the nfurzeri_gene_ensembl dataset</code></pre>
<p>Building a query, requires three things:</p>
<ol style="list-style-type: decimal">
<li>filters</li>
<li>attributes</li>
<li>values</li>
</ol>
<p>Use <code>listFilters()</code> to show available filters.</p>
<pre class="r"><code>avail_filters &lt;- listFilters(ensembl)
head(avail_filters)</code></pre>
<pre><code>                        name                            description
1            chromosome_name               Chromosome/scaffold name
2                      start                                  Start
3                        end                                    End
4                     strand                                 Strand
5         chromosomal_region e.g. 1:100:10000:-1, 1:100000:200000:1
6 with_entrezgene_trans_name  With EntrezGene transcript name ID(s)</code></pre>
<p>Use <code>listAttributes()</code> to show available attributes.</p>
<pre class="r"><code>avail_attributes &lt;- listAttributes(ensembl)
avail_attributes |&gt;
  head()</code></pre>
<pre><code>                           name                  description         page
1               ensembl_gene_id               Gene stable ID feature_page
2       ensembl_gene_id_version       Gene stable ID version feature_page
3         ensembl_transcript_id         Transcript stable ID feature_page
4 ensembl_transcript_id_version Transcript stable ID version feature_page
5            ensembl_peptide_id            Protein stable ID feature_page
6    ensembl_peptide_id_version    Protein stable ID version feature_page</code></pre>
<p>The <code>getBM()</code> function is the main query function in
{biomaRt}; use it once you have identified your attributes of interest
and filters to use.</p>
<pre class="r"><code>gene_ids &lt;- c(&quot;ENSNFUG00015000040&quot;, &quot;ENSNFUG00015000042&quot;, &quot;ENSNFUG00015000043&quot;, &quot;ENSNFUG00015000127&quot;)

getBM(
  attributes=c(&#39;ensembl_gene_id&#39;, &#39;external_gene_name&#39;, &#39;description&#39;),
  filters = &#39;ensembl_gene_id&#39;,
  values = gene_ids,
  mart = ensembl
) -&gt; res

res</code></pre>
<pre><code>     ensembl_gene_id external_gene_name
1 ENSNFUG00015000040                   
2 ENSNFUG00015000042              irf2b
3 ENSNFUG00015000043              CASP3
4 ENSNFUG00015000127            SLC38A2
                                                               description
1                                                                         
2     interferon regulatory factor 2b [Source:ZFIN;Acc:ZDB-GENE-041212-38]
3                    caspase 3 [Source:UniProtKB Gene Name;Acc:A0A8C6K862]
4 solute carrier family 38 member 2 [Source:ZFIN;Acc:ZDB-GENE-030131-9659]</code></pre>
<p>All gene IDs.</p>
<pre class="r"><code>gtf_file &lt;- &quot;https://ftp.ensembl.org/pub/release-113/gtf/nothobranchius_furzeri/Nothobranchius_furzeri.Nfu_20140520.113.gtf.gz&quot;
if(file.exists(basename(gtf_file)) == FALSE){
  download.file(url = gtf_file, destfile = basename(gtf_file))
}

gtf_cols &lt;- c(
  &quot;seqname&quot;,
  &quot;source&quot;,
  &quot;feature&quot;,
  &quot;start&quot;,
  &quot;end&quot;,
  &quot;score&quot;,
  &quot;strand&quot;,
  &quot;frame&quot;,
  &quot;attribute&quot;
)

gtf &lt;- readr::read_tsv(file = gtf_file, comment = &quot;#&quot;, col_names = gtf_cols, show_col_types = FALSE)
gtf |&gt;
  dplyr::filter(feature == &quot;gene&quot;) |&gt;
  dplyr::select(attribute) |&gt;
  tidyr::separate_rows(attribute, sep = &quot;;\\s*&quot;) |&gt;
  dplyr::filter(grepl(&quot;gene_id&quot;, attribute)) |&gt;
  tidyr::separate(attribute, c(&#39;key&#39;, &#39;value&#39;), sep = &quot;\\s&quot;) |&gt;
  dplyr::pull(value) |&gt;
  gsub(pattern =&#39;&quot;&#39;, replacement = &quot;&quot;) -&gt; all_gene_ids

length(all_gene_ids) == length(unique(all_gene_ids))</code></pre>
<pre><code>[1] TRUE</code></pre>
<p>Get gene names and descriptions.</p>
<pre class="r"><code>getBM(
  attributes=c(&#39;ensembl_gene_id&#39;, &#39;external_gene_name&#39;, &#39;description&#39;),
  filters = &#39;ensembl_gene_id&#39;,
  values = all_gene_ids,
  mart = ensembl
) -&gt; all_gene_info

head(all_gene_info)</code></pre>
<pre><code>     ensembl_gene_id external_gene_name
1 ENSNFUG00015000040                   
2 ENSNFUG00015000041                   
3 ENSNFUG00015000042              irf2b
4 ENSNFUG00015000043              CASP3
5 ENSNFUG00015000044              cenpu
6 ENSNFUG00015000045            ST3GAL5
                                                                             description
1                                                                                       
2                                                                                       
3                   interferon regulatory factor 2b [Source:ZFIN;Acc:ZDB-GENE-041212-38]
4                                  caspase 3 [Source:UniProtKB Gene Name;Acc:A0A8C6K862]
5                                                                                       
6 ST3 beta-galactoside alpha-2,3-sialyltransferase 5 [Source:ZFIN;Acc:ZDB-GENE-060322-1]</code></pre>
<p>Save lookup table.</p>
<pre class="r"><code>readr::write_csv(x = all_gene_info, file = &quot;data/nfurzeri_gene_info.csv.gz&quot;)</code></pre>
</div>
<div id="gene-ontology-terms" class="section level2">
<h2>Gene Ontology terms</h2>
<p>Find GO attribute names.</p>
<pre class="r"><code>grep(&quot;^go&quot;, avail_attributes$name, ignore.case=TRUE, value=TRUE)</code></pre>
<pre><code>[1] &quot;go_id&quot;                  &quot;go_linkage_type&quot;        &quot;goslim_goa_accession&quot;  
[4] &quot;goslim_goa_description&quot;</code></pre>
<p>Query.</p>
<pre class="r"><code>getBM(
  attributes=c(&quot;ensembl_gene_id&quot;, &quot;go_id&quot;),
  filters=&quot;ensembl_gene_id&quot;,
  values = all_gene_ids,
  mart = ensembl
) -&gt; all_gene_go_ids

head(all_gene_go_ids)</code></pre>
<pre><code>     ensembl_gene_id      go_id
1 ENSNFUG00015000040           
2 ENSNFUG00015000041 GO:0007156
3 ENSNFUG00015000041 GO:0005886
4 ENSNFUG00015000041 GO:0007399
5 ENSNFUG00015000041 GO:0098609
6 ENSNFUG00015000041 GO:0050808</code></pre>
<p>Save GO table.</p>
<pre class="r"><code>readr::write_csv(x = all_gene_go_ids, file = &quot;data/nfurzeri_gene_id_to_go_id.csv.gz&quot;)</code></pre>
</div>
<div id="ncbi" class="section level2">
<h2>NCBI</h2>
<p>NCBI <a
href="https://www.ncbi.nlm.nih.gov/datasets/genome/GCF_043380555.1/">Genome
assembly NfurGRZ-RIMD1</a> reference.</p>
<p>Download GAF file <code>*_gene_ontology.gaf.gz</code>, which contains
Gene Ontology (GO) annotation of the annotated genes in GO Annotation
File <a
href="https://geneontology.org/docs/go-annotation-file-gaf-format-2.2/">GAF
format</a>.</p>
<pre class="r"><code>gaf_url &lt;- &#39;https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/043/380/555/GCF_043380555.1_NfurGRZ-RIMD1/GCF_043380555.1-RS_2024_12_gene_ontology.gaf.gz&#39;
gaf_file &lt;- paste0(&quot;data/&quot;, basename(gaf_url))

if(file.exists(gaf_file) == FALSE){
  download.file(url = gaf_url, destfile = gaf_file)
}</code></pre>
<p>The annotation flat file format is comprised of 17 tab-delimited
fields.</p>
<pre class="r"><code>gaf_cols &lt;- c(&quot;DB&quot;,&quot;GeneID&quot;,&quot;Symbol&quot;,&quot;Qualifier&quot;,&quot;GO_ID&quot;,&quot;Reference&quot;,&quot;Evidence_Code&quot;,&quot;With_From&quot;,&quot;Aspect&quot;,&quot;Gene_Name&quot;,&quot;Gene_Synonym&quot;,&quot;Type&quot;,&quot;Taxon&quot;,&quot;Date&quot;,&quot;Assigned_By&quot;,&quot;Annot_Ext&quot;,&quot;Gene_Product_Form_ID&quot;)

gaf &lt;- readr::read_tsv(file = gaf_file, comment = &#39;!&#39;, show_col_types = FALSE, col_names = gaf_cols)
head(gaf)</code></pre>
<pre><code># A tibble: 6 × 17
  DB      GeneID Symbol Qualifier GO_ID Reference Evidence_Code With_From Aspect
  &lt;chr&gt;    &lt;dbl&gt; &lt;chr&gt;  &lt;chr&gt;     &lt;chr&gt; &lt;chr&gt;     &lt;chr&gt;         &lt;chr&gt;     &lt;chr&gt; 
1 NCBIGe… 1.07e8 psd2   enables   GO:0… PMID:223… IEA           InterPro… F     
2 NCBIGe… 1.07e8 psd2   enables   GO:0… PMID:223… IEA           InterPro… F     
3 NCBIGe… 1.07e8 psd2   enables   GO:0… PMID:223… IEA           InterPro… F     
4 NCBIGe… 1.07e8 psd2   involved… GO:0… PMID:223… IEA           InterPro… P     
5 NCBIGe… 1.07e8 cnn2   involved… GO:0… PMID:300… IEA           PANTHER:… P     
6 NCBIGe… 1.07e8 cnn2   located_… GO:0… PMID:300… IEA           PANTHER:… C     
# ℹ 8 more variables: Gene_Name &lt;chr&gt;, Gene_Synonym &lt;lgl&gt;, Type &lt;chr&gt;,
#   Taxon &lt;chr&gt;, Date &lt;dbl&gt;, Assigned_By &lt;chr&gt;, Annot_Ext &lt;lgl&gt;,
#   Gene_Product_Form_ID &lt;lgl&gt;</code></pre>
<p>Qualifiers.</p>
<pre class="r"><code>table(gaf$Qualifier)</code></pre>
<pre><code>
    enables involved_in  located_in     part_of 
      36393       36175       23210        5183 </code></pre>
<p>Evidence code.</p>
<pre class="r"><code>table(gaf$Evidence_Code)</code></pre>
<pre><code>
   IEA 
100961 </code></pre>
<p>Lookup table; check whether lookup is unique.</p>
<pre class="r"><code>library(dplyr)</code></pre>
<pre><code>
Attaching package: &#39;dplyr&#39;</code></pre>
<pre><code>The following object is masked from &#39;package:biomaRt&#39;:

    select</code></pre>
<pre><code>The following objects are masked from &#39;package:stats&#39;:

    filter, lag</code></pre>
<pre><code>The following objects are masked from &#39;package:base&#39;:

    intersect, setdiff, setequal, union</code></pre>
<pre class="r"><code>select(gaf, GeneID, GO_ID) |&gt;
  group_by(GeneID, GO_ID) |&gt;
  summarise(n = n(), .groups = &quot;keep&quot;) |&gt;
  arrange(-n) -&gt; lookup_table

head(lookup_table)</code></pre>
<pre><code># A tibble: 6 × 3
# Groups:   GeneID, GO_ID [6]
     GeneID GO_ID          n
      &lt;dbl&gt; &lt;chr&gt;      &lt;int&gt;
1 107372345 GO:0005975     2
2 107372345 GO:0016020     2
3 107372345 GO:0016758     2
4 107372351 GO:0004674     2
5 107372352 GO:0017025     2
6 107372360 GO:0004930     2</code></pre>
<p>Look at an example of a duplicated entry.</p>
<pre class="r"><code>eg &lt;- head(lookup_table, n = 1)

dplyr::filter(gaf, GeneID == eg$GeneID, GO_ID == eg$GO_ID)</code></pre>
<pre><code># A tibble: 2 × 17
  DB      GeneID Symbol Qualifier GO_ID Reference Evidence_Code With_From Aspect
  &lt;chr&gt;    &lt;dbl&gt; &lt;chr&gt;  &lt;chr&gt;     &lt;chr&gt; &lt;chr&gt;     &lt;chr&gt;         &lt;chr&gt;     &lt;chr&gt; 
1 NCBIGe… 1.07e8 LOC10… involved… GO:0… PMID:223… IEA           InterPro… P     
2 NCBIGe… 1.07e8 LOC10… involved… GO:0… PMID:300… IEA           UniProtK… P     
# ℹ 8 more variables: Gene_Name &lt;chr&gt;, Gene_Synonym &lt;lgl&gt;, Type &lt;chr&gt;,
#   Taxon &lt;chr&gt;, Date &lt;dbl&gt;, Assigned_By &lt;chr&gt;, Annot_Ext &lt;lgl&gt;,
#   Gene_Product_Form_ID &lt;lgl&gt;</code></pre>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span>
Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 4.4.1 (2024-06-14)
Platform: x86_64-pc-linux-gnu
Running under: Ubuntu 22.04.5 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 
LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

time zone: Etc/UTC
tzcode source: system (glibc)

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] dplyr_1.1.4     biomaRt_2.62.1  workflowr_1.7.1

loaded via a namespace (and not attached):
 [1] KEGGREST_1.46.0         xfun_0.48               bslib_0.8.0            
 [4] httr2_1.1.0             processx_3.8.4          Biobase_2.66.0         
 [7] tzdb_0.4.0              callr_3.7.6             generics_0.1.3         
[10] vctrs_0.6.5             tools_4.4.1             ps_1.8.1               
[13] parallel_4.4.1          curl_6.2.1              stats4_4.4.1           
[16] tibble_3.2.1            AnnotationDbi_1.68.0    RSQLite_2.3.9          
[19] blob_1.2.4              pkgconfig_2.0.3         dbplyr_2.5.0           
[22] S4Vectors_0.44.0        lifecycle_1.0.4         GenomeInfoDbData_1.2.13
[25] compiler_4.4.1          stringr_1.5.1           git2r_0.35.0           
[28] Biostrings_2.74.1       progress_1.2.3          getPass_0.2-4          
[31] httpuv_1.6.15           GenomeInfoDb_1.42.3     htmltools_0.5.8.1      
[34] sass_0.4.9              yaml_2.3.10             tidyr_1.3.1            
[37] later_1.3.2             pillar_1.10.1           crayon_1.5.3           
[40] jquerylib_0.1.4         whisker_0.4.1           cachem_1.1.0           
[43] tidyselect_1.2.1        digest_0.6.37           stringi_1.8.4          
[46] purrr_1.0.2             rprojroot_2.0.4         fastmap_1.2.0          
[49] cli_3.6.3               magrittr_2.0.3          utf8_1.2.4             
[52] readr_2.1.5             withr_3.0.2             filelock_1.0.3         
[55] prettyunits_1.2.0       UCSC.utils_1.2.0        promises_1.3.2         
[58] rappdirs_0.3.3          bit64_4.5.2             rmarkdown_2.28         
[61] XVector_0.46.0          httr_1.4.7              bit_4.5.0              
[64] png_0.1-8               hms_1.1.3               memoise_2.0.1          
[67] evaluate_1.0.1          knitr_1.48              IRanges_2.40.1         
[70] BiocFileCache_2.14.0    rlang_1.1.4             Rcpp_1.0.13            
[73] glue_1.8.0              DBI_1.2.3               xml2_1.3.6             
[76] BiocGenerics_0.52.0     vroom_1.6.5             rstudioapi_0.17.1      
[79] jsonlite_1.8.9          R6_2.5.1                fs_1.6.4               
[82] zlibbioc_1.52.0        </code></pre>
</div>
</div>
</div>
</div>
</div>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
https://docs.mathjax.org/en/latest/web/configuration.html. This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>




</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
