---
title: "Using biomaRt to get homologues"
date: "`r Sys.Date()`"
output:
  workflowr::wflow_html:
    toc: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Installation

To begin, install the {biomaRt} package.

```{r install_biomart, eval=FALSE}
if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("biomaRt")
```

## Package

Load package.

```{r load_biomart}
suppressPackageStartupMessages(library(biomaRt))
packageVersion("biomaRt")
```

## Getting started

List the available BioMart databases.

```{r list_marts}
listMarts()
```

Connect to the selected BioMart database by using `useMart()`.

```{r ensembl}
ensembl <- useMart("ENSEMBL_MART_ENSEMBL")
avail_datasets <- listDatasets(ensembl)
head(avail_datasets)
```

Look for human datasets by searching the description column.

```{r human}
idx <- grep('human', avail_datasets$description, ignore.case = TRUE)
avail_datasets[idx, ]
```

Connect to the selected BioMart database and human dataset.

```{r usemart_with_dataset}
ensembl <- useMart("ensembl", dataset=avail_datasets[idx, 'dataset'])
ensembl
```

Building a query, requires three things:

1. filters
2. attributes
3. values

Use `listFilters()` to show available filters.

```{r avail_filters}
avail_filters <- listFilters(ensembl)
head(avail_filters)
```

Use `listAttributes()` to show available attributes.

```{r avail_attributes}
avail_attributes <- listAttributes(ensembl)
head(avail_attributes)
```

Look for mouse homologues.

```{r look_for_homolog}
grep('homolog', avail_attributes$name, ignore.case = TRUE, value = TRUE) |>
  grep('mmus', x = _, ignore.case = TRUE, value = TRUE) -> wanted_attr

wanted_attr <- c('ensembl_gene_id', wanted_attr)
wanted_attr
```

ENSG00000206172 (HBA1).

```{r hba1}
my_gene <- 'ENSG00000206172'

getBM(
  attributes = wanted_attr,
  filters = "ensembl_gene_id",
  values = my_gene,
  mart = ensembl
) -> my_res

t(my_res)
```

I [manually slugged through the Compara database](https://github.com/davetang/ensembldb?tab=readme-ov-file#database) and found that ENSG00000207721 (MIR186) should have a one-to-one ortholog with ENSMUSG00000065431 (Mir186).

```{r mir186}
my_gene <- 'ENSG00000207721'

getBM(
  attributes = wanted_attr,
  filters = "ensembl_gene_id",
  values = my_gene,
  mart = ensembl
) -> my_res

t(my_res)
```

## Older release

List releases.

```{r list_archives}
listEnsemblArchives()
```

Use Ensembl 113.

```{r use_ensembl_113}
ensembl <- useMart(
  "ENSEMBL_MART_ENSEMBL",
  dataset=avail_datasets[idx, 'dataset'],
  host = "https://oct2024.archive.ensembl.org"
)
```

Check again.

```{r mir186_113}
my_gene <- 'ENSG00000207721'

getBM(
  attributes = wanted_attr,
  filters = "ensembl_gene_id",
  values = my_gene,
  mart = ensembl
) -> my_res

t(my_res)
```

Get all human genes for Ensembl 113.

```{r all_genes_113}
all_genes_113 <- getBM(
  attributes = "ensembl_gene_id",
  mart = ensembl
)

length(unique(all_genes_113$ensembl_gene_id))
```

Get all homologues.

```{r human_mouse_homologues}
getBM(
  attributes = wanted_attr,
  filters = "ensembl_gene_id",
  values = all_genes_113$ensembl_gene_id,
  mart = ensembl
) -> human_mouse_homologues

dim(human_mouse_homologues)
```

Check out the results!

```{r check_out_human_mouse_homologues}
head(human_mouse_homologues)
```

```{r include=FALSE, eval=FALSE}
readr::write_csv(x = human_mouse_homologues, file = "../data/ensembl_113_human_mouse_homologues.csv")
```
