---
title: "Pokedex for Cell Types"
date: "`r Sys.Date()`"
output:
  workflowr::wflow_html:
    toc: true
editor_options:
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

> The celldex package provides convenient access to several cell type reference datasets. Most of these references are derived from bulk RNA-seq or microarray data of cell populations that (hopefully) consist of a pure cell type after sorting and/or culturing. The aim is to provide a common resource for further analysis like cell type annotation of single cell expression data or deconvolution of cell type proportions in bulk expression datasets.

## Installation

Install [celldex](https://bioconductor.org/packages/3.20/data/experiment/html/celldex.html).

```{r install_celldex, eval=FALSE}
if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("celldex")
```

## References

We can examine the available references using the `surveyReferences()` function. This returns a DataFrame of the reference's name and version, along with additional information like the title, description, species, number of samples, available labels, and so on.

```{r survey_references}
suppressPackageStartupMessages(library(celldex))
surveyReferences()
```

Alternatively, users can search the text of each reference's metadata to identify relevant datasets. This may require some experimentation as it depends on the level of detail in the metadata supplied by the uploader.

```{r search_references}
searchReferences(
    defineTextQuery("immun%", partial=TRUE) & 
    defineTextQuery("9606", field="taxonomy_id")
)
```

### Human primary cell atlas

The HPCA reference consists of publicly available microarray datasets derived from human primary cells (Mabbott et al. 2013). Most of the labels refer to blood subpopulations but cell types from other tissues are also available.

```{r hpca}
hpca <- fetchReference("hpca", "2024-02-26")
colData(hpca) |>
  as.data.frame() |>
  dplyr::pull("label.main") |>
  table() |>
  as.data.frame() |>
  dplyr::arrange(-Freq)
```

### Monaco reference

The [Monaco reference](https://bioconductor.org/packages/3.20/data/experiment/vignettes/celldex/inst/doc/userguide.html#44_Monaco_immune_data) consists of bulk RNA-seq samples of sorted immune cell populations from GSE107011 (Monaco et al. 2019).

This is the human immune reference that best covers all of the bases for a typical PBMC sample. It provides expansive B and T cell subsets, differentiates between classical and non-classical monocytes, includes basic dendritic cell subsets, and also includes neutrophil and basophil samples to help identify small contaminating populations that may have slipped into a PBMC preparation.

```{r monaco_immune}
monaco_immune <- fetchReference("monaco_immune", "2024-02-26")
colData(monaco_immune) |>
  as.data.frame()
```
