---
title: "Getting started with clusterProfiler"
date: "`r Sys.Date()`"
output:
  workflowr::wflow_html
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

options(repos=c(CRAN="https://cran.r-project.org"))
# getOption("repos")
knitr::opts_knit$set(root.dir = rprojroot::find_rstudio_root_file())
```

[clusterProfiler](https://bioconductor.org/packages/release/bioc/html/clusterProfiler.html):

> This package supports functional characteristics of both coding and non-coding genomics data for thousands of species with up-to-date gene annotation. It provides a univeral interface for gene functional annotation from a variety of sources and thus can be applied in diverse scenarios. It provides a tidy interface to access, manipulate, and visualize enrichment results to help users achieve efficient data interpretation. Datasets obtained from multiple treatments and time points can be analyzed and compared in a single run, easily revealing functional consensus and differences among distinct conditions.

## Getting started

Install {clusterProfiler} and some dependencies.

```{r install_cluster_profiler, eval=FALSE}
if (!require("BiocManager", quietly = TRUE))
  install.packages("BiocManager")

install.packages('ggarchery')
install.packages('ggtangle')
BiocManager::install("clusterProfiler")
BiocManager::install("org.Hs.eg.db")
BiocManager::install("GO.db")
```

Load required libraries.

```{r load_libraries}
suppressPackageStartupMessages(library(tidyverse))
suppressPackageStartupMessages(library(ggarchery))
suppressPackageStartupMessages(library(clusterProfiler))
suppressPackageStartupMessages(library(org.Hs.eg.db))
suppressPackageStartupMessages(library(GO.db))
```

## Example data

We will use [An example differential gene expression results table](https://davetang.org/muse/2024/10/31/an-example-differential-gene-expression-results-table/) that contains {edgeR} results comparing normal versus cancer samples.

```{r de_res}
de_res <- read_csv("https://raw.githubusercontent.com/davetang/muse/refs/heads/main/data/13970886_edger_res.csv", show_col_types = FALSE)
head(de_res)
```

## Gene IDs

The {clusterProfiler} package uses the `enrichGO()` function for performing a [Gene Ontology over-representation test](https://yulab-smu.top/biomedical-knowledge-mining-book/clusterprofiler-go.html#clusterprofiler-go-ora). The arguments are:

* `gene` - a vector of entrez gene id.
* `OrgDb` - OrgDb database
* `keyType` - keytype of input gene
* `ont` - One of "BP", "MF", and "CC" subontologies, or "ALL" for all three.
* `pvalueCutoff` - adjusted pvalue cutoff on enrichment tests to report
* `pAdjustMethod` - one of "holm", "hochberg", "hommel", "bonferroni", "BH", "BY", "fdr", "none"
* `universe` - background genes. If missing, the all genes listed in the database (e.g. TERM2GENE table) will be used as background.
* `qvalueCutoff` - qvalue cutoff on enrichment tests to report as significant. Tests must pass i) pvalueCutoff on unadjusted pvalues, ii) pvalueCutoff on adjusted pvalues and iii) qvalueCutoff on qvalues to be reported.
* `minGSSize` - minimal size of genes annotated by Ontology term for testing.
* `maxGSSize` - maximal size of genes annotated for testing
* `readable` - whether mapping gene ID to gene Name
* `pool` - If ont='ALL', whether pool 3 GO sub-ontologies

Our example data uses Ensembl gene IDs, so we need to convert these into Entrez Gene IDs. We can use the {org.Hs.eg.db} package provided by [Bioconductor](https://bioconductor.org/packages/release/data/annotation/html/org.Hs.eg.db.html), which provides genome wide annotation for human, primarily based on mapping using Entrez Gene identifiers. 

```{r ensembl_to_entrez}
ensembl_to_entrez <- AnnotationDbi::select(
  org.Hs.eg.db,
  keys = de_res$ensembl_gene_id,
  columns = c("ENSEMBL", "ENTREZID"), 
  keytype = "ENSEMBL"
)

de_res |>
  dplyr::inner_join(ensembl_to_entrez, by = dplyr::join_by(ensembl_gene_id == ENSEMBL)) |>
  dplyr::select(ensembl_gene_id, ENTREZID, dplyr::everything()) -> de_res

head(de_res)
```

## clusterProfiler

The example data contains results of a comparison between cancer samples and normal samples; positive fold change indicates that genes were expressed higher in cancer. There are a lot of genes up-regulated in cancer.

```{r up_regulated_in_cancer}
de_res |>
  dplyr::filter(logFC > 0, adjusted_pvalue < 0.05) |>
  nrow()
```

Instead of examining the list of up-regulated genes individually, which would take a long time, we can see whether these genes are associated with Gene Ontology terms more often than expected. Our expectation will be based on the background.

First we'll get the top 500 (an arbitrary number) most significantly up-regulated genes.

```{r sig_up_genes}
de_res |>
  dplyr::filter(logFC > 0, adjusted_pvalue < 0.05) |>
  dplyr::slice_min(order_by = adjusted_pvalue, n = 500) |>
  dplyr::filter(!is.na(ENTREZID)) |>
  dplyr::pull(ENTREZID) -> sig_up_genes
```

Create the background, i.e., the universe, using all genes tested for differential expression so we can test for over-representation.

```{r the_universe}
de_res |>
  dplyr::filter(!is.na(ENTREZID)) |>
  dplyr::pull(ENTREZID) -> the_universe
```

Gene Ontology terms are grouped into three ontologies:

1. Molecular Function (MF)
2. Biological Process (BP), and
3. Cellular Component (CC)

Here, we'll perform an over-representation analysis using biological processes.

```{r ego_bp}
ego_bp <- enrichGO(
  gene          = sig_up_genes,
  universe      = the_universe,
  OrgDb         = org.Hs.eg.db,
  ont           = "BP",
  pAdjustMethod = "BH",
  pvalueCutoff  = 0.01,
  qvalueCutoff  = 0.05,
  minGSSize     = 10,
  maxGSSize     = 500,
  readable      = TRUE
)

head(ego_bp)
```

Here, we'll perform an over-representation analysis using molecular functions.

```{r ego_mf}
ego_mf <- enrichGO(
  gene          = sig_up_genes,
  universe      = the_universe,
  OrgDb         = org.Hs.eg.db,
  ont           = "MF",
  pAdjustMethod = "BH",
  pvalueCutoff  = 0.01,
  qvalueCutoff  = 0.05,
  minGSSize     = 10,
  maxGSSize     = 500,
  readable      = TRUE
)

head(ego_mf)
```

### Examining the results

Results are stored in an `enrichResult` class.

```{r enrich_result_obj}
class(ego_mf)
```

Slot names of the class.

```{r enrich_result_slot}
slotNames(ego_mf)
```

One question that has been asked (and not conclusively answered) on the interweb is why do the universe lengths differ when the same list of background genes is provided?

```{r universe_lengths}
length(the_universe)
length(ego_mf@universe)
length(ego_bp@universe)
```

In order to address this, we will first create a Entrez Gene ID to GO term lookup table.

```{r entrez_to_go}
entrez_to_go <- select(
  org.Hs.eg.db,
  keys = keys(org.Hs.eg.db, keytype = "ENTREZID"),
  columns = c("GO", "ONTOLOGY"),
  keytype = "ENTREZID"
)

head(entrez_to_go)
```

Intuitively (to me), the only reason why the universe lengths are different per ontology group is because there is some sort of filtering performed per ontology group. One gene is associated to various GO terms and if you group them into the three respective groups, their numbers might be different. Here's an example.

```{r entrez_gene_1}
entrez_to_go |>
  dplyr::filter(ENTREZID == "1") |>
  dplyr::summarise(n = n(), .by = ONTOLOGY)
```

Let's reproduce the universe list for molecular functions. The pipeline below is:

* Filter the Entrez Gene ID to GO term lookup table to only contain GO terms in the MF group
* Output only the `ENTREZID` and `GO` columns
* Keep only the unique rows
* Keep only the Entrez Gene IDs in our background
* Output all the Entrez Gene IDs
* Keep only unique IDs
* Calculate the length

```{r universe_lengths_mf}
length(ego_mf@universe)

dplyr::filter(entrez_to_go, ONTOLOGY == "MF") |>
  dplyr::select(ENTREZID, GO) |>
  dplyr::distinct() |>
  dplyr::filter(ENTREZID %in% the_universe) |>
  dplyr::pull(ENTREZID) |>
  unique() |>
  length()
```

Repeat the same workflow on biological processes.

```{r universe_lengths_bp}
length(ego_bp@universe)

dplyr::filter(entrez_to_go, ONTOLOGY == "BP") |>
  dplyr::select(ENTREZID, GO) |>
  dplyr::distinct() |>
  dplyr::filter(ENTREZID %in% the_universe) |>
  dplyr::pull(ENTREZID) |>
  unique() |>
  length()
```

Getting the same lengths confirms our intuition that gene lists are created by only keeping genes that have at least one GO term in the Gene Ontology group of interest.

### Visualisations

Bar plot showing each enriched GO term coloured by the adjusted p-value.

```{r barplot}
barplot(ego_bp, showCategory=10)
```

Dot plot showing each enriched GO term with associated statistics.

```{r dotplot}
dotplot(ego_bp, showCategory=10)
```

Heat plot showing the enriched GO terms on the y-axis and the genes on the x-axis. Genes with the associated GO term are highlighted.

```{r heatplot}
heatplot(ego_bp, showCategory=10)
```

`goplot` shows the gene ontology graph with the enriched GO terms highlighted.

```{r goplot, fig.height=10, fig.width=12}
goplot(ego_bp)
```

Another nice feature of {clusterProfiler} is that you can plot multiple gene lists together. We can create a list of down-regulated genes.

```{r sig_down_genes}
de_res |>
  dplyr::filter(logFC < 0, adjusted_pvalue < 0.05) |>
  dplyr::slice_min(order_by = adjusted_pvalue, n = 500) |>
  dplyr::filter(!is.na(ENTREZID)) |>
  dplyr::pull(ENTREZID) -> sig_down_genes
```

Perform GO enrichment on two gene lists.

```{r ego_bp_both}
my_gene_list <- list(
  up_gene = sig_up_genes,
  down_gene = sig_down_genes
)

ego_bp_both <- compareCluster(
  geneCluster = my_gene_list,
  fun = "enrichGO",
  universe = the_universe,
  OrgDb = org.Hs.eg.db,
  keyType = "ENTREZID",
  ont = "BP",
  pvalueCutoff = 0.01,
  pAdjustMethod = "BH",
  qvalueCutoff = 0.05,
  minGSSize = 10,
  maxGSSize = 500,
  readable = TRUE
)

head(as.data.frame(ego_bp_both))
```

Dot plot with enriched GO terms by gene list.

```{r dotplot_gene_list, fig.width=12, fig.height=12}
dotplot(ego_bp_both, showCategory = 15)
```

## enricher

`enricher()` is a universal enrichment analyzer.

* gene - a vector of gene id
* universe - background genes. If missing, the all genes listed in the database (eg TERM2GENE table) will be used as background.
* minGSSize - minimal size of genes annotated for testing
* maxGSSize - maximal size of genes annotated for testing
* TERM2GENE - user input annotation of TERM TO GENE mapping, a data.frame of 2 column with term and gene. Only used when gson is NULL.
* TERM2NAME - user input of TERM TO NAME mapping, a data.frame of 2 column with term and name. Only used when gson is NULL.

[TERM2NAME](https://support.bioconductor.org/p/9144621/) is needed to map the GOID to its term (description).

An easy way of obtaining this mapping is by extracting the information from the Bioconductor GO annotation database package ({GO.db}).

```{r term2name}
# extract a named vector of all terms
goterms <- AnnotationDbi::Term(GOTERM)

#convert into a data frame
term2name <- data.frame(
  "term"=names(goterms),
  "name"=goterms
)

dim(term2name)
head(term2name)
```

Get the ontologies and store in `term2name`.

```{r get_ontology}
ontologies <- AnnotationDbi::select(x = GO.db, keys = names(goterms), columns = c("GOID", "ONTOLOGY"))

length(unique(term2name$term))
length(unique(ontologies$GOID))

stopifnot(all(term2name$term == ontologies$GOID))

term2name$ontology <- ontologies$ONTOLOGY
head(term2name)

any(is.na(term2name$term))
```

`term2name` can then be used when calling `enrichr()`, by specifying `TERM2NAME=term2name`. Be sure, though, to check that `term2name` contains all GOIDs present in your TERM2GENE mapping.

Demo input.

```{r demo}
demo <- readr::read_csv("data/nfurzeri_gene_id_to_go_id.csv.gz", show_col_types = FALSE)
head(demo)
```

Check if we have all the GO IDs.

```{r check_go_id}
goids <- demo$go_id
goids <- goids[!is.na(goids)]

table(unique(goids) %in% term2name$term)
```

Which GO IDs are missing?

```{r missing_goids}
missing_goids <- setdiff(unique(goids), term2name$term)
missing_goids
```

They are missing because they have become obsolete.

```{r obsolete}
missing_goids %in% keys(GOOBSOLETE)
```

Prepare data frame of 2 column with term and gene.

```{r term2gene}
demo |>
  dplyr::filter(!is.na(go_id)) |>
  dplyr::rename(term = go_id, gene = ensembl_gene_id) |>
  dplyr::select(term, gene) |>
  dplyr::arrange(term) -> term2gene

lookup <- AnnotationDbi::select(x = GO.db, keys = term2gene$term, columns = c("GOID", "ONTOLOGY"))
stopifnot(all(lookup$GOID == term2gene$term))

term2gene$ontology <- lookup$ONTOLOGY

head(term2gene)
```

Test with random genes.

```{r enricher_random}
set.seed(1984)
my_genes <- sample(x = unique(term2gene$gene), size = 100, replace = FALSE)

res <- enricher(
  gene = my_genes,
  TERM2GENE = dplyr::filter(term2gene, ontology == "BP"),
  TERM2NAME = dplyr::filter(term2name, ontology == "BP")
)

res
```

Test with genes associated with same term.

```{r enricher_same_term}
term2gene |>
  dplyr::group_by(term) |>
  dplyr::summarise(n = n()) |>
  dplyr::arrange(-n) -> dev_null

my_term <- "GO:0006397"
dplyr::filter(term2name, term == my_term)

term2gene |>
  dplyr::filter(term == my_term) |>
  dplyr::pull(gene) |>
  head(75) -> enriched_genes

set.seed(1984)
my_genes <- sample(x = unique(term2gene$gene), size = 25, replace = FALSE)
my_genes <- union(enriched_genes, my_genes)

res <- enricher(
  gene = my_genes,
  TERM2GENE = dplyr::filter(term2gene, ontology == "BP"),
  TERM2NAME = dplyr::filter(term2name, ontology == "BP")
)

res |>
  as.data.frame()
```

Check some of the results.

```{r check_res}
go_term <- "GO:0007156"

dplyr::filter(term2name, term == go_term)

dplyr::filter(term2gene, gene %in% my_genes, term == go_term) |>
  nrow()

dplyr::filter(term2gene, term == go_term) |>
  nrow()
```

Dot plot.

```{r dotplot_res, fig.height=7, fig.width=8}
dotplot(res, showCategory=10) +
  ggtitle("Biological Processes")
```
