---
title: "Differential gene expression analysis with DESeq2"
date: "`r Sys.Date()`"
output:
  workflowr::wflow_html:
    toc: true
editor_options:
  chunk_output_type: console
---

```{r setup, include=FALSE}
library(tidyverse)
library(DESeq2)
library(pasilla)
library(matrixStats)
library(hexbin)
knitr::opts_chunk$set(echo = TRUE)
```

[DESeq2](https://bioconductor.org/packages/release/bioc/html/DESeq2.html) is used to:

> Estimate variance-mean dependence in count data from high-throughput sequencing assays and test for differential expression based on a model using the negative binomial distribution. 

## Installation

Install using `BiocManager::install()`.

```{r install_edger, eval=FALSE}
if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("DESeq2")
```

We will use data from the [pasilla package](https://www.bioconductor.org/packages/release/data/experiment/html/pasilla.html) so install it too.

```{r install_pasilla, eval=FALSE}
BiocManager::install("pasilla")
```

## Example data

Example dataset in the experiment data package {pasilla}.

```{r counts}
fn <- system.file("extdata", "pasilla_gene_counts.tsv", package = "pasilla", mustWork = TRUE)
counts <- as.matrix(read.csv(fn, sep="\t", row.names = "gene_id"))
dim(counts)
```

The matrix tallies the number of reads assigned for each gene in each sample.

```{r tail_counts}
tail(counts)
```

## Size factors

Estimate size factors.

```{r est_sz_fc}
DESeq2::estimateSizeFactorsForMatrix(counts)
```

## Mean-variance relationship

Variance versus mean for the (size factor adjusted) `counts` data. The axes are logarithmic. Also shown are lines through the origin with slopes 1 (green) and 2 (red).

```{r mean_variance}
sf <- DESeq2::estimateSizeFactorsForMatrix(counts)

ncounts <- t(t(counts) / sf)

# untreated samples
uncounts <- ncounts[, grep("^untreated", colnames(ncounts)), drop = FALSE]

ggplot(
  tibble(
    mean = rowMeans(uncounts),
    var  = rowVars(uncounts)
  ),
  aes(x = log1p(mean), y = log1p(var))
) +
  geom_hex() +
  coord_fixed() +
  theme_minimal() +
  theme(legend.position = "none") +
  geom_abline(slope = 1:2, color = c("forestgreen", "red"))
```

The green line is what we expect if the variance equals the mean, as is the case for a Poisson-distributed random variable. This approximately fits the data in the lower range. The red line corresponds to the quadratic mean-variance relationship $v = m^2$. We can see that in the upper range of the data, the quadratic relationship approximately fits the data.

## A basic analysis

The {pasilla} data set is from an experiment on _Drosophila melanogaster_ cell cultures that tested the effect of RNAi knockdown of the splicing factor _pasilla_ on the cells' transcriptome. There were two experimental conditions, termed _untreated_ and _treated_ in the column headers and they correspond to negative control and siRNA against _pasilla_. The experimental metadata of the seven samples in this dataset are loaded below.

```{r pasilla_sample_anno}
annotationFile <- system.file("extdata", "pasilla_sample_annotation.csv", package = "pasilla", mustWork = TRUE)
pasillaSampleAnno <- readr::read_csv(annotationFile)
pasillaSampleAnno
```

The dataset was produced in two batches, the first consisting of three sequencing libraries using single read sequencing and the second using using paired-end sequencing.

Create factors.

```{r format_pasilla_sample_anno}
mutate(
  pasillaSampleAnno,
  condition = factor(condition, levels = c("untreated", "treated")),
  type = factor(sub("-.*", "", type), levels = c("single", "paired"))
) -> pasillaSampleAnno
```

Note that the design is approximately balanced between the factor of interest, `condition`, and the "nuisance factor", `type`.

```{r table_interest_versus_nuisance}
with(pasillaSampleAnno, table(condition, type))
```

{DESeq2} uses `DESeqDataSet` objects and is an extension of the `SummarizedExperiment` class in Bioconductor. We can create a `DESeqDataSet` object using the constructor function `DESeqDataSetFromMatrix`.

```{r create_deseq_data_set}
mt <- match(colnames(counts), sub("fb$", "", pasillaSampleAnno$file))
stopifnot(!any(is.na(mt)))

pasilla <- DESeqDataSetFromMatrix(
  countData = counts,
  colData   = pasillaSampleAnno[mt, ],
  design    = ~ condition
)

class(pasilla)
is(pasilla, "SummarizedExperiment")
```

After creating a `DESeqDataSet` we are ready to carry out a differential expression analysis. The aim is to identify genes that are differentially abundant between the treated and untreated cells. A test that is conceptually similar to the _t_-test is used. A choice of standard analysis steps are wrapped into a single function, `DESeq()`.

```{r deseq}
pasilla <- DESeq(pasilla)
```

The `DESeq()` function is a wrapper that calls:

* `estimateSizeFactors` for normalisation
* `estimateDispersions` for dispersion estimation
* `nbinomWaldTest` for hypothesis tests for differential abundance

The test is between the two levels `untreated` and `treated` of the factor `condition`, since this is what was specified in the design: `design = ~ condition`.

```{r deseq2_res}
res <- results(pasilla)
res[order(res$padj), ] |> head()
```

### Exploring the results

The first step after a differential expression analysis is to visualise the following plots:

* the histogram of p-values
* the MA plot
* an ordination plot
* a heatmap

Histogram of p-values of a differential expression analysis."

```{r pvalue_hist}
ggplot(as(res, "data.frame"), aes(x = pvalue)) +
  geom_histogram(binwidth = 0.01, fill = "Royalblue", boundary = 0) +
  geom_hline(yintercept = 100, lty = 3) +
  theme_minimal()
```

The p-value distribution shows a uniform background (non-DE genes) with values between 0 and 1 and a peak of small p-values at the left (DE genes). The ratio of the level of the background to the height of the peak provides us with a rough indication of the false discovery rate (FDR) that would be associated with calling the genes in the leftmost bin differentially expressed.

In this case, the leftmost bin contains 993 genes and the background level (dotted line) is around 100, so the FDR associated with calling all genes in the leftmost bin would be around 10%.

```{r fdr_est}
table(res$pvalue < 0.01)
```

A background distribution that is not uniform, but shows a tilted shape with an increase towards the right, tends to indicate batch effects, that is, underlying systematic variation that makes replicates look more different than expected.

The {DESeq2} package can produce the MA plot.

```{r plot_ma}
plotMA(pasilla, ylim = c(-2, 2))
```

{DESeq2} also has a function `plotPCA()` that can produce PCA plots. PC1 is mostly aligned with the experimental covariate of interest (`untreated`/`treated`), while PC2 is roughly aligned with the sequencing protocol (`single`/`paired`).

```{r deseq2_pca}
# transform using regularised logarithm
pas_rlog <- rlogTransformation(pasilla)
plotPCA(pas_rlog, intgroup=c("condition", "type")) + coord_fixed()
```

Heatmap using the regularised logarithm (rlog) transformed data of the 30 most variable genes.

```{r pheatmap, fig.height=10, fig.width=8}
library("pheatmap")
wanted <- order(rowMeans(assay(pas_rlog)), decreasing = TRUE)[1:30]

pheatmap(
  assay(pas_rlog)[wanted, ],
  scale = "row",
  annotation_col = as.data.frame(
    colData(pas_rlog)[, c("condition", "type")]
  )
)
```

Note the clustering of the sequencing protocol and this highlights that our differential expression analysis was probably too naive and that we should adjust for the "nuisance" factor.
