---
title: "Differential gene expression analysis with DESeq2"
date: "`r Sys.Date()`"
output:
  workflowr::wflow_html:
    toc: true
editor_options:
  chunk_output_type: console
---

```{r setup, include=FALSE}
library(tidyverse)
library(DESeq2)
library(pasilla)
library(matrixStats)
library(hexbin)
knitr::opts_chunk$set(echo = TRUE)
```

[DESeq2](https://bioconductor.org/packages/release/bioc/html/DESeq2.html) is used to:

> Estimate variance-mean dependence in count data from high-throughput sequencing assays and test for differential expression based on a model using the negative binomial distribution. 

## Installation

Install using `BiocManager::install()`.

```{r install_edger, eval=FALSE}
if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("DESeq2")
```

We will use data from the [pasilla package](https://www.bioconductor.org/packages/release/data/experiment/html/pasilla.html) so install it too.

```{r install_pasilla, eval=FALSE}
BiocManager::install("pasilla")
```

## Example data

Example dataset in the experiment data package {pasilla}.

```{r counts}
fn <- system.file("extdata", "pasilla_gene_counts.tsv", package = "pasilla", mustWork = TRUE)
counts <- as.matrix(read.csv(fn, sep="\t", row.names = "gene_id"))
dim(counts)
```

The matrix tallies the number of reads assigned for each gene in each sample.

```{r tail_counts}
tail(counts)
```

## Size factors

Estimate size factors.

```{r est_sz_fc}
DESeq2::estimateSizeFactorsForMatrix(counts)
```

## Mean-variance relationship

Variance versus mean for the (size factor adjusted) `counts` data. The axes are logarithmic. Also shown are lines through the origin with slopes 1 (green) and 2 (red).

```{r mean_variance}
sf <- DESeq2::estimateSizeFactorsForMatrix(counts)

ncounts <- t(t(counts) / sf)

# untreated samples
uncounts <- ncounts[, grep("^untreated", colnames(ncounts)), drop = FALSE]

ggplot(
  tibble(
    mean = rowMeans(uncounts),
    var  = rowVars(uncounts)
  ),
  aes(x = log1p(mean), y = log1p(var))
) +
  geom_hex() +
  coord_fixed() +
  theme_minimal() +
  theme(legend.position = "none") +
  geom_abline(slope = 1:2, color = c("forestgreen", "red"))
```

The green line is what we expect if the variance equals the mean, as is the case for a Poisson-distributed random variable. This approximately fits the data in the lower range. The red line corresponds to the quadratic mean-variance relationship $v = m^2$. We can see that in the upper range of the data, the quadratic relationship approximately fits the data.

## A basic analysis

The {pasilla} data set is from an experiment on _Drosophila melanogaster_ cell cultures that tested the effect of RNAi knockdown of the splicing factor _pasilla_ on the cells' transcriptome. There were two experimental conditions, termed _untreated_ and _treated_ in the column headers and they correspond to negative control and siRNA against _pasilla_. The experimental metadata of the seven samples in this dataset are loaded below.

```{r pasilla_sample_anno}
annotationFile <- system.file("extdata", "pasilla_sample_annotation.csv", package = "pasilla", mustWork = TRUE)
pasillaSampleAnno <- readr::read_csv(annotationFile)
pasillaSampleAnno
```

The dataset was produced in two batches, the first consisting of three sequencing libraries using single read sequencing and the second using using paired-end sequencing.

Create factors.

```{r format_pasilla_sample_anno}
mutate(
  pasillaSampleAnno,
  condition = factor(condition, levels = c("untreated", "treated")),
  type = factor(sub("-.*", "", type), levels = c("single", "paired"))
) -> pasillaSampleAnno
```

Note that the design is approximately balanced between the factor of interest, `condition`, and the "nuisance factor", `type`.

```{r table_interest_versus_nuisance}
with(pasillaSampleAnno, table(condition, type))
```

{DESeq2} uses `DESeqDataSet` objects and is an extension of the `SummarizedExperiment` class in Bioconductor. We can create a `DESeqDataSet` object using the constructor function `DESeqDataSetFromMatrix`.

```{r create_deseq_data_set}
mt <- match(colnames(counts), sub("fb$", "", pasillaSampleAnno$file))
stopifnot(!any(is.na(mt)))

pasilla <- DESeqDataSetFromMatrix(
  countData = counts,
  colData   = pasillaSampleAnno[mt, ],
  design    = ~ condition
)

class(pasilla)
is(pasilla, "SummarizedExperiment")
```

After creating a `DESeqDataSet` we are ready to carry out a differential expression analysis. The aim is to identify genes that are differentially abundant between the treated and untreated cells. A test that is conceptually similar to the _t_-test is used. A choice of standard analysis steps are wrapped into a single function, `DESeq()`.

```{r deseq}
pasilla <- DESeq(pasilla)
```

The `DESeq()` function is a wrapper that calls:

* `estimateSizeFactors` for normalisation
* `estimateDispersions` for dispersion estimation
* `nbinomWaldTest` for hypothesis tests for differential abundance

The test is between the two levels `untreated` and `treated` of the factor `condition`, since this is what was specified in the design: `design = ~ condition`.

```{r deseq2_res}
res <- results(pasilla)
res[order(res$padj), ] |> head()
```

### Exploring the results

The first step after a differential expression analysis is to visualise the following plots:

* the histogram of p-values
* the MA plot
* an ordination plot
* a heatmap

Histogram of p-values of a differential expression analysis."

```{r pvalue_hist}
ggplot(as(res, "data.frame"), aes(x = pvalue)) +
  geom_histogram(binwidth = 0.01, fill = "Royalblue", boundary = 0) +
  geom_hline(yintercept = 100, lty = 3) +
  theme_minimal()
```

The p-value distribution shows a uniform background (non-DE genes) with values between 0 and 1 and a peak of small p-values at the left (DE genes). The ratio of the level of the background to the height of the peak provides us with a rough indication of the false discovery rate (FDR) that would be associated with calling the genes in the leftmost bin differentially expressed.

In this case, the leftmost bin contains 993 genes and the background level (dotted line) is around 100, so the FDR associated with calling all genes in the leftmost bin would be around 10%.

```{r fdr_est}
table(res$pvalue < 0.01)
```

A background distribution that is not uniform, but shows a tilted shape with an increase towards the right, tends to indicate batch effects, that is, underlying systematic variation that makes replicates look more different than expected.

The {DESeq2} package can produce the MA plot.

```{r plot_ma}
plotMA(pasilla, ylim = c(-2, 2))
```

{DESeq2} also has a function `plotPCA()` that can produce PCA plots. PC1 is mostly aligned with the experimental covariate of interest (`untreated`/`treated`), while PC2 is roughly aligned with the sequencing protocol (`single`/`paired`).

```{r deseq2_pca}
# transform using regularised logarithm
pas_rlog <- rlogTransformation(pasilla)
plotPCA(pas_rlog, intgroup=c("condition", "type")) + coord_fixed()
```

Heatmap using the regularised logarithm (rlog) transformed data of the 30 most variable genes.

```{r pheatmap, fig.height=10, fig.width=8}
library("pheatmap")
wanted <- order(rowMeans(assay(pas_rlog)), decreasing = TRUE)[1:30]

pheatmap(
  assay(pas_rlog)[wanted, ],
  scale = "row",
  annotation_col = as.data.frame(
    colData(pas_rlog)[, c("condition", "type")]
  )
)
```

Note the clustering of the sequencing protocol and this highlights that our differential expression analysis was probably too naive and that we should adjust for the "nuisance" factor.

## Multifactorial designs and linear models

Let's assume that, in addition to the siRNA knockdown of the pasilla gene, we also want to test the effect of a certain drug. We would then envisage an experiment in which the experimenter treats the cells either with negative control, with the siRNA against pasilla, with the drug, or with both. To analyse this experiment, we can introduce the notation:

\begin{equation}
   y = \beta_0 + x_1 \beta_1 + x_2 \beta_2 + x_1 x_2 \beta_{12}
\end{equation} 

The left-hand size, $y$, is the experimental measurement of interest, which in our case is the suitably transformed expression level of a gene. Since in an RNA-seq experiment there are lots of genes, we'll have as many copies of the above equation, one for each.

The coefficient $\beta_0$ is the base level of the measurement in the negative control; often it is called the **intercept**.

The design factors $x_1$ and $x_2$ are binary indicator variables: $x_1$ takes the value 1 if the siRNA was transfected and 0 if not, and similarly, $x_2$ indicates whether the drug was administered.

In the experiment where only where only the siRNA is used, $x_1$ = 1 and $x_2$ = 0, and the third and fourth terms of the equation vanish. The equation then simplifies to $y = \beta_0 + \beta_1$. This means that $\beta_1$ represents the difference between treatment and control. If our measurements are on a logarithmic scale, then:

$$
y = \beta_0 + \beta_1 \\
\beta_1 = y - \beta_0 = log_2(expression_{treated}) - log2(expression_{untreated}) \\
= log_2\frac{expression_{treated}}{expression_{untreated}}
$$

is the logarithmic fold change due to treatment with the siRNA. In exactly the same way, $\beta_2$ is the logarithmic fold change due to treatment with the drug.

If we treat the cells with both siRNA and drug, $x_1$ = $x_2$ = 1:

$$
y = \beta_0 + x_1 \beta_1 + x_2 \beta_2 + x_1 x_2 \beta_{12} \\
y = \beta_0 + \beta_1 + \beta_2 + \beta_{12} \\
\beta_{12} = y - (\beta_0 + \beta_1 + \beta_2)
$$

This means that $\beta_{12}$ is the difference between the observed outcome, $y$, and the outcome expected from the individual treatments, obtained by adding to the baseline the effect of siRNA alone, $\beta_1$, and of drug alone, $\beta_2$.

We call $\beta_{12}$ the _interaction_ effect of siRNA and drug. It has nothing to do with a physical interaction: the terminology indicates that the effects of these two different experimental factors do not simply add up, but combine in a more complicated fashion. For instance, if the target of the drug and of the siRNA were equivalent, leading to the same effect on the cells, then we would biologically expect that $\beta_1$ = $\beta_2$. We would also expect that their combination has no further effect, so that $\beta_{12}$ = $-\beta_1$.

If, on the other hand, the targets of the drug and of the siRNA are in parallel pathways that can buffer each other, we would expect that $\beta_1$ and $\beta_2$ are both relatively small, but the combined effect is synergistic, and $\beta_{12}$ is large.

We don't always care about interactions. Many experiments are designed with multiple factors where we care most about each of their individual effects. In that case, the combinatorial treatment might not be present in the experimental design, and the model to use for the analysis is a version of the equation with the rightmost term removed.

We can succinctly encode the design of the experiment in the **design matrix**. The combinatorial experiment would be:

| $x_0$ | $x_1$ | $x_2$ |
| :---: | :---: | :---: |
| 1     | 0     | 0     |
| 1     | 1     | 0     |
| 1     | 0     | 1     |
| 1     | 1     | 1     |

The columns of the design matrix correspond to the experimental factors, and its rows represent the different experimental conditions, four in this case. If the combinatorial treatment is not performed, then the design matrix is reduced to only the first three rows.

## Noise and replicates

The equation:

\begin{equation}
   y = \beta_0 + x_1 \beta_1 + x_2 \beta_2 + x_1 x_2 \beta_{12}
\end{equation} 

provides a conceptual decomposition of the observed data into the effects caused by the different experimental variables. If our data (the $y$s) were absolutely precise, we could set up a linear system of equations, one equation for each of the four possible experimental conditions represented by the $x$s, and solve for the $\beta$s.

However real data is usually affected by noise and we need replicates to estimate the levels of noise and assess the uncertainty of our estimated $\beta$s. Only then can we empirically assess whether any of the observed changes between conditions are significantly larger than those occurring due just to experimental or natural variation. We extend the equation to:

$$
y_j = x_{j0} \beta_0 + x_{j1} \beta_1 + x_{j2} \beta_2 + x_{j1} x_{j2} \beta_{12} + \epsilon_j
$$

We have included the index $j$ and a new term $\epsilon_j$. The index $j$ now explicitly counts over our individual replicate experiments; for example, if for each of the four conditions we perform three replicates, then $j$ counts from 1 to 12. The design matrix now has 12 rows, and $x_{jk}$ is the value of the matrix in its $j$th row and $k$th column.

The additional terms $\epsilon_j$, which we call the **residuals**, are there to absorb differences between replicates. However, one additional modeling component is needed: the system of 12 equations would be underdetermined without further information, since it now has more variables (12 $\epsilon$s and four $\beta$s) than it has equations (12, one for each $j$). To address this, we require that the $\epsilon_j$ be small. One popular way to overcome this is to minimise the sum of squared residuals,

$$
\sum_j \epsilon^2_j \rightarrow min.
$$

It turns out that with this requirement satisfied, the $\beta$s represent the _average_ effects of each of the experimental factors, while the residuals $\epsilon_j$ reflect the experimental fluntuations around the mean between the replicates. This approach, which is called **least sum of squares fitting**, is mathematically convenient, since it can be achieved by straight-forward matrix algebra and it is what the `lm()` function does.

## GSE40419

Processed data available at <https://zenodo.org/records/13970886>.

```{r gene_counts}
my_url <- 'https://zenodo.org/records/13970886/files/rsem.merged.gene_counts.tsv?download=1'
my_file <- 'rsem.merged.gene_counts.tsv'

if(file.exists(my_file) == FALSE){
  download.file(url = my_url, destfile = my_file)
}

gene_counts <- read_tsv("rsem.merged.gene_counts.tsv", show_col_types = FALSE)
head(gene_counts)
```

Metadata.

```{r metadata}
tibble::tribble(
  ~sample, ~run_id, ~group,
  "C2_norm", "ERR160122", "normal",
  "C3_norm", "ERR160123", "normal",
  "C5_norm", "ERR160124", "normal",
  "C1_norm", "ERR164473", "normal",
  "C1_cancer", "ERR164550", "cancer",
  "C2_cancer", "ERR164551", "cancer",
  "C3_cancer", "ERR164552", "cancer",
  "C5_cancer", "ERR164554", "cancer"
) -> my_metadata

my_metadata$group <- factor(my_metadata$group, levels = c('normal', 'cancer'))
my_metadata
```

Prepare DESeqDataSet.

```{r lung_cancer}
gene_counts |>
  dplyr::select(starts_with("ERR")) |>
  mutate(across(everything(), as.integer)) |>
  as.matrix() -> gene_counts_mat

row.names(gene_counts_mat) <- gene_counts$gene_id

idx <- match(colnames(gene_counts_mat), my_metadata$run_id)
colnames(gene_counts_mat) <- my_metadata$sample[idx]

lung_cancer <- DESeqDataSetFromMatrix(
  countData = gene_counts_mat,
  colData   = my_metadata[idx, ],
  design    = ~ group
)

lung_cancer
```

Run `DESeq`.

```{r lung_cancer_deseq}
lung_cancer <- DESeq(lung_cancer)
```

DE results.

```{r lung_cancer_res}
lung_cancer_res <- results(lung_cancer)
lung_cancer_res[order(lung_cancer_res$padj), ] |> head()
```
