---
title: "Normalisation methods implemented in edgeR"
date: "`r Sys.Date()`"
output:
  workflowr::wflow_html:
    toc: true
editor_options:
  chunk_output_type: console
---

```{r setup, include=FALSE}
library(tidyverse)
library(edgeR)
knitr::opts_chunk$set(echo = TRUE)
```

[edgeR](https://bioconductor.org/packages/release/bioc/html/edgeR.html) carries out:

> Differential expression analysis of RNA-seq expression profiles with biological replication. Implements a range of statistical methodology based on the negative binomial distributions, including empirical Bayes estimation, exact tests, generalized linear models and quasi-likelihood tests. As well as RNA-seq, it be applied to differential signal analysis of other types of genomic data that produce read counts, including ChIP-seq, ATAC-seq, Bisulfite-seq, SAGE and CAGE. 

## Installation

Install using `BiocManager::install()`.

```{r install_edger, eval=FALSE}
if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("edgeR")
```

## Normalize library sizes

The function `normLibSizes()` calculates scaling factors to convert the raw library sizes for a set of sequenced samples into normalized effective library sizes. Additional information on each method is provided in the help page for the function, i.e., `?edgeR::normLibSizes`:

> This function computes scaling factors to convert observed library sizes into normalized library sizes, also called "effective library sizes". The effective library sizes for use in downstream analysis are lib.size * norm.factors where lib.size contains the original library sizes and norm.factors is the vector of scaling factors computed by this function.
>
> The TMM method implements the trimmed mean of M-values method proposed by Robinson and Oshlack (2010). By default, the M-values are weighted according to inverse variances, as computed by the delta method for logarithms of binomial random variables. If refColumn is unspecified, then the column whose count-per-million upper quartile is closest to the mean upper quartile is set as the reference library.
>
> The TMMwsp method stands for "TMM with singleton pairing". This is a variant of TMM that is intended to have more stable performance when the counts have a high proportion of zeros. In the TMM method, genes that have zero count in either library are ignored when comparing pairs of libraries. In the TMMwsp method, the positive counts from such genes are reused to increase the number of features by which the libraries are compared. The singleton positive counts are paired up between the libraries in decreasing order of size and then a slightly modified TMM method is applied to the re-ordered libraries. If refColumn is unspecified, then the column with largest sum of square-root counts is used as the reference library.
>
> RLE is the scaling factor method proposed by Anders and Huber (2010). We call it "relative log expression", as median library is calculated from the geometric mean of all columns and the median ratio of each sample to the median library is taken as the scale factor.
>
> The upperquartile method is the upper-quartile normalization method of Bullard et al (2010), in which the scale factors are calculated from the 75% quantile of the counts for each library, after removing genes that are zero in all libraries. The idea is generalized here to allow normalization by any quantile of the count distributions.
>
> If method="none", then the normalization factors are set to 1.
>
> For symmetry, normalization factors are adjusted to multiply to 1. Rows of object that have zero counts for all columns are removed before normalization factors are computed. The number of such rows does not affect the estimated normalization factors. 

## A hypothetical situation

We will recreate a hypothetical situation described in [Robinson and Oshlack Genome Biology 2010](https://genomebiology.biomedcentral.com/articles/10.1186/gb-2010-11-3-r25) to test the different normalisation methods.

There are four samples; columns one and two are the controls and columns three and four are the cases. The total number of transcripts across all the samples is 50. The first 25 transcripts are in all four samples in equal counts. The remaining 25 transcripts are only present in the controls and their counts are the same as the first 25 transcripts.

The four samples have exactly the same depth (500 counts). However, since the case samples have half the number of transcripts than the controls (25 versus 50), the first 25 transcripts are sequenced at twice the depth.

```{r create_df}
control_1 <- rep(10, 50)
control_2 <- rep(10, 50)
case_1 <- c(rep(20, 25),rep(0,25))
case_2 <- c(rep(20, 25),rep(0,25))

df <- data.frame(
  cont1 = control_1,
  cont2 = control_2,
  case1 = case_1,
  case2 = case_2
)

head(df)
```

Tail of the dataset.

```{r tail_df}
tail(df)
```

Equal depth.

```{r depth_df}
colSums(df)
```

## No normalisation

In order to use {edgeR} we need to create a `DGEList` object with the groups.

```{r dgelist_no_norm}
group <- rep(c('control', 'case'), each = 2)
d <- DGEList(counts=df, group=group)
d
```

Let's run the differential expression analysis without any normalisation step.

```{r no_norm_de}
d <- estimateCommonDisp(d)

# perform the DE test
de <- exactTest(d)

# how many differentially expressed transcripts?
table(p.adjust(de$table$PValue, method="BH")<0.05)
```

Without normalisation, every transcript is differentially expressed.

## TMM normalisation

From the [edgeR manual](https://bioconductor.org/packages/release/bioc/vignettes/edgeR/inst/doc/edgeRUsersGuide.pdf):

> {edgeR} is concerned with differential expression analysis rather than with the quantification of expression levels. It is concerned with relative changes in expression levels between conditions, but not directly with estimating absolute expression levels.
>
> The `normLibSizes()` function normalises the library sizes in such a way to minimise the log-fold changes between the samples for most genes. The default method for computing these scale factors uses a trimmed mean of M-values (TMM) between each pair of samples. We call the product of the original library size and the scaling factor the effective library size, i.e., the normalised library size. The effective library size replaces the original library size in all downstream analyses

Let's test the weighted trimmed mean of M-values method:

```{r d_tmm}
d_tmm <- normLibSizes(d, method="TMM")
d_tmm
```

The effective library size.

```{r effective_lib_size}
effective_lib_size <- d_tmm$samples$lib.size * d_tmm$samples$norm.factors
effective_lib_size
```

Counts before scaling.

```{r head_d_tmm}
head(d$counts)
```

Scaling with the effective library size eliminates the differences in the first 25 transcripts.

```{r scale_with_effective_lib_size}
t(t(d$counts) / effective_lib_size) |>
  head()
```

Perform the differential expression test.

```{r d_tmm_de}
d_tmm <- estimateCommonDisp(d_tmm)
d_tmm <- exactTest(d_tmm)
table(p.adjust(d_tmm$table$PValue, method="BH")<0.05)
```

Only half of the transcripts are differentially expressed (the last 25 transcripts), which is the correct conclusion.

### The need for normalisation

From the Sampling framework section of Robinson and Oshlack.

A more formal explanation for the requirement of normalisation uses the following framework. Define $Y_{gk}$ as the observed count for gene $g$ in library $k$ summarised from the raw reads, $\mu_{gk}$ as the true and unknown expression level (number of transcripts), $L_g$ as the length of gene $g$ and $N_k$ as total number of reads for library $k$. We can model the expected value of $Y_{gk}$ as:

$$
E[Y_{gk}] = \frac{\mu_{gk} \times L_g}{S_k} \times N_k \\
\\
where\ S_k = \sum^G_{g=1} \mu_{gk} \times L_g
$$

$S_k$ represents the total RNA output of a sample. The problem underlying the analysis of RNA-seq data is that while $N_k$ is known, $S_k$ is unknown and can vary drastically from sample to sample, depending on the RNA composition. If a population has a larger total RNA output, then RNA-seq experiments will under-sample many genes, relative to another sample.

At this stage, we leave the variance in the above model for $Y_{gk}$ unspecified. Depending on the experimental situation, Poisson seems appropriate for technical replicates and Negative Binomial may be appropriate for the additional variation observed from biological replicates. It is also worth noting that, in practice, the $L_g$ is generally absorbed into the $\mu_{gk}$ parameter and does not get used in the inference procedure. However, it has been well established that gene length biases are prominent in the analysis of gene expression.

### The trimmed mean of M-values normalisation method

The total RNA production, $S_k$, cannot be estimated directly, since we do not know the expression levels and true lengths of every gene. However, the relative RNA production of two samples, $f_k = S_k /S_{k'}$, essentially a global fold change, can more easily be determined. We propose an empirical strategy that equates the overall expression levels of genes between samples under the assumption that the majority of them are not DE. One simple yet robust way to estimate the ratio of RNA production uses a weighted trimmed mean of the log expression ratios (trimmed mean of M values (TMM)). For sequencing data, we define the gene-wise log-fold-changes as:

$$
M_g = log_2\frac{y_{gk}/N_k}{y_{gk'}/N_{k'}}
$$

and absolute expression levels:

$$
A_g = \frac{1}{2}log_2(Y_{gk}/N_k \times Y_{gk'}/N_{k'})\ for\ Y_{gk'} \ne 0
$$

To robustly summarise the observed M values, we trim both the M values and the A values before taking the weighted average. Precision (inverse of the variance) weights are used to account for the fact that log fold changes (effectively, a log relative risk) from genes with larger read counts have lower variance on the logarithm scale.

For a two-sample comparison, only one relative scaling factor ($f_k$) is required. It can be used to adjust both library sizes (divide the reference by $\sqrt{f_k}$ and multiply non-reference by $\sqrt{f_k}$) in the statistical analysis (for example, Fisher's exact test).

Normalisation factors across several samples can be calculated by selecting one sample as a reference and calculating the TMM factor for each non-reference sample. Similar to two-sample comparisons, the TMM normalization factors can be built into the statistical model used to test for DE. For example, a Poisson model would modify the observed library size to an effective library size, which adjusts the modeled mean (for example, using an additional offset in a generalized linear model).

## RLE normalisation

Let's test the relative log expression normalisation method:

```{r d_rle}
d_rle <- normLibSizes(d, method="RLE")
d_rle
```

Perform the differential gene expression analysis.

```{r d_rle_de}
d_rle <- estimateCommonDisp(d_rle)
d_rle <- exactTest(d_rle)
table(p.adjust(d_rle$table$PValue, method="BH")<0.05)
```

## Upper-quartile normalisation

Lastly let's try the upper-quartile normalisation method:

```{r d_uq}
d_uq <- normLibSizes(d, method="upperquartile")
d_uq
```

DE test.

```{r d_uq_de}
d_uq <- estimateCommonDisp(d_uq)
d_uq <- exactTest(d_uq)
table(p.adjust(d_uq$table$PValue, method="BH")<0.05)
```

## Testing a real dataset

Perform differential gene expression analysis using various normalisation methods on the [pnas_expression.txt](https://pubmed.ncbi.nlm.nih.gov/19088194/) dataset.

```{r data}
my_url <- "https://davetang.org/file/pnas_expression.txt"
data <- read.table(my_url, header=TRUE, sep="\t")

dim(data)
```

Prepare a DGEList object.

```{r data_d}
d <- data[,2:8]
rownames(d) <- data[,1]
group <- c(rep("Control",4),rep("DHT",3))
d <- DGEList(counts = d, group=group)
```

Filter out lowly expressed genes.

```{r keep}
keep <- rowSums(cpm(d) > 0.5) >= 2
d <- d[keep, , keep.lib.sizes=FALSE]
```

Function to run edgeR.

```{r run_edger}
run_edger <- function(d, method = NULL){
  if(!is.null(method)){
    d <- normLibSizes(d, method = method)
  }
  design <- model.matrix(~d$samples$group)
  d <- estimateDisp(d, design)
  # use Quasi-likelihood F-tests
  fit <- glmQLFit(d, design)
  qlf <- glmQLFTest(fit)
}
```

Carry out differential gene expression analysis with no normalisation.

```{r run_edger_no_norm}
no_norm <- run_edger(d)
table(p.adjust(no_norm$table$PValue, method="fdr")<0.05)
```

With TMM normalisation.

```{r run_edger_tmm}
TMM <- run_edger(d, method="TMM")
table(p.adjust(TMM$table$PValue, method="fdr")<0.05)
```

With RLE.

```{r run_edger_rle}
RLE <- run_edger(d, method="RLE")
table(p.adjust(RLE$table$PValue, method="fdr")<0.05)
```

With the upper quartile method.

```{r run_edger_uq}
uq <- run_edger(d, method="upperquartile")
table(p.adjust(uq$table$PValue, method="fdr")<0.05)
```

Finding the overlaps between the differential gene expression analyses.

```{r run_edger_overlap}
library(gplots)

get_de <- function(x, pvalue){
  my_i <- p.adjust(x$PValue, method="fdr") < pvalue
  row.names(x)[my_i]
}

my_de_no_norm <- get_de(no_norm$table, 0.05)
my_de_tmm <- get_de(TMM$table, 0.05)
my_de_rle <- get_de(RLE$table, 0.05)
my_de_uq <- get_de(uq$table, 0.05)

gplots::venn(list(no_norm = my_de_no_norm, TMM = my_de_tmm, RLE = my_de_rle, UQ = my_de_uq))
```

There is a large overlap of differentially expressed genes in the different normalisation approaches.

```{r run_edger_norm}
gplots::venn(list(TMM = my_de_tmm, RLE = my_de_rle, UQ = my_de_uq))
```

## Summary

The normalisation factors were quite similar between all normalisation methods, which is why the results of the differential expression were quite concordant. Most methods down sized the DHT samples with a normalisation factor of less than one to account for the larger library sizes of these samples.
