---
title: "4.10 Single cell RNA-seq differential expression with pseudobulking"
date: "`r Sys.Date()`"
output:
  workflowr::wflow_html:
    toc: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
# https://stackoverflow.com/questions/30237310/setting-work-directory-in-knitr-using-opts-chunksetroot-dir-doesnt-wor
knitr::opts_knit$set(root.dir = rprojroot::find_rstudio_root_file())
```

```{r load_libraries, include=FALSE}
suppressPackageStartupMessages({
  library(tidyverse)
  library(edgeR)
  library(Seurat)
  library(pheatmap)
})

options(mc.cores = 8)
```

```{r start_time, include=FALSE}
start_time <- Sys.time()
```

## Data

The single cell RNA-seq data used in this notebook is from the human breast single cell RNA atlas generated by Pal et al. The preprocessing of the data and the complete bioinformatics analyses of the entire atlas study are described in detail in Chen et al. Most of the single cell analysis, such as dimensionality reduction and integration, were performed using Seurat. All the generated Seurat objects are publicly available on Figshare.

The Seurat object used in this notebook was downloaded directly from the website of the edgeR maintainers. This object contains breast tissue micro-environment samples from 13 individual healthy donors. This object has been subsetted to contain 10,000 cells of the total 24,751 cells from the original object.

```{r so}
so <- readRDS("data/SeuratObj.rds")
so
```

Distribution of cell counts across 13 healthy donors and 7 clusters; note that some samples don't have cells belonging to a certain cluster.

```{r group_versus_cluster}
table(so@meta.data$group, so@meta.data$seurat_clusters)
```

## Pseudobulking

Pseudo-bulk samples are created by aggregating read counts together for all the cells with the same combination of human donor and cluster. Here, we generate pseudo-bulk expression profiles from the Seurat object using the `Seurat2PB()` function. The human donor and cell cluster information of the integrated single cell data is stored in the `group` and `seurat_clusters` columns of the `meta.data` component of the Seurat object.

```{r y}
y <- Seurat2PB(so, sample="group", cluster="seurat_clusters")
dim(y$samples)
sum(table(so@meta.data$group, so@meta.data$seurat_clusters) > 0)
```

Counts are aggregated into samples + clusters; note that there aren't 13 * 7 samples because as we noted in the table, some combinations have 0 counts.

```{r y_counts}
colnames(y$counts)
```

Wide range of expression sums and note that the minimum is not 0.

```{r y_counts_sum}
summary(colSums(y$counts))
```

## Filtering and normalisation

Filter samples that have less than 50,000 total UMI.

```{r filter_samples}
keep.samples <- y$samples$lib.size > 5e4
y <- y[, keep.samples]

dim(y$samples)
```

Filter genes.

```{r filter_genes}
keep.genes <- filterByExpr(y, group=y$samples$cluster)
y <- y[keep.genes, , keep=FALSE]
```

TMM normalisation.

```{r tmm}
y <- normLibSizes(y)
```

## Design matrix

To perform differential expression analysis between cell clusters, we create a design matrix
using both cluster and donor information.

```{r design}
donor <- factor(y$samples$sample)
cluster <- as.factor(y$samples$cluster)
design <- model.matrix(~ cluster + donor)
colnames(design) <- gsub("donor", "", colnames(design))
colnames(design)[1] <- "Int"
dim(design)
```

There are 19 columns because 1 + 6 (clusters) + 12 (samples) = 19; the first sample + first cluster is the intercept. Each column represents one model parameter to be estimated.

There are 59 rows because each row represents a single sample + cluster combination.

```{r design_head}
head(design)
```

The first row (`cluster0` and `N_0019_total`) shows `Int`=1 and everything else as 0; this is the baseline/reference.

Next we will estimate the dispersion and fit a model using the design matrix structure.

```{r disp}
y <- estimateDisp(y, design, robust=TRUE)
fit <- glmQLFit(y, design, robust=TRUE)
```

To confirm the identities of cell clusters, we perform differential expression analysis to identify marker genes of each cluster. In particular, we compare each cluster with all the other clusters. Since there are 7 clusters in total, we construct a contrast matrix as follows so that each column of the contrast matrix represents a testing contrast for one cell cluster. Each contrast is a different linear combination of the same coefficients.

Each column of the contrast matrix represents one comparison: testing whether a specific cluster is different from the average of all other clusters.

The contrast is only testing cluster effects. The donor effects are being controlled for and they are not part of the comparison; all the donor rows are 0 meaning that we are not adding or subtracting donor effects in this contrast.

```{r contr}
ncls <- nlevels(cluster)
contr <- rbind( matrix(1/(1-ncls), ncls, ncls),
matrix(0, ncol(design)-ncls, ncls) )
diag(contr) <- 1
contr[1,] <- 0
rownames(contr) <- colnames(design)
colnames(contr) <- paste0("cluster", levels(cluster))
contr
```

We then perform quasi-likelihood F-test for each testing contrast. The results are stored as a list of DGELRT objects, one for each comparison.

```{r test_contrast}
qlf <- list()
for(i in 1:ncls){
  qlf[[i]] <- glmQLFTest(fit, contrast=contr[,i])
  qlf[[i]]$comparison <- paste0("cluster", levels(cluster)[i], "_vs_others")
}

length(qlf)
```

The top most significant DE genes of cluster 0 vs other clusters can be examined with topTags.

```{r}
topTags(qlf[[1]], n=10L)
```

The numbers of DE genes under each comparison are shown below

```{r}
dt <- lapply(lapply(qlf, decideTests), summary)
dt.all <- do.call("cbind", dt)
dt.all
```

```{r}
top <- 20
topMarkers <- list()
for(i in 1:ncls) {
  ord <- order(qlf[[i]]$table$PValue, decreasing=FALSE)
  up <- qlf[[i]]$table$logFC[ord] > 0
  topMarkers[[i]] <- rownames(y)[ord[up][1:top]]
}
topMarkers <- unique(unlist(topMarkers))
topMarkers
```

```{r heatmap}
lcpm <- edgeR::cpm(y, log=TRUE)
annot <- data.frame(cluster=paste0("cluster ", cluster))
rownames(annot) <- colnames(y)
ann_colors <- list(cluster=2:8)
names(ann_colors$cluster) <- paste0("cluster ", levels(cluster))
pheatmap::pheatmap(lcpm[topMarkers, ], breaks=seq(-2,2,length.out=101),
                   color=colorRampPalette(c("blue","white","red"))(100), scale="row",
                   cluster_cols=TRUE, border_color="NA", fontsize_row=5,
                   treeheight_row=70, treeheight_col=70, cutree_cols=7,
                   clustering_method="ward.D2", show_colnames=FALSE,
                   annotation_col=annot, annotation_colors=ann_colors)
```

## Session info

```{r}
#| label: session_info

sessionInfo()
```

Time taken to render notebook.

```{r}
#| label: notebook_runtime
end_time <- Sys.time()
end_time - start_time
```
