---
title: "Plot ARCHS4 gene expression data"
date: "`r Sys.Date()`"
output:
  workflowr::wflow_html:
    toc: false
editor_options:
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(pheatmap)
library(plotly)
```

## ARCHS4

Use [gget](https://github.com/pachterlab/gget) and the [archs4](https://pachterlab.github.io/gget/en/archs4.html) subtool to find the most correlated genes to a gene of interest or find the gene's tissue expression atlas using [ARCHS4](https://maayanlab.cloud/archs4/).

[Install](https://pachterlab.github.io/gget/en/installation.html) using pip:

```
pip install --upgrade gget
```

Usage.

```
gget archs4 -h
```
```
usage: gget archs4 [-h] [-e] [-w {correlation,tissue}] [-gc GENE_COUNT] [-s {human,mouse}] [-csv] [-o OUT] [-q] [-g GENE_DEPRECATED] [-j] gene

Find the most correlated genes or the tissue expression atlas of a gene using data from the human and mouse RNA-seq database ARCHS4 (https://maayanlab.cloud/archs4/).

positional arguments:
  gene                  Gene symbol or Ensembl gene ID of gene of interest, e.g. 'STAT4'.

options:
  -h, --help            show this help message and exit
  -e, --ensembl         Add this flag if gene is given as an Ensembl gene ID. (default: False)
  -w {correlation,tissue}, --which {correlation,tissue}
                        'correlation' (default) or 'tissue'.
                        - 'correlation' returns a gene correlation table that contains the 100 most correlated genes to the gene of interest. The Pearson correlation is calculated over all samples and tissues in ARCHS4.
                        - 'tissue' returns a tissue expression atlas calculated from human or mouse samples (as defined by 'species') in ARCHS4. (default: correlation)
  -gc GENE_COUNT, --gene_count GENE_COUNT
                        Number of correlated genes to return (default: 100).
                        (Only for gene correlation.) (default: 100)
  -s {human,mouse}, --species {human,mouse}
                        'human' (default) or 'mouse'. (Only for tissue expression atlas.) (default: human)
  -csv, --csv           Returns results in csv format instead of json. (default: True)
  -o OUT, --out OUT     Path to the csv file the results will be saved in, e.g. path/to/directory/results.csv.
                        Default: Standard out.
  -q, --quiet           Does not print progress information. (default: True)
  -g GENE_DEPRECATED, --gene GENE_DEPRECATED
                        DEPRECATED - use positional argument instead. Gene symbol or Ensembl gene ID of gene of interest (str), e.g. 'STAT4'.
  -j, --json            DEPRECATED - json is now the default output format (convert to csv using flag [--csv]). (default: False)
```

Get tissue expression for TNF in CSV.

```
gget archs4 --which tissue --csv --out TNF_tissue.csv TNF
```

## TNF tissue expression

Load data.

```{r tnf_exp}
tnf_exp <- readr::read_csv(file = "data/archs4/TNF_tissue.csv", show_col_types = FALSE)
head(tnf_exp)
```

Split the ID into multiple columns.

```{r split_id}
tnf_exp |>
  tidyr::separate(col = id, sep = "\\.", into = c("junk", "system", "organ", "tissue")) |>
  dplyr::mutate(tissue = stringr::str_to_title(tissue)) |>
  dplyr::arrange(system) |>
  dplyr::mutate(tissue = factor(tissue, levels = tissue)) -> tnf_exp

head(tnf_exp)
```

Plot tissue expression as box plots.

```{r tnf_box_plot}
library(ggplot2)
ggplot(
  tnf_exp,
  aes(
    x = tissue,
    ymin = min,
    lower = q1,
    middle = median,
    upper = q3,
    ymax = max,
    fill = system
  )
) +
  geom_boxplot(stat = 'identity') +
  theme_minimal() +
  theme(
    legend.position = 'right',
    axis.title.x = element_blank(),
    axis.text.x = element_text(angle = 70, hjust = 1)
  ) +
  labs(y = "Expression (TPM)")
```

## Load data

Prepare data using base R.

```{r my_df}
lapply(
  list.files("data/archs4/cancer", pattern = ".csv$", full.names = TRUE),
  function(x){
    cbind(gene = sub("\\.\\w+$", "", basename(x)), read.csv(x))
  }
) |>
  do.call("rbind", args = _) -> my_df

# Split `id` column.
do.call("rbind", strsplit(x = my_df$id, split = "\\.")) |>
  as.data.frame() -> id_split

colnames(id_split) <- c('root', 'system', 'organ', 'tissue')

# Rename tissues.
cap_first <- function(x){
  s <- strsplit(x, "")[[1]][1]
  return(sub(s, toupper(s), x))
}

id_split$tissue <- tolower(id_split$tissue)
id_split$tissue <- sapply(id_split$tissue, cap_first)

my_df <- cbind(my_df, id_split)

# Order `my_df` by system.
my_df <- my_df[order(my_df$gene, my_df$system), ]
my_df$tissue <- factor(my_df$tissue, levels = unique(my_df$tissue))

head(my_df)
```

Back to wide format.

```{r pivot_wider}
my_df |>
  dplyr::select(gene, median, tissue) |>
  tidyr::pivot_wider(names_from = tissue, values_from = median) -> my_df_wide
```

Convert to matrix and plot.

```{r pheatmap, fig.width=12, fig.height=6}
my_mat <- as.matrix(my_df_wide[, -1])
row.names(my_mat) <- my_df_wide$gene

pheatmap(my_mat)
```

Create sample annotation.

```{r sample_anno}
my_order <- colnames(my_mat)

my_df |>
  dplyr::select(system, tissue) |>
  dplyr::distinct() |>
  dplyr::arrange(match(tissue, my_order)) |>
  dplyr::select(-tissue) -> sample_anno

row.names(sample_anno) <- my_order
head(sample_anno)
```

Heatmap with system annotation.

```{r pheatmap_with_anno, fig.width=12, fig.height=6}
pheatmap(my_mat, annotation_col = sample_anno)
```

Interactive heatmap.

```{r plotly_heatmap, fig.height=6, fig.width=12}
plot_ly(
  x=colnames(my_mat),
  y=rownames(my_mat),
  z = my_mat,
  colors = colorRamp(c("green", "red")),
  type = "heatmap"
)
```
