---
title: "PCA with DESeq2"
date: "`r Sys.Date()`"
output:
  workflowr::wflow_html:
    toc: true
editor_options:
  chunk_output_type: console
---

```{r setup, include=FALSE}
library(tidyverse)
library(DESeq2)
knitr::opts_chunk$set(echo = TRUE)
```

[DESeq2](https://bioconductor.org/packages/release/bioc/html/DESeq2.html) is used to:

> Estimate variance-mean dependence in count data from high-throughput sequencing assays and test for differential expression based on a model using the negative binomial distribution. 

## Installation

Install using `BiocManager::install()`.

```{r install_edger, eval=FALSE}
if (!require("BiocManager", quietly = TRUE))
  install.packages("BiocManager")

BiocManager::install("DESeq2")
```

Package version.

```{r package_version}
packageVersion("DESeq2")
```

## Count table

https://zenodo.org/records/13970886.

```{r gene_counts}
my_url <- 'https://zenodo.org/records/13970886/files/rsem.merged.gene_counts.tsv?download=1'
my_file <- 'rsem.merged.gene_counts.tsv'

if(file.exists(my_file) == FALSE){
  download.file(url = my_url, destfile = my_file)
}

gene_counts <- read_tsv("rsem.merged.gene_counts.tsv", show_col_types = FALSE)
head(gene_counts)
```

Metadata.

```{r metadata}
tibble::tribble(
  ~sample, ~run_id, ~group,
  "C2_norm", "ERR160122", "normal",
  "C3_norm", "ERR160123", "normal",
  "C5_norm", "ERR160124", "normal",
  "C1_norm", "ERR164473", "normal",
  "C1_cancer", "ERR164550", "cancer",
  "C2_cancer", "ERR164551", "cancer",
  "C3_cancer", "ERR164552", "cancer",
  "C5_cancer", "ERR164554", "cancer"
) -> my_metadata

my_metadata$group <- factor(my_metadata$group, levels = c('normal', 'cancer'))
```

Matrix.

```{r gene_counts_mat}
gene_counts |>
  dplyr::select(starts_with("ERR")) |>
  mutate(across(everything(), as.integer)) |>
  as.matrix() -> gene_counts_mat

row.names(gene_counts_mat) <- gene_counts$gene_id

idx <- match(colnames(gene_counts_mat), my_metadata$run_id)
colnames(gene_counts_mat) <- my_metadata$sample[idx]

tail(gene_counts_mat)
```

## PCA

Create `DESeqDataSet` object.

```{r deseq_data_set}
lung_cancer <- DESeqDataSetFromMatrix(
  countData = gene_counts_mat,
  colData   = my_metadata,
  design    = ~ group
)
lung_cancer
```

Quickly estimate dispersion trend and apply a variance stabilizing transformation.

```{r vst}
pas_vst <- vst(lung_cancer)
pas_vst
```

Plot PCA.

```{r plot_PCA, fig.width=8, fig.height=8}
plotPCA(pas_vst, intgroup = "group") +
  theme_minimal()
```

PCA data.

```{r pca_data}
pca_data <- plotPCA(pas_vst, intgroup = "group", returnData = TRUE)
pca_data
```

[Source code](https://github.com/thelovelab/DESeq2/blob/devel/R/plots.R).

```{r plot_pca_source}
plotPCA_copied = function(object, intgroup="condition", ntop=500, returnData=FALSE, pcsToUse=1:2){
  message(paste0("using ntop=",ntop," top features by variance"))
  
  # calculate the variance for each gene
  rv <- rowVars(assay(object))

  # select the ntop genes by variance
  select <- order(rv, decreasing=TRUE)[seq_len(min(ntop, length(rv)))]

  # perform a PCA on the data in assay(x) for the selected genes
  pca <- prcomp(t(assay(object)[select,]))

  # the contribution to the total variance for each component
  percentVar <- pca$sdev^2 / sum( pca$sdev^2 )

  if (!all(intgroup %in% names(colData(object)))) {
    stop("the argument 'intgroup' should specify columns of colData(dds)")
  }

  # add the intgroup factors together to create a new grouping factor
  group <- if (length(intgroup) > 1) {
    intgroup.df <- as.data.frame(colData(object)[, intgroup, drop=FALSE])
    factor(apply( intgroup.df, 1, paste, collapse=":"))
  } else {
    colData(object)[[intgroup]]
  }

  # assembly the data for the plot
  pcs <- paste0("PC", pcsToUse)
  d <- data.frame(V1=pca$x[,pcsToUse[1]],
                  V2=pca$x[,pcsToUse[2]],
                  group=group, name=colnames(object), colData(object))
  colnames(d)[1:2] <- pcs
  
  if (returnData) {
    attr(d, "percentVar") <- percentVar[pcsToUse]
    return(d)
  }

  ggplot(data=d, aes_string(x=pcs[1], y=pcs[2], color="group")) +
    geom_point(size=3) + 
    xlab(paste0(pcs[1],": ",round(percentVar[pcsToUse[1]] * 100),"% variance")) +
      ylab(paste0(pcs[2],": ",round(percentVar[pcsToUse[2]] * 100),"% variance")) +
        coord_fixed()
}
```

Plot using the copied function.

```{r plot_PCA_copied, fig.width=8, fig.height=8}
plotPCA_copied(pas_vst, intgroup = "group") +
  theme_minimal()
```

Perform PCA as per `plotPCA()`.

```{r plot_pca_analysis}
# calculate the variance for each gene
rv <- rowVars(assay(pas_vst))
head(rv)

# select the ntop genes by variance
ntop <- 500
topgenes <- order(rv, decreasing=TRUE)[seq_len(min(ntop, length(rv)))]
head(topgenes)

# perform a PCA on the data in assay(x) for the selected genes
pca <- prcomp(t(assay(pas_vst)[topgenes,]))
names(pca)
```

1. **`sdev`** (Standard Deviations)  
   - A numeric vector containing the standard deviations of the principal components.
   - These values correspond to the square roots of the eigenvalues of the covariance (or correlation) matrix.
   - They indicate the amount of variance explained by each principal component.

2. **`rotation`** (Loadings)
   - A matrix where each column represents a principal component.
   - The rows correspond to the original variables, and the values indicate how much each variable contributes to each principal component.
   - This is sometimes referred to as the eigenvectors or loadings matrix.

3. **`center`**
   - A numeric vector containing the means of the original variables if centering was performed.
   - If `center = TRUE` in `prcomp()`, each column of the input data was mean-centered before PCA.
   - If centering was not performed, this slot will be `NULL`.

4. **`scale`**
   - A numeric vector containing the scaling values applied to the original variables if scaling was performed.
   - If `scale = TRUE`, each column of the input data was scaled to have unit variance before PCA.
   - If scaling was not performed, this slot will be `NULL`.

5. **`x`** (Principal Component Scores)
   - A matrix containing the transformed data in the new principal component space.
   - The rows correspond to observations, and the columns correspond to the principal components.
   - This is obtained by projecting the original data onto the principal component directions.

```{r loadings}
loadings <- pca$rotation
loadings[1:10, 1:2]
```

The loadings matrix contains the coefficients that define how the original variables contribute to each principal component. Each element in the matrix tells you how much a given original variable contributes (positively or negatively) to a principal component.

When you square the loadings, you get the proportion of variance in a variable explained by each principal component. Squaring eliminates negative signs, so you lose the direction of influence. It is useful when you want to: (i) determine the relative importance of each variable in a principal component and (ii) see how much of the original variable's variance is captured by each principal component.

```{r contribution}
contribution <- loadings^2
contribution[1:10, 1:2]
```

Gene contributing most to PC1.

```{r gene_pc1}
sort(contribution[, 1], decreasing = TRUE) |> head()
```

Note that the order is similar, because `plotPCA()` already sorted the genes by their variance!

## Gene contribution

Generate normalised counts for visualisation.

```{r norm_counts}
lung_cancer <- estimateSizeFactors(lung_cancer)
norm_counts <- counts(lung_cancer, normalized=TRUE)
head(norm_counts)
```

Top 10 genes contributing to PC1.

```{r top_genes}
sort(contribution[, 1], decreasing = TRUE) |>
  head(10) |>
  names() -> top_genes

norm_counts[top_genes, ] |>
  as.data.frame() |>
  tibble::rownames_to_column(var = "ensembl_gene_id") |>
  tidyr::pivot_longer(-ensembl_gene_id) |>
  ggplot(aes(name, ensembl_gene_id, fill = value)) +
  geom_tile() +
  theme_minimal() +
  theme(axis.title = element_blank())
```
