---
title: "Home"
site: workflowr::wflow_site
output:
  workflowr::wflow_html:
    toc: true
---

I refer to [my blog](https://davetang.org/muse/) often to look up notes and code I have written. I will recreate my blog posts here, to make it easier to reuse and maintain my code. Some posts only exist here and not on my blog, so there is no `original` hyperlink back to my blog.

## Visualisation

* Making a heatmap in R with the pheatmap package - [original](https://davetang.org/muse/2018/05/15/making-a-heatmap-in-r-with-the-pheatmap-package/) / [updated](pheatmap.html)
* Visualising Google Trends results with R - [original](https://davetang.org/muse/2018/12/31/visualising-google-trends-results-with-r/) / [updated](google_trends.html)
* Complex UpSet plots - original / [updated](complex_upset.html)
* Gene expression heatmap using ARCHS4 data - original / [updated](gene_heatmap.html)
* Network graph using R Plotly - original / [updated](plotly_network.html)
* Making a heatmap in R with the ComplexHeatmap package - original / [updated](complex_heatmap.html)

## Bioinformatics

### General

* Read GTF file into R - [original](https://davetang.org/muse/2017/08/04/read-gtf-file-r/) / [updated](read_gtf.html)
* Position weight matrix - [original](https://davetang.org/muse/2013/10/01/position-weight-matrix/) / [updated](pwm.html)
* Quantile normalisation in R - [original](https://davetang.org/muse/2014/07/07/quantile-normalisation-in-r/) / [updated](qn.html)
* Design matrices for gene expression experiments - original / [updated](design.html)
* Cell Ontology - original / [updated](cell_ontology.html)
* Using the R `seqinr` package - [original](https://davetang.org/muse/2013/05/09/using-the-r-seqinr-package/) / [updated](seqinr.html)
* Tissue specificity - [original](https://davetang.org/muse/2013/08/28/tissue-specificity/) / [updated](tissue.html)
* Convert gene symbols across species - original / [updated](cross_species.html)
* BPCells - original / [updated](bpcells.html)
* Matrix Market - original / [updated](matrix_market.html)

### Bioconductor

* Learning to use biomaRt - [original](https://davetang.org/muse/2012/04/27/learning-to-use-biomart/) / [updated](biomart.html)
* Bioconductor annotation packages - [original](https://davetang.org/muse/2013/12/16/bioconductor-annotation-packages/) / [updated](annotation_dbi.html)
* Using the Bioconductor ensembldb package - original / [updated](ensembldb.html)
* Using the Bioconductor GenomicRanges package - [original](https://davetang.org/muse/2013/01/02/iranges-and-genomicranges/) / [updated](genomic_ranges.html)
* Using the cBioPortalData R package - original / [updated](cbioportal.html)
* Using the GenomicDataCommons package - [original](https://davetang.org/muse/2025/01/06/using-the-genomicdatacommons-package/) / [updated](gdc.html)
* Downloading molecular signatures from MSigDB in R - [original](https://davetang.org/muse/2025/01/06/downloading-molecular-signatures-from-msigdb-in-r/) / [updated](msigdb.html)
* Interfacing with the Sequence Read Archive in R - original / [updated](sradb.html)
* Access the AnnotationHub Web Service - original / [updated](annotation_hub.html)
* The SingleCellExperiment class - original / [updated](sce.html)
* Mouse gastrulation and organogeneis datasets - original / [updated](mouse_gastrulation_data.html)
* zellkonverter - original / [updated](zellkonverter.html)
* Use biomaRt with _Rattus norvegicus_ - original / [updated](biomart_rattus_norvegicus.html)
* IsoformSwitchAnalyzeR - original / [updated](isoform_switch_analyzer.html)
* Using biomaRt to get homologues - original / [updated](biomart_homologues.html)

### Functional genomics

* Differential gene expression analysis with DESeq2 - original / [updated](deseq2.html)
* Differential gene expression analysis using edgeR - original / [updated](edger_de.html)
* edgeR versus DESeq2 - original / [updated](edger_vs_deseq2.html)
* Normalisation methods implemented in edgeR - [original](https://davetang.org/muse/2011/01/24/normalisation-methods-for-dge-data/) / [updated](edger.html)
* Gene Ontology Enrichment Analysis - [original](https://davetang.org/muse/2010/11/10/gene-ontology-enrichment-analysis/) / [updated](go_enrichment.html)
* GSVA: gene set variation analysis - [original](https://davetang.org/muse/2025/01/06/gene-set-variation-analysis/) / [updated](gsva.html)
* Using the fast preranked gene set enrichment analysis (fgsea) package - [original](https://davetang.org/muse/2018/01/10/using-fast-preranked-gene-set-enrichment-analysis-fgsea-package/) / [updated](fgsea.html)
* Using fgsea with edgeR results - original / [updated](fgsea_edger.html)
* Using the GSEA GUI tool - original / [updated](gsea.html)
* Getting started with clusterProfiler - original / [updated](cluster_profiler.html)
* Nothobranchius furzeri - original / [updated](turquoise_killifish.html)
* PCA with DESeq2 - original / [updated](gene_exp_pca.html)
* Checking out the presto package - original / [updated](presto.html)
* Caenorhabditis elegans - original / [updated](celegans.html)
* Differential gene expression analysis using edgeR with batch effects - original / [updated](edger_batch.html)

### Single cell RNA-seq

* Getting started with Cell Ranger - [original](https://davetang.org/muse/2018/08/09/getting-started-with-cell-ranger/) / [updated](cell_ranger.html)
* Getting started with Seurat - [original](https://davetang.org/muse/2017/08/01/getting-started-seurat/) / [updated](seurat.html)
* Analysing 20k cells - original / [updated](pbmc20k.html)
* Creating a Seurat object - original / [updated](seurat_object.html)
* Finding Markers with Seurat - original / [updated](seurat_markers.html)
* Seurat's AddModuleScore function - original / [updated](add_module_score.html)
* Using BPCells with Seurat - original / [updated](seurat_bpcells.html)
* Seurat analysis using memory versus on disk - original / [updated](bpcells_vs_memory.html)
* Getting started with harmony - original / [updated](harmony.html)
* Pokedex for Cell Types - original / [updated](celldex.html)
* Reference-Based Single-Cell RNA-Seq Annotation - original / [updated](singler.html)
* Differential abundance testing with Milo - original / [updated](miloR.html)
* Getting started with CELLxGENE Discover Census - original / [updated](cellxgene_census.html)
* Trajectory Analysis - original / [updated](trajectory.html)
* Doublet prediction using {scDblFinder} - original / [updated](scdblfinder.html)
* pbmc3k - original / [updated](pbmc3k.html)
* Seurat v4 versus v5 - original / [updated](seurat_v4_vs_v5.html)
* Re-running Seurat - original / [updated](seurat_rerun.html)
* Saving Seurat objects - original / [updated](seurat_save.html)
* Is single-cell expression bimodal? - original / [updated](scrnaseq_bimodal.html)

## Machine learning

* Getting started with tidymodels - original / [updated](tidymodels.html)
* Sequence composition and random forests - [original](https://davetang.org/muse/2013/09/14/sequence-composition-random-forests/) / [updated](tss_predict.html)
* Performance of binary classifiers - original / [updated](bin_class.html)
* Ordering Objects using Seriation in R - original / [updated](seriation.html)
* R function for calculating confusion matrix rates - [original](https://davetang.org/muse/2023/03/09/r-function-for-calculating-confusion-matrix-rates/) / [updated](confusion_matrix_rates.html)

## Programming

* Object-Oriented Programming - original / [updated](oo.html)
* Backticks in R - [original](https://davetang.org/muse/2022/09/07/backticks-in-r/) / [updated](r_backticks.html)
* Split single column of annotations into multiple columns - original / [updated](split_column.html)
* Matrix to adjacency list in R - [original](https://davetang.org/muse/2017/03/16/matrix-to-adjacency-list-in-r/) / [updated](adj_list.html)
* Using dplyr to aggregate in R - [original](https://davetang.org/muse/2016/10/13/using-dplyr-aggregate-r/) / [updated](dplyr_agg.html)
* Parallel computation in R - [original](https://davetang.org/muse/2016/11/04/parallel-distance-calculation/) / [updated](parallel.html)
* Reading irregular data into R - [original](https://davetang.org/muse/2023/06/02/reading-irregular-data-into-r/) / [updated](r_parsing.html)
* The tidyr pivot_longer and pivot_wider functions using base R - [original](https://davetang.org/muse/2023/07/26/the-tidyr-pivot_longer-and-pivot_wider-functions-using-base-r/) / [updated](reshape.html)
* Reading a list of files into a single R data frame - [original](https://davetang.org/muse/2023/06/23/reading-list-of-files-into-single-r-data-frame/) / [updated](read_files.html)
* Benchmarking R expressions - original / [updated](bench.html)
* Non-invasive pretty printing of R code - original / [updated](styler.html)
* Exporting files out of R - original / [updated](exporting.html)
* S3 - original / [updated](s3.html)
* Tidyverse - original / [updated](tidyverse.html)
* Trace - original / [updated](trace.html)
* Monkey patch - original / [updated](monkey_patch.html)

## Maths/Stats

* Monte Carlo integration - [original](https://davetang.org/muse/2020/10/09/monte-carlo-integration/) / [updated](monte_carlo.html)
* Set notation - [original](https://davetang.org/muse/2013/08/04/set-notation/) / [updated](set_notation.html)
* Step by step Principal Component Analysis using R [original](https://davetang.org/muse/2012/02/01/step-by-step-principal-components-analysis-using-r/) / [updated](pca.html)
* Distance Matrix Computation - [original](https://davetang.org/muse/2013/08/15/distance-matrix-computation/) / [updated](distance.html)
* Solving problems with graphs - [original](https://davetang.org/muse/2016/10/16/solving-problems-graphs/) / [updated](graph.html)
* Markov chain - [original](https://davetang.org/muse/2014/01/23/markov-chain/) / [updated](markov_chain.html)
* Hidden Markov Models - original / [updated](hmm.html)
* Survival analysis using the survival and survminer R packages - original / [updated](survival.html)
* Creating a correlation matrix with R - [original](https://davetang.org/muse/2012/01/31/creating-a-correlation-matrix-with-r/) / [updated](cor_mat.html)
* Introduction to Canonical Correlation Analysis original / [updated](cca.html)
* Euclidean vs. Cosine Distance - original / [updated](cosine.html)
* Linear algebra basics - original / [updated](linear_algebra.html)
* The Rand index - [original](https://davetang.org/muse/2017/09/21/the-rand-index/) / [updated](rand_index.html)
* The Adjusted Rand index - [original](https://davetang.org/muse/2017/09/21/adjusted-rand-index/) / [updated](adjusted_rand_index.html)
* Rand Index versus the Adjusted Rand Index - [original](https://davetang.org/muse/2017/09/28/rand-index-versus-adjusted-rand-index/) / [updated](rand_index_vs_adjusted.html)
* The Poisson distribution - [original](https://davetang.org/muse/2012/02/14/the-poisson-distribution/) / [updated](poisson.html)
* Generative models - original / [updated](generative_models.html)
* Hierarchical clustering - [original](https://davetang.org/muse/2010/10/22/hierarchical-clustering/) / [updated](hclust.html)
* _k_-methods clustering - [original](https://davetang.org/muse/2012/12/23/k-means-clustering/) / [updated](kmethods.html)
* Benjamini-Hochberg correction - original / [updated](bh.html)
* Manual linear regression analysis using R - [original](https://davetang.org/muse/2012/02/10/manual-linear-regression-analysis-using-r/) / [updated](lm.html)
* Fitting Generalised Linear Models - original / [updated](glm.html)
* Fisher's Exact Test - original / [updated](fisher.html)
* ANOVA - original / [updated](anova.html)
* Linear Models in R - original / [updated](linear_models.html)
* Information theory - original / [updated](information.html)
* Fitting a distribution - original / [updated](fitting.html)
* Intraclass correlation coefficient - original / [updated](icc.html)
* Community Ecology Package - original / [updated](vegan.html)
* Inter-rater reliability - original / [updated](rater.html)
* Mixed-effects models - original / [updated](mixed_effects_models.html)
* Matthews Correlation Coefficient - original / [updated](mcc.html)
* Likelihood Ratio Tests - original / [updated](lrt.html)
* Chi-squared Tests - original / [updated](chi.html)
* Bartlett's Test - original / [updated](bartlett.html)
* Odds ratio - original / [updated](odds_ratio.html)
* Understanding R's Formula Syntax - original / [updated](formula.html)

## Misc

* Learning about antibodies - original / [updated](antibody.html)
* JSON versus YAML - original / [updated](json_vs_yaml.html)
* Working with HDF5 data in R - original / [updated](hdf5.html)
* Text mining using the tm package - original / [updated](tm.html)
* Downloading files using the curl package - original / [updated](curl.html)
* Getting started with the R arrow package - original / [updated](arrow.html)
* Interfacing with databases in R - original / [updated](dbi.html)
* Schwartz circumplex - original / [updated](schwartz_circumplex.html)
