---
title: "The Poisson distribution"
date: "`r Sys.Date()`"
output:
  workflowr::wflow_html:
    toc: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
```

## Introduction

A [Poisson distribution](https://en.wikipedia.org/wiki/Poisson_distribution) is the probability distribution that results from a Poisson experiment. A probability distribution assigns a probability to possible outcomes of a random experiment. A Poisson experiment has the following properties:

1. The outcomes of the experiment can be classified as either successes or failures.
2. The average number of successes that occurs in a specified region is known.
3. The probability that a success will occur is proportional to the size of the region.
4. The probability that a success will occur in an extremely small region is virtually zero.

A Poisson random variable is the number of successes that result from a Poisson experiment. Given the mean number of successes that occur in a specified region, we can compute the Poisson probability based on the following formula:

$$ P(x; \mu) = \frac{(e^{-\mu})(\mu^x)}{x!} $$

which is also written as:

$$ Pr(X = k) = e^{-\lambda} \frac{\lambda^k}{k!} \ \ k = 0, 1, 2, \dotsc $$

## Examples

### Homes sold

The average number of homes sold is 2 homes per day. What is the probability that exactly 3 homes will be sold tomorrow?

$$ P(3; 2) = \frac{(e^{-2}) (2^3)}{3!} $$

Calculating this manually in R:

```{r homes_sold}
e <- exp(1)
((e^-2)*(2^3))/factorial(3)
```

Using `dpois()`:

```{r dpois}
dpois(x = 3, lambda = 2)
```

### HIV mutations

When the probability of success $p$ is small and the number of trials $n$ is large, the binomial distribution $B(n,p)$ can be faithfully approximated by the Poisson distribution with rate parameter $\lambda = np$.

Question: What is the probability mass distribution of observing `0:12` mutations in a genome of $n = 10^4$ nucleotides, when the probability of mutation is $p = 5 \times 10^{-4}$ per nucleotide?

Note that, unlike the binomial distribution, the Poisson does not depend on two separate parameters $n$ and $p$, but only on their product $np$.

```{r hiv_mutation_poisson}
dpois(x = 0:12, lambda = 10^4 * 5 * 10^-4) |>
  barplot(names.arg = 0:12, col = 2)
```

### RNA-seq

The Poisson distribution can be used to estimate the [technical variance in high-throughput sequencing experiments](https://genome.cshlp.org/content/early/2008/06/11/gr.079558.108). My basic understanding is that the variance between technical replicates can be modelled using the Poisson distribution. Check out [Why Does Rna-Seq Read Count Fit Poisson Distribution?](https://www.biostars.org/p/84445/) on Biostars.

From Chris Miller:

> Picture a process whereby you take the genome and choose a location at random to produce a read. This is a Poisson process. If you plot the depth of sequence along this theoretical genome, it will be a poisson distribution.

Expanding on the idea above in [Modern Statistics for Modern Biology](https://web.stanford.edu/class/bios221/book/):

Consider a sequencing library that contains $n_1$ fragments corresponding to gene 1, $n_2$ fragments for gene 2, and so on, with a total library size of $n = n_1 + n_2 + \ldots$. This library is then sequenced and the identity of $r$ randomly sampled fragments are determined. To paint a better mental picture, the following are some typical numbers. The number of genes will be in the order of tens of thousands; the value of $n$ depends on the amount of cells that were used to prepare the library and typically this is in the order of billions or trillions; and the number of reads $r$ is usually in the tens of millions, which is much smaller than $n$. Sequencing is sampling from $n$.

From this we can conclude that the probability that a given read maps to the $i$ th gene is $p_i = n_i/n$ (ratio of a specific fragment to all fragments) and this is independent of the outcomes for all the other reads. So we can model the number of reads for gene $i$ by a Poisson distribution, where the _rate_ of the Poisson process is the product of $p_i$, the initial proportion of fragments for the $i$ th gene, times $r$, the number of reads sequenced; that is $\lambda_i = rp_i$.

In practice, we are usually not interested in modeling the read counts within a single library, but in comparing the counts between libraries. That is, we want to know whether any differences that we see between different biological conditions are larger than what we might expect even between biological replicates. Empirically, it turns out that replicate experiments vary more than the Poisson distribution predicts.

Intuitively, what happens is that $p_i$, and therefore also $\lambda_i$, varies even between biological replicates. To account for that variation, we need to add another layer of modeling on top and it turns out that the gamma-Poisson (a.k.a. negative binomial) distribution suits our modeling requirements. Instead of a single $\lambda$, which represents both mean and variance, this distribution has two parameters. In principle, these can be different for each gene and we can estimate them from the data.

## Calculating confidence intervals

Calculate the [confidence intervals](https://www.math.mcmaster.ca/peter/s743/poissonalpha.html) using R. Create data with 1,000,000 values that follow a Poisson distribution with lambda = 20.

```{r poisson_data}
set.seed(1984)
n <- 1000000
data <- rpois(n, 20)
```

Functions for calculating the lower and upper tails.

```{r poisson_tails}
poisson_lower_tail <- function(n) {
   qchisq(0.025, 2*n)/2
}
poisson_upper_tail <- function(n) {
   qchisq(0.975, 2*(n+1))/2
}
```

Lower limit for lambda = 20.

```{r lower_limit}
poisson_lower_tail(20)
```

Upper limit for lambda = 20.

```{r upper_limit}
poisson_upper_tail(20)
```

How many values in data are lower than the lower limit?

```{r check_lower}
table(data<poisson_lower_tail(20))
```

How many values in data are higher than the upper limit?

```{r check_upper}
table(data>poisson_upper_tail(20))
```

What percentage of values were outside of the 95% CI?

```{r check_outside}
(sum(data<poisson_lower_tail(20)) + sum(data>poisson_upper_tail(20))) * 100 / n
```

Plot.

```{r plot_pois_cis}
hist(data)
abline(v=poisson_lower_tail(20))
abline(v=poisson_upper_tail(20))
```

### Webtool

Using the [Poisson Confidence Interval Calculator](https://www.danielsoper.com/statcalc/calculator.aspx?id=86) and lambda = 20 returns:

* 99% confidence interval: 10.35327 - 34.66800
* 95% confidence interval: 12.21652 - 30.88838
* 90% confidence interval: 13.25465 - 29.06202

which matches our 95% CI values.

## Checking whether distribution is Poisson

To assess whether count data follows a Poisson distribution, check whether observed counts match expected counts under a Poisson model.

The **Chi-Square test** compares observed vs. expected frequencies in count data.

```{r chi_square_test}
# Simulated count data
counts <- c(0, 1, 2, 3, 4, 5)  # Unique count values
observed <- c(50, 80, 70, 40, 30, 10)  # Frequency of each count

# Estimate lambda (mean count per observation)
lambda_hat <- sum(counts * observed) / sum(observed)

# Compute expected counts using Poisson probabilities
expected <- dpois(counts, lambda_hat) * sum(observed)

# Perform Chi-Square test
chisq.test(observed, p = expected / sum(expected))
```
* A **p-value > 0.05** suggests no significant deviation from a Poisson distribution.
* A **p-value < 0.05** suggests the data does not follow a Poisson distribution.

The Chi-Square test assumes **sufficient counts per category** (expected values should be > 5). If expected values are too low, consider **collapsing categories**.

If data follows a Poisson model, the **deviance** from a Poisson regression should be roughly chi-square distributed.

```{r poisson_deviance}
# Example dataset
set.seed(1984)
data <- data.frame(counts = rpois(100, lambda = 2))

# Fit Poisson model
poisson_model <- glm(counts ~ 1, family = poisson, data = data)
summary(poisson_model)
```

Kolmogorov-Smirnov (KS) test for Poisson Fit.

```{r poisson_ks}
set.seed(1984)
observed_data <- rpois(100, lambda = 2)

ks.test(observed_data, "ppois", lambda = mean(observed_data))
```
