---
title: "Finding Markers with Seurat"
date: "`r Sys.Date()`"
output:
  workflowr::wflow_html:
    toc: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
# https://stackoverflow.com/questions/30237310/setting-work-directory-in-knitr-using-opts-chunksetroot-dir-doesnt-wor
knitr::opts_knit$set(root.dir = rprojroot::find_rstudio_root_file())
```

Use the Peripheral Blood Mononuclear Cells (PBMCs) 2,700 cells dataset to test finding markers with Seurat.

Install the following packages, if necessary.

```{r install_seurat, eval=FALSE}
install.packages("remotes")
remotes::install_github("immunogenomics/presto")
install.packages("Seurat")
install.packages("bench")
```

Load `Seurat` and `bench` for some benchmarking.

```{r load_seurat}
suppressPackageStartupMessages(library("Seurat"))
suppressPackageStartupMessages(library("bench"))
suppressPackageStartupMessages(library("presto"))
```

## Data

To follow the tutorial, you'll need the 10X data, which can be download from AWS.

```{bash, eval=FALSE}
mkdir -p data/pbmc3k && cd data/pbmc3k
wget -c https://s3-us-west-2.amazonaws.com/10x.files/samples/cell/pbmc3k/pbmc3k_filtered_gene_bc_matrices.tar.gz
tar -xzf pbmc3k_filtered_gene_bc_matrices.tar.gz
```

## Seurat object

Load 10x data into a matrix using `Read10X()`.

```{r read_10x}
pbmc.data <- Read10X(
  data.dir = "data/pbmc3k/filtered_gene_bc_matrices/hg19/"
)
```

Create the Seurat object using `CreateSeuratObject`; see `?SeuratObject` for more information on the class.

```{r create_seurat_object}
seurat_obj <- CreateSeuratObject(
  counts = pbmc.data,
  min.cells = 3,
  min.features = 200,
  project = "pbmc3k"
)

class(seurat_obj)
```

## Seurat workflow

Run the workflow as separate steps; they can be piped together but sometimes errors occur, so it is useful to split up the steps.

```{r seurat_workflow}
debug_flag <- FALSE
seurat_obj <- NormalizeData(seurat_obj, normalization.method = "LogNormalize", scale.factor = 1e4, verbose = debug_flag)
seurat_obj <- FindVariableFeatures(seurat_obj, selection.method = 'vst', nfeatures = 2000, verbose = debug_flag)
seurat_obj <- ScaleData(seurat_obj, verbose = debug_flag)
seurat_obj <- RunPCA(seurat_obj, verbose = debug_flag)
seurat_obj <- RunUMAP(seurat_obj, dims = 1:30, verbose = debug_flag)
seurat_obj <- FindNeighbors(seurat_obj, dims = 1:30, verbose = debug_flag)
seurat_obj <- FindClusters(seurat_obj, resolution = 0.5, verbose = debug_flag)

seurat_obj
```

## Find all markers

`FindAllMarkers()` will find markers (differentially expressed genes) for each of the identity classes in a dataset.

```{r ident_levels}
levels(Idents(seurat_obj))
```

Find all markers.

```{r seurat_markers}
all_markers <- FindAllMarkers(seurat_obj, verbose = debug_flag)
dim(all_markers)
```

## Find markers

`FindMarkers()` finds markers (differentially expressed genes) for identity classes. Things to note:

* Default is to use the `data` slot/layer; this contains normalised values (after running `NormalizeData()`)
* `ident.1` - Identity class to define markers for; pass an object of class `phylo` or `clustertree` to find markers for a node in a cluster tree; passing `clustertree` requires `BuildClusterTree()` to have been run
* `ident.2` - A second identity class for comparison; if NULL, use all other cells for comparison; if an object of class `phylo` or `clustertree` is passed to `ident.1`, must pass a node to find markers for
* `group.by` - Regroup cells into a different identity class prior to performing differential expression
* `subset.ident` - Subset a particular identity class prior to regrouping. Only relevant if group.by is set

`pbmc_small` dataset.

```{r pbmc_small}
pbmc_small
```

`pbmc_small` metadata.

```{r pbmc_small_metadata}
table(
  pbmc_small@meta.data$RNA_snn_res.1,
  pbmc_small@meta.data$groups
)
```

Take all cells in cluster 2, and find markers that separate cells in the 'g1' group (metadata variable 'group').

```{r find_markers_ex}
markers <- FindMarkers(pbmc_small, ident.1 = "g1", group.by = 'groups', subset.ident = "2")
head(markers)
```

Perform some sanity checks.

```{r plot_gene}
plot_gene <- function(gene){
  gene_exp <- pbmc_small[['RNA']]['data'][gene, ]

  pbmc_small@meta.data |>
    dplyr::filter(RNA_snn_res.1 == 2, groups == 'g1') |>
    row.names() -> g1_c2
  
  pbmc_small@meta.data |>
    dplyr::filter(RNA_snn_res.1 == 2, groups == 'g2') |>
    row.names() -> g2_c2
  
  g1 <- gene_exp[g1_c2]
  g2 <- gene_exp[g2_c2]
  rbind(
    data.frame(exp = g1, group = "g1"),
    data.frame(exp = g2, group = "g2")
  ) -> my_df
  
  boxplot(
    exp~group,
    data = my_df,
    main = gene
  )
}

head(markers, 3) |>
  row.names() -> genes_to_check

sapply(genes_to_check, plot_gene) -> dev_null
```

Find markers for cluster 0 in pbmc3k.

```{r cluster_0_markers}
cluster_0_markers <- FindMarkers(seurat_obj, ident.1 = "0")
dim(cluster_0_markers)
```

Cluster 0 markers from `FindAllMarkers()`.

```{r all_markers_cluster_0}
all_markers |>
  dplyr::filter(cluster == 0) |>
  dim()
```

The start of the results are the same.

```{r compare_head_of_results}
head(cluster_0_markers)

all_markers |>
  dplyr::filter(cluster == 0) |>
  dplyr::select(-cluster, -gene) |>
  head()
```

The tail of the results are the same too, except that in `FindAllMarkers()` results have been trimmed.

```{r compare_tail_of_results}
cluster_0_markers[3134:3139, ]

all_markers |>
  dplyr::filter(cluster == 0) |>
  dplyr::select(-cluster, -gene) |>
  tail()
```

Trimming seems to be from `p_val < 0.01`

```{r trim_mark}
cluster_0_markers[3139:3142, ]
```

Find markers in parallel to speed up `FindAllMarkers()`. Use `imap()` to get the name of each list (`.y`); `.x` is each element of the list.

```{r find_markers_parallel}
library(future)
library(future.apply)

clusters <- levels(Idents(seurat_obj))
plan(multisession, workers = 4)
markers <- future_lapply(
  clusters,
  function(x){
    FindMarkers(seurat_obj, ident.1 = x)
  },
  future.seed = TRUE
)

names(markers) <- clusters

purrr::map(
  markers,
  \(x) tibble::rownames_to_column(.data = x, var = "gene") |> tibble::remove_rownames()
) |>
  purrr::imap(~ dplyr::mutate(.x, cluster = .y)) |>
  purrr::list_rbind() |>
  dplyr::filter(p_val < 0.01) |>
  dplyr::mutate(cluster = factor(cluster, levels = clusters)) |>
  dplyr::select(p_val, avg_log2FC, pct.1, pct.2, p_val_adj, cluster, gene) -> all_markers_parallel

all.equal(
  all_markers_parallel,
  tibble::remove_rownames(all_markers)
)
```





