---
title: "Reference-Based Single-Cell RNA-Seq Annotation"
date: "`r Sys.Date()`"
output:
  workflowr::wflow_html:
    toc: true
editor_options:
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

> Performs unbiased cell type recognition from single-cell RNA sequencing data, by leveraging reference transcriptomic datasets of pure cell types to infer the cell of origin of each single cell independently. 

## Installation

Install [SingleR](https://www.bioconductor.org/packages/release/bioc/html/SingleR.html).

```{r install_packages, eval=FALSE}
if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("SingleR")
BiocManager::install("scRNAseq")
BiocManager::install("scuttle")
BiocManager::install("scran")
install.packages("viridis")
install.packages("pheatmap")
```

## Vignette

Following [Using SingleR to annotate single-cell RNA-seq data](https://www.bioconductor.org/packages/release/bioc/vignettes/SingleR/inst/doc/SingleR.html).

> SingleR is an automatic annotation method for single-cell RNA sequencing (scRNAseq) data (Aran et al. 2019). Given a reference dataset of samples (single-cell or bulk) with known labels, it labels new cells from a test dataset based on similarity to the reference. Thus, the burden of manually interpreting clusters and defining marker genes only has to be done once, for the reference dataset, and this biological knowledge can be propagated to new datasets in an automated manner.

> The easiest way to use SingleR is to annotate cells against built-in references. In particular, the celldex package provides access to several reference datasets (mostly derived from bulk RNA-seq or microarray data) through dedicated retrieval functions. Here, we will use the Human Primary Cell Atlas (Mabbott et al. 2013), represented as a SummarizedExperiment object containing a matrix of log-expression values with sample-level labels.

```{r hpca_se}
suppressPackageStartupMessages(library(celldex))
hpca.se <- HumanPrimaryCellAtlasData()
hpca.se
```

## Human Embryonic Stem Cells

Our test dataset consists of some human embryonic stem cells (La Manno et al. 2016) from the scRNAseq package. For the sake of speed, we will only label the first 100 cells from this dataset.

```{r hescs}
suppressPackageStartupMessages(library(scRNAseq))
hESCs <- LaMannoBrainData('human-es')
hESCs <- hESCs[,1:100]
```

We use our hpca.se reference to annotate each cell in hESCs via the SingleR() function. This identifies marker genes from the reference and uses them to compute assignment scores (based on the Spearman correlation across markers) for each cell in the test dataset against each label in the reference. The label with the highest score is the assigned to the test cell, possibly with further fine-tuning to resolve closely related labels.

```{r hesc_pred}
suppressPackageStartupMessages(library(SingleR))
pred.hesc <- SingleR(
  test = hESCs,
  ref = hpca.se,
  assay.type.test=1,
  labels = hpca.se$label.main
)
```

Each row of the output DataFrame contains prediction results for a single cell. Labels are shown before (labels) and after pruning (pruned.labels), along with the associated scores.

```{r glimpse_pred_hesc}
pred.hesc
```

SingleR is workflow/package agnostic. The above example uses `SummarizedExperiment` objects, but the same functions will accept any (log-)normalized expression matrix.

## Using single-cell references

Here, we will use two human pancreas datasets from the scRNAseq package. The aim is to use one pre-labelled dataset to annotate the other unlabelled dataset. First, we set up the Muraro et al. (2016) dataset to be our reference.

```{r scem}
suppressPackageStartupMessages(library(scuttle))
sceM <- MuraroPancreasData()

# One should normally do cell-based quality control at this point, but for
# brevity's sake, we will just remove the unlabelled libraries here.
sceM <- sceM[,!is.na(sceM$label)]

# SingleR() expects reference datasets to be normalized and log-transformed.
sceM <- logNormCounts(sceM)
sceM
```

Label tally.

```{r scem_label}
table(colData(sceM)$label)
```

We then set up our test dataset from Grun et al. (2016).

```{r sceg}
sceG <- GrunPancreasData()
sceG <- sceG[,colSums(counts(sceG)) > 0] # Remove libraries with no counts.
sceG <- logNormCounts(sceG) 
sceG
```

We then run SingleR() as described previously but with a marker detection mode that considers the variance of expression across cells. Here, we will use the Wilcoxon ranked sum test to identify the top markers for each pairwise comparison between labels. This is slower but more appropriate for single-cell data compared to the default marker detection algorithm (which may fail for low-coverage data where the median is frequently zero).

```{r grun_pred}
pred.grun <- SingleR(
  test=sceG,
  ref=sceM,
  labels=sceM$label,
  de.method="wilcox"
)

table(pred.grun$labels)
```

## Annotation diagnostics

`plotScoreHeatmap()` displays the scores for all cells across all reference labels, which allows users to inspect the confidence of the predicted labels across the dataset. Ideally, each cell (i.e., column of the heatmap) should have one score that is obviously larger than the rest, indicating that it is unambiguously assigned to a single label. A spread of similar scores for a given cell indicates that the assignment is uncertain, though this may be acceptable if the uncertainty is distributed across similar cell types that cannot be easily resolved.

```{r plot_score_heatmap}
plotScoreHeatmap(pred.grun)
```

Another diagnostic is based on the per-cell “deltas”, i.e., the difference between the score for the assigned label and the median across all labels for each cell. Low deltas indicate that the assignment is uncertain, which is especially relevant if the cell’s true label does not exist in the reference. We can inspect these deltas across cells for each label using the plotDeltaDistribution() function.

```{r plot_delta_dist}
plotDeltaDistribution(pred.grun, ncol = 3)
```

The `pruneScores()` function will remove potentially poor-quality or ambiguous assignments based on the deltas. The minimum threshold on the deltas is defined using an outlier-based approach that accounts for differences in the scale of the correlations in various contexts - see ?pruneScores for more details. SingleR() will also report the pruned scores automatically in the pruned.labels field where low-quality assignments are replaced with NA.

```{r}
summary(is.na(pred.grun$pruned.labels))
```

Finally, a simple yet effective diagnostic is to examine the expression of the marker genes for each label in the test dataset. We extract the identity of the markers from the metadata of the SingleR() results and use them in the plotMarkerHeatmap() function, as shown below for beta cell markers. If a cell in the test dataset is confidently assigned to a particular label, we would expect it to have strong expression of that label’s markers. At the very least, it should exhibit upregulation of those markers relative to cells assigned to other labels.

```{r plot_marker_heatmap}
plotMarkerHeatmap(pred.grun, sceG, label="beta")
```

## pbmc3k

pbmc3k.

```{r pbmc3k}
suppressPackageStartupMessages(library(Seurat))
pbmc.data <- Read10X(data.dir = "data/pbmc3k/filtered_gene_bc_matrices/hg19/")

pbmc3k <- CreateSeuratObject(
  counts = pbmc.data,
  min.cells = 3,
  min.features = 200,
  project = "pbmc3k"
)

pbmc3k <- NormalizeData(pbmc3k)
```

Use Monaco.

```{r monaco_immune}
suppressPackageStartupMessages(library(celldex))
monaco_immune <- fetchReference("monaco_immune", "2024-02-26")
monaco_immune
```

Annotate.

```{r pbmc3k_anno}
pbmc3k.anno <- SingleR(
  test=pbmc3k@assays$RNA$data,
  ref=monaco_immune,
  labels=colData(monaco_immune)$label.main
)

cbind(
  pbmc3k@meta.data,
  as.data.frame(pbmc3k.anno)
) -> pbmc3k@meta.data

head(pbmc3k@meta.data)
```

### Reference

Split in half.

```{r split_cells}
total <- nrow(pbmc3k@meta.data)
half <- floor(total / 2)
first_half <- row.names(pbmc3k@meta.data)[1:half]
second_half <- row.names(pbmc3k@meta.data)[(half+1):total]
```

Create reference.

```{r pbmc3k_ref}
sce_ref <- SingleCellExperiment(
  assays = list(counts = pbmc3k@assays$RNA$counts[, first_half])
)
colLabels(sce_ref) <- pbmc3k@meta.data[first_half, 'labels']
sce_ref <- logNormCounts(sce_ref)
sce_ref
```

Create query.

```{r pbmc3k_query}
sce_query <- SingleCellExperiment(
  assays = list(counts = pbmc3k@assays$RNA$counts[, second_half])
)
sce_query <- logNormCounts(sce_query)
sce_query
```

Annotate.

```{r sce_pred}
sce.pred <- SingleR(
  test=sce_query,
  ref=sce_ref,
  labels=sce_ref$label,
  de.method="wilcox"
)

table(sce_ref$label)
table(sce.pred$labels)
```

Self annotate.

```{r self_annotate}
pbmc3k_ref <- SingleCellExperiment(
  assays = list(counts = pbmc3k@assays$RNA$counts)
)
colLabels(pbmc3k_ref) <- pbmc3k@meta.data$labels
pbmc3k_ref <- logNormCounts(pbmc3k_ref)

pbmc3k_query <- SingleCellExperiment(
  assays = list(counts = pbmc3k@assays$RNA$counts)
)
pbmc3k_query <- logNormCounts(pbmc3k_query)

pbmc3k.self.pred <- SingleR(
  test=pbmc3k_query,
  ref=pbmc3k_ref,
  labels=pbmc3k_ref$label,
  de.method="wilcox"
)

stopifnot(all(row.names(pbmc3k.self.pred) == row.names(pbmc3k@meta.data)))
table(
  pbmc3k@meta.data$labels,
  pbmc3k.self.pred$labels
)

sum(pbmc3k@meta.data$labels == pbmc3k.self.pred$labels) / length(pbmc3k.self.pred$labels)
```

Self annotate score heatmap.

```{r self_annotate_heatmap}
plotScoreHeatmap(pbmc3k.self.pred, show.pruned = TRUE)
```
