---
title: "Interfacing with the Sequence Read Archive in R"
date: "`r Sys.Date()`"
output:
  workflowr::wflow_html:
    toc: true
editor_options:
  chunk_output_type: console
---

```{r setup, include=FALSE}
library(tidyverse)
library(SRAdb)
knitr::opts_chunk$set(echo = TRUE)
```

## SRAdb

The [SRAdb package](https://www.bioconductor.org/packages/release/bioc/html/SRAdb.html) is a compilation of metadata from NCBI SRA and tools. Specifically:

> The Sequence Read Archive (SRA) is the largest public repository of sequencing data from the next generation of sequencing platforms including Roche 454 GS System, Illumina Genome Analyzer, Applied Biosystems SOLiD System, Helicos Heliscope, and others. However, finding data of interest can be challenging using current tools.
>
> SRAdb is an attempt to make access to the metadata associated with submission, study, sample, experiment and run much more feasible. This is accomplished by parsing all the NCBI SRA metadata into a SQLite database that can be stored and queried locally. Fulltext search in the package make querying metadata very flexible and powerful.
>
> fastq and sra files can be downloaded for doing alignment locally. Beside ftp protocol, the SRAdb has funcitons supporting fastp protocol (ascp from Aspera Connect) for faster downloading large data files over long distance. The SQLite database is updated regularly as new data is added to SRA and can be downloaded at will for the most up-to-date metadata. 

## SQLite database

First download `SRAmetadb.sqlite.gz` and gunzip it; the function `getSRAdbFile()` can do this but I recommend downloading this file externally. The gzipped file is around 6.8G in size (2024-07-26) and when uncompressed is 138G in size.

```{r get_sra_db_file, eval=FALSE}
sqlfile <- getSRAdbFile(method = "wget")
```

Download database file externally using `wget`.

```
wget -c https://gbnci.cancer.gov/backup/SRAmetadb.sqlite.gz
```

Point to the location of downloaded and gunzipped file.

```{r sqlfile}
sqlfile <- '/data2/sradb/SRAmetadb.sqlite'
```

Create a connection for queries. The standard DBI functionality as implemented in RSQLite function dbConnect makes the connection to the database. The dbDisconnect function disconnects the connection.

```{r dbconnect}
sra_con <- dbConnect(SQLite(), sqlfile)
```

## SRA tables

Database tables.

```{r sra_tables}
sra_tables <- dbListTables(sra_con)
sra_tables
```

Study fields.

```{r dblistfields_study}
dbListFields(sra_con, "study")
```

Sample fields.

```{r dblistfields_sample}
dbListFields(sra_con, "sample")
```

Query.

```{r select_from_study}
rs <- dbGetQuery(sra_con, "select * from study limit 3")
rs[, 1:5]
```

Query matching specific study accession.

```{r select_where}
dbGetQuery(sra_con, "select * from study where study_accession == 'DRP000001'")
```

Match a list of study accessions.

```{r list_of_accessions}
rs <- dbGetQuery(sra_con, "select * from study where study_accession in ('DRP000001', 'DRP000003')")
rs[, 1:5]
```

## SRX to SRR

How do we get the SRR accession for [SRX510365](https://www.ncbi.nlm.nih.gov/sra/SRX510365[accn])? (Should be SRR1216135)

```{r lookup}
exps <- c("SRX510281", "SRX510280", "SRX510279", "SRX510278", "SRX510277", "SRX510276")

lookup <- sraConvert(exps, sra_con = sra_con)
lookup
```

Get sample information.

```{r get_sample_info}
purrr::map_df(lookup$sample, \(x){
  dbGetQuery(sra_con, paste0("select * from sample where sample_accession == '", x, "'"))
}) -> sample_info
```

The required sample information is stored in `sample_attribute` and that needs to be further formatted.

```{r separate_sample_attribute}
sample_info |>
  as_tibble() |>
  select(sample_accession, sample_attribute) |>
  tidyr::separate_longer_delim(cols = sample_attribute, delim = " || ") |>
  tidyr::separate_wider_delim(cols = sample_attribute, delim = ": ", names = c('key', 'value')) |>
  tidyr::pivot_wider(id_cols = sample_accession, names_from = key, values_from = value) -> sample_attribute
```

Join into one table.

```{r lookup_join_sample_attribute}
dplyr::inner_join(x = lookup, y = sample_attribute, by = c('sample' = "sample_accession"))
```

## End

Disconnect.

```{r dbdisconnect}
dbDisconnect(conn = sra_con)
```
