---
title: "Nothobranchius furzeri"
date: "`r Sys.Date()`"
output:
  workflowr::wflow_html:
    toc: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
# https://stackoverflow.com/questions/30237310/setting-work-directory-in-knitr-using-opts-chunksetroot-dir-doesnt-wor
knitr::opts_knit$set(root.dir = rprojroot::find_rstudio_root_file())
```

The [biomaRt package](https://bioconductor.org/packages/release/bioc/html/biomaRt.html) provides an interface to BioMart databases provided by Ensembl.

> biomaRt provides an interface to a growing collection of databases implementing the BioMart software suite. The package enables retrieval of large amounts of data in a uniform way without the need to know the underlying database schemas or write complex SQL queries. The most prominent examples of BioMart databases are maintain by Ensembl, which provides biomaRt users direct access to a diverse set of data and enables a wide range of powerful online queries from gene annotation to database mining. 

For more information, check out the [Accessing Ensembl annotation with biomaRt](https://bioconductor.org/packages/release/bioc/vignettes/biomaRt/inst/doc/accessing_ensembl.html) guide.

## Installation

To begin, install the {biomaRt} package.

```{r install_biomart, eval=FALSE}
if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("biomaRt")
```

## Package

Load package.

```{r load_biomart}
packageVersion("biomaRt")
suppressPackageStartupMessages(library(biomaRt))
```

## Getting started

Use mirro.

```{r mirror_mart}
ensembl <- useEnsembl(biomart = "ensembl", mirror = "asia")
ensembl
```

Connect to the selected BioMart database by using `useMart()`.

```{r ensembl}
avail_datasets <- listDatasets(ensembl)
head(avail_datasets)
```

Look for human datasets by searching the description column.

```{r furzeri}
idx <- grep(pattern = "furzeri", avail_datasets$dataset, ignore.case = TRUE)
avail_datasets[idx, ]
```

Connect to the selected BioMart database and turquoise killifish dataset.

```{r usemart_with_dataset}
ensembl <- useEnsembl(biomart = "ensembl", mirror = "asia", dataset=avail_datasets[idx, 'dataset'])
ensembl
```

Building a query, requires three things:

1. filters
2. attributes
3. values

Use `listFilters()` to show available filters.

```{r avail_filters}
avail_filters <- listFilters(ensembl)
head(avail_filters)
```

Use `listAttributes()` to show available attributes.

```{r avail_attributes}
avail_attributes <- listAttributes(ensembl)
avail_attributes |>
  head()
```

The `getBM()` function is the main query function in {biomaRt}; use it once you have identified your attributes of interest and filters to use.

```{r affy_to_entrez_gene}
gene_ids <- c("ENSNFUG00015000040", "ENSNFUG00015000042", "ENSNFUG00015000043", "ENSNFUG00015000127")

getBM(
  attributes=c('ensembl_gene_id', 'external_gene_name', 'description'),
  filters = 'ensembl_gene_id',
  values = gene_ids,
  mart = ensembl
) -> res

res
```

All gene IDs.

```{r gtf_file}
gtf_file <- "https://ftp.ensembl.org/pub/release-113/gtf/nothobranchius_furzeri/Nothobranchius_furzeri.Nfu_20140520.113.gtf.gz"
if(file.exists(basename(gtf_file)) == FALSE){
  download.file(url = gtf_file, destfile = basename(gtf_file))
}

gtf_cols <- c(
  "seqname",
  "source",
  "feature",
  "start",
  "end",
  "score",
  "strand",
  "frame",
  "attribute"
)

gtf <- readr::read_tsv(file = gtf_file, comment = "#", col_names = gtf_cols, show_col_types = FALSE)
gtf |>
  dplyr::filter(feature == "gene") |>
  dplyr::select(attribute) |>
  tidyr::separate_rows(attribute, sep = ";\\s*") |>
  dplyr::filter(grepl("gene_id", attribute)) |>
  tidyr::separate(attribute, c('key', 'value'), sep = "\\s") |>
  dplyr::pull(value) |>
  gsub(pattern ='"', replacement = "") -> all_gene_ids

length(all_gene_ids) == length(unique(all_gene_ids))
```

Get gene names and descriptions.

```{r all_gene_info}
getBM(
  attributes=c('ensembl_gene_id', 'external_gene_name', 'description'),
  filters = 'ensembl_gene_id',
  values = all_gene_ids,
  mart = ensembl
) -> all_gene_info

head(all_gene_info)
```

Save lookup table.

```{r save_all_gene_info}
readr::write_csv(x = all_gene_info, file = "data/nfurzeri_gene_info.csv.gz")
```

## Gene Ontology terms

Find GO attribute names.

```{r find_go}
grep("^go", avail_attributes$name, ignore.case=TRUE, value=TRUE)
```

Query.

```{r go_id}
getBM(
  attributes=c("ensembl_gene_id", "go_id"),
  filters="ensembl_gene_id",
  values = all_gene_ids,
  mart = ensembl
) -> all_gene_go_ids

head(all_gene_go_ids)
```

Save GO table.

```{r save_all_gene_go_ids}
readr::write_csv(x = all_gene_go_ids, file = "data/nfurzeri_gene_id_to_go_id.csv.gz")
```

## NCBI

NCBI [Genome assembly NfurGRZ-RIMD1](https://www.ncbi.nlm.nih.gov/datasets/genome/GCF_043380555.1/) reference.

Download GAF file `*_gene_ontology.gaf.gz`, which contains Gene Ontology (GO) annotation of the annotated genes in GO Annotation File [GAF format](https://geneontology.org/docs/go-annotation-file-gaf-format-2.2/).

```{r download_gaf}
gaf_url <- 'https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/043/380/555/GCF_043380555.1_NfurGRZ-RIMD1/GCF_043380555.1-RS_2024_12_gene_ontology.gaf.gz'
gaf_file <- paste0("data/", basename(gaf_url))

if(file.exists(gaf_file) == FALSE){
  download.file(url = gaf_url, destfile = gaf_file)
}
```

The annotation flat file format is comprised of 17 tab-delimited fields.

```{r read_gaf}
gaf_cols <- c("DB","GeneID","Symbol","Qualifier","GO_ID","Reference","Evidence_Code","With_From","Aspect","Gene_Name","Gene_Synonym","Type","Taxon","Date","Assigned_By","Annot_Ext","Gene_Product_Form_ID")

gaf <- readr::read_tsv(file = gaf_file, comment = '!', show_col_types = FALSE, col_names = gaf_cols)
head(gaf)
```

Qualifiers.

```{r table_of_qualifiers}
table(gaf$Qualifier)
```

Evidence code.

```{r table_of_evidence_codes}
table(gaf$Evidence_Code)
```

Lookup table; check whether lookup is unique.

```{r entrez_gene_id_to_go_id}
library(dplyr)
select(gaf, GeneID, GO_ID) |>
  group_by(GeneID, GO_ID) |>
  summarise(n = n(), .groups = "keep") |>
  arrange(-n) -> lookup_table

head(lookup_table)
```

Look at an example of a duplicated entry.

```{r duplicated_entry}
eg <- head(lookup_table, n = 1)

dplyr::filter(gaf, GeneID == eg$GeneID, GO_ID == eg$GO_ID)
```
